// @c8y/ngx-components/esm2015/context-dashboard/device-management-home-dashboard/device-management-home-dashboard.component.js
import { __awaiter } from "tslib";
import { Component } from '@angular/core';
import { gettext } from '@c8y/ngx-components';
import { WidgetService } from '../widget.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../widget.service';
import * as ɵngcc2 from '@c8y/ngx-components';

const _c0 = function () { return {}; };
const _c1 = function () { return { isDisabled: false, canDelete: false, canCopy: false, translateWidgetTitle: true, allowFullscreen: false }; };
export class DeviceManagementHomeDashboardComponent {
    constructor(widgetService) {
        this.widgetService = widgetService;
        this.widgets = [];
        this.DEFAULT_HOME_WIDGETS = [
            {
                name: 'Asset Alarms',
                title: gettext('Active, critical alarms'),
                _x: 0,
                _y: 0,
                _height: 4,
                _width: 6
            },
            {
                name: 'Map',
                title: gettext('Map'),
                _x: 6,
                _y: 0,
                _height: 8,
                _width: 6
            }
        ];
    }
    ngOnInit() {
        this.DEFAULT_HOME_WIDGETS.map((widget) => __awaiter(this, void 0, void 0, function* () {
            yield this.widgetService
                .mapLegacy(widget)
                .then(transformedWidget => this.widgets.push(transformedWidget));
        }));
    }
}
DeviceManagementHomeDashboardComponent.ɵfac = function DeviceManagementHomeDashboardComponent_Factory(t) { return new (t || DeviceManagementHomeDashboardComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.WidgetService)); };
DeviceManagementHomeDashboardComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: DeviceManagementHomeDashboardComponent, selectors: [["c8y-device-management-home-dashboard"]], decls: 4, vars: 8, consts: [[3, "contextDashboard", "widgets", "settings"]], template: function DeviceManagementHomeDashboardComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵpipe(2, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(3, "c8y-widgets-dashboard", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 4, "Home"), "\n");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("contextDashboard", ɵngcc0.ɵɵpureFunction0(6, _c0))("widgets", ctx.widgets)("settings", ɵngcc0.ɵɵpureFunction0(7, _c1));
    } }, directives: [ɵngcc2.TitleComponent, ɵngcc2.WidgetsDashboardComponent], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
DeviceManagementHomeDashboardComponent.ctorParameters = () => [
    { type: WidgetService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DeviceManagementHomeDashboardComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-device-management-home-dashboard',
                template: "<c8y-title>\n  {{ 'Home' | translate }}\n</c8y-title>\n<c8y-widgets-dashboard\n  [contextDashboard]=\"{}\"\n  [widgets]=\"widgets\"\n  [settings]=\"{\n    isDisabled: false,\n    canDelete: false,\n    canCopy: false,\n    translateWidgetTitle: true,\n    allowFullscreen: false\n  }\"\n></c8y-widgets-dashboard>\n"
            }]
    }], function () { return [{ type: ɵngcc1.WidgetService }]; }, null); })();
