// @c8y/ngx-components/esm2015/context-dashboard/device-management-home-dashboard/device-management-home-dashboard.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { CoreModule, gettext, HOOK_NAVIGATOR_NODES, HOOK_ONCE_ROUTE } from '@c8y/ngx-components';
import { ContextDashboardModule } from '../context-dashboard.module';
import { DeviceManagementHomeDashboardComponent } from './device-management-home-dashboard.component';
import * as ɵngcc0 from '@angular/core';
const ɵ0 = {
    path: '',
    component: DeviceManagementHomeDashboardComponent
}, ɵ1 = [{ label: gettext('Home'), icon: 'home', priority: 10000, path: '/' }];
export class DeviceManagementHomeDashboardModule {
}
DeviceManagementHomeDashboardModule.ɵfac = function DeviceManagementHomeDashboardModule_Factory(t) { return new (t || DeviceManagementHomeDashboardModule)(); };
DeviceManagementHomeDashboardModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DeviceManagementHomeDashboardModule });
DeviceManagementHomeDashboardModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        {
            provide: HOOK_ONCE_ROUTE,
            useValue: ɵ0,
            multi: true
        },
        {
            provide: HOOK_NAVIGATOR_NODES,
            useValue: ɵ1,
            multi: true
        }
    ], imports: [[CommonModule, CoreModule, ContextDashboardModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DeviceManagementHomeDashboardModule, [{
        type: NgModule,
        args: [{
                declarations: [DeviceManagementHomeDashboardComponent],
                imports: [CommonModule, CoreModule, ContextDashboardModule],
                exports: [],
                entryComponents: [DeviceManagementHomeDashboardComponent],
                providers: [
                    {
                        provide: HOOK_ONCE_ROUTE,
                        useValue: ɵ0,
                        multi: true
                    },
                    {
                        provide: HOOK_NAVIGATOR_NODES,
                        useValue: ɵ1,
                        multi: true
                    }
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DeviceManagementHomeDashboardModule, { declarations: function () { return [DeviceManagementHomeDashboardComponent]; }, imports: function () { return [CommonModule, CoreModule, ContextDashboardModule]; } }); })();
export { ɵ0, ɵ1 };
