// @c8y/ngx-components/esm2015/core/header/right-drawer/right-drawer.component.js
import { Component, Input } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '../../i18n/c8y-translate.directive';
import * as ɵngcc3 from '../../product-experience/product-experience.directive';
import * as ɵngcc4 from '../../common/icon.directive';
import * as ɵngcc5 from '../../i18n/c8y-translate.pipe';
import * as ɵngcc6 from '../../common/humanize-app-name.pipe';

function RightDrawerComponent_div_7_ng_container_4_h2_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "h2");
    ɵngcc0.ɵɵelement(1, "i");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const quickLink_r3 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵclassMap(quickLink_r3.icon);
} }
function RightDrawerComponent_div_7_ng_container_4_img_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "img", 9);
} if (rf & 2) {
    const quickLink_r3 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("src", quickLink_r3.iconSrc, ɵngcc0.ɵɵsanitizeUrl);
} }
function RightDrawerComponent_div_7_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "a", 7);
    ɵngcc0.ɵɵlistener("click", function RightDrawerComponent_div_7_ng_container_4_Template_a_click_1_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r9); const quickLink_r3 = restoredCtx.$implicit; quickLink_r3.click && $event.preventDefault(); return quickLink_r3.click && quickLink_r3.click(); });
    ɵngcc0.ɵɵtemplate(2, RightDrawerComponent_div_7_ng_container_4_h2_2_Template, 2, 3, "h2", 3);
    ɵngcc0.ɵɵtemplate(3, RightDrawerComponent_div_7_ng_container_4_img_3_Template, 1, 1, "img", 8);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const quickLink_r3 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("href", quickLink_r3.url || "#", ɵngcc0.ɵɵsanitizeUrl)("actionName", "quickLinkRightDrawer:" + quickLink_r3.label);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", quickLink_r3.icon);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", quickLink_r3.iconSrc);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 5, quickLink_r3.label), " ");
} }
function RightDrawerComponent_div_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "div", 4);
    ɵngcc0.ɵɵtext(2, "Quick links");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div", 5);
    ɵngcc0.ɵɵtemplate(4, RightDrawerComponent_div_7_ng_container_4_Template, 6, 7, "ng-container", 6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r0.quickLinks);
} }
function RightDrawerComponent_div_8_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 12);
    ɵngcc0.ɵɵelementStart(1, "a", 13);
    ɵngcc0.ɵɵelement(2, "i", 14);
    ɵngcc0.ɵɵelementStart(3, "span", 15);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const link_r11 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("href", link_r11.url, ɵngcc0.ɵɵsanitizeUrl)("actionName", "quickLinkRightDrawer:" + link_r11.label);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yIcon", link_r11.icon);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(4, 5, link_r11.label));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(6, 7, link_r11.label), " ");
} }
function RightDrawerComponent_div_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "div", 10);
    ɵngcc0.ɵɵtext(2, "Help & support");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(3, RightDrawerComponent_div_8_div_3_Template, 7, 9, "div", 11);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r1.helpAndSupport);
} }
export class RightDrawerComponent {
    constructor() {
        this.quickLinks = [];
        this.helpAndSupport = [];
    }
    set _quickLinks(data) {
        if (data) {
            this.quickLinks = data.filter(d => d.type === 'quicklink');
        }
    }
    set _helpAndSupport(data) {
        if (data) {
            this.helpAndSupport = data.filter(d => d.type === 'doc');
        }
    }
}
RightDrawerComponent.ɵfac = function RightDrawerComponent_Factory(t) { return new (t || RightDrawerComponent)(); };
RightDrawerComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: RightDrawerComponent, selectors: [["c8y-right-drawer"]], inputs: { _quickLinks: ["quickLinks", "_quickLinks"], _helpAndSupport: ["helpAndSupport", "_helpAndSupport"], app: "app" }, decls: 9, vars: 10, consts: [[1, "c8y-right-drawer"], [1, "summary"], [1, "text-light"], [4, "ngIf"], ["translate", "", 1, "legend", "form-block"], [1, "quick-links-drawer"], [4, "ngFor", "ngForOf"], ["c8yProductExperience", "", 1, "quick-link", 3, "href", "actionName", "actionData", "click"], ["style", "max-width:50px;", "class", "img-responsive center-block", "onerror", "this.style.display = 'none'", 3, "src", 4, "ngIf"], ["onerror", "this.style.display = 'none'", 1, "img-responsive", "center-block", 2, "max-width", "50px", 3, "src"], ["translate", "", 1, "legend", "form-block", "m-t-24"], ["class", "help-support", 4, "ngFor", "ngForOf"], [1, "help-support"], ["target", "_blank", "rel", "noopener noreferrer", "c8yProductExperience", "", 1, "help-and-support-link", "text-truncate", 3, "href", "actionName", "actionData"], [3, "c8yIcon"], [1, "d-inline", 3, "title"]], template: function RightDrawerComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "h4", 2);
        ɵngcc0.ɵɵtext(3);
        ɵngcc0.ɵɵpipe(4, "translate");
        ɵngcc0.ɵɵpipe(5, "async");
        ɵngcc0.ɵɵpipe(6, "humanizeAppName");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(7, RightDrawerComponent_div_7_Template, 5, 1, "div", 3);
        ɵngcc0.ɵɵtemplate(8, RightDrawerComponent_div_8_Template, 4, 1, "div", 3);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate2(" ", ɵngcc0.ɵɵpipeBind1(4, 4, "Welcome to"), " ", ɵngcc0.ɵɵpipeBind1(5, 6, ɵngcc0.ɵɵpipeBind1(6, 8, ctx.app)), " ");
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngIf", ctx.quickLinks.length);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.helpAndSupport.length);
    } }, directives: [ɵngcc1.NgIf, ɵngcc2.C8yTranslateDirective, ɵngcc1.NgForOf, ɵngcc3.ProductExperienceDirective, ɵngcc4.IconDirective], pipes: [ɵngcc5.C8yTranslatePipe, ɵngcc1.AsyncPipe, ɵngcc6.HumanizeAppNamePipe], encapsulation: 2 });
RightDrawerComponent.propDecorators = {
    _quickLinks: [{ type: Input, args: ['quickLinks',] }],
    _helpAndSupport: [{ type: Input, args: ['helpAndSupport',] }],
    app: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(RightDrawerComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-right-drawer',
                template: "<div class=\"c8y-right-drawer\">\n  <div class=\"summary\">\n    <h4 class=\"text-light\">\n      {{ 'Welcome to' | translate }} {{ app | humanizeAppName | async }}\n    </h4>\n  </div>\n\n  <div *ngIf=\"quickLinks.length\">\n    <div class=\"legend form-block\" translate>Quick links</div>\n    <div class=\"quick-links-drawer\">\n      <ng-container *ngFor=\"let quickLink of quickLinks\">\n        <a\n          class=\"quick-link\"\n          [href]=\"quickLink.url || '#'\"\n          (click)=\"quickLink.click && $event.preventDefault(); quickLink.click && quickLink.click();\"\n          c8yProductExperience\n          [actionName]=\"'quickLinkRightDrawer:' + quickLink.label\"\n          [actionData]=\"\"\n        >\n          <h2 *ngIf=\"quickLink.icon\"><i class=\"{{ quickLink.icon }}\"></i></h2>\n          <img\n            *ngIf=\"quickLink.iconSrc\"\n            [src]=\"quickLink.iconSrc\"\n            style=\"max-width:50px;\"\n            class=\"img-responsive center-block\"\n            onerror=\"this.style.display = 'none'\">\n          {{ quickLink.label | translate }}\n        </a>\n      </ng-container>\n    </div>\n  </div>\n\n  <div *ngIf=\"helpAndSupport.length\">\n    <div class=\"legend form-block m-t-24\" translate>Help &amp; support</div>\n    <div class=\"help-support\" *ngFor=\"let link of helpAndSupport\">\n      <a\n        class=\"help-and-support-link text-truncate\"\n        target=\"_blank\"\n        rel=\"noopener noreferrer\"\n        [href]=\"link.url\"\n        c8yProductExperience\n        [actionName]=\"'quickLinkRightDrawer:' + link.label\"\n        [actionData]=\"\"\n      >\n        <i [c8yIcon]=\"link.icon\"></i>\n        <span title=\"{{ link.label | translate }}\" class=\"d-inline\">\n          {{ link.label | translate }}\n        </span>\n      </a>\n    </div>\n  </div>\n</div>\n"
            }]
    }], function () { return []; }, { _quickLinks: [{
            type: Input,
            args: ['quickLinks']
        }], _helpAndSupport: [{
            type: Input,
            args: ['helpAndSupport']
        }], app: [{
            type: Input
        }] }); })();
