// @c8y/ngx-components/esm2015/operations/operations-list/operations-list.module.js
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { CoreModule, gettext, HOOK_ROUTE, ViewContext } from '@c8y/ngx-components';
import { StatusFilterModule } from '@c8y/ngx-components/operations/status-filter';
import { OperationsListItemDetailsModule } from '@c8y/ngx-components/operations/operations-list-item-details';
import { BulkOperationListItemService } from '@c8y/ngx-components/operations/bulk-operation-list-item';
import { OperationsSharedModule } from '@c8y/ngx-components/operations/shared';
import { OperationsListComponent } from './operations-list.component';
import { OperationsListItemComponent } from './operations-list-item.component';
import { OperationsListService } from './operations-list.service';
export const singleOperationListRoutes = [
    {
        path: 'devicecontrol/single',
        component: OperationsListComponent
    }
];
const ɵ0 = [
    {
        context: ViewContext.Device,
        path: 'control',
        component: OperationsListComponent,
        label: gettext('Control'),
        icon: 'c8y-icon c8y-icon-device-control',
        priority: 600
    }
];
/**
 * This module allows for displaying the list item details of single and bulk operations.
 */
export class OperationsListModule {
}
OperationsListModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CoreModule,
                    StatusFilterModule,
                    OperationsListItemDetailsModule,
                    OperationsSharedModule,
                    TooltipModule,
                    RouterModule.forChild(singleOperationListRoutes)
                ],
                declarations: [OperationsListComponent, OperationsListItemComponent],
                exports: [OperationsListComponent, OperationsListItemComponent],
                providers: [
                    OperationsListService,
                    BulkOperationListItemService,
                    {
                        provide: HOOK_ROUTE,
                        useValue: ɵ0,
                        multi: true
                    }
                ]
            },] }
];
export { ɵ0 };
