// @c8y/ngx-components/esm2015/operations/operations-list/operations-list-item.component.js
import { __awaiter } from "tslib";
import { Component, Input, ViewChild } from '@angular/core';
import { ActivatedRoute, Router } from '@angular/router';
import { OperationStatus } from '@c8y/client';
import { gettext } from '@c8y/ngx-components';
import { BulkOperationListItemService } from '@c8y/ngx-components/operations/bulk-operation-list-item';
import { OperationsService, OPERATION_STATUS_OPTIONS_MAP } from '@c8y/ngx-components/operations/shared';
import { OperationsListService } from './operations-list.service';
export class OperationsListItemComponent {
    constructor(bulkOperationListItemService, sharedService, operationsListService, route, router) {
        var _a, _b, _c, _d;
        this.bulkOperationListItemService = bulkOperationListItemService;
        this.sharedService = sharedService;
        this.operationsListService = operationsListService;
        this.route = route;
        this.router = router;
        this.collapsed = true;
        this.readOnly = false;
        this.noExpandToggle = false;
        this.refreshLoading = false;
        this.OPERATION_STATUS = OperationStatus;
        this.OPERATION_STATUS_OPTIONS_MAP = OPERATION_STATUS_OPTIONS_MAP;
        this.deviceId = ((_b = (_a = this.route.snapshot.parent) === null || _a === void 0 ? void 0 : _a.data.contextData) === null || _b === void 0 ? void 0 : _b.id) ||
            ((_d = (_c = this.route.snapshot.firstChild) === null || _c === void 0 ? void 0 : _c.data.contextData) === null || _d === void 0 ? void 0 : _d.id);
        this.isSmartRulesMicroserviceAvailable = false;
    }
    ngOnInit() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            this.isSmartRulesMicroserviceAvailable =
                yield this.operationsListService.isSmartRulesMicroserviceAvailable();
            this.operationTitle =
                ((_a = this.operation.c8y_Command) === null || _a === void 0 ? void 0 : _a.text) ||
                    this.operation.description ||
                    gettext('No command text or description available');
            const smartRulesCreatePermissions = this.operationsListService.getSmartRulesCreatePermissions();
            if (smartRulesCreatePermissions) {
                this.inventoryPermissions = smartRulesCreatePermissions[0].anyRole;
                this.smartRuleAdminPermissions = smartRulesCreatePermissions[1].anyRole;
            }
        });
    }
    createCustomizedBulkOperation($event) {
        if ($event) {
            $event.stopPropagation();
        }
        this.router.navigateByUrl(`/devicecontrol/single/create-bulk/${this.operation.id}`);
    }
}
OperationsListItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'c8y-operations-list-item',
                template: "<c8y-li\n  class=\"c8y-list__item--double-actions\"\n  [ngClass]=\"{ 'c8y-list__item--no-expand': noExpandToggle }\"\n  [collapsed]=\"collapsed\"\n  #listItem\n  id=\"{{ operation.id }}\"\n>\n  <c8y-li-icon>\n    <i\n      [c8yIcon]=\"OPERATION_STATUS_OPTIONS_MAP[operation.status].icon\"\n      [ngClass]=\"OPERATION_STATUS_OPTIONS_MAP[operation.status].styleClass\"\n      [tooltip]=\"OPERATION_STATUS_OPTIONS_MAP[operation.status].label | translate\"\n    ></i>\n  </c8y-li-icon>\n  <div\n    [ngClass]=\"{ 'content-flex-58': !readOnly, 'content-flex-50': readOnly }\"\n    [style.cursor]=\"!noExpandToggle ? 'pointer' : ''\"\n    (click)=\"!noExpandToggle && listItem.toggleCollapsed()\"\n  >\n    <div class=\"col-5\">\n      <span class=\"text-truncate\" title=\"{{ operationTitle | translate }}\">\n        {{ operationTitle | translate }}\n      </span>\n    </div>\n    <div class=\"col-3\" *ngIf=\"!deviceId\">\n      <a [routerLink]=\"['/device', operation.deviceId, 'device-info']\" routerLinkActive=\"active\">\n        {{ operation.deviceName }}\n      </a>\n    </div>\n    <div class=\"col-3\">\n      <small class=\"icon-flex text-muted\">\n        <i c8yIcon=\"calendar\" class=\"m-r-4\"></i>\n        {{ operation.creationTime | c8yDate }}\n      </small>\n    </div>\n  </div>\n  <ng-container *ngIf=\"!readOnly\">\n    <c8y-li-action\n      label=\"{{ 'Schedule as bulk operation' | translate }}\"\n      (click)=\"createCustomizedBulkOperation($event)\"\n      icon=\"c8y-icon c8y-icon-energy\"\n    >\n    </c8y-li-action>\n    <ng-container *c8yIfAllowed=\"inventoryPermissions; allowAny: true\">\n      <ng-container *c8yIfAllowed=\"smartRuleAdminPermissions; allowAny: true\">\n        <c8y-li-action\n          *ngIf=\"isSmartRulesMicroserviceAvailable\"\n          label=\"{{ 'Create smart rule' | translate }}\"\n          (click)=\"operationsListService.createSmartRule(operation)\"\n          icon=\"sliders\"\n        >\n        </c8y-li-action>\n      </ng-container>\n    </ng-container>\n    <c8y-li-action\n      *ngIf=\"operation.status === OPERATION_STATUS.PENDING\"\n      label=\"{{ 'Cancel operation' | translate }}\"\n      (click)=\"$event.stopPropagation(); sharedService.cancel(operation)\"\n      icon=\"times\"\n    >\n    </c8y-li-action>\n  </ng-container>\n  <c8y-li-collapse class=\"m-b-16\">\n    <c8y-operation-details-tabs\n      *ngIf=\"!listItem.collapsed\"\n      [operation]=\"operation\"\n      [bulkOperationModalDetailsService]=\"bulkOperationListItemService\"\n    ></c8y-operation-details-tabs>\n  </c8y-li-collapse>\n</c8y-li>\n"
            },] }
];
OperationsListItemComponent.ctorParameters = () => [
    { type: BulkOperationListItemService },
    { type: OperationsService },
    { type: OperationsListService },
    { type: ActivatedRoute },
    { type: Router }
];
OperationsListItemComponent.propDecorators = {
    operation: [{ type: Input }],
    collapsed: [{ type: Input }],
    readOnly: [{ type: Input }],
    noExpandToggle: [{ type: Input }],
    listItem: [{ type: ViewChild, args: ['listItem', { static: true },] }]
};
