// @c8y/ngx-components/esm2015/operations/operations-list/operations-list.service.js
import { __awaiter } from "tslib";
import { Injectable, Optional } from '@angular/core';
import { OperationService, SmartRulesService } from '@c8y/client';
import { Ng1SmartRulesService } from '@c8y/ngx-components/upgrade';
export class OperationsListService {
    constructor(operationService, smartRulesService, ng1SmartRulesService) {
        this.operationService = operationService;
        this.smartRulesService = smartRulesService;
        this.ng1SmartRulesService = ng1SmartRulesService;
    }
    getOperations(customFilter = {}) {
        const filter = Object.assign({ withTotalPages: true, withDeleted: true, pageSize: 50, dateFrom: new Date(0).toISOString(), revert: true }, customFilter);
        return this.operationService.list(filter);
    }
    isSmartRulesMicroserviceAvailable() {
        return this.smartRulesService.isMicroserviceAvailable();
    }
    getSmartRulesCreatePermissions() {
        if (this.ng1SmartRulesService && this.ng1SmartRulesService.permissionsCfgs) {
            return this.ng1SmartRulesService.permissionsCfgs.create;
        }
    }
    createSmartRule(operation) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.ng1SmartRulesService) {
                return yield this.ng1SmartRulesService.addNewForOutputOperationWithUI(operation);
            }
        });
    }
}
OperationsListService.decorators = [
    { type: Injectable }
];
OperationsListService.ctorParameters = () => [
    { type: OperationService },
    { type: SmartRulesService },
    { type: Ng1SmartRulesService, decorators: [{ type: Optional }] }
];
