// @c8y/ngx-components/esm2015/context-dashboard/report-dashboard/report-dashboard-list.component.js
import { __awaiter, __rest } from "tslib";
import { Component, ViewChild } from '@angular/core';
import { BehaviorSubject, combineLatest } from 'rxjs';
import { BsModalService } from 'ngx-bootstrap/modal';
import { InventoryService } from '@c8y/client';
import { debounceTime, distinctUntilChanged, shareReplay, switchMap, tap } from 'rxjs/operators';
import { FilterInputComponent, gettext, ModalService, Status, AlertService, Permissions } from '@c8y/ngx-components';
import { DashboardDetailComponent } from '../dashboard-detail.component';
import { ContextDashboardService } from '../context-dashboard.service';
import { ContextDashboardType } from '../context-dashboard.model';
import { TranslateService } from '@ngx-translate/core';
import { ReportDashboardService } from './report-dashboard.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/client';
import * as ɵngcc2 from '../context-dashboard.service';
import * as ɵngcc3 from 'ngx-bootstrap/modal';
import * as ɵngcc4 from '@ngx-translate/core';
import * as ɵngcc5 from '@c8y/ngx-components';
import * as ɵngcc6 from './report-dashboard.service';
import * as ɵngcc7 from '@angular/forms';
import * as ɵngcc8 from '@angular/common';
import * as ɵngcc9 from 'ngx-bootstrap/popover';
import * as ɵngcc10 from '@angular/router';

function ReportDashboardListComponent_i_11_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 18);
} }
function ReportDashboardListComponent_i_13_Template(rf, ctx) { if (rf & 1) {
    const _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "i", 19);
    ɵngcc0.ɵɵlistener("click", function ReportDashboardListComponent_i_13_Template_i_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r7); const ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.textFilter$.next(""); });
    ɵngcc0.ɵɵelementEnd();
} }
function ReportDashboardListComponent_button_21_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "button", 20);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 21);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("popover", ɵngcc0.ɵɵpipeBind1(1, 2, "You don't have permission to add reports"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", "info-circle");
} }
function ReportDashboardListComponent_c8y_li_31_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 22);
    ɵngcc0.ɵɵelementStart(1, "c8y-li-icon");
    ɵngcc0.ɵɵelement(2, "i", 23);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "c8y-li-body", 24);
    ɵngcc0.ɵɵelementStart(4, "div", 25);
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "div", 26);
    ɵngcc0.ɵɵtext(8);
    ɵngcc0.ɵɵpipe(9, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(10, "div", 27);
    ɵngcc0.ɵɵtext(11);
    ɵngcc0.ɵɵpipe(12, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(6, 3, "Report"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(9, 5, "Description"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(12, 7, "Show in navigator"), " ");
} }
function ReportDashboardListComponent_c8y_li_33_em_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "em", 45);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "No description available"), " ");
} }
function ReportDashboardListComponent_c8y_li_33_ng_template_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtext(0);
} if (rf & 2) {
    const report_r8 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵtextInterpolate1(" ", report_r8.description, " ");
} }
function ReportDashboardListComponent_c8y_li_33_Template(rf, ctx) { if (rf & 1) {
    const _r15 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li");
    ɵngcc0.ɵɵelement(1, "c8y-li-icon", 28);
    ɵngcc0.ɵɵelementStart(2, "c8y-li-body", 29);
    ɵngcc0.ɵɵelementStart(3, "div", 25);
    ɵngcc0.ɵɵelementStart(4, "button", 30);
    ɵngcc0.ɵɵelementStart(5, "span", 31);
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "div", 26);
    ɵngcc0.ɵɵelementStart(8, "small", 32);
    ɵngcc0.ɵɵtemplate(9, ReportDashboardListComponent_c8y_li_33_em_9_Template, 3, 3, "em", 33);
    ɵngcc0.ɵɵtemplate(10, ReportDashboardListComponent_c8y_li_33_ng_template_10_Template, 1, 1, "ng-template", null, 34, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(12, "div", 27);
    ɵngcc0.ɵɵelement(13, "span", 35);
    ɵngcc0.ɵɵelementStart(14, "label", 36);
    ɵngcc0.ɵɵpipe(15, "translate");
    ɵngcc0.ɵɵelementStart(16, "input", 37);
    ɵngcc0.ɵɵlistener("ngModelChange", function ReportDashboardListComponent_c8y_li_33_Template_input_ngModelChange_16_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r15); const report_r8 = restoredCtx.$implicit; return !!(report_r8.c8y_IsNavigatorNode = $event); })("change", function ReportDashboardListComponent_c8y_li_33_Template_input_change_16_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r15); const report_r8 = restoredCtx.$implicit; const ctx_r16 = ɵngcc0.ɵɵnextContext(); return ctx_r16.update(report_r8); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(17, "span");
    ɵngcc0.ɵɵelementStart(18, "span", 38);
    ɵngcc0.ɵɵtext(19);
    ɵngcc0.ɵɵpipe(20, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(21, "div", 39);
    ɵngcc0.ɵɵelementStart(22, "button", 40);
    ɵngcc0.ɵɵlistener("click", function ReportDashboardListComponent_c8y_li_33_Template_button_click_22_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r15); const report_r8 = restoredCtx.$implicit; const ctx_r17 = ɵngcc0.ɵɵnextContext(); return ctx_r17.delete(report_r8); });
    ɵngcc0.ɵɵpipe(23, "translate");
    ɵngcc0.ɵɵelement(24, "i", 41);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(25, "div", 42);
    ɵngcc0.ɵɵelementStart(26, "button", 43);
    ɵngcc0.ɵɵlistener("click", function ReportDashboardListComponent_c8y_li_33_Template_button_click_26_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r15); const report_r8 = restoredCtx.$implicit; const ctx_r18 = ɵngcc0.ɵɵnextContext(); return ctx_r18.delete(report_r8); });
    ɵngcc0.ɵɵpipe(27, "translate");
    ɵngcc0.ɵɵelement(28, "i", 44);
    ɵngcc0.ɵɵtext(29);
    ɵngcc0.ɵɵpipe(30, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const report_r8 = ctx.$implicit;
    const _r11 = ɵngcc0.ɵɵreference(11);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("icon", report_r8.icon);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", report_r8.name);
    ɵngcc0.ɵɵpropertyInterpolate1("routerLink", "/reports/", report_r8.id, "");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", report_r8.name, " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", !report_r8.description)("ngIfElse", _r11);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(15, 12, "Show in navigator"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngModel", !!report_r8.c8y_IsNavigatorNode);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(20, 14, "Show in navigator"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(23, 16, "Remove report"));
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(27, 18, "Remove report"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(30, 20, "Remove report"), " ");
} }
function ReportDashboardListComponent_div_34_Template(rf, ctx) { if (rf & 1) {
    const _r20 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 46);
    ɵngcc0.ɵɵelement(1, "h1", 47);
    ɵngcc0.ɵɵelementStart(2, "h3", 0);
    ɵngcc0.ɵɵtext(3, "There are no reports defined");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "p", 0);
    ɵngcc0.ɵɵtext(5, "Add a report first.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "div");
    ɵngcc0.ɵɵelementStart(7, "button", 48);
    ɵngcc0.ɵɵlistener("click", function ReportDashboardListComponent_div_34_Template_button_click_7_listener() { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r19 = ɵngcc0.ɵɵnextContext(); return ctx_r19.add(); });
    ɵngcc0.ɵɵtext(8, "Add report");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(9, "p", 49);
    ɵngcc0.ɵɵelementStart(10, "small", 0);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(11, "Find out more in the ");
    ɵngcc0.ɵɵelementStart(12, "a", 50);
    ɵngcc0.ɵɵtext(13, " User guide`KEEP_ORIGINAL` ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(14, ". ");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
const _c0 = function (a0) { return { "icon-spin": a0 }; };
export class ReportDashboardListComponent {
    constructor(inventoryService, contextDashboardService, bsModal, translateService, modal, alertService, reportDashboardService, permissions) {
        this.inventoryService = inventoryService;
        this.contextDashboardService = contextDashboardService;
        this.bsModal = bsModal;
        this.translateService = translateService;
        this.modal = modal;
        this.alertService = alertService;
        this.reportDashboardService = reportDashboardService;
        this.permissions = permissions;
        this.textFilter$ = new BehaviorSubject('');
        this.reload$ = new BehaviorSubject(null);
        this.reloading = false;
        this.reports$ = combineLatest(this.textFilter$.pipe(debounceTime(400), distinctUntilChanged()), this.reload$).pipe(tap(() => {
            this.reloading = true;
        }), switchMap(([text]) => this.loadReports(text)), tap(() => {
            this.reloading = false;
        }), shareReplay(1));
        this.canAddReport = false;
        this.DELETED_SUCCESS_MSG = gettext('Report deleted.');
    }
    ngOnInit() {
        this.canAddReport = this.permissions.hasAnyRole([
            'ROLE_INVENTORY_CREATE',
            'ROLE_INVENTORY_ADMIN'
        ]);
    }
    loadReports(partialName) {
        return __awaiter(this, void 0, void 0, function* () {
            return partialName
                ? this.reportDashboardService.listReports({ filter: { name: `*${partialName}*` } })
                : this.reportDashboardService.listReports();
        });
    }
    add() {
        return __awaiter(this, void 0, void 0, function* () {
            const initialState = {
                isReport: true,
                isNamedDashboard: true
            };
            const modal = this.bsModal.show(DashboardDetailComponent, {
                class: 'modal-lg',
                ignoreBackdropClick: true,
                initialState
            }).content;
            try {
                const cfg = yield modal.result;
                const { name, icon, c8y_IsNavigatorNode, priority, description } = cfg, dashboardCfg = __rest(cfg, ["name", "icon", "c8y_IsNavigatorNode", "priority", "description"]);
                const report = (yield this.reportDashboardService.createReport({
                    name,
                    icon,
                    c8y_IsNavigatorNode,
                    priority,
                    description
                })).data;
                yield this.contextDashboardService.create(dashboardCfg, undefined, `${this.contextDashboardService.REPORT_PARTIAL_NAME}${report.id}`);
                if (report.c8y_IsNavigatorNode) {
                    this.reportDashboardService.addReportNavigatorNode(report);
                }
                this.reload$.next();
                modal.close();
            }
            catch (ex) {
                if (ex) {
                    throw new Error(`Something went wrong: ${ex}`);
                }
            }
        });
    }
    delete(report) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const dashboard = yield this.contextDashboardService
                    .getDashboard$(`report_${report.id}`, [ContextDashboardType.Named])
                    .toPromise();
                const isDashboardFrozen = !!((_a = dashboard === null || dashboard === void 0 ? void 0 : dashboard.c8y_Dashboard) === null || _a === void 0 ? void 0 : _a.isFrozen);
                const msg = isDashboardFrozen
                    ? gettext(`You are about to delete the report "{{ reportName }}". Be aware that the report has a locked dashboard assigned. Do you want to proceed?`)
                    : gettext(`You are about to delete the report "{{ reportName }}". Do you want to proceed?`);
                yield this.modal.confirm(gettext('Delete report'), this.translateService.instant(msg, {
                    reportName: report.name
                }), Status.DANGER, {
                    ok: gettext('Delete'),
                    cancel: gettext('Cancel')
                });
                yield this.contextDashboardService.delete(dashboard, false);
                yield this.inventoryService.delete(report.id);
                this.alertService.success(this.DELETED_SUCCESS_MSG);
                if (report.c8y_IsNavigatorNode) {
                    this.reportDashboardService.removeNavigatorNode(report);
                }
                this.reload$.next();
            }
            catch (ex) {
                if (ex) {
                    throw new Error(`Something went wrong: ${ex}`);
                }
            }
        });
    }
    update(report) {
        return __awaiter(this, void 0, void 0, function* () {
            report.c8y_IsNavigatorNode = report.c8y_IsNavigatorNode ? {} : null;
            yield this.inventoryService.update(report);
            report.c8y_IsNavigatorNode
                ? this.reportDashboardService.addReportNavigatorNode(report)
                : this.reportDashboardService.removeNavigatorNode(report);
        });
    }
}
ReportDashboardListComponent.ɵfac = function ReportDashboardListComponent_Factory(t) { return new (t || ReportDashboardListComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.InventoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ContextDashboardService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.BsModalService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.TranslateService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.ModalService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc6.ReportDashboardService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.Permissions)); };
ReportDashboardListComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ReportDashboardListComponent, selectors: [["c8y-report-dashboard-list"]], viewQuery: function ReportDashboardListComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(FilterInputComponent, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.filter = _t.first);
    } }, decls: 36, vars: 43, consts: [["translate", ""], ["itemClass", "navbar-form", 3, "placement"], [1, "input-group", "input-group-search"], ["type", "search", 1, "form-control", 3, "title", "placeholder", "ngModel", "ngModelChange"], [1, "input-group-addon"], ["c8yIcon", "search", 4, "ngIf"], ["class", "text-muted", "c8yIcon", "times", "px-event", "Clear filtering reports", 3, "click", 4, "ngIf"], [3, "placement"], [1, "btn", "btn-link", 3, "title", "disabled", "click"], ["c8yIcon", "plus-circle"], ["class", "btn btn-clean m-r-16", "triggers", "focus", "placement", "top", "container", "body", 3, "popover", 4, "ngIf"], [1, "btn", "btn-link", 3, "title", "click"], ["c8yIcon", "refresh", 3, "ngClass"], ["src", "/users-guide/cockpit/#reports"], [1, "p-b-32"], ["class", "page-sticky-header hidden-xs hidden-sm", 4, "ngIf"], [4, "c8yFor", "c8yForOf", "c8yForLoadMore"], ["class", "c8y-empty-state text-center m-t-40", 4, "ngIf"], ["c8yIcon", "search"], ["c8yIcon", "times", "px-event", "Clear filtering reports", 1, "text-muted", 3, "click"], ["triggers", "focus", "placement", "top", "container", "body", 1, "btn", "btn-clean", "m-r-16", 3, "popover"], [1, "text-primary", 3, "c8yIcon"], [1, "page-sticky-header", "hidden-xs", "hidden-sm"], [1, "p-l-24"], [1, "content-flex-80"], [1, "col-3"], [1, "col-6"], [1, "col-2"], [3, "icon"], [1, "content-flex-70"], [1, "btn-clean", 3, "title", "routerLink"], [1, "text-truncate"], [1, "text-truncate-wrap"], ["class", "text-muted", 4, "ngIf", "ngIfElse"], ["showDescription", ""], [1, "m-t-8", "visible-xs"], [1, "c8y-switch", "c8y-switch--inline", 3, "title"], ["type", "checkbox", 3, "ngModel", "ngModelChange", "change"], [1, "visible-xs"], [1, "col-1", "text-right", "hidden-xs"], [1, "btn-dot", "showOnHover", "pull-right", 3, "title", "click"], ["c8yIcon", "minus-circle", 1, "text-danger"], [1, "visible-xs", "p-t-8", "text-right"], [1, "btn-danger", "btn", "btn-xs", 3, "title", "click"], ["c8yIcon", "trash"], [1, "text-muted"], [1, "c8y-empty-state", "text-center", "m-t-40"], ["c8yIcon", "c8y-reports", 1, "c8y-icon-duocolor"], ["translate", "", 1, "btn", "btn-primary", 3, "click"], ["c8y-guide-docs", ""], ["c8y-guide-href", "users-guide/cockpit/#reports"]], template: function ReportDashboardListComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵelementStart(1, "span", 0);
        ɵngcc0.ɵɵtext(2, " Reports ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtext(3, "\u00A0 ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "c8y-action-bar-item", 1);
        ɵngcc0.ɵɵelementStart(5, "div", 2);
        ɵngcc0.ɵɵelementStart(6, "input", 3);
        ɵngcc0.ɵɵlistener("ngModelChange", function ReportDashboardListComponent_Template_input_ngModelChange_6_listener($event) { return ctx.textFilter$.next($event); });
        ɵngcc0.ɵɵpipe(7, "translate");
        ɵngcc0.ɵɵpipe(8, "translate");
        ɵngcc0.ɵɵpipe(9, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(10, "span", 4);
        ɵngcc0.ɵɵtemplate(11, ReportDashboardListComponent_i_11_Template, 1, 0, "i", 5);
        ɵngcc0.ɵɵpipe(12, "async");
        ɵngcc0.ɵɵtemplate(13, ReportDashboardListComponent_i_13_Template, 1, 0, "i", 6);
        ɵngcc0.ɵɵpipe(14, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(15, "c8y-action-bar-item", 7);
        ɵngcc0.ɵɵelementStart(16, "button", 8);
        ɵngcc0.ɵɵlistener("click", function ReportDashboardListComponent_Template_button_click_16_listener() { return ctx.add(); });
        ɵngcc0.ɵɵpipe(17, "translate");
        ɵngcc0.ɵɵelement(18, "i", 9);
        ɵngcc0.ɵɵtext(19);
        ɵngcc0.ɵɵpipe(20, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(21, ReportDashboardListComponent_button_21_Template, 3, 4, "button", 10);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(22, "c8y-action-bar-item", 7);
        ɵngcc0.ɵɵelementStart(23, "button", 11);
        ɵngcc0.ɵɵlistener("click", function ReportDashboardListComponent_Template_button_click_23_listener() { return ctx.loadReports(); });
        ɵngcc0.ɵɵpipe(24, "translate");
        ɵngcc0.ɵɵelement(25, "i", 12);
        ɵngcc0.ɵɵtext(26);
        ɵngcc0.ɵɵpipe(27, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(28, "c8y-help", 13);
        ɵngcc0.ɵɵelementStart(29, "div", 14);
        ɵngcc0.ɵɵelementStart(30, "c8y-list-group");
        ɵngcc0.ɵɵtemplate(31, ReportDashboardListComponent_c8y_li_31_Template, 13, 9, "c8y-li", 15);
        ɵngcc0.ɵɵpipe(32, "async");
        ɵngcc0.ɵɵtemplate(33, ReportDashboardListComponent_c8y_li_33_Template, 31, 22, "c8y-li", 16);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(34, ReportDashboardListComponent_div_34_Template, 15, 0, "div", 17);
        ɵngcc0.ɵɵpipe(35, "async");
    } if (rf & 2) {
        let tmp_15_0;
        let tmp_18_0;
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("placement", "left");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(7, 19, "Filter\u2026"));
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(8, 21, "Filter\u2026"));
        ɵngcc0.ɵɵproperty("ngModel", ɵngcc0.ɵɵpipeBind1(9, 23, ctx.textFilter$));
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(12, 25, ctx.textFilter$).length === 0);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(14, 27, ctx.textFilter$).length > 0);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("placement", "right");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(17, 29, "Add report"));
        ɵngcc0.ɵɵproperty("disabled", !ctx.canAddReport);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(20, 31, "Add report"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.canAddReport);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("placement", "right");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(24, 33, "Reload"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(41, _c0, ctx.reloading));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(27, 35, "Reload"), " ");
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("ngIf", ((tmp_15_0 = ɵngcc0.ɵɵpipeBind1(32, 37, ctx.reports$)) == null ? null : tmp_15_0.data.length) > 0);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yForOf", ctx.reports$)("c8yForLoadMore", "auto");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ((tmp_18_0 = ɵngcc0.ɵɵpipeBind1(35, 39, ctx.reports$)) == null ? null : tmp_18_0.data.length) === 0);
    } }, directives: [ɵngcc5.TitleComponent, ɵngcc5.C8yTranslateDirective, ɵngcc5.ActionBarItemComponent, ɵngcc7.DefaultValueAccessor, ɵngcc7.NgControlStatus, ɵngcc7.NgModel, ɵngcc8.NgIf, ɵngcc5.IconDirective, ɵngcc8.NgClass, ɵngcc5.HelpComponent, ɵngcc5.ListGroupComponent, ɵngcc5.ForOfDirective, ɵngcc9.PopoverDirective, ɵngcc5.ListItemComponent, ɵngcc5.ListItemIconComponent, ɵngcc5.ListItemBodyComponent, ɵngcc10.RouterLink, ɵngcc7.CheckboxControlValueAccessor, ɵngcc5.GuideDocsComponent, ɵngcc5.GuideHrefDirective], pipes: [ɵngcc5.C8yTranslatePipe, ɵngcc8.AsyncPipe], encapsulation: 2 });
ReportDashboardListComponent.ctorParameters = () => [
    { type: InventoryService },
    { type: ContextDashboardService },
    { type: BsModalService },
    { type: TranslateService },
    { type: ModalService },
    { type: AlertService },
    { type: ReportDashboardService },
    { type: Permissions }
];
ReportDashboardListComponent.propDecorators = {
    filter: [{ type: ViewChild, args: [FilterInputComponent, { static: false },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ReportDashboardListComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-report-dashboard-list',
                template: "<c8y-title> <span translate> Reports </span>&nbsp; </c8y-title>\n\n<c8y-action-bar-item [placement]=\"'left'\" itemClass=\"navbar-form\">\n  <div class=\"input-group input-group-search\">\n    <input\n      class=\"form-control\"\n      type=\"search\"\n      title=\"{{ 'Filter\u2026' | translate }}\"\n      placeholder=\"{{ 'Filter\u2026' | translate }}\"\n      [ngModel]=\"textFilter$ | async\"\n      (ngModelChange)=\"textFilter$.next($event)\"\n    />\n    <span class=\"input-group-addon\">\n      <i c8yIcon=\"search\" *ngIf=\"(textFilter$ | async).length === 0\"></i>\n      <i\n        class=\"text-muted\"\n        c8yIcon=\"times\"\n        *ngIf=\"(textFilter$ | async).length > 0\"\n        (click)=\"textFilter$.next('')\"\n        px-event=\"Clear filtering reports\"\n      ></i>\n    </span>\n  </div>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button\n    class=\"btn btn-link\"\n    title=\"{{ 'Add report' | translate }}\"\n    (click)=\"add()\"\n    [disabled]=\"!canAddReport\"\n  >\n    <i c8yIcon=\"plus-circle\"></i>\n    {{ 'Add report' | translate }}\n  </button>\n  <button *ngIf=\"!canAddReport\"\n    class=\"btn btn-clean m-r-16\"\n    popover=\"{{ 'You don\\'t have permission to add reports' | translate }}\"\n    triggers=\"focus\"\n    placement=\"top\"\n    container=\"body\"\n  >\n    <i [c8yIcon]=\"'info-circle'\" class=\"text-primary\"></i>\n  </button>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button class=\"btn btn-link\" title=\"{{ 'Reload' | translate }}\" (click)=\"loadReports()\">\n    <i [ngClass]=\"{ 'icon-spin': reloading }\" c8yIcon=\"refresh\"></i>\n    {{ 'Reload' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<c8y-help src=\"/users-guide/cockpit/#reports\"></c8y-help>\n\n<div class=\"p-b-32\">\n  <c8y-list-group>\n    <c8y-li\n      class=\"page-sticky-header hidden-xs hidden-sm\"\n      *ngIf=\"(reports$ | async)?.data.length > 0\"\n    >\n      <c8y-li-icon>\n        <i class=\"p-l-24\"></i>\n      </c8y-li-icon>\n      <c8y-li-body class=\"content-flex-80\">\n        <div class=\"col-3\">\n          {{ 'Report' | translate }}\n        </div>\n        <div class=\"col-6\">\n          {{ 'Description' | translate }}\n        </div>\n        <div class=\"col-2\">\n          {{ 'Show in navigator' | translate }}\n        </div>\n      </c8y-li-body>\n    </c8y-li>\n\n    <c8y-li *c8yFor=\"let report of reports$; let i = index; loadMore: 'auto'\">\n      <c8y-li-icon [icon]=\"report.icon\"></c8y-li-icon>\n      <c8y-li-body class=\"content-flex-70\">\n        <div class=\"col-3\">\n          <button class=\"btn-clean\" title=\"{{ report.name }}\" routerLink=\"/reports/{{ report.id }}\">\n            <span class=\"text-truncate\">\n              {{ report.name }}\n            </span>\n          </button>\n        </div>\n        <div class=\"col-6\">\n          <small class=\"text-truncate-wrap\">\n            <em class=\"text-muted\" *ngIf=\"!report.description; else showDescription\">\n              {{ 'No description available' | translate }}\n            </em>\n            <ng-template #showDescription>\n              {{ report.description }}\n            </ng-template>\n          </small>\n        </div>\n        <div class=\"col-2\">\n          <span class=\"m-t-8 visible-xs\"></span>\n          <label\n            class=\"c8y-switch c8y-switch--inline\"\n            title=\"{{ 'Show in navigator' | translate }}\"\n          >\n            <input\n              [(ngModel)]=\"!!report.c8y_IsNavigatorNode\"\n              type=\"checkbox\"\n              (change)=\"update(report)\"\n            />\n            <span></span>\n            <span class=\"visible-xs\">\n              {{ 'Show in navigator' | translate }}\n            </span>\n          </label>\n        </div>\n        <div class=\"col-1 text-right hidden-xs\">\n          <button\n            class=\"btn-dot showOnHover pull-right\"\n            (click)=\"delete(report)\"\n            title=\"{{ 'Remove report' | translate }}\"\n          >\n            <i class=\"text-danger\" c8yIcon=\"minus-circle\"></i>\n          </button>\n        </div>\n        <div class=\"visible-xs p-t-8 text-right\">\n          <button\n            class=\"btn-danger btn btn-xs\"\n            (click)=\"delete(report)\"\n            title=\"{{ 'Remove report' | translate }}\"\n          >\n            <i c8yIcon=\"trash\"></i>\n            {{ 'Remove report' | translate }}\n          </button>\n        </div>\n      </c8y-li-body>\n    </c8y-li>\n  </c8y-list-group>\n</div>\n\n<div class=\"c8y-empty-state text-center m-t-40\" *ngIf=\"(reports$ | async)?.data.length === 0\">\n  <h1 c8yIcon=\"c8y-reports\" class=\"c8y-icon-duocolor\"></h1>\n  <h3 translate>There are no reports defined</h3>\n  <p translate>Add a report first.</p>\n  <div>\n    <button class=\"btn btn-primary\" (click)=\"add()\" translate>Add report</button>\n  </div>\n  <p c8y-guide-docs>\n    <small translate ngNonBindable\n      >Find out more in the\n      <a c8y-guide-href=\"users-guide/cockpit/#reports\"> User guide`KEEP_ORIGINAL` </a>.\n    </small>\n  </p>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.InventoryService }, { type: ɵngcc2.ContextDashboardService }, { type: ɵngcc3.BsModalService }, { type: ɵngcc4.TranslateService }, { type: ɵngcc5.ModalService }, { type: ɵngcc5.AlertService }, { type: ɵngcc6.ReportDashboardService }, { type: ɵngcc5.Permissions }]; }, { filter: [{
            type: ViewChild,
            args: [FilterInputComponent, { static: false }]
        }] }); })();
