// @c8y/ngx-components/esm2015/context-dashboard/report-dashboard/report-dashboard.guard.js
import { Injectable } from '@angular/core';
import { ContextDashboardType } from '../context-dashboard.model';
import { ContextDashboardService } from '../context-dashboard.service';
import { tap } from 'rxjs/operators';
import { from, of } from 'rxjs';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../context-dashboard.service';
export class ReportDashboardGuard {
    constructor(contextDashboardService) {
        this.contextDashboardService = contextDashboardService;
    }
    canActivate(route) {
        const reportId = route.data.contextData && route.data.contextData.id;
        return from(this.contextDashboardService.activateDashboards(route, [ContextDashboardType.Named])).pipe(tap(res => {
            return Array.isArray(res) && res.length === 0
                ? this.createDashboardIfNotExist(reportId)
                : of(true);
        }));
    }
    createDashboardIfNotExist(reportId) {
        const dashboardCfg = {};
        return this.contextDashboardService
            .create(dashboardCfg, undefined, `report_${reportId}`)
            .then(() => true);
    }
}
ReportDashboardGuard.ɵfac = function ReportDashboardGuard_Factory(t) { return new (t || ReportDashboardGuard)(ɵngcc0.ɵɵinject(ɵngcc1.ContextDashboardService)); };
ReportDashboardGuard.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: ReportDashboardGuard, factory: ReportDashboardGuard.ɵfac });
ReportDashboardGuard.ctorParameters = () => [
    { type: ContextDashboardService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ReportDashboardGuard, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.ContextDashboardService }]; }, null); })();
