// @c8y/ngx-components/esm2015/device-grid/decorator.service.js
import { Injectable } from '@angular/core';
import { DeviceGridConfigMapperService } from './device-grid-config-mapper.service';
import * as i0 from "@angular/core";
import * as i1 from "./device-grid-config-mapper.service";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './device-grid-config-mapper.service';
export class DecoratorService {
    constructor(service) {
        DecoratorService.mapperService = service;
    }
    static getMapperService() {
        if (!DecoratorService.mapperService) {
            throw new Error('DecoratorService not initialized');
        }
        return DecoratorService.mapperService;
    }
}
DecoratorService.ɵfac = function DecoratorService_Factory(t) { return new (t || DecoratorService)(ɵngcc0.ɵɵinject(ɵngcc1.DeviceGridConfigMapperService)); };
DecoratorService.mapperService = undefined;
DecoratorService.ɵprov = i0.ɵɵdefineInjectable({ factory: function DecoratorService_Factory() { return new DecoratorService(i0.ɵɵinject(i1.DeviceGridConfigMapperService)); }, token: DecoratorService, providedIn: "root" });
DecoratorService.ctorParameters = () => [
    { type: DeviceGridConfigMapperService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DecoratorService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: ɵngcc1.DeviceGridConfigMapperService }]; }, null); })();
