// @c8y/ngx-components/esm2015/widgets/kpi-widget/kpi-widget-config/kpi-widget-config.component.js
import { __awaiter } from "tslib";
import { Component, Input } from '@angular/core';
import { ControlContainer, FormBuilder, NgForm, Validators } from '@angular/forms';
import { WidgetConfigComponent } from '@c8y/ngx-components/context-dashboard';
import { DatapointLibraryService } from '@c8y/ngx-components/datapoint-selector';
import { IconSelectorService } from '@c8y/ngx-components/icon-selector';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/forms';
import * as ɵngcc2 from '@c8y/ngx-components/icon-selector';
import * as ɵngcc3 from '@c8y/ngx-components/context-dashboard';
import * as ɵngcc4 from '@c8y/ngx-components/datapoint-selector';
import * as ɵngcc5 from '@c8y/ngx-components';
import * as ɵngcc6 from 'ngx-bootstrap/popover';

const _c0 = function () { return { example: 1 }; };
const _c1 = function () { return { example: 36 }; };
export function exactlyASingleDatapointActive() {
    return (control) => {
        const datapoints = control.value;
        if (!datapoints || !datapoints.length) {
            return null;
        }
        const activeDatapoints = datapoints.filter(datapoint => datapoint.__active);
        if (activeDatapoints.length === 1) {
            return null;
        }
        return { exactlyOneDatapointNeedsToBeActive: true };
    };
}
export class KpiWidgetConfigComponent {
    constructor(formBuilder, form, iconSelector, widgetConfig, datapointLibrary) {
        this.formBuilder = formBuilder;
        this.form = form;
        this.iconSelector = iconSelector;
        this.widgetConfig = widgetConfig;
        this.datapointLibrary = datapointLibrary;
        this.datapointSelectionConfig = {};
        this.defaultFormOptions = {
            showRedRange: true,
            showYellowRange: true
        };
        this.availableIcons = [];
        this.limits = {
            fontSizeMax: 72,
            fontSizeMin: 18,
            numberOfDecimalPlacesMax: 10,
            numberOfDecimalPlacesMin: 0
        };
    }
    onBeforeSave(config) {
        if (this.formGroup.valid) {
            Object.assign(config, this.formGroup.value);
            return true;
        }
        return false;
    }
    ngOnInit() {
        var _a, _b, _c, _d;
        return __awaiter(this, void 0, void 0, function* () {
            if ((_a = this.widgetConfig.context) === null || _a === void 0 ? void 0 : _a.id) {
                this.datapointSelectionConfig.contextAsset = (_b = this.widgetConfig) === null || _b === void 0 ? void 0 : _b.context;
            }
            this.initForm();
            if ((_c = this.config) === null || _c === void 0 ? void 0 : _c.datapoints) {
                this.config.datapoints = yield this.datapointLibrary.updateDatapoints((_d = this.config) === null || _d === void 0 ? void 0 : _d.datapoints);
                this.formGroup.patchValue({ datapoints: this.config.datapoints });
            }
        });
    }
    openIconSelector() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const icon = yield this.iconSelector.selectIcon({
                    currentSelection: this.formGroup.value.icon
                });
                this.formGroup.patchValue({ icon });
            }
            catch (_a) {
                // nothing to do
            }
        });
    }
    initForm() {
        this.formGroup = this.formBuilder.group({
            numberOfDecimalPlaces: [
                1,
                [
                    Validators.required,
                    Validators.min(this.limits.numberOfDecimalPlacesMin),
                    Validators.max(this.limits.numberOfDecimalPlacesMax)
                ]
            ],
            showTimestamp: [true, []],
            showTrend: [true, []],
            showIcon: [true, []],
            icon: ['water', [Validators.required, Validators.minLength(1)]],
            fontSize: [
                36,
                [
                    Validators.required,
                    Validators.min(this.limits.fontSizeMin),
                    Validators.max(this.limits.fontSizeMax)
                ]
            ],
            datapoints: [
                [],
                [Validators.required, Validators.minLength(1), exactlyASingleDatapointActive()]
            ]
        });
        this.form.form.addControl('config', this.formGroup);
        this.formGroup.patchValue(this.config);
    }
}
KpiWidgetConfigComponent.ɵfac = function KpiWidgetConfigComponent_Factory(t) { return new (t || KpiWidgetConfigComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.FormBuilder), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.NgForm), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.IconSelectorService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.WidgetConfigComponent), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.DatapointLibraryService)); };
KpiWidgetConfigComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: KpiWidgetConfigComponent, selectors: [["c8y-kpi-widget-config"]], inputs: { config: "config" }, features: [ɵngcc0.ɵɵProvidersFeature([], [{ provide: ControlContainer, useExisting: NgForm }])], decls: 56, vars: 44, consts: [[1, "row", 3, "formGroup"], [1, "col-sm-6"], ["formControlName", "datapoints", "name", "datapoints", 3, "defaultFormOptions", "config", "minActiveCount", "maxActiveCount"], ["translate", ""], [1, "d-flex", "a-i-center"], [1, "p-r-8", "icon-32", "text-muted"], [3, "c8yIcon"], [1, "btn", "btn-default", "btn-xs", 3, "click"], ["translate", "", 3, "title"], ["formControlName", "numberOfDecimalPlaces", "name", "numberOfDecimalPlaces", "type", "number", 1, "form-control", 3, "placeholder"], [3, "show"], [1, "d-flex", "gap-16", "flex-wrap"], [1, "c8y-checkbox", 3, "title"], ["type", "checkbox", "formControlName", "showTimestamp", "name", "showTimestamp"], ["type", "checkbox", "formControlName", "showIcon", "name", "showIcon"], ["type", "checkbox", "formControlName", "showTrend", "name", "showTrend"], ["container", "body", 1, "btn", "btn-clean", 3, "popover", "outsideClick"], [1, "text-info", 3, "c8yIcon"], ["formControlName", "fontSize", "name", "fontSize", "type", "number", 1, "form-control", 3, "placeholder"]], template: function KpiWidgetConfigComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "form", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelement(2, "c8y-datapoint-selection-list", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "div", 1);
        ɵngcc0.ɵɵelementStart(4, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(5, "label", 3);
        ɵngcc0.ɵɵtext(6, "Icon");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(7, "div", 4);
        ɵngcc0.ɵɵelementStart(8, "div", 5);
        ɵngcc0.ɵɵelement(9, "i", 6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(10, "button", 7);
        ɵngcc0.ɵɵlistener("click", function KpiWidgetConfigComponent_Template_button_click_10_listener() { return ctx.openIconSelector(); });
        ɵngcc0.ɵɵtext(11);
        ɵngcc0.ɵɵpipe(12, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(13, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(14, "label", 8);
        ɵngcc0.ɵɵpipe(15, "translate");
        ɵngcc0.ɵɵtext(16, " Number of decimal places ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(17, "input", 9);
        ɵngcc0.ɵɵpipe(18, "translate");
        ɵngcc0.ɵɵelement(19, "c8y-messages", 10);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(20, "div");
        ɵngcc0.ɵɵelementStart(21, "label");
        ɵngcc0.ɵɵtext(22);
        ɵngcc0.ɵɵpipe(23, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(24, "div", 11);
        ɵngcc0.ɵɵelementStart(25, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(26, "label", 12);
        ɵngcc0.ɵɵpipe(27, "translate");
        ɵngcc0.ɵɵelement(28, "input", 13);
        ɵngcc0.ɵɵelement(29, "span");
        ɵngcc0.ɵɵelementStart(30, "span", 3);
        ɵngcc0.ɵɵtext(31, "Show timestamp");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(32, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(33, "label", 12);
        ɵngcc0.ɵɵpipe(34, "translate");
        ɵngcc0.ɵɵelement(35, "input", 14);
        ɵngcc0.ɵɵelement(36, "span");
        ɵngcc0.ɵɵelementStart(37, "span", 3);
        ɵngcc0.ɵɵtext(38, "Show icon");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(39, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(40, "label", 12);
        ɵngcc0.ɵɵpipe(41, "translate");
        ɵngcc0.ɵɵelement(42, "input", 15);
        ɵngcc0.ɵɵelement(43, "span");
        ɵngcc0.ɵɵelementStart(44, "span", 3);
        ɵngcc0.ɵɵtext(45, "Show trend icon");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(46, "button", 16);
        ɵngcc0.ɵɵpipe(47, "translate");
        ɵngcc0.ɵɵelement(48, "i", 17);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(49, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(50, "label", 8);
        ɵngcc0.ɵɵpipe(51, "translate");
        ɵngcc0.ɵɵtext(52, " Font size of measurement value (px) ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(53, "input", 18);
        ɵngcc0.ɵɵpipe(54, "translate");
        ɵngcc0.ɵɵelement(55, "c8y-messages", 10);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("formGroup", ctx.formGroup);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("defaultFormOptions", ctx.defaultFormOptions)("config", ctx.datapointSelectionConfig)("minActiveCount", 1)("maxActiveCount", 1);
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵproperty("c8yIcon", ctx.formGroup.value.icon);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(12, 20, "Change"), " ");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(15, 22, "Number of decimal places"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("placeholder", ɵngcc0.ɵɵpipeBind2(18, 24, "e.g. {{ example }}", ɵngcc0.ɵɵpureFunction0(42, _c0)));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("show", (ctx.formGroup.controls == null ? null : ctx.formGroup.controls.numberOfDecimalPlaces == null ? null : ctx.formGroup.controls.numberOfDecimalPlaces.touched) && (ctx.formGroup == null ? null : ctx.formGroup.controls == null ? null : ctx.formGroup.controls.numberOfDecimalPlaces == null ? null : ctx.formGroup.controls.numberOfDecimalPlaces.errors));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(23, 27, "Display"), "");
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(27, 29, "Show timestamp"));
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(34, 31, "Show icon"));
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(41, 33, "Show trend icon"));
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵpropertyInterpolate("popover", ɵngcc0.ɵɵpipeBind1(47, 35, "Indicates the trend between the last two measurement values."));
        ɵngcc0.ɵɵproperty("outsideClick", true);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yIcon", "question-circle-o");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(51, 37, "Font size of measurement value (px)"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("placeholder", ɵngcc0.ɵɵpipeBind2(54, 39, "e.g. {{ example }}", ɵngcc0.ɵɵpureFunction0(43, _c1)));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("show", (ctx.formGroup.controls == null ? null : ctx.formGroup.controls.fontSize == null ? null : ctx.formGroup.controls.fontSize.touched) && (ctx.formGroup == null ? null : ctx.formGroup.controls == null ? null : ctx.formGroup.controls.fontSize == null ? null : ctx.formGroup.controls.fontSize.errors));
    } }, directives: [ɵngcc1.ɵNgNoValidate, ɵngcc1.NgControlStatusGroup, ɵngcc1.FormGroupDirective, ɵngcc4.DatapointSelectionListComponent, ɵngcc1.NgControlStatus, ɵngcc1.FormControlName, ɵngcc5.FormGroupComponent, ɵngcc5.C8yTranslateDirective, ɵngcc5.IconDirective, ɵngcc1.DefaultValueAccessor, ɵngcc5.RequiredInputPlaceholderDirective, ɵngcc1.NumberValueAccessor, ɵngcc5.MessagesComponent, ɵngcc1.CheckboxControlValueAccessor, ɵngcc6.PopoverDirective], pipes: [ɵngcc5.C8yTranslatePipe], encapsulation: 2 });
KpiWidgetConfigComponent.ctorParameters = () => [
    { type: FormBuilder },
    { type: NgForm },
    { type: IconSelectorService },
    { type: WidgetConfigComponent },
    { type: DatapointLibraryService }
];
KpiWidgetConfigComponent.propDecorators = {
    config: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(KpiWidgetConfigComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-kpi-widget-config',
                template: "<form [formGroup]=\"formGroup\" class=\"row\">\n  <div class=\"col-sm-6\">\n    <c8y-datapoint-selection-list\n      [defaultFormOptions]=\"defaultFormOptions\"\n      [config]=\"datapointSelectionConfig\"\n      [minActiveCount]=\"1\"\n      [maxActiveCount]=\"1\"\n      formControlName=\"datapoints\"\n      name=\"datapoints\"\n    >\n    </c8y-datapoint-selection-list>\n  </div>\n  <div class=\"col-sm-6\">\n    <c8y-form-group>\n      <label translate>Icon</label>\n      <div class=\"d-flex a-i-center\">\n        <div class=\"p-r-8 icon-32 text-muted\">\n          <i [c8yIcon]=\"formGroup.value.icon\"></i>\n        </div>\n        <button class=\"btn btn-default btn-xs\" (click)=\"openIconSelector()\">\n          {{ 'Change' | translate }}\n        </button>\n      </div>\n    </c8y-form-group>\n    <c8y-form-group>\n      <label [title]=\"'Number of decimal places' | translate\" translate>\n        Number of decimal places\n      </label>\n      <input\n        class=\"form-control\"\n        formControlName=\"numberOfDecimalPlaces\"\n        name=\"numberOfDecimalPlaces\"\n        type=\"number\"\n        [placeholder]=\"'e.g. {{ example }}' | translate: { example: 1 }\"\n      />\n      <c8y-messages [show]=\"formGroup.controls?.numberOfDecimalPlaces?.touched && formGroup?.controls?.numberOfDecimalPlaces?.errors\">\n      </c8y-messages>\n    </c8y-form-group>\n    <div>\n      <label> {{ 'Display' | translate }}</label>\n      <div class=\"d-flex gap-16 flex-wrap\">\n        <c8y-form-group>\n          <label [title]=\"'Show timestamp' | translate\" class=\"c8y-checkbox\">\n            <input type=\"checkbox\" formControlName=\"showTimestamp\" name=\"showTimestamp\" />\n            <span></span>\n            <span translate>Show timestamp</span>\n          </label>\n        </c8y-form-group>\n\n        <c8y-form-group>\n          <label [title]=\"'Show icon' | translate\" class=\"c8y-checkbox\">\n            <input type=\"checkbox\" formControlName=\"showIcon\" name=\"showIcon\" />\n            <span></span>\n            <span translate>Show icon</span>\n          </label>\n        </c8y-form-group>\n\n        <c8y-form-group>\n          <label [title]=\"'Show trend icon' | translate\" class=\"c8y-checkbox\">\n            <input type=\"checkbox\" formControlName=\"showTrend\" name=\"showTrend\" />\n            <span></span>\n            <span translate>Show trend icon</span>\n            <button\n              class=\"btn btn-clean\"\n              popover=\"{{\n                'Indicates the trend between the last two measurement values.' | translate\n              }}\"\n              container=\"body\"\n              [outsideClick]=\"true\"\n            >\n              <i [c8yIcon]=\"'question-circle-o'\" class=\"text-info\"></i>\n            </button>\n          </label>\n        </c8y-form-group>\n      </div>\n\n      <c8y-form-group>\n        <label [title]=\"'Font size of measurement value (px)' | translate\" translate>\n          Font size of measurement value (px)\n        </label>\n        <input\n          class=\"form-control\"\n          formControlName=\"fontSize\"\n          name=\"fontSize\"\n          type=\"number\"\n          [placeholder]=\"'e.g. {{ example }}' | translate: { example: 36 }\"\n        />\n        <c8y-messages [show]=\"formGroup.controls?.fontSize?.touched && formGroup?.controls?.fontSize?.errors\">\n        </c8y-messages>\n      </c8y-form-group>\n    </div>\n  </div>\n</form>\n",
                viewProviders: [{ provide: ControlContainer, useExisting: NgForm }]
            }]
    }], function () { return [{ type: ɵngcc1.FormBuilder }, { type: ɵngcc1.NgForm }, { type: ɵngcc2.IconSelectorService }, { type: ɵngcc3.WidgetConfigComponent }, { type: ɵngcc4.DatapointLibraryService }]; }, { config: [{
            type: Input
        }] }); })();
