// @c8y/ngx-components/esm2015/repository/firmware/device-tab/firmware-repository-device-tab.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FirmwareDeviceTabComponent } from './firmware-device-tab.component';
import { FirmwareDeviceTabGuard } from './firmware-device-tab.guard';
import { CoreModule, gettext, HOOK_ROUTE, ViewContext } from '@c8y/ngx-components';
import { SharedRepositoryModule } from '@c8y/ngx-components/repository/shared';
import { OperationDetailsModule } from '@c8y/ngx-components/operations/operation-details';
export class FirmwareRepositoryDeviceTabModule {
    static forRoot() {
        return {
            ngModule: FirmwareRepositoryDeviceTabModule,
            providers: [
                FirmwareDeviceTabGuard,
                {
                    provide: HOOK_ROUTE,
                    useValue: [
                        {
                            context: ViewContext.Device,
                            path: 'firmware',
                            component: FirmwareDeviceTabComponent,
                            label: gettext('Firmware'),
                            icon: 'c8y-firmware',
                            priority: 500,
                            canActivate: [FirmwareDeviceTabGuard]
                        }
                    ],
                    multi: true
                }
            ]
        };
    }
}
FirmwareRepositoryDeviceTabModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, CoreModule, SharedRepositoryModule, OperationDetailsModule],
                declarations: [FirmwareDeviceTabComponent],
                entryComponents: [FirmwareDeviceTabComponent]
            },] }
];
