// @c8y/ngx-components/esm2015/cockpit-config/setup/cockpit-setup-step2.component.js
import { CdkStep } from '@angular/cdk/stepper';
import { Component } from '@angular/core';
import { AlertService, AppStateService, C8yStepper, SetupComponent } from '@c8y/ngx-components';
import { CockpitSetupStep } from './cockpit-setup-step';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from '@angular/cdk/stepper';
import * as ɵngcc3 from '@angular/forms';
import * as ɵngcc4 from '../root-node-config.component';
import * as ɵngcc5 from './cockpit-setup-stepper-buttons.component';
export class CockpitSetupStep2Component extends CockpitSetupStep {
    constructor(stepper, step, setup, appState, alert) {
        super(stepper, step, setup, appState, alert);
        this.stepper = stepper;
        this.step = step;
        this.setup = setup;
        this.appState = appState;
        this.alert = alert;
    }
}
CockpitSetupStep2Component.ɵfac = function CockpitSetupStep2Component_Factory(t) { return new (t || CockpitSetupStep2Component)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.C8yStepper), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.CdkStep), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.SetupComponent), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AppStateService), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AlertService)); };
CockpitSetupStep2Component.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: CockpitSetupStep2Component, selectors: [["c8y-cockpit-setup-step2"]], hostAttrs: [1, "d-contents"], features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 15, vars: 2, consts: [["name", "form", 1, "d-contents"], ["stepForm", "ngForm"], [1, "container-fluid", "flex-no-shrink", "fit-w"], [1, "row", "separator-bottom"], [1, "col-md-8", "col-md-offset-2", "col-lg-6", "col-lg-offset-3", "p-t-24", "p-l-16", "p-r-16"], ["translate", "", 1, "text-medium", "l-h-base"], ["translate", "", 1, "lead", "text-normal"], [1, "inner-scroll", "flex-grow"], [1, "container-fluid", "fit-w", "p-t-16"], [1, "row"], [1, "col-md-8", "col-md-offset-2", "col-lg-6", "col-lg-offset-3"], [3, "config"], [3, "index", "onNext", "onBack"]], template: function CockpitSetupStep2Component_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "form", 0, 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelementStart(3, "div", 3);
        ɵngcc0.ɵɵelementStart(4, "div", 4);
        ɵngcc0.ɵɵelementStart(5, "h3", 5);
        ɵngcc0.ɵɵtext(6, "Top level nodes");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(7, "p", 6);
        ɵngcc0.ɵɵtext(8, " Select which nodes to display in the top level of the navigator menu. By default, only Groups is shown. ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "div", 7);
        ɵngcc0.ɵɵelementStart(10, "div", 8);
        ɵngcc0.ɵɵelementStart(11, "div", 9);
        ɵngcc0.ɵɵelementStart(12, "div", 10);
        ɵngcc0.ɵɵelement(13, "c8y-root-node-config", 11);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(14, "c8y-cockpit-setup-stepper-buttons", 12);
        ɵngcc0.ɵɵlistener("onNext", function CockpitSetupStep2Component_Template_c8y_cockpit_setup_stepper_buttons_onNext_14_listener() { return ctx.next(); })("onBack", function CockpitSetupStep2Component_Template_c8y_cockpit_setup_stepper_buttons_onBack_14_listener() { return ctx.back(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(13);
        ɵngcc0.ɵɵproperty("config", ctx.config);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("index", ctx.stepper.selectedIndex);
    } }, directives: [ɵngcc3.ɵNgNoValidate, ɵngcc3.NgControlStatusGroup, ɵngcc3.NgForm, ɵngcc1.C8yTranslateDirective, ɵngcc4.RootNodeConfigComponent, ɵngcc5.CockpitSetupStepperButtonsComponent], encapsulation: 2 });
CockpitSetupStep2Component.ctorParameters = () => [
    { type: C8yStepper },
    { type: CdkStep },
    { type: SetupComponent },
    { type: AppStateService },
    { type: AlertService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CockpitSetupStep2Component, [{
        type: Component,
        args: [{
                selector: 'c8y-cockpit-setup-step2',
                template: "<form #stepForm=\"ngForm\" name=\"form\" class=\"d-contents\">\n  <div class=\"container-fluid flex-no-shrink fit-w\">\n    <div class=\"row separator-bottom\">\n      <div class=\"col-md-8 col-md-offset-2 col-lg-6 col-lg-offset-3 p-t-24 p-l-16 p-r-16\">\n        <h3 translate class=\"text-medium l-h-base\">Top level nodes</h3>\n        <p class=\"lead text-normal\" translate>\n          Select which nodes to display in the top level of the navigator menu. By default, only\n          Groups is shown.\n        </p>\n      </div>\n    </div>\n  </div>\n  <div class=\"inner-scroll flex-grow\">\n    <div class=\"container-fluid fit-w p-t-16\">\n      <div class=\"row\">\n        <div class=\"col-md-8 col-md-offset-2 col-lg-6 col-lg-offset-3\">\n          <c8y-root-node-config [config]=\"config\"></c8y-root-node-config>\n        </div>\n      </div>\n    </div>\n  </div>\n  <c8y-cockpit-setup-stepper-buttons\n    [index]=\"stepper.selectedIndex\"\n    (onNext)=\"next()\"\n    (onBack)=\"back()\"\n  >\n  </c8y-cockpit-setup-stepper-buttons>\n</form>\n",
                host: { class: 'd-contents' }
            }]
    }], function () { return [{ type: ɵngcc1.C8yStepper }, { type: ɵngcc2.CdkStep }, { type: ɵngcc1.SetupComponent }, { type: ɵngcc1.AppStateService }, { type: ɵngcc1.AlertService }]; }, null); })();
