// @c8y/ngx-components/esm2015/cockpit-config/setup/cockpit-setup-step.js
import { __awaiter } from "tslib";
import { DEFAULT_CONFIG } from '../cockpit-config.model';
export class CockpitSetupStep {
    constructor(stepper, step, setup, appState, alert) {
        this.stepper = stepper;
        this.step = step;
        this.setup = setup;
        this.appState = appState;
        this.alert = alert;
        this.config = DEFAULT_CONFIG;
        this.pending = false;
    }
    next() {
        return __awaiter(this, void 0, void 0, function* () {
            this.pending = true;
            try {
                const newConfig = Object.assign(Object.assign({}, this.setup.data$.value), this.config);
                yield this.appState.updateApplicationConfig(newConfig);
                this.setup.stepCompleted(this.stepper.selectedIndex);
                this.setup.data$.next(newConfig);
                this.stepper.next();
            }
            catch (ex) {
                this.alert.addServerFailure(ex);
            }
            finally {
                this.pending = false;
            }
        });
    }
    back() {
        this.stepper.previous();
    }
}
