// @c8y/ngx-components/esm2015/cockpit-config/setup/cockpit-setup-stepper-buttons.component.js
import { Component, EventEmitter, Input, Output } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '@c8y/ngx-components';

function CockpitSetupStepperButtonsComponent_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r2 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 3);
    ɵngcc0.ɵɵlistener("click", function CockpitSetupStepperButtonsComponent_button_1_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r2); const ctx_r1 = ɵngcc0.ɵɵnextContext(); return ctx_r1.onBack.emit(); });
    ɵngcc0.ɵɵtext(1, " Previous ");
    ɵngcc0.ɵɵelementEnd();
} }
export class CockpitSetupStepperButtonsComponent {
    constructor() {
        this.onNext = new EventEmitter();
        this.onBack = new EventEmitter();
    }
}
CockpitSetupStepperButtonsComponent.ɵfac = function CockpitSetupStepperButtonsComponent_Factory(t) { return new (t || CockpitSetupStepperButtonsComponent)(); };
CockpitSetupStepperButtonsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: CockpitSetupStepperButtonsComponent, selectors: [["c8y-cockpit-setup-stepper-buttons"]], inputs: { index: "index" }, outputs: { onNext: "onNext", onBack: "onBack" }, decls: 4, vars: 1, consts: [[1, "card-footer", "separator", "d-flex", "j-c-center"], ["class", "btn btn-default", "type", "button", "translate", "", 3, "click", 4, "ngIf"], ["type", "submit", "translate", "", 1, "btn", "btn-primary", 3, "click"], ["type", "button", "translate", "", 1, "btn", "btn-default", 3, "click"]], template: function CockpitSetupStepperButtonsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, CockpitSetupStepperButtonsComponent_button_1_Template, 2, 0, "button", 1);
        ɵngcc0.ɵɵelementStart(2, "button", 2);
        ɵngcc0.ɵɵlistener("click", function CockpitSetupStepperButtonsComponent_Template_button_click_2_listener() { return ctx.onNext.emit(); });
        ɵngcc0.ɵɵtext(3, " Save and continue ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.index !== 0);
    } }, directives: [ɵngcc1.NgIf, ɵngcc2.C8yTranslateDirective], encapsulation: 2 });
CockpitSetupStepperButtonsComponent.propDecorators = {
    index: [{ type: Input }],
    onNext: [{ type: Output }],
    onBack: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CockpitSetupStepperButtonsComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-cockpit-setup-stepper-buttons',
                template: "<div class=\"card-footer separator d-flex j-c-center\">\n  <button\n    class=\"btn btn-default\"\n    type=\"button\"\n    (click)=\"onBack.emit()\"\n    *ngIf=\"index !== 0\"\n    translate\n  >\n    Previous\n  </button>\n  <button class=\"btn btn-primary\" type=\"submit\" (click)=\"onNext.emit()\" translate>\n    Save and continue\n  </button>\n</div>\n"
            }]
    }], function () { return []; }, { onNext: [{
            type: Output
        }], onBack: [{
            type: Output
        }], index: [{
            type: Input
        }] }); })();
