// @c8y/ngx-components/esm2015/register-device/dropdown/register-device-dropdown.component.js
import { Component } from '@angular/core';
import { map } from 'rxjs/operators';
import { RegisterDeviceService } from '../register-device.service';
import { RegisterDeviceExtensionService } from './register-device-extension.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './register-device-extension.service';
import * as ɵngcc2 from '../register-device.service';
import * as ɵngcc3 from 'ngx-bootstrap/dropdown';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from '@c8y/ngx-components';

function RegisterDeviceDropdownComponent_button_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "button", 7);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelementStart(2, "span", 8);
    ɵngcc0.ɵɵtext(3, "Register device");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(4, "i", 9);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, "Register device"));
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("c8yIcon", "caret-down");
} }
function RegisterDeviceDropdownComponent_ng_template_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "button", 10);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelementStart(2, "span", 8);
    ɵngcc0.ɵɵtext(3, "Register device");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(4, "i", 9);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, "Device registration disabled"));
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("c8yIcon", "caret-down");
} }
function RegisterDeviceDropdownComponent_ul_5_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function RegisterDeviceDropdownComponent_ul_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "ul", 11);
    ɵngcc0.ɵɵtemplate(1, RegisterDeviceDropdownComponent_ul_5_ng_container_1_Template, 1, 0, "ng-container", 5);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵnextContext();
    const _r5 = ɵngcc0.ɵɵreference(9);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r5);
} }
function RegisterDeviceDropdownComponent_ng_container_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function RegisterDeviceDropdownComponent_ng_template_8_ng_container_0_li_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 15);
    ɵngcc0.ɵɵtext(1, "Single registration");
    ɵngcc0.ɵɵelementEnd();
} }
function RegisterDeviceDropdownComponent_ng_template_8_ng_container_0_li_2_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function RegisterDeviceDropdownComponent_ng_template_8_ng_container_0_li_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li");
    ɵngcc0.ɵɵtemplate(1, RegisterDeviceDropdownComponent_ng_template_8_ng_container_0_li_2_ng_container_1_Template, 1, 0, "ng-container", 16);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r13 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yOutlet", item_r13.template);
} }
function RegisterDeviceDropdownComponent_ng_template_8_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, RegisterDeviceDropdownComponent_ng_template_8_ng_container_0_li_1_Template, 2, 0, "li", 13);
    ɵngcc0.ɵɵtemplate(2, RegisterDeviceDropdownComponent_ng_template_8_ng_container_0_li_2_Template, 2, 1, "li", 14);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const single_r10 = ctx.ngIf;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", single_r10.length > 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", single_r10);
} }
function RegisterDeviceDropdownComponent_ng_template_8_ng_container_2_li_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 15);
    ɵngcc0.ɵɵtext(1, "Bulk registration");
    ɵngcc0.ɵɵelementEnd();
} }
function RegisterDeviceDropdownComponent_ng_template_8_ng_container_2_li_2_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function RegisterDeviceDropdownComponent_ng_template_8_ng_container_2_li_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li");
    ɵngcc0.ɵɵtemplate(1, RegisterDeviceDropdownComponent_ng_template_8_ng_container_2_li_2_ng_container_1_Template, 1, 0, "ng-container", 16);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r18 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yOutlet", item_r18.template);
} }
function RegisterDeviceDropdownComponent_ng_template_8_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, RegisterDeviceDropdownComponent_ng_template_8_ng_container_2_li_1_Template, 2, 0, "li", 13);
    ɵngcc0.ɵɵtemplate(2, RegisterDeviceDropdownComponent_ng_template_8_ng_container_2_li_2_Template, 2, 1, "li", 14);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const bulk_r15 = ctx.ngIf;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", bulk_r15.length > 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", bulk_r15);
} }
function RegisterDeviceDropdownComponent_ng_template_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, RegisterDeviceDropdownComponent_ng_template_8_ng_container_0_Template, 3, 2, "ng-container", 12);
    ɵngcc0.ɵɵpipe(1, "async");
    ɵngcc0.ɵɵtemplate(2, RegisterDeviceDropdownComponent_ng_template_8_ng_container_2_Template, 3, 2, "ng-container", 12);
    ɵngcc0.ɵɵpipe(3, "async");
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(1, 2, ctx_r6.single$));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(3, 4, ctx_r6.bulk$));
} }
export class RegisterDeviceDropdownComponent {
    constructor(registerDeviceExtensionService, registerDeviceService) {
        this.registerDeviceExtensionService = registerDeviceExtensionService;
        this.registerDeviceService = registerDeviceService;
        this.single$ = this.registerDeviceExtensionService.items$.pipe(map(items => items.filter(item => item.category === 'single').sort((a, b) => b.priority - a.priority)));
        this.bulk$ = this.registerDeviceExtensionService.items$.pipe(map(items => items.filter(item => item.category === 'bulk').sort((a, b) => b.priority - a.priority)));
        this.limit$ = this.registerDeviceService.limit$.pipe(map(limit => limit.isReached));
    }
}
RegisterDeviceDropdownComponent.ɵfac = function RegisterDeviceDropdownComponent_Factory(t) { return new (t || RegisterDeviceDropdownComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.RegisterDeviceExtensionService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.RegisterDeviceService)); };
RegisterDeviceDropdownComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: RegisterDeviceDropdownComponent, selectors: [["c8y-register-device-dropdown"]], decls: 10, vars: 5, consts: [["dropdown", "", 1, "dropdown"], ["class", "dropdown-toggle c8y-dropdown d-flex", "dropdownToggle", "", "aria-haspopup", "true", "aria-expanded", "false", 3, "title", 4, "ngIf", "ngIfElse"], ["disable", ""], ["class", "dropdown-menu dropdown-menu-right hidden-xs", 4, "dropdownMenu"], [1, "dropdown-menu", "dropdown-menu", "visible-xs"], [4, "ngTemplateOutlet"], ["dropdown", ""], ["dropdownToggle", "", "aria-haspopup", "true", "aria-expanded", "false", 1, "dropdown-toggle", "c8y-dropdown", "d-flex", 3, "title"], ["translate", "", 1, "text-truncate"], [3, "c8yIcon"], ["disabled", "", 1, "btn", "btn-clean", "d-flex", "p-l-8", 3, "title"], [1, "dropdown-menu", "dropdown-menu-right", "hidden-xs"], [4, "ngIf"], ["class", "dropdown-header", "translate", "", 4, "ngIf"], [4, "ngFor", "ngForOf"], ["translate", "", 1, "dropdown-header"], [4, "c8yOutlet"]], template: function RegisterDeviceDropdownComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, RegisterDeviceDropdownComponent_button_1_Template, 5, 4, "button", 1);
        ɵngcc0.ɵɵpipe(2, "async");
        ɵngcc0.ɵɵtemplate(3, RegisterDeviceDropdownComponent_ng_template_3_Template, 5, 4, "ng-template", null, 2, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(5, RegisterDeviceDropdownComponent_ul_5_Template, 2, 1, "ul", 3);
        ɵngcc0.ɵɵelementStart(6, "ul", 4);
        ɵngcc0.ɵɵtemplate(7, RegisterDeviceDropdownComponent_ng_container_7_Template, 1, 0, "ng-container", 5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(8, RegisterDeviceDropdownComponent_ng_template_8_Template, 4, 6, "ng-template", null, 6, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r1 = ɵngcc0.ɵɵreference(4);
        const _r5 = ɵngcc0.ɵɵreference(9);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ɵngcc0.ɵɵpipeBind1(2, 3, ctx.limit$))("ngIfElse", _r1);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r5);
    } }, directives: [ɵngcc3.BsDropdownDirective, ɵngcc4.NgIf, ɵngcc3.BsDropdownMenuDirective, ɵngcc4.NgTemplateOutlet, ɵngcc3.BsDropdownToggleDirective, ɵngcc5.C8yTranslateDirective, ɵngcc5.IconDirective, ɵngcc4.NgForOf, ɵngcc5.OutletDirective], pipes: [ɵngcc4.AsyncPipe, ɵngcc5.C8yTranslatePipe], encapsulation: 2 });
RegisterDeviceDropdownComponent.ctorParameters = () => [
    { type: RegisterDeviceExtensionService },
    { type: RegisterDeviceService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(RegisterDeviceDropdownComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-register-device-dropdown',
                template: "<div class=\"dropdown\" dropdown>\n  <button\n    *ngIf=\"!(limit$ | async); else disable\"\n    title=\"{{ 'Register device' | translate }}\"\n    class=\"dropdown-toggle c8y-dropdown d-flex\"\n    dropdownToggle\n    aria-haspopup=\"true\"\n    aria-expanded=\"false\"\n  >\n    <span class=\"text-truncate\" translate>Register device</span>\n    <i [c8yIcon]=\"'caret-down'\"></i>\n  </button>\n  <ng-template #disable>\n    <button\n      title=\"{{ 'Device registration disabled' | translate }}\"\n      class=\"btn btn-clean d-flex p-l-8\"\n      disabled\n    >\n      <span class=\"text-truncate\" translate>Register device</span>\n      <i [c8yIcon]=\"'caret-down'\"></i>\n    </button>\n  </ng-template>\n\n  <!-- dropdown for normal screen sizes -->\n  <ul class=\"dropdown-menu dropdown-menu-right hidden-xs\" *dropdownMenu>\n    <ng-container *ngTemplateOutlet=\"dropdown\"></ng-container>\n  </ul>\n\n  <!-- fake dropdown for mobile screen sizes. *dropdownMenu is missing by design! -->\n  <ul class=\"dropdown-menu dropdown-menu visible-xs\">\n    <ng-container *ngTemplateOutlet=\"dropdown\"></ng-container>\n  </ul>\n\n  <ng-template #dropdown>\n    <ng-container *ngIf=\"single$ | async as single\">\n      <li class=\"dropdown-header\" *ngIf=\"single.length > 0\" translate>Single registration</li>\n      <li *ngFor=\"let item of single\">\n        <ng-container *c8yOutlet=\"item.template\"></ng-container>\n      </li>\n    </ng-container>\n    <ng-container *ngIf=\"bulk$ | async as bulk\">\n      <li class=\"dropdown-header\" *ngIf=\"bulk.length > 0\" translate>Bulk registration</li>\n      <li *ngFor=\"let item of bulk\">\n        <ng-container *c8yOutlet=\"item.template\"></ng-container>\n      </li>\n    </ng-container>\n  </ng-template>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.RegisterDeviceExtensionService }, { type: ɵngcc2.RegisterDeviceService }]; }, null); })();
