// @c8y/ngx-components/esm2015/datapoint-selector/datapoint-selector-list-item/datapoint-selector-list-item.component.js
import { Component, ContentChild, EventEmitter, forwardRef, Input, Output } from '@angular/core';
import { FormBuilder, NG_VALIDATORS, NG_VALUE_ACCESSOR } from '@angular/forms';
import { pipe } from 'rxjs';
import { map, startWith, take } from 'rxjs/operators';
import { ListItemDragHandleComponent } from '@c8y/ngx-components';
import { DATAPOINT_LIBRARY_FRAGMENT } from '../datapoint-selection.model';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/forms';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from 'ngx-bootstrap/popover';
import * as ɵngcc5 from '../datapoint-template-popover/datapoint-template-popover.component';
import * as ɵngcc6 from '../datapoint-attributes-form/datapoint-attributes-form.component';

function DatapointSelectorListItemComponent_c8y_li_icon_4_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r14 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 26);
    ɵngcc0.ɵɵlistener("click", function DatapointSelectorListItemComponent_c8y_li_icon_4_button_1_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r14); const ctx_r13 = ɵngcc0.ɵɵnextContext(2); return ctx_r13.addOrRemoveItem(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 27);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(1, 1, "Remove"));
} }
function DatapointSelectorListItemComponent_c8y_li_icon_4_button_2_Template(rf, ctx) { if (rf & 1) {
    const _r16 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 26);
    ɵngcc0.ɵɵlistener("click", function DatapointSelectorListItemComponent_c8y_li_icon_4_button_2_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r16); const ctx_r15 = ɵngcc0.ɵɵnextContext(2); return ctx_r15.addOrRemoveItem(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 28);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(1, 1, "Select"));
} }
function DatapointSelectorListItemComponent_c8y_li_icon_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li-icon", 24);
    ɵngcc0.ɵɵtemplate(1, DatapointSelectorListItemComponent_c8y_li_icon_4_button_1_Template, 3, 3, "button", 25);
    ɵngcc0.ɵɵtemplate(2, DatapointSelectorListItemComponent_c8y_li_icon_4_button_2_Template, 3, 3, "button", 25);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.isSelected);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r1.isSelected);
} }
function DatapointSelectorListItemComponent_c8y_li_checkbox_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li-checkbox", 29);
    ɵngcc0.ɵɵlistener("click", function DatapointSelectorListItemComponent_c8y_li_checkbox_5_Template_c8y_li_checkbox_click_0_listener($event) { return $event.stopPropagation(); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("displayAsSwitch", true);
} }
function DatapointSelectorListItemComponent_small_15_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 30);
    ɵngcc0.ɵɵelement(1, "i", 31);
    ɵngcc0.ɵɵelementStart(2, "span", 9);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ctx_r3.formGroup.value.__target.name);
} }
function DatapointSelectorListItemComponent_button_16_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "button", 32);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 33);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("popover", ɵngcc0.ɵɵpipeBind1(1, 2, "Some entries are invalid. Check the input fields with red borders."))("outsideClick", true);
} }
function DatapointSelectorListItemComponent_c8y_li_action_18_Template(rf, ctx) { if (rf & 1) {
    const _r19 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li-action", 34);
    ɵngcc0.ɵɵlistener("click", function DatapointSelectorListItemComponent_c8y_li_action_18_Template_c8y_li_action_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r19); const ctx_r18 = ɵngcc0.ɵɵnextContext(); return ctx_r18.remove(ctx_r18.formGroup.value); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("icon", "minus-circle")("label", ɵngcc0.ɵɵpipeBind1(1, 2, "Remove from list"));
} }
function DatapointSelectorListItemComponent_c8y_li_action_19_Template(rf, ctx) { if (rf & 1) {
    const _r22 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li-action", 34);
    ɵngcc0.ɵɵlistener("click", function DatapointSelectorListItemComponent_c8y_li_action_19_Template_c8y_li_action_click_0_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r22); const action_r20 = restoredCtx.$implicit; const ctx_r21 = ɵngcc0.ɵɵnextContext(); return action_r20.callback(ctx_r21.formGroup.value); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const action_r20 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("icon", action_r20.icon)("label", action_r20.label);
} }
function DatapointSelectorListItemComponent_div_33_ng_template_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtext(0);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(1, 1, "Using a data point template sets color, label, unit and ranges. Removing the template allows you to set the values manually."), " ");
} }
function DatapointSelectorListItemComponent_div_33_c8y_typeahead_8_c8y_li_2_Template(rf, ctx) { if (rf & 1) {
    const _r34 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 47);
    ɵngcc0.ɵɵlistener("click", function DatapointSelectorListItemComponent_div_33_c8y_typeahead_8_c8y_li_2_Template_c8y_li_click_0_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r34); const item_r32 = restoredCtx.$implicit; const ctx_r33 = ɵngcc0.ɵɵnextContext(3); return ctx_r33.dataPointTemplateSelected(item_r32); });
    ɵngcc0.ɵɵelement(1, "c8y-highlight", 48);
    ɵngcc0.ɵɵelement(2, "c8y-li-icon", 49);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r32 = ctx.$implicit;
    const ctx_r29 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("active", ctx_r29.formGroup.value.__template === item_r32.id);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("text", item_r32.c8y_Kpi == null ? null : item_r32.c8y_Kpi.label)("pattern", ctx_r29.pattern);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵstyleProp("color", item_r32.c8y_Kpi == null ? null : item_r32.c8y_Kpi.color);
} }
function DatapointSelectorListItemComponent_div_33_c8y_typeahead_8_ng_template_3_c8y_li_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 51);
    ɵngcc0.ɵɵelementStart(1, "span");
    ɵngcc0.ɵɵtext(2, "No match found.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function DatapointSelectorListItemComponent_div_33_c8y_typeahead_8_ng_template_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, DatapointSelectorListItemComponent_div_33_c8y_typeahead_8_ng_template_3_c8y_li_0_Template, 3, 0, "c8y-li", 50);
} if (rf & 2) {
    const ctx_r31 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r31.pattern.length > 0);
} }
const _c0 = function () { return { standalone: true }; };
function DatapointSelectorListItemComponent_div_33_c8y_typeahead_8_Template(rf, ctx) { if (rf & 1) {
    const _r37 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-typeahead", 44);
    ɵngcc0.ɵɵlistener("onSearch", function DatapointSelectorListItemComponent_div_33_c8y_typeahead_8_Template_c8y_typeahead_onSearch_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r37); const ctx_r36 = ɵngcc0.ɵɵnextContext(2); return ctx_r36.setPipe($event); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtemplate(2, DatapointSelectorListItemComponent_div_33_c8y_typeahead_8_c8y_li_2_Template, 3, 5, "c8y-li", 45);
    ɵngcc0.ɵɵtemplate(3, DatapointSelectorListItemComponent_div_33_c8y_typeahead_8_ng_template_3_Template, 1, 1, "ng-template", null, 46, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _r30 = ɵngcc0.ɵɵreference(4);
    const ctx_r26 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("placeholder", ɵngcc0.ɵɵpipeBind1(1, 10, "No template"))("ngModel", ctx_r26.formGroup.value.__template ? ctx_r26.formGroup.value.details : undefined)("ngModelOptions", ɵngcc0.ɵɵpureFunction0(12, _c0))("displayProperty", "label")("hideNew", true)("disabled", ctx_r26.isSelected && ctx_r26.disableTypeaheadIfSelected);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yForOf", ctx_r26.datapointLibraryEntries)("c8yForLoadMore", "auto")("c8yForPipe", ctx_r26.filterPipe)("c8yForNotFound", _r30);
} }
function DatapointSelectorListItemComponent_div_33_button_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "button", 52);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 53);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵnextContext(2);
    const _r8 = ɵngcc0.ɵɵreference(36);
    ɵngcc0.ɵɵproperty("popover", _r8)("outsideClick", true)("title", ɵngcc0.ɵɵpipeBind1(1, 3, "Info"));
} }
function DatapointSelectorListItemComponent_div_33_button_10_Template(rf, ctx) { if (rf & 1) {
    const _r39 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 54);
    ɵngcc0.ɵɵlistener("click", function DatapointSelectorListItemComponent_div_33_button_10_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r39); const ctx_r38 = ɵngcc0.ɵɵnextContext(2); return ctx_r38.unlinkDatapointTemplate(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 55);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(1, 1, "Unlink data point template"));
} }
function DatapointSelectorListItemComponent_div_33_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 35);
    ɵngcc0.ɵɵelementStart(1, "label", 36);
    ɵngcc0.ɵɵtext(2, "Data point template");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "button", 37);
    ɵngcc0.ɵɵlistener("click", function DatapointSelectorListItemComponent_div_33_Template_button_click_3_listener($event) { return $event.stopPropagation(); });
    ɵngcc0.ɵɵelement(4, "i", 38);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(5, DatapointSelectorListItemComponent_div_33_ng_template_5_Template, 2, 3, "ng-template", null, 39, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementStart(7, "div", 40);
    ɵngcc0.ɵɵtemplate(8, DatapointSelectorListItemComponent_div_33_c8y_typeahead_8_Template, 5, 13, "c8y-typeahead", 41);
    ɵngcc0.ɵɵtemplate(9, DatapointSelectorListItemComponent_div_33_button_9_Template, 3, 5, "button", 42);
    ɵngcc0.ɵɵtemplate(10, DatapointSelectorListItemComponent_div_33_button_10_Template, 3, 3, "button", 43);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _r24 = ɵngcc0.ɵɵreference(6);
    const ctx_r7 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("popover", _r24)("outsideClick", true);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yIcon", "question-circle-o");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r7.datapointLibraryEntries);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r7.formGroup.value.__template);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r7.formGroup.value.__template && ctx_r7.hasUnlinkTemplateOption);
} }
function DatapointSelectorListItemComponent_ng_template_35_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-datapoint-template-popover", 56);
} if (rf & 2) {
    const ctx_r9 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("datapoint", ctx_r9.formGroup.value.details);
} }
function DatapointSelectorListItemComponent_c8y_datapoint_attributes_form_37_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-datapoint-attributes-form", 57);
} if (rf & 2) {
    const ctx_r10 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("showTarget", ctx_r10.defaultFormOptions.showTarget)("showRange", ctx_r10.defaultFormOptions.showRange)("showYellowRange", ctx_r10.defaultFormOptions.showYellowRange)("showRedRange", ctx_r10.defaultFormOptions.showRedRange)("showChart", ctx_r10.defaultFormOptions.showChart)("showFormIfTemplateWasSelected", ctx_r10.defaultFormOptions.showFormIfTemplateWasSelected);
} }
const _c1 = [[["c8y-li-drag-handle"]]];
const _c2 = ["c8y-li-drag-handle"];
export class DatapointSelectorListItemComponent {
    constructor(formBuilder) {
        this.formBuilder = formBuilder;
        this.defaultFormOptions = {};
        this.isSelected = false;
        this.isCollapsed = true;
        this.showAddRemoveButton = true;
        this.editable = true;
        this.showActiveToggle = false;
        this.activeToggleDisabled = false;
        this.showOptions = false;
        this.actions = [];
        this.optionToRemove = false;
        this.hasUnlinkTemplateOption = false;
        this.added = new EventEmitter();
        this.removed = new EventEmitter();
        this.colorPickerDisabled = true;
        this.disableTypeaheadIfSelected = false;
        this.pattern = '';
        this.formGroup = this.formBuilder.group({
            details: [],
            color: [],
            __active: [],
            __target: [],
            fragment: [],
            series: [],
            __template: []
        });
        this.isValid$ = this.formGroup.statusChanges.pipe(map((status) => status === 'VALID'), startWith(this.formGroup.valid));
    }
    validate(control) {
        var _a;
        return ((_a = this.formGroup) === null || _a === void 0 ? void 0 : _a.valid) ? null : { formInvalid: {} };
    }
    writeValue(obj) {
        this.formGroup.patchValue(Object.assign(Object.assign({}, obj), { details: obj }));
    }
    registerOnChange(fn) {
        this.formGroup.valueChanges.pipe(map(tmp => this.transformFormValue(tmp))).subscribe(fn);
    }
    registerOnTouched(fn) {
        this.formGroup.valueChanges.pipe(take(1)).subscribe(fn);
    }
    setDisabledState(isDisabled) {
        isDisabled ? this.formGroup.disable() : this.formGroup.enable();
    }
    collapse() {
        this.isCollapsed = !this.isCollapsed;
    }
    addOrRemoveItem() {
        const value = this.transformFormValue(this.formGroup.value);
        if (this.isSelected) {
            this.removed.emit(value);
        }
        else {
            this.added.emit(value);
        }
    }
    remove(dp) {
        this.removed.emit(this.transformFormValue(this.formGroup.value));
    }
    setPipe(filterStr) {
        this.pattern = filterStr;
        this.filterPipe = pipe(map((data) => {
            return this.filterDatapointLabel(data, filterStr);
        }));
    }
    unlinkDatapointTemplate() {
        const details = this.formGroup.value.details || {};
        this.resetUnusedProperties(details);
        this.formGroup.patchValue({ __template: undefined, details });
    }
    dataPointTemplateSelected(template) {
        const attributesToAssign = [
            'color',
            'label',
            'min',
            'max',
            'yellowRangeMax',
            'yellowRangeMin',
            'redRangeMax',
            'redRangeMin',
            'target',
            'orientation',
            'unit'
        ];
        const { fragment, series, __target, __active } = this.formGroup.value;
        const dataPoint = {
            fragment,
            series,
            __active,
            __target,
            __template: template.id
        };
        for (const attribute of attributesToAssign) {
            const value = template[DATAPOINT_LIBRARY_FRAGMENT][attribute];
            dataPoint[attribute] = value;
        }
        this.writeValue(dataPoint);
        this.setPipe('');
    }
    resetUnusedProperties(details) {
        const { showTarget, showYellowRange, showRedRange } = this.defaultFormOptions;
        details.__template = undefined;
        if (!showTarget) {
            details.target = undefined;
        }
        if (!showYellowRange) {
            details.yellowRangeMin = undefined;
            details.yellowRangeMax = undefined;
        }
        if (!showRedRange) {
            details.redRangeMin = undefined;
            details.redRangeMax = undefined;
        }
    }
    filterDatapointLabel(kpis, filterStr) {
        return kpis.filter((mo) => mo[DATAPOINT_LIBRARY_FRAGMENT] &&
            mo[DATAPOINT_LIBRARY_FRAGMENT].label &&
            typeof mo[DATAPOINT_LIBRARY_FRAGMENT].label === 'string' &&
            mo[DATAPOINT_LIBRARY_FRAGMENT].label.toLowerCase().indexOf(filterStr.toLowerCase()) > -1);
    }
    transformFormValue(formValue) {
        const obj = Object.assign({}, formValue.details || {}, formValue);
        delete obj.details;
        return obj;
    }
}
DatapointSelectorListItemComponent.ɵfac = function DatapointSelectorListItemComponent_Factory(t) { return new (t || DatapointSelectorListItemComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.FormBuilder)); };
DatapointSelectorListItemComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: DatapointSelectorListItemComponent, selectors: [["c8y-datapoint-selector-list-item"]], contentQueries: function DatapointSelectorListItemComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, ListItemDragHandleComponent, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.dragHandle = _t.first);
    } }, inputs: { defaultFormOptions: "defaultFormOptions", isSelected: "isSelected", isCollapsed: "isCollapsed", showAddRemoveButton: "showAddRemoveButton", editable: "editable", showActiveToggle: "showActiveToggle", activeToggleDisabled: "activeToggleDisabled", showOptions: "showOptions", actions: "actions", optionToRemove: "optionToRemove", hasUnlinkTemplateOption: "hasUnlinkTemplateOption", colorPickerDisabled: "colorPickerDisabled", disableTypeaheadIfSelected: "disableTypeaheadIfSelected", datapointLibraryEntries: "datapointLibraryEntries", highlightText: "highlightText" }, outputs: { added: "added", removed: "removed" }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: NG_VALUE_ACCESSOR,
                useExisting: forwardRef(() => DatapointSelectorListItemComponent),
                multi: true
            },
            {
                provide: NG_VALIDATORS,
                useExisting: forwardRef(() => DatapointSelectorListItemComponent),
                multi: true
            }
        ])], ngContentSelectors: _c2, decls: 38, vars: 28, consts: [[1, "c8y-list__item__collapse--container-small", 3, "formGroup"], ["li", ""], ["class", "a-s-center p-r-4", 4, "ngIf"], ["formControlName", "__active", "class", "a-s-center p-r-0", 3, "displayAsSwitch", "click", 4, "ngIf"], [1, "d-flex", "a-i-center", "p-l-4"], [1, "c8y-list__item__colorpicker", "p-t-0", "p-b-0", "p-l-0", 3, "title"], [1, "c8y-colorpicker"], ["type", "color", "formControlName", "color", 3, "click"], [1, "btn-clean", "data-point-label", "text-truncate", 3, "title", "click"], [1, "text-truncate"], [3, "text", "pattern", "shouldTrimPattern"], ["class", "text-truncate text-muted icon-flex", 4, "ngIf"], ["class", "btn-clean m-l-auto", "container", "body", 3, "popover", "outsideClick", 4, "ngIf"], [3, "icon", "label", "click", 4, "ngIf"], [3, "icon", "label", "click", 4, "ngFor", "ngForOf"], [1, "data-point-details"], [1, "list-unstyled", "small", "m-b-16"], [1, "p-t-4", "p-b-4", "flex-row", "separator-top-bottom"], ["translate", "", 1, "small", "m-b-0", "m-r-8", "flex-item-v-start", "text-muted"], [1, "flex-item-right"], [1, "p-t-4", "p-b-4", "flex-row", "separator-bottom"], ["class", "form-group", 4, "ngIf"], ["datapointOverviewPopoverTemplate", ""], ["formControlName", "details", 3, "showTarget", "showRange", "showYellowRange", "showRedRange", "showChart", "showFormIfTemplateWasSelected", 4, "ngIf"], [1, "a-s-center", "p-r-4"], ["class", "btn btn-clean text-primary", 3, "title", "click", 4, "ngIf"], [1, "btn", "btn-clean", "text-primary", 3, "title", "click"], ["c8yIcon", "minus-circle", 1, "text-danger"], ["c8yIcon", "plus-circle", 1, "text-primary"], ["formControlName", "__active", 1, "a-s-center", "p-r-0", 3, "displayAsSwitch", "click"], [1, "text-truncate", "text-muted", "icon-flex"], ["c8yIcon", "exchange"], ["container", "body", 1, "btn-clean", "m-l-auto", 3, "popover", "outsideClick"], ["c8yIcon", "warning", 1, "text-danger"], [3, "icon", "label", "click"], [1, "form-group"], ["translate", "", 1, "d-inline-block"], ["container", "body", 1, "btn", "btn-clean", 3, "popover", "outsideClick", "click"], [1, "m-l-4", "text-info", 3, "c8yIcon"], ["datapointHintPopoverTemplate", ""], [1, "d-flex"], ["class", "flex-grow", 3, "placeholder", "ngModel", "ngModelOptions", "displayProperty", "hideNew", "disabled", "onSearch", 4, "ngIf"], ["class", "btn btn-info btn-xs btn-icon a-s-center m-l-8", "placement", "left", "container", "body", 3, "popover", "outsideClick", "title", 4, "ngIf"], ["class", "btn btn-danger btn-xs btn-icon a-s-center m-l-8", 3, "title", "click", 4, "ngIf"], [1, "flex-grow", 3, "placeholder", "ngModel", "ngModelOptions", "displayProperty", "hideNew", "disabled", "onSearch"], ["class", "p-l-8 p-r-8 c8y-list__item--link interact", 3, "active", "click", 4, "c8yFor", "c8yForOf", "c8yForLoadMore", "c8yForPipe", "c8yForNotFound"], ["notFoundTemplate", ""], [1, "p-l-8", "p-r-8", "c8y-list__item--link", "interact", 3, "active", "click"], [3, "text", "pattern"], ["icon", "circle"], ["class", "bg-gray-lighter p-8", 4, "ngIf"], [1, "bg-gray-lighter", "p-8"], ["placement", "left", "container", "body", 1, "btn", "btn-info", "btn-xs", "btn-icon", "a-s-center", "m-l-8", 3, "popover", "outsideClick", "title"], ["c8yIcon", "info", 1, "text-info"], [1, "btn", "btn-danger", "btn-xs", "btn-icon", "a-s-center", "m-l-8", 3, "title", "click"], ["c8yIcon", "unlink"], [3, "datapoint"], ["formControlName", "details", 3, "showTarget", "showRange", "showYellowRange", "showRedRange", "showChart", "showFormIfTemplateWasSelected"]], template: function DatapointSelectorListItemComponent_Template(rf, ctx) { if (rf & 1) {
        const _r41 = ɵngcc0.ɵɵgetCurrentView();
        ɵngcc0.ɵɵprojectionDef(_c1);
        ɵngcc0.ɵɵelementStart(0, "c8y-li", 0, 1);
        ɵngcc0.ɵɵelementStart(2, "c8y-li-drag-handle");
        ɵngcc0.ɵɵprojection(3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(4, DatapointSelectorListItemComponent_c8y_li_icon_4_Template, 3, 2, "c8y-li-icon", 2);
        ɵngcc0.ɵɵtemplate(5, DatapointSelectorListItemComponent_c8y_li_checkbox_5_Template, 1, 1, "c8y-li-checkbox", 3);
        ɵngcc0.ɵɵelementStart(6, "div", 4);
        ɵngcc0.ɵɵelementStart(7, "div", 5);
        ɵngcc0.ɵɵpipe(8, "translate");
        ɵngcc0.ɵɵelementStart(9, "div", 6);
        ɵngcc0.ɵɵelementStart(10, "input", 7);
        ɵngcc0.ɵɵlistener("click", function DatapointSelectorListItemComponent_Template_input_click_10_listener($event) { return $event.stopPropagation(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(11, "span");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(12, "button", 8);
        ɵngcc0.ɵɵlistener("click", function DatapointSelectorListItemComponent_Template_button_click_12_listener() { ɵngcc0.ɵɵrestoreView(_r41); const _r0 = ɵngcc0.ɵɵreference(1); return _r0.collapsed = !_r0.collapsed; });
        ɵngcc0.ɵɵelementStart(13, "span", 9);
        ɵngcc0.ɵɵelement(14, "c8y-highlight", 10);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(15, DatapointSelectorListItemComponent_small_15_Template, 4, 1, "small", 11);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(16, DatapointSelectorListItemComponent_button_16_Template, 3, 4, "button", 12);
        ɵngcc0.ɵɵpipe(17, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(18, DatapointSelectorListItemComponent_c8y_li_action_18_Template, 2, 4, "c8y-li-action", 13);
        ɵngcc0.ɵɵtemplate(19, DatapointSelectorListItemComponent_c8y_li_action_19_Template, 1, 2, "c8y-li-action", 14);
        ɵngcc0.ɵɵelementStart(20, "c8y-li-collapse");
        ɵngcc0.ɵɵelementStart(21, "div", 15);
        ɵngcc0.ɵɵelementStart(22, "ul", 16);
        ɵngcc0.ɵɵelementStart(23, "li", 17);
        ɵngcc0.ɵɵelementStart(24, "label", 18);
        ɵngcc0.ɵɵtext(25, "Fragment");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(26, "span", 19);
        ɵngcc0.ɵɵelement(27, "c8y-highlight", 10);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(28, "li", 20);
        ɵngcc0.ɵɵelementStart(29, "label", 18);
        ɵngcc0.ɵɵtext(30, "Series");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(31, "span", 19);
        ɵngcc0.ɵɵelement(32, "c8y-highlight", 10);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(33, DatapointSelectorListItemComponent_div_33_Template, 11, 6, "div", 21);
        ɵngcc0.ɵɵpipe(34, "async");
        ɵngcc0.ɵɵtemplate(35, DatapointSelectorListItemComponent_ng_template_35_Template, 1, 1, "ng-template", null, 22, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(37, DatapointSelectorListItemComponent_c8y_datapoint_attributes_form_37_Template, 1, 6, "c8y-datapoint-attributes-form", 23);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("formGroup", ctx.formGroup);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showAddRemoveButton);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showActiveToggle);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(8, 22, "Change color"));
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵstyleProp("background-color", ctx.formGroup.value.color);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("title", ctx.formGroup.value.details == null ? null : ctx.formGroup.value.details.label);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("text", ctx.formGroup.value.details == null ? null : ctx.formGroup.value.details.label)("pattern", ctx.highlightText)("shouldTrimPattern", true);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.formGroup.value.__target);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ɵngcc0.ɵɵpipeBind1(17, 24, ctx.isValid$));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.optionToRemove);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.actions);
        ɵngcc0.ɵɵadvance(8);
        ɵngcc0.ɵɵproperty("text", ctx.formGroup.value.fragment)("pattern", ctx.highlightText)("shouldTrimPattern", true);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("text", ctx.formGroup.value.series)("pattern", ctx.highlightText)("shouldTrimPattern", true);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(34, 26, ctx.datapointLibraryEntries && ctx.datapointLibraryEntries));
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngIf", ctx.defaultFormOptions);
    } }, directives: [ɵngcc2.ListItemComponent, ɵngcc1.NgControlStatusGroup, ɵngcc1.FormGroupDirective, ɵngcc2.ListItemDragHandleComponent, ɵngcc3.NgIf, ɵngcc1.DefaultValueAccessor, ɵngcc2.RequiredInputPlaceholderDirective, ɵngcc1.NgControlStatus, ɵngcc1.FormControlName, ɵngcc2.HighlightComponent, ɵngcc3.NgForOf, ɵngcc2.ListItemCollapseComponent, ɵngcc2.C8yTranslateDirective, ɵngcc2.ListItemIconComponent, ɵngcc2.IconDirective, ɵngcc2.ListItemCheckboxComponent, ɵngcc4.PopoverDirective, ɵngcc2.ListItemActionComponent, ɵngcc2.TypeaheadComponent, ɵngcc1.NgModel, ɵngcc2.ForOfDirective, ɵngcc5.DatapointTemplatePopoverComponent, ɵngcc6.DatapointAttributesFormComponent], pipes: [ɵngcc2.C8yTranslatePipe, ɵngcc3.AsyncPipe], encapsulation: 2 });
DatapointSelectorListItemComponent.ctorParameters = () => [
    { type: FormBuilder }
];
DatapointSelectorListItemComponent.propDecorators = {
    defaultFormOptions: [{ type: Input }],
    isSelected: [{ type: Input }],
    isCollapsed: [{ type: Input }],
    showAddRemoveButton: [{ type: Input }],
    editable: [{ type: Input }],
    showActiveToggle: [{ type: Input }],
    activeToggleDisabled: [{ type: Input }],
    showOptions: [{ type: Input }],
    datapointLibraryEntries: [{ type: Input }],
    actions: [{ type: Input }],
    optionToRemove: [{ type: Input }],
    hasUnlinkTemplateOption: [{ type: Input }],
    added: [{ type: Output }],
    removed: [{ type: Output }],
    colorPickerDisabled: [{ type: Input }],
    disableTypeaheadIfSelected: [{ type: Input }],
    highlightText: [{ type: Input }],
    dragHandle: [{ type: ContentChild, args: [ListItemDragHandleComponent,] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DatapointSelectorListItemComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-datapoint-selector-list-item',
                template: "<c8y-li class=\"c8y-list__item__collapse--container-small\" [formGroup]=\"formGroup\" #li>\n  <c8y-li-drag-handle><ng-content select=\"c8y-li-drag-handle\"></ng-content></c8y-li-drag-handle>\n  <c8y-li-icon *ngIf=\"showAddRemoveButton\" class=\"a-s-center p-r-4\">\n    <button\n      *ngIf=\"isSelected\"\n      class=\"btn btn-clean text-primary\"\n      (click)=\"addOrRemoveItem()\"\n      [title]=\"'Remove' | translate\"\n    >\n      <i c8yIcon=\"minus-circle\" class=\"text-danger\"></i>\n    </button>\n    <button\n      *ngIf=\"!isSelected\"\n      class=\"btn btn-clean text-primary\"\n      (click)=\"addOrRemoveItem()\"\n      [title]=\"'Select' | translate\"\n    >\n      <i c8yIcon=\"plus-circle\" class=\"text-primary\"></i>\n    </button>\n  </c8y-li-icon>\n\n  <c8y-li-checkbox\n    *ngIf=\"showActiveToggle\"\n    [displayAsSwitch]=\"true\"\n    formControlName=\"__active\"\n    (click)=\"$event.stopPropagation()\"\n    class=\"a-s-center p-r-0\"\n  ></c8y-li-checkbox>\n\n  <div class=\"d-flex a-i-center p-l-4\">\n    <div class=\"c8y-list__item__colorpicker p-t-0 p-b-0 p-l-0\" [title]=\"'Change color' | translate\">\n      <div class=\"c8y-colorpicker\">\n        <input type=\"color\" formControlName=\"color\" (click)=\"$event.stopPropagation()\" />\n        <span [style.background-color]=\"formGroup.value.color\"></span>\n      </div>\n    </div>\n    <button\n      class=\"btn-clean data-point-label text-truncate\"\n      [title]=\"formGroup.value.details?.label\"\n      (click)=\"li.collapsed = !li.collapsed\"\n    >\n      <span class=\"text-truncate\">\n        <c8y-highlight\n          [text]=\"formGroup.value.details?.label\"\n          [pattern]=\"highlightText\"\n          [shouldTrimPattern]=\"true\"\n        ></c8y-highlight>\n      </span>\n      <small class=\"text-truncate text-muted icon-flex\" *ngIf=\"formGroup.value.__target\">\n        <i c8yIcon=\"exchange\"></i>\n        <span class=\"text-truncate\">{{ formGroup.value.__target.name }}</span>\n      </small>\n    </button>\n\n    <button\n      class=\"btn-clean m-l-auto\"\n      *ngIf=\"!(isValid$ | async)\"\n      [popover]=\"'Some entries are invalid. Check the input fields with red borders.' | translate\"\n      container=\"body\"\n      [outsideClick]=\"true\"\n    >\n      <i class=\"text-danger\" c8yIcon=\"warning\"></i>\n    </button>\n  </div>\n\n  <c8y-li-action\n    *ngIf=\"optionToRemove\"\n    [icon]=\"'minus-circle'\"\n    [label]=\"'Remove from list' | translate\"\n    (click)=\"remove(formGroup.value)\"\n  ></c8y-li-action>\n  <c8y-li-action\n    *ngFor=\"let action of actions\"\n    [icon]=\"action.icon\"\n    [label]=\"action.label\"\n    (click)=\"action.callback(formGroup.value)\"\n  ></c8y-li-action>\n  <c8y-li-collapse>\n    <div class=\"data-point-details\">\n      <ul class=\"list-unstyled small m-b-16\">\n        <li class=\"p-t-4 p-b-4 flex-row separator-top-bottom\">\n          <label class=\"small m-b-0 m-r-8 flex-item-v-start text-muted\" translate>Fragment</label>\n          <span class=\"flex-item-right\">\n            <c8y-highlight\n              [text]=\"formGroup.value.fragment\"\n              [pattern]=\"highlightText\"\n              [shouldTrimPattern]=\"true\"\n            ></c8y-highlight>\n          </span>\n        </li>\n        <li class=\"p-t-4 p-b-4 flex-row separator-bottom\">\n          <label class=\"small m-b-0 m-r-8 flex-item-v-start text-muted\" translate>Series</label>\n          <span class=\"flex-item-right\">\n            <c8y-highlight\n              [text]=\"formGroup.value.series\"\n              [pattern]=\"highlightText\"\n              [shouldTrimPattern]=\"true\"\n            ></c8y-highlight>\n          </span>\n        </li>\n      </ul>\n      <div\n        class=\"form-group\"\n        *ngIf=\"datapointLibraryEntries && datapointLibraryEntries | async as libraryEntries\"\n      >\n        <label class=\"d-inline-block\" translate>Data point template</label>\n        <button\n          class=\"btn btn-clean\"\n          [popover]=\"datapointHintPopoverTemplate\"\n          (click)=\"$event.stopPropagation()\"\n          container=\"body\"\n          [outsideClick]=\"true\"\n        >\n          <i [c8yIcon]=\"'question-circle-o'\" class=\"m-l-4 text-info\"></i>\n        </button>\n\n        <ng-template #datapointHintPopoverTemplate>\n          {{\n            'Using a data point template sets color, label, unit and ranges. Removing the template allows you to set the values manually.'\n              | translate\n          }}\n        </ng-template>\n        <div class=\"d-flex\">\n          <c8y-typeahead\n            [placeholder]=\"'No template' | translate\"\n            [ngModel]=\"formGroup.value.__template ? formGroup.value.details : undefined\"\n            [ngModelOptions]=\"{ standalone: true }\"\n            (onSearch)=\"setPipe($event)\"\n            [displayProperty]=\"'label'\"\n            [hideNew]=\"true\"\n            *ngIf=\"datapointLibraryEntries\"\n            [disabled]=\"isSelected && disableTypeaheadIfSelected\"\n            class=\"flex-grow\"\n          >\n            <c8y-li\n              *c8yFor=\"\n                let item of datapointLibraryEntries;\n                loadMore: 'auto';\n                pipe: filterPipe;\n                notFound: notFoundTemplate\n              \"\n              class=\"p-l-8 p-r-8 c8y-list__item--link interact\"\n              (click)=\"dataPointTemplateSelected(item)\"\n              [active]=\"formGroup.value.__template === item.id\"\n            >\n              <c8y-highlight [text]=\"item.c8y_Kpi?.label\" [pattern]=\"pattern\"></c8y-highlight>\n              <c8y-li-icon icon=\"circle\" [style.color]=\"item.c8y_Kpi?.color\"></c8y-li-icon>\n            </c8y-li>\n            <ng-template #notFoundTemplate>\n              <c8y-li class=\"bg-gray-lighter p-8\" *ngIf=\"pattern.length > 0\">\n                <span>No match found.</span>\n              </c8y-li>\n            </ng-template>\n          </c8y-typeahead>\n          <button\n            *ngIf=\"formGroup.value.__template\"\n            class=\"btn btn-info btn-xs btn-icon a-s-center m-l-8\"\n            [popover]=\"datapointOverviewPopoverTemplate\"\n            placement=\"left\"\n            container=\"body\"\n            [outsideClick]=\"true\"\n            [title]=\"'Info' | translate\"\n          >\n            <i c8yIcon=\"info\" class=\"text-info\"></i>\n          </button>\n          <button\n            *ngIf=\"formGroup.value.__template && hasUnlinkTemplateOption\"\n            class=\"btn btn-danger btn-xs btn-icon a-s-center m-l-8\"\n            (click)=\"unlinkDatapointTemplate()\"\n            [title]=\"'Unlink data point template' | translate\"\n          >\n            <i c8yIcon=\"unlink\"></i>\n          </button>\n        </div>\n      </div>\n      <ng-template #datapointOverviewPopoverTemplate>\n        <c8y-datapoint-template-popover\n          [datapoint]=\"formGroup.value.details\"\n        ></c8y-datapoint-template-popover>\n      </ng-template>\n      <c8y-datapoint-attributes-form\n        *ngIf=\"defaultFormOptions\"\n        [showTarget]=\"defaultFormOptions.showTarget\"\n        [showRange]=\"defaultFormOptions.showRange\"\n        [showYellowRange]=\"defaultFormOptions.showYellowRange\"\n        [showRedRange]=\"defaultFormOptions.showRedRange\"\n        [showChart]=\"defaultFormOptions.showChart\"\n        [showFormIfTemplateWasSelected]=\"defaultFormOptions.showFormIfTemplateWasSelected\"\n        formControlName=\"details\"\n      ></c8y-datapoint-attributes-form>\n    </div>\n  </c8y-li-collapse>\n</c8y-li>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => DatapointSelectorListItemComponent),
                        multi: true
                    },
                    {
                        provide: NG_VALIDATORS,
                        useExisting: forwardRef(() => DatapointSelectorListItemComponent),
                        multi: true
                    }
                ]
            }]
    }], function () { return [{ type: ɵngcc1.FormBuilder }]; }, { defaultFormOptions: [{
            type: Input
        }], isSelected: [{
            type: Input
        }], isCollapsed: [{
            type: Input
        }], showAddRemoveButton: [{
            type: Input
        }], editable: [{
            type: Input
        }], showActiveToggle: [{
            type: Input
        }], activeToggleDisabled: [{
            type: Input
        }], showOptions: [{
            type: Input
        }], actions: [{
            type: Input
        }], optionToRemove: [{
            type: Input
        }], hasUnlinkTemplateOption: [{
            type: Input
        }], added: [{
            type: Output
        }], removed: [{
            type: Output
        }], colorPickerDisabled: [{
            type: Input
        }], disableTypeaheadIfSelected: [{
            type: Input
        }], datapointLibraryEntries: [{
            type: Input
        }], highlightText: [{
            type: Input
        }], dragHandle: [{
            type: ContentChild,
            args: [ListItemDragHandleComponent]
        }] }); })();
