// @c8y/ngx-components/esm2015/datapoint-selector/datapoint-selector.service.js
import { Injectable } from '@angular/core';
import { BsModalService } from 'ngx-bootstrap/modal';
import { DatapointSelectorModalComponent } from './datapoint-selector-modal/datapoint-selector-modal.component';
import * as i0 from "@angular/core";
import * as i1 from "ngx-bootstrap/modal";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/modal';
export class DatapointSelectorService {
    constructor(modal) {
        this.modal = modal;
    }
    selectDataPoints(initialState = {}) {
        const modal = this.modal.show(DatapointSelectorModalComponent, {
            ignoreBackdropClick: true,
            keyboard: false,
            initialState,
            class: 'modal-lg'
        });
        const content = modal.content;
        return content.result;
    }
}
DatapointSelectorService.ɵfac = function DatapointSelectorService_Factory(t) { return new (t || DatapointSelectorService)(ɵngcc0.ɵɵinject(ɵngcc1.BsModalService)); };
DatapointSelectorService.ɵprov = i0.ɵɵdefineInjectable({ factory: function DatapointSelectorService_Factory() { return new DatapointSelectorService(i0.ɵɵinject(i1.BsModalService)); }, token: DatapointSelectorService, providedIn: "root" });
DatapointSelectorService.ctorParameters = () => [
    { type: BsModalService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DatapointSelectorService, [{
        type: Injectable,
        args: [{ providedIn: 'root' }]
    }], function () { return [{ type: ɵngcc1.BsModalService }]; }, null); })();
