// @c8y/ngx-components/esm2015/datapoint-selector/datapoint-selector.module.js
import { DragDropModule } from '@angular/cdk/drag-drop';
import { NgModule } from '@angular/core';
import { CoreModule } from '@c8y/ngx-components';
import { AssetSelectorModule } from '@c8y/ngx-components/assets-navigator';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { ModalModule } from 'ngx-bootstrap/modal';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { DatapointAttributesFormComponent } from './datapoint-attributes-form/datapoint-attributes-form.component';
import { DatapointSelectionListComponent } from './datapoint-selection-list/datapoint-selection-list.component';
import { DatapointSelectorListItemComponent } from './datapoint-selector-list-item/datapoint-selector-list-item.component';
import { DatapointSelectorModalComponent } from './datapoint-selector-modal/datapoint-selector-modal.component';
import { DatapointSelectorComponent } from './datapoint-selector.component';
import { DatapointTemplatePopoverComponent } from './datapoint-template-popover/datapoint-template-popover.component';
import { DatapointLabelPipe } from './pipes/datapoint-label.pipe';
import { FilterDatapointsPipe } from './pipes/filter-datapoints.pipe';
import { IncludesDatapointPipe } from './pipes/includes-datapoint.pipe';
import * as ɵngcc0 from '@angular/core';
export class DatapointSelectorModule {
}
DatapointSelectorModule.ɵfac = function DatapointSelectorModule_Factory(t) { return new (t || DatapointSelectorModule)(); };
DatapointSelectorModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DatapointSelectorModule });
DatapointSelectorModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            AssetSelectorModule,
            CollapseModule,
            CoreModule,
            BsDropdownModule,
            ModalModule,
            PopoverModule,
            DragDropModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DatapointSelectorModule, [{
        type: NgModule,
        args: [{
                imports: [
                    AssetSelectorModule,
                    CollapseModule,
                    CoreModule,
                    BsDropdownModule,
                    ModalModule,
                    PopoverModule,
                    DragDropModule
                ],
                declarations: [
                    DatapointSelectorComponent,
                    DatapointSelectorModalComponent,
                    DatapointSelectorListItemComponent,
                    DatapointSelectionListComponent,
                    DatapointTemplatePopoverComponent,
                    IncludesDatapointPipe,
                    DatapointLabelPipe,
                    FilterDatapointsPipe,
                    DatapointAttributesFormComponent
                ],
                exports: [
                    DatapointSelectorComponent,
                    DatapointSelectorModalComponent,
                    DatapointSelectorListItemComponent,
                    DatapointSelectionListComponent,
                    DatapointTemplatePopoverComponent,
                    IncludesDatapointPipe,
                    DatapointLabelPipe,
                    DatapointAttributesFormComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DatapointSelectorModule, { declarations: function () { return [DatapointSelectorComponent,
        DatapointSelectorModalComponent,
        DatapointSelectorListItemComponent,
        DatapointSelectionListComponent,
        DatapointTemplatePopoverComponent,
        IncludesDatapointPipe,
        DatapointLabelPipe,
        FilterDatapointsPipe,
        DatapointAttributesFormComponent]; }, imports: function () { return [AssetSelectorModule,
        CollapseModule,
        CoreModule,
        BsDropdownModule,
        ModalModule,
        PopoverModule,
        DragDropModule]; }, exports: function () { return [DatapointSelectorComponent,
        DatapointSelectorModalComponent,
        DatapointSelectorListItemComponent,
        DatapointSelectionListComponent,
        DatapointTemplatePopoverComponent,
        IncludesDatapointPipe,
        DatapointLabelPipe,
        DatapointAttributesFormComponent]; } }); })();
