// @c8y/ngx-components/esm2015/datapoint-selector/datapoint-selector.component.js
import { Component, forwardRef, Input } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { BehaviorSubject, combineLatest, from } from 'rxjs';
import { debounceTime, distinctUntilChanged, map, shareReplay, switchMap, tap } from 'rxjs/operators';
import { DatapointLibraryService } from './datapoint-library.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './datapoint-library.service';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from '@c8y/ngx-components';
import * as ɵngcc4 from '@c8y/ngx-components/assets-navigator';
import * as ɵngcc5 from '@angular/forms';
import * as ɵngcc6 from './datapoint-selector-list-item/datapoint-selector-list-item.component';
import * as ɵngcc7 from './pipes/includes-datapoint.pipe';
import * as ɵngcc8 from './pipes/datapoint-label.pipe';

const _c0 = function (a6) { return { view: "miller", groupsSelectable: true, columnHeaders: true, showChildDevices: true, showUnassignedDevices: true, singleColumn: true, search: a6 }; };
function DatapointSelectorComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 11);
    ɵngcc0.ɵɵelementStart(1, "c8y-asset-selector-miller", 12);
    ɵngcc0.ɵɵlistener("ngModelChange", function DatapointSelectorComponent_div_1_Template_c8y_asset_selector_miller_ngModelChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r9); const ctx_r8 = ɵngcc0.ɵɵnextContext(); return ctx_r8.contextAsset = $event; })("onSelected", function DatapointSelectorComponent_div_1_Template_c8y_asset_selector_miller_onSelected_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r9); const ctx_r10 = ɵngcc0.ɵɵnextContext(); return ctx_r10.selectionChanged($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r0.contextAsset)("asset", ctx_r0.contextAsset)("container", "")("config", ɵngcc0.ɵɵpureFunction1(4, _c0, ctx_r0.allowSearch));
} }
function DatapointSelectorComponent_ng_template_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 13);
    ɵngcc0.ɵɵelementStart(1, "div", 14);
    ɵngcc0.ɵɵelement(2, "h1", 15);
    ɵngcc0.ɵɵelementStart(3, "div");
    ɵngcc0.ɵɵelementStart(4, "p");
    ɵngcc0.ɵɵelementStart(5, "strong", 16);
    ɵngcc0.ɵɵtext(6, "No data points to display.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "small", 16);
    ɵngcc0.ɵɵtext(8, "Select an asset from the list.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function DatapointSelectorComponent_ng_template_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 17);
    ɵngcc0.ɵɵelement(1, "c8y-loading");
    ɵngcc0.ɵɵelementEnd();
} }
function DatapointSelectorComponent_div_7_div_6_i_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 28);
} }
function DatapointSelectorComponent_div_7_div_6_ng_template_4_i_0_Template(rf, ctx) { if (rf & 1) {
    const _r19 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "i", 30);
    ɵngcc0.ɵɵlistener("click", function DatapointSelectorComponent_div_7_div_6_ng_template_4_i_0_Template_i_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r19); const ctx_r18 = ɵngcc0.ɵɵnextContext(4); return ctx_r18.searchString = ""; });
    ɵngcc0.ɵɵelementEnd();
} }
function DatapointSelectorComponent_div_7_div_6_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, DatapointSelectorComponent_div_7_div_6_ng_template_4_i_0_Template, 1, 0, "i", 29);
} if (rf & 2) {
    const ctx_r16 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r16.searchString);
} }
function DatapointSelectorComponent_div_7_div_6_Template(rf, ctx) { if (rf & 1) {
    const _r21 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 23);
    ɵngcc0.ɵɵelementStart(1, "input", 24);
    ɵngcc0.ɵɵlistener("ngModelChange", function DatapointSelectorComponent_div_7_div_6_Template_input_ngModelChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r21); const ctx_r20 = ɵngcc0.ɵɵnextContext(2); return ctx_r20.searchStringChanged($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(2, "span", 25);
    ɵngcc0.ɵɵtemplate(3, DatapointSelectorComponent_div_7_div_6_i_3_Template, 1, 0, "i", 26);
    ɵngcc0.ɵɵtemplate(4, DatapointSelectorComponent_div_7_div_6_ng_template_4_Template, 1, 1, "ng-template", null, 27, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _r15 = ɵngcc0.ɵɵreference(5);
    const ctx_r12 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r12.searchString);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r12.searchString)("ngIfElse", _r15);
} }
function DatapointSelectorComponent_div_7_ng_container_7_ng_container_1_ng_container_1_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 13);
    ɵngcc0.ɵɵelement(1, "c8y-ui-empty-state", 34);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const datapoints_r25 = ɵngcc0.ɵɵnextContext().ngIf;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("icon", "c8y-data-points")("title", ɵngcc0.ɵɵpipeBind1(2, 4, "No data points to display."))("subtitle", datapoints_r25.length ? ɵngcc0.ɵɵpipeBind1(3, 6, "Try another search term.") : ɵngcc0.ɵɵpipeBind1(4, 8, "Select an asset with data points from the list."))("horizontal", true);
} }
function DatapointSelectorComponent_div_7_ng_container_7_ng_container_1_ng_container_1_c8y_list_item_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-list-item", 35);
    ɵngcc0.ɵɵelementStart(1, "div", 36);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 1, "Due to the large number, only a subset of data points are displayed. Use search to narrow down the number of results."), " ");
} }
function DatapointSelectorComponent_div_7_ng_container_7_ng_container_1_ng_container_1_c8y_datapoint_selector_list_item_4_Template(rf, ctx) { if (rf & 1) {
    const _r32 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-datapoint-selector-list-item", 37);
    ɵngcc0.ɵɵlistener("added", function DatapointSelectorComponent_div_7_ng_container_7_ng_container_1_ng_container_1_c8y_datapoint_selector_list_item_4_Template_c8y_datapoint_selector_list_item_added_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r32); const ctx_r31 = ɵngcc0.ɵɵnextContext(5); return ctx_r31.datapointAdded($event); })("removed", function DatapointSelectorComponent_div_7_ng_container_7_ng_container_1_ng_container_1_c8y_datapoint_selector_list_item_4_Template_c8y_datapoint_selector_list_item_removed_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r32); const ctx_r33 = ɵngcc0.ɵɵnextContext(5); return ctx_r33.datapointRemoved($event); });
    ɵngcc0.ɵɵpipe(1, "includesDatapoint");
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const dp_r30 = ctx.$implicit;
    const ctx_r28 = ɵngcc0.ɵɵnextContext(5);
    ɵngcc0.ɵɵproperty("ngModel", dp_r30)("isSelected", ɵngcc0.ɵɵpipeBind2(1, 5, ctx_r28.selectedDatapoints, dp_r30))("datapointLibraryEntries", ctx_r28.datapointLibraryEntries)("disableTypeaheadIfSelected", true)("highlightText", ɵngcc0.ɵɵpipeBind1(2, 8, ctx_r28.searchStringChanges$));
} }
function DatapointSelectorComponent_div_7_ng_container_7_ng_container_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, DatapointSelectorComponent_div_7_ng_container_7_ng_container_1_ng_container_1_div_1_Template, 5, 10, "div", 10);
    ɵngcc0.ɵɵelementStart(2, "c8y-list-group");
    ɵngcc0.ɵɵtemplate(3, DatapointSelectorComponent_div_7_ng_container_7_ng_container_1_ng_container_1_c8y_list_item_3_Template, 4, 3, "c8y-list-item", 32);
    ɵngcc0.ɵɵtemplate(4, DatapointSelectorComponent_div_7_ng_container_7_ng_container_1_ng_container_1_c8y_datapoint_selector_list_item_4_Template, 3, 10, "c8y-datapoint-selector-list-item", 33);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const datapoints_r25 = ctx.ngIf;
    const filteredDatapoints_r22 = ɵngcc0.ɵɵnextContext(2).ngIf;
    const ctx_r24 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !filteredDatapoints_r22.length);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", datapoints_r25.length > ctx_r24.maxNumberOfDatapoints && filteredDatapoints_r22.length >= ctx_r24.maxNumberOfDatapoints);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", filteredDatapoints_r22)("ngForTrackBy", ctx_r24.trackByFn);
} }
function DatapointSelectorComponent_div_7_ng_container_7_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, DatapointSelectorComponent_div_7_ng_container_7_ng_container_1_ng_container_1_Template, 5, 4, "ng-container", 31);
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r23 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(2, 1, ctx_r23.datapoints$));
} }
function DatapointSelectorComponent_div_7_ng_container_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, DatapointSelectorComponent_div_7_ng_container_7_ng_container_1_Template, 3, 3, "ng-container", 22);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r13 = ɵngcc0.ɵɵnextContext(2);
    const _r3 = ɵngcc0.ɵɵreference(6);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r13.loadingDatapoints)("ngIfElse", _r3);
} }
function DatapointSelectorComponent_div_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 18);
    ɵngcc0.ɵɵelementStart(1, "div", 19);
    ɵngcc0.ɵɵelementStart(2, "p", 20);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(6, DatapointSelectorComponent_div_7_div_6_Template, 6, 3, "div", 21);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(7, DatapointSelectorComponent_div_7_ng_container_7_Template, 2, 2, "ng-container", 22);
    ɵngcc0.ɵɵpipe(8, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    const _r3 = ɵngcc0.ɵɵreference(6);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(3, 5, "Available data points"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 7, "Available data points"), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r5.loadingDatapoints);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(8, 9, ctx_r5.filteredDatapoints$))("ngIfElse", _r3);
} }
function DatapointSelectorComponent_div_14_small_11_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 45);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const selectedDp_r35 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", selectedDp_r35 == null ? null : selectedDp_r35.__target == null ? null : selectedDp_r35.__target.name, " ");
} }
const _c1 = function () { return { doNotUseLabel: true, includeDevice: true }; };
function DatapointSelectorComponent_div_14_Template(rf, ctx) { if (rf & 1) {
    const _r39 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 38);
    ɵngcc0.ɵɵelementStart(1, "button", 39);
    ɵngcc0.ɵɵlistener("click", function DatapointSelectorComponent_div_14_Template_button_click_1_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r39); const selectedDp_r35 = restoredCtx.$implicit; const ctx_r38 = ɵngcc0.ɵɵnextContext(); return ctx_r38.datapointRemoved(selectedDp_r35); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 40);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "div", 41);
    ɵngcc0.ɵɵpipe(5, "datapointLabel");
    ɵngcc0.ɵɵelement(6, "i", 42);
    ɵngcc0.ɵɵelementStart(7, "span", 43);
    ɵngcc0.ɵɵelementStart(8, "span", 43);
    ɵngcc0.ɵɵtext(9);
    ɵngcc0.ɵɵpipe(10, "datapointLabel");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(11, DatapointSelectorComponent_div_14_small_11_Template, 2, 1, "small", 44);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const selectedDp_r35 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(2, 6, "Remove"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind2(5, 8, selectedDp_r35, ɵngcc0.ɵɵpureFunction0(13, _c1)));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵstyleProp("color", selectedDp_r35.color);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(10, 11, selectedDp_r35));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", selectedDp_r35 == null ? null : selectedDp_r35.__target == null ? null : selectedDp_r35.__target.name);
} }
function DatapointSelectorComponent_div_15_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 13);
    ɵngcc0.ɵɵelementStart(1, "div", 14);
    ɵngcc0.ɵɵelement(2, "h1", 15);
    ɵngcc0.ɵɵelementStart(3, "p");
    ɵngcc0.ɵɵelementStart(4, "strong", 16);
    ɵngcc0.ɵɵtext(5, "No data points selected.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
export class DatapointSelectorComponent {
    constructor(datapointService) {
        this.datapointService = datapointService;
        this.allowChangingContext = true;
        this.allowDatapointsFromMultipleAssets = true;
        this.selectedDatapoints = new Array();
        this.defaultActiveState = true;
        this.ignoreDatapointTemplates = false;
        this.guessDatapointUnit = true;
        this.allowSearch = true;
        this.searchString = '';
        this.maxNumberOfDatapoints = 50;
        this.loadingDatapoints = false;
        this.assetSelection = new BehaviorSubject(null);
        this.searchString$ = new BehaviorSubject('');
        this.touched = false;
        this.setupObservables();
    }
    ngOnInit() {
        if (!this.ignoreDatapointTemplates) {
            this.datapointLibraryEntries = from(this.datapointService.getFirstDatapointLibraryPage());
        }
        if (this.contextAsset) {
            this.selectionChanged(this.contextAsset);
        }
    }
    writeValue(obj) {
        this.selectedDatapoints = obj;
    }
    registerOnChange(fn) {
        this.onChange = fn;
    }
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    datapointAdded(dp) {
        this.markAsTouched();
        dp.__active = this.defaultActiveState;
        if (this.guessDatapointUnit && !dp.unit) {
            this.datapointService.guessUnitOfDatapoint(dp.fragment, dp.series, dp.__target).then(unit => {
                dp.unit = unit;
            });
        }
        this.selectedDatapoints = [...this.selectedDatapoints, dp];
        this.emitCurrentSelection();
    }
    datapointRemoved(dp) {
        this.markAsTouched();
        this.selectedDatapoints = this.selectedDatapoints.filter(tmp => {
            var _a, _b;
            return tmp.fragment !== dp.fragment ||
                tmp.series !== dp.series ||
                ((_a = tmp.__target) === null || _a === void 0 ? void 0 : _a.id) !== ((_b = dp.__target) === null || _b === void 0 ? void 0 : _b.id);
        });
        this.emitCurrentSelection();
    }
    selectionChanged(evt) {
        if (Array.isArray(evt) && evt.length !== 0) {
            return this.selectAsset(evt[0]);
        }
        if (!Array.isArray(evt) && evt.items) {
            return this.selectionChanged(evt.items);
        }
        if (!Array.isArray(evt) && evt.id) {
            return this.selectAsset(evt);
        }
        // reset selection
        this.assetSelection.next(null);
    }
    trackByFn(_index, item) {
        var _a;
        return `${item.fragment}-${(_a = item.__target) === null || _a === void 0 ? void 0 : _a.id}-${item.series}`;
    }
    searchStringChanged(newValue = '') {
        this.searchString$.next(newValue);
        this.searchString = newValue;
    }
    setupObservables() {
        this.datapoints$ = this.assetSelection.pipe(tap(() => {
            this.loadingDatapoints = true;
        }), switchMap(asset => (asset === null || asset === void 0 ? void 0 : asset.id)
            ? this.datapointService.getDatapointsOfAsset(asset, this.ignoreDatapointTemplates)
            : []), tap(() => (this.loadingDatapoints = false)), shareReplay(1));
        this.searchStringChanges$ = this.searchString$.pipe(distinctUntilChanged(), debounceTime(500), shareReplay(1));
        this.filteredDatapoints$ = combineLatest([this.searchStringChanges$, this.datapoints$]).pipe(map(([searchString, datapoints]) => {
            if (!searchString) {
                return datapoints;
            }
            const lowerCaseSearchString = searchString.toLowerCase();
            return datapoints.filter(datapoint => this.includesSearchString(datapoint, lowerCaseSearchString));
        }), map(filtered => filtered.slice(0, this.maxNumberOfDatapoints)));
    }
    selectAsset(asset) {
        this.assetSelection.next(asset);
        this.searchStringChanged();
        if (!this.allowDatapointsFromMultipleAssets) {
            this.clearSelection();
        }
    }
    clearSelection() {
        this.selectedDatapoints = [];
        this.emitCurrentSelection();
    }
    emitCurrentSelection() {
        this.onChange(this.selectedDatapoints);
    }
    markAsTouched() {
        if (!this.touched) {
            this.onTouched();
            this.touched = true;
        }
    }
    includesSearchString(datapoint, lowerCaseSearchString) {
        var _a, _b, _c;
        const label = (_a = datapoint.label) === null || _a === void 0 ? void 0 : _a.toLowerCase();
        if (label && label.includes(lowerCaseSearchString)) {
            return true;
        }
        const fragment = (_b = datapoint.fragment) === null || _b === void 0 ? void 0 : _b.toLowerCase();
        if (fragment && fragment.includes(lowerCaseSearchString)) {
            return true;
        }
        const series = (_c = datapoint.series) === null || _c === void 0 ? void 0 : _c.toLowerCase();
        if (series && series.includes(lowerCaseSearchString)) {
            return true;
        }
        return false;
    }
}
DatapointSelectorComponent.ɵfac = function DatapointSelectorComponent_Factory(t) { return new (t || DatapointSelectorComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.DatapointLibraryService)); };
DatapointSelectorComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: DatapointSelectorComponent, selectors: [["c8y-datapoint-selector"]], inputs: { allowChangingContext: "allowChangingContext", allowDatapointsFromMultipleAssets: "allowDatapointsFromMultipleAssets", defaultActiveState: "defaultActiveState", ignoreDatapointTemplates: "ignoreDatapointTemplates", guessDatapointUnit: "guessDatapointUnit", allowSearch: "allowSearch", contextAsset: "contextAsset" }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: NG_VALUE_ACCESSOR,
                multi: true,
                useExisting: forwardRef(() => DatapointSelectorComponent)
            }
        ])], decls: 16, vars: 11, consts: [[1, "d-grid", "grid__row--1", "fit-h", 3, "ngClass"], ["class", "d-flex d-col p-relative bg-gray-white", 4, "ngIf"], [1, "inner-scroll", "bg-component"], ["noDeviceEmptyState", ""], ["loadingData", ""], ["class", "bg-inherit", 4, "ngIf", "ngIfElse"], [1, "inner-scroll", "bg-gray-white"], ["translate", "", 1, "text-medium", "m-b-4", "p-l-16", "p-r-16", "p-t-8", "p-b-8", "separator-bottom", "sticky-top", "text-truncate", 3, "title"], [1, "d-flex", "flex-wrap", "gap-8", "p-l-16", "p-r-16", "p-b-16"], ["class", "c8y-datapoint-pill", 4, "ngFor", "ngForOf"], ["class", "p-16", 4, "ngIf"], [1, "d-flex", "d-col", "p-relative", "bg-gray-white"], [1, "d-contents", 3, "ngModel", "asset", "container", "config", "ngModelChange", "onSelected"], [1, "p-16"], [1, "c8y-empty-state", "text-left"], [1, "c8y-icon", "c8y-icon-data-points", "c8y-icon-duocolor"], ["translate", ""], [1, "p-16", "text-center"], [1, "bg-inherit"], [1, "p-l-16", "p-r-16", "p-t-8", "p-b-8", "sticky-top", "bg-inherit", "separator-bottom"], [1, "text-medium", "text-truncate", 3, "title"], ["id", "search", "class", "input-group input-group-search m-t-4", 4, "ngIf"], [4, "ngIf", "ngIfElse"], ["id", "search", 1, "input-group", "input-group-search", "m-t-4"], ["type", "search", "placeholder", "Search\u2026", 1, "form-control", 3, "ngModel", "ngModelChange"], [1, "input-group-addon"], ["c8yIcon", "search", 4, "ngIf", "ngIfElse"], ["clearSearchString", ""], ["c8yIcon", "search"], ["c8yIcon", "times", "class", "text-muted", 3, "click", 4, "ngIf"], ["c8yIcon", "times", 1, "text-muted", 3, "click"], [4, "ngIf"], ["class", "sticky-top", "style", "top: 72px", "translate", "", 4, "ngIf"], ["class", "d-contents", 3, "ngModel", "isSelected", "datapointLibraryEntries", "disableTypeaheadIfSelected", "highlightText", "added", "removed", 4, "ngFor", "ngForOf", "ngForTrackBy"], [3, "icon", "title", "subtitle", "horizontal"], ["translate", "", 1, "sticky-top", 2, "top", "72px"], [1, "alert", "alert-warning", "m-b-0"], [1, "d-contents", 3, "ngModel", "isSelected", "datapointLibraryEntries", "disableTypeaheadIfSelected", "highlightText", "added", "removed"], [1, "c8y-datapoint-pill"], ["type", "button", 1, "c8y-datapoint-pill__btn", 3, "title", "click"], ["c8yIcon", "remove", 1, "icon-14"], [1, "c8y-datapoint-pill__label", 3, "title"], ["c8yIcon", "circle", 1, "m-r-4", "icon-14"], [1, "text-truncate"], ["class", "text-muted text-10", 4, "ngIf"], [1, "text-muted", "text-10"]], template: function DatapointSelectorComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, DatapointSelectorComponent_div_1_Template, 2, 6, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵtemplate(3, DatapointSelectorComponent_ng_template_3_Template, 9, 0, "ng-template", null, 3, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(5, DatapointSelectorComponent_ng_template_5_Template, 2, 0, "ng-template", null, 4, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(7, DatapointSelectorComponent_div_7_Template, 9, 11, "div", 5);
        ɵngcc0.ɵɵpipe(8, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "div", 6);
        ɵngcc0.ɵɵelementStart(10, "p", 7);
        ɵngcc0.ɵɵpipe(11, "translate");
        ɵngcc0.ɵɵtext(12, " Selected data points ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(13, "div", 8);
        ɵngcc0.ɵɵtemplate(14, DatapointSelectorComponent_div_14_Template, 12, 14, "div", 9);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(15, DatapointSelectorComponent_div_15_Template, 6, 0, "div", 10);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r1 = ɵngcc0.ɵɵreference(4);
        ɵngcc0.ɵɵproperty("ngClass", ctx.allowChangingContext ? "grid__col--3-6-3--md" : "grid__col--8-4--md");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.allowChangingContext);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(8, 7, ctx.assetSelection))("ngIfElse", _r1);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(11, 9, "Selected data points"));
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.selectedDatapoints);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.selectedDatapoints || !ctx.selectedDatapoints.length);
    } }, directives: [ɵngcc2.NgClass, ɵngcc2.NgIf, ɵngcc3.C8yTranslateDirective, ɵngcc2.NgForOf, ɵngcc4.MillerViewComponent, ɵngcc5.NgControlStatus, ɵngcc5.NgModel, ɵngcc3.LoadingComponent, ɵngcc5.DefaultValueAccessor, ɵngcc3.IconDirective, ɵngcc3.ListGroupComponent, ɵngcc3.EmptyStateComponent, ɵngcc3.ListItemComponent, ɵngcc6.DatapointSelectorListItemComponent], pipes: [ɵngcc2.AsyncPipe, ɵngcc3.C8yTranslatePipe, ɵngcc7.IncludesDatapointPipe, ɵngcc8.DatapointLabelPipe], encapsulation: 2 });
DatapointSelectorComponent.ctorParameters = () => [
    { type: DatapointLibraryService }
];
DatapointSelectorComponent.propDecorators = {
    contextAsset: [{ type: Input }],
    allowChangingContext: [{ type: Input }],
    allowDatapointsFromMultipleAssets: [{ type: Input }],
    defaultActiveState: [{ type: Input }],
    ignoreDatapointTemplates: [{ type: Input }],
    guessDatapointUnit: [{ type: Input }],
    allowSearch: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DatapointSelectorComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-datapoint-selector',
                template: "<div class=\"d-grid grid__row--1 fit-h\" [ngClass]=\"allowChangingContext ? 'grid__col--3-6-3--md' : 'grid__col--8-4--md'\">\n  <div class=\"d-flex d-col p-relative bg-gray-white\" *ngIf=\"allowChangingContext\">\n    <c8y-asset-selector-miller\n      [(ngModel)]=\"contextAsset\"\n      [asset]=\"contextAsset\"\n      (onSelected)=\"selectionChanged($event)\"\n      [container]=\"''\"\n      [config]=\"{\n        view: 'miller',\n        groupsSelectable: true,\n        columnHeaders: true,\n        showChildDevices: true,\n        showUnassignedDevices: true,\n        singleColumn: true,\n        search: allowSearch\n      }\"\n      class=\"d-contents\"\n    ></c8y-asset-selector-miller>\n  </div>\n  <!-- center column -->\n  <div class=\"inner-scroll bg-component\">\n    <ng-template #noDeviceEmptyState>\n      <div class=\"p-16\">\n        <div class=\"c8y-empty-state text-left\">\n          <h1 class=\"c8y-icon c8y-icon-data-points c8y-icon-duocolor\"></h1>\n          <div>\n            <p>\n              <strong translate>No data points to display.</strong>\n            </p>\n            <small translate>Select an asset from the list.</small>\n          </div>\n        </div>\n      </div>\n    </ng-template>\n    <ng-template #loadingData>\n      <div class=\"p-16 text-center\">\n        <c8y-loading></c8y-loading>\n      </div>\n    </ng-template>\n    <div *ngIf=\"assetSelection | async as asset; else noDeviceEmptyState\" class=\"bg-inherit\">\n      <div class=\"p-l-16 p-r-16 p-t-8 p-b-8 sticky-top bg-inherit separator-bottom\">\n        <p\n          class=\"text-medium text-truncate\"\n          [title]=\"'Available data points' | translate\"\n        >\n          {{ 'Available data points' | translate }}\n        </p>\n        <div *ngIf=\"!loadingDatapoints\" id=\"search\" class=\"input-group input-group-search m-t-4\">\n          <input\n            type=\"search\"\n            class=\"form-control\"\n            placeholder=\"Search\u2026\"\n            [ngModel]=\"searchString\"\n            (ngModelChange)=\"searchStringChanged($event)\"\n          />\n          <span class=\"input-group-addon\">\n            <i c8yIcon=\"search\" *ngIf=\"!searchString; else clearSearchString\"></i>\n            <ng-template #clearSearchString>\n                <i\n                  c8yIcon=\"times\"\n                  class=\"text-muted\"\n                  *ngIf=\"searchString\"\n                  (click)=\"searchString = ''\"\n                ></i>\n            </ng-template>\n          </span>\n        </div>\n      </div>\n      <ng-container *ngIf=\"filteredDatapoints$ | async as filteredDatapoints; else loadingData\">\n        <ng-container *ngIf=\"!loadingDatapoints; else loadingData\">\n          <ng-container *ngIf=\"datapoints$ | async as datapoints\">\n            <div class=\"p-16\" *ngIf=\"!filteredDatapoints.length\">\n              <c8y-ui-empty-state\n                [icon]=\"'c8y-data-points'\"\n                [title]=\"'No data points to display.' | translate\"\n                [subtitle]=\"datapoints.length ? ('Try another search term.' | translate) : ('Select an asset with data points from the list.' | translate)\"\n                [horizontal]=\"true\"\n              ></c8y-ui-empty-state>\n            </div>\n\n\n\n            <c8y-list-group>\n              <c8y-list-item *ngIf=\"datapoints.length > maxNumberOfDatapoints && filteredDatapoints.length >= maxNumberOfDatapoints\"\n                class=\"sticky-top\"\n                style=\"top: 72px\"\n                translate\n              >\n                <div class=\"alert alert-warning m-b-0\">\n                  {{ 'Due to the large number, only a subset of data points are displayed. Use search to narrow down the number of results.' | translate}}\n                </div>\n              </c8y-list-item>\n              <c8y-datapoint-selector-list-item\n                [ngModel]=\"dp\"\n                [isSelected]=\"selectedDatapoints | includesDatapoint: dp\"\n                [datapointLibraryEntries]=\"datapointLibraryEntries\"\n                [disableTypeaheadIfSelected]=\"true\"\n                (added)=\"datapointAdded($event)\"\n                (removed)=\"datapointRemoved($event)\"\n                [highlightText]=\"searchStringChanges$ | async\"\n                class=\"d-contents\"\n                *ngFor=\"let dp of filteredDatapoints; trackBy: trackByFn\"\n              ></c8y-datapoint-selector-list-item>\n            </c8y-list-group>\n          </ng-container>\n        </ng-container>\n      </ng-container>\n    </div>\n  </div>\n  <!-- last column  -->\n  <div class=\"inner-scroll bg-gray-white\">\n    <p\n      class=\"text-medium m-b-4 p-l-16 p-r-16 p-t-8 p-b-8 separator-bottom sticky-top text-truncate\"\n      [title]=\"'Selected data points' | translate\"\n      translate\n    >\n      Selected data points\n    </p>\n    <div class=\"d-flex flex-wrap gap-8 p-l-16 p-r-16 p-b-16\">\n      <div class=\"c8y-datapoint-pill\" *ngFor=\"let selectedDp of selectedDatapoints\">\n        <button\n          [title]=\"'Remove' | translate\"\n          type=\"button\"\n          class=\"c8y-datapoint-pill__btn\"\n          (click)=\"datapointRemoved(selectedDp)\"\n        >\n          <i c8yIcon=\"remove\" class=\"icon-14\"></i>\n        </button>\n        <div\n          class=\"c8y-datapoint-pill__label\"\n          [title]=\"selectedDp | datapointLabel: { doNotUseLabel: true, includeDevice: true }\"\n        >\n          <i\n            c8yIcon=\"circle\"\n            class=\"m-r-4 icon-14\"\n            [style.color]=\"selectedDp.color\"\n          ></i>\n          <span class=\"text-truncate\">\n            <span class=\"text-truncate\">{{ selectedDp | datapointLabel }}</span>\n            <small class=\"text-muted text-10\" *ngIf=\"selectedDp?.__target?.name\">\n              {{ selectedDp?.__target?.name }}\n            </small>\n          </span>\n        </div>\n      </div>\n    </div>\n    <div class=\"p-16\" *ngIf=\"!selectedDatapoints || !selectedDatapoints.length\">\n      <div class=\"c8y-empty-state text-left\">\n        <h1 class=\"c8y-icon c8y-icon-data-points c8y-icon-duocolor\"></h1>\n        <p>\n          <strong translate>No data points selected.</strong>\n        </p>\n      </div>\n    </div>\n  </div>\n</div>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        multi: true,
                        useExisting: forwardRef(() => DatapointSelectorComponent)
                    }
                ]
            }]
    }], function () { return [{ type: ɵngcc1.DatapointLibraryService }]; }, { allowChangingContext: [{
            type: Input
        }], allowDatapointsFromMultipleAssets: [{
            type: Input
        }], defaultActiveState: [{
            type: Input
        }], ignoreDatapointTemplates: [{
            type: Input
        }], guessDatapointUnit: [{
            type: Input
        }], allowSearch: [{
            type: Input
        }], contextAsset: [{
            type: Input
        }] }); })();
