// @c8y/ngx-components/esm2015/ecosystem/applications/duplicate-application/duplicate-application-properties/duplicate-application-properties.component.js
import { __awaiter } from "tslib";
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { ApplicationService } from '@c8y/client';
import { BsModalRef } from 'ngx-bootstrap/modal';
import { distinctUntilKeyChanged, map, tap } from 'rxjs/operators';
import { APP_STATE } from '../../../ecosystem.constants';
import { EcosystemService } from '../../../ecosystem.service';
import { ApplicationPropertiesFormComponent } from '../../../shared/application-properties-form.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/modal';
import * as ɵngcc2 from '../../../ecosystem.service';
import * as ɵngcc3 from '@c8y/client';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from '@c8y/ngx-components';
import * as ɵngcc6 from '../../../shared/application-properties-form.component';
import * as ɵngcc7 from '@angular/forms';

function DuplicateApplicationPropertiesComponent_p_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p", 6);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Provide application details"), "\n");
} }
function DuplicateApplicationPropertiesComponent_ng_container_1_label_1_Template(rf, ctx) { if (rf & 1) {
    const _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "label", 9);
    ɵngcc0.ɵɵelementStart(1, "input", 10);
    ɵngcc0.ɵɵlistener("ngModelChange", function DuplicateApplicationPropertiesComponent_ng_container_1_label_1_Template_input_ngModelChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r5 = ɵngcc0.ɵɵnextContext(2); return ctx_r5.disableForm = $event; })("change", function DuplicateApplicationPropertiesComponent_ng_container_1_label_1_Template_input_change_1_listener() { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r7 = ɵngcc0.ɵɵnextContext(2); return ctx_r7.getAppConfig(); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(2, "span");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r4.disableForm);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 2, "Overrule subscribed application"), " ");
} }
function DuplicateApplicationPropertiesComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, DuplicateApplicationPropertiesComponent_ng_container_1_label_1_Template, 5, 4, "label", 7);
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵelement(3, "c8y-application-properties-form", 8);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(2, 3, ctx_r1.isSubscribedApp$));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("application", ctx_r1.newAppConfig)("disabled", ctx_r1.disableForm);
} }
function DuplicateApplicationPropertiesComponent_c8y_progress_bar_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-progress-bar", 11);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("message", ɵngcc0.ɵɵpipeBind1(1, 1, "Duplicating\u2026"));
} }
function DuplicateApplicationPropertiesComponent_button_4_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 12);
    ɵngcc0.ɵɵlistener("click", function DuplicateApplicationPropertiesComponent_button_4_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r9); const ctx_r8 = ɵngcc0.ɵɵnextContext(); return ctx_r8.back(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 3, "Back"));
    ɵngcc0.ɵɵproperty("disabled", ctx_r3.inProgress);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 5, "Back"), " ");
} }
export class DuplicateApplicationPropertiesComponent {
    constructor(bsModalRef, ecosystemService, applicationService) {
        this.bsModalRef = bsModalRef;
        this.ecosystemService = ecosystemService;
        this.applicationService = applicationService;
        this.isFirstStep = false;
        this.duplicatedApp = new EventEmitter();
        this.disableForm = false;
        this.inProgress = false;
    }
    ngOnInit() {
        this.isSubscribedApp$ = this.selectedApp.pipe(distinctUntilKeyChanged('id'), map(app => (app ? this.ecosystemService.getAppState(app) === APP_STATE.SUBSCRIBED : true)), tap(isSubscribed => {
            this.disableForm = isSubscribed;
            this.getAppConfig();
        }));
        this.getAppConfig();
    }
    duplicateApp() {
        return __awaiter(this, void 0, void 0, function* () {
            const formGroupValue = this.applicationPropertiesForm.formGroup.getRawValue();
            this.inProgress = true;
            const clonedApp = (yield this.applicationService.clone(this.selectedApp.value)).data;
            Object.assign(clonedApp, formGroupValue);
            delete clonedApp.type;
            yield this.updateApp(clonedApp);
            this.inProgress = false;
        });
    }
    cancel() {
        this.bsModalRef.hide();
    }
    back() {
        this.stepper.previous();
    }
    getAppConfig() {
        if (this.disableForm) {
            const { name, key, contextPath } = this.selectedApp.value;
            this.newAppConfig = { name, key, contextPath };
        }
        else {
            this.newAppConfig = this.ecosystemService.getUniqueAppConfig(this.selectedApp.value, this.existingApps);
        }
    }
    updateApp(clonedAppConfig) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const { data: app } = yield this.ecosystemService.updateApp(clonedAppConfig, true);
                const manifest = yield this.getAppManifest(app);
                if (manifest) {
                    yield this.updateManifest(app, manifest);
                }
                this.duplicatedApp.emit(app);
                this.stepper.next();
            }
            catch (err) {
                this.inProgress = false;
            }
        });
    }
    updateManifest(app, manifest) {
        return __awaiter(this, void 0, void 0, function* () {
            const keysToUpdate = ['name', 'key', 'contextPath'];
            const someKeyDiffers = keysToUpdate.some(key => app[key] !== manifest[key]);
            if (someKeyDiffers) {
                keysToUpdate.forEach(key => {
                    manifest[key] = app[key];
                });
                yield this.applicationService.storeAppManifest(app, manifest);
            }
        });
    }
    getAppManifest(app) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!(app === null || app === void 0 ? void 0 : app.contextPath)) {
                return;
            }
            try {
                const manifest = yield this.applicationService.getAppManifest(app);
                return manifest;
            }
            catch (ex) {
                return;
            }
        });
    }
}
DuplicateApplicationPropertiesComponent.ɵfac = function DuplicateApplicationPropertiesComponent_Factory(t) { return new (t || DuplicateApplicationPropertiesComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BsModalRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.EcosystemService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.ApplicationService)); };
DuplicateApplicationPropertiesComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: DuplicateApplicationPropertiesComponent, selectors: [["c8y-duplicate-application-properties"]], viewQuery: function DuplicateApplicationPropertiesComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(ApplicationPropertiesFormComponent, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.applicationPropertiesForm = _t.first);
    } }, inputs: { isFirstStep: "isFirstStep", stepper: "stepper", existingApps: "existingApps", selectedApp: "selectedApp" }, outputs: { duplicatedApp: "duplicatedApp" }, decls: 13, vars: 17, consts: [["class", "p-16 text-center text-medium separator-bottom sticky-top bg-white fit-w", 4, "ngIf"], [4, "ngIf"], ["class", "text-center d-block", 3, "message", 4, "ngIf"], ["class", "btn btn-default", "type", "button", 3, "disabled", "title", "click", 4, "ngIf"], [1, "btn", "btn-default", 3, "title", "click"], ["type", "button", 1, "btn", "btn-primary", 3, "disabled", "title", "click"], [1, "p-16", "text-center", "text-medium", "separator-bottom", "sticky-top", "bg-white", "fit-w"], ["class", "c8y-switch", 4, "ngIf"], [1, "d-block", "fit-w", 3, "application", "disabled"], [1, "c8y-switch"], ["type", "checkbox", 3, "ngModel", "ngModelChange", "change"], [1, "text-center", "d-block", 3, "message"], ["type", "button", 1, "btn", "btn-default", 3, "disabled", "title", "click"]], template: function DuplicateApplicationPropertiesComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, DuplicateApplicationPropertiesComponent_p_0_Template, 3, 3, "p", 0);
        ɵngcc0.ɵɵtemplate(1, DuplicateApplicationPropertiesComponent_ng_container_1_Template, 4, 5, "ng-container", 1);
        ɵngcc0.ɵɵtemplate(2, DuplicateApplicationPropertiesComponent_c8y_progress_bar_2_Template, 2, 3, "c8y-progress-bar", 2);
        ɵngcc0.ɵɵelementStart(3, "c8y-wizard-footer");
        ɵngcc0.ɵɵtemplate(4, DuplicateApplicationPropertiesComponent_button_4_Template, 4, 7, "button", 3);
        ɵngcc0.ɵɵelementStart(5, "button", 4);
        ɵngcc0.ɵɵlistener("click", function DuplicateApplicationPropertiesComponent_Template_button_click_5_listener() { return ctx.cancel(); });
        ɵngcc0.ɵɵpipe(6, "translate");
        ɵngcc0.ɵɵtext(7);
        ɵngcc0.ɵɵpipe(8, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "button", 5);
        ɵngcc0.ɵɵlistener("click", function DuplicateApplicationPropertiesComponent_Template_button_click_9_listener() { return ctx.duplicateApp(); });
        ɵngcc0.ɵɵpipe(10, "translate");
        ɵngcc0.ɵɵtext(11);
        ɵngcc0.ɵɵpipe(12, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", !ctx.inProgress);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.inProgress);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.inProgress);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isFirstStep);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(6, 9, "Cancel"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(8, 11, "Cancel"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(10, 13, "Duplicate"));
        ɵngcc0.ɵɵproperty("disabled", ctx.inProgress);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(12, 15, "Duplicate"), " ");
    } }, directives: [ɵngcc4.NgIf, ɵngcc5.WizardFooterComponent, ɵngcc6.ApplicationPropertiesFormComponent, ɵngcc7.CheckboxControlValueAccessor, ɵngcc7.NgControlStatus, ɵngcc7.NgModel, ɵngcc5.ProgressBarComponent], pipes: [ɵngcc5.C8yTranslatePipe, ɵngcc4.AsyncPipe], encapsulation: 2 });
DuplicateApplicationPropertiesComponent.ctorParameters = () => [
    { type: BsModalRef },
    { type: EcosystemService },
    { type: ApplicationService }
];
DuplicateApplicationPropertiesComponent.propDecorators = {
    stepper: [{ type: Input }],
    existingApps: [{ type: Input }],
    selectedApp: [{ type: Input }],
    isFirstStep: [{ type: Input }],
    duplicatedApp: [{ type: Output }],
    applicationPropertiesForm: [{ type: ViewChild, args: [ApplicationPropertiesFormComponent,] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DuplicateApplicationPropertiesComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-duplicate-application-properties',
                template: "<p\n  class=\"p-16 text-center text-medium separator-bottom sticky-top bg-white fit-w\"\n  *ngIf=\"!inProgress\"\n>\n  {{ 'Provide application details' | translate }}\n</p>\n\n<ng-container *ngIf=\"!inProgress\">\n  <label class=\"c8y-switch\" *ngIf=\"isSubscribedApp$ | async\">\n    <input type=\"checkbox\" [(ngModel)]=\"disableForm\" (change)=\"getAppConfig()\" />\n    <span></span> {{ 'Overrule subscribed application' | translate }}\n  </label>\n\n  <c8y-application-properties-form\n    [application]=\"newAppConfig\"\n    class=\"d-block fit-w\"\n    [disabled]=\"disableForm\"\n  ></c8y-application-properties-form>\n</ng-container>\n<c8y-progress-bar\n  *ngIf=\"inProgress\"\n  [message]=\"'Duplicating\u2026' | translate\"\n  class=\"text-center d-block\"\n></c8y-progress-bar>\n\n<c8y-wizard-footer>\n  <button\n    *ngIf=\"!isFirstStep\"\n    (click)=\"back()\"\n    [disabled]=\"inProgress\"\n    class=\"btn btn-default\"\n    type=\"button\"\n    title=\"{{ 'Back' | translate }}\"\n  >\n    {{ 'Back' | translate }}\n  </button>\n  <button (click)=\"cancel()\" class=\"btn btn-default\" title=\"{{ 'Cancel' | translate }}\">\n    {{ 'Cancel' | translate }}\n  </button>\n  <button\n    (click)=\"duplicateApp()\"\n    [disabled]=\"inProgress\"\n    class=\"btn btn-primary\"\n    type=\"button\"\n    title=\"{{ 'Duplicate' | translate }}\"\n  >\n    {{ 'Duplicate' | translate }}\n  </button>\n</c8y-wizard-footer>\n"
            }]
    }], function () { return [{ type: ɵngcc1.BsModalRef }, { type: ɵngcc2.EcosystemService }, { type: ɵngcc3.ApplicationService }]; }, { isFirstStep: [{
            type: Input
        }], duplicatedApp: [{
            type: Output
        }], stepper: [{
            type: Input
        }], existingApps: [{
            type: Input
        }], selectedApp: [{
            type: Input
        }], applicationPropertiesForm: [{
            type: ViewChild,
            args: [ApplicationPropertiesFormComponent]
        }] }); })();
