// @c8y/ngx-components/esm2015/register-device/bulk/bulk-device-registration-modal.component.js
import { Component, ViewChild } from '@angular/core';
import { C8yJSONSchema, C8yStepper, gettext } from '@c8y/ngx-components';
import { FormGroup } from '@angular/forms';
import { saveAs } from 'file-saver';
import { DeviceRegistrationBulkService } from '@c8y/client';
import { Router } from '@angular/router';
import { RegisterDeviceService } from '../register-device.service';
import { BsModalRef } from 'ngx-bootstrap/modal';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from '@c8y/client';
import * as ɵngcc4 from '../register-device.service';
import * as ɵngcc5 from 'ngx-bootstrap/modal';
import * as ɵngcc6 from '@angular/cdk/stepper';
import * as ɵngcc7 from '@ngx-formly/core';
import * as ɵngcc8 from '@angular/common';

function BulkDeviceRegistrationModalComponent_div_38_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelement(1, "c8y-operation-result", 23);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("text", ɵngcc0.ɵɵpipeBind1(2, 3, ctx_r0.message));
    ɵngcc0.ɵɵproperty("size", 84)("vertical", true);
} }
function BulkDeviceRegistrationModalComponent_ng_template_39_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-operation-result", 24);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("text", ɵngcc0.ɵɵpipeBind1(1, 3, ctx_r2.message));
    ɵngcc0.ɵɵproperty("size", 84)("vertical", true);
} }
function BulkDeviceRegistrationModalComponent_ng_container_42_c8y_li_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li");
    ɵngcc0.ɵɵelement(1, "c8y-li-icon", 26);
    ɵngcc0.ɵɵelementStart(2, "p");
    ɵngcc0.ɵɵtext(3, "All devices processed successfully.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "c8y-li-collapse");
    ɵngcc0.ɵɵelementStart(5, "pre");
    ɵngcc0.ɵɵelementStart(6, "code");
    ɵngcc0.ɵɵtext(7);
    ɵngcc0.ɵɵpipe(8, "json");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(7);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(8, 1, ctx_r6.result));
} }
const _c0 = function (a0, a1) { return { count: a0, total: a1 }; };
function BulkDeviceRegistrationModalComponent_ng_container_42_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li");
    ɵngcc0.ɵɵelement(1, "c8y-li-icon", 27);
    ɵngcc0.ɵɵelementStart(2, "p", 28);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(3, " Failed to process {{ count }} out of {{ total }}. ");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "c8y-li-collapse");
    ɵngcc0.ɵɵelementStart(5, "pre");
    ɵngcc0.ɵɵelementStart(6, "code");
    ɵngcc0.ɵɵtext(7);
    ɵngcc0.ɵɵpipe(8, "json");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r8 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction2(4, _c0, ctx_r8.result == null ? null : ctx_r8.result.numberOfFailed, ctx_r8.result == null ? null : ctx_r8.result.numberOfAll));
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(8, 2, ctx_r8.result));
} }
function BulkDeviceRegistrationModalComponent_ng_container_42_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, BulkDeviceRegistrationModalComponent_ng_container_42_c8y_li_1_Template, 9, 3, "c8y-li", 18);
    ɵngcc0.ɵɵtemplate(2, BulkDeviceRegistrationModalComponent_ng_container_42_ng_template_2_Template, 9, 7, "ng-template", null, 25, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const _r7 = ɵngcc0.ɵɵreference(3);
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.success)("ngIfElse", _r7);
} }
function BulkDeviceRegistrationModalComponent_ng_template_43_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li");
    ɵngcc0.ɵɵelement(1, "c8y-li-icon", 29);
    ɵngcc0.ɵɵelementStart(2, "small");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "c8y-li-collapse");
    ɵngcc0.ɵɵelementStart(6, "pre");
    ɵngcc0.ɵɵelementStart(7, "code");
    ɵngcc0.ɵɵtext(8);
    ɵngcc0.ɵɵpipe(9, "json");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("icon", "ban");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(4, 3, ctx_r5.failedResult == null ? null : ctx_r5.failedResult.message));
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(9, 5, ctx_r5.failedResult));
} }
const _c1 = function () { return { cancel: true, next: true }; };
const _c2 = function () { return { next: "Upload" }; };
const _c3 = function () { return { next: true }; };
const _c4 = function (a0) { return { next: a0 }; };
const bulkSchema = {
    $schema: 'https://json-schema.org/draft/2019-09/schema',
    type: 'object',
    properties: {
        csvBulkFile: {
            type: 'string',
            title: gettext('CSV file upload'),
            description: gettext('You can use file upload component to let users send files. This input accepts only a single CSV file.'),
            contentEncoding: 'base64',
            contentMediaType: 'csv'
        }
    },
    required: ['csvBulkFile'],
    additionalProperties: false
};
const simpleCsvHeaders = ['ID', 'PATH'];
const fullCsvHeaders = [
    'ID',
    'CREDENTIALS',
    'TYPE',
    'NAME',
    'ICCID',
    'IDTYPE',
    'PATH',
    'SHELL',
    'AUTH_TYPE'
];
export class BulkDeviceRegistrationModalComponent {
    constructor(router, jsonschema, deviceRegistrationService, registerDeviceService, bsModalRef) {
        this.router = router;
        this.jsonschema = jsonschema;
        this.deviceRegistrationService = deviceRegistrationService;
        this.registerDeviceService = registerDeviceService;
        this.bsModalRef = bsModalRef;
        this.form = new FormGroup({});
        this.model = {};
    }
    ngOnInit() {
        this.template = [this.jsonschema.toFieldConfig(bulkSchema)];
    }
    upload() {
        this.pending = true;
        const file = this.getFile(this.model);
        this.deviceRegistrationService
            .create(file)
            .then(({ res, data }) => {
            if (res.status < 400) {
                this.result = data;
                this.success = data.numberOfFailed === 0 && data.numberOfSuccessful === data.numberOfAll;
                this.message = this.success
                    ? gettext('Device registration created.')
                    : (this.message = gettext('Device registration failed.'));
            }
            else {
                this.failedResult = data;
                this.message = gettext('Device registration failed.');
            }
            this.model = {};
            this.pending = false;
            this.stepper.next();
        })
            .catch(reason => {
            this.message = gettext('Error occurred while processing the uploaded file.');
            this.pending = false;
            this.stepper.next();
        });
    }
    downloadSimple() {
        return this.download(simpleCsvHeaders, gettext('Simple bulk registration - template.csv'));
    }
    downloadFull() {
        return this.download(fullCsvHeaders, gettext('Full bulk registration - template.csv'));
    }
    download(headers, fileName) {
        const headerRaw = headers.map(header => `"${header}"`).join(';');
        const binaryFile = new Blob([headerRaw], { type: 'text/csv' });
        saveAs(binaryFile, fileName);
    }
    complete() {
        this.registerDeviceService.list();
        this.bsModalRef.hide();
    }
    cancel() {
        this.bsModalRef.hide();
    }
    getFile(model) {
        var _a, _b;
        const csvBulkFile = (_a = model) === null || _a === void 0 ? void 0 : _a.csvBulkFile;
        return csvBulkFile ? (_b = csvBulkFile[0]) === null || _b === void 0 ? void 0 : _b.file : undefined;
    }
}
BulkDeviceRegistrationModalComponent.ɵfac = function BulkDeviceRegistrationModalComponent_Factory(t) { return new (t || BulkDeviceRegistrationModalComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Router), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.C8yJSONSchema), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.DeviceRegistrationBulkService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.RegisterDeviceService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.BsModalRef)); };
BulkDeviceRegistrationModalComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: BulkDeviceRegistrationModalComponent, selectors: [["bulk-device-registration"]], viewQuery: function BulkDeviceRegistrationModalComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(C8yStepper, 7);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.stepper = _t.first);
    } }, decls: 46, vars: 31, consts: [[3, "headerClasses", "customFooter"], ["c8y-modal-title", ""], ["c8yIcon", "upload"], ["linear", "", 3, "hideStepProgress"], [3, "stepControl"], [1, "sticky-top", "bg-component", "separator-bottom"], ["translate", "", 1, "p-l-24", "p-r-24", "p-t-16", "p-b-16", "m-b-0", "sticky-top", "separator-bottom", "text-16", "text-medium", "text-center", "bg-component"], [1, "p-24", "p-t-16", "p-b-8", "m-b-0"], [3, "form", "fields", "model"], [1, "bg-gray-white", "separator-bottom", "p-t-16", "p-b-16", "p-l-24", "p-r-24"], ["translate", "", 1, "m-b-8", "text-medium"], ["translate", "", 1, "small", "m-b-8"], ["target", "_self", 1, "btn", "btn-default", "btn-sm", 3, "title", "click"], ["c8yIcon", "download"], ["translate", ""], [1, "sticky-bottom", "d-block", "p-t-16", "p-b-16", "separator-top", "bg-white", 3, "showButtons", "disabled", "pending", "labels", "onCancel", "onNext"], ["state", "final"], [1, "m-24"], [4, "ngIf", "ngIfElse"], ["warning", ""], [1, "separator-top", "m-t-16"], ["failedResponse", ""], [1, "sticky-bottom", "d-block", "p-t-16", "p-b-16", "separator-top", "bg-white", 3, "showButtons", "labels", "onNext"], ["type", "success", 1, "lead", 3, "text", "size", "vertical"], ["type", "error", 1, "lead", 3, "text", "size", "vertical"], ["fail", ""], ["icon", "check-circle", 1, "text-success"], ["icon", "ban", 1, "text-danger"], ["translate", "", 3, "translateParams"], [1, "text-danger", 3, "icon"]], template: function BulkDeviceRegistrationModalComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-modal", 0);
        ɵngcc0.ɵɵelementContainerStart(1, 1);
        ɵngcc0.ɵɵelement(2, "span", 2);
        ɵngcc0.ɵɵelementStart(3, "h4");
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementContainerEnd();
        ɵngcc0.ɵɵelementStart(6, "c8y-stepper", 3);
        ɵngcc0.ɵɵelementStart(7, "cdk-step", 4);
        ɵngcc0.ɵɵelementStart(8, "div", 5);
        ɵngcc0.ɵɵelementStart(9, "p", 6);
        ɵngcc0.ɵɵtext(10, " Register general devices in bulk ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "c8y-form-group", 7);
        ɵngcc0.ɵɵelement(12, "formly-form", 8);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(13, "div", 9);
        ɵngcc0.ɵɵelementStart(14, "div");
        ɵngcc0.ɵɵelementStart(15, "p", 10);
        ɵngcc0.ɵɵtext(16, "Simple registration");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(17, "p", 11);
        ɵngcc0.ɵɵtext(18, " Creates all registration requests at once, then each one needs to go through regular acceptance process. ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(19, "a", 12);
        ɵngcc0.ɵɵlistener("click", function BulkDeviceRegistrationModalComponent_Template_a_click_19_listener() { return ctx.downloadSimple(); });
        ɵngcc0.ɵɵpipe(20, "translate");
        ɵngcc0.ɵɵelement(21, "i", 13);
        ɵngcc0.ɵɵelementStart(22, "span", 14);
        ɵngcc0.ɵɵtext(23, "Download template");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(24, "div", 9);
        ɵngcc0.ɵɵelementStart(25, "div");
        ɵngcc0.ɵɵelementStart(26, "p", 10);
        ɵngcc0.ɵɵtext(27, "Full registration");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(28, "p", 11);
        ɵngcc0.ɵɵtext(29, " Creates all device credentials and devices using provided list of property values. Devices can start communicating with the platform immediately. ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(30, "a", 12);
        ɵngcc0.ɵɵlistener("click", function BulkDeviceRegistrationModalComponent_Template_a_click_30_listener() { return ctx.downloadFull(); });
        ɵngcc0.ɵɵpipe(31, "translate");
        ɵngcc0.ɵɵelement(32, "i", 13);
        ɵngcc0.ɵɵelementStart(33, "span", 14);
        ɵngcc0.ɵɵtext(34, "Download template");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(35, "c8y-stepper-buttons", 15);
        ɵngcc0.ɵɵlistener("onCancel", function BulkDeviceRegistrationModalComponent_Template_c8y_stepper_buttons_onCancel_35_listener() { return ctx.cancel(); })("onNext", function BulkDeviceRegistrationModalComponent_Template_c8y_stepper_buttons_onNext_35_listener() { return ctx.upload(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(36, "cdk-step", 16);
        ɵngcc0.ɵɵelementStart(37, "div", 17);
        ɵngcc0.ɵɵtemplate(38, BulkDeviceRegistrationModalComponent_div_38_Template, 3, 5, "div", 18);
        ɵngcc0.ɵɵtemplate(39, BulkDeviceRegistrationModalComponent_ng_template_39_Template, 2, 5, "ng-template", null, 19, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementStart(41, "c8y-list-group", 20);
        ɵngcc0.ɵɵtemplate(42, BulkDeviceRegistrationModalComponent_ng_container_42_Template, 4, 2, "ng-container", 18);
        ɵngcc0.ɵɵtemplate(43, BulkDeviceRegistrationModalComponent_ng_template_43_Template, 10, 7, "ng-template", null, 21, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(45, "c8y-stepper-buttons", 22);
        ɵngcc0.ɵɵlistener("onNext", function BulkDeviceRegistrationModalComponent_Template_c8y_stepper_buttons_onNext_45_listener() { return ctx.complete(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r1 = ɵngcc0.ɵɵreference(40);
        const _r4 = ɵngcc0.ɵɵreference(44);
        ɵngcc0.ɵɵproperty("headerClasses", "dialog-header")("customFooter", true);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(5, 20, "Bulk device registration"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("hideStepProgress", true);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("stepControl", ctx.form);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("form", ctx.form)("fields", ctx.template)("model", ctx.model);
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(20, 22, "Download template"));
        ɵngcc0.ɵɵadvance(11);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(31, 24, "Download template"));
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("showButtons", ɵngcc0.ɵɵpureFunction0(26, _c1))("disabled", ctx.form.invalid)("pending", ctx.pending)("labels", ɵngcc0.ɵɵpureFunction0(27, _c2));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.success)("ngIfElse", _r1);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngIf", ctx.result)("ngIfElse", _r4);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("showButtons", ɵngcc0.ɵɵpureFunction0(28, _c3))("labels", ɵngcc0.ɵɵpureFunction1(29, _c4, ctx.success ? "Close" : "Cancel"));
    } }, directives: [ɵngcc2.ModalComponent, ɵngcc2.IconDirective, ɵngcc2.C8yStepper, ɵngcc6.CdkStep, ɵngcc2.C8yTranslateDirective, ɵngcc2.FormGroupComponent, ɵngcc7.FormlyForm, ɵngcc2.C8yStepperButtons, ɵngcc8.NgIf, ɵngcc2.ListGroupComponent, ɵngcc2.OperationResultComponent, ɵngcc2.ListItemComponent, ɵngcc2.ListItemIconComponent, ɵngcc2.ListItemCollapseComponent], pipes: [ɵngcc2.C8yTranslatePipe, ɵngcc8.JsonPipe], encapsulation: 2 });
BulkDeviceRegistrationModalComponent.ctorParameters = () => [
    { type: Router },
    { type: C8yJSONSchema },
    { type: DeviceRegistrationBulkService },
    { type: RegisterDeviceService },
    { type: BsModalRef }
];
BulkDeviceRegistrationModalComponent.propDecorators = {
    stepper: [{ type: ViewChild, args: [C8yStepper, { static: true },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BulkDeviceRegistrationModalComponent, [{
        type: Component,
        args: [{
                selector: 'bulk-device-registration',
                template: "<c8y-modal [headerClasses]=\"'dialog-header'\" [customFooter]=\"true\">\n  <ng-container c8y-modal-title>\n    <span c8yIcon=\"upload\"></span>\n    <h4>{{ 'Bulk device registration' | translate }}</h4>\n  </ng-container>\n  <c8y-stepper [hideStepProgress]=\"true\" linear>\n    <cdk-step [stepControl]=\"form\">\n      <div class=\"sticky-top bg-component separator-bottom\">\n        <p\n          class=\"\n            p-l-24 p-r-24 p-t-16 p-b-16\n            m-b-0\n            sticky-top\n            separator-bottom\n            text-16 text-medium text-center\n            bg-component\n          \"\n          translate\n        >\n          Register general devices in bulk\n        </p>\n        <c8y-form-group class=\"p-24 p-t-16 p-b-8 m-b-0\">\n          <formly-form [form]=\"form\" [fields]=\"template\" [model]=\"model\"></formly-form>\n        </c8y-form-group>\n      </div>\n      <div class=\"bg-gray-white separator-bottom p-t-16 p-b-16 p-l-24 p-r-24\">\n        <div>\n          <p class=\"m-b-8 text-medium\" translate>Simple registration</p>\n          <p class=\"small m-b-8\" translate>\n            Creates all registration requests at once, then each one needs to go through regular\n            acceptance process.\n          </p>\n        </div>\n        <a\n          title=\"{{ 'Download template' | translate }}\"\n          class=\"btn btn-default btn-sm\"\n          target=\"_self\"\n          (click)=\"downloadSimple()\"\n        >\n          <i c8yIcon=\"download\"></i>\n          <span translate>Download template</span>\n        </a>\n      </div>\n      <div class=\"bg-gray-white separator-bottom p-t-16 p-b-16 p-l-24 p-r-24\">\n        <div>\n          <p class=\"m-b-8 text-medium\" translate>Full registration</p>\n          <p class=\"small m-b-8\" translate>\n            Creates all device credentials and devices using provided list of property values.\n            Devices can start communicating with the platform immediately.\n          </p>\n        </div>\n        <a\n          title=\"{{ 'Download template' | translate }}\"\n          class=\"btn btn-default btn-sm\"\n          target=\"_self\"\n          (click)=\"downloadFull()\"\n        >\n          <i c8yIcon=\"download\"></i>\n          <span translate>Download template</span>\n        </a>\n      </div>\n      <c8y-stepper-buttons\n        class=\"sticky-bottom d-block p-t-16 p-b-16 separator-top bg-white\"\n        [showButtons]=\"{ cancel: true, next: true }\"\n        [disabled]=\"form.invalid\"\n        [pending]=\"pending\"\n        (onCancel)=\"cancel()\"\n        (onNext)=\"upload()\"\n        [labels]=\"{ next: 'Upload' }\"\n      ></c8y-stepper-buttons>\n    </cdk-step>\n    <cdk-step state=\"final\">\n      <div class=\"m-24\">\n        <div *ngIf=\"success; else warning\">\n          <c8y-operation-result\n            text=\"{{ message | translate }}\"\n            [size]=\"84\"\n            [vertical]=\"true\"\n            type=\"success\"\n            class=\"lead\"\n          ></c8y-operation-result>\n        </div>\n        <ng-template #warning>\n          <c8y-operation-result\n            text=\"{{ message | translate }}\"\n            [size]=\"84\"\n            [vertical]=\"true\"\n            type=\"error\"\n            class=\"lead\"\n          ></c8y-operation-result>\n        </ng-template>\n        <c8y-list-group class=\"separator-top m-t-16\">\n          <ng-container *ngIf=\"result; else failedResponse\">\n            <c8y-li *ngIf=\"success; else fail\">\n              <c8y-li-icon class=\"text-success\" icon=\"check-circle\"></c8y-li-icon>\n              <p>All devices processed successfully.</p>\n              <c8y-li-collapse>\n                <pre><code>{{ result | json }}</code></pre>\n              </c8y-li-collapse>\n            </c8y-li>\n            <ng-template #fail>\n              <c8y-li>\n                <c8y-li-icon class=\"text-danger\" icon=\"ban\"></c8y-li-icon>\n                <p\n                  ngNonBindable\n                  [translateParams]=\"{ count: result?.numberOfFailed, total: result?.numberOfAll }\"\n                  translate\n                >\n                  Failed to process {{ count }} out of {{ total }}.\n                </p>\n                <c8y-li-collapse>\n                  <pre><code>{{ result | json }}</code></pre>\n                </c8y-li-collapse>\n              </c8y-li>\n            </ng-template>\n          </ng-container>\n          <ng-template #failedResponse>\n            <c8y-li>\n              <c8y-li-icon class=\"text-danger\" [icon]=\"'ban'\"></c8y-li-icon>\n              <small>{{ failedResult?.message | translate }}</small>\n              <c8y-li-collapse>\n                <pre><code>{{ failedResult | json }}</code></pre>\n              </c8y-li-collapse>\n            </c8y-li>\n          </ng-template>\n        </c8y-list-group>\n      </div>\n      <c8y-stepper-buttons\n        class=\"sticky-bottom d-block p-t-16 p-b-16 separator-top bg-white\"\n        [showButtons]=\"{ next: true }\"\n        (onNext)=\"complete()\"\n        [labels]=\"{ next: success ? 'Close' : 'Cancel' }\"\n      ></c8y-stepper-buttons>\n    </cdk-step>\n  </c8y-stepper>\n</c8y-modal>\n"
            }]
    }], function () { return [{ type: ɵngcc1.Router }, { type: ɵngcc2.C8yJSONSchema }, { type: ɵngcc3.DeviceRegistrationBulkService }, { type: ɵngcc4.RegisterDeviceService }, { type: ɵngcc5.BsModalRef }]; }, { stepper: [{
            type: ViewChild,
            args: [C8yStepper, { static: true }]
        }] }); })();
