// @c8y/ngx-components/esm2015/register-device/bulk/bulk-device-registration-button.component.js
import { Component } from '@angular/core';
import { BsModalService } from 'ngx-bootstrap/modal';
import { BulkDeviceRegistrationModalComponent } from './bulk-device-registration-modal.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/modal';
import * as ɵngcc2 from '@c8y/ngx-components';
export class BulkDeviceRegistrationButtonComponent {
    constructor(modalService) {
        this.modalService = modalService;
    }
    click() {
        this.modalService.show(BulkDeviceRegistrationModalComponent, {
            class: 'modal-sm',
            ignoreBackdropClick: true
        });
    }
}
BulkDeviceRegistrationButtonComponent.ɵfac = function BulkDeviceRegistrationButtonComponent_Factory(t) { return new (t || BulkDeviceRegistrationButtonComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BsModalService)); };
BulkDeviceRegistrationButtonComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: BulkDeviceRegistrationButtonComponent, selectors: [["extensible-device-registration-component"]], decls: 4, vars: 3, consts: [[3, "click"], ["c8yIcon", "upload"]], template: function BulkDeviceRegistrationButtonComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "button", 0);
        ɵngcc0.ɵɵlistener("click", function BulkDeviceRegistrationButtonComponent_Template_button_click_0_listener() { return ctx.click(); });
        ɵngcc0.ɵɵelement(1, "i", 1);
        ɵngcc0.ɵɵtext(2);
        ɵngcc0.ɵɵpipe(3, "translate");
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(3, 1, "General"));
    } }, directives: [ɵngcc2.IconDirective], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
BulkDeviceRegistrationButtonComponent.ctorParameters = () => [
    { type: BsModalService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BulkDeviceRegistrationButtonComponent, [{
        type: Component,
        args: [{
                selector: 'extensible-device-registration-component',
                template: "<button (click)=\"click()\"><i c8yIcon=\"upload\"></i>{{ 'General' | translate }}</button>\n"
            }]
    }], function () { return [{ type: ɵngcc1.BsModalService }]; }, null); })();
