// @c8y/ngx-components/esm2015/repository/software/list/software-repository-list.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NameCellRendererComponent } from './columns/name.cell-renderer.component';
import { SoftwareTypeCellRendererComponent } from './columns/software-type.cell-renderer.component';
import { SoftwareTypeFilteringFormRendererComponent } from './columns/software-type.filtering-form-renderer.component';
import { DeviceTypeCellRendererComponent } from './columns/device-type.cell-renderer.component';
import { VersionsCellRendererComponent } from './columns/versions.cell-renderer.component';
import { SoftwareListComponent } from './software-list.component';
import { SoftwareDetailsComponent } from './software-details.component';
import { AddSoftwareModalComponent } from './add-software-modal.component';
import { CoreModule, FormsModule, HOOK_NAVIGATOR_NODES, HOOK_ROUTE } from '@c8y/ngx-components';
import { SoftwareRepositoryNavigationFactory } from './software-repository-navigation-factory';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { RouterModule } from '@angular/router';
import { SharedRepositoryModule } from '@c8y/ngx-components/repository/shared';
import { DeviceGridModule } from '@c8y/ngx-components/device-grid';
export class SoftwareRepositoryListModule {
    static forRoot() {
        return {
            ngModule: SoftwareRepositoryListModule,
            providers: [
                {
                    provide: HOOK_NAVIGATOR_NODES,
                    useClass: SoftwareRepositoryNavigationFactory,
                    multi: true
                },
                {
                    provide: HOOK_ROUTE,
                    useValue: [
                        {
                            path: 'software',
                            component: SoftwareListComponent
                        },
                        {
                            path: 'software/:id',
                            component: SoftwareDetailsComponent
                        }
                    ],
                    multi: true
                }
            ]
        };
    }
}
SoftwareRepositoryListModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    CoreModule,
                    FormsModule,
                    DeviceGridModule,
                    PopoverModule,
                    RouterModule,
                    SharedRepositoryModule
                ],
                declarations: [
                    SoftwareListComponent,
                    SoftwareDetailsComponent,
                    AddSoftwareModalComponent,
                    NameCellRendererComponent,
                    SoftwareTypeCellRendererComponent,
                    SoftwareTypeFilteringFormRendererComponent,
                    DeviceTypeCellRendererComponent,
                    VersionsCellRendererComponent
                ],
                entryComponents: [SoftwareListComponent, SoftwareDetailsComponent, AddSoftwareModalComponent]
            },] }
];
