// @c8y/ngx-components/esm2015/core/wizard/wizard-outlet/wizard-outlet.component.js
import { Component, ComponentFactoryResolver, EventEmitter, Host, Input, Output, ViewChild, ViewContainerRef } from '@angular/core';
import { Router } from '@angular/router';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { WizardComponent } from '../wizard.component';
import { WizardService } from '../wizard.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../wizard.component';
import * as ɵngcc2 from '@angular/router';
import * as ɵngcc3 from '../wizard.service';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from '../../i18n/c8y-translate.directive';
import * as ɵngcc6 from '../../common/icon.directive';
import * as ɵngcc7 from '../../i18n/c8y-translate.pipe';

const _c0 = ["container"];
function WizardOutletComponent_div_1_div_4_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r8 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 10);
    ɵngcc0.ɵɵlistener("click", function WizardOutletComponent_div_1_div_4_button_1_Template_button_click_0_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r8); const entry_r6 = restoredCtx.$implicit; const ctx_r7 = ɵngcc0.ɵɵnextContext(3); return ctx_r7.handleEntry(entry_r6); });
    ɵngcc0.ɵɵelementStart(1, "div", 11);
    ɵngcc0.ɵɵelement(2, "i", 12);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div", 13);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const entry_r6 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", entry_r6.c8yIcon || "cogs");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(4, 3, entry_r6.name));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(6, 5, entry_r6.name), " ");
} }
function WizardOutletComponent_div_1_div_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 8);
    ɵngcc0.ɵɵtemplate(1, WizardOutletComponent_div_1_div_4_button_1_Template, 7, 7, "button", 9);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r2.entries);
} }
function WizardOutletComponent_div_1_ng_template_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 14);
    ɵngcc0.ɵɵelementStart(1, "p");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(3, 1, "No options available."));
} }
function WizardOutletComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 3);
    ɵngcc0.ɵɵelementStart(1, "p", 4);
    ɵngcc0.ɵɵelementStart(2, "span", 5);
    ɵngcc0.ɵɵtext(3, "Select method");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(4, WizardOutletComponent_div_1_div_4_Template, 2, 1, "div", 6);
    ɵngcc0.ɵɵtemplate(5, WizardOutletComponent_div_1_ng_template_5_Template, 4, 3, "ng-template", null, 7, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _r3 = ɵngcc0.ɵɵreference(6);
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.entries.length > 0)("ngIfElse", _r3);
} }
export class WizardOutletComponent {
    constructor(wizardComponent, router, c8yWizardService, resolver) {
        this.wizardComponent = wizardComponent;
        this.router = router;
        this.c8yWizardService = c8yWizardService;
        this.resolver = resolver;
        /**
         * The event that is emitted when the wizard entry is clicked.
         */
        this.onSelect = new EventEmitter();
        /**
         * The initial state to be assigned to the created component.
         */
        this.initialState = {};
        /**
         * List of wizard entries.
         */
        this.entries = [];
        /**
         * @ignore
         */
        this.showList = true;
        /**
         * @ignore
         */
        this.destroy$ = new Subject();
    }
    ngAfterViewInit() {
        this.c8yWizardService
            .getEntriesListById(this.wizardComponent.id)
            .pipe(takeUntil(this.destroy$))
            .subscribe((entries) => {
            this.container.clear();
            this.entries = entries;
            this.onEntriesLoaded();
        });
        this.wizardComponent.onReset.pipe(takeUntil(this.destroy$)).subscribe(() => {
            if (this.componentInstance) {
                this.componentInstance = undefined;
                this.container.clear();
                this.showList = true;
            }
        });
    }
    /**
     * Handles the wizard entry, if the entry contains a path, the user will be redirected to it.
     * Otherwise, an attempt will be made to create a dynamic component provided in the entry.
     * @param entry Wizard menu entry.
     */
    handleEntry(entry = {}) {
        const { path, component } = entry;
        if (path) {
            this.handlePath(path);
            return;
        }
        if (component) {
            this.handleComponent(component);
            this.onSelect.emit(entry);
        }
    }
    ngOnDestroy() {
        if (this.componentInstance) {
            this.container.clear();
        }
        this.destroy$.next(true);
        this.destroy$.unsubscribe();
    }
    onEntriesLoaded() {
        const [firstEntry] = this.entries;
        const isSingleEntry = this.entries.length === 1;
        const shouldRedirectToPath = firstEntry && !!firstEntry.path;
        const shouldOmitListOfEntries = isSingleEntry && !shouldRedirectToPath;
        if (shouldOmitListOfEntries) {
            this.handleEntry(firstEntry);
        }
    }
    handlePath(path) {
        this.router.navigateByUrl(path);
        if (this.wizardComponent) {
            this.wizardComponent.close();
        }
    }
    handleComponent(component) {
        this.showList = false;
        const factory = this.resolver.resolveComponentFactory(component);
        const componentInstance = this.container.createComponent(factory);
        Object.assign(componentInstance.instance, this.initialState);
        this.componentInstance = componentInstance;
    }
}
WizardOutletComponent.ɵfac = function WizardOutletComponent_Factory(t) { return new (t || WizardOutletComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.WizardComponent, 1), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.Router), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.WizardService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ComponentFactoryResolver)); };
WizardOutletComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: WizardOutletComponent, selectors: [["c8y-wizard-outlet"]], viewQuery: function WizardOutletComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 5, ViewContainerRef);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.container = _t.first);
    } }, inputs: { initialState: "initialState" }, outputs: { onSelect: "onSelect" }, decls: 4, vars: 2, consts: [["class", "modal-inner-scroll animated fadeIn", 4, "ngIf"], [3, "hidden"], ["container", ""], [1, "modal-inner-scroll", "animated", "fadeIn"], [1, "p-16", "text-center", "text-medium", "separator-bottom"], ["translate", ""], ["class", "list-group list-group-links", 4, "ngIf", "ngIfElse"], ["emptyState", ""], [1, "list-group", "list-group-links"], ["type", "button", "class", "list-group-item d-flex a-i-center p-t-8 p-b-8", 3, "click", 4, "ngFor", "ngForOf"], ["type", "button", 1, "list-group-item", "d-flex", "a-i-center", "p-t-8", "p-b-8", 3, "click"], [1, "list-item-icon"], [3, "c8yIcon"], [1, "list-item-body", "text-truncate", 3, "title"], [1, "c8y-empty-state"]], template: function WizardOutletComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementContainerStart(0);
        ɵngcc0.ɵɵtemplate(1, WizardOutletComponent_div_1_Template, 7, 2, "div", 0);
        ɵngcc0.ɵɵelementContainerEnd();
        ɵngcc0.ɵɵelement(2, "template", 1, 2);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showList);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("hidden", ctx.showList);
    } }, directives: [ɵngcc4.NgIf, ɵngcc5.C8yTranslateDirective, ɵngcc4.NgForOf, ɵngcc6.IconDirective], pipes: [ɵngcc7.C8yTranslatePipe], encapsulation: 2 });
WizardOutletComponent.ctorParameters = () => [
    { type: WizardComponent, decorators: [{ type: Host }] },
    { type: Router },
    { type: WizardService },
    { type: ComponentFactoryResolver }
];
WizardOutletComponent.propDecorators = {
    container: [{ type: ViewChild, args: ['container', { read: ViewContainerRef },] }],
    onSelect: [{ type: Output }],
    initialState: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(WizardOutletComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-wizard-outlet',
                template: "<ng-container>\n  <div class=\"modal-inner-scroll animated fadeIn\" *ngIf=\"showList\">\n    <p class=\"p-16 text-center text-medium separator-bottom\">\n      <span translate>Select method</span>\n    </p>\n    <div class=\"list-group list-group-links\" *ngIf=\"entries.length > 0; else emptyState\">\n      <button\n        *ngFor=\"let entry of entries\"\n        type=\"button\"\n        class=\"list-group-item d-flex a-i-center p-t-8 p-b-8\"\n        (click)=\"handleEntry(entry)\"\n      >\n        <div class=\"list-item-icon\">\n          <i [c8yIcon]=\"entry.c8yIcon || 'cogs'\"></i>\n        </div>\n        <div title=\"{{ entry.name | translate }}\" class=\"list-item-body text-truncate\">\n          {{ entry.name | translate }}\n        </div>\n      </button>\n    </div>\n\n    <ng-template #emptyState>\n      <div class=\"c8y-empty-state\">\n        <p>{{ 'No options available.' | translate }}</p>\n      </div>\n    </ng-template>\n  </div>\n</ng-container>\n<template #container [hidden]=\"showList\"></template>\n"
            }]
    }], function () { return [{ type: ɵngcc1.WizardComponent, decorators: [{
                type: Host
            }] }, { type: ɵngcc2.Router }, { type: ɵngcc3.WizardService }, { type: ɵngcc0.ComponentFactoryResolver }]; }, { onSelect: [{
            type: Output
        }], initialState: [{
            type: Input
        }], container: [{
            type: ViewChild,
            args: ['container', { read: ViewContainerRef }]
        }] }); })();
