// @c8y/ngx-components/esm2015/core/forms/forms.module.js
import { NgModule } from '@angular/core';
import { FormsModule as NgFormsModule, ReactiveFormsModule } from '@angular/forms';
import { CommonModule } from '../common/common.module';
import { DefaultValidationDirective } from './default-validation.directive';
import { EmailsValidatorDirective } from './emails-validator.directive';
import { FilterInputComponent } from './filter-input.component';
import { FormGroupComponent } from './form-group.component';
import { InputGroupListContainerDirective } from './input-group-list-container.directive';
import { InputGroupListComponent } from './input-group-list.component';
import { IpRangeInputListComponent } from './ip-range-input-list.component';
import { JsonValidationPrettifierDirective } from './json-validation-prettifier.directive';
import { MaxValidationDirective } from './max-validation.directive';
import { MessageDirective } from './message.directive';
import { MessagesComponent } from './messages.component';
import { MinValidationDirective } from './min-validation.directive';
import { PhoneValidationDirective } from './phone-validation.directive';
import { RangeComponent } from './range.component';
import { RangeDirective } from './range.directive';
import { RequiredInputPlaceholderDirective } from './required-input-placeholder.directive';
import { TextAreaRowHeightDirective } from './text-area-row-height.directive';
import { UniqueInCollectionByPathValidationDirective } from './unique-in-collection-by-path-validation.directive';
import { PopoverModule } from 'ngx-bootstrap/popover';
/**
 * @exports FormGroupComponent A component that allows to group an input and can display warnings and errors.
 * @exports PhoneValidationDirective Allows to validate international phone numbers.
 * @exports NewsletterDirective Allows to subscribe to the newsletter.
 * @exports DefaultValidationDirective Allows to validate some predefined patterns.
 * @exports MessageDirective A directive that allows to define messages.
 * @exports MessagesComponent Allows to display error messages.
 * @exports RangeComponent A component that allows to display a styled range slider.
 */
import * as ɵngcc0 from '@angular/core';
export class FormsModule {
}
FormsModule.ɵfac = function FormsModule_Factory(t) { return new (t || FormsModule)(); };
FormsModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: FormsModule });
FormsModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule, NgFormsModule, ReactiveFormsModule, PopoverModule], NgFormsModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(FormsModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, NgFormsModule, ReactiveFormsModule, PopoverModule],
                exports: [
                    PhoneValidationDirective,
                    DefaultValidationDirective,
                    MinValidationDirective,
                    MaxValidationDirective,
                    NgFormsModule,
                    FormGroupComponent,
                    MessageDirective,
                    MessagesComponent,
                    InputGroupListComponent,
                    InputGroupListContainerDirective,
                    JsonValidationPrettifierDirective,
                    TextAreaRowHeightDirective,
                    RequiredInputPlaceholderDirective,
                    IpRangeInputListComponent,
                    FilterInputComponent,
                    EmailsValidatorDirective,
                    RangeDirective,
                    RangeComponent,
                    UniqueInCollectionByPathValidationDirective
                ],
                declarations: [
                    PhoneValidationDirective,
                    DefaultValidationDirective,
                    MinValidationDirective,
                    MaxValidationDirective,
                    FormGroupComponent,
                    MessageDirective,
                    MessagesComponent,
                    InputGroupListComponent,
                    InputGroupListContainerDirective,
                    JsonValidationPrettifierDirective,
                    TextAreaRowHeightDirective,
                    RequiredInputPlaceholderDirective,
                    IpRangeInputListComponent,
                    FilterInputComponent,
                    EmailsValidatorDirective,
                    RangeDirective,
                    RangeComponent,
                    UniqueInCollectionByPathValidationDirective
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(FormsModule, { declarations: function () { return [PhoneValidationDirective,
        DefaultValidationDirective,
        MinValidationDirective,
        MaxValidationDirective,
        FormGroupComponent,
        MessageDirective,
        MessagesComponent,
        InputGroupListComponent,
        InputGroupListContainerDirective,
        JsonValidationPrettifierDirective,
        TextAreaRowHeightDirective,
        RequiredInputPlaceholderDirective,
        IpRangeInputListComponent,
        FilterInputComponent,
        EmailsValidatorDirective,
        RangeDirective,
        RangeComponent,
        UniqueInCollectionByPathValidationDirective]; }, imports: function () { return [CommonModule, NgFormsModule, ReactiveFormsModule, PopoverModule]; }, exports: function () { return [PhoneValidationDirective,
        DefaultValidationDirective,
        MinValidationDirective,
        MaxValidationDirective,
        NgFormsModule,
        FormGroupComponent,
        MessageDirective,
        MessagesComponent,
        InputGroupListComponent,
        InputGroupListContainerDirective,
        JsonValidationPrettifierDirective,
        TextAreaRowHeightDirective,
        RequiredInputPlaceholderDirective,
        IpRangeInputListComponent,
        FilterInputComponent,
        EmailsValidatorDirective,
        RangeDirective,
        RangeComponent,
        UniqueInCollectionByPathValidationDirective]; } }); })();
