// @c8y/ngx-components/esm2015/core/forms/max-validation.directive.js
import { NG_VALIDATORS } from '@angular/forms';
import { Directive, Input } from '@angular/core';
import { isUndefined, isFinite } from 'lodash-es';
import * as ɵngcc0 from '@angular/core';
export class MaxValidationDirective {
    constructor() {
        this.max = '';
    }
    validate(control) {
        const parsedMax = typeof this.max === 'number' ? this.max : parseFloat(this.max);
        if (isUndefined(control.value) || !isFinite(parsedMax)) {
            return null;
        }
        const valid = !(control.value > parsedMax);
        return valid ? null : { max: { max: parsedMax, actual: control.value } };
    }
}
MaxValidationDirective.ɵfac = function MaxValidationDirective_Factory(t) { return new (t || MaxValidationDirective)(); };
MaxValidationDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: MaxValidationDirective, selectors: [["", "max", ""]], inputs: { max: "max" }, features: [ɵngcc0.ɵɵProvidersFeature([{ provide: NG_VALIDATORS, useExisting: MaxValidationDirective, multi: true }])] });
MaxValidationDirective.propDecorators = {
    max: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(MaxValidationDirective, [{
        type: Directive,
        args: [{
                selector: '[max]',
                providers: [{ provide: NG_VALIDATORS, useExisting: MaxValidationDirective, multi: true }]
            }]
    }], function () { return []; }, { max: [{
            type: Input
        }] }); })();
