// @c8y/ngx-components/esm2015/core/forms/min-validation.directive.js
import { NG_VALIDATORS } from '@angular/forms';
import { Directive, Input } from '@angular/core';
import { isUndefined, isFinite } from 'lodash-es';
import * as ɵngcc0 from '@angular/core';
export class MinValidationDirective {
    constructor() {
        this.min = '';
    }
    validate(control) {
        const parsedMin = typeof this.min === 'number' ? this.min : parseFloat(this.min);
        if (isUndefined(control.value) || !isFinite(parsedMin)) {
            return null;
        }
        const valid = (control.value !== null) ? !(control.value < parsedMin) : true;
        return valid ? null : { min: { min: parsedMin, actual: control.value } };
    }
}
MinValidationDirective.ɵfac = function MinValidationDirective_Factory(t) { return new (t || MinValidationDirective)(); };
MinValidationDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: MinValidationDirective, selectors: [["", "min", ""]], inputs: { min: "min" }, features: [ɵngcc0.ɵɵProvidersFeature([{ provide: NG_VALIDATORS, useExisting: MinValidationDirective, multi: true }])] });
MinValidationDirective.propDecorators = {
    min: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(MinValidationDirective, [{
        type: Directive,
        args: [{
                selector: '[min]',
                providers: [{ provide: NG_VALIDATORS, useExisting: MinValidationDirective, multi: true }]
            }]
    }], function () { return []; }, { min: [{
            type: Input
        }] }); })();
