// @c8y/ngx-components/esm2015/core/forms/phone-validation.directive.js
import { NG_VALIDATORS } from '@angular/forms';
import { Directive } from '@angular/core';
import { parsePhoneNumberFromString } from 'libphonenumber-js/max';
import * as ɵngcc0 from '@angular/core';
export class PhoneValidationDirective {
    validate(control) {
        const valid = this.validateInternationalPhoneNumber(control);
        return valid ? null : { internationalPhoneNumber: true };
    }
    validateInternationalPhoneNumber(control) {
        let numberValid;
        try {
            const phoneNumber = parsePhoneNumberFromString(control.value);
            numberValid = phoneNumber.isValid();
            if (numberValid && control.value !== phoneNumber.format('E.164')) {
                control.setValue(phoneNumber.format('E.164'), {
                    emitEvent: false,
                    emitModelToViewChange: false
                });
            }
        }
        catch (e) {
            numberValid = !control.value;
        }
        return numberValid;
    }
}
PhoneValidationDirective.ɵfac = function PhoneValidationDirective_Factory(t) { return new (t || PhoneValidationDirective)(); };
PhoneValidationDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: PhoneValidationDirective, selectors: [["", "c8yPhoneValidation", ""]], features: [ɵngcc0.ɵɵProvidersFeature([{ provide: NG_VALIDATORS, useExisting: PhoneValidationDirective, multi: true }])] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(PhoneValidationDirective, [{
        type: Directive,
        args: [{
                selector: '[c8yPhoneValidation]',
                providers: [{ provide: NG_VALIDATORS, useExisting: PhoneValidationDirective, multi: true }]
            }]
    }], null, null); })();
