// @c8y/ngx-components/esm2015/core/forms/unique-in-collection-by-path-validation.directive.js
import { Directive, Input } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { find, get, isEqualWith } from 'lodash-es';
import * as ɵngcc0 from '@angular/core';
export class UniqueInCollectionByPathValidationDirective {
    validate(control) {
        const match = find(this.collection, element => isEqualWith(get(element, this.path), control.value, this.comparator));
        return !!match && !!control.value
            ? { uniqueByPath: { key: this.path, value: get(match, this.path) } }
            : null;
    }
}
UniqueInCollectionByPathValidationDirective.ɵfac = function UniqueInCollectionByPathValidationDirective_Factory(t) { return new (t || UniqueInCollectionByPathValidationDirective)(); };
UniqueInCollectionByPathValidationDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: UniqueInCollectionByPathValidationDirective, selectors: [["", "uniqueByPath", ""]], inputs: { collection: "collection", path: "path", comparator: "comparator" }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: NG_VALIDATORS,
                useExisting: UniqueInCollectionByPathValidationDirective,
                multi: true
            }
        ])] });
UniqueInCollectionByPathValidationDirective.propDecorators = {
    collection: [{ type: Input }],
    path: [{ type: Input }],
    comparator: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(UniqueInCollectionByPathValidationDirective, [{
        type: Directive,
        args: [{
                selector: '[uniqueByPath]',
                providers: [
                    {
                        provide: NG_VALIDATORS,
                        useExisting: UniqueInCollectionByPathValidationDirective,
                        multi: true
                    }
                ]
            }]
    }], null, { collection: [{
            type: Input
        }], path: [{
            type: Input
        }], comparator: [{
            type: Input
        }] }); })();
