// @c8y/ngx-components/esm2015/services/status-options/status-options.service.js
import { Inject, Injectable, Optional } from '@angular/core';
import { flatten } from 'lodash-es';
import { HOOK_SERVICE_STATUS, SERVICE_STATUS_OPTIONS } from './status-options.model';
import * as i0 from "@angular/core";
import * as i1 from "./status-options.model";
/**
 * The service for system wide registered service status options.
 * It provides the standard service statuses plus all custom status options
 * registered with the <code>HOOK_SERVICE_STATUS</code> extension hook.
 */
import * as ɵngcc0 from '@angular/core';
export class StatusOptionsService {
    constructor(options) {
        this._options = [...SERVICE_STATUS_OPTIONS, ...flatten(options)];
    }
    get options() {
        return this._options;
    }
}
StatusOptionsService.ɵfac = function StatusOptionsService_Factory(t) { return new (t || StatusOptionsService)(ɵngcc0.ɵɵinject(HOOK_SERVICE_STATUS, 8)); };
StatusOptionsService.ɵprov = i0.ɵɵdefineInjectable({ factory: function StatusOptionsService_Factory() { return new StatusOptionsService(i0.ɵɵinject(i1.HOOK_SERVICE_STATUS, 8)); }, token: StatusOptionsService, providedIn: "root" });
StatusOptionsService.ctorParameters = () => [
    { type: Array, decorators: [{ type: Optional }, { type: Inject, args: [HOOK_SERVICE_STATUS,] }] }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(StatusOptionsService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: Array, decorators: [{
                type: Optional
            }, {
                type: Inject,
                args: [HOOK_SERVICE_STATUS]
            }] }]; }, null); })();
