// @c8y/ngx-components/esm2015/core/action-bar/action-bar.component.js
import { Component, HostBinding, HostListener, Input } from '@angular/core';
import { filter, sortBy } from 'lodash-es';
import { of, Subject } from 'rxjs';
import { map, takeUntil } from 'rxjs/operators';
import { GainsightService } from '../product-experience/gainsight.service';
/**
 * This component is used as the outlet to show the action bars.
 * In a c8ycli app it is by default placed on the bootstrap component.
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../product-experience/gainsight.service';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from '../common/outlet.directive';
import * as ɵngcc4 from 'ngx-bootstrap/dropdown';
import * as ɵngcc5 from '../i18n/c8y-translate.directive';
import * as ɵngcc6 from '../common/icon.directive';
import * as ɵngcc7 from '../i18n/c8y-translate.pipe';

function ActionBarComponent_ng_container_7_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function ActionBarComponent_ng_container_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ActionBarComponent_ng_container_7_ng_container_1_Template, 1, 0, "ng-container", 8);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r3 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yOutlet", item_r3.template);
} }
function ActionBarComponent_ng_container_10_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function ActionBarComponent_ng_container_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ActionBarComponent_ng_container_10_ng_container_1_Template, 1, 0, "ng-container", 8);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r5 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yOutlet", item_r5.template);
} }
function ActionBarComponent_li_12_ng_container_8_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function ActionBarComponent_li_12_ng_container_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ActionBarComponent_li_12_ng_container_8_ng_container_1_Template, 1, 0, "ng-container", 8);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r9 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yOutlet", item_r9.template);
} }
function ActionBarComponent_li_12_ul_10_ng_container_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function ActionBarComponent_li_12_ul_10_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ActionBarComponent_li_12_ul_10_ng_container_1_ng_container_1_Template, 1, 0, "ng-container", 8);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r12 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yOutlet", item_r12.template);
} }
function ActionBarComponent_li_12_ul_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "ul", 16);
    ɵngcc0.ɵɵtemplate(1, ActionBarComponent_li_12_ul_10_ng_container_1_Template, 2, 1, "ng-container", 5);
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r8 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ɵngcc0.ɵɵpipeBind1(2, 1, ctx_r8.more$));
} }
function ActionBarComponent_li_12_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 9);
    ɵngcc0.ɵɵelementStart(1, "div", 10);
    ɵngcc0.ɵɵelementStart(2, "button", 11);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementStart(4, "span", 12);
    ɵngcc0.ɵɵtext(5, "More\u2026");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(6, "i", 13);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "ul", 14);
    ɵngcc0.ɵɵtemplate(8, ActionBarComponent_li_12_ng_container_8_Template, 2, 1, "ng-container", 5);
    ɵngcc0.ɵɵpipe(9, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(10, ActionBarComponent_li_12_ul_10_Template, 3, 3, "ul", 15);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate1("title", "", ɵngcc0.ɵɵpipeBind1(3, 3, "More"), "\u2026");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("c8yIcon", "caret-down");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", ɵngcc0.ɵɵpipeBind1(9, 5, ctx_r2.more$));
} }
export class ActionBarComponent {
    constructor(gainsightService) {
        this.gainsightService = gainsightService;
        /**
         * Identifies if the navigator is opened. If yes, the action-bar
         * needs to move to the right.
         */
        this.navigatorOpen = false;
        /**
         * Identifies if the current view has tabs. If yes, the action bar needs
         * to move down.
         */
        this.hasTabs = false;
        /**
         * Identifies if the tabs are aligned horizontally. If yes, the tabs don't
         * need to move to the left.
         */
        this.isTabsHorizontal = false;
        /**
         * Identifies if in a mobile view the toolbar is expanded or not.
         */
        this.isPageToolbarExpanded = false;
        /**
         * Adds the default `c8y-ui-action-bar` class.
         */
        this.uiActionBar = true;
        /**
         * The current items to display in the ActionBar.
         */
        this.items$ = of([]);
        /**
         * Provides observable for right action items.
         */
        this.right$ = of([]);
        /**
         * Provides observable for left action items.
         */
        this.left$ = of([]);
        /**
         * Provides observable for more action items.
         */
        this.more$ = of([]);
        /**
         * Returns true if no items are present and hides the action bar.
         */
        this.hidden$ = of(true);
        /**
         * Binds hidden property to this component's native element
         */
        this.hidden = true;
        this.destroy$ = new Subject();
    }
    mouseDown(event) {
        if (event.getAttribute('id') === 'page-toolbar') {
            return;
        }
        const element = event.closest('[px-event]') || event.closest('[title]') || event.closest('[uib-tooltip]');
        if (!element) {
            return;
        }
        const eventName = element.getAttribute('px-event') || element.getAttribute('title') || element.getAttribute('uib-tooltip');
        const translatedEventName = this.gainsightService.translateToEnglish(eventName);
        this.gainsightService.triggerEvent(`actionBarItem:${translatedEventName}`, { url: element.baseURI });
    }
    ngOnInit() {
        this.right$ = this.items$.pipe(map(items => filter(items, { placement: 'right' })), map(items => sortBy(items, this.byPriority)));
        this.left$ = this.items$.pipe(map(items => filter(items, { placement: 'left' })), map(items => sortBy(items, this.byPriority)));
        this.more$ = this.items$.pipe(map(items => filter(items, { placement: 'more' })), map(items => sortBy(items, this.byPriority)));
        this.hidden$ = this.items$.pipe(map(items => {
            return !(items && items.length);
        }));
        this.hidden$
            .pipe(takeUntil(this.destroy$)).subscribe(hidden => {
            this.hidden = hidden;
        });
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
    byPriority(item) {
        return -item.priority;
    }
}
ActionBarComponent.ɵfac = function ActionBarComponent_Factory(t) { return new (t || ActionBarComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.GainsightService)); };
ActionBarComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ActionBarComponent, selectors: [["c8y-action-bar"]], hostAttrs: [1, "c8y-ui-action-bar"], hostVars: 9, hostBindings: function ActionBarComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("mousedown", function ActionBarComponent_mousedown_HostBindingHandler($event) { return ctx.mouseDown($event.target); });
    } if (rf & 2) {
        ɵngcc0.ɵɵhostProperty("hidden", ctx.hidden);
        ɵngcc0.ɵɵclassProp("navigator-open", ctx.navigatorOpen)("has-tabs", ctx.hasTabs)("horizontal-tabs", ctx.isTabsHorizontal)("c8y-ui-action-bar", ctx.uiActionBar);
    } }, inputs: { navigatorOpen: "navigatorOpen", hasTabs: "hasTabs", isTabsHorizontal: "isTabsHorizontal", items$: "items$" }, decls: 14, vars: 12, consts: [[1, "navbar-header"], [1, "text-label-small", "p-l-16", "p-t-8", "p-b-8", "visible-xs"], [1, "text-primary"], ["id", "page-toolbar", 1, "navbar-collapse"], [1, "nav", "navbar-nav", "navbar-left"], [4, "ngFor", "ngForOf"], [1, "nav", "navbar-nav", "navbar-right"], ["class", "no-remove more-wrapper", 4, "ngIf"], [4, "c8yOutlet"], [1, "no-remove", "more-wrapper"], ["dropdown", "", 1, "dropdown"], ["dropdownToggle", "", "aria-haspopup", "true", "aria-expanded", "false", 1, "dropdown-toggle", "c8y-dropdown", "d-flex", 3, "title"], ["translate", "", 1, "text-truncate"], [3, "c8yIcon"], [1, "dropdown-menu", "visible-xs"], ["class", "dropdown-menu dropdown-menu-right hidden-xs", 4, "dropdownMenu"], [1, "dropdown-menu", "dropdown-menu-right", "hidden-xs"]], template: function ActionBarComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "p", 1);
        ɵngcc0.ɵɵelementStart(2, "span", 2);
        ɵngcc0.ɵɵtext(3);
        ɵngcc0.ɵɵpipe(4, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "div", 3);
        ɵngcc0.ɵɵelementStart(6, "ul", 4);
        ɵngcc0.ɵɵtemplate(7, ActionBarComponent_ng_container_7_Template, 2, 1, "ng-container", 5);
        ɵngcc0.ɵɵpipe(8, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "ul", 6);
        ɵngcc0.ɵɵtemplate(10, ActionBarComponent_ng_container_10_Template, 2, 1, "ng-container", 5);
        ɵngcc0.ɵɵpipe(11, "async");
        ɵngcc0.ɵɵtemplate(12, ActionBarComponent_li_12_Template, 11, 7, "li", 7);
        ɵngcc0.ɵɵpipe(13, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(4, 4, "Action bar"));
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngForOf", ɵngcc0.ɵɵpipeBind1(8, 6, ctx.left$));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngForOf", ɵngcc0.ɵɵpipeBind1(11, 8, ctx.right$));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(13, 10, ctx.more$).length > 0);
    } }, directives: [ɵngcc2.NgForOf, ɵngcc2.NgIf, ɵngcc3.OutletDirective, ɵngcc4.BsDropdownDirective, ɵngcc4.BsDropdownToggleDirective, ɵngcc5.C8yTranslateDirective, ɵngcc6.IconDirective, ɵngcc4.BsDropdownMenuDirective], pipes: [ɵngcc7.C8yTranslatePipe, ɵngcc2.AsyncPipe], encapsulation: 2 });
ActionBarComponent.ctorParameters = () => [
    { type: GainsightService }
];
ActionBarComponent.propDecorators = {
    navigatorOpen: [{ type: Input }, { type: HostBinding, args: ['class.navigator-open',] }],
    hasTabs: [{ type: Input }, { type: HostBinding, args: ['class.has-tabs',] }],
    isTabsHorizontal: [{ type: Input }, { type: HostBinding, args: ['class.horizontal-tabs',] }],
    uiActionBar: [{ type: HostBinding, args: ['class.c8y-ui-action-bar',] }],
    items$: [{ type: Input }],
    hidden: [{ type: HostBinding }],
    mouseDown: [{ type: HostListener, args: ['mousedown', ['$event.target'],] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ActionBarComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-action-bar',
                template: "<div class=\"navbar-header\">\n  <p class=\"text-label-small p-l-16 p-t-8 p-b-8 visible-xs\">\n    <span class=\"text-primary\">{{ 'Action bar' | translate }}</span>\n  </p>\n</div>\n\n<div id=\"page-toolbar\" class=\"navbar-collapse\">\n  <ul class=\"nav navbar-nav navbar-left\">\n    <ng-container *ngFor=\"let item of left$ | async\">\n      <ng-container *c8yOutlet=\"item.template\"></ng-container>\n    </ng-container>\n  </ul>\n  <ul class=\"nav navbar-nav navbar-right\">\n    <ng-container *ngFor=\"let item of right$ | async\">\n      <ng-container *c8yOutlet=\"item.template\"></ng-container>\n    </ng-container>\n    <li class=\"no-remove more-wrapper\" *ngIf=\"(more$ | async).length > 0\">\n      <div class=\"dropdown\" dropdown>\n        <button\n          title=\"{{ 'More' | translate }}\u2026\"\n          class=\"dropdown-toggle c8y-dropdown d-flex\"\n          dropdownToggle\n          aria-haspopup=\"true\"\n          aria-expanded=\"false\"\n        >\n          <span class=\"text-truncate\" translate>More\u2026</span>\n          <i [c8yIcon]=\"'caret-down'\"></i>\n        </button>\n        <ul class=\"dropdown-menu visible-xs\">\n          <ng-container *ngFor=\"let item of more$ | async\">\n            <ng-container *c8yOutlet=\"item.template\"></ng-container>\n          </ng-container>\n        </ul>\n        <ul class=\"dropdown-menu dropdown-menu-right hidden-xs\" *dropdownMenu>\n          <ng-container *ngFor=\"let item of more$ | async\">\n            <ng-container *c8yOutlet=\"item.template\"></ng-container>\n          </ng-container>\n        </ul>\n      </div>\n    </li>\n  </ul>\n</div>\n",
                host: { class: 'c8y-ui-action-bar' }
            }]
    }], function () { return [{ type: ɵngcc1.GainsightService }]; }, { navigatorOpen: [{
            type: Input
        }, {
            type: HostBinding,
            args: ['class.navigator-open']
        }], hasTabs: [{
            type: Input
        }, {
            type: HostBinding,
            args: ['class.has-tabs']
        }], isTabsHorizontal: [{
            type: Input
        }, {
            type: HostBinding,
            args: ['class.horizontal-tabs']
        }], uiActionBar: [{
            type: HostBinding,
            args: ['class.c8y-ui-action-bar']
        }], items$: [{
            type: Input
        }], hidden: [{
            type: HostBinding
        }], mouseDown: [{
            type: HostListener,
            args: ['mousedown', ['$event.target']]
        }] }); })();
