// @c8y/ngx-components/esm2015/core/search/search-input.component.js
import { __awaiter, __decorate } from "tslib";
import { Component, ViewChild, Input, Output, EventEmitter, HostListener } from '@angular/core';
import { Router } from '@angular/router';
import { InventoryService } from '@c8y/client';
import { TypeaheadComponent } from '../select/typeahead.component';
import { defer, empty, merge, of, pipe } from 'rxjs';
import { map, switchMap, tap } from 'rxjs/operators';
import { InventorySearchService } from './inventory-search.service';
import { AssetTypesService } from '../asset-types/asset-types.service';
import { memoize } from '../common/memoize.decorator';
import { property } from 'lodash-es';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from '@c8y/client';
import * as ɵngcc3 from './inventory-search.service';
import * as ɵngcc4 from '../asset-types/asset-types.service';
import * as ɵngcc5 from '@angular/common';
import * as ɵngcc6 from 'ngx-bootstrap/dropdown';
import * as ɵngcc7 from '../common/icon.directive';
import * as ɵngcc8 from '@angular/forms';
import * as ɵngcc9 from '../select/typeahead.component';
import * as ɵngcc10 from '../common/forOf.directive';
import * as ɵngcc11 from '../list-group/list-item.component';
import * as ɵngcc12 from '../i18n/c8y-translate.directive';
import * as ɵngcc13 from '../list-group/list-item-icon.component';
import * as ɵngcc14 from '../device-status/device-status.component';
import * as ɵngcc15 from '../i18n/c8y-translate.pipe';

const _c0 = ["dropdown"];
function SearchInputComponent_div_0_div_5_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function SearchInputComponent_div_0_div_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 8);
    ɵngcc0.ɵɵtemplate(1, SearchInputComponent_div_0_div_5_ng_container_1_Template, 1, 0, "ng-container", 9);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵnextContext(2);
    const _r2 = ɵngcc0.ɵɵreference(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r2);
} }
function SearchInputComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    const _r8 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 3, 4);
    ɵngcc0.ɵɵlistener("isOpenChange", function SearchInputComponent_div_0_Template_div_isOpenChange_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r7 = ɵngcc0.ɵɵnextContext(); return ctx_r7.onOpenChange($event); });
    ɵngcc0.ɵɵelementStart(2, "button", 5);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelement(4, "i", 6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(5, SearchInputComponent_div_0_div_5_Template, 2, 1, "div", 7);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("insideClick", true);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(3, 2, "Search"));
} }
function SearchInputComponent_div_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function SearchInputComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 10);
    ɵngcc0.ɵɵtemplate(1, SearchInputComponent_div_1_ng_container_1_Template, 1, 0, "ng-container", 9);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵnextContext();
    const _r2 = ɵngcc0.ɵɵreference(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r2);
} }
function SearchInputComponent_ng_template_2_c8y_li_4_div_1_Template(rf, ctx) { if (rf & 1) {
    const _r26 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 25);
    ɵngcc0.ɵɵelementStart(1, "p", 26);
    ɵngcc0.ɵɵelementStart(2, "em", 27);
    ɵngcc0.ɵɵtext(3, "Searching by exact match. Click for other search options:");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "div", 28);
    ɵngcc0.ɵɵelementStart(5, "button", 29);
    ɵngcc0.ɵɵlistener("click", function SearchInputComponent_ng_template_2_c8y_li_4_div_1_Template_button_click_5_listener() { ɵngcc0.ɵɵrestoreView(_r26); const ctx_r25 = ɵngcc0.ɵɵnextContext(3); return ctx_r25.onFilter(ctx_r25.term + "*"); });
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵtext(7);
    ɵngcc0.ɵɵpipe(8, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(9, "button", 29);
    ɵngcc0.ɵɵlistener("click", function SearchInputComponent_ng_template_2_c8y_li_4_div_1_Template_button_click_9_listener() { ɵngcc0.ɵɵrestoreView(_r26); const ctx_r27 = ɵngcc0.ɵɵnextContext(3); return ctx_r27.onFilter("*" + ctx_r27.term + "*"); });
    ɵngcc0.ɵɵpipe(10, "translate");
    ɵngcc0.ɵɵtext(11);
    ɵngcc0.ɵɵpipe(12, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(13, "button", 29);
    ɵngcc0.ɵɵlistener("click", function SearchInputComponent_ng_template_2_c8y_li_4_div_1_Template_button_click_13_listener() { ɵngcc0.ɵɵrestoreView(_r26); const ctx_r28 = ɵngcc0.ɵɵnextContext(3); return ctx_r28.onFilter("*" + ctx_r28.term); });
    ɵngcc0.ɵɵpipe(14, "translate");
    ɵngcc0.ɵɵtext(15);
    ɵngcc0.ɵɵpipe(16, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(6, 6, "Starts with"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(8, 8, "Starts with"), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(10, 10, "Contains"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(12, 12, "Contains"), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(14, 14, "Ends with"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(16, 16, "Ends with"), " ");
} }
function SearchInputComponent_ng_template_2_c8y_li_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 23);
    ɵngcc0.ɵɵtemplate(1, SearchInputComponent_ng_template_2_c8y_li_4_div_1_Template, 17, 18, "div", 24);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r11 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r11.mode === "search");
} }
function SearchInputComponent_ng_template_2_c8y_li_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 30);
    ɵngcc0.ɵɵelementStart(1, "div", 31);
    ɵngcc0.ɵɵelementStart(2, "span", 27);
    ɵngcc0.ɵɵtext(3, "Recent search views");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("selectable", false);
} }
function SearchInputComponent_ng_template_2_c8y_li_6_device_status_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "device-status", 35);
} if (rf & 2) {
    const result_r29 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("mo", result_r29);
} }
function SearchInputComponent_ng_template_2_c8y_li_6_i_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 36);
} if (rf & 2) {
    const result_r29 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r31 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("c8yIcon", ctx_r31.getIcon(result_r29));
} }
function SearchInputComponent_ng_template_2_c8y_li_6_Template(rf, ctx) { if (rf & 1) {
    const _r35 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 32);
    ɵngcc0.ɵɵlistener("click", function SearchInputComponent_ng_template_2_c8y_li_6_Template_c8y_li_click_0_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r35); const result_r29 = restoredCtx.$implicit; const ctx_r34 = ɵngcc0.ɵɵnextContext(2); return ctx_r34.open($event, result_r29, result_r29.name); });
    ɵngcc0.ɵɵelementStart(1, "c8y-li-icon");
    ɵngcc0.ɵɵtemplate(2, SearchInputComponent_ng_template_2_c8y_li_6_device_status_2_Template, 1, 1, "device-status", 33);
    ɵngcc0.ɵɵtemplate(3, SearchInputComponent_ng_template_2_c8y_li_6_i_3_Template, 1, 1, "i", 34);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const result_r29 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !result_r29.c8y_IsDeviceGroup);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", result_r29.c8y_IsDeviceGroup);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", result_r29.name || "--", " ");
} }
function SearchInputComponent_ng_template_2_c8y_li_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 30);
    ɵngcc0.ɵɵelementStart(1, "div", 31);
    ɵngcc0.ɵɵelementStart(2, "span", 27);
    ɵngcc0.ɵɵtext(3, "Recently registered devices");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("selectable", false);
} }
function SearchInputComponent_ng_template_2_c8y_li_9_device_status_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "device-status", 35);
} if (rf & 2) {
    const result_r36 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("mo", result_r36);
} }
function SearchInputComponent_ng_template_2_c8y_li_9_i_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 36);
} if (rf & 2) {
    const result_r36 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r38 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("c8yIcon", ctx_r38.getIcon(result_r36));
} }
function SearchInputComponent_ng_template_2_c8y_li_9_Template(rf, ctx) { if (rf & 1) {
    const _r42 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 32);
    ɵngcc0.ɵɵlistener("click", function SearchInputComponent_ng_template_2_c8y_li_9_Template_c8y_li_click_0_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r42); const result_r36 = restoredCtx.$implicit; const ctx_r41 = ɵngcc0.ɵɵnextContext(2); return ctx_r41.open($event, result_r36, result_r36.name); });
    ɵngcc0.ɵɵelementStart(1, "c8y-li-icon");
    ɵngcc0.ɵɵtemplate(2, SearchInputComponent_ng_template_2_c8y_li_9_device_status_2_Template, 1, 1, "device-status", 33);
    ɵngcc0.ɵɵtemplate(3, SearchInputComponent_ng_template_2_c8y_li_9_i_3_Template, 1, 1, "i", 34);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const result_r36 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !result_r36.c8y_IsDeviceGroup);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", result_r36.c8y_IsDeviceGroup);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", result_r36.name || "--", " ");
} }
function SearchInputComponent_ng_template_2_c8y_li_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 30);
    ɵngcc0.ɵɵelementStart(1, "div", 31);
    ɵngcc0.ɵɵelementStart(2, "span", 27);
    ɵngcc0.ɵɵtext(3, "Search results");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("selectable", false);
} }
function SearchInputComponent_ng_template_2_c8y_li_11_device_status_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "device-status", 35);
} if (rf & 2) {
    const result_r43 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("mo", result_r43);
} }
function SearchInputComponent_ng_template_2_c8y_li_11_i_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 36);
} if (rf & 2) {
    const result_r43 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r45 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("c8yIcon", ctx_r45.getIcon(result_r43));
} }
function SearchInputComponent_ng_template_2_c8y_li_11_Template(rf, ctx) { if (rf & 1) {
    const _r49 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 37);
    ɵngcc0.ɵɵlistener("click", function SearchInputComponent_ng_template_2_c8y_li_11_Template_c8y_li_click_0_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r49); const result_r43 = restoredCtx.$implicit; const ctx_r48 = ɵngcc0.ɵɵnextContext(2); return ctx_r48.open($event, result_r43, result_r43.name); });
    ɵngcc0.ɵɵelementStart(1, "c8y-li-icon");
    ɵngcc0.ɵɵtemplate(2, SearchInputComponent_ng_template_2_c8y_li_11_device_status_2_Template, 1, 1, "device-status", 33);
    ɵngcc0.ɵɵtemplate(3, SearchInputComponent_ng_template_2_c8y_li_11_i_3_Template, 1, 1, "i", 34);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const result_r43 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("title", result_r43.name);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !result_r43.c8y_IsDeviceGroup);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", result_r43.c8y_IsDeviceGroup);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", result_r43.name || "--", " ");
} }
function SearchInputComponent_ng_template_2_ng_template_12_c8y_li_0_small_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 27);
    ɵngcc0.ɵɵtext(1, " Try to filter or open the asset grid to show all devices and groups. ");
    ɵngcc0.ɵɵelementEnd();
} }
function SearchInputComponent_ng_template_2_ng_template_12_c8y_li_0_small_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 27);
    ɵngcc0.ɵɵtext(1, " Try to rephrase your search word. ");
    ɵngcc0.ɵɵelementEnd();
} }
function SearchInputComponent_ng_template_2_ng_template_12_c8y_li_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 39);
    ɵngcc0.ɵɵelement(1, "c8y-li-icon", 40);
    ɵngcc0.ɵɵelementStart(2, "p");
    ɵngcc0.ɵɵelementStart(3, "strong", 27);
    ɵngcc0.ɵɵtext(4, "No match found.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(5, SearchInputComponent_ng_template_2_ng_template_12_c8y_li_0_small_5_Template, 2, 0, "small", 41);
    ɵngcc0.ɵɵtemplate(6, SearchInputComponent_ng_template_2_ng_template_12_c8y_li_0_small_6_Template, 2, 0, "small", 41);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r50 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("selectable", false);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("icon", "search");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r50.mode === "search");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r50.mode === "select");
} }
function SearchInputComponent_ng_template_2_ng_template_12_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, SearchInputComponent_ng_template_2_ng_template_12_c8y_li_0_Template, 7, 4, "c8y-li", 38);
} if (rf & 2) {
    const ctx_r19 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r19.noMatch);
} }
function SearchInputComponent_ng_template_2_c8y_li_14_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 42);
    ɵngcc0.ɵɵelementStart(1, "div", 43);
    ɵngcc0.ɵɵelement(2, "div", 44);
    ɵngcc0.ɵɵelement(3, "div", 45);
    ɵngcc0.ɵɵelement(4, "div", 46);
    ɵngcc0.ɵɵelement(5, "div", 47);
    ɵngcc0.ɵɵelement(6, "div", 48);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function SearchInputComponent_ng_template_2_ng_template_15_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 49);
    ɵngcc0.ɵɵelementStart(1, "div", 43);
    ɵngcc0.ɵɵelement(2, "div", 44);
    ɵngcc0.ɵɵelement(3, "div", 45);
    ɵngcc0.ɵɵelement(4, "div", 46);
    ɵngcc0.ɵɵelement(5, "div", 47);
    ɵngcc0.ɵɵelement(6, "div", 48);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function SearchInputComponent_ng_template_2_c8y_li_17_Template(rf, ctx) { if (rf & 1) {
    const _r54 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 50);
    ɵngcc0.ɵɵelementStart(1, "div", 25);
    ɵngcc0.ɵɵelement(2, "i", 51);
    ɵngcc0.ɵɵelementStart(3, "p", 52);
    ɵngcc0.ɵɵtext(4, "Need more filter possibilities?");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "button", 53);
    ɵngcc0.ɵɵlistener("mousedown", function SearchInputComponent_ng_template_2_c8y_li_17_Template_button_mousedown_5_listener() { ɵngcc0.ɵɵrestoreView(_r54); const ctx_r53 = ɵngcc0.ɵɵnextContext(2); return ctx_r53.onOpenAssetTable(); });
    ɵngcc0.ɵɵtext(6, " Go to the asset data table ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("selectable", false);
} }
const _c1 = function (a0) { return { "c8y-search-form": a0 }; };
const _c2 = function () { return []; };
const _c3 = function (a0) { return { data: a0 }; };
function SearchInputComponent_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    const _r56 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "form", 11, 12);
    ɵngcc0.ɵɵelementStart(2, "c8y-typeahead", 13);
    ɵngcc0.ɵɵlistener("ngModelChange", function SearchInputComponent_ng_template_2_Template_c8y_typeahead_ngModelChange_2_listener($event) { ɵngcc0.ɵɵrestoreView(_r56); const ctx_r55 = ɵngcc0.ɵɵnextContext(); return ctx_r55.selected = $event; })("keydown", function SearchInputComponent_ng_template_2_Template_c8y_typeahead_keydown_2_listener($event) { ɵngcc0.ɵɵrestoreView(_r56); const ctx_r57 = ɵngcc0.ɵɵnextContext(); return ctx_r57.keyDown($event); })("onIconClick", function SearchInputComponent_ng_template_2_Template_c8y_typeahead_onIconClick_2_listener($event) { ɵngcc0.ɵɵrestoreView(_r56); const ctx_r58 = ɵngcc0.ɵɵnextContext(); return ctx_r58.onReset($event); });
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵtemplate(4, SearchInputComponent_ng_template_2_c8y_li_4_Template, 2, 1, "c8y-li", 14);
    ɵngcc0.ɵɵtemplate(5, SearchInputComponent_ng_template_2_c8y_li_5_Template, 4, 1, "c8y-li", 15);
    ɵngcc0.ɵɵtemplate(6, SearchInputComponent_ng_template_2_c8y_li_6_Template, 5, 3, "c8y-li", 16);
    ɵngcc0.ɵɵtemplate(7, SearchInputComponent_ng_template_2_c8y_li_7_Template, 4, 1, "c8y-li", 15);
    ɵngcc0.ɵɵpipe(8, "async");
    ɵngcc0.ɵɵtemplate(9, SearchInputComponent_ng_template_2_c8y_li_9_Template, 5, 3, "c8y-li", 17);
    ɵngcc0.ɵɵtemplate(10, SearchInputComponent_ng_template_2_c8y_li_10_Template, 4, 1, "c8y-li", 15);
    ɵngcc0.ɵɵtemplate(11, SearchInputComponent_ng_template_2_c8y_li_11_Template, 5, 4, "c8y-li", 18);
    ɵngcc0.ɵɵtemplate(12, SearchInputComponent_ng_template_2_ng_template_12_Template, 1, 1, "ng-template", null, 19, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵtemplate(14, SearchInputComponent_ng_template_2_c8y_li_14_Template, 7, 0, "c8y-li", 20);
    ɵngcc0.ɵɵtemplate(15, SearchInputComponent_ng_template_2_ng_template_15_Template, 7, 0, "ng-template", null, 21, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵtemplate(17, SearchInputComponent_ng_template_2_c8y_li_17_Template, 7, 1, "c8y-li", 22);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _r18 = ɵngcc0.ɵɵreference(13);
    const _r21 = ɵngcc0.ɵɵreference(16);
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    let tmp_9_0;
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(26, _c1, ctx_r3.mode === "search"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(3, 22, "Search for groups or assets\u2026"));
    ɵngcc0.ɵɵproperty("ngModel", ctx_r3.selected)("icon", ctx_r3.term ? "times" : "search")("allowFreeEntries", false)("container", ctx_r3.container);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.term.length !== 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.term.length === 0 && ctx_r3.recentSearchResults.length > 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r3.term.length === 0 ? ctx_r3.recentSearchResults : ɵngcc0.ɵɵpureFunction0(28, _c2));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.term.length === 0 && ((tmp_9_0 = ɵngcc0.ɵɵpipeBind1(8, 24, ctx_r3.recentlyRegisteredResults$)) == null ? null : tmp_9_0.data == null ? null : tmp_9_0.data.length) > 0);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yForOf", ctx_r3.term.length === 0 ? ctx_r3.recentlyRegisteredResults$ : ɵngcc0.ɵɵpureFunction1(30, _c3, ɵngcc0.ɵɵpureFunction0(29, _c2)))("c8yForLoadMore", "none")("c8yForPipe", ctx_r3.filterPipe);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.term.length !== 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yForOf", ctx_r3.results$)("c8yForLoadMore", "auto")("c8yForPipe", ctx_r3.filterPipe)("c8yForNotFound", _r18)("c8yForLoadingTemplate", _r21)("c8yForLoadNextLabel", "Find more\u2026");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.isLoading);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.mode === "search");
} }
export class SearchInputComponent {
    constructor(router, inventory, searchService, assetTypesService) {
        this.router = router;
        this.inventory = inventory;
        this.searchService = searchService;
        this.assetTypesService = assetTypesService;
        this.mode = 'search';
        this.container = '';
        this.filter = new EventEmitter();
        this.search = new EventEmitter();
        this.reset = new EventEmitter();
        this.onClick = new EventEmitter();
        this.term = '';
        this.filterPipe = pipe(map((data) => {
            return this.searchService.filterOnlyAssets(data);
        }));
        this.recentSearchResults = [];
        this.isLoading = false;
        this.noMatch = false;
        this.RECENT_SEARCH_STORAGE_KEY = 'recent_search_view';
        this.MAX_RECENT_SEARCH_RESULTS = 5;
        this.DEFAULT_FILTER = {
            withTotalPages: true,
            pageSize: 5,
            withChildren: false
        };
        this.KEYCODE_ENTER = 'Enter';
        this.KEYCODE_ESC = 'Escape';
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            const recentSearchIds = JSON.parse(localStorage.getItem(this.RECENT_SEARCH_STORAGE_KEY));
            if (recentSearchIds && recentSearchIds.length > 0) {
                const { data } = yield this.inventory.list({ ids: recentSearchIds.join(',') });
                this.recentSearchResults = data;
                this.recentlyRegisteredResults$ = defer(() => this.inventory.list(Object.assign({ q: '$orderby=creationTime desc' }, this.DEFAULT_FILTER)));
            }
            if (this.mode === 'select') {
                requestAnimationFrame(() => {
                    this.subscribeOnSearch();
                });
            }
        });
    }
    onKeydownHandler(event) {
        if (event.key === this.KEYCODE_ESC) {
            this.hideDropdown();
        }
    }
    getIcon(mo) {
        return this.assetTypesService.getIcon(mo.type) || 'c8y-group-open';
    }
    onOpenChange(isOpen) {
        if (isOpen) {
            // needs to request an animation frame as
            // otherwise the typeahead is undefined
            requestAnimationFrame(() => {
                this.subscribeOnSearch();
                this.typeahead.dropdown.show();
                this.typeahead.searchControl.nativeElement.focus();
            });
        }
    }
    open(event, mo, term) {
        event.stopPropagation();
        const isAlreadyRecent = this.recentSearchResults.find(({ id }) => id === mo.id);
        if (!isAlreadyRecent) {
            this.recentSearchResults.unshift(mo);
            this.recentSearchResults = this.recentSearchResults.slice(0, this.MAX_RECENT_SEARCH_RESULTS);
        }
        const recentSearchIds = this.recentSearchResults.map(({ id }) => id);
        localStorage.setItem(this.RECENT_SEARCH_STORAGE_KEY, JSON.stringify(recentSearchIds));
        if (term) {
            this.selected = mo;
            this.term = term;
        }
        this.onClick.emit(mo);
        this.hideDropdown();
    }
    onReset(status) {
        status.$event.stopPropagation();
        this.reset.emit(this.term);
        this.typeahead.onSearch.emit('');
        this.selected = undefined;
        this.typeahead.searchControl.nativeElement.focus();
    }
    keyDown(event) {
        if (event.key === this.KEYCODE_ENTER) {
            // enter hit can be faster then typeahead debounce,
            // therefore we take the term from the DOM element
            // itself:
            const searchTerm = event.target.value;
            this.onSearch(searchTerm);
        }
    }
    onSearch(search) {
        this.search.emit(search);
        this.hideDropdown();
    }
    onFilter(search) {
        this.filter.emit(search);
        this.hideDropdown();
    }
    onOpenAssetTable() {
        this.router.navigateByUrl('/assetsearch');
        this.hideDropdown();
    }
    hideDropdown() {
        if (this.dropdown) {
            this.dropdown.hide();
            return;
        }
        if (this.typeahead && this.typeahead.dropdown) {
            this.typeahead.dropdown.hide();
            return;
        }
    }
    subscribeOnSearch() {
        if (!this.results$) {
            this.results$ = this.typeahead.onSearch.pipe(tap(term => this.onTypingStarted(term)), switchMap(term => this.mergeRequest(term)));
        }
    }
    navigate(commands, extras) {
        this.router
            .navigateByUrl('/', { skipLocationChange: true })
            .then(() => this.router.navigate(commands, extras));
    }
    mergeRequest(term) {
        return merge(of({ data: [] }), this.queryInventoryService(term).pipe(tap(({ data, paging }) => this.onLoadingDone(data, paging))));
    }
    queryInventoryService(term) {
        if (term) {
            return defer(() => this.searchService.search(term));
        }
        return empty();
    }
    onLoadingDone(data, paging) {
        this.isLoading = false;
        this.noMatch =
            paging && paging.nextPage === null && this.searchService.filterOnlyAssets(data).length === 0;
    }
    onTypingStarted(term) {
        this.noMatch = false;
        this.term = term;
        this.isLoading = term.length > 0;
    }
}
SearchInputComponent.ɵfac = function SearchInputComponent_Factory(t) { return new (t || SearchInputComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Router), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.InventoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.InventorySearchService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.AssetTypesService)); };
SearchInputComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SearchInputComponent, selectors: [["c8y-search-input"]], viewQuery: function SearchInputComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(TypeaheadComponent, 5);
        ɵngcc0.ɵɵviewQuery(_c0, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.typeahead = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.dropdown = _t.first);
    } }, hostBindings: function SearchInputComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("keydown", function SearchInputComponent_keydown_HostBindingHandler($event) { return ctx.onKeydownHandler($event); }, false, ɵngcc0.ɵɵresolveDocument);
    } }, inputs: { mode: "mode", container: "container" }, outputs: { filter: "filter", search: "search", reset: "reset", onClick: "onClick" }, decls: 4, vars: 2, consts: [["class", "dropdown", "dropdown", "", 3, "insideClick", "isOpenChange", 4, "ngIf"], ["class", "search-header-inline", 4, "ngIf"], ["form", ""], ["dropdown", "", 1, "dropdown", 3, "insideClick", "isOpenChange"], ["dropdown", "bs-dropdown"], ["dropdownToggle", "", "type", "button", "aria-controls", "searchDropdown", 1, "main-header-button", "dropdown-toggle", "c8y-dropdown", 3, "title"], ["c8yIcon", "search", 1, "icon-2x"], ["id", "searchDropdown", "class", "search-header-menu dropdown-menu dropdown-menu-center", 4, "dropdownMenu"], ["id", "searchDropdown", 1, "search-header-menu", "dropdown-menu", "dropdown-menu-center"], [4, "ngTemplateOutlet"], [1, "search-header-inline"], ["novalidate", "", 3, "ngClass"], ["searchForm", "ngForm"], ["name", "selected", 3, "ngModel", "placeholder", "icon", "allowFreeEntries", "container", "ngModelChange", "keydown", "onIconClick"], ["class", "p-l-16 p-r-16", 4, "ngIf"], ["class", "p-l-24 p-r-24", 3, "selectable", 4, "ngIf"], ["class", "c8y-list__item--link m-l-16 m-r-16", 3, "click", 4, "ngFor", "ngForOf"], ["class", "c8y-list__item--link m-l-16 m-r-16", 3, "click", 4, "c8yFor", "c8yForOf", "c8yForLoadMore", "c8yForPipe"], ["class", "c8y-list__item--link  m-l-16 m-r-16", 3, "title", "click", 4, "c8yFor", "c8yForOf", "c8yForLoadMore", "c8yForPipe", "c8yForNotFound", "c8yForLoadingTemplate", "c8yForLoadNextLabel"], ["notFoundTemplate", ""], ["class", "p-t-32 p-b-0", "style", "position: relative;", 4, "ngIf"], ["loading", ""], ["class", "m-t-24 bg-gray-lighter p-t-16 p-b-16 p-l-24 p-r-24", 3, "selectable", 4, "ngIf"], [1, "p-l-16", "p-r-16"], ["class", "flex-row", 4, "ngIf"], [1, "flex-row"], [1, "m-r-4", "text-muted"], ["translate", ""], [1, "btn-group", "btn-group-sm"], [1, "btn", "btn-default", 3, "title", "click"], [1, "p-l-24", "p-r-24", 3, "selectable"], [1, "legend", "form-block"], [1, "c8y-list__item--link", "m-l-16", "m-r-16", 3, "click"], [3, "mo", 4, "ngIf"], ["class", "c8y-icon-duocolor", 3, "c8yIcon", 4, "ngIf"], [3, "mo"], [1, "c8y-icon-duocolor", 3, "c8yIcon"], [1, "c8y-list__item--link", "m-l-16", "m-r-16", 3, "title", "click"], ["class", "p-16 c8y-empty-state", 3, "selectable", 4, "ngIf"], [1, "p-16", "c8y-empty-state", 3, "selectable"], [3, "icon"], ["translate", "", 4, "ngIf"], [1, "p-t-32", "p-b-0", 2, "position", "relative"], [1, "spinner", 2, "right", "0"], [1, "rect1"], [1, "rect2"], [1, "rect3"], [1, "rect4"], [1, "rect5"], [1, "text-center", "p-t-32", "p-b-0", 2, "position", "relative"], [1, "m-t-24", "bg-gray-lighter", "p-t-16", "p-b-16", "p-l-24", "p-r-24", 3, "selectable"], ["c8yIcon", "info-circle", 1, "text-info", "m-r-4"], ["translate", "", 1, "m-r-8"], ["type", "button", "translate", "", 1, "m-l-auto", "btn", "btn-default", "btn-sm", 3, "mousedown"]], template: function SearchInputComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, SearchInputComponent_div_0_Template, 6, 4, "div", 0);
        ɵngcc0.ɵɵtemplate(1, SearchInputComponent_div_1_Template, 2, 1, "div", 1);
        ɵngcc0.ɵɵtemplate(2, SearchInputComponent_ng_template_2_Template, 18, 32, "ng-template", null, 2, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.mode === "search");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.mode === "select");
    } }, directives: [ɵngcc5.NgIf, ɵngcc6.BsDropdownDirective, ɵngcc6.BsDropdownToggleDirective, ɵngcc7.IconDirective, ɵngcc6.BsDropdownMenuDirective, ɵngcc5.NgTemplateOutlet, ɵngcc8.ɵNgNoValidate, ɵngcc8.NgControlStatusGroup, ɵngcc8.NgForm, ɵngcc5.NgClass, ɵngcc9.TypeaheadComponent, ɵngcc8.NgControlStatus, ɵngcc8.NgModel, ɵngcc5.NgForOf, ɵngcc10.ForOfDirective, ɵngcc11.ListItemComponent, ɵngcc12.C8yTranslateDirective, ɵngcc13.ListItemIconComponent, ɵngcc14.DeviceStatusComponent], pipes: [ɵngcc15.C8yTranslatePipe, ɵngcc5.AsyncPipe], encapsulation: 2 });
SearchInputComponent.ctorParameters = () => [
    { type: Router },
    { type: InventoryService },
    { type: InventorySearchService },
    { type: AssetTypesService }
];
SearchInputComponent.propDecorators = {
    mode: [{ type: Input }],
    container: [{ type: Input }],
    filter: [{ type: Output }],
    search: [{ type: Output }],
    reset: [{ type: Output }],
    onClick: [{ type: Output }],
    typeahead: [{ type: ViewChild, args: [TypeaheadComponent, { static: false },] }],
    dropdown: [{ type: ViewChild, args: ['dropdown', { static: false },] }],
    onKeydownHandler: [{ type: HostListener, args: ['document:keydown', ['$event'],] }]
};
__decorate([
    memoize(property('id'))
], SearchInputComponent.prototype, "getIcon", null);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SearchInputComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-search-input',
                template: "<div\n  class=\"dropdown\"\n  dropdown\n  #dropdown=\"bs-dropdown\"\n  [insideClick]=\"true\"\n  (isOpenChange)=\"onOpenChange($event)\"\n  *ngIf=\"mode === 'search'\"\n>\n  <button\n    class=\"main-header-button dropdown-toggle c8y-dropdown\"\n    dropdownToggle\n    type=\"button\"\n    [title]=\"'Search' | translate\"\n    aria-controls=\"searchDropdown\"\n  >\n    <i c8yIcon=\"search\" class=\"icon-2x\"></i>\n  </button>\n\n  <div\n    id=\"searchDropdown\"\n    *dropdownMenu\n    class=\"search-header-menu dropdown-menu dropdown-menu-center\"\n  >\n    <ng-container *ngTemplateOutlet=\"form\"></ng-container>\n  </div>\n</div>\n\n<div class=\"search-header-inline\" *ngIf=\"mode === 'select'\">\n  <ng-container *ngTemplateOutlet=\"form\"></ng-container>\n</div>\n\n<ng-template #form>\n  <form [ngClass]=\"{ 'c8y-search-form': mode === 'search' }\" novalidate #searchForm=\"ngForm\">\n    <c8y-typeahead\n      [(ngModel)]=\"selected\"\n      placeholder=\"{{ 'Search for groups or assets\u2026' | translate }}\"\n      (keydown)=\"keyDown($event)\"\n      (onIconClick)=\"onReset($event)\"\n      [icon]=\"term ? 'times' : 'search'\"\n      [allowFreeEntries]=\"false\"\n      [container]=\"container\"\n      name=\"selected\"\n    >\n      <!-- filter buttons -->\n      <c8y-li *ngIf=\"term.length !== 0\" class=\"p-l-16 p-r-16\">\n        <div class=\"flex-row\" *ngIf=\"mode === 'search'\">\n          <p class=\"m-r-4 text-muted\">\n            <em translate>Searching by exact match. Click for other search options:</em>\n          </p>\n          <div class=\"btn-group btn-group-sm\">\n            <button\n              class=\"btn btn-default\"\n              title=\"{{ 'Starts with' | translate }}\"\n              (click)=\"onFilter(term + '*')\"\n            >\n              {{ 'Starts with' | translate }}\n            </button>\n            <button\n              class=\"btn btn-default\"\n              title=\"{{ 'Contains' | translate }}\"\n              (click)=\"onFilter('*' + term + '*')\"\n            >\n              {{ 'Contains' | translate }}\n            </button>\n            <button\n              class=\"btn btn-default\"\n              title=\"{{ 'Ends with' | translate }}\"\n              (click)=\"onFilter('*' + term)\"\n            >\n              {{ 'Ends with' | translate }}\n            </button>\n          </div>\n        </div>\n      </c8y-li>\n\n      <!-- Recent search -->\n      <c8y-li\n        *ngIf=\"term.length === 0 && recentSearchResults.length > 0\"\n        [selectable]=\"false\"\n        class=\"p-l-24 p-r-24\"\n      >\n        <div class=\"legend form-block\">\n          <span translate>Recent search views</span>\n        </div>\n      </c8y-li>\n      <c8y-li\n        *ngFor=\"let result of term.length === 0 ? recentSearchResults : []\"\n        class=\"c8y-list__item--link m-l-16 m-r-16\"\n        (click)=\"open($event, result, result.name)\"\n      >\n        <c8y-li-icon>\n          <device-status [mo]=\"result\" *ngIf=\"!result.c8y_IsDeviceGroup\"></device-status>\n          <i\n            [c8yIcon]=\"getIcon(result)\"\n            class=\"c8y-icon-duocolor\"\n            *ngIf=\"result.c8y_IsDeviceGroup\"\n          ></i>\n        </c8y-li-icon>\n        {{ result.name || '--' }}\n      </c8y-li>\n\n      <!-- Recently registered devices -->\n      <c8y-li\n        *ngIf=\"term.length === 0 && (recentlyRegisteredResults$ | async)?.data?.length > 0\"\n        class=\"p-l-24 p-r-24\"\n        [selectable]=\"false\"\n      >\n        <div class=\"legend form-block\">\n          <span translate>Recently registered devices</span>\n        </div>\n      </c8y-li>\n      <c8y-li\n        *c8yFor=\"\n          let result of term.length === 0 ? recentlyRegisteredResults$ : { data: [] };\n          loadMore: 'none';\n          pipe: filterPipe\n        \"\n        class=\"c8y-list__item--link m-l-16 m-r-16\"\n        (click)=\"open($event, result, result.name)\"\n      >\n        <c8y-li-icon>\n          <device-status [mo]=\"result\" *ngIf=\"!result.c8y_IsDeviceGroup\"></device-status>\n          <i\n          [c8yIcon]=\"getIcon(result)\"\n            class=\"c8y-icon-duocolor\"\n            *ngIf=\"result.c8y_IsDeviceGroup\"\n          ></i>\n        </c8y-li-icon>\n        {{ result.name || '--' }}\n      </c8y-li>\n\n      <!-- Search results -->\n      <c8y-li *ngIf=\"term.length !== 0\" class=\"p-l-24 p-r-24\" [selectable]=\"false\">\n        <div class=\"legend form-block\">\n          <span translate>Search results</span>\n        </div>\n      </c8y-li>\n      <c8y-li\n        *c8yFor=\"\n          let result of results$;\n          loadMore: 'auto';\n          pipe: filterPipe;\n          notFound: notFoundTemplate;\n          loadingTemplate: loading;\n          loadNextLabel: 'Find more\u2026'\n        \"\n        class=\"c8y-list__item--link  m-l-16 m-r-16\"\n        (click)=\"open($event, result, result.name)\"\n        [title]=\"result.name\"\n      >\n        <c8y-li-icon>\n          <device-status [mo]=\"result\" *ngIf=\"!result.c8y_IsDeviceGroup\"></device-status>\n          <i\n            [c8yIcon]=\"getIcon(result)\"\n            class=\"c8y-icon-duocolor\"\n            *ngIf=\"result.c8y_IsDeviceGroup\"\n          ></i>\n        </c8y-li-icon>\n        {{ result.name || '--' }}\n      </c8y-li>\n\n      <!-- No search results found entry -->\n      <ng-template #notFoundTemplate>\n        <c8y-li *ngIf=\"noMatch\" class=\"p-16 c8y-empty-state\" [selectable]=\"false\">\n          <c8y-li-icon [icon]=\"'search'\"></c8y-li-icon>\n          <p><strong translate>No match found.</strong></p>\n          <small translate *ngIf=\"mode === 'search'\">\n            Try to filter or open the asset grid to show all devices and groups.\n          </small>\n          <small translate *ngIf=\"mode === 'select'\">\n            Try to rephrase your search word.\n          </small>\n        </c8y-li>\n      </ng-template>\n\n      <!-- loading bar first entries -->\n      <c8y-li *ngIf=\"isLoading\" class=\"p-t-32 p-b-0\" style=\"position: relative;\">\n        <div class=\"spinner\" style=\"right:0;\">\n          <div class=\"rect1\"></div>\n          <div class=\"rect2\"></div>\n          <div class=\"rect3\"></div>\n          <div class=\"rect4\"></div>\n          <div class=\"rect5\"></div>\n        </div>\n      </c8y-li>\n\n      <!-- loading bar for loading more entries (inventory roles) -->\n      <ng-template #loading>\n        <c8y-li class=\"text-center p-t-32 p-b-0\" style=\"position:relative;\">\n          <div class=\"spinner\" style=\"right:0;\">\n            <div class=\"rect1\"></div>\n            <div class=\"rect2\"></div>\n            <div class=\"rect3\"></div>\n            <div class=\"rect4\"></div>\n            <div class=\"rect5\"></div>\n          </div>\n        </c8y-li>\n      </ng-template>\n\n      <!-- more filter possibilities -->\n      <c8y-li\n        class=\"m-t-24 bg-gray-lighter p-t-16 p-b-16 p-l-24 p-r-24\"\n        [selectable]=\"false\"\n        *ngIf=\"mode === 'search'\"\n      >\n        <div class=\"flex-row\">\n          <i c8yIcon=\"info-circle\" class=\"text-info m-r-4\"></i>\n          <p translate class=\"m-r-8\">Need more filter possibilities?</p>\n          <button\n            type=\"button\"\n            class=\"m-l-auto btn btn-default btn-sm\"\n            translate\n            (mousedown)=\"onOpenAssetTable()\"\n          >\n            Go to the asset data table\n          </button>\n        </div>\n      </c8y-li>\n    </c8y-typeahead>\n  </form>\n</ng-template>\n"
            }]
    }], function () { return [{ type: ɵngcc1.Router }, { type: ɵngcc2.InventoryService }, { type: ɵngcc3.InventorySearchService }, { type: ɵngcc4.AssetTypesService }]; }, { mode: [{
            type: Input
        }], container: [{
            type: Input
        }], filter: [{
            type: Output
        }], search: [{
            type: Output
        }], reset: [{
            type: Output
        }], onClick: [{
            type: Output
        }], onKeydownHandler: [{
            type: HostListener,
            args: ['document:keydown', ['$event']]
        }], typeahead: [{
            type: ViewChild,
            args: [TypeaheadComponent, { static: false }]
        }], dropdown: [{
            type: ViewChild,
            args: ['dropdown', { static: false }]
        }] }); })();
