// @c8y/ngx-components/esm2015/repository/shared/shared-repository.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CoreModule, FormsModule, OperationRealtimeService } from '@c8y/ngx-components';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { RepositoryService } from './repository.service';
import { RepositorySelectModalComponent } from './select-modal/repository-select-modal.component';
import { FileDownloadComponent } from './file-download/file-download.component';
import { AdvancedSoftwareService } from './advanced-software.service';
import { SoftwareTypeComponent } from './software-type/software-type.component';
import * as ɵngcc0 from '@angular/core';
export class SharedRepositoryModule {
}
SharedRepositoryModule.ɵfac = function SharedRepositoryModule_Factory(t) { return new (t || SharedRepositoryModule)(); };
SharedRepositoryModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: SharedRepositoryModule });
SharedRepositoryModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [AdvancedSoftwareService, RepositoryService, OperationRealtimeService], imports: [[CommonModule, CoreModule, FormsModule, PopoverModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SharedRepositoryModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, CoreModule, FormsModule, PopoverModule],
                declarations: [RepositorySelectModalComponent, FileDownloadComponent, SoftwareTypeComponent],
                entryComponents: [RepositorySelectModalComponent],
                providers: [AdvancedSoftwareService, RepositoryService, OperationRealtimeService],
                exports: [RepositorySelectModalComponent, FileDownloadComponent, SoftwareTypeComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(SharedRepositoryModule, { declarations: function () { return [RepositorySelectModalComponent, FileDownloadComponent, SoftwareTypeComponent]; }, imports: function () { return [CommonModule, CoreModule, FormsModule, PopoverModule]; }, exports: function () { return [RepositorySelectModalComponent, FileDownloadComponent, SoftwareTypeComponent]; } }); })();
