// @c8y/ngx-components/esm2015/core/list-group/list-item-timeline.component.js
import { Component, ContentChild, HostBinding } from '@angular/core';
import { ListItemComponent } from './list-item.component';
/**
 * Allows to display a timeline. Should be used with
 * content projection.
 *
 * @example
 * ```html
 * <c8y-list-group>
 *   <c8y-li-timeline>
 *     24.April 2020
 *     <c8y-li>
 *       Hello
 *     </c8y-li>
 *   </c8y-li-timeline>
 * </c8y-list-group>
 * ```
 */
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*", [["c8y-li"], ["c8y-list-item"]]];
const _c1 = ["*", "c8y-li, c8y-list-item"];
export class ListItemTimelineComponent {
    get expanded() {
        var _a;
        return !((_a = this.listItemComponent) === null || _a === void 0 ? void 0 : _a.collapsed);
    }
}
ListItemTimelineComponent.ɵfac = function ListItemTimelineComponent_Factory(t) { return new (t || ListItemTimelineComponent)(); };
ListItemTimelineComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ListItemTimelineComponent, selectors: [["c8y-list-item-timeline"], ["c8y-li-timeline"]], contentQueries: function ListItemTimelineComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, ListItemComponent, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.listItemComponent = _t.first);
    } }, hostAttrs: [1, "c8y-list--timeline__item"], hostVars: 2, hostBindings: function ListItemTimelineComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("expanded", ctx.expanded);
    } }, ngContentSelectors: _c1, decls: 3, vars: 0, consts: [[1, "c8y-list--timeline__item__date"]], template: function ListItemTimelineComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c0);
        ɵngcc0.ɵɵelementStart(0, "small", 0);
        ɵngcc0.ɵɵprojection(1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵprojection(2, 1);
    } }, encapsulation: 2 });
ListItemTimelineComponent.propDecorators = {
    listItemComponent: [{ type: ContentChild, args: [ListItemComponent, { static: false },] }],
    expanded: [{ type: HostBinding, args: ['class.expanded',] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ListItemTimelineComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-list-item-timeline, c8y-li-timeline',
                template: "<small class=\"c8y-list--timeline__item__date\">\n  <ng-content></ng-content>\n</small>\n<ng-content select=\"c8y-li, c8y-list-item\"></ng-content>\n",
                host: {
                    class: 'c8y-list--timeline__item'
                }
            }]
    }], null, { expanded: [{
            type: HostBinding,
            args: ['class.expanded']
        }], listItemComponent: [{
            type: ContentChild,
            args: [ListItemComponent, { static: false }]
        }] }); })();
