// @c8y/ngx-components/esm2015/core/list-group/list-item-drag-handle.component.js
import { Component } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class ListItemDragHandleComponent {
}
ListItemDragHandleComponent.ɵfac = function ListItemDragHandleComponent_Factory(t) { return new (t || ListItemDragHandleComponent)(); };
ListItemDragHandleComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ListItemDragHandleComponent, selectors: [["c8y-list-item-drag-handle"], ["c8y-li-drag-handle"]], ngContentSelectors: _c0, decls: 1, vars: 0, template: function ListItemDragHandleComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ListItemDragHandleComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-list-item-drag-handle, c8y-li-drag-handle',
                template: `<ng-content></ng-content>`
            }]
    }], null, null); })();
