// @c8y/ngx-components/esm2015/core/list-group/list-item.component.js
import { Component, ContentChild, ContentChildren, ElementRef, EventEmitter, HostBinding, Input, Output } from '@angular/core';
import { Subject } from 'rxjs';
import { delay, filter, takeUntil, tap } from 'rxjs/operators';
import { ListItemActionComponent } from './list-item-action.component';
import { ListItemCollapseComponent } from './list-item-collapse.component';
import { ListItemFooterComponent } from './list-item-footer.component';
/**
 * A list item is a representation of an item inside a list and
 * can be used to compose different styles in a list (mainly in a
 * <c8y-group-list>-component but also in others like the
 * <c8y-typeahead>-component):
 *  - a list with a footer
 *  - a list with one or multiple actions
 *  - an icon
 *  - a checkbox or radio input
 *
 * The item can be composed via content projection. The following
 * example shows a radio list with an icon, two actions
 * and a footer:
 * ```html
 * <c8y-list-group>
 *   <c8y-li
 *     *c8yFor="
 *       let device of devices;
 *       let i = index;
 *     "
 *   >
 *     <c8y-li-radio (onSelect)="updateSelected($event, device)"></c8y-li-radio>
 *     <c8y-li-icon [icon]="'rocket'"></c8y-li-icon>
 *     {{ i + 1 }}. {{ device.name || '-' }}
 *     <c8y-li-footer>
 *       Device id: <a [routerLink]="['/device', device.id]">{{ device.id }}</a>
 *     </c8y-li-footer>
 *     <c8y-li-action (click)="deleteDevice(device.id)" icon="times">
 *       Delete
 *     </c8y-li-action>
 *     <c8y-li-action (click)="(false)" icon="rocket">
 *       Launch to space
 *     </c8y-li-action>
 *   </c8y-li>
 * </c8y-list-group>
 * ```
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '../common/icon.directive';
import * as ɵngcc3 from 'ngx-bootstrap/dropdown';
import * as ɵngcc4 from 'ngx-bootstrap/collapse';
import * as ɵngcc5 from '../i18n/c8y-translate.pipe';

function ListItemComponent_div_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 6);
    ɵngcc0.ɵɵprojection(1, 6);
    ɵngcc0.ɵɵelementEnd();
} }
function ListItemComponent_div_10_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 11);
    ɵngcc0.ɵɵlistener("click", function ListItemComponent_div_10_button_1_Template_button_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r7); const ctx_r6 = ɵngcc0.ɵɵnextContext(2); return ctx_r6.toggleCollapsed($event); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 12);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 3, "Expand"));
    ɵngcc0.ɵɵattribute("aria-expanded", !ctx_r3.collapsed);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", "chevron-down");
} }
function ListItemComponent_div_10_div_2_ul_4_ng_container_2_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function ListItemComponent_div_10_div_2_ul_4_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ListItemComponent_div_10_div_2_ul_4_ng_container_2_ng_container_1_Template, 1, 0, "ng-container", 18);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const action_r10 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", action_r10.template);
} }
function ListItemComponent_div_10_div_2_ul_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "ul", 16);
    ɵngcc0.ɵɵprojection(1, 7);
    ɵngcc0.ɵɵtemplate(2, ListItemComponent_div_10_div_2_ul_4_ng_container_2_Template, 2, 1, "ng-container", 17);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r8 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r8.actions);
} }
function ListItemComponent_div_10_div_2_Template(rf, ctx) { if (rf & 1) {
    const _r13 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 13);
    ɵngcc0.ɵɵlistener("onShown", function ListItemComponent_div_10_div_2_Template_div_onShown_0_listener() { ɵngcc0.ɵɵrestoreView(_r13); const ctx_r12 = ɵngcc0.ɵɵnextContext(2); return ctx_r12.showDropdownUp(); });
    ɵngcc0.ɵɵelementStart(1, "button", 14);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 12);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(4, ListItemComponent_div_10_div_2_ul_4_Template, 3, 1, "ul", 15);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("dropup", ctx_r4.dropdownUp);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 3, "Actions"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", "ellipsis-v");
} }
function ListItemComponent_div_10_ng_content_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵprojection(0, 8, ["*ngIf", "showActions"]);
} }
function ListItemComponent_div_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 7);
    ɵngcc0.ɵɵtemplate(1, ListItemComponent_div_10_button_1_Template, 3, 5, "button", 8);
    ɵngcc0.ɵɵtemplate(2, ListItemComponent_div_10_div_2_Template, 5, 5, "div", 9);
    ɵngcc0.ɵɵtemplate(3, ListItemComponent_div_10_ng_content_3_Template, 1, 0, "ng-content", 10);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.showCollapse && ctx_r1.itemCollapse.collapseWay === "button");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.showActions);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.showActions);
} }
function ListItemComponent_div_11_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 19);
    ɵngcc0.ɵɵelementStart(1, "div", 20);
    ɵngcc0.ɵɵprojection(2, 9);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("collapse", ctx_r2.collapsed)("isAnimated", true);
} }
const _c0 = [[["c8y-list-item-drag-handle"], ["c8y-li-drag-handle"]], [["c8y-list-item-radio"], ["c8y-li-radio"]], [["c8y-list-item-checkbox"], ["c8y-li-checkbox"]], [["c8y-list-item-icon"], ["c8y-li-icon"]], [["c8y-list-item-body"], ["c8y-li-body"]], "*", [["c8y-list-item-footer"], ["c8y-li-footer"]], [["c8y-list-item-action"], ["c8y-li-action"]], [["c8y-list-item-action"], ["c8y-li-action"]], [["c8y-list-item-collapse"], ["c8y-li-collapse"]]];
const _c1 = function (a0, a1) { return { "expanded": a0, interact: a1 }; };
const _c2 = ["c8y-list-item-drag-handle, c8y-li-drag-handle", "c8y-list-item-radio, c8y-li-radio", "c8y-list-item-checkbox, c8y-li-checkbox", "c8y-list-item-icon, c8y-li-icon", "c8y-list-item-body, c8y-li-body", "*", "c8y-list-item-footer, c8y-li-footer", "c8y-list-item-action, c8y-li-action", "c8y-list-item-action, c8y-li-action", "c8y-list-item-collapse, c8y-li-collapse"];
export class ListItemComponent {
    /**
     * @ignore
     */
    constructor(element) {
        this.element = element;
        /**
         * If set to true, the class "active" is added which
         * indicates that the current row is active.
         */
        this.active = false;
        /**
         * If set to true, the class "c8y-list__item--empty-actions" is added which adds
         * additional padding to compensate for the "actions" column in other list items.
         */
        this.emptyActions = false;
        /**
         * Indicates if the current list item is collapsed. You can trigger the collapsing from
         * any element event by toggling this value.
         *
         * @example
         * ```html
         * <c8y-li #li (click)="li.collapsed = !li.collapsed">
         *  Toggle
         *  <c8y-li-collapse>
         *    I can be toggled by clicking on the row.
         *  </c8y-li-collapse
         * </c8y-li>
         * ``
         */
        this.collapsed = true;
        /**
         * Indicates if the current list item is selectable.
         */
        this.selectable = true;
        /**
         * An event emitter which is triggered when the user collapses the content
         * via the chevron on the right.
         */
        this.collapsedChange = new EventEmitter();
        /**
         * @ignore
         */
        this.showFooter = false;
        /**
         * @ignore
         */
        this.showActions = false;
        /**
         * @ignore
         */
        this.showCollapse = false;
        /**
         * @ignore
         */
        this.dropdownUp = false;
        this.pulse$ = new Subject();
        this.destroy$ = new Subject();
        this.ACTION_ITEM_HEIGHT = 40;
    }
    /**
     * @ignore
     */
    ngOnInit() {
        this.pulse$
            .pipe(filter(apply => !!apply), tap(() => this.element.nativeElement.classList.remove('c8y-list--pulse')), delay(100), tap(() => this.element.nativeElement.classList.add('c8y-list--pulse')), takeUntil(this.destroy$))
            .subscribe();
    }
    /**
     * @ignore
     */
    ngAfterContentChecked() {
        this.actions = this.itemActions.toArray();
        this.showFooter = this.itemFooter != null;
        this.showActions = this.itemActions.length > 0;
        this.showCollapse = this.itemCollapse != null;
    }
    /**
     * @ignore
     */
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
    /**
     * A helper function which helps to determine if the dropdown
     * should toggle up.
     */
    showDropdownUp() {
        const listItemRect = this.element.nativeElement.getBoundingClientRect();
        const dropdownTop = listItemRect.top;
        const dropdownHeightWithToggle = (1 + this.itemActions.length) * this.ACTION_ITEM_HEIGHT;
        const availableHeight = window.innerHeight;
        const dropdownDownFits = dropdownTop + dropdownHeightWithToggle <= availableHeight;
        this.dropdownUp = !dropdownDownFits;
    }
    /**
     * Toggles the collapse state and emits this state as
     * collapsedChange output as boolean.
     * @param $event Pass optional an event to stop propagation.
     */
    toggleCollapsed($event) {
        if ($event) {
            $event.stopPropagation();
        }
        this.collapsed = !this.collapsed;
        this.collapsedChange.emit(this.collapsed);
    }
    /**
     * Highlights the list-item. This method should be used to show the
     * user that something within this item was changed.
     */
    pulse() {
        this.pulse$.next(true);
    }
}
ListItemComponent.ɵfac = function ListItemComponent_Factory(t) { return new (t || ListItemComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
ListItemComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ListItemComponent, selectors: [["c8y-list-item"], ["c8y-li"]], contentQueries: function ListItemComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, ListItemFooterComponent, 5);
        ɵngcc0.ɵɵcontentQuery(dirIndex, ListItemCollapseComponent, 5);
        ɵngcc0.ɵɵcontentQuery(dirIndex, ListItemActionComponent, 4);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.itemFooter = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.itemCollapse = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.itemActions = _t);
    } }, hostAttrs: [1, "c8y-list__item"], hostVars: 4, hostBindings: function ListItemComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("active", ctx.active)("c8y-list__item--empty-actions", ctx.emptyActions);
    } }, inputs: { active: "active", emptyActions: "emptyActions", collapsed: "collapsed", selectable: "selectable" }, outputs: { collapsedChange: "collapsedChange" }, ngContentSelectors: _c2, decls: 12, vars: 7, consts: [[3, "ngClass"], [1, "c8y-list__item__block"], [1, "c8y-list__item__body", "text-truncate-wrap", 3, "click"], ["class", "c8y-list__item__footer", 4, "ngIf"], ["class", "c8y-list__item__actions", 4, "ngIf"], [3, "collapse", "isAnimated", 4, "ngIf"], [1, "c8y-list__item__footer"], [1, "c8y-list__item__actions"], ["class", "collapse-btn", "type", "button", 3, "title", "click", 4, "ngIf"], ["class", "dropdown", "dropdown", "", 3, "dropup", "onShown", 4, "ngIf"], [4, "ngIf"], ["type", "button", 1, "collapse-btn", 3, "title", "click"], [3, "c8yIcon"], ["dropdown", "", 1, "dropdown", 3, "dropup", "onShown"], ["dropdownToggle", "", "type", "button", "aria-haspopup", "true", 1, "dropdown-toggle", "c8y-dropdown", 3, "title"], ["class", "dropdown-menu dropdown-menu-right", 4, "dropdownMenu"], [1, "dropdown-menu", "dropdown-menu-right"], [4, "ngFor", "ngForOf"], [4, "ngTemplateOutlet"], [3, "collapse", "isAnimated"], [1, "c8y-list__item__collapse--container"]], template: function ListItemComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c0);
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵprojection(2);
        ɵngcc0.ɵɵprojection(3, 1);
        ɵngcc0.ɵɵprojection(4, 2);
        ɵngcc0.ɵɵprojection(5, 3);
        ɵngcc0.ɵɵelementStart(6, "div", 2);
        ɵngcc0.ɵɵlistener("click", function ListItemComponent_Template_div_click_6_listener() { return ctx.itemCollapse && ctx.itemCollapse.collapseWay === "row" && ctx.toggleCollapsed(); });
        ɵngcc0.ɵɵprojection(7, 4);
        ɵngcc0.ɵɵprojection(8, 5);
        ɵngcc0.ɵɵtemplate(9, ListItemComponent_div_9_Template, 2, 0, "div", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(10, ListItemComponent_div_10_Template, 4, 3, "div", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(11, ListItemComponent_div_11_Template, 3, 2, "div", 5);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(4, _c1, !ctx.collapsed, ctx.itemCollapse && ctx.itemCollapse.collapseWay === "row"));
        ɵngcc0.ɵɵadvance(9);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showFooter);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showCollapse && (ctx.itemCollapse && ctx.itemCollapse.collapseWay === "button") || ctx.showActions);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showCollapse);
    } }, directives: [ɵngcc1.NgClass, ɵngcc1.NgIf, ɵngcc2.IconDirective, ɵngcc3.BsDropdownDirective, ɵngcc3.BsDropdownToggleDirective, ɵngcc3.BsDropdownMenuDirective, ɵngcc1.NgForOf, ɵngcc1.NgTemplateOutlet, ɵngcc4.CollapseDirective], pipes: [ɵngcc5.C8yTranslatePipe], encapsulation: 2 });
ListItemComponent.ctorParameters = () => [
    { type: ElementRef }
];
ListItemComponent.propDecorators = {
    active: [{ type: Input }, { type: HostBinding, args: ['class.active',] }],
    emptyActions: [{ type: Input }, { type: HostBinding, args: ['class.c8y-list__item--empty-actions',] }],
    collapsed: [{ type: Input }],
    selectable: [{ type: Input }],
    collapsedChange: [{ type: Output }],
    itemFooter: [{ type: ContentChild, args: [ListItemFooterComponent, { static: false },] }],
    itemActions: [{ type: ContentChildren, args: [ListItemActionComponent,] }],
    itemCollapse: [{ type: ContentChild, args: [ListItemCollapseComponent, { static: false },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ListItemComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-list-item, c8y-li',
                template: "<div [ngClass]=\"{ 'expanded': !collapsed, interact: (itemCollapse && itemCollapse.collapseWay === 'row')}\">\n  <div class=\"c8y-list__item__block\">\n    <ng-content select=\"c8y-list-item-drag-handle, c8y-li-drag-handle\"></ng-content>\n    <ng-content select=\"c8y-list-item-radio, c8y-li-radio\"></ng-content>\n    <ng-content select=\"c8y-list-item-checkbox, c8y-li-checkbox\"></ng-content>\n\n    <ng-content select=\"c8y-list-item-icon, c8y-li-icon\"></ng-content>\n\n    <div class=\"c8y-list__item__body text-truncate-wrap\" (click)=\"(itemCollapse && itemCollapse.collapseWay === 'row') && toggleCollapsed()\">\n      <ng-content select=\"c8y-list-item-body, c8y-li-body\"></ng-content>\n      <ng-content></ng-content>\n      <div class=\"c8y-list__item__footer\" *ngIf=\"showFooter\">\n        <ng-content select=\"c8y-list-item-footer, c8y-li-footer\"></ng-content>\n      </div>\n    </div>\n\n    <div class=\"c8y-list__item__actions\" *ngIf=\"(showCollapse && (itemCollapse && itemCollapse.collapseWay === 'button')) || showActions\">\n      <button\n        class=\"collapse-btn\"\n        title=\"{{ 'Expand' | translate }}\"\n        type=\"button\"\n        (click)=\"toggleCollapsed($event)\"\n        [attr.aria-expanded]=\"!collapsed\"\n        *ngIf=\"showCollapse && itemCollapse.collapseWay === 'button'\"\n      >\n        <i [c8yIcon]=\"'chevron-down'\"></i>\n      </button>\n      <div\n        class=\"dropdown\"\n        dropdown\n        [dropup]=\"dropdownUp\"\n        *ngIf=\"showActions\"\n        (onShown)=\"showDropdownUp()\"\n      >\n        <button\n          class=\"dropdown-toggle c8y-dropdown\"\n          dropdownToggle\n          title=\"{{ 'Actions' | translate }}\"\n          type=\"button\"\n          aria-haspopup=\"true\"\n        >\n          <i [c8yIcon]=\"'ellipsis-v'\"></i>\n        </button>\n        <ul class=\"dropdown-menu dropdown-menu-right\" *dropdownMenu>\n          <ng-content select=\"c8y-list-item-action, c8y-li-action\"></ng-content>\n          <ng-container *ngFor=\"let action of actions\">\n            <ng-container *ngTemplateOutlet=\"action.template\"></ng-container>\n          </ng-container>\n        </ul>\n      </div>\n\n      <ng-content select=\"c8y-list-item-action, c8y-li-action\" *ngIf=\"showActions\"></ng-content>\n    </div>\n  </div>\n\n  <div *ngIf=\"showCollapse\" [collapse]=\"collapsed\" [isAnimated]=\"true\">\n    <div class=\"c8y-list__item__collapse--container\">\n      <ng-content select=\"c8y-list-item-collapse, c8y-li-collapse\"></ng-content>\n    </div>\n  </div>\n</div>\n",
                host: {
                    class: 'c8y-list__item'
                }
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { active: [{
            type: Input
        }, {
            type: HostBinding,
            args: ['class.active']
        }], emptyActions: [{
            type: Input
        }, {
            type: HostBinding,
            args: ['class.c8y-list__item--empty-actions']
        }], collapsed: [{
            type: Input
        }], selectable: [{
            type: Input
        }], collapsedChange: [{
            type: Output
        }], itemFooter: [{
            type: ContentChild,
            args: [ListItemFooterComponent, { static: false }]
        }], itemActions: [{
            type: ContentChildren,
            args: [ListItemActionComponent]
        }], itemCollapse: [{
            type: ContentChild,
            args: [ListItemCollapseComponent, { static: false }]
        }] }); })();
