// @c8y/ngx-components/esm2015/core/list-group/list-group.module.js
import { NgModule } from '@angular/core';
import { ListGroupComponent } from './list-group.component';
import { ListItemComponent } from './list-item.component';
import { CommonModule } from '../common/common.module';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { ListItemIconComponent } from './list-item-icon.component';
import { ListItemBodyComponent } from './list-item-body.component';
import { ListItemFooterComponent } from './list-item-footer.component';
import { ListItemActionComponent } from './list-item-action.component';
import { ListItemCollapseComponent } from './list-item-collapse.component';
import { ListItemCheckboxComponent } from './list-item-checkbox.component';
import { ListItemRadioComponent } from './list-item-radio.component';
import { ListItemTimelineComponent } from './list-item-timeline.component';
import { ListItemDragHandleComponent } from './list-item-drag-handle.component';
/**
 * The list-group module allows you to compose different styles
 * of lists used in Cumulocity IoT. They can consist of plain
 * text, actions, icons and more.
 *
 * @example
 * ```html
 * <c8y-list-group>
 *   <c8y-li>Item 1<c8y-li>
 *   <c8y-li>Item 2<c8y-li>
 * </c8y-list-group>
 * ```
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/collapse';
export class ListGroupModule {
}
ListGroupModule.ɵfac = function ListGroupModule_Factory(t) { return new (t || ListGroupModule)(); };
ListGroupModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: ListGroupModule });
ListGroupModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule, CollapseModule.forRoot(), BsDropdownModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ListGroupModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, CollapseModule.forRoot(), BsDropdownModule],
                declarations: [
                    ListGroupComponent,
                    ListItemComponent,
                    ListItemIconComponent,
                    ListItemBodyComponent,
                    ListItemFooterComponent,
                    ListItemActionComponent,
                    ListItemCollapseComponent,
                    ListItemCheckboxComponent,
                    ListItemRadioComponent,
                    ListItemTimelineComponent,
                    ListItemDragHandleComponent
                ],
                exports: [
                    ListGroupComponent,
                    ListItemComponent,
                    ListItemIconComponent,
                    ListItemBodyComponent,
                    ListItemFooterComponent,
                    ListItemActionComponent,
                    ListItemCollapseComponent,
                    ListItemCheckboxComponent,
                    ListItemRadioComponent,
                    ListItemTimelineComponent,
                    ListItemDragHandleComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ListGroupModule, { declarations: function () { return [ListGroupComponent,
        ListItemComponent,
        ListItemIconComponent,
        ListItemBodyComponent,
        ListItemFooterComponent,
        ListItemActionComponent,
        ListItemCollapseComponent,
        ListItemCheckboxComponent,
        ListItemRadioComponent,
        ListItemTimelineComponent,
        ListItemDragHandleComponent]; }, imports: function () { return [CommonModule, ɵngcc1.CollapseModule, BsDropdownModule]; }, exports: function () { return [ListGroupComponent,
        ListItemComponent,
        ListItemIconComponent,
        ListItemBodyComponent,
        ListItemFooterComponent,
        ListItemActionComponent,
        ListItemCollapseComponent,
        ListItemCheckboxComponent,
        ListItemRadioComponent,
        ListItemTimelineComponent,
        ListItemDragHandleComponent]; } }); })();
