// @c8y/ngx-components/esm2015/core/dynamic-forms/dynamic-forms-ui-theme.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { FormlyModule, FORMLY_CONFIG } from '@ngx-formly/core';
import { FormlySelectModule } from '@ngx-formly/core/select';
import { BsDatepickerModule } from 'ngx-bootstrap/datepicker';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { CommonModule as C8YCommonModule } from '../common/common.module';
import { DropAreaModule } from '../drop-area/drop-area.module';
import { FormsModule } from '../forms/index';
import { ListGroupModule } from '../list-group/list-group.module';
import { SearchModule } from '../search/search.module';
import { SelectModule } from '../select/select.module';
import { C8yTranslateModule } from '../i18n/c8y-translate.module';
import { ArrayTypeComponent } from './array/array.type.component';
import { FieldCheckbox } from './checkbox/checkbox.type.component';
import { FieldDate } from './date/date.type.component';
import { FieldFile } from './file/file.type.component';
import { WrapperFormField } from './form-field/form-field.wrapper.component';
import { FieldInput } from './input/input.type.component';
import { ObjectTypeComponent } from './object/object.type.component';
import { FieldRadio } from './radio/radio.type.component';
import { FieldDateTime } from './date-time/date-time.type.component';
import { DateTimePickerModule } from '../date-time-picker/date-time-picker.module';
import { TypeaheadTypeComponent } from './typeahead/typeahead.type.component';
import * as ɵngcc0 from '@angular/core';
export const FIELD_TYPES = [
    FieldCheckbox,
    ArrayTypeComponent,
    WrapperFormField,
    FieldFile,
    FieldInput,
    FieldRadio,
    ObjectTypeComponent,
    FieldDate,
    FieldDateTime,
    TypeaheadTypeComponent
];
export function fieldTypeFactory() {
    // tslint:disable-next-line:no-string-literal
    return FIELD_TYPES.map(type => type['CONFIG']).reduce((formlyConfig, typeConfig) => (Object.assign(Object.assign({}, formlyConfig), { types: [...formlyConfig.types, ...(typeConfig.types || [])], wrappers: [...formlyConfig.wrappers, ...(typeConfig.wrappers || [])] })), { types: [], wrappers: [] });
}
/**
 * This module declares field type components for dynamic forms that match C8Y styleguide.
 *
 * @dynamic
 */
export class DynamicFormsUIThemeModule {
}
DynamicFormsUIThemeModule.ɵfac = function DynamicFormsUIThemeModule_Factory(t) { return new (t || DynamicFormsUIThemeModule)(); };
DynamicFormsUIThemeModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DynamicFormsUIThemeModule });
DynamicFormsUIThemeModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        {
            provide: FORMLY_CONFIG,
            useFactory: fieldTypeFactory,
            multi: true
        }
    ], imports: [[
            CommonModule,
            C8YCommonModule,
            PopoverModule,
            ReactiveFormsModule,
            C8yTranslateModule,
            DropAreaModule,
            FormsModule,
            FormlyModule,
            FormlySelectModule,
            DateTimePickerModule,
            SelectModule,
            ListGroupModule,
            SearchModule,
            BsDatepickerModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DynamicFormsUIThemeModule, [{
        type: NgModule,
        args: [{
                declarations: [...FIELD_TYPES],
                entryComponents: [...FIELD_TYPES],
                imports: [
                    CommonModule,
                    C8YCommonModule,
                    PopoverModule,
                    ReactiveFormsModule,
                    C8yTranslateModule,
                    DropAreaModule,
                    FormsModule,
                    FormlyModule,
                    FormlySelectModule,
                    DateTimePickerModule,
                    SelectModule,
                    ListGroupModule,
                    SearchModule,
                    BsDatepickerModule
                ],
                providers: [
                    {
                        provide: FORMLY_CONFIG,
                        useFactory: fieldTypeFactory,
                        multi: true
                    }
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DynamicFormsUIThemeModule, { declarations: function () { return [FieldCheckbox,
        ArrayTypeComponent,
        WrapperFormField,
        FieldFile,
        FieldInput,
        FieldRadio,
        ObjectTypeComponent,
        FieldDate,
        FieldDateTime,
        TypeaheadTypeComponent]; }, imports: function () { return [CommonModule,
        C8YCommonModule,
        PopoverModule,
        ReactiveFormsModule,
        C8yTranslateModule,
        DropAreaModule,
        FormsModule,
        FormlyModule,
        FormlySelectModule,
        DateTimePickerModule,
        SelectModule,
        ListGroupModule,
        SearchModule,
        BsDatepickerModule]; } }); })();
