// @c8y/ngx-components/esm2015/core/dashboard/widgets-dashboard.component.js
import { Component, EventEmitter, Input, Output } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '../header/title/title.component';
import * as ɵngcc3 from '../breadcrumb/breadcrumb.component';
import * as ɵngcc4 from '../breadcrumb/breadcrumb-item.component';
import * as ɵngcc5 from '../action-bar/action-bar-item.component';
import * as ɵngcc6 from '../common/icon.directive';
import * as ɵngcc7 from '../i18n/c8y-translate.directive';
import * as ɵngcc8 from '../docs/guide-docs.component';
import * as ɵngcc9 from '../docs/guide-href.directive';
import * as ɵngcc10 from './dashboard.component';
import * as ɵngcc11 from './dashboard-child.component';
import * as ɵngcc12 from './dashboard-child-title.component';
import * as ɵngcc13 from './dashboard-child-action.component';
import * as ɵngcc14 from '../dynamic-component/dynamic-component.component';
import * as ɵngcc15 from '../i18n/c8y-translate.pipe';

function WidgetsDashboardComponent_c8y_title_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-title");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, ctx_r0.settings.title), "\n");
} }
function WidgetsDashboardComponent_c8y_breadcrumb_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-breadcrumb");
    ɵngcc0.ɵɵelement(1, "c8y-breadcrumb-item", 7);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("icon", ctx_r1.breadcrumb.icon)("label", ctx_r1.breadcrumb.label)("path", ctx_r1.breadcrumb.path);
} }
function WidgetsDashboardComponent_c8y_action_bar_item_2_Template(rf, ctx) { if (rf & 1) {
    const _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-action-bar-item", 8);
    ɵngcc0.ɵɵelementStart(1, "button", 9);
    ɵngcc0.ɵɵlistener("click", function WidgetsDashboardComponent_c8y_action_bar_item_2_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r12); const ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.onAddWidget.emit(); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 10);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("placement", "right");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 4, "Add widget"));
    ɵngcc0.ɵɵproperty("disabled", ctx_r2.settings.isFrozen || ctx_r2.settings.isDisabled);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 6, "Add widget"), " ");
} }
function WidgetsDashboardComponent_c8y_action_bar_item_3_Template(rf, ctx) { if (rf & 1) {
    const _r14 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-action-bar-item", 8);
    ɵngcc0.ɵɵelementStart(1, "button", 9);
    ɵngcc0.ɵɵlistener("click", function WidgetsDashboardComponent_c8y_action_bar_item_3_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r14); const ctx_r13 = ɵngcc0.ɵɵnextContext(); return ctx_r13.onEditDashboard.emit(); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 11);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("placement", "right");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 4, "Edit"));
    ɵngcc0.ɵɵproperty("disabled", ctx_r3.settings.isFrozen || ctx_r3.settings.isDisabled);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 6, "Edit"), " ");
} }
function WidgetsDashboardComponent_c8y_action_bar_item_4_Template(rf, ctx) { if (rf & 1) {
    const _r16 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-action-bar-item", 8);
    ɵngcc0.ɵɵelementStart(1, "button", 12);
    ɵngcc0.ɵɵlistener("click", function WidgetsDashboardComponent_c8y_action_bar_item_4_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r16); const ctx_r15 = ɵngcc0.ɵɵnextContext(); return ctx_r15.toggleFullscreen(); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 13);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("placement", "right");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 4, "Full screen"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", ctx_r4.fullScreen() ? "compress" : "expand");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 6, "Full screen"), " ");
} }
const _c0 = function (a0) { return { active: a0 }; };
function WidgetsDashboardComponent_c8y_action_bar_item_5_Template(rf, ctx) { if (rf & 1) {
    const _r18 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-action-bar-item", 14);
    ɵngcc0.ɵɵelementStart(1, "button", 15);
    ɵngcc0.ɵɵlistener("click", function WidgetsDashboardComponent_c8y_action_bar_item_5_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r18); const ctx_r17 = ɵngcc0.ɵɵnextContext(); return ctx_r17.onFreeze.emit(ctx_r17.settings); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 13);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("placement", "right");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 5, "Lock/unlock this dashboard"));
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(7, _c0, ctx_r5.settings.isFrozen))("disabled", ctx_r5.settings.isDisabled);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", ctx_r5.settings.isFrozen ? "lock" : "unlock");
} }
function WidgetsDashboardComponent_c8y_action_bar_item_6_Template(rf, ctx) { if (rf & 1) {
    const _r20 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-action-bar-item", 16);
    ɵngcc0.ɵɵelementStart(1, "button", 17);
    ɵngcc0.ɵɵlistener("click", function WidgetsDashboardComponent_c8y_action_bar_item_6_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r19 = ɵngcc0.ɵɵnextContext(); return ctx_r19.onCopyDashboard.emit(); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 18);
    ɵngcc0.ɵɵelementStart(4, "span", 19);
    ɵngcc0.ɵɵtext(5, "Copy dashboard");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("placement", "more")("priority", -2000);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 3, "Copy dashboard"));
} }
function WidgetsDashboardComponent_c8y_action_bar_item_7_Template(rf, ctx) { if (rf & 1) {
    const _r22 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-action-bar-item", 16);
    ɵngcc0.ɵɵelementStart(1, "button", 20);
    ɵngcc0.ɵɵlistener("click", function WidgetsDashboardComponent_c8y_action_bar_item_7_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r22); const ctx_r21 = ɵngcc0.ɵɵnextContext(); return ctx_r21.onDeleteDashboard.emit(); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 21);
    ɵngcc0.ɵɵelementStart(4, "span", 19);
    ɵngcc0.ɵɵtext(5, "Delete dashboard");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("placement", "more")("priority", -3000);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 4, "Delete dashboard"));
    ɵngcc0.ɵɵproperty("disabled", ctx_r7.settings.isFrozen || ctx_r7.settings.isDisabled);
} }
function WidgetsDashboardComponent_div_8_div_4_Template(rf, ctx) { if (rf & 1) {
    const _r25 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "p", 24);
    ɵngcc0.ɵɵtext(2, "Add widgets to this dashboard.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div");
    ɵngcc0.ɵɵelementStart(4, "button", 25);
    ɵngcc0.ɵɵlistener("click", function WidgetsDashboardComponent_div_8_div_4_Template_button_click_4_listener() { ɵngcc0.ɵɵrestoreView(_r25); const ctx_r24 = ɵngcc0.ɵɵnextContext(2); return ctx_r24.onAddWidget.emit(); });
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵtext(6, " Add widget ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "p", 26);
    ɵngcc0.ɵɵelementStart(8, "small", 24);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(9, " Find out more in the ");
    ɵngcc0.ɵɵelementStart(10, "a", 27);
    ɵngcc0.ɵɵtext(11, " User guide`KEEP_ORIGINAL` ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(12, ". ");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r23 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(5, 2, "Add widget"));
    ɵngcc0.ɵɵproperty("disabled", ctx_r23.settings.isFrozen || ctx_r23.settings.isDisabled);
} }
function WidgetsDashboardComponent_div_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 22);
    ɵngcc0.ɵɵelement(1, "h1", 23);
    ɵngcc0.ɵɵelementStart(2, "h3", 24);
    ɵngcc0.ɵɵtext(3, "No widgets to display.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(4, WidgetsDashboardComponent_div_8_div_4_Template, 13, 4, "div", 0);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r8 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r8.settings.isFrozen && ctx_r8.onAddWidget.observers.length);
} }
function WidgetsDashboardComponent_div_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 28);
    ɵngcc0.ɵɵelementStart(1, "div", 29);
    ɵngcc0.ɵɵelement(2, "div", 30);
    ɵngcc0.ɵɵelement(3, "div", 31);
    ɵngcc0.ɵɵelement(4, "div", 32);
    ɵngcc0.ɵɵelement(5, "div", 33);
    ɵngcc0.ɵɵelement(6, "div", 34);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function WidgetsDashboardComponent_c8y_dashboard_10_c8y_dashboard_child_2_span_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const widget_r28 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, widget_r28.title), " ");
} }
function WidgetsDashboardComponent_c8y_dashboard_10_c8y_dashboard_child_2_span_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const widget_r28 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", widget_r28.title, " ");
} }
function WidgetsDashboardComponent_c8y_dashboard_10_c8y_dashboard_child_2_c8y_dashboard_child_action_5_Template(rf, ctx) { if (rf & 1) {
    const _r39 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-dashboard-child-action");
    ɵngcc0.ɵɵelementStart(1, "button", 17);
    ɵngcc0.ɵɵlistener("click", function WidgetsDashboardComponent_c8y_dashboard_10_c8y_dashboard_child_2_c8y_dashboard_child_action_5_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r39); const widget_r28 = ɵngcc0.ɵɵnextContext().$implicit; const _r29 = ɵngcc0.ɵɵreference(1); ɵngcc0.ɵɵnextContext(); const _r26 = ɵngcc0.ɵɵreference(1); const ctx_r37 = ɵngcc0.ɵɵnextContext(); return ctx_r37.onEditWidget.emit({ widget: widget_r28, source: _r29, dashboard: _r26 }); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 11);
    ɵngcc0.ɵɵelementStart(4, "span", 19);
    ɵngcc0.ɵɵtext(5, "Edit");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 1, "Edit widget"));
} }
function WidgetsDashboardComponent_c8y_dashboard_10_c8y_dashboard_child_2_c8y_dashboard_child_action_6_Template(rf, ctx) { if (rf & 1) {
    const _r42 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-dashboard-child-action");
    ɵngcc0.ɵɵelementStart(1, "button", 17);
    ɵngcc0.ɵɵlistener("click", function WidgetsDashboardComponent_c8y_dashboard_10_c8y_dashboard_child_2_c8y_dashboard_child_action_6_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r42); const widget_r28 = ɵngcc0.ɵɵnextContext().$implicit; const _r29 = ɵngcc0.ɵɵreference(1); ɵngcc0.ɵɵnextContext(); const _r26 = ɵngcc0.ɵɵreference(1); const ctx_r40 = ɵngcc0.ɵɵnextContext(); return ctx_r40.onDeleteWidget.emit({ widget: widget_r28, source: _r29, dashboard: _r26 }); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 21);
    ɵngcc0.ɵɵelementStart(4, "span", 19);
    ɵngcc0.ɵɵtext(5, "Remove");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 1, "Remove widget"));
} }
const _c1 = function (a0, a1, a2) { return { child: a0, dashboard: a1, context: a2 }; };
function WidgetsDashboardComponent_c8y_dashboard_10_c8y_dashboard_child_2_c8y_dynamic_component_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-dynamic-component", 41);
} if (rf & 2) {
    const widget_r28 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r34 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("componentId", widget_r28.componentId || widget_r28.name)("config", widget_r28.templateUrl || widget_r28.widgetComponent ? ɵngcc0.ɵɵpureFunction3(2, _c1, widget_r28, ctx_r34.contextDashboard, ctx_r34.context) : widget_r28.config);
} }
function WidgetsDashboardComponent_c8y_dashboard_10_c8y_dashboard_child_2_Template(rf, ctx) { if (rf & 1) {
    const _r45 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-dashboard-child", 38, 39);
    ɵngcc0.ɵɵlistener("changeStart", function WidgetsDashboardComponent_c8y_dashboard_10_c8y_dashboard_child_2_Template_c8y_dashboard_child_changeStart_0_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r45); const widget_r28 = restoredCtx.$implicit; const _r29 = ɵngcc0.ɵɵreference(1); ɵngcc0.ɵɵnextContext(); const _r26 = ɵngcc0.ɵɵreference(1); const ctx_r44 = ɵngcc0.ɵɵnextContext(); return ctx_r44.onChangeStart.emit({ widget: widget_r28, source: _r29, dashboard: _r26 }); })("changeEnd", function WidgetsDashboardComponent_c8y_dashboard_10_c8y_dashboard_child_2_Template_c8y_dashboard_child_changeEnd_0_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r45); const widget_r28 = restoredCtx.$implicit; const _r29 = ɵngcc0.ɵɵreference(1); ɵngcc0.ɵɵnextContext(); const _r26 = ɵngcc0.ɵɵreference(1); const ctx_r46 = ɵngcc0.ɵɵnextContext(); return ctx_r46.onChangeEnd.emit({ widget: widget_r28, source: _r29, dashboard: _r26 }); });
    ɵngcc0.ɵɵelementStart(2, "c8y-dashboard-child-title");
    ɵngcc0.ɵɵtemplate(3, WidgetsDashboardComponent_c8y_dashboard_10_c8y_dashboard_child_2_span_3_Template, 3, 3, "span", 0);
    ɵngcc0.ɵɵtemplate(4, WidgetsDashboardComponent_c8y_dashboard_10_c8y_dashboard_child_2_span_4_Template, 2, 1, "span", 0);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(5, WidgetsDashboardComponent_c8y_dashboard_10_c8y_dashboard_child_2_c8y_dashboard_child_action_5_Template, 6, 3, "c8y-dashboard-child-action", 0);
    ɵngcc0.ɵɵtemplate(6, WidgetsDashboardComponent_c8y_dashboard_10_c8y_dashboard_child_2_c8y_dashboard_child_action_6_Template, 6, 3, "c8y-dashboard-child-action", 0);
    ɵngcc0.ɵɵtemplate(7, WidgetsDashboardComponent_c8y_dashboard_10_c8y_dashboard_child_2_c8y_dynamic_component_7_Template, 1, 6, "c8y-dynamic-component", 40);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const widget_r28 = ctx.$implicit;
    const _r29 = ɵngcc0.ɵɵreference(1);
    const ctx_r27 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassMap(widget_r28.classes);
    ɵngcc0.ɵɵproperty("x", widget_r28._x)("y", widget_r28._y)("width", widget_r28._width || ctx_r27.settings.defaultWidth)("height", widget_r28._height || ctx_r27.settings.defaultHeight)("margin", ctx_r27.settings.widgetMargin)("isFrozen", ctx_r27.settings.isFrozen || ctx_r27.settings.isDisabled)("data", widget_r28)("useIntersection", true);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r27.settings.translateWidgetTitle);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r27.settings.translateWidgetTitle);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r27.onEditWidget.observers.length);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r27.onDeleteWidget.observers.length);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", _r29.intersected);
} }
function WidgetsDashboardComponent_c8y_dashboard_10_Template(rf, ctx) { if (rf & 1) {
    const _r48 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-dashboard", 35, 36);
    ɵngcc0.ɵɵlistener("dashboardChange", function WidgetsDashboardComponent_c8y_dashboard_10_Template_c8y_dashboard_dashboardChange_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r48); const ctx_r47 = ɵngcc0.ɵɵnextContext(); return ctx_r47.onChangeDashboard.emit($event); });
    ɵngcc0.ɵɵtemplate(2, WidgetsDashboardComponent_c8y_dashboard_10_c8y_dashboard_child_2_Template, 8, 15, "c8y-dashboard-child", 37);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r10 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r10.widgets);
} }
export class WidgetsDashboardComponent {
    constructor() {
        this.contextDashboard = { updateTarget: true };
        this.settings = {
            isFrozen: false,
            isDisabled: false,
            widgetMargin: 12,
            translateWidgetTitle: false,
            defaultHeight: 4,
            defaultWidth: 4,
            allowFullscreen: false,
            canCopy: true,
            canDelete: true,
            isLoading: false
        };
        this.onAddWidget = new EventEmitter();
        this.onEditWidget = new EventEmitter();
        this.onDeleteWidget = new EventEmitter();
        this.onChangeDashboard = new EventEmitter();
        this.onEditDashboard = new EventEmitter();
        this.onCopyDashboard = new EventEmitter();
        this.onDeleteDashboard = new EventEmitter();
        this.onFreeze = new EventEmitter();
        this.onChangeStart = new EventEmitter();
        this.onChangeEnd = new EventEmitter();
    }
    set _settings(settings) {
        this.settings = Object.assign(Object.assign({}, this.settings), settings);
    }
    toggleFullscreen() {
        const elem = document.body;
        const doc = document;
        if (!this.fullScreen()) {
            if (elem.requestFullscreen) {
                elem.requestFullscreen();
            }
            else if (elem.msRequestFullscreen) {
                elem.msRequestFullscreen();
            }
            else if (elem.mozRequestFullScreen) {
                elem.mozRequestFullScreen();
            }
            else if (elem.webkitRequestFullscreen) {
                elem.webkitRequestFullscreen(Element.ALLOW_KEYBOARD_INPUT);
            }
        }
        else if (doc.exitFullscreen) {
            doc.exitFullscreen();
        }
        else if (doc.msExitFullscreen) {
            doc.msExitFullscreen();
        }
        else if (doc.mozCancelFullScreen) {
            doc.mozCancelFullScreen();
        }
        else if (doc.webkitExitFullscreen) {
            doc.webkitExitFullscreen();
        }
    }
    fullScreen() {
        const doc = document;
        return !!(doc.fullscreenElement ||
            doc.mozFullScreenElement ||
            doc.webkitFullscreenElement ||
            doc.msFullscreenElement);
    }
}
WidgetsDashboardComponent.ɵfac = function WidgetsDashboardComponent_Factory(t) { return new (t || WidgetsDashboardComponent)(); };
WidgetsDashboardComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: WidgetsDashboardComponent, selectors: [["c8y-widgets-dashboard"]], hostAttrs: [1, "dashboard", "c8y-grid-dashboard", 2, "display", "block"], inputs: { contextDashboard: "contextDashboard", _settings: ["settings", "_settings"], widgets: "widgets", context: "context", breadcrumb: "breadcrumb" }, outputs: { onAddWidget: "onAddWidget", onEditWidget: "onEditWidget", onDeleteWidget: "onDeleteWidget", onChangeDashboard: "onChangeDashboard", onEditDashboard: "onEditDashboard", onCopyDashboard: "onCopyDashboard", onDeleteDashboard: "onDeleteDashboard", onFreeze: "onFreeze", onChangeStart: "onChangeStart", onChangeEnd: "onChangeEnd" }, decls: 11, vars: 11, consts: [[4, "ngIf"], [3, "placement", 4, "ngIf"], ["itemClass", "navbar-form", 3, "placement", 4, "ngIf"], [3, "placement", "priority", 4, "ngIf"], ["class", "c8y-empty-state text-center", 4, "ngIf"], ["class", "col-xs-12 text-center", 4, "ngIf"], [3, "dashboardChange", 4, "ngIf"], [3, "icon", "label", "path"], [3, "placement"], [1, "btn", "btn-link", 3, "title", "disabled", "click"], ["c8yIcon", "plus-circle"], ["c8yIcon", "pencil"], [1, "btn", "btn-link", 3, "title", "click"], [3, "c8yIcon"], ["itemClass", "navbar-form", 3, "placement"], ["uib-btn-checkbox", "", 1, "btn", "btn-default", 3, "title", "ngClass", "disabled", "click"], [3, "placement", "priority"], [3, "title", "click"], ["c8yIcon", "clone"], ["translate", "", 1, "m-l-4"], [3, "title", "disabled", "click"], ["c8yIcon", "trash"], [1, "c8y-empty-state", "text-center"], [1, "c8y-icon", "c8y-icon-device", "c8y-icon-duocolor"], ["translate", ""], ["translate", "", 1, "btn", "btn-primary", 3, "title", "disabled", "click"], ["c8y-guide-docs", ""], ["c8y-guide-href", "users-guide/cockpit/#dashboards"], [1, "col-xs-12", "text-center"], [1, "spinner", "p-relative"], [1, "rect1"], [1, "rect2"], [1, "rect3"], [1, "rect4"], [1, "rect5"], [3, "dashboardChange"], ["dashboard", ""], [3, "x", "y", "width", "height", "margin", "isFrozen", "class", "data", "useIntersection", "changeStart", "changeEnd", 4, "ngFor", "ngForOf"], [3, "x", "y", "width", "height", "margin", "isFrozen", "data", "useIntersection", "changeStart", "changeEnd"], ["child", ""], [3, "componentId", "config", 4, "ngIf"], [3, "componentId", "config"]], template: function WidgetsDashboardComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, WidgetsDashboardComponent_c8y_title_0_Template, 3, 3, "c8y-title", 0);
        ɵngcc0.ɵɵtemplate(1, WidgetsDashboardComponent_c8y_breadcrumb_1_Template, 2, 3, "c8y-breadcrumb", 0);
        ɵngcc0.ɵɵtemplate(2, WidgetsDashboardComponent_c8y_action_bar_item_2_Template, 6, 8, "c8y-action-bar-item", 1);
        ɵngcc0.ɵɵtemplate(3, WidgetsDashboardComponent_c8y_action_bar_item_3_Template, 6, 8, "c8y-action-bar-item", 1);
        ɵngcc0.ɵɵtemplate(4, WidgetsDashboardComponent_c8y_action_bar_item_4_Template, 6, 8, "c8y-action-bar-item", 1);
        ɵngcc0.ɵɵtemplate(5, WidgetsDashboardComponent_c8y_action_bar_item_5_Template, 4, 9, "c8y-action-bar-item", 2);
        ɵngcc0.ɵɵtemplate(6, WidgetsDashboardComponent_c8y_action_bar_item_6_Template, 6, 5, "c8y-action-bar-item", 3);
        ɵngcc0.ɵɵtemplate(7, WidgetsDashboardComponent_c8y_action_bar_item_7_Template, 6, 6, "c8y-action-bar-item", 3);
        ɵngcc0.ɵɵtemplate(8, WidgetsDashboardComponent_div_8_Template, 5, 1, "div", 4);
        ɵngcc0.ɵɵtemplate(9, WidgetsDashboardComponent_div_9_Template, 7, 0, "div", 5);
        ɵngcc0.ɵɵtemplate(10, WidgetsDashboardComponent_c8y_dashboard_10_Template, 3, 1, "c8y-dashboard", 6);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", !!ctx.settings.title);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !!ctx.breadcrumb);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.onAddWidget.observers.length);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.onEditDashboard.observers.length);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.settings.allowFullscreen);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.onFreeze.observers.length);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.settings.canCopy);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.settings.canDelete && ctx.onDeleteDashboard.observers.length);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.settings.isLoading && (ctx.widgets == null ? null : ctx.widgets.length) === 0);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.settings.isLoading);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.settings.isLoading);
    } }, directives: [ɵngcc1.NgIf, ɵngcc2.TitleComponent, ɵngcc3.BreadcrumbComponent, ɵngcc4.BreadcrumbItemComponent, ɵngcc5.ActionBarItemComponent, ɵngcc6.IconDirective, ɵngcc1.NgClass, ɵngcc7.C8yTranslateDirective, ɵngcc8.GuideDocsComponent, ɵngcc9.GuideHrefDirective, ɵngcc10.DashboardComponent, ɵngcc1.NgForOf, ɵngcc11.DashboardChildComponent, ɵngcc12.DashboardChildTitleComponent, ɵngcc13.DashboardChildActionComponent, ɵngcc14.DynamicComponentComponent], pipes: [ɵngcc15.C8yTranslatePipe], encapsulation: 2 });
WidgetsDashboardComponent.propDecorators = {
    widgets: [{ type: Input }],
    context: [{ type: Input }],
    contextDashboard: [{ type: Input }],
    _settings: [{ type: Input, args: ['settings',] }],
    breadcrumb: [{ type: Input }],
    onAddWidget: [{ type: Output }],
    onEditWidget: [{ type: Output }],
    onDeleteWidget: [{ type: Output }],
    onChangeDashboard: [{ type: Output }],
    onEditDashboard: [{ type: Output }],
    onCopyDashboard: [{ type: Output }],
    onDeleteDashboard: [{ type: Output }],
    onFreeze: [{ type: Output }],
    onChangeStart: [{ type: Output }],
    onChangeEnd: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(WidgetsDashboardComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-widgets-dashboard',
                template: "<c8y-title *ngIf=\"!!settings.title\">\n  {{ settings.title | translate }}\n</c8y-title>\n\n<c8y-breadcrumb *ngIf=\"!!breadcrumb\">\n  <c8y-breadcrumb-item\n    [icon]=\"breadcrumb.icon\"\n    [label]=\"breadcrumb.label\"\n    [path]=\"breadcrumb.path\"\n  ></c8y-breadcrumb-item>\n</c8y-breadcrumb>\n\n<c8y-action-bar-item [placement]=\"'right'\" *ngIf=\"onAddWidget.observers.length\">\n  <button\n    title=\"{{ 'Add widget' | translate }}\"\n    class=\"btn btn-link\"\n    [disabled]=\"settings.isFrozen || settings.isDisabled\"\n    (click)=\"onAddWidget.emit()\"\n  >\n    <i c8yIcon=\"plus-circle\"></i>\n    {{ 'Add widget' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\" *ngIf=\"onEditDashboard.observers.length\">\n  <button\n    title=\"{{ 'Edit' | translate }}\"\n    class=\"btn btn-link\"\n    [disabled]=\"settings.isFrozen || settings.isDisabled\"\n    (click)=\"onEditDashboard.emit()\"\n  >\n    <i c8yIcon=\"pencil\"></i>\n    {{ 'Edit' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\" *ngIf=\"settings.allowFullscreen\">\n  <button title=\"{{ 'Full screen' | translate }}\" class=\"btn btn-link\" (click)=\"toggleFullscreen()\">\n    <i [c8yIcon]=\"fullScreen() ? 'compress' : 'expand'\"></i>\n    {{ 'Full screen' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item\n  [placement]=\"'right'\"\n  itemClass=\"navbar-form\"\n  *ngIf=\"onFreeze.observers.length\"\n>\n  <button\n    title=\"{{ 'Lock/unlock this dashboard' | translate }}\"\n    class=\"btn btn-default\"\n    [ngClass]=\"{ active: settings.isFrozen }\"\n    (click)=\"onFreeze.emit(settings)\"\n    [disabled]=\"settings.isDisabled\"\n    uib-btn-checkbox\n  >\n    <i [c8yIcon]=\"settings.isFrozen ? 'lock' : 'unlock'\"></i>\n  </button>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'more'\" [priority]=\"-2000\" *ngIf=\"settings.canCopy\">\n  <button title=\"{{ 'Copy dashboard' | translate }}\" (click)=\"onCopyDashboard.emit()\">\n    <i c8yIcon=\"clone\"></i>\n    <span class=\"m-l-4\" translate>Copy dashboard</span>\n  </button>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item\n  [placement]=\"'more'\"\n  [priority]=\"-3000\"\n  *ngIf=\"settings.canDelete && onDeleteDashboard.observers.length\"\n>\n  <button\n    title=\"{{ 'Delete dashboard' | translate }}\"\n    (click)=\"onDeleteDashboard.emit()\"\n    [disabled]=\"settings.isFrozen || settings.isDisabled\"\n  >\n    <i c8yIcon=\"trash\"></i>\n    <span class=\"m-l-4\" translate>Delete dashboard</span>\n  </button>\n</c8y-action-bar-item>\n\n<!-- empty state -->\n<div class=\"c8y-empty-state text-center\" *ngIf=\"!settings.isLoading && widgets?.length === 0\">\n  <h1 class=\"c8y-icon c8y-icon-device c8y-icon-duocolor\"></h1>\n  <h3 translate>No widgets to display.</h3>\n  <div *ngIf=\"!settings.isFrozen && onAddWidget.observers.length\">\n    <p translate>Add widgets to this dashboard.</p>\n    <div>\n      <button\n        title=\"{{ 'Add widget' | translate }}\"\n        [disabled]=\"settings.isFrozen || settings.isDisabled\"\n        (click)=\"onAddWidget.emit()\"\n        class=\"btn btn-primary\"\n        translate\n      >\n        Add widget\n      </button>\n    </div>\n    <p c8y-guide-docs>\n      <small translate ngNonBindable>\n        Find out more in the\n        <a c8y-guide-href=\"users-guide/cockpit/#dashboards\"> User guide`KEEP_ORIGINAL` </a>.\n      </small>\n    </p>\n  </div>\n</div>\n<!-- /.empty settings -->\n\n<div class=\"col-xs-12 text-center\" *ngIf=\"settings.isLoading\">\n  <div class=\"spinner p-relative\">\n    <div class=\"rect1\"></div>\n    <div class=\"rect2\"></div>\n    <div class=\"rect3\"></div>\n    <div class=\"rect4\"></div>\n    <div class=\"rect5\"></div>\n  </div>\n</div>\n\n<c8y-dashboard\n  (dashboardChange)=\"onChangeDashboard.emit($event)\"\n  #dashboard\n  *ngIf=\"!settings.isLoading\"\n>\n  <c8y-dashboard-child\n    *ngFor=\"let widget of widgets\"\n    [x]=\"widget._x\"\n    [y]=\"widget._y\"\n    [width]=\"widget._width || settings.defaultWidth\"\n    [height]=\"widget._height || settings.defaultHeight\"\n    [margin]=\"settings.widgetMargin\"\n    [isFrozen]=\"settings.isFrozen || settings.isDisabled\"\n    [class]=\"widget.classes\"\n    [data]=\"widget\"\n    [useIntersection]=\"true\"\n    (changeStart)=\"onChangeStart.emit({ widget: widget, source: child, dashboard: dashboard })\"\n    (changeEnd)=\"onChangeEnd.emit({ widget: widget, source: child, dashboard: dashboard })\"\n    #child\n  >\n    <c8y-dashboard-child-title>\n      <span *ngIf=\"settings.translateWidgetTitle\">\n        {{ widget.title | translate }}\n      </span>\n      <span *ngIf=\"!settings.translateWidgetTitle\">\n        {{ widget.title }}\n      </span>\n    </c8y-dashboard-child-title>\n    <c8y-dashboard-child-action *ngIf=\"onEditWidget.observers.length\">\n      <button\n        title=\"{{ 'Edit widget' | translate }}\"\n        (click)=\"onEditWidget.emit({ widget: widget, source: child, dashboard: dashboard })\"\n      >\n        <i c8yIcon=\"pencil\"></i>\n        <span class=\"m-l-4\" translate>Edit</span>\n      </button>\n    </c8y-dashboard-child-action>\n    <c8y-dashboard-child-action *ngIf=\"onDeleteWidget.observers.length\">\n      <button\n        title=\"{{ 'Remove widget' | translate }}\"\n        (click)=\"onDeleteWidget.emit({ widget: widget, source: child, dashboard: dashboard })\"\n      >\n        <i c8yIcon=\"trash\"></i>\n        <span class=\"m-l-4\" translate>Remove</span>\n      </button>\n    </c8y-dashboard-child-action>\n    <c8y-dynamic-component\n      [componentId]=\"widget.componentId || widget.name\"\n      [config]=\"\n        widget.templateUrl || widget.widgetComponent\n          ? { child: widget, dashboard: contextDashboard, context: context }\n          : widget.config\n      \"\n      *ngIf=\"child.intersected\"\n    ></c8y-dynamic-component>\n  </c8y-dashboard-child>\n</c8y-dashboard>\n",
                host: {
                    style: `
      display: block;
    `,
                    class: 'dashboard c8y-grid-dashboard'
                }
            }]
    }], function () { return []; }, { contextDashboard: [{
            type: Input
        }], onAddWidget: [{
            type: Output
        }], onEditWidget: [{
            type: Output
        }], onDeleteWidget: [{
            type: Output
        }], onChangeDashboard: [{
            type: Output
        }], onEditDashboard: [{
            type: Output
        }], onCopyDashboard: [{
            type: Output
        }], onDeleteDashboard: [{
            type: Output
        }], onFreeze: [{
            type: Output
        }], onChangeStart: [{
            type: Output
        }], onChangeEnd: [{
            type: Output
        }], _settings: [{
            type: Input,
            args: ['settings']
        }], widgets: [{
            type: Input
        }], context: [{
            type: Input
        }], breadcrumb: [{
            type: Input
        }] }); })();
