// @c8y/ngx-components/esm2015/core/select-modal/select-modal.component.js
import { __awaiter } from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { assign } from 'lodash-es';
import { BsModalRef } from 'ngx-bootstrap/modal';
import { Subject } from 'rxjs';
import { debounceTime } from 'rxjs/operators';
import { gettext } from '../i18n/gettext';
import { ModalSelectionMode } from './select-modal.model';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/modal';
import * as ɵngcc2 from '../common/icon.directive';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from '../list-group/list-group.component';
import * as ɵngcc5 from '../forms/filter-input.component';
import * as ɵngcc6 from '../list-group/list-item.component';
import * as ɵngcc7 from '../list-group/list-item-icon.component';
import * as ɵngcc8 from '../list-group/list-item-body.component';
import * as ɵngcc9 from '../list-group/list-item-collapse.component';
import * as ɵngcc10 from '../search/highlight.component';
import * as ɵngcc11 from '../list-group/list-item-radio.component';
import * as ɵngcc12 from '../i18n/c8y-translate.directive';
import * as ɵngcc13 from '../i18n/c8y-translate.pipe';
import * as ɵngcc14 from './select-modal-filter.pipe';

const _c0 = function (a0, a1) { return { "col-xs-6": a0, "col-xs-12": a1 }; };
function SelectModalComponent_div_11_Template(rf, ctx) { if (rf & 1) {
    const _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 16);
    ɵngcc0.ɵɵelementStart(1, "c8y-filter", 17);
    ɵngcc0.ɵɵlistener("onSearch", function SelectModalComponent_div_11_Template_c8y_filter_onSearch_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r7); const ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.updatePipe($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(2, _c0, ctx_r0.showFilter && ctx_r0.additionalFilterTemplate, !ctx_r0.showFilter || !ctx_r0.additionalFilterTemplate));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("icon", "search");
} }
function SelectModalComponent_div_12_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function SelectModalComponent_div_12_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 16);
    ɵngcc0.ɵɵtemplate(1, SelectModalComponent_div_12_ng_container_1_Template, 1, 0, "ng-container", 18);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(2, _c0, ctx_r1.showFilter && ctx_r1.additionalFilterTemplate, !ctx_r1.showFilter || !ctx_r1.additionalFilterTemplate));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r1.additionalFilterTemplate);
} }
function SelectModalComponent_div_15_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 19);
    ɵngcc0.ɵɵelementStart(1, "div", 20);
    ɵngcc0.ɵɵelement(2, "h1", 21);
    ɵngcc0.ɵɵelementStart(3, "p");
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    let tmp_1_0;
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate1("c8yIcon", "", ctx_r2.icon, " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(5, 2, (tmp_1_0 = ɵngcc0.ɵɵpipeBind1(6, 4, ctx_r2.noItemsMessage)) !== null && tmp_1_0 !== undefined ? tmp_1_0 : "No items to display."));
} }
function SelectModalComponent_c8y_li_17_div_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 16);
    ɵngcc0.ɵɵelement(1, "c8y-highlight", 26);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const bodyPart_r13 = ctx.$implicit;
    const ctx_r10 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngClass", bodyPart_r13.class);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("title", bodyPart_r13.value)("pattern", ctx_r10.filterTerm)("text", bodyPart_r13.value);
} }
function SelectModalComponent_c8y_li_17_div_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 27);
    ɵngcc0.ɵɵelementStart(1, "div", 16);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r9 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", item_r9.additionalInformation.class);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", item_r9.additionalInformation.value, " ");
} }
function SelectModalComponent_c8y_li_17_c8y_li_9_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 16);
    ɵngcc0.ɵɵelement(1, "c8y-highlight", 31);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const optionPart_r18 = ctx.$implicit;
    const ctx_r16 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("ngClass", optionPart_r18.class);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("pattern", ctx_r16.filterTerm)("text", optionPart_r18.value);
} }
const _c1 = function (a0, a1) { return { $implicit: a0, option: a1 }; };
function SelectModalComponent_c8y_li_17_c8y_li_9_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0, 32);
} if (rf & 2) {
    const option_r15 = ɵngcc0.ɵɵnextContext().$implicit;
    const item_r9 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", option_r15.template)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(2, _c1, item_r9, option_r15));
} }
function SelectModalComponent_c8y_li_17_c8y_li_9_Template(rf, ctx) { if (rf & 1) {
    const _r24 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li");
    ɵngcc0.ɵɵelementStart(1, "c8y-li-radio", 28);
    ɵngcc0.ɵɵlistener("onSelect", function SelectModalComponent_c8y_li_17_c8y_li_9_Template_c8y_li_radio_onSelect_1_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r24); const option_r15 = restoredCtx.$implicit; const item_r9 = ɵngcc0.ɵɵnextContext().$implicit; const ctx_r22 = ɵngcc0.ɵɵnextContext(); return ctx_r22.updateChoice({ item: item_r9, id: option_r15.obj.id }); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(2, "c8y-li-body", 29);
    ɵngcc0.ɵɵtemplate(3, SelectModalComponent_c8y_li_17_c8y_li_9_div_3_Template, 2, 3, "div", 24);
    ɵngcc0.ɵɵtemplate(4, SelectModalComponent_c8y_li_17_c8y_li_9_ng_container_4_Template, 1, 5, "ng-container", 30);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const option_r15 = ctx.$implicit;
    const item_r9 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r12 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("name", ctx_r12.mode === "single" ? "single" : item_r9.groupId)("disabled", option_r15.disabled)("selected", option_r15.selected);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", option_r15.body);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", option_r15.template);
} }
function SelectModalComponent_c8y_li_17_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li");
    ɵngcc0.ɵɵelementStart(1, "c8y-li-icon");
    ɵngcc0.ɵɵelement(2, "i", 2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "c8y-li-body", 22);
    ɵngcc0.ɵɵelementStart(4, "div", 23);
    ɵngcc0.ɵɵtemplate(5, SelectModalComponent_c8y_li_17_div_5_Template, 2, 4, "div", 24);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(6, SelectModalComponent_c8y_li_17_div_6_Template, 3, 2, "div", 25);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "c8y-li-collapse");
    ɵngcc0.ɵɵelementStart(8, "c8y-list-group");
    ɵngcc0.ɵɵtemplate(9, SelectModalComponent_c8y_li_17_c8y_li_9_Template, 5, 5, "c8y-li", 11);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r9 = ctx.$implicit;
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("c8yIcon", ctx_r3.icon);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngForOf", item_r9.body);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r9.additionalInformation);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngForOf", item_r9.options);
} }
function SelectModalComponent_div_19_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "div", 33);
    ɵngcc0.ɵɵtext(2, " Some entries might not be shown. Try narrowing search criteria. ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function SelectModalComponent_button_21_Template(rf, ctx) { if (rf & 1) {
    const _r27 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 34);
    ɵngcc0.ɵɵlistener("click", function SelectModalComponent_button_21_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r27); const ctx_r26 = ɵngcc0.ɵɵnextContext(); return ctx_r26.dismiss(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, ctx_r5.labels.cancel));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 4, ctx_r5.labels.cancel), " ");
} }
export class SelectModalComponent {
    constructor(bsModalRef) {
        this.bsModalRef = bsModalRef;
        this.subTitle = gettext('Select from the list of items matching the device type');
        this.mode = ModalSelectionMode.MULTI;
        this.disableSelected = true;
        this.showFilter = true;
        this.areMoreEntries = false;
        this.result = new EventEmitter();
        this.search = new EventEmitter();
        this.onChoiceUpdated = new EventEmitter();
        this.selected = false;
        this.filterTerm = '';
        this.listItems = [];
        this.debouncer = new Subject();
        this._labels = { ok: gettext('Confirm'), cancel: gettext('Cancel') };
        this.debouncer.pipe(debounceTime(500)).subscribe(value => {
            this.search.emit(value);
        });
    }
    set labels(labels) {
        const { ok = this.labels.ok, cancel = this.labels.cancel } = labels || {};
        this._labels = { ok, cancel };
    }
    get labels() {
        return this._labels;
    }
    ngOnChanges(changes) {
        return __awaiter(this, void 0, void 0, function* () {
            if (changes.items && changes.items.currentValue) {
                const itemsPromise = changes.items.currentValue.map((item) => __awaiter(this, void 0, void 0, function* () {
                    item.options = yield item.options;
                    const selected = item.options.find(option => option.selected);
                    if (selected) {
                        item.selectedId = selected.obj.id;
                        if (this.disableSelected) {
                            item.options.map(option => assign(option, { disabled: true }));
                        }
                    }
                    return item;
                }));
                this.listItems = yield Promise.all(itemsPromise);
            }
        });
    }
    updatePipe(filterTerm) {
        this.debouncer.next(filterTerm);
        this.filterTerm = filterTerm;
    }
    updateChoice({ item, id }) {
        if (this.mode === 'single') {
            this.listItems.map(value => (value.selectedId = undefined));
        }
        item.selectedId = id;
        this.selected = true;
        this.onChoiceUpdated.emit(item);
    }
    dismiss() {
        this.bsModalRef.hide();
    }
    select() {
        this.result.emit(this.getOutput());
        this.bsModalRef.hide();
    }
    ngOnDestroy() {
        this.debouncer.complete();
        this.result.complete();
        this.search.complete();
    }
    getOutput() {
        return this.listItems
            .filter(item => item.selectedId)
            .map(item => item.options.find(option => item.selectedId === option.obj.id))
            .filter(option => !option.selected)
            .map(selectedOption => selectedOption.obj);
    }
}
SelectModalComponent.ɵfac = function SelectModalComponent_Factory(t) { return new (t || SelectModalComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BsModalRef)); };
SelectModalComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SelectModalComponent, selectors: [["c8y-select-modal"]], inputs: { subTitle: "subTitle", mode: "mode", disableSelected: "disableSelected", showFilter: "showFilter", areMoreEntries: "areMoreEntries", labels: "labels", icon: "icon", title: "title", items: "items", additionalFilterTemplate: "additionalFilterTemplate", noItemsMessage: "noItemsMessage" }, outputs: { result: "result", search: "search", onChoiceUpdated: "onChoiceUpdated" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], decls: 26, vars: 23, consts: [[1, "viewport-modal"], [1, "modal-header", "dialog-header"], [3, "c8yIcon"], [1, "text-uppercase"], [1, "p-16", "text-center", "separator-bottom", "min-height-fit"], [1, "m-b-8"], [1, "row"], [3, "ngClass", 4, "ngIf"], [1, "modal-inner-scroll"], [1, "p-l-16", "p-r-16"], ["class", "panel m-t-8 m-b-8", 4, "ngIf"], [4, "ngFor", "ngForOf"], [4, "ngIf"], [1, "modal-footer"], ["class", "btn btn-default", 3, "title", "click", 4, "ngIf"], [1, "btn", "btn-primary", 3, "title", "disabled", "click"], [3, "ngClass"], [3, "icon", "onSearch"], [4, "ngTemplateOutlet"], [1, "panel", "m-t-8", "m-b-8"], [1, "c8y-empty-state", "text-left"], [1, "c8y-icon-duocolor", 3, "c8yIcon"], [1, "content-flex-30"], [1, "col-9"], [3, "ngClass", 4, "ngFor", "ngForOf"], ["class", "col-3 text-right", 4, "ngIf"], [3, "title", "pattern", "text"], [1, "col-3", "text-right"], [3, "name", "disabled", "selected", "onSelect"], [1, "content-flex-20"], [3, "ngTemplateOutlet", "ngTemplateOutletContext", 4, "ngIf"], [3, "pattern", "text"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], ["translate", "", 1, "alert", "alert-info", "m-t-16", "m-r-8", "m-l-8"], [1, "btn", "btn-default", 3, "title", "click"]], template: function SelectModalComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelement(2, "span", 2);
        ɵngcc0.ɵɵelementStart(3, "h4", 3);
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "div", 4);
        ɵngcc0.ɵɵelementStart(7, "p", 5);
        ɵngcc0.ɵɵtext(8);
        ɵngcc0.ɵɵpipe(9, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(10, "div", 6);
        ɵngcc0.ɵɵtemplate(11, SelectModalComponent_div_11_Template, 2, 5, "div", 7);
        ɵngcc0.ɵɵtemplate(12, SelectModalComponent_div_12_Template, 2, 5, "div", 7);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(13, "div", 8);
        ɵngcc0.ɵɵelementStart(14, "div", 9);
        ɵngcc0.ɵɵtemplate(15, SelectModalComponent_div_15_Template, 7, 6, "div", 10);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(16, "c8y-list-group");
        ɵngcc0.ɵɵtemplate(17, SelectModalComponent_c8y_li_17_Template, 10, 4, "c8y-li", 11);
        ɵngcc0.ɵɵpipe(18, "selectModalFilterPipe");
        ɵngcc0.ɵɵtemplate(19, SelectModalComponent_div_19_Template, 3, 0, "div", 12);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(20, "div", 13);
        ɵngcc0.ɵɵtemplate(21, SelectModalComponent_button_21_Template, 4, 6, "button", 14);
        ɵngcc0.ɵɵelementStart(22, "button", 15);
        ɵngcc0.ɵɵlistener("click", function SelectModalComponent_Template_button_click_22_listener() { return ctx.select(); });
        ɵngcc0.ɵɵpipe(23, "translate");
        ɵngcc0.ɵɵtext(24);
        ɵngcc0.ɵɵpipe(25, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("c8yIcon", ctx.icon);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 12, ctx.title), " ");
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(9, 14, ctx.subTitle));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showFilter);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.additionalFilterTemplate);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.items || ctx.items.length === 0);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngForOf", ɵngcc0.ɵɵpipeBind2(18, 16, ctx.listItems, ctx.filterTerm));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.areMoreEntries);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.labels.cancel);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(23, 19, ctx.labels.ok));
        ɵngcc0.ɵɵproperty("disabled", !ctx.selected);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(25, 21, ctx.labels.ok), " ");
    } }, directives: [ɵngcc2.IconDirective, ɵngcc3.NgIf, ɵngcc4.ListGroupComponent, ɵngcc3.NgForOf, ɵngcc3.NgClass, ɵngcc5.FilterInputComponent, ɵngcc3.NgTemplateOutlet, ɵngcc6.ListItemComponent, ɵngcc7.ListItemIconComponent, ɵngcc8.ListItemBodyComponent, ɵngcc9.ListItemCollapseComponent, ɵngcc10.HighlightComponent, ɵngcc11.ListItemRadioComponent, ɵngcc12.C8yTranslateDirective], pipes: [ɵngcc13.C8yTranslatePipe, ɵngcc14.SelectModalFilterPipe], encapsulation: 2 });
SelectModalComponent.ctorParameters = () => [
    { type: BsModalRef }
];
SelectModalComponent.propDecorators = {
    icon: [{ type: Input }],
    title: [{ type: Input }],
    subTitle: [{ type: Input }],
    items: [{ type: Input }],
    mode: [{ type: Input }],
    disableSelected: [{ type: Input }],
    showFilter: [{ type: Input }],
    additionalFilterTemplate: [{ type: Input }],
    areMoreEntries: [{ type: Input }],
    labels: [{ type: Input }],
    noItemsMessage: [{ type: Input }],
    result: [{ type: Output }],
    search: [{ type: Output }],
    onChoiceUpdated: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SelectModalComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-select-modal',
                template: "<div class=\"viewport-modal\">\n  <div class=\"modal-header dialog-header\">\n    <span c8yIcon=\"{{ icon }}\"></span>\n    <h4 class=\"text-uppercase\">\n      {{ title | translate }}\n    </h4>\n  </div>\n  <div class=\"p-16 text-center separator-bottom min-height-fit\">\n    <p class=\"m-b-8\">{{ subTitle | translate }}</p>\n    <div class=\"row\">\n      <div\n        *ngIf=\"showFilter\"\n        [ngClass]=\"{\n          'col-xs-6': showFilter && additionalFilterTemplate,\n          'col-xs-12': !showFilter || !additionalFilterTemplate\n        }\"\n      >\n        <c8y-filter [icon]=\"'search'\" (onSearch)=\"updatePipe($event)\"></c8y-filter>\n      </div>\n      <div\n        *ngIf=\"additionalFilterTemplate\"\n        [ngClass]=\"{\n          'col-xs-6': showFilter && additionalFilterTemplate,\n          'col-xs-12': !showFilter || !additionalFilterTemplate\n        }\"\n      >\n        <ng-container *ngTemplateOutlet=\"additionalFilterTemplate\"></ng-container>\n      </div>\n    </div>\n  </div>\n  <div class=\"modal-inner-scroll\">\n    <div class=\"p-l-16 p-r-16\">\n      <div class=\"panel m-t-8 m-b-8\" *ngIf=\"!items || items.length === 0\">\n        <div class=\"c8y-empty-state text-left\">\n          <h1 c8yIcon=\"{{ icon }} \" class=\"c8y-icon-duocolor\"></h1>\n          <p>{{ (noItemsMessage | translate) ?? 'No items to display.' | translate }}</p>\n        </div>\n      </div>\n    </div>\n    <c8y-list-group>\n      <c8y-li *ngFor=\"let item of listItems | selectModalFilterPipe: filterTerm\">\n        <c8y-li-icon>\n          <i c8yIcon=\"{{ icon }}\"></i>\n        </c8y-li-icon>\n\n        <c8y-li-body class=\"content-flex-30\">\n          <div class=\"col-9\">\n            <div *ngFor=\"let bodyPart of item.body\" [ngClass]=\"bodyPart.class\">\n              <c8y-highlight\n                [title]=\"bodyPart.value\"\n                [pattern]=\"filterTerm\"\n                [text]=\"bodyPart.value\"\n              ></c8y-highlight>\n            </div>\n          </div>\n\n          <div class=\"col-3 text-right\" *ngIf=\"item.additionalInformation\">\n            <div [ngClass]=\"item.additionalInformation.class\">\n              {{ item.additionalInformation.value }}\n            </div>\n          </div>\n        </c8y-li-body>\n\n        <c8y-li-collapse>\n          <c8y-list-group>\n            <c8y-li *ngFor=\"let option of item.options\">\n              <c8y-li-radio\n                [name]=\"mode === 'single' ? 'single' : item.groupId\"\n                (onSelect)=\"updateChoice({ item: item, id: option.obj.id })\"\n                [disabled]=\"option.disabled\"\n                [selected]=\"option.selected\"\n              >\n              </c8y-li-radio>\n              <c8y-li-body class=\"content-flex-20\">\n                <div\n                  *ngFor=\"let optionPart of option.body; let i = index\"\n                  [ngClass]=\"optionPart.class\"\n                >\n                  <c8y-highlight [pattern]=\"filterTerm\" [text]=\"optionPart.value\"></c8y-highlight>\n                </div>\n                <ng-container\n                  *ngIf=\"option.template\"\n                  [ngTemplateOutlet]=\"option.template\"\n                  [ngTemplateOutletContext]=\"{ $implicit: item, option: option }\"\n                ></ng-container>\n              </c8y-li-body>\n            </c8y-li>\n          </c8y-list-group>\n        </c8y-li-collapse>\n      </c8y-li>\n      <div *ngIf=\"areMoreEntries\">\n        <div class=\"alert alert-info m-t-16 m-r-8 m-l-8\" translate>\n          Some entries might not be shown. Try narrowing search criteria.\n        </div>\n      </div>\n    </c8y-list-group>\n  </div>\n\n  <div class=\"modal-footer\">\n    <button\n      title=\"{{ labels.cancel | translate }}\"\n      *ngIf=\"labels.cancel\"\n      class=\"btn btn-default\"\n      (click)=\"dismiss()\"\n    >\n      {{ labels.cancel | translate }}\n    </button>\n    <button\n      title=\"{{ labels.ok | translate }}\"\n      class=\"btn btn-primary\"\n      (click)=\"select()\"\n      [disabled]=\"!selected\"\n    >\n      {{ labels.ok | translate }}\n    </button>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.BsModalRef }]; }, { subTitle: [{
            type: Input
        }], mode: [{
            type: Input
        }], disableSelected: [{
            type: Input
        }], showFilter: [{
            type: Input
        }], areMoreEntries: [{
            type: Input
        }], result: [{
            type: Output
        }], search: [{
            type: Output
        }], onChoiceUpdated: [{
            type: Output
        }], labels: [{
            type: Input
        }], icon: [{
            type: Input
        }], title: [{
            type: Input
        }], items: [{
            type: Input
        }], additionalFilterTemplate: [{
            type: Input
        }], noItemsMessage: [{
            type: Input
        }] }); })();
