// @c8y/ngx-components/esm2015/operations/stepper-frames/preview-block.directive.js
import { Directive, Input, TemplateRef } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class PreviewBlockDirective {
    constructor(templateRef) {
        this.templateRef = templateRef;
        this.show = true;
    }
}
PreviewBlockDirective.ɵfac = function PreviewBlockDirective_Factory(t) { return new (t || PreviewBlockDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.TemplateRef)); };
PreviewBlockDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: PreviewBlockDirective, selectors: [["", "previewBlock", ""]], inputs: { show: ["previewBlockIf", "show"], label: ["previewBlock", "label"] } });
PreviewBlockDirective.ctorParameters = () => [
    { type: TemplateRef }
];
PreviewBlockDirective.propDecorators = {
    label: [{ type: Input, args: ['previewBlock',] }],
    show: [{ type: Input, args: ['previewBlockIf',] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(PreviewBlockDirective, [{
        type: Directive,
        args: [{
                selector: '[previewBlock]'
            }]
    }], function () { return [{ type: ɵngcc0.TemplateRef }]; }, { show: [{
            type: Input,
            args: ['previewBlockIf']
        }], label: [{
            type: Input,
            args: ['previewBlock']
        }] }); })();
