// @c8y/ngx-components/esm2015/operations/stepper-frames/select-step-frame.component.js
import { Component, Input } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '@c8y/ngx-components';

function SelectStepFrameComponent_c8y_search_result_empty_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-search-result-empty");
} }
const _c0 = [[["", "filters", ""]], [["", "header", ""]], [["", "list", ""]]];
const _c1 = ["[filters]", "[header]", "[list]"];
export class SelectStepFrameComponent {
}
SelectStepFrameComponent.ɵfac = function SelectStepFrameComponent_Factory(t) { return new (t || SelectStepFrameComponent)(); };
SelectStepFrameComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SelectStepFrameComponent, selectors: [["c8y-select-step-frame"]], inputs: { header: "header", noResults: "noResults" }, ngContentSelectors: _c1, decls: 13, vars: 4, consts: [[1, "card-block", "overflow-visible", "flex-no-shrink", "col-xs-12", "p-b-0", "p-t-4", "separator-bottom"], [1, "flex-row", "j-c-center", "m-b-24"], [1, "col-xs-12", "col-sm-8"], [1, "text-normal", "text-center", "m-b-16"], [4, "ngIf"], [1, "col-xs-12", "flex-grow", "no-gutter"], [1, "card-inner-scroll", "fit-h"], [1, "card-block", "p-t-0", "p-b-0"]], template: function SelectStepFrameComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c0);
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelementStart(3, "h4", 3);
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵprojection(6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵprojection(7, 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(8, SelectStepFrameComponent_c8y_search_result_empty_8_Template, 1, 0, "c8y-search-result-empty", 4);
        ɵngcc0.ɵɵelementStart(9, "div", 5);
        ɵngcc0.ɵɵelementStart(10, "div", 6);
        ɵngcc0.ɵɵelementStart(11, "div", 7);
        ɵngcc0.ɵɵprojection(12, 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 2, ctx.header), " ");
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngIf", ctx.noResults);
    } }, directives: [ɵngcc1.NgIf, ɵngcc2.SearchResultEmptyComponent], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
SelectStepFrameComponent.propDecorators = {
    header: [{ type: Input }],
    noResults: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SelectStepFrameComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-select-step-frame',
                template: "<div class=\"card-block overflow-visible flex-no-shrink col-xs-12 p-b-0 p-t-4 separator-bottom\">\n  <div class=\"flex-row j-c-center m-b-24\">\n    <div class=\"col-xs-12 col-sm-8\">\n      <h4 class=\"text-normal text-center m-b-16\">\n        {{ header | translate }}\n      </h4>\n      <ng-content select=\"[filters]\"></ng-content>\n    </div>\n  </div>\n  <ng-content select=\"[header]\"></ng-content>\n</div>\n<c8y-search-result-empty *ngIf=\"noResults\"> </c8y-search-result-empty>\n<div class=\"col-xs-12 flex-grow no-gutter\">\n  <div class=\"card-inner-scroll fit-h\">\n    <div class=\"card-block p-t-0 p-b-0\">\n      <ng-content select=\"[list]\"></ng-content>\n    </div>\n  </div>\n</div>\n"
            }]
    }], null, { header: [{
            type: Input
        }], noResults: [{
            type: Input
        }] }); })();
