// @c8y/ngx-components/esm2015/operations/stepper-frames/stepper-frames.modules.js
import { NgModule } from '@angular/core';
import { CoreModule } from '@c8y/ngx-components';
import { PreviewBlockDirective } from './preview-block.directive';
import { PreviewSelectionFrameComponent } from './preview-selection-frame.component';
import { SelectStepFrameComponent } from './select-step-frame.component';
import * as ɵngcc0 from '@angular/core';
export class StepperFramesModule {
}
StepperFramesModule.ɵfac = function StepperFramesModule_Factory(t) { return new (t || StepperFramesModule)(); };
StepperFramesModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: StepperFramesModule });
StepperFramesModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CoreModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(StepperFramesModule, [{
        type: NgModule,
        args: [{
                imports: [CoreModule],
                declarations: [PreviewSelectionFrameComponent, SelectStepFrameComponent, PreviewBlockDirective],
                exports: [PreviewSelectionFrameComponent, SelectStepFrameComponent, PreviewBlockDirective]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(StepperFramesModule, { declarations: function () { return [PreviewSelectionFrameComponent, SelectStepFrameComponent, PreviewBlockDirective]; }, imports: function () { return [CoreModule]; }, exports: function () { return [PreviewSelectionFrameComponent, SelectStepFrameComponent, PreviewBlockDirective]; } }); })();
