// @c8y/ngx-components/esm2015/core/bootstrap/bootstrap.component.js
import { __awaiter } from "tslib";
import { Component } from '@angular/core';
import { of } from 'rxjs';
import { distinctUntilChanged, map, switchMap } from 'rxjs/operators';
import { ActionBarService } from '../action-bar/action-bar.service';
import { OptionsService } from '../common/options.service';
import { Permissions } from '../common/permissions.service';
import { AppStateService } from '../common/ui-state.service';
import { HeaderService } from '../header/header.service';
import { NavigatorService } from '../navigator/navigator.service';
import { PluginsResolveService } from '../plugins/plugins-resolve.service';
import { NEEDED_ROLE_FOR_SETUP, SETUP_FINISHED_STEP_ID } from '../setup/setup.model';
import { Steppers } from '../stepper/stepper.model';
import { StepperService } from '../stepper/stepper.service';
import { TabsService } from '../tabs/tabs.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../tabs/tabs.service';
import * as ɵngcc2 from '../common/ui-state.service';
import * as ɵngcc3 from '../navigator/navigator.service';
import * as ɵngcc4 from '../action-bar/action-bar.service';
import * as ɵngcc5 from '../header/header.service';
import * as ɵngcc6 from '../common/options.service';
import * as ɵngcc7 from '../plugins/plugins-resolve.service';
import * as ɵngcc8 from '../stepper/stepper.service';
import * as ɵngcc9 from '../common/permissions.service';

function BootstrapComponent_c8y_login_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-login");
} }
function BootstrapComponent_div_2_div_16_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 15);
    ɵngcc0.ɵɵelementStart(1, "p", 16);
    ɵngcc0.ɵɵtext(2, "powered by Cumulocity");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
const _c0 = function (a0, a1, a2, a3, a4, a5) { return { open: a0, "no-apps-margin": a1, "horizontal-tabs": a2, "vertical-tabs": a3, "has-tabs": a4, "has-action-bar": a5 }; };
function BootstrapComponent_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelement(1, "c8y-header-bar", null, 2);
    ɵngcc0.ɵɵelementStart(3, "c8y-navigator-outlet", 3);
    ɵngcc0.ɵɵpipe(4, "async");
    ɵngcc0.ɵɵpipe(5, "async");
    ɵngcc0.ɵɵelementStart(6, "header", 4);
    ɵngcc0.ɵɵelement(7, "div", 5);
    ɵngcc0.ɵɵelement(8, "c8y-app-icon", 6);
    ɵngcc0.ɵɵpipe(9, "async");
    ɵngcc0.ɵɵpipe(10, "async");
    ɵngcc0.ɵɵelementStart(11, "span");
    ɵngcc0.ɵɵtext(12);
    ɵngcc0.ɵɵpipe(13, "async");
    ɵngcc0.ɵɵpipe(14, "humanizeAppName");
    ɵngcc0.ɵɵpipe(15, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(16, BootstrapComponent_div_2_div_16_Template, 3, 0, "div", 7);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(17, "div", 8);
    ɵngcc0.ɵɵelement(18, "c8y-alert-outlet");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(19, "c8y-tabs-outlet", 9, 10);
    ɵngcc0.ɵɵpipe(21, "async");
    ɵngcc0.ɵɵpipe(22, "async");
    ɵngcc0.ɵɵpipe(23, "async");
    ɵngcc0.ɵɵelement(24, "c8y-action-bar", 11, 12);
    ɵngcc0.ɵɵpipe(26, "async");
    ɵngcc0.ɵɵelementStart(27, "div", 13);
    ɵngcc0.ɵɵpipe(28, "async");
    ɵngcc0.ɵɵpipe(29, "async");
    ɵngcc0.ɵɵelementStart(30, "div", 14);
    ɵngcc0.ɵɵelement(31, "router-outlet");
    ɵngcc0.ɵɵprojection(32);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _r5 = ɵngcc0.ɵɵreference(20);
    const _r6 = ɵngcc0.ɵɵreference(25);
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassProp("head-open", ctx_r1.headerService.headerOpen);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("nodes", ɵngcc0.ɵɵpipeBind1(4, 16, ctx_r1.navigator.items$))("open", ɵngcc0.ɵɵpipeBind1(5, 18, ctx_r1.headerService.navigatorOpen$));
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵproperty("name", ɵngcc0.ɵɵpipeBind1(9, 20, ctx_r1.ui.state$).app.name)("contextPath", ɵngcc0.ɵɵpipeBind1(10, 22, ctx_r1.ui.state$).app.contextPath);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(13, 24, ɵngcc0.ɵɵpipeBind1(14, 26, ɵngcc0.ɵɵpipeBind1(15, 28, ctx_r1.ui.state$).app)));
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.showPoweredBy);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("tabs", ɵngcc0.ɵɵpipeBind1(21, 30, ctx_r1.tabs.items$))("navigatorOpen", ɵngcc0.ɵɵpipeBind1(22, 32, ctx_r1.headerService.navigatorOpen$))("orientation", ɵngcc0.ɵɵpipeBind1(23, 34, ctx_r1.tabs.orientation$));
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵproperty("navigatorOpen", ɵngcc0.ɵɵpipeBind1(26, 36, ctx_r1.headerService.navigatorOpen$))("hasTabs", _r5.hasTabs)("isTabsHorizontal", _r5 == null ? null : _r5.isHorizontal)("items$", ctx_r1.actionBar.items$);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction6(42, _c0, ɵngcc0.ɵɵpipeBind1(28, 38, ctx_r1.headerService.navigatorOpen$), ɵngcc0.ɵɵpipeBind1(29, 40, ctx_r1.noAppsMargin$), _r5.isHorizontal, !_r5.isHorizontal, _r5.hasTabs, !(_r6 == null ? null : _r6.hidden)));
} }
function BootstrapComponent_div_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelement(1, "c8y-header-bar", 17);
    ɵngcc0.ɵɵelementStart(2, "div", 18);
    ɵngcc0.ɵɵelementStart(3, "div", 14);
    ɵngcc0.ɵɵelement(4, "c8y-setup");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("simple", true);
} }
const _c1 = [[["", "id", "c8y-legacy-view"]]];
const _c2 = ["#c8y-legacy-view"];
export class BootstrapComponent {
    constructor(tabs, ui, navigator, actionBar, headerService, options, pluginsResolve, stepperService, permissions) {
        this.tabs = tabs;
        this.ui = ui;
        this.navigator = navigator;
        this.actionBar = actionBar;
        this.headerService = headerService;
        this.options = options;
        this.pluginsResolve = pluginsResolve;
        this.stepperService = stepperService;
        this.permissions = permissions;
        this.showPoweredBy = true;
        this.noAppsMargin$ = this.headerService.map(({ nav }) => !nav.open && nav.hiddenOnStartup);
        this.tabsOrientation = this.options.tabsHorizontal ? 'horizontal' : 'vertical';
        this.ui
            .map(({ lang }) => lang)
            .pipe(distinctUntilChanged())
            .subscribe(() => {
            this.actionBar.refresh();
        });
        this.showPoweredBy = !this.options.get('hidePowered');
        this.isSetupNeeded$ = this.ui.currentApplication.pipe(map(app => this.ui.isOwnerOfApplication(app)), switchMap(isOwner => {
            if (!isOwner) {
                return of([]);
            }
            return this.stepperService.getById$(Steppers.SETUP);
        }), map((steps) => this.getNotCompletedSetupSteps(steps)), map(notCompletedSetupSteps => this.options.forceSetup || this.options.isSetup && this.needsSetup(notCompletedSetupSteps)));
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            const { remoteModules } = window.C8Y_APP;
            if (remoteModules.length) {
                yield this.pluginsResolve.resolveRemotePlugins(remoteModules);
            }
        });
    }
    needsSetup(notCompletedSetupSteps) {
        notCompletedSetupSteps = this.excludeSetupFinishStep(notCompletedSetupSteps);
        if (this.hasPermission()) {
            return notCompletedSetupSteps.length > 0;
        }
        // Show setup screen as there are required steps but the user
        // has no permission. The Setup screen will show an error to
        // the user.
        return this.hasRequiredSteps(notCompletedSetupSteps);
    }
    excludeSetupFinishStep(notCompletedSetupSteps) {
        return notCompletedSetupSteps.filter(({ setupId }) => setupId !== SETUP_FINISHED_STEP_ID);
    }
    hasRequiredSteps(notCompletedSetupSteps) {
        return notCompletedSetupSteps.some(step => step.required);
    }
    hasPermission() {
        return this.permissions.hasRole(NEEDED_ROLE_FOR_SETUP);
    }
    getNotCompletedSetupSteps(steps) {
        if (!this.ui.currentApplicationConfig.value) {
            return steps;
        }
        return steps.filter(step => step.setupId && !(this.ui.currentApplicationConfig.value.setup || []).includes(step.setupId));
    }
}
BootstrapComponent.ɵfac = function BootstrapComponent_Factory(t) { return new (t || BootstrapComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.TabsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.AppStateService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.NavigatorService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.ActionBarService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.HeaderService), ɵngcc0.ɵɵdirectiveInject(ɵngcc6.OptionsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc7.PluginsResolveService), ɵngcc0.ɵɵdirectiveInject(ɵngcc8.StepperService), ɵngcc0.ɵɵdirectiveInject(ɵngcc9.Permissions)); };
BootstrapComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: BootstrapComponent, selectors: [["c8y-bootstrap"]], ngContentSelectors: _c2, decls: 9, vars: 13, consts: [[4, "ngIf"], [3, "head-open", 4, "ngIf"], ["header", ""], [3, "nodes", "open"], [1, "title", "navigator-slot-top"], [1, "tenant-brand"], [3, "name", "contextPath"], ["class", "powered-by navigator-slot-bottom", 4, "ngIf"], [1, "alerts"], [3, "tabs", "navigatorOpen", "orientation"], ["tabsComponent", ""], [3, "navigatorOpen", "hasTabs", "isTabsHorizontal", "items$"], ["actionBarComponent", ""], [1, "mcontainer", 3, "ngClass"], [1, "container-fluid"], [1, "powered-by", "navigator-slot-bottom"], ["translate", ""], [3, "simple"], [1, "mcontainer"]], template: function BootstrapComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c1);
        ɵngcc0.ɵɵtemplate(0, BootstrapComponent_c8y_login_0_Template, 1, 0, "c8y-login", 0);
        ɵngcc0.ɵɵpipe(1, "async");
        ɵngcc0.ɵɵtemplate(2, BootstrapComponent_div_2_Template, 33, 49, "div", 1);
        ɵngcc0.ɵɵpipe(3, "async");
        ɵngcc0.ɵɵpipe(4, "async");
        ɵngcc0.ɵɵtemplate(5, BootstrapComponent_div_5_Template, 5, 1, "div", 0);
        ɵngcc0.ɵɵpipe(6, "async");
        ɵngcc0.ɵɵpipe(7, "async");
        ɵngcc0.ɵɵelement(8, "c8y-cookie-banner");
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", !ɵngcc0.ɵɵpipeBind1(1, 3, ctx.ui.currentUser));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(3, 5, ctx.ui.currentUser) && !ɵngcc0.ɵɵpipeBind1(4, 7, ctx.isSetupNeeded$));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(6, 9, ctx.ui.currentUser) && ɵngcc0.ɵɵpipeBind1(7, 11, ctx.isSetupNeeded$));
    } }, encapsulation: 2 });
BootstrapComponent.ctorParameters = () => [
    { type: TabsService },
    { type: AppStateService },
    { type: NavigatorService },
    { type: ActionBarService },
    { type: HeaderService },
    { type: OptionsService },
    { type: PluginsResolveService },
    { type: StepperService },
    { type: Permissions }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BootstrapComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-bootstrap',
                template: "<c8y-login *ngIf=\"!(ui.currentUser | async)\"></c8y-login>\n\n<div\n  *ngIf=\"(ui.currentUser | async) && !(isSetupNeeded$ | async)\"\n  [class.head-open]=\"headerService.headerOpen\"\n>\n  <c8y-header-bar #header></c8y-header-bar>\n  <c8y-navigator-outlet\n    [nodes]=\"navigator.items$ | async\"\n    [open]=\"headerService.navigatorOpen$ | async\"\n  >\n    <header class=\"title navigator-slot-top\">\n      <div class=\"tenant-brand\"></div>\n      <c8y-app-icon\n        [name]=\"(ui.state$ | async).app.name\"\n        [contextPath]=\"(ui.state$ | async).app.contextPath\"\n      ></c8y-app-icon>\n      <span>{{ (ui.state$ | async).app | humanizeAppName | async }}</span>\n    </header>\n    <div class=\"powered-by navigator-slot-bottom\" *ngIf=\"showPoweredBy\">\n      <p translate>powered by Cumulocity</p>\n    </div>\n  </c8y-navigator-outlet>\n\n  <div class=\"alerts\">\n    <c8y-alert-outlet></c8y-alert-outlet>\n  </div>\n  <c8y-tabs-outlet\n    #tabsComponent\n    [tabs]=\"tabs.items$ | async\"\n    [navigatorOpen]=\"headerService.navigatorOpen$ | async\"\n    [orientation]=\"tabs.orientation$ | async\"\n  >\n  </c8y-tabs-outlet>\n  <c8y-action-bar\n    #actionBarComponent\n    [navigatorOpen]=\"headerService.navigatorOpen$ | async\"\n    [hasTabs]=\"tabsComponent.hasTabs\"\n    [isTabsHorizontal]=\"tabsComponent?.isHorizontal\"\n    [items$]=\"actionBar.items$\"\n  >\n  </c8y-action-bar>\n\n  <div\n    class=\"mcontainer\"\n    [ngClass]=\"{\n      open: headerService.navigatorOpen$ | async,\n      'no-apps-margin': noAppsMargin$ | async,\n      'horizontal-tabs': tabsComponent.isHorizontal,\n      'vertical-tabs': !tabsComponent.isHorizontal,\n      'has-tabs': tabsComponent.hasTabs,\n      'has-action-bar': !actionBarComponent?.hidden\n    }\"\n  >\n    <div class=\"container-fluid\">\n      <router-outlet></router-outlet>\n      <ng-content select=\"#c8y-legacy-view\"></ng-content>\n      <!-- legacy ng-view, will not be migrated atm -->\n    </div>\n  </div>\n</div>\n\n<div *ngIf=\"(ui.currentUser | async) && (isSetupNeeded$ | async)\">\n  <c8y-header-bar [simple]=\"true\"></c8y-header-bar>\n  <div class=\"mcontainer\">\n    <div class=\"container-fluid\">\n      <c8y-setup></c8y-setup>\n    </div>\n  </div>\n</div>\n\n<c8y-cookie-banner></c8y-cookie-banner>\n"
            }]
    }], function () { return [{ type: ɵngcc1.TabsService }, { type: ɵngcc2.AppStateService }, { type: ɵngcc3.NavigatorService }, { type: ɵngcc4.ActionBarService }, { type: ɵngcc5.HeaderService }, { type: ɵngcc6.OptionsService }, { type: ɵngcc7.PluginsResolveService }, { type: ɵngcc8.StepperService }, { type: ɵngcc9.Permissions }]; }, null); })();
