// @c8y/ngx-components/esm2015/core/bootstrap/bootstrap.module.js
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { ActionBarModule } from '../action-bar/action-bar.module';
import { AlertModule } from '../alert/alert.module';
import { CommonModule } from '../common/common.module';
import { HeaderModule } from '../header/header.module';
import { LoginModule } from '../login/login.module';
import { ModalModule } from '../modal/modal.module';
import { NavigatorModule } from '../navigator/navigator.module';
import { SetupModule } from '../setup/setup.module';
import { TabsModule } from '../tabs/tabs.module';
import { BootstrapComponent } from './bootstrap.component';
import { CookieBannerPreferencesModalComponent } from './cookie-banner/cookie-banner-preferences-modal/cookie-banner-preferences-modal.component';
import { CookieBannerComponent } from './cookie-banner/cookie-banner.component';
import { CookieBannerService } from './cookie-banner/cookie-banner.service';
/**
 * The module which allows to bootstrap a custom application.
 * Use it as `bootstrap: []` component in your main app module
 * or as `<c8y-bootstrap>` in your main app.
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '../login/login.component';
import * as ɵngcc3 from '../header/header-bar/header-bar.component';
import * as ɵngcc4 from '../navigator/navigator-outlet.component';
import * as ɵngcc5 from '../header/application/app-icon.component';
import * as ɵngcc6 from '../i18n/c8y-translate.directive';
import * as ɵngcc7 from '../alert/alert-outlet.component';
import * as ɵngcc8 from '../tabs/tabs-outlet.component';
import * as ɵngcc9 from '../action-bar/action-bar.component';
import * as ɵngcc10 from '@angular/router';
import * as ɵngcc11 from '../setup/setup.component';
import * as ɵngcc12 from '../common/humanize-app-name.pipe';
export class BootstrapModule {
}
BootstrapModule.ɵfac = function BootstrapModule_Factory(t) { return new (t || BootstrapModule)(); };
BootstrapModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: BootstrapModule });
BootstrapModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [CookieBannerService], imports: [[
            FormsModule,
            CommonModule,
            HeaderModule,
            LoginModule,
            RouterModule,
            TabsModule,
            ActionBarModule,
            AlertModule,
            NavigatorModule,
            TooltipModule,
            ModalModule,
            SetupModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BootstrapModule, [{
        type: NgModule,
        args: [{
                declarations: [BootstrapComponent, CookieBannerComponent, CookieBannerPreferencesModalComponent],
                imports: [
                    FormsModule,
                    CommonModule,
                    HeaderModule,
                    LoginModule,
                    RouterModule,
                    TabsModule,
                    ActionBarModule,
                    AlertModule,
                    NavigatorModule,
                    TooltipModule,
                    ModalModule,
                    SetupModule
                ],
                exports: [BootstrapComponent],
                entryComponents: [BootstrapComponent, CookieBannerPreferencesModalComponent],
                providers: [CookieBannerService]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(BootstrapModule, { declarations: function () { return [BootstrapComponent, CookieBannerComponent, CookieBannerPreferencesModalComponent]; }, imports: function () { return [FormsModule,
        CommonModule,
        HeaderModule,
        LoginModule,
        RouterModule,
        TabsModule,
        ActionBarModule,
        AlertModule,
        NavigatorModule,
        TooltipModule,
        ModalModule,
        SetupModule]; }, exports: function () { return [BootstrapComponent]; } }); })();
ɵngcc0.ɵɵsetComponentScope(BootstrapComponent, [ɵngcc1.NgIf, ɵngcc2.LoginComponent, ɵngcc3.HeaderBarComponent, ɵngcc4.NavigatorOutletComponent, ɵngcc5.AppIconComponent, ɵngcc6.C8yTranslateDirective, ɵngcc7.AlertOutletComponent, ɵngcc8.TabsOutletComponent, ɵngcc9.ActionBarComponent, ɵngcc1.NgClass, ɵngcc10.RouterOutlet, ɵngcc11.SetupComponent, CookieBannerComponent], [ɵngcc1.AsyncPipe, ɵngcc12.HumanizeAppNamePipe]);
