// @c8y/ngx-components/esm2015/core/action/action-outlet.component.js
import { Component, HostBinding, Input } from '@angular/core';
import { sortBy } from 'lodash-es';
/**
 * This component is used as the outlet to show the action list.
 * In a c8ycli app it is by default placed on the header component near app-switcher.
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from 'ngx-bootstrap/dropdown';
import * as ɵngcc3 from '../common/icon.directive';
import * as ɵngcc4 from '../common/outlet.directive';
import * as ɵngcc5 from '../i18n/c8y-translate.pipe';

function ActionOutletComponent_div_0_ul_4_ng_container_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function ActionOutletComponent_div_0_ul_4_ng_container_1_li_2_i_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 11);
} if (rf & 2) {
    const item_r3 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵproperty("c8yIcon", item_r3.icon);
} }
function ActionOutletComponent_div_0_ul_4_ng_container_1_li_2_Template(rf, ctx) { if (rf & 1) {
    const _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li");
    ɵngcc0.ɵɵelementStart(1, "button", 9);
    ɵngcc0.ɵɵlistener("click", function ActionOutletComponent_div_0_ul_4_ng_container_1_li_2_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r10); const item_r3 = ɵngcc0.ɵɵnextContext().$implicit; const ctx_r8 = ɵngcc0.ɵɵnextContext(3); return ctx_r8.invoke(item_r3); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵtemplate(3, ActionOutletComponent_div_0_ul_4_ng_container_1_li_2_i_3_Template, 1, 1, "i", 10);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r3 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 4, item_r3.label));
    ɵngcc0.ɵɵattribute("disabled", item_r3.disabled ? "disabled" : null);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", item_r3.icon);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1("", ɵngcc0.ɵɵpipeBind1(5, 6, item_r3.label), " ");
} }
function ActionOutletComponent_div_0_ul_4_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ActionOutletComponent_div_0_ul_4_ng_container_1_ng_container_1_Template, 1, 0, "ng-container", 7);
    ɵngcc0.ɵɵtemplate(2, ActionOutletComponent_div_0_ul_4_ng_container_1_li_2_Template, 6, 8, "li", 8);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r3 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yOutlet", item_r3.template);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r3.label);
} }
function ActionOutletComponent_div_0_ul_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "ul", 5);
    ɵngcc0.ɵɵtemplate(1, ActionOutletComponent_div_0_ul_4_ng_container_1_Template, 3, 2, "ng-container", 6);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r1.getByPriority());
} }
function ActionOutletComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 1);
    ɵngcc0.ɵɵelementStart(1, "button", 2);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(4, ActionOutletComponent_div_0_ul_4_Template, 2, 1, "ul", 4);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 2, "Actions"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", "plus-circle");
} }
export class ActionOutletComponent {
    constructor() {
        /**
         * Identifies if in a mobile view the toolbar is expanded or not.
         */
        this.isPageToolbarExpanded = false;
        /**
         * The current items to display in the action dropdown.
         */
        this.items = [];
    }
    /**
     * @ignore
     */
    get hasItems() {
        return !!this.items.length;
    }
    /**
     * Returns the items ordered by priority.
     */
    getByPriority() {
        return sortBy(this.items, this.byPriority);
    }
    /**
     * Invokes action after an action is clicked.
     */
    invoke(action) {
        if (action.action) {
            action.action.call();
        }
    }
    byPriority(item) {
        return -item.priority;
    }
}
ActionOutletComponent.ɵfac = function ActionOutletComponent_Factory(t) { return new (t || ActionOutletComponent)(); };
ActionOutletComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ActionOutletComponent, selectors: [["c8y-action-outlet"]], hostVars: 2, hostBindings: function ActionOutletComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("main-header-item", ctx.hasItems);
    } }, inputs: { items: "items" }, decls: 1, vars: 1, consts: [["class", "dropdown", "dropdown", "", 4, "ngIf"], ["dropdown", "", 1, "dropdown"], ["type", "button", "dropdownToggle", "", "aria-haspopup", "true", "aria-expanded", "true", 1, "main-header-button", "dropdown-toggle", "c8y-dropdown", 3, "title"], [1, "icon-2x", 3, "c8yIcon"], ["class", "dropdown-menu cog-dropdown", 4, "dropdownMenu"], [1, "dropdown-menu", "cog-dropdown"], [4, "ngFor", "ngForOf"], [4, "c8yOutlet"], [4, "ngIf"], [3, "title", "click"], [3, "c8yIcon", 4, "ngIf"], [3, "c8yIcon"]], template: function ActionOutletComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ActionOutletComponent_div_0_Template, 5, 4, "div", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.items && ctx.items.length > 0);
    } }, directives: [ɵngcc1.NgIf, ɵngcc2.BsDropdownDirective, ɵngcc2.BsDropdownToggleDirective, ɵngcc3.IconDirective, ɵngcc2.BsDropdownMenuDirective, ɵngcc1.NgForOf, ɵngcc4.OutletDirective], pipes: [ɵngcc5.C8yTranslatePipe], encapsulation: 2 });
ActionOutletComponent.propDecorators = {
    hasItems: [{ type: HostBinding, args: ['class.main-header-item',] }],
    items: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ActionOutletComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-action-outlet',
                template: "<div class=\"dropdown\" dropdown *ngIf=\"items && items.length > 0\">\n  <button\n    type=\"button\"\n    class=\"main-header-button dropdown-toggle c8y-dropdown\"\n    dropdownToggle\n    aria-haspopup=\"true\"\n    aria-expanded=\"true\"\n    title=\"{{'Actions' | translate}}\">\n    <i [c8yIcon]=\"'plus-circle'\" class=\"icon-2x\"></i>\n  </button>\n\n  <ul class=\"dropdown-menu cog-dropdown\" *dropdownMenu>\n    <ng-container *ngFor=\"let item of getByPriority()\">\n      <ng-container *c8yOutlet=\"item.template\"></ng-container>\n      <li *ngIf=\"item.label\">\n        <button title=\"{{ item.label | translate }}\" [attr.disabled]=\"item.disabled ? 'disabled' : null\" (click)=\"invoke(item)\">\n          <i *ngIf=\"item.icon\" [c8yIcon]=\"item.icon\"></i>{{item.label | translate}}\n        </button>\n      </li>\n    </ng-container>\n  </ul>\n</div>\n"
            }]
    }], function () { return []; }, { items: [{
            type: Input
        }], hasItems: [{
            type: HostBinding,
            args: ['class.main-header-item']
        }] }); })();
