// @c8y/ngx-components/esm2015/core/user/user-edit.component.js
import { __awaiter } from "tslib";
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { UserService, TenantLoginOptionsService, TenantService } from '@c8y/client';
import { clone } from 'lodash-es';
import { AppStateService } from '../common/ui-state.service';
import { TranslateService } from '../i18n/translate.service';
import { BsModalService } from 'ngx-bootstrap/modal';
import { UserTotpSetupComponent } from './user-totp-setup.component';
import { AlertService } from '../alert/alert.service';
import { UserPreferencesService } from '../common/user-preferences/user-preferences.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../common/ui-state.service';
import * as ɵngcc2 from '../i18n/translate.service';
import * as ɵngcc3 from 'ngx-bootstrap/modal';
import * as ɵngcc4 from '../alert/alert.service';
import * as ɵngcc5 from '@c8y/client';
import * as ɵngcc6 from '../common/user-preferences/user-preferences.service';
import * as ɵngcc7 from '@angular/forms';
import * as ɵngcc8 from '@angular/common';
import * as ɵngcc9 from '../forms/form-group.component';
import * as ɵngcc10 from '../i18n/c8y-translate.directive';
import * as ɵngcc11 from '../forms/required-input-placeholder.directive';
import * as ɵngcc12 from '../forms/default-validation.directive';
import * as ɵngcc13 from '../forms/phone-validation.directive';
import * as ɵngcc14 from '../authentication/new-password.component';
import * as ɵngcc15 from '../i18n/c8y-translate.pipe';

function UserEditComponent_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 30);
    ɵngcc0.ɵɵtext(1, " Some of the user settings are not editable here because they are managed via your authorization server. ");
    ɵngcc0.ɵɵelementEnd();
} }
function UserEditComponent_option_41_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "option", 31);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const lang_r7 = ctx.$implicit;
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("value", lang_r7);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r3.translate.getNativeLanguage(lang_r7), " ");
} }
function UserEditComponent_c8y_form_group_43_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-form-group", 32);
    ɵngcc0.ɵɵelementStart(1, "strong", 33);
    ɵngcc0.ɵɵtext(2, "Product experience");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "label", 34);
    ɵngcc0.ɵɵelementStart(4, "input", 35);
    ɵngcc0.ɵɵlistener("ngModelChange", function UserEditComponent_c8y_form_group_43_Template_input_ngModelChange_4_listener($event) { ɵngcc0.ɵɵrestoreView(_r9); const ctx_r8 = ɵngcc0.ɵɵnextContext(); return ctx_r8.isUsageTrackingEnabled = $event; });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(5, "span");
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵpipe(7, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r4.isUsageTrackingEnabled);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(7, 2, "Enable anonymous tracking to enhance the product experience"), " ");
} }
function UserEditComponent_div_44_button_5_Template(rf, ctx) { if (rf & 1) {
    const _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 28);
    ɵngcc0.ɵɵlistener("click", function UserEditComponent_div_44_button_5_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r12); const ctx_r11 = ɵngcc0.ɵɵnextContext(2); return ctx_r11.setupTotp(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, "Set up two-factor authentication"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 4, "Set up two-factor authentication"), " ");
} }
function UserEditComponent_div_44_Template(rf, ctx) { if (rf & 1) {
    const _r14 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 36);
    ɵngcc0.ɵɵelementStart(1, "label", 37);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "c8y-new-password", 38);
    ɵngcc0.ɵɵlistener("password", function UserEditComponent_div_44_Template_c8y_new_password_password_4_listener($event) { ɵngcc0.ɵɵrestoreView(_r14); const ctx_r13 = ɵngcc0.ɵɵnextContext(); return ctx_r13.onNewPasswordChanged($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(5, UserEditComponent_div_44_button_5_Template, 4, 6, "button", 39);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(3, 2, "Login options"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r5.userCanSetupTotp && !ctx_r5.userHasActiveTotp && ctx_r5.isTfaEnabled);
} }
function UserEditComponent_c8y_form_group_45_Template(rf, ctx) { if (rf & 1) {
    const _r16 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(1, "label", 33);
    ɵngcc0.ɵɵtext(2, "Newsletter");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "label", 40);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementStart(5, "input", 41);
    ɵngcc0.ɵɵlistener("ngModelChange", function UserEditComponent_c8y_form_group_45_Template_input_ngModelChange_5_listener($event) { ɵngcc0.ɵɵrestoreView(_r16); const ctx_r15 = ɵngcc0.ɵɵnextContext(); return ctx_r15.user.newsletter = $event; });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(6, "span");
    ɵngcc0.ɵɵelementStart(7, "span");
    ɵngcc0.ɵɵtext(8);
    ɵngcc0.ɵɵpipe(9, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(4, 4, "Send me information about outages, maintenance or updates."));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r6.user.newsletter)("disabled", ctx_r6.userIsExternal);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(9, 6, "Send me information about outages, maintenance or updates."), " ");
} }
export class UserEditComponent {
    constructor(state, translate, bsModalService, alert, userService, tenantLoginOptionsService, tenantService, userPreferencesService) {
        this.state = state;
        this.translate = translate;
        this.bsModalService = bsModalService;
        this.alert = alert;
        this.userService = userService;
        this.tenantLoginOptionsService = tenantLoginOptionsService;
        this.tenantService = tenantService;
        this.userPreferencesService = userPreferencesService;
        this.loading = false;
        this.showProductUsageSetting = false;
        this.isUsageTrackingEnabled = true;
        this.onUser = new EventEmitter();
        this.onLanguage = new EventEmitter();
        this.onProductExperience = new EventEmitter();
        this.onCancel = new EventEmitter();
        this.userHasActiveTotp = false;
        this.userCanSetupTotp = false;
        this.isPhoneRequired = false;
    }
    set user(u) {
        if (u) {
            this._user = clone(u);
            this.userIsExternal = u.customProperties.userOrigin === 'OAUTH2';
            this.isPhoneRequired = this.isPhoneRequired && u.twoFactorAuthenticationEnabled;
        }
    }
    get user() {
        return this._user;
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            const currentTenant = (yield this.tenantService.current()).data;
            const { enabledOnSystemLevel, enabledOnTenantLevel } = yield this.tenantService.getTfaSettings(currentTenant);
            this.isTfaEnabled = enabledOnSystemLevel || enabledOnTenantLevel;
            yield this.initializeTotpSettings();
            if (this.user.twoFactorAuthenticationEnabled && !this.userCanSetupTotp) {
                this.isPhoneRequired = true;
            }
        });
    }
    get langs() {
        return this.state.state.langs;
    }
    setupTotp() {
        this.bsModalService.show(UserTotpSetupComponent, { class: 'modal-sm' });
        this.cancel(); // to close the user edit modal and prevent console errors on logout
    }
    cancel() {
        this.onCancel.emit();
    }
    save() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.loading) {
                return;
            }
            if (this.showProductUsageSetting) {
                this.onProductExperience.emit(this.isUsageTrackingEnabled);
                this.userPreferencesService.set('gainsightEnabled', this.isUsageTrackingEnabled);
            }
            this.onUser.emit(this._user);
        });
    }
    onNewPasswordChanged(newPassword) {
        this._user.password = newPassword.password;
    }
    initializeTotpSettings() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                this.userCanSetupTotp = yield this.canUserSetupTotp();
                if (this.userCanSetupTotp) {
                    const { data: totpActivity } = yield this.userService.getActivityTotp();
                    this.userHasActiveTotp = totpActivity.isActive;
                }
            }
            catch (ex) {
                this.alert.removeLastDanger();
            }
        });
    }
    canUserSetupTotp() {
        return __awaiter(this, void 0, void 0, function* () {
            // we don't check for tenant options here due to permission restrictions on that end-point
            const loginOptions = (yield this.tenantLoginOptionsService.listForCurrentTenant()).data;
            return loginOptions.some(({ tfaStrategy = '' }) => tfaStrategy.toLowerCase() === 'totp');
        });
    }
}
UserEditComponent.ɵfac = function UserEditComponent_Factory(t) { return new (t || UserEditComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AppStateService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.TranslateService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.BsModalService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.UserService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.TenantLoginOptionsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.TenantService), ɵngcc0.ɵɵdirectiveInject(ɵngcc6.UserPreferencesService)); };
UserEditComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: UserEditComponent, selectors: [["c8y-user-edit"]], inputs: { loading: "loading", showProductUsageSetting: "showProductUsageSetting", isUsageTrackingEnabled: "isUsageTrackingEnabled", user: "user", lang: "lang" }, outputs: { onUser: "onUser", onLanguage: "onLanguage", onProductExperience: "onProductExperience", onCancel: "onCancel" }, decls: 56, vars: 49, consts: [[3, "ngSubmit"], ["userForm", "ngForm"], [1, "d-block", "p-24", "p-b-0"], ["class", "alert alert-warning", "role", "alert", "translate", "", 4, "ngIf"], ["translate", "", "for", "userName"], ["id", "userName", "name", "userName", "autocomplete", "off", "required", "", "maxlength", "254", "c8yDefaultValidation", "user", 1, "form-control", 3, "ngModel", "placeholder", "disabled", "ngModelChange"], ["translate", "", "for", "displayName"], ["id", "displayName", "name", "displayName", "autocomplete", "off", "maxlength", "254", "c8yDefaultValidation", "loginAlias", 1, "form-control", 3, "ngModel", "placeholder", "disabled", "ngModelChange"], [3, "hasWarning"], ["translate", "", "for", "userEmail"], ["id", "userEmail", "type", "email", "name", "email", "autocomplete", "off", "email", "", 1, "form-control", 3, "maxlength", "placeholder", "ngModel", "required", "disabled", "ngModelChange"], [1, "row"], [1, "col-sm-6"], ["translate", "", "for", "userFirstName"], ["id", "userFirstName", "autocomplete", "off", "maxlength", "50", "name", "firstName", 1, "form-control", 3, "ngModel", "disabled", "ngModelChange"], ["translate", "", "for", "userLastName"], ["id", "userLastName", "autocomplete", "off", "maxlength", "50", "name", "lastName", 1, "form-control", 3, "ngModel", "disabled", "ngModelChange"], ["translate", "", "for", "userTelephone"], ["id", "userTelephone", "autocomplete", "off", "name", "phone", "maxlength", "254", "c8yPhoneValidation", "", "c8yDefaultValidation", "phoneNumber", 1, "form-control", 3, "ngModel", "placeholder", "required", "disabled", "ngModelChange"], ["translate", "", "for", "userLang"], [1, "c8y-select-wrapper"], ["id", "userLang", "name", "lang", 1, "form-control", 3, "ngModel", "ngModelChange", "change"], ["selectLang", ""], [3, "value", 4, "ngFor", "ngForOf"], ["class", "p-t-8 p-b-8 separator-top", 4, "ngIf"], ["class", "form-group", 4, "ngIf"], [4, "ngIf"], [1, "modal-footer", "separator-top", "bg-white", "sticky-bottom"], ["type", "button", 1, "btn", "btn-default", 3, "title", "click"], ["type", "submit", 1, "btn", "btn-primary", 3, "title", "disabled"], ["role", "alert", "translate", "", 1, "alert", "alert-warning"], [3, "value"], [1, "p-t-8", "p-b-8", "separator-top"], ["translate", ""], ["for", "productUsageTracking", 1, "c8y-switch"], ["id", "productUsageTracking", "name", "productUsageTracking", "type", "checkbox", 3, "ngModel", "ngModelChange"], [1, "form-group"], [1, "control-label"], [3, "password"], ["class", "btn btn-default", "type", "button", 3, "title", "click", 4, "ngIf"], [1, "c8y-checkbox", 3, "title"], ["type", "checkbox", "name", "newsletter", 3, "ngModel", "disabled", "ngModelChange"]], template: function UserEditComponent_Template(rf, ctx) { if (rf & 1) {
        const _r17 = ɵngcc0.ɵɵgetCurrentView();
        ɵngcc0.ɵɵelementStart(0, "form", 0, 1);
        ɵngcc0.ɵɵlistener("ngSubmit", function UserEditComponent_Template_form_ngSubmit_0_listener() { ɵngcc0.ɵɵrestoreView(_r17); const _r0 = ɵngcc0.ɵɵreference(1); return _r0.form.valid && ctx.save(); });
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵtemplate(3, UserEditComponent_div_3_Template, 2, 0, "div", 3);
        ɵngcc0.ɵɵelementStart(4, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(5, "label", 4);
        ɵngcc0.ɵɵtext(6, "Username (for example, email)");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(7, "input", 5);
        ɵngcc0.ɵɵlistener("ngModelChange", function UserEditComponent_Template_input_ngModelChange_7_listener($event) { return ctx.user.userName = $event; });
        ɵngcc0.ɵɵpipe(8, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(10, "label", 6);
        ɵngcc0.ɵɵtext(11, "Login alias");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(12, "input", 7);
        ɵngcc0.ɵɵlistener("ngModelChange", function UserEditComponent_Template_input_ngModelChange_12_listener($event) { return ctx.user.displayName = $event; });
        ɵngcc0.ɵɵpipe(13, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(14, "c8y-form-group", 8);
        ɵngcc0.ɵɵelementStart(15, "label", 9);
        ɵngcc0.ɵɵtext(16, "Email");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(17, "input", 10);
        ɵngcc0.ɵɵlistener("ngModelChange", function UserEditComponent_Template_input_ngModelChange_17_listener($event) { return ctx.user.email = $event; });
        ɵngcc0.ɵɵpipe(18, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(19, "div", 11);
        ɵngcc0.ɵɵelementStart(20, "div", 12);
        ɵngcc0.ɵɵelementStart(21, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(22, "label", 13);
        ɵngcc0.ɵɵtext(23, "First name");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(24, "input", 14);
        ɵngcc0.ɵɵlistener("ngModelChange", function UserEditComponent_Template_input_ngModelChange_24_listener($event) { return ctx.user.firstName = $event; });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(25, "div", 12);
        ɵngcc0.ɵɵelementStart(26, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(27, "label", 15);
        ɵngcc0.ɵɵtext(28, "Last name");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(29, "input", 16);
        ɵngcc0.ɵɵlistener("ngModelChange", function UserEditComponent_Template_input_ngModelChange_29_listener($event) { return ctx.user.lastName = $event; });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(30, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(31, "label", 17);
        ɵngcc0.ɵɵtext(32, "Telephone");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(33, "input", 18);
        ɵngcc0.ɵɵlistener("ngModelChange", function UserEditComponent_Template_input_ngModelChange_33_listener($event) { return ctx.user.phone = $event; });
        ɵngcc0.ɵɵpipe(34, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(35, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(36, "label", 19);
        ɵngcc0.ɵɵtext(37, "Language");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(38, "div", 20);
        ɵngcc0.ɵɵelementStart(39, "select", 21, 22);
        ɵngcc0.ɵɵlistener("ngModelChange", function UserEditComponent_Template_select_ngModelChange_39_listener($event) { return ctx.lang = $event; })("change", function UserEditComponent_Template_select_change_39_listener() { ɵngcc0.ɵɵrestoreView(_r17); const _r2 = ɵngcc0.ɵɵreference(40); return ctx.onLanguage.emit(_r2.value); });
        ɵngcc0.ɵɵtemplate(41, UserEditComponent_option_41_Template, 2, 2, "option", 23);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(42, "span");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(43, UserEditComponent_c8y_form_group_43_Template, 8, 4, "c8y-form-group", 24);
        ɵngcc0.ɵɵtemplate(44, UserEditComponent_div_44_Template, 6, 4, "div", 25);
        ɵngcc0.ɵɵtemplate(45, UserEditComponent_c8y_form_group_45_Template, 10, 8, "c8y-form-group", 26);
        ɵngcc0.ɵɵpipe(46, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(47, "div", 27);
        ɵngcc0.ɵɵelementStart(48, "button", 28);
        ɵngcc0.ɵɵlistener("click", function UserEditComponent_Template_button_click_48_listener() { return ctx.cancel(); });
        ɵngcc0.ɵɵpipe(49, "translate");
        ɵngcc0.ɵɵtext(50);
        ɵngcc0.ɵɵpipe(51, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(52, "button", 29);
        ɵngcc0.ɵɵpipe(53, "translate");
        ɵngcc0.ɵɵtext(54);
        ɵngcc0.ɵɵpipe(55, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r0 = ɵngcc0.ɵɵreference(1);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.userIsExternal);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(8, 31, "e.g. joe.doe@example.com`LOCALIZE`"));
        ɵngcc0.ɵɵproperty("ngModel", ctx.user.userName)("disabled", ctx.user.id);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(13, 33, "e.g. joe.doe`LOCALIZE`"));
        ɵngcc0.ɵɵproperty("ngModel", ctx.user.displayName)("disabled", ctx.userIsExternal);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("hasWarning", !ctx.user.email);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(18, 35, "e.g. joe.doe@example.com`LOCALIZE`"));
        ɵngcc0.ɵɵproperty("maxlength", 254)("ngModel", ctx.user.email)("required", true)("disabled", ctx.userIsExternal);
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵproperty("ngModel", ctx.user.firstName)("disabled", ctx.userIsExternal);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("ngModel", ctx.user.lastName)("disabled", ctx.userIsExternal);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(34, 37, "e.g. +49 9 876 543 210`LOCALIZE`"));
        ɵngcc0.ɵɵproperty("ngModel", ctx.user.phone)("required", ctx.isPhoneRequired)("disabled", ctx.userIsExternal);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵproperty("ngModel", ctx.lang);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.langs);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showProductUsageSetting);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.userIsExternal);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !!ɵngcc0.ɵɵpipeBind1(46, 39, ctx.state.state$).newsletter);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(49, 41, "Cancel"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(51, 43, "Cancel"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(53, 45, "Save"));
        ɵngcc0.ɵɵproperty("disabled", !_r0.form.valid || _r0.form.pristine || ctx.loading);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(55, 47, "Save"), " ");
    } }, directives: [ɵngcc7.ɵNgNoValidate, ɵngcc7.NgControlStatusGroup, ɵngcc7.NgForm, ɵngcc8.NgIf, ɵngcc9.FormGroupComponent, ɵngcc10.C8yTranslateDirective, ɵngcc11.RequiredInputPlaceholderDirective, ɵngcc7.DefaultValueAccessor, ɵngcc7.RequiredValidator, ɵngcc7.MaxLengthValidator, ɵngcc12.DefaultValidationDirective, ɵngcc7.NgControlStatus, ɵngcc7.NgModel, ɵngcc7.EmailValidator, ɵngcc13.PhoneValidationDirective, ɵngcc7.SelectControlValueAccessor, ɵngcc8.NgForOf, ɵngcc7.NgSelectOption, ɵngcc7.ɵNgSelectMultipleOption, ɵngcc7.CheckboxControlValueAccessor, ɵngcc14.NewPasswordComponent], pipes: [ɵngcc15.C8yTranslatePipe, ɵngcc8.AsyncPipe], encapsulation: 2 });
UserEditComponent.ctorParameters = () => [
    { type: AppStateService },
    { type: TranslateService },
    { type: BsModalService },
    { type: AlertService },
    { type: UserService },
    { type: TenantLoginOptionsService },
    { type: TenantService },
    { type: UserPreferencesService }
];
UserEditComponent.propDecorators = {
    lang: [{ type: Input }],
    loading: [{ type: Input }],
    user: [{ type: Input }],
    showProductUsageSetting: [{ type: Input }],
    isUsageTrackingEnabled: [{ type: Input }],
    onUser: [{ type: Output }],
    onLanguage: [{ type: Output }],
    onProductExperience: [{ type: Output }],
    onCancel: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(UserEditComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-user-edit',
                template: "<form #userForm=\"ngForm\" (ngSubmit)=\"userForm.form.valid && save()\">\n  <div class=\"d-block p-24 p-b-0\">\n    <div class=\"alert alert-warning\" role=\"alert\" *ngIf=\"userIsExternal\" translate>\n      Some of the user settings are not editable here because they are managed via your\n      authorization server.\n    </div>\n    <c8y-form-group>\n      <label translate for=\"userName\">Username (for example, email)</label>\n      <input\n        id=\"userName\"\n        class=\"form-control\"\n        [(ngModel)]=\"user.userName\"\n        name=\"userName\"\n        autocomplete=\"off\"\n        required\n        maxlength=\"254\"\n        placeholder=\"{{ 'e.g. joe.doe@example.com`LOCALIZE`' | translate }}\"\n        [disabled]=\"user.id\"\n        c8yDefaultValidation=\"user\"\n      />\n    </c8y-form-group>\n\n    <c8y-form-group>\n      <label translate for=\"displayName\">Login alias</label>\n      <input\n        id=\"displayName\"\n        class=\"form-control\"\n        [(ngModel)]=\"user.displayName\"\n        name=\"displayName\"\n        autocomplete=\"off\"\n        maxlength=\"254\"\n        placeholder=\"{{ 'e.g. joe.doe`LOCALIZE`' | translate }}\"\n        [disabled]=\"userIsExternal\"\n        c8yDefaultValidation=\"loginAlias\"\n      />\n    </c8y-form-group>\n\n    <c8y-form-group [hasWarning]=\"!user.email\">\n      <label translate for=\"userEmail\">Email</label>\n      <input\n        id=\"userEmail\"\n        class=\"form-control\"\n        type=\"email\"\n        name=\"email\"\n        [maxlength]=\"254\"\n        autocomplete=\"off\"\n        placeholder=\"{{ 'e.g. joe.doe@example.com`LOCALIZE`' | translate }}\"\n        [(ngModel)]=\"user.email\"\n        email\n        [required]=\"true\"\n        [disabled]=\"userIsExternal\"\n      />\n    </c8y-form-group>\n\n    <div class=\"row\">\n      <div class=\"col-sm-6\">\n        <c8y-form-group>\n          <label translate for=\"userFirstName\">First name</label>\n          <input\n            id=\"userFirstName\"\n            class=\"form-control\"\n            autocomplete=\"off\"\n            maxlength=\"50\"\n            name=\"firstName\"\n            [(ngModel)]=\"user.firstName\"\n            [disabled]=\"userIsExternal\"\n          />\n        </c8y-form-group>\n      </div>\n      <div class=\"col-sm-6\">\n        <c8y-form-group>\n          <label translate for=\"userLastName\">Last name</label>\n          <input\n            id=\"userLastName\"\n            class=\"form-control\"\n            autocomplete=\"off\"\n            maxlength=\"50\"\n            name=\"lastName\"\n            [(ngModel)]=\"user.lastName\"\n            [disabled]=\"userIsExternal\"\n          />\n        </c8y-form-group>\n      </div>\n    </div>\n\n    <c8y-form-group>\n      <label translate for=\"userTelephone\">Telephone</label>\n      <input\n        id=\"userTelephone\"\n        class=\"form-control\"\n        autocomplete=\"off\"\n        name=\"phone\"\n        maxlength=\"254\"\n        [(ngModel)]=\"user.phone\"\n        placeholder=\"{{ 'e.g. +49 9 876 543 210`LOCALIZE`' | translate }}\"\n        c8yPhoneValidation\n        c8yDefaultValidation=\"phoneNumber\"\n        [required]=\"isPhoneRequired\"\n        [disabled]=\"userIsExternal\"\n      />\n    </c8y-form-group>\n\n    <c8y-form-group>\n      <label translate for=\"userLang\">Language</label>\n      <div class=\"c8y-select-wrapper\">\n        <select\n          id=\"userLang\"\n          class=\"form-control\"\n          #selectLang\n          name=\"lang\"\n          [(ngModel)]=\"lang\"\n          (change)=\"onLanguage.emit(selectLang.value)\"\n        >\n          <option *ngFor=\"let lang of langs\" [value]=\"lang\">\n            {{ translate.getNativeLanguage(lang) }}\n          </option>\n        </select>\n        <span></span>\n      </div>\n    </c8y-form-group>\n\n    <c8y-form-group class=\"p-t-8 p-b-8 separator-top\" *ngIf=\"showProductUsageSetting\">\n      <strong translate>Product experience</strong>\n      <label class=\"c8y-switch\" for=\"productUsageTracking\">\n        <input\n          id=\"productUsageTracking\"\n          name=\"productUsageTracking\"\n          type=\"checkbox\"\n          [(ngModel)]=\"isUsageTrackingEnabled\"\n        />\n        <span></span>\n        {{ 'Enable anonymous tracking to enhance the product experience' | translate }}\n      </label>\n    </c8y-form-group>\n\n    <div class=\"form-group\" *ngIf=\"!userIsExternal\">\n      <label class=\"control-label\">{{ 'Login options' | translate }}</label>\n      <c8y-new-password (password)=\"onNewPasswordChanged($event)\"></c8y-new-password>\n      <button\n        title=\"{{ 'Set up two-factor authentication' | translate }}\"\n        class=\"btn btn-default\"\n        type=\"button\"\n        (click)=\"setupTotp()\"\n        *ngIf=\"userCanSetupTotp && !userHasActiveTotp && isTfaEnabled\"\n      >\n        {{ 'Set up two-factor authentication' | translate }}\n      </button>\n    </div>\n\n    <c8y-form-group *ngIf=\"!!(state.state$ | async).newsletter\">\n      <label translate>Newsletter</label>\n      <label\n        title=\"{{ 'Send me information about outages, maintenance or updates.' | translate }}\"\n        class=\"c8y-checkbox\"\n      >\n        <input\n          type=\"checkbox\"\n          name=\"newsletter\"\n          [(ngModel)]=\"user.newsletter\"\n          [disabled]=\"userIsExternal\"\n        />\n        <span></span>\n        <span>\n          {{ 'Send me information about outages, maintenance or updates.' | translate }}\n        </span>\n      </label>\n    </c8y-form-group>\n  </div>\n  <div class=\"modal-footer separator-top bg-white sticky-bottom\">\n    <button\n      title=\"{{ 'Cancel' | translate }}\"\n      class=\"btn btn-default\"\n      type=\"button\"\n      (click)=\"cancel()\"\n    >\n      {{ 'Cancel' | translate }}\n    </button>\n    <button\n      title=\"{{ 'Save' | translate }}\"\n      class=\"btn btn-primary\"\n      type=\"submit\"\n      [disabled]=\"!userForm.form.valid || userForm.form.pristine || loading\"\n    >\n      {{ 'Save' | translate }}\n    </button>\n  </div>\n</form>\n"
            }]
    }], function () { return [{ type: ɵngcc1.AppStateService }, { type: ɵngcc2.TranslateService }, { type: ɵngcc3.BsModalService }, { type: ɵngcc4.AlertService }, { type: ɵngcc5.UserService }, { type: ɵngcc5.TenantLoginOptionsService }, { type: ɵngcc5.TenantService }, { type: ɵngcc6.UserPreferencesService }]; }, { loading: [{
            type: Input
        }], showProductUsageSetting: [{
            type: Input
        }], isUsageTrackingEnabled: [{
            type: Input
        }], onUser: [{
            type: Output
        }], onLanguage: [{
            type: Output
        }], onProductExperience: [{
            type: Output
        }], onCancel: [{
            type: Output
        }], user: [{
            type: Input
        }], lang: [{
            type: Input
        }] }); })();
