// @c8y/ngx-components/esm2015/device-grid/columns/base.filtering-form-renderer.component.js
import { Component, HostListener } from '@angular/core';
import { cloneDeep } from 'lodash-es';
import { C8yJSONSchema, FilteringFormRendererContext } from '@c8y/ngx-components';
import { FormGroup } from '@angular/forms';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from '@ngx-formly/core';
export class BaseFilteringFormRendererComponent {
    constructor(context, jsonschema) {
        this.context = context;
        this.jsonschema = jsonschema;
    }
    onEnterKeyDown(event) {
        event.stopPropagation();
        this.applyFilter();
    }
    onEscapeKeyDown(event) {
        event.stopPropagation();
        this.resetFilter();
    }
    ngOnInit() {
        const column = this.context.property;
        if (column.filteringConfig.fields) {
            this.fields = cloneDeep(column.filteringConfig.fields);
        }
        else if (!column.filteringConfig.fields && column.filteringConfig.schema) {
            this.schema = cloneDeep(column.filteringConfig.schema);
            this.fields = [this.jsonschema.toFieldConfig(this.schema)];
        }
        else {
            throw new Error(`Neither fields nor schema configuration provided for Formly.`);
        }
        this.form = column.filteringConfig.formGroup || new FormGroup({});
        this.model = column.externalFilterQuery || column.filteringConfig.model || {};
        this.options = { formDefaults: { ngModelOptions: { debounce: 100 } } };
    }
    applyFilter() {
        this.context.applyFilter({
            externalFilterQuery: this.form.value
        });
    }
    resetFilter() {
        this.context.resetFilter();
    }
}
BaseFilteringFormRendererComponent.ɵfac = function BaseFilteringFormRendererComponent_Factory(t) { return new (t || BaseFilteringFormRendererComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.FilteringFormRendererContext), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.C8yJSONSchema)); };
BaseFilteringFormRendererComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: BaseFilteringFormRendererComponent, selectors: [["ng-component"]], hostBindings: function BaseFilteringFormRendererComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("keydown.enter", function BaseFilteringFormRendererComponent_keydown_enter_HostBindingHandler($event) { return ctx.onEnterKeyDown($event); })("keydown.escape", function BaseFilteringFormRendererComponent_keydown_escape_HostBindingHandler($event) { return ctx.onEscapeKeyDown($event); });
    } }, decls: 8, vars: 10, consts: [[3, "form", "fields", "model"], [1, "data-grid__dropdown__footer", "d-flex", "separator-top"], ["translate", "", 1, "btn", "btn-default", "btn-sm", "m-r-8", "flex-grow", 3, "title", "click"], ["translate", "", 1, "btn", "btn-primary", "btn-sm", "flex-grow", 3, "disabled", "title", "click"]], template: function BaseFilteringFormRendererComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelement(0, "formly-form", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "button", 2);
        ɵngcc0.ɵɵlistener("click", function BaseFilteringFormRendererComponent_Template_button_click_2_listener() { return ctx.resetFilter(); });
        ɵngcc0.ɵɵpipe(3, "translate");
        ɵngcc0.ɵɵtext(4, " Reset ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "button", 3);
        ɵngcc0.ɵɵlistener("click", function BaseFilteringFormRendererComponent_Template_button_click_5_listener() { return ctx.applyFilter(); });
        ɵngcc0.ɵɵpipe(6, "translate");
        ɵngcc0.ɵɵtext(7, " Apply ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("form", ctx.form)("fields", ctx.fields)("model", ctx.model);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(3, 6, "Reset"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(6, 8, "Apply"));
        ɵngcc0.ɵɵproperty("disabled", ctx.form.pristine && ctx.form.untouched || ctx.form.invalid);
    } }, directives: [ɵngcc2.FormlyForm, ɵngcc1.C8yTranslateDirective], pipes: [ɵngcc1.C8yTranslatePipe], encapsulation: 2 });
BaseFilteringFormRendererComponent.ctorParameters = () => [
    { type: FilteringFormRendererContext },
    { type: C8yJSONSchema }
];
BaseFilteringFormRendererComponent.propDecorators = {
    onEnterKeyDown: [{ type: HostListener, args: ['keydown.enter', ['$event'],] }],
    onEscapeKeyDown: [{ type: HostListener, args: ['keydown.escape', ['$event'],] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BaseFilteringFormRendererComponent, [{
        type: Component,
        args: [{
                template: "<formly-form [form]=\"form\" [fields]=\"fields\" [model]=\"model\"></formly-form>\n<div class=\"data-grid__dropdown__footer d-flex separator-top\">\n  <button\n    class=\"btn btn-default btn-sm m-r-8 flex-grow\"\n    (click)=\"resetFilter()\"\n    title=\"{{ 'Reset' | translate }}\"\n    translate\n  >\n    Reset\n  </button>\n\n  <button\n    class=\"btn btn-primary btn-sm flex-grow\"\n    (click)=\"applyFilter()\"\n    [disabled]=\"(form.pristine && form.untouched) || form.invalid\"\n    title=\"{{ 'Apply' | translate }}\"\n    translate\n  >\n    Apply\n  </button>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.FilteringFormRendererContext }, { type: ɵngcc1.C8yJSONSchema }]; }, { onEnterKeyDown: [{
            type: HostListener,
            args: ['keydown.enter', ['$event']]
        }], onEscapeKeyDown: [{
            type: HostListener,
            args: ['keydown.escape', ['$event']]
        }] }); })();
