// @c8y/ngx-components/esm2015/core/time-picker/time-picker.module.js
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { CommonModule as C8yCommonModule } from '../common/common.module';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { FormsModule as C8yFormsModule } from '../forms/index';
import { BsDatepickerModule } from 'ngx-bootstrap/datepicker';
import { TimePickerComponent } from './time-picker.component';
/**
 * The angular module definition for a time picker - it provides FormControl that handles time.
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/datepicker';
export class TimePickerModule {
}
TimePickerModule.ɵfac = function TimePickerModule_Factory(t) { return new (t || TimePickerModule)(); };
TimePickerModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: TimePickerModule });
TimePickerModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            C8yFormsModule,
            CommonModule,
            FormsModule,
            ReactiveFormsModule,
            BsDatepickerModule.forRoot(),
            C8yCommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TimePickerModule, [{
        type: NgModule,
        args: [{
                imports: [
                    C8yFormsModule,
                    CommonModule,
                    FormsModule,
                    ReactiveFormsModule,
                    BsDatepickerModule.forRoot(),
                    C8yCommonModule
                ],
                exports: [TimePickerComponent],
                declarations: [TimePickerComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(TimePickerModule, { declarations: function () { return [TimePickerComponent]; }, imports: function () { return [C8yFormsModule,
        CommonModule,
        FormsModule,
        ReactiveFormsModule, ɵngcc1.BsDatepickerModule, C8yCommonModule]; }, exports: function () { return [TimePickerComponent]; } }); })();
