// @c8y/ngx-components/esm2015/core/wizard/wizard.component.js
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { BsModalRef } from 'ngx-bootstrap/modal';
import { filter } from 'rxjs/operators';
import { WizardService } from './wizard.service';
/**
 * A generic component that is intended to serve as the building block of more wizards.
 *
 * **Example**
 * ```html
 * <button
 *    title="Add application"
 *    class="btn btn-primary"
 *    (click)="addApplication()"
 *  >
 *    Add application
 *  </button>
 * ```
 *
 * ```ts
 * import { Component } from '@angular/core';
 * import { WizardConfig, WizardService, Wizard } from '@c8y/ngx-components';
 * import { ModalOptions } from 'ngx-bootstrap/modal';
 *
 * @Component({
 *  selector: 'c8y-add-application-wizard',
 *  templateUrl: './templatePath'
 * })
 * export class AddApplicationWizardComponent {
 *   constructor(private wizardService: WizardService) {}
 *
 *   addApplication() {
 *     const wizardConfig: WizardConfig = {
 *       headerText: 'Add Application',
 *       headerIcon: 'c8y-icon-modules',
 *       bodyHeaderText: 'Select methods',
 *       bodyHeaderIcon: 'c8y-icon-modules'
 *     };
 *
 *     const initialState: Wizard = {
 *       wizardConfig,
 *       id: 'uploadApplication'
 *     };
 *
 *     const modalOptions: ModalOptions = { initialState };
 *
 *     this.wizardService.show(modalOptions);
 *   }
 * }
 * ```
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './wizard.service';
import * as ɵngcc2 from 'ngx-bootstrap/modal';

const _c0 = ["headerRef"];
const _c1 = ["bodyRef"];
const _c2 = ["footerRef"];
function WizardComponent_1_ng_template_0_Template(rf, ctx) { }
function WizardComponent_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, WizardComponent_1_ng_template_0_Template, 0, 0, "ng-template");
} }
function WizardComponent_ng_template_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 6);
    ɵngcc0.ɵɵelement(1, "h1", 7);
    ɵngcc0.ɵɵelementStart(2, "h4");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yIcon", ctx_r2.wizardConfig.headerIcon);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(4, 2, ctx_r2.wizardConfig.headerText));
} }
function WizardComponent_5_ng_template_0_Template(rf, ctx) { }
function WizardComponent_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, WizardComponent_5_ng_template_0_Template, 0, 0, "ng-template");
} }
function WizardComponent_ng_template_7_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 9);
    ɵngcc0.ɵɵelementStart(1, "div", 10);
    ɵngcc0.ɵɵelement(2, "i", 7);
    ɵngcc0.ɵɵelementStart(3, "span");
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r11 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", ctx_r11.wizardConfig.bodyHeaderIcon);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(5, 2, ctx_r11.wizardConfig.bodyHeaderText));
} }
function WizardComponent_ng_template_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, WizardComponent_ng_template_7_div_0_Template, 6, 4, "div", 8);
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngIf", ctx_r5.wizardConfig.bodyHeaderText);
} }
function WizardComponent_10_ng_template_0_Template(rf, ctx) { }
function WizardComponent_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, WizardComponent_10_ng_template_0_Template, 0, 0, "ng-template");
} }
function WizardComponent_ng_template_12_Template(rf, ctx) { if (rf & 1) {
    const _r14 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 11);
    ɵngcc0.ɵɵelementStart(1, "button", 12);
    ɵngcc0.ɵɵlistener("click", function WizardComponent_ng_template_12_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r14); const ctx_r13 = ɵngcc0.ɵɵnextContext(); return ctx_r13.close(); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementStart(3, "span", 13);
    ɵngcc0.ɵɵtext(4, "Cancel");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 1, "Cancel"));
} }
export class WizardComponent {
    constructor(wizardService, bsModalRef) {
        this.wizardService = wizardService;
        this.bsModalRef = bsModalRef;
        /**
         * A configuration object that allows you to set header values, both text and icon.
         */
        this.wizardConfig = {};
        /**
         * A state that will be assigned to the actual component contained by the wizard.
         */
        this.componentInitialState = {};
        /**
         * The event that is emitted when the wizard is closed.
         */
        this.onClose = new EventEmitter();
        /**
         * The event that is emitted when the wizard is reseted.
         */
        this.onReset = new EventEmitter();
        /**
         * The event that is emitted when the wizard entry is clicked.
         */
        this.onSelect = new EventEmitter();
        this.headerTemplate$ = this.wizardService.headerTemplate$.pipe(filter(header => header.wizardId === this.id));
        this.bodyTemplate$ = this.wizardService.bodyTemplate$.pipe(filter(body => body.wizardId === this.id));
        this.footerTemplate$ = this.wizardService.footerTemplate$.pipe(filter(footer => footer.wizardId === this.id));
    }
    ngOnInit() {
        this.emitTemplates();
    }
    /**
     * Resets the wizard.
     * @param result Custom result object which will be emitted via onReset output.
     */
    reset(result) {
        this.onReset.emit(result);
        this.emitTemplates();
    }
    /**
     * Closes the wizard.
     * @param result Custom result object which will be emitted via onClose output.
     */
    close(result) {
        this.onClose.emit(result);
        this.onClose.complete();
        this.bsModalRef.hide();
    }
    emitTemplates() {
        setTimeout(() => {
            this.wizardService.headerTemplate$.next({ template: this.header, wizardId: this.id });
            this.wizardService.bodyTemplate$.next({ template: this.body, wizardId: this.id });
            this.wizardService.footerTemplate$.next({ template: this.footer, wizardId: this.id });
        });
    }
}
WizardComponent.ɵfac = function WizardComponent_Factory(t) { return new (t || WizardComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.WizardService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.BsModalRef)); };
WizardComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: WizardComponent, selectors: [["c8y-wizard"]], viewQuery: function WizardComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 7);
        ɵngcc0.ɵɵviewQuery(_c1, 7);
        ɵngcc0.ɵɵviewQuery(_c2, 7);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.header = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.body = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.footer = _t.first);
    } }, inputs: { wizardConfig: "wizardConfig", id: "id" }, outputs: { onClose: "onClose", onReset: "onReset", onSelect: "onSelect" }, decls: 14, vars: 10, consts: [[1, "viewport-modal", "animated", "fadeIn"], [4, "ngTemplateOutlet"], ["headerRef", ""], ["bodyRef", ""], [3, "initialState", "onSelect"], ["footerRef", ""], [1, "modal-header", "dialog-header", "animated", "fadeIn"], [3, "c8yIcon"], ["class", "p-16 p-t-8 text-center", 4, "ngIf"], [1, "p-16", "p-t-8", "text-center"], [1, "c8y-wizard-nav"], [1, "modal-footer", "animated", "fadeIn"], ["type", "button", 1, "btn", "btn-default", 3, "title", "click"], ["translate", ""]], template: function WizardComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, WizardComponent_1_Template, 1, 0, undefined, 1);
        ɵngcc0.ɵɵpipe(2, "async");
        ɵngcc0.ɵɵtemplate(3, WizardComponent_ng_template_3_Template, 5, 4, "ng-template", null, 2, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(5, WizardComponent_5_Template, 1, 0, undefined, 1);
        ɵngcc0.ɵɵpipe(6, "async");
        ɵngcc0.ɵɵtemplate(7, WizardComponent_ng_template_7_Template, 1, 1, "ng-template", null, 3, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementStart(9, "c8y-wizard-outlet", 4);
        ɵngcc0.ɵɵlistener("onSelect", function WizardComponent_Template_c8y_wizard_outlet_onSelect_9_listener($event) { return ctx.onSelect.emit($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(10, WizardComponent_10_Template, 1, 0, undefined, 1);
        ɵngcc0.ɵɵpipe(11, "async");
        ɵngcc0.ɵɵtemplate(12, WizardComponent_ng_template_12_Template, 5, 3, "ng-template", null, 5, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        let tmp_0_0;
        let tmp_1_0;
        let tmp_3_0;
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", (tmp_0_0 = ɵngcc0.ɵɵpipeBind1(2, 4, ctx.headerTemplate$)) == null ? null : tmp_0_0.template);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", (tmp_1_0 = ɵngcc0.ɵɵpipeBind1(6, 6, ctx.bodyTemplate$)) == null ? null : tmp_1_0.template);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("initialState", ctx.componentInitialState);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", (tmp_3_0 = ɵngcc0.ɵɵpipeBind1(11, 8, ctx.footerTemplate$)) == null ? null : tmp_3_0.template);
    } }, encapsulation: 2 });
WizardComponent.ctorParameters = () => [
    { type: WizardService },
    { type: BsModalRef }
];
WizardComponent.propDecorators = {
    header: [{ type: ViewChild, args: ['headerRef', { static: true },] }],
    body: [{ type: ViewChild, args: ['bodyRef', { static: true },] }],
    footer: [{ type: ViewChild, args: ['footerRef', { static: true },] }],
    id: [{ type: Input }],
    wizardConfig: [{ type: Input }],
    onClose: [{ type: Output }],
    onReset: [{ type: Output }],
    onSelect: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(WizardComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-wizard',
                template: "<div class=\"viewport-modal animated fadeIn\">\n  <ng-template *ngTemplateOutlet=\"(headerTemplate$ | async)?.template\"></ng-template>\n\n  <ng-template #headerRef>\n    <div class=\"modal-header dialog-header animated fadeIn\">\n      <h1 [c8yIcon]=\"wizardConfig.headerIcon\"></h1>\n      <h4>{{ wizardConfig.headerText | translate }}</h4>\n    </div>\n  </ng-template>\n\n  <ng-template *ngTemplateOutlet=\"(bodyTemplate$ | async)?.template\"></ng-template>\n\n  <ng-template #bodyRef>\n    <div class=\"p-16 p-t-8 text-center\" *ngIf=\"wizardConfig.bodyHeaderText\">\n      <div class=\"c8y-wizard-nav\">\n        <i [c8yIcon]=\"wizardConfig.bodyHeaderIcon\"></i>\n        <span>{{ wizardConfig.bodyHeaderText | translate }}</span>\n      </div>\n    </div>\n  </ng-template>\n  <c8y-wizard-outlet [initialState]=\"componentInitialState\" (onSelect)=\"onSelect.emit($event)\"></c8y-wizard-outlet>\n\n  <ng-template *ngTemplateOutlet=\"(footerTemplate$ | async)?.template\"></ng-template>\n  <ng-template #footerRef>\n    <div class=\"modal-footer animated fadeIn\">\n      <button\n        (click)=\"close()\"\n        type=\"button\"\n        class=\"btn btn-default\"\n        title=\"{{ 'Cancel' | translate }}\"\n      >\n        <span translate>Cancel</span>\n      </button>\n    </div>\n  </ng-template>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.WizardService }, { type: ɵngcc2.BsModalRef }]; }, { wizardConfig: [{
            type: Input
        }], onClose: [{
            type: Output
        }], onReset: [{
            type: Output
        }], onSelect: [{
            type: Output
        }], header: [{
            type: ViewChild,
            args: ['headerRef', { static: true }]
        }], body: [{
            type: ViewChild,
            args: ['bodyRef', { static: true }]
        }], footer: [{
            type: ViewChild,
            args: ['footerRef', { static: true }]
        }], id: [{
            type: Input
        }] }); })();
