// @c8y/ngx-components/esm2015/core/dynamic-forms/form-field/form-field.wrapper.component.js
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { FieldWrapper } from '@ngx-formly/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '../../common/icon.directive';
import * as ɵngcc3 from 'ngx-bootstrap/popover';
import * as ɵngcc4 from '@ngx-formly/core';
import * as ɵngcc5 from '../../common/humanize.pipe';
import * as ɵngcc6 from '../../i18n/c8y-translate.pipe';

function WrapperFormField_label_1_i_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 7);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 1, ctx_r5.to.tooltip));
} }
function WrapperFormField_label_1_button_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "button", 8);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 9);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("popover", ɵngcc0.ɵɵpipeBind1(1, 1, ctx_r6.to.description));
} }
function WrapperFormField_label_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "label");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "humanize");
    ɵngcc0.ɵɵtemplate(3, WrapperFormField_label_1_i_3_Template, 2, 3, "i", 5);
    ɵngcc0.ɵɵtemplate(4, WrapperFormField_label_1_button_4_Template, 3, 3, "button", 6);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵattribute("for", ctx_r0.id);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 4, ctx_r0.to.label), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.to.tooltip);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.to.description && ctx_r0.showDescriptionAsPopup);
} }
function WrapperFormField_ng_template_2_Template(rf, ctx) { }
function WrapperFormField_div_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 10);
    ɵngcc0.ɵɵelement(1, "formly-validation-message", 11);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("field", ctx_r3.field);
} }
function WrapperFormField_div_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 12);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, ctx_r4.to.description), " ");
} }
export class WrapperFormField extends FieldWrapper {
    constructor() {
        super(...arguments);
        this.maxHelpBlockLength = 64;
    }
    ngOnInit() {
        this.showDescriptionAsPopup =
            this.field.type === 'radio' || this.field.type === 'typeahead' ||
                (this.to.description && this.to.description.length > this.maxHelpBlockLength);
    }
}
WrapperFormField.ɵfac = /*@__PURE__*/ function () { let ɵWrapperFormField_BaseFactory; return function WrapperFormField_Factory(t) { return (ɵWrapperFormField_BaseFactory || (ɵWrapperFormField_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(WrapperFormField)))(t || WrapperFormField); }; }();
WrapperFormField.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: WrapperFormField, selectors: [["c8y-wrapper-form-field"]], features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 6, vars: 5, consts: [[1, "form-group"], [4, "ngIf"], ["fieldComponent", ""], ["class", "c8y-messages", 4, "ngIf"], ["class", "help-block", 4, "ngIf"], ["class", "text-info", "c8yIcon", "info-circle", 3, "title", 4, "ngIf"], ["class", "btn-clean m-l-4 flex-item-middle", "placement", "auto", "outsideClick", "true", "type", "button", 3, "popover", 4, "ngIf"], ["c8yIcon", "info-circle", 1, "text-info", 3, "title"], ["placement", "auto", "outsideClick", "true", "type", "button", 1, "btn-clean", "m-l-4", "flex-item-middle", 3, "popover"], ["c8yIcon", "question-circle-o", 1, "text-info"], [1, "c8y-messages"], [1, "form-control-feedback-message", 3, "field"], [1, "help-block"]], template: function WrapperFormField_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, WrapperFormField_label_1_Template, 5, 6, "label", 1);
        ɵngcc0.ɵɵtemplate(2, WrapperFormField_ng_template_2_Template, 0, 0, "ng-template", null, 2, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(4, WrapperFormField_div_4_Template, 2, 1, "div", 3);
        ɵngcc0.ɵɵtemplate(5, WrapperFormField_div_5_Template, 3, 3, "div", 4);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassProp("has-error", ctx.showError);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.to.label && ctx.to.hideLabel !== true);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showError);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.showError && ctx.to.description && !ctx.showDescriptionAsPopup);
    } }, directives: [ɵngcc1.NgIf, ɵngcc2.IconDirective, ɵngcc3.PopoverDirective, ɵngcc4.ɵc], pipes: [ɵngcc5.HumanizePipe, ɵngcc6.C8yTranslatePipe], encapsulation: 2, changeDetection: 0 });
WrapperFormField.CONFIG = {
    wrappers: [{ name: 'c8y-form-field', component: WrapperFormField }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(WrapperFormField, [{
        type: Component,
        args: [{
                selector: 'c8y-wrapper-form-field',
                template: "<div class=\"form-group\" [class.has-error]=\"showError\">\n  <label *ngIf=\"to.label && to.hideLabel !== true\" [attr.for]=\"id\">\n    {{ to.label | humanize }} <i\n      *ngIf=\"to.tooltip\"\n      class=\"text-info\"\n      c8yIcon=\"info-circle\"\n      title=\"{{ to.tooltip | translate }}\"\n    ></i>\n    <button\n      *ngIf=\"to.description && showDescriptionAsPopup\"\n      class=\"btn-clean m-l-4 flex-item-middle\"\n      [popover]=\"to.description | translate\"\n      placement=\"auto\"\n      outsideClick=\"true\"\n      type=\"button\"\n    >\n      <i c8yIcon=\"question-circle-o\" class=\"text-info\"></i>\n    </button>\n  </label>\n\n  <ng-template #fieldComponent></ng-template>\n\n  <div *ngIf=\"showError\" class=\"c8y-messages\">\n    <formly-validation-message\n      class=\"form-control-feedback-message\"\n      [field]=\"field\"\n    ></formly-validation-message>\n  </div>\n\n  <div class=\"help-block\" *ngIf=\"!showError && to.description && !showDescriptionAsPopup\">\n    {{ to.description | translate }}\n  </div>\n</div>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], null, null); })();
