// @c8y/ngx-components/fesm2015/c8y-ngx-components-operations-stepper-bulk-type-configuration.js
import { Component, NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { gettext, CoreModule, FormsModule } from '@c8y/ngx-components';
import { RepositoryType, RepositoryService, SharedRepositoryModule } from '@c8y/ngx-components/repository/shared';
import { BaseStepperComponent, BulkOperationStepperModule } from '@c8y/ngx-components/operations/bulk-operation-stepper';
import { BulkOperationsService, BulkOperationType, baseUrl, HOOK_LIST_BULK_TYPE } from '@c8y/ngx-components/operations/bulk-operations-service';
import { StepperFramesModule } from '@c8y/ngx-components/operations/stepper-frames';
import { __awaiter } from 'tslib';
import { TranslateService } from '@ngx-translate/core';
import { uniqWith, isEqual, get } from 'lodash-es';
import { BehaviorSubject, combineLatest, from } from 'rxjs';
import { switchMap, shareReplay, debounceTime, distinctUntilChanged } from 'rxjs/operators';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components/operations/bulk-operations-service';
import * as ɵngcc2 from '@c8y/ngx-components/repository/shared';
import * as ɵngcc3 from '@ngx-translate/core';
import * as ɵngcc4 from '@c8y/ngx-components/operations/bulk-operation-stepper';
import * as ɵngcc5 from '@c8y/ngx-components/operations/stepper-frames';
import * as ɵngcc6 from '@angular/forms';
import * as ɵngcc7 from '@angular/common';
import * as ɵngcc8 from '@c8y/ngx-components';

function StepperBulkTypeConfigurationComponent_ng_container_1_i_14_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 24);
} }
function StepperBulkTypeConfigurationComponent_ng_container_1_i_16_Template(rf, ctx) { if (rf & 1) {
    const _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "i", 25);
    ɵngcc0.ɵɵlistener("click", function StepperBulkTypeConfigurationComponent_ng_container_1_i_16_Template_i_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r7); const ctx_r6 = ɵngcc0.ɵɵnextContext(2); return ctx_r6.textFilter$.next(""); });
    ɵngcc0.ɵɵelementEnd();
} }
function StepperBulkTypeConfigurationComponent_ng_container_1_c8y_li_27_Template(rf, ctx) { if (rf & 1) {
    const _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 26);
    ɵngcc0.ɵɵlistener("click", function StepperBulkTypeConfigurationComponent_ng_container_1_c8y_li_27_Template_c8y_li_click_0_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r10); const configType_r8 = restoredCtx.$implicit; const ctx_r9 = ɵngcc0.ɵɵnextContext(2); ctx_r9.selectedConfigType = configType_r8; return ctx_r9.configType$.next(configType_r8.name); });
    ɵngcc0.ɵɵelement(1, "c8y-highlight", 27);
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const configType_r8 = ctx.$implicit;
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("active", ctx_r4.selectedConfigType === configType_r8);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("text", configType_r8.name)("pattern", ɵngcc0.ɵɵpipeBind1(2, 3, ctx_r4.configType$));
} }
function StepperBulkTypeConfigurationComponent_ng_container_1_c8y_li_49_em_15_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "em", 38);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Undefined`device type`"), " ");
} }
function StepperBulkTypeConfigurationComponent_ng_container_1_c8y_li_49_span_19_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 39);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "c8y-highlight", 40);
    ɵngcc0.ɵɵpipe(3, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const configuration_r11 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r14 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵpropertyInterpolate2("title", "", ɵngcc0.ɵɵpipeBind1(1, 4, "Configuration type"), ": ", configuration_r11.configurationType, "");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("text", configuration_r11.configurationType)("pattern", ɵngcc0.ɵɵpipeBind1(3, 6, ctx_r14.configType$));
} }
function StepperBulkTypeConfigurationComponent_ng_container_1_c8y_li_49_span_20_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 41);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate2("title", "", ɵngcc0.ɵɵpipeBind1(1, 3, "Configuration type"), ": ", ɵngcc0.ɵɵpipeBind1(2, 5, "Undefined"), "");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 7, "Undefined"), " ");
} }
function StepperBulkTypeConfigurationComponent_ng_container_1_c8y_li_49_Template(rf, ctx) { if (rf & 1) {
    const _r18 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li", null, 28);
    ɵngcc0.ɵɵelementStart(2, "c8y-li-radio", 29);
    ɵngcc0.ɵɵlistener("onSelect", function StepperBulkTypeConfigurationComponent_ng_container_1_c8y_li_49_Template_c8y_li_radio_onSelect_2_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r18); const configuration_r11 = restoredCtx.$implicit; const ctx_r17 = ɵngcc0.ɵɵnextContext(2); return ctx_r17.selectConfiguration(configuration_r11); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(3, "c8y-li-icon", 30);
    ɵngcc0.ɵɵelementStart(4, "c8y-li-body", 17);
    ɵngcc0.ɵɵelementStart(5, "div", 31);
    ɵngcc0.ɵɵelementStart(6, "div", 19);
    ɵngcc0.ɵɵelement(7, "c8y-highlight", 27);
    ɵngcc0.ɵɵpipe(8, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(9, "div", 32);
    ɵngcc0.ɵɵelementStart(10, "div", 19);
    ɵngcc0.ɵɵpipe(11, "translate");
    ɵngcc0.ɵɵelementStart(12, "span", 33);
    ɵngcc0.ɵɵtext(13, "Device type");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(14);
    ɵngcc0.ɵɵtemplate(15, StepperBulkTypeConfigurationComponent_ng_container_1_c8y_li_49_em_15_Template, 3, 3, "em", 34);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(16, "div", 35);
    ɵngcc0.ɵɵelementStart(17, "span", 33);
    ɵngcc0.ɵɵtext(18, " Configuration type ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(19, StepperBulkTypeConfigurationComponent_ng_container_1_c8y_li_49_span_19_Template, 4, 8, "span", 36);
    ɵngcc0.ɵɵtemplate(20, StepperBulkTypeConfigurationComponent_ng_container_1_c8y_li_49_span_20_Template, 5, 9, "span", 37);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const configuration_r11 = ctx.$implicit;
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵpropertyInterpolate("title", configuration_r11.name || "-");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("text", configuration_r11.name || "-")("pattern", ɵngcc0.ɵɵpipeBind1(8, 9, ctx_r5.textFilter$));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate2("title", "", ɵngcc0.ɵɵpipeBind1(11, 11, "Device type"), ": ", ctx_r5.getDeviceTypeTitle(configuration_r11), "");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate1(" ", configuration_r11.deviceType, " ");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !configuration_r11.deviceType);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngIf", configuration_r11.configurationType);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !configuration_r11.configurationType);
} }
const _c0 = function (a0) { return { "dd-low": a0 }; };
function StepperBulkTypeConfigurationComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    const _r20 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "c8y-select-step-frame", 2);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵpipe(3, "async");
    ɵngcc0.ɵɵpipe(4, "async");
    ɵngcc0.ɵɵpipe(5, "async");
    ɵngcc0.ɵɵelementStart(6, "div", 3);
    ɵngcc0.ɵɵelementStart(7, "div", 4);
    ɵngcc0.ɵɵelementStart(8, "div", 5);
    ɵngcc0.ɵɵelementStart(9, "input", 6);
    ɵngcc0.ɵɵlistener("ngModelChange", function StepperBulkTypeConfigurationComponent_ng_container_1_Template_input_ngModelChange_9_listener($event) { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r19 = ɵngcc0.ɵɵnextContext(); return ctx_r19.textFilter$.next($event); });
    ɵngcc0.ɵɵpipe(10, "translate");
    ɵngcc0.ɵɵpipe(11, "translate");
    ɵngcc0.ɵɵpipe(12, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(13, "span", 7);
    ɵngcc0.ɵɵtemplate(14, StepperBulkTypeConfigurationComponent_ng_container_1_i_14_Template, 1, 0, "i", 8);
    ɵngcc0.ɵɵpipe(15, "async");
    ɵngcc0.ɵɵtemplate(16, StepperBulkTypeConfigurationComponent_ng_container_1_i_16_Template, 1, 0, "i", 9);
    ɵngcc0.ɵɵpipe(17, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(18, "div", 10);
    ɵngcc0.ɵɵelementStart(19, "div", 4);
    ɵngcc0.ɵɵelementStart(20, "c8y-form-group", 11);
    ɵngcc0.ɵɵelementStart(21, "c8y-typeahead", 12);
    ɵngcc0.ɵɵlistener("ngModelChange", function StepperBulkTypeConfigurationComponent_ng_container_1_Template_c8y_typeahead_ngModelChange_21_listener($event) { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r21 = ɵngcc0.ɵɵnextContext(); return ctx_r21.selectedConfigType = $event; })("onSearch", function StepperBulkTypeConfigurationComponent_ng_container_1_Template_c8y_typeahead_onSearch_21_listener($event) { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r22 = ɵngcc0.ɵɵnextContext(); return ctx_r22.configType$.next($event); });
    ɵngcc0.ɵɵpipe(22, "translate");
    ɵngcc0.ɵɵelementStart(23, "c8y-li", 13);
    ɵngcc0.ɵɵlistener("click", function StepperBulkTypeConfigurationComponent_ng_container_1_Template_c8y_li_click_23_listener() { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r23 = ɵngcc0.ɵɵnextContext(); ctx_r23.selectedConfigType = { name: "" }; return ctx_r23.configType$.next(""); });
    ɵngcc0.ɵɵelementStart(24, "span");
    ɵngcc0.ɵɵtext(25);
    ɵngcc0.ɵɵpipe(26, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(27, StepperBulkTypeConfigurationComponent_ng_container_1_c8y_li_27_Template, 3, 5, "c8y-li", 14);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(28, "c8y-li", 15);
    ɵngcc0.ɵɵelementStart(29, "c8y-li-icon");
    ɵngcc0.ɵɵelement(30, "i", 16);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(31, "c8y-li-body", 17);
    ɵngcc0.ɵɵelementStart(32, "div", 18);
    ɵngcc0.ɵɵelementStart(33, "span", 19);
    ɵngcc0.ɵɵpipe(34, "translate");
    ɵngcc0.ɵɵtext(35);
    ɵngcc0.ɵɵpipe(36, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(37, "div", 20);
    ɵngcc0.ɵɵelementStart(38, "span", 19);
    ɵngcc0.ɵɵpipe(39, "translate");
    ɵngcc0.ɵɵtext(40);
    ɵngcc0.ɵɵpipe(41, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(42, "div", 21);
    ɵngcc0.ɵɵelementStart(43, "span", 19);
    ɵngcc0.ɵɵpipe(44, "translate");
    ɵngcc0.ɵɵtext(45);
    ɵngcc0.ɵɵpipe(46, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(47, "c8y-list-group", 22);
    ɵngcc0.ɵɵpipe(48, "async");
    ɵngcc0.ɵɵtemplate(49, StepperBulkTypeConfigurationComponent_ng_container_1_c8y_li_49_Template, 21, 13, "c8y-li", 23);
    ɵngcc0.ɵɵpipe(50, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    let tmp_1_0;
    let tmp_17_0;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("header", ɵngcc0.ɵɵpipeBind1(2, 19, "Select configuration"))("noResults", (((tmp_1_0 = ɵngcc0.ɵɵpipeBind1(3, 21, ctx_r0.textFilter$)) == null ? null : tmp_1_0.length) || ((tmp_1_0 = ɵngcc0.ɵɵpipeBind1(4, 23, ctx_r0.configType$)) == null ? null : tmp_1_0.length)) && ((tmp_1_0 = ɵngcc0.ɵɵpipeBind1(5, 25, ctx_r0.configurations$)) == null ? null : tmp_1_0.data.length) === 0);
    ɵngcc0.ɵɵadvance(8);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(10, 27, "Filter\u2026"));
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(11, 29, "Filter\u2026"));
    ɵngcc0.ɵɵproperty("ngModel", ɵngcc0.ɵɵpipeBind1(12, 31, ctx_r0.textFilter$));
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(15, 33, ctx_r0.textFilter$).length === 0);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(17, 35, ctx_r0.textFilter$).length > 0);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(22, 37, "Type to filter configuration types\u2026"));
    ɵngcc0.ɵɵproperty("ngModel", ctx_r0.selectedConfigType);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(26, 39, "All configuration types"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r0.configTypes);
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵpropertyInterpolate1("title", " ", ɵngcc0.ɵɵpipeBind1(34, 41, "Name"), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(36, 43, "Name"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(39, 45, "Device type"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(41, 47, "Device type"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(44, 49, "Configuration type"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(46, 51, "Configuration type"), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(57, _c0, ((tmp_17_0 = ɵngcc0.ɵɵpipeBind1(48, 53, ctx_r0.configurations$)) == null ? null : tmp_17_0.data == null ? null : tmp_17_0.data.length) < ctx_r0.DD_LOW_COUNT));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yForOf", ɵngcc0.ɵɵpipeBind1(50, 55, ctx_r0.configurations$));
} }
function StepperBulkTypeConfigurationComponent_ng_container_3_ng_container_3_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "span");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r27 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r27.selectedConfiguration == null ? null : ctx_r27.selectedConfiguration.deviceType);
} }
function StepperBulkTypeConfigurationComponent_ng_container_3_ng_container_3_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "em", 46);
    ɵngcc0.ɵɵtext(1, "Undefined`device type`");
    ɵngcc0.ɵɵelementEnd();
} }
function StepperBulkTypeConfigurationComponent_ng_container_3_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, StepperBulkTypeConfigurationComponent_ng_container_3_ng_container_3_ng_container_1_Template, 3, 1, "ng-container", 44);
    ɵngcc0.ɵɵtemplate(2, StepperBulkTypeConfigurationComponent_ng_container_3_ng_container_3_ng_template_2_Template, 2, 0, "ng-template", null, 45, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const _r28 = ɵngcc0.ɵɵreference(3);
    const ctx_r24 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r24.selectedConfiguration == null ? null : ctx_r24.selectedConfiguration.deviceType)("ngIfElse", _r28);
} }
function StepperBulkTypeConfigurationComponent_ng_container_3_ng_container_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "span");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "c8yDate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r25 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(3, 1, ctx_r25.selectedConfiguration == null ? null : ctx_r25.selectedConfiguration.lastUpdated));
} }
function StepperBulkTypeConfigurationComponent_ng_container_3_ng_container_7_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 49);
    ɵngcc0.ɵɵelement(1, "h1", 50);
    ɵngcc0.ɵɵelementStart(2, "p");
    ɵngcc0.ɵɵelementStart(3, "strong", 51);
    ɵngcc0.ɵɵtext(4, "No preview available.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(5, "br");
    ɵngcc0.ɵɵelementStart(6, "small", 51);
    ɵngcc0.ɵɵtext(7, "Could not fetch the file.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yIcon", "file-code-o");
} }
function StepperBulkTypeConfigurationComponent_ng_container_3_ng_container_7_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "pre", 52);
    ɵngcc0.ɵɵelementStart(1, "code");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r32 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r32.selectedConfiguration == null ? null : ctx_r32.selectedConfiguration.binary);
} }
function StepperBulkTypeConfigurationComponent_ng_container_3_ng_container_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, StepperBulkTypeConfigurationComponent_ng_container_3_ng_container_7_div_1_Template, 8, 1, "div", 47);
    ɵngcc0.ɵɵtemplate(2, StepperBulkTypeConfigurationComponent_ng_container_3_ng_container_7_ng_template_2_Template, 3, 1, "ng-template", null, 48, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const _r31 = ɵngcc0.ɵɵreference(3);
    const ctx_r26 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !(ctx_r26.selectedConfiguration == null ? null : ctx_r26.selectedConfiguration.binary))("ngIfElse", _r31);
} }
function StepperBulkTypeConfigurationComponent_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "c8y-preview-selection-frame", 42);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵtemplate(3, StepperBulkTypeConfigurationComponent_ng_container_3_ng_container_3_Template, 4, 2, "ng-container", 43);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵtemplate(5, StepperBulkTypeConfigurationComponent_ng_container_3_ng_container_5_Template, 4, 3, "ng-container", 43);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵtemplate(7, StepperBulkTypeConfigurationComponent_ng_container_3_ng_container_7_Template, 4, 2, "ng-container", 43);
    ɵngcc0.ɵɵpipe(8, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("header", ɵngcc0.ɵɵpipeBind1(2, 6, "Preview configuration"))("name", ctx_r1.selectedConfiguration == null ? null : ctx_r1.selectedConfiguration.name)("type", ctx_r1.selectedConfiguration == null ? null : ctx_r1.selectedConfiguration.configurationType);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("previewBlock", ɵngcc0.ɵɵpipeBind1(4, 8, "Device type"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("previewBlock", ɵngcc0.ɵɵpipeBind1(6, 10, "Updated"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("previewBlock", ɵngcc0.ɵɵpipeBind1(8, 12, "Preview"));
} }
class StepperBulkTypeConfigurationComponent extends BaseStepperComponent {
    constructor(bulkOperationService, repositoryService, translate) {
        super();
        this.bulkOperationService = bulkOperationService;
        this.repositoryService = repositoryService;
        this.translate = translate;
        this.NO_DEVICE_TYPE_AVAILABLE = gettext('Undefined`device type`');
        this.DD_LOW_COUNT = 10;
        this.textFilter$ = new BehaviorSubject('');
        this.configType$ = new BehaviorSubject('');
        this.configTypes = [];
        this.selectedConfigType = { name: '' };
        this.configurations$ = combineLatest(this.textFilter$, this.configType$).pipe(switchMap(([name, configType]) => this.getConfiguration(name, configType)), shareReplay(1));
        this.DD_LOW_COUNT = this.bulkOperationService.DD_LOW_COUNT;
        this.loadConfigurationTypes();
    }
    loadConfigurationTypes() {
        this.configTypeSubscription = this.configType$
            .pipe(debounceTime(300), distinctUntilChanged(), switchMap(searchStr => {
            const query = { configurationType: `*${searchStr}*` };
            return from(this.repositoryService.listRepositoryEntries(RepositoryType.CONFIGURATION, { query }));
        }))
            .subscribe(result => {
            const { data } = result;
            this.configTypes = uniqWith(data.map(val => ({ name: val.configurationType })), isEqual);
        });
    }
    ngOnDestroy() {
        this.configTypeSubscription.unsubscribe();
    }
    selectConfiguration(configuration) {
        this.selectedConfiguration = configuration;
    }
    goToSecondStep($event) {
        $event.stepper.next();
        this.getConfigBinary();
        this.deviceTypes = this.selectedConfiguration.deviceType;
    }
    getConfigBinary() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.selectedConfiguration.url) {
                this.selectedConfiguration.binary = yield this.repositoryService.getBinaryText(this.selectedConfiguration.url, { allowExternal: true, noAlerts: true });
            }
        });
    }
    getDeviceTypeTitle(configuration) {
        return get(configuration, 'deviceType', this.translate.instant(this.NO_DEVICE_TYPE_AVAILABLE));
    }
    retrieveOperationPrototype() {
        const configuration = {
            type: this.selectedConfiguration.configurationType,
            url: this.selectedConfiguration.url
        };
        return {
            name: gettext('Configuration update'),
            description: get(this.selectedConfiguration, 'name'),
            prototype: {
                description: `Update configuration to: ${this.selectedConfiguration.name}.`,
                c8y_DownloadConfigFile: configuration
            }
        };
    }
    getConfiguration(name, configurationType) {
        const query = name ? { name: `*${name}*` } : {};
        if (configurationType) {
            query.__or = [{ configurationType }, { __not: { __has: `configurationType` } }];
        }
        return this.repositoryService.listRepositoryEntries(RepositoryType.CONFIGURATION, { query });
    }
}
StepperBulkTypeConfigurationComponent.ɵfac = function StepperBulkTypeConfigurationComponent_Factory(t) { return new (t || StepperBulkTypeConfigurationComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BulkOperationsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.RepositoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.TranslateService)); };
StepperBulkTypeConfigurationComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: StepperBulkTypeConfigurationComponent, selectors: [["c8y-stepper-bulk-type-configuration"]], features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 5, vars: 9, consts: [[4, "customStep", "customStepCompleted", "customStepButtonsDisabled", "customStepOnNext"], [4, "customStep"], [3, "header", "noResults"], ["filters", "", 1, "row"], [1, "col-md-6"], [1, "input-group", "input-group-search"], ["type", "search", 1, "form-control", 3, "title", "placeholder", "ngModel", "ngModelChange"], [1, "input-group-addon"], ["c8yIcon", "filter", 4, "ngIf"], ["c8yIcon", "times", "class", "text-muted", 3, "click", 4, "ngIf"], [1, "col-xs-12", "p-b-8", "visible-xs", "visible-sm"], [1, "m-0"], ["name", "configType", 3, "ngModel", "placeholder", "ngModelChange", "onSearch"], [1, "p-l-8", "p-r-8", "c8y-list__item--link", 3, "click"], ["class", "p-l-8 p-r-8 c8y-list__item--link", 3, "active", "click", 4, "ngFor", "ngForOf"], ["header", "", 1, "hidden-sm", "hidden-xs", "m-r-8", "m-l-8"], [1, "p-l-32", "p-r-40"], [1, "content-flex-40"], [1, "col-5"], [1, "text-truncate", 3, "title"], [1, "col-4"], [1, "col-3"], ["list", "", 3, "ngClass"], [4, "c8yFor", "c8yForOf"], ["c8yIcon", "filter"], ["c8yIcon", "times", 1, "text-muted", 3, "click"], [1, "p-l-8", "p-r-8", "c8y-list__item--link", 3, "active", "click"], [3, "text", "pattern"], ["listItem", ""], [3, "onSelect"], ["icon", "gears"], [1, "col-5", "m-b-xs-8"], [1, "col-4", "m-b-xs-8"], ["translate", "", 1, "text-label-small", "m-r-8", "visible-xs", "visible-sm"], ["class", "text-muted", 4, "ngIf"], [1, "col-3", "m-b-xs-8"], ["class", "label label-info", 3, "title", 4, "ngIf"], ["class", "label label-default", 3, "title", 4, "ngIf"], [1, "text-muted"], [1, "label", "label-info", 3, "title"], ["elementClass", "text-gray-lighter", 3, "text", "pattern"], [1, "label", "label-default", 3, "title"], [3, "header", "name", "type"], [4, "previewBlock"], [4, "ngIf", "ngIfElse"], ["noType", ""], ["translate", "", 1, "text-muted"], ["class", "c8y-empty-state text-left", 4, "ngIf", "ngIfElse"], ["binaryPreview", ""], [1, "c8y-empty-state", "text-left"], [3, "c8yIcon"], ["translate", ""], [2, "min-height", "98px"]], template: function StepperBulkTypeConfigurationComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-bulk-operation-stepper");
        ɵngcc0.ɵɵtemplate(1, StepperBulkTypeConfigurationComponent_ng_container_1_Template, 51, 59, "ng-container", 0);
        ɵngcc0.ɵɵpipe(2, "translate");
        ɵngcc0.ɵɵtemplate(3, StepperBulkTypeConfigurationComponent_ng_container_3_Template, 9, 14, "ng-container", 1);
        ɵngcc0.ɵɵpipe(4, "translate");
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("customStep", ɵngcc0.ɵɵpipeBind1(2, 5, "Select configuration"))("customStepCompleted", !!ctx.selectedConfiguration)("customStepButtonsDisabled", !ctx.selectedConfiguration)("customStepOnNext", ctx.goToSecondStep.bind(ctx));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("customStep", ɵngcc0.ɵɵpipeBind1(4, 7, "Preview configuration"));
    } }, directives: [ɵngcc4.BulkOperationStepper, ɵngcc4.CustomStep, ɵngcc5.SelectStepFrameComponent, ɵngcc6.DefaultValueAccessor, ɵngcc6.NgControlStatus, ɵngcc6.NgModel, ɵngcc7.NgIf, ɵngcc8.FormGroupComponent, ɵngcc8.TypeaheadComponent, ɵngcc8.ListItemComponent, ɵngcc7.NgForOf, ɵngcc8.ListItemIconComponent, ɵngcc8.ListItemBodyComponent, ɵngcc8.ListGroupComponent, ɵngcc7.NgClass, ɵngcc8.ForOfDirective, ɵngcc8.IconDirective, ɵngcc8.HighlightComponent, ɵngcc8.ListItemRadioComponent, ɵngcc8.C8yTranslateDirective, ɵngcc5.PreviewSelectionFrameComponent, ɵngcc5.PreviewBlockDirective], pipes: [ɵngcc8.C8yTranslatePipe, ɵngcc7.AsyncPipe, ɵngcc8.DatePipe], encapsulation: 2 });
StepperBulkTypeConfigurationComponent.ctorParameters = () => [
    { type: BulkOperationsService },
    { type: RepositoryService },
    { type: TranslateService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(StepperBulkTypeConfigurationComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-stepper-bulk-type-configuration',
                template: "<c8y-bulk-operation-stepper>\n  <ng-container\n    *customStep=\"\n      'Select configuration' | translate;\n      completed: !!selectedConfiguration;\n      buttonsDisabled: !selectedConfiguration;\n      onNext: goToSecondStep.bind(this)\n    \"\n  >\n    <c8y-select-step-frame\n      [header]=\"'Select configuration' | translate\"\n      [noResults]=\"\n        ((textFilter$ | async)?.length || (configType$ | async)?.length) &&\n        (configurations$ | async)?.data.length === 0\n      \"\n    >\n      <div filters class=\"row\">\n        <div class=\"col-md-6\">\n          <div class=\"input-group input-group-search\">\n            <input\n              type=\"search\"\n              class=\"form-control\"\n              title=\"{{ 'Filter\u2026' | translate }}\"\n              placeholder=\"{{ 'Filter\u2026' | translate }}\"\n              [ngModel]=\"textFilter$ | async\"\n              (ngModelChange)=\"textFilter$.next($event)\"\n            />\n            <span class=\"input-group-addon\">\n              <i c8yIcon=\"filter\" *ngIf=\"(textFilter$ | async).length === 0\"></i>\n              <i\n                c8yIcon=\"times\"\n                class=\"text-muted\"\n                *ngIf=\"(textFilter$ | async).length > 0\"\n                (click)=\"textFilter$.next('')\"\n              ></i>\n            </span>\n          </div>\n        </div>\n        <div class=\"col-xs-12 p-b-8 visible-xs visible-sm\"></div>\n        <div class=\"col-md-6\">\n          <c8y-form-group class=\"m-0\">\n            <c8y-typeahead\n              name=\"configType\"\n              [(ngModel)]=\"selectedConfigType\"\n              placeholder=\"{{ 'Type to filter configuration types\u2026' | translate }}\"\n              (onSearch)=\"configType$.next($event)\"\n            >\n              <c8y-li\n                class=\"p-l-8 p-r-8 c8y-list__item--link\"\n                (click)=\"selectedConfigType = { name: '' }; configType$.next('')\"\n              >\n                <span>{{ 'All configuration types' | translate }}</span>\n              </c8y-li>\n              <c8y-li\n                *ngFor=\"let configType of configTypes\"\n                class=\"p-l-8 p-r-8 c8y-list__item--link\"\n                (click)=\"selectedConfigType = configType; configType$.next(configType.name)\"\n                [active]=\"selectedConfigType === configType\"\n              >\n                <c8y-highlight\n                  [text]=\"configType.name\"\n                  [pattern]=\"configType$ | async\"\n                ></c8y-highlight>\n              </c8y-li>\n            </c8y-typeahead>\n          </c8y-form-group>\n        </div>\n      </div>\n      <c8y-li header class=\"hidden-sm hidden-xs m-r-8 m-l-8\">\n        <c8y-li-icon><i class=\"p-l-32 p-r-40\"></i></c8y-li-icon>\n        <c8y-li-body class=\"content-flex-40\">\n          <div class=\"col-5\">\n            <span class=\"text-truncate\" title=\" {{ 'Name' | translate }} \">\n              {{ 'Name' | translate }}\n            </span>\n          </div>\n          <div class=\"col-4\">\n            <span class=\"text-truncate\" title=\"{{ 'Device type' | translate }}\">\n              {{ 'Device type' | translate }}\n            </span>\n          </div>\n          <div class=\"col-3\">\n            <span class=\"text-truncate\" title=\"{{ 'Configuration type' | translate }}\">\n              {{ 'Configuration type' | translate }}\n            </span>\n          </div>\n        </c8y-li-body>\n      </c8y-li>\n      <c8y-list-group\n        list\n        [ngClass]=\"{ 'dd-low': (configurations$ | async)?.data?.length < DD_LOW_COUNT }\"\n      >\n        <c8y-li #listItem *c8yFor=\"let configuration of configurations$ | async\">\n          <c8y-li-radio (onSelect)=\"selectConfiguration(configuration)\"></c8y-li-radio>\n          <c8y-li-icon icon=\"gears\"></c8y-li-icon>\n          <c8y-li-body class=\"content-flex-40\">\n            <div class=\"col-5 m-b-xs-8\">\n              <div class=\"text-truncate\" title=\"{{ configuration.name || '-' }}\">\n                <c8y-highlight\n                  [text]=\"configuration.name || '-'\"\n                  [pattern]=\"textFilter$ | async\"\n                ></c8y-highlight>\n              </div>\n            </div>\n            <div class=\"col-4 m-b-xs-8\">\n              <div\n                class=\"text-truncate\"\n                title=\"{{ 'Device type' | translate }}: {{ getDeviceTypeTitle(configuration) }}\"\n              >\n                <span translate class=\"text-label-small m-r-8 visible-xs visible-sm\"\n                  >Device type</span\n                >\n                {{ configuration.deviceType }}\n                <em class=\"text-muted\" *ngIf=\"!configuration.deviceType\">\n                  {{ 'Undefined`device type`' | translate }}\n                </em>\n              </div>\n            </div>\n            <div class=\"col-3 m-b-xs-8\">\n              <span class=\"text-label-small m-r-8 visible-xs visible-sm\" translate>\n                Configuration type\n              </span>\n              <span\n                class=\"label label-info\"\n                *ngIf=\"configuration.configurationType\"\n                title=\"{{ 'Configuration type' | translate }}: {{\n                  configuration.configurationType\n                }}\"\n              >\n                <c8y-highlight\n                  [text]=\"configuration.configurationType\"\n                  elementClass=\"text-gray-lighter\"\n                  [pattern]=\"configType$ | async\"\n                ></c8y-highlight>\n              </span>\n              <span\n                class=\"label label-default\"\n                *ngIf=\"!configuration.configurationType\"\n                title=\"{{ 'Configuration type' | translate }}: {{ 'Undefined' | translate }}\"\n              >\n                {{ 'Undefined' | translate }}\n              </span>\n            </div>\n          </c8y-li-body>\n        </c8y-li>\n      </c8y-list-group>\n    </c8y-select-step-frame>\n  </ng-container>\n  <ng-container *customStep=\"'Preview configuration' | translate\">\n    <c8y-preview-selection-frame\n      [header]=\"'Preview configuration' | translate\"\n      [name]=\"selectedConfiguration?.name\"\n      [type]=\"selectedConfiguration?.configurationType\"\n    >\n      <ng-container *previewBlock=\"'Device type' | translate\">\n        <ng-container *ngIf=\"selectedConfiguration?.deviceType; else noType\">\n          <span>{{ selectedConfiguration?.deviceType }}</span>\n        </ng-container>\n        <ng-template #noType>\n          <em class=\"text-muted\" translate>Undefined`device type`</em>\n        </ng-template>\n      </ng-container>\n      <ng-container *previewBlock=\"'Updated' | translate\">\n        <span>{{ selectedConfiguration?.lastUpdated | c8yDate }}</span>\n      </ng-container>\n      <ng-container *previewBlock=\"'Preview' | translate\">\n        <div\n          class=\"c8y-empty-state text-left\"\n          *ngIf=\"!this.selectedConfiguration?.binary; else binaryPreview\"\n        >\n          <h1 [c8yIcon]=\"'file-code-o'\"></h1>\n          <p>\n            <strong translate>No preview available.</strong><br />\n            <small translate>Could not fetch the file.</small>\n          </p>\n        </div>\n        <ng-template #binaryPreview>\n          <pre style=\"min-height: 98px\"><code>{{ this.selectedConfiguration?.binary }}</code></pre>\n        </ng-template>\n      </ng-container>\n    </c8y-preview-selection-frame>\n  </ng-container>\n</c8y-bulk-operation-stepper>\n"
            }]
    }], function () { return [{ type: ɵngcc1.BulkOperationsService }, { type: ɵngcc2.RepositoryService }, { type: ɵngcc3.TranslateService }]; }, null); })();

const ɵ0 = {
    type: BulkOperationType.CONFIGURATION,
    c8yIcon: 'cogs',
    name: gettext('Configuration update'),
    path: `${baseUrl}configuration`,
    component: StepperBulkTypeConfigurationComponent,
    fragments: ['c8y_DownloadConfigFile', 'c8y_Configuration'],
    selected: false
};
/** Module for the 'Configuration update' operation type stepper */
class StepperBulkTypeConfigurationModule {
}
StepperBulkTypeConfigurationModule.ɵfac = function StepperBulkTypeConfigurationModule_Factory(t) { return new (t || StepperBulkTypeConfigurationModule)(); };
StepperBulkTypeConfigurationModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: StepperBulkTypeConfigurationModule });
StepperBulkTypeConfigurationModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        {
            provide: HOOK_LIST_BULK_TYPE,
            useValue: ɵ0,
            multi: true
        }
    ], imports: [[
            CoreModule,
            FormsModule,
            ReactiveFormsModule,
            BulkOperationStepperModule,
            SharedRepositoryModule,
            StepperFramesModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(StepperBulkTypeConfigurationModule, [{
        type: NgModule,
        args: [{
                declarations: [StepperBulkTypeConfigurationComponent],
                imports: [
                    CoreModule,
                    FormsModule,
                    ReactiveFormsModule,
                    BulkOperationStepperModule,
                    SharedRepositoryModule,
                    StepperFramesModule
                ],
                providers: [
                    {
                        provide: HOOK_LIST_BULK_TYPE,
                        useValue: ɵ0,
                        multi: true
                    }
                ],
                entryComponents: [StepperBulkTypeConfigurationComponent],
                exports: [StepperBulkTypeConfigurationComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(StepperBulkTypeConfigurationModule, { declarations: function () { return [StepperBulkTypeConfigurationComponent]; }, imports: function () { return [CoreModule,
        FormsModule,
        ReactiveFormsModule,
        BulkOperationStepperModule,
        SharedRepositoryModule,
        StepperFramesModule]; }, exports: function () { return [StepperBulkTypeConfigurationComponent]; } }); })();

/**
 * Generated bundle index. Do not edit.
 */

export { StepperBulkTypeConfigurationComponent, StepperBulkTypeConfigurationModule, ɵ0 };
