// @c8y/ngx-components/fesm2015/c8y-ngx-components-operations-operations-list-item-details.js
import { EventEmitter, Component, Input, Output, ViewChild, NgModule } from '@angular/core';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { gettext, CoreModule } from '@c8y/ngx-components';
import { BulkSingleOperationsListModule } from '@c8y/ngx-components/operations/bulk-single-operations-list';
import { StatusFilterModule } from '@c8y/ngx-components/operations/status-filter';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from '@c8y/ngx-components/operations/bulk-single-operations-list';
import * as ɵngcc4 from 'ngx-bootstrap/tooltip';

const _c0 = ["details"];
const _c1 = ["audit"];
const _c2 = function (a0) { return { active: a0 }; };
function OperationDetailsTabsComponent_li_3_Template(rf, ctx) { if (rf & 1) {
    const _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 8);
    ɵngcc0.ɵɵelementStart(1, "button", 9);
    ɵngcc0.ɵɵlistener("click", function OperationDetailsTabsComponent_li_3_Template_button_click_1_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r10); const tab_r8 = restoredCtx.$implicit; const ctx_r9 = ɵngcc0.ɵɵnextContext(); return ctx_r9.template = tab_r8.template; });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 10);
    ɵngcc0.ɵɵelementStart(4, "span", 11);
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const tab_r8 = ctx.$implicit;
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(8, _c2, tab_r8.template === ctx_r1.template));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(2, 4, tab_r8.label));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", tab_r8.icon);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(6, 6, tab_r8.label));
} }
function OperationDetailsTabsComponent_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function OperationDetailsTabsComponent_ng_template_5_Template(rf, ctx) { if (rf & 1) {
    const _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-operation-details", 12);
    ɵngcc0.ɵɵlistener("showFailedOperation", function OperationDetailsTabsComponent_ng_template_5_Template_c8y_operation_details_showFailedOperation_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r12); const ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.showFailedOperation.emit($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("operation", ctx_r4.operation)("bulkOperationModalDetailsService", ctx_r4.bulkOperationModalDetailsService);
} }
function OperationDetailsTabsComponent_ng_template_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-audit-log", 13);
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("source", ctx_r6.operation.id);
} }
function OperationDetailsTabsComponent_c8y_single_operations_list_9_Template(rf, ctx) { if (rf & 1) {
    const _r14 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-single-operations-list", 14);
    ɵngcc0.ɵɵlistener("onRetryFailedOperations", function OperationDetailsTabsComponent_c8y_single_operations_list_9_Template_c8y_single_operations_list_onRetryFailedOperations_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r14); const ctx_r13 = ɵngcc0.ɵɵnextContext(); return ctx_r13.onRetryFailedOperations.emit($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("bulkOperation", ctx_r7.operation)("readOnly", ctx_r7.readOnly);
} }
function OperationDetailsComponent_li_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 5);
    ɵngcc0.ɵɵelementStart(1, "label", 6);
    ɵngcc0.ɵɵtext(2, "Start date");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "span", 10);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "c8yDate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(5, 1, ctx_r0.operation.startDate));
} }
function OperationDetailsComponent_li_6_Template(rf, ctx) { if (rf & 1) {
    const _r8 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 5);
    ɵngcc0.ɵɵelementStart(1, "label", 11);
    ɵngcc0.ɵɵtext(2, "Bulk operation");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div", 12);
    ɵngcc0.ɵɵelementStart(4, "a", 13);
    ɵngcc0.ɵɵlistener("click", function OperationDetailsComponent_li_6_Template_a_click_4_listener() { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r7 = ɵngcc0.ɵɵnextContext(); return ctx_r7.openBulkOperationDetails(ctx_r7.operation.bulkOperationId); });
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(6, 1, "See details"), " ");
} }
function OperationDetailsComponent_li_7_ng_template_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtext(0);
    ɵngcc0.ɵɵelementStart(1, "span", 20);
    ɵngcc0.ɵɵtext(2, "second");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r9 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r9.operation.creationRamp, " ");
} }
function OperationDetailsComponent_li_7_ng_template_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtext(0);
    ɵngcc0.ɵɵelementStart(1, "span", 20);
    ɵngcc0.ɵɵtext(2, "seconds");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r10 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r10.operation.creationRamp, " ");
} }
function OperationDetailsComponent_li_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 5);
    ɵngcc0.ɵɵelementStart(1, "label", 14);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementStart(4, "a", 15);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelement(6, "i", 16);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "span", 17);
    ɵngcc0.ɵɵtemplate(8, OperationDetailsComponent_li_7_ng_template_8_Template, 3, 1, "ng-template", 18);
    ɵngcc0.ɵɵtemplate(9, OperationDetailsComponent_li_7_ng_template_9_Template, 3, 1, "ng-template", 19);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 4, "Delay"), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("tooltip", ɵngcc0.ɵɵpipeBind1(5, 6, "Delay before sending the operation to the next device."));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", "info-circle");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngPlural", ctx_r2.operation.creationRamp);
} }
function OperationDetailsComponent_li_15_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 5);
    ɵngcc0.ɵɵelementStart(1, "label", 6);
    ɵngcc0.ɵɵtext(2, "Failure reason");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "span", 21);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(4, 2, ctx_r3.operation.failureReason));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(6, 4, ctx_r3.operation.failureReason));
} }
const _c3 = function (a0) { return { bulkOperationsCountSuccessful: a0 }; };
function OperationDetailsComponent_li_16_div_3_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 26);
    ɵngcc0.ɵɵelement(1, "i", 27);
    ɵngcc0.ɵɵelementStart(2, "span", 28);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(3, " {{ bulkOperationsCountSuccessful }} successful ");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r14 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction1(1, _c3, ctx_r14.operation.progress.successful));
} }
const _c4 = function (a0) { return { bulkOperationsCountFailed: a0 }; };
function OperationDetailsComponent_li_16_div_3_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 26);
    ɵngcc0.ɵɵelement(1, "i", 29);
    ɵngcc0.ɵɵelementStart(2, "span", 28);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(3, " {{ bulkOperationsCountFailed }} failed ");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r15 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction1(1, _c4, ctx_r15.operation.progress.failed));
} }
const _c5 = function (a0) { return { bulkOperationsCountExecuting: a0 }; };
function OperationDetailsComponent_li_16_div_3_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 26);
    ɵngcc0.ɵɵelement(1, "i", 30);
    ɵngcc0.ɵɵelementStart(2, "span", 28);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(3, " {{ bulkOperationsCountExecuting }} executing ");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r16 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction1(1, _c5, ctx_r16.operation.progress.executing));
} }
const _c6 = function (a0) { return { bulkOperationsCountPending: a0 }; };
function OperationDetailsComponent_li_16_div_3_div_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 26);
    ɵngcc0.ɵɵelement(1, "i", 31);
    ɵngcc0.ɵɵelementStart(2, "span", 28);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(3, " {{ bulkOperationsCountPending }} pending ");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r17 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction1(1, _c6, ctx_r17.operation.progress.pending));
} }
function OperationDetailsComponent_li_16_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 24);
    ɵngcc0.ɵɵtemplate(1, OperationDetailsComponent_li_16_div_3_div_1_Template, 4, 3, "div", 25);
    ɵngcc0.ɵɵtemplate(2, OperationDetailsComponent_li_16_div_3_div_2_Template, 4, 3, "div", 25);
    ɵngcc0.ɵɵtemplate(3, OperationDetailsComponent_li_16_div_3_div_3_Template, 4, 3, "div", 25);
    ɵngcc0.ɵɵtemplate(4, OperationDetailsComponent_li_16_div_3_div_4_Template, 4, 3, "div", 25);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r11 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r11.operation.progress.successful > 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r11.operation.progress.failed > 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r11.operation.progress.executing > 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r11.operation.progress.pending > 0);
} }
function OperationDetailsComponent_li_16_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 21);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r13 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ctx_r13.operation.result);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r13.operation.result);
} }
function OperationDetailsComponent_li_16_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 5);
    ɵngcc0.ɵɵelementStart(1, "label", 11);
    ɵngcc0.ɵɵtext(2, "Result");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(3, OperationDetailsComponent_li_16_div_3_Template, 5, 4, "div", 22);
    ɵngcc0.ɵɵtemplate(4, OperationDetailsComponent_li_16_ng_template_4_Template, 2, 2, "ng-template", null, 23, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _r12 = ɵngcc0.ɵɵreference(5);
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r4.operation.progress)("ngIfElse", _r12);
} }
function OperationDetailsComponent_li_17_Template(rf, ctx) { if (rf & 1) {
    const _r19 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 5);
    ɵngcc0.ɵɵelementStart(1, "label", 6);
    ɵngcc0.ɵɵtext(2, "Retried from");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "span", 10);
    ɵngcc0.ɵɵelementStart(4, "button", 32);
    ɵngcc0.ɵɵlistener("click", function OperationDetailsComponent_li_17_Template_button_click_4_listener() { ɵngcc0.ɵɵrestoreView(_r19); const ctx_r18 = ɵngcc0.ɵɵnextContext(); return ctx_r18.scrollToRetriedFromOperation(ctx_r18.operation.failedParentId); });
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(5, 2, "Scroll to parent operation."));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r5.operation.failedParentId, " ");
} }
function OperationDetailsComponent_li_18_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 5);
    ɵngcc0.ɵɵelementStart(1, "label", 11);
    ɵngcc0.ɵɵtext(2, "Description");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "span", 21);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ctx_r6.operation.note ? ctx_r6.operation.note : ɵngcc0.ɵɵpipeBind1(4, 2, ctx_r6.operation.description));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r6.operation.note ? ctx_r6.operation.note : ɵngcc0.ɵɵpipeBind1(6, 4, ctx_r6.operation.description), " ");
} }
class OperationDetailsTabsComponent {
    constructor() {
        this.readOnly = false;
        this.onRetryFailedOperations = new EventEmitter();
        this.showFailedOperation = new EventEmitter();
        this.tabs = [];
    }
    ngOnInit() {
        this.template = this.detailsTemplate;
        this.tabs = [
            {
                label: gettext('Details'),
                path: null,
                icon: 'asterisk',
                template: this.detailsTemplate
            },
            {
                label: gettext('History of changes'),
                path: null,
                icon: 'archive',
                template: this.auditTemplate
            }
        ];
    }
}
OperationDetailsTabsComponent.ɵfac = function OperationDetailsTabsComponent_Factory(t) { return new (t || OperationDetailsTabsComponent)(); };
OperationDetailsTabsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: OperationDetailsTabsComponent, selectors: [["c8y-operation-details-tabs"]], viewQuery: function OperationDetailsTabsComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 7);
        ɵngcc0.ɵɵviewQuery(_c1, 7);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.detailsTemplate = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.auditTemplate = _t.first);
    } }, inputs: { readOnly: "readOnly", operation: "operation", bulkOperationModalDetailsService: "bulkOperationModalDetailsService" }, outputs: { onRetryFailedOperations: "onRetryFailedOperations", showFailedOperation: "showFailedOperation" }, decls: 10, vars: 3, consts: [[1, "tabContainer"], ["container", ""], [1, "nav", "nav-tabs", "nav-tabsc8y"], [3, "ngClass", 4, "ngFor", "ngForOf"], [4, "ngTemplateOutlet"], ["details", ""], ["audit", ""], ["class", "d-flex d-col", 3, "bulkOperation", "readOnly", "onRetryFailedOperations", 4, "ngIf"], [3, "ngClass"], [1, "btn", 3, "title", "click"], [3, "c8yIcon"], [1, "txt"], [3, "operation", "bulkOperationModalDetailsService", "showFailedOperation"], [3, "source"], [1, "d-flex", "d-col", 3, "bulkOperation", "readOnly", "onRetryFailedOperations"]], template: function OperationDetailsTabsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0, 1);
        ɵngcc0.ɵɵelementStart(2, "ul", 2);
        ɵngcc0.ɵɵtemplate(3, OperationDetailsTabsComponent_li_3_Template, 7, 10, "li", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(4, OperationDetailsTabsComponent_ng_container_4_Template, 1, 0, "ng-container", 4);
        ɵngcc0.ɵɵtemplate(5, OperationDetailsTabsComponent_ng_template_5_Template, 1, 2, "ng-template", null, 5, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(7, OperationDetailsTabsComponent_ng_template_7_Template, 1, 1, "ng-template", null, 6, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(9, OperationDetailsTabsComponent_c8y_single_operations_list_9_Template, 1, 2, "c8y-single-operations-list", 7);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.tabs);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.template);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("ngIf", ctx.operation.operationPrototype);
    } }, directives: function () { return [ɵngcc1.NgForOf, ɵngcc1.NgTemplateOutlet, ɵngcc1.NgIf, ɵngcc1.NgClass, ɵngcc2.IconDirective, OperationDetailsComponent, ɵngcc2.ɵbd, ɵngcc3.SingleOperationsListComponent]; }, pipes: function () { return [ɵngcc2.C8yTranslatePipe]; }, encapsulation: 2 });
OperationDetailsTabsComponent.propDecorators = {
    operation: [{ type: Input }],
    readOnly: [{ type: Input }],
    bulkOperationModalDetailsService: [{ type: Input }],
    onRetryFailedOperations: [{ type: Output }],
    showFailedOperation: [{ type: Output }],
    detailsTemplate: [{ type: ViewChild, args: ['details', { static: true },] }],
    auditTemplate: [{ type: ViewChild, args: ['audit', { static: true },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OperationDetailsTabsComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-operation-details-tabs',
                template: "<div #container class=\"tabContainer\">\n  <ul class=\"nav nav-tabs nav-tabsc8y\">\n    <li [ngClass]=\"{ active: tab.template === template }\" *ngFor=\"let tab of tabs\">\n      <button [title]=\"tab.label | translate\" class=\"btn\" (click)=\"template = tab.template\">\n        <i [c8yIcon]=\"tab.icon\"></i>\n        <span class=\"txt\">{{ tab.label | translate }}</span>\n      </button>\n    </li>\n  </ul>\n</div>\n\n<ng-container *ngTemplateOutlet=\"template\"></ng-container>\n\n<ng-template #details>\n  <c8y-operation-details\n    [operation]=\"operation\"\n    (showFailedOperation)=\"showFailedOperation.emit($event)\"\n    [bulkOperationModalDetailsService]=\"bulkOperationModalDetailsService\"\n  >\n  </c8y-operation-details>\n</ng-template>\n\n<ng-template #audit>\n  <c8y-audit-log [source]=\"operation.id\"></c8y-audit-log>\n</ng-template>\n\n<c8y-single-operations-list\n  *ngIf=\"operation.operationPrototype\"\n  [bulkOperation]=\"operation\"\n  [readOnly]=\"readOnly\"\n  (onRetryFailedOperations)=\"onRetryFailedOperations.emit($event)\"\n  class=\"d-flex d-col\"\n></c8y-single-operations-list>\n"
            }]
    }], function () { return []; }, { readOnly: [{
            type: Input
        }], onRetryFailedOperations: [{
            type: Output
        }], showFailedOperation: [{
            type: Output
        }], operation: [{
            type: Input
        }], bulkOperationModalDetailsService: [{
            type: Input
        }], detailsTemplate: [{
            type: ViewChild,
            args: ['details', { static: true }]
        }], auditTemplate: [{
            type: ViewChild,
            args: ['audit', { static: true }]
        }] }); })();

class OperationDetailsComponent {
    constructor() {
        this.showFailedOperation = new EventEmitter();
    }
    scrollToRetriedFromOperation(failedOperationId) {
        this.showFailedOperation.emit(failedOperationId);
    }
    openBulkOperationDetails(bulkOperationId) {
        this.bulkOperationModalDetailsService.openDetails(bulkOperationId);
    }
}
OperationDetailsComponent.ɵfac = function OperationDetailsComponent_Factory(t) { return new (t || OperationDetailsComponent)(); };
OperationDetailsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: OperationDetailsComponent, selectors: [["c8y-operation-details"]], inputs: { operation: "operation", bulkOperationModalDetailsService: "bulkOperationModalDetailsService" }, outputs: { showFailedOperation: "showFailedOperation" }, decls: 25, vars: 15, consts: [[1, "row", "d-flex-md", "fit-h"], [1, "col-md-4", "flex-item-top"], ["translate", "", 1, "legend", "form-block"], [1, "list-unstyled", "small"], ["class", "p-t-4 p-b-4 flex-row separator-bottom", 4, "ngIf"], [1, "p-t-4", "p-b-4", "flex-row", "separator-bottom"], ["translate", "", 1, "small", "m-0"], [1, "flex-item-right", "text-uppercase"], [1, "col-md-8", "fit-h", "d-flex", "d-col"], [1, "inner-scroll", "p-16", "m-b-0", "text-pre-wrap"], [1, "flex-item-right"], ["translate", "", 1, "small", "flex-item-top"], [1, "flex-item-right", "text-right", "text-truncate"], [1, "interact", 3, "click"], [1, "small", "m-0"], ["container", "body", 1, "text-primary", 3, "tooltip"], [3, "c8yIcon"], [1, "flex-item-right", 3, "ngPlural"], ["ngPluralCase", "=1"], ["ngPluralCase", "other"], ["translate", ""], [1, "inner-scroll", "p-l-16", "m-b-0", "flex-item-right", "text-truncate-wrap", 3, "title"], ["class", "flex-row flex-grow flex-item-right flex-end", 4, "ngIf", "ngIfElse"], ["singleOpResult", ""], [1, "flex-row", "flex-grow", "flex-item-right", "flex-end"], ["class", "icon-flex flex-item-left m-r-8", 4, "ngIf"], [1, "icon-flex", "flex-item-left", "m-r-8"], ["c8yIcon", "check-circle", 1, "text-success", "m-r-4"], ["translate", "", 3, "translateParams"], ["c8yIcon", "warning", 1, "text-danger", "m-r-4"], ["c8yIcon", "refresh", 1, "m-r-4", "text-info"], ["c8yIcon", "clock-o", 1, "m-r-4", "text-info"], [1, "btn", "btn-default", "btn-xs", 3, "title", "click"]], template: function OperationDetailsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵtext(3, "Details");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "ul", 3);
        ɵngcc0.ɵɵtemplate(5, OperationDetailsComponent_li_5_Template, 6, 3, "li", 4);
        ɵngcc0.ɵɵtemplate(6, OperationDetailsComponent_li_6_Template, 7, 3, "li", 4);
        ɵngcc0.ɵɵtemplate(7, OperationDetailsComponent_li_7_Template, 10, 8, "li", 4);
        ɵngcc0.ɵɵelementStart(8, "li", 5);
        ɵngcc0.ɵɵelementStart(9, "label", 6);
        ɵngcc0.ɵɵtext(10, "Status");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "span", 7);
        ɵngcc0.ɵɵtext(12);
        ɵngcc0.ɵɵpipe(13, "translate");
        ɵngcc0.ɵɵpipe(14, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(15, OperationDetailsComponent_li_15_Template, 7, 6, "li", 4);
        ɵngcc0.ɵɵtemplate(16, OperationDetailsComponent_li_16_Template, 6, 2, "li", 4);
        ɵngcc0.ɵɵtemplate(17, OperationDetailsComponent_li_17_Template, 7, 4, "li", 4);
        ɵngcc0.ɵɵtemplate(18, OperationDetailsComponent_li_18_Template, 7, 6, "li", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(19, "div", 8);
        ɵngcc0.ɵɵelementStart(20, "div", 2);
        ɵngcc0.ɵɵtext(21, "Operation");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(22, "pre", 9);
        ɵngcc0.ɵɵtext(23);
        ɵngcc0.ɵɵpipe(24, "json");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("ngIf", ctx.operation.startDate);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.operation.bulkOperationId);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.operation.creationRamp);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵtextInterpolate1(" ", ctx.operation.operationPrototype ? ɵngcc0.ɵɵpipeBind1(13, 9, ctx.operation.generalStatus.toString()) : ɵngcc0.ɵɵpipeBind1(14, 11, ctx.operation.status.toString()), " ");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.operation.failureReason);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.operation.result || ctx.operation.progress);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.operation.failedParentId);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.operation.note || ctx.operation.description);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(24, 13, ctx.operation.operationPrototype || ctx.operation));
    } }, directives: [ɵngcc2.C8yTranslateDirective, ɵngcc1.NgIf, ɵngcc4.TooltipDirective, ɵngcc2.IconDirective, ɵngcc1.NgPlural, ɵngcc1.NgPluralCase], pipes: [ɵngcc2.C8yTranslatePipe, ɵngcc1.JsonPipe, ɵngcc2.DatePipe], encapsulation: 2 });
OperationDetailsComponent.propDecorators = {
    operation: [{ type: Input }],
    bulkOperationModalDetailsService: [{ type: Input }],
    showFailedOperation: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OperationDetailsComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-operation-details',
                template: "<div class=\"row d-flex-md fit-h\">\n  <div class=\"col-md-4 flex-item-top\">\n    <div class=\"legend form-block\" translate>Details</div>\n    <ul class=\"list-unstyled small\">\n      <li class=\"p-t-4 p-b-4 flex-row separator-bottom\" *ngIf=\"operation.startDate\">\n        <label class=\"small m-0\" translate>Start date</label>\n        <span class=\"flex-item-right\">{{ operation.startDate | c8yDate }}</span>\n      </li>\n      <li class=\"p-t-4 p-b-4 flex-row separator-bottom\" *ngIf=\"operation.bulkOperationId\">\n        <label class=\"small flex-item-top\" translate>Bulk operation</label>\n        <div class=\"flex-item-right text-right text-truncate\">\n          <a (click)=\"openBulkOperationDetails(operation.bulkOperationId)\" class=\"interact\">\n            {{ 'See details' | translate }}\n          </a>\n        </div>\n      </li>\n      <li class=\"p-t-4 p-b-4 flex-row separator-bottom\" *ngIf=\"operation.creationRamp\">\n        <label class=\"small m-0\">\n          {{ 'Delay' | translate }}\n          <a\n            class=\"text-primary\"\n            tooltip=\"{{ 'Delay before sending the operation to the next device.' | translate }}\"\n            container=\"body\"\n          >\n            <i [c8yIcon]=\"'info-circle'\"></i>\n          </a>\n        </label>\n        <span [ngPlural]=\"operation.creationRamp\" class=\"flex-item-right\">\n          <ng-template ngPluralCase=\"=1\">\n            {{ operation.creationRamp }} <span translate>second</span>\n          </ng-template>\n          <ng-template ngPluralCase=\"other\">\n            {{ operation.creationRamp }} <span translate>seconds</span>\n          </ng-template>\n        </span>\n      </li>\n      <li class=\"p-t-4 p-b-4 flex-row separator-bottom\">\n        <label class=\"small m-0\" translate>Status</label>\n        <span class=\"flex-item-right text-uppercase\">\n          {{\n            operation.operationPrototype\n              ? (operation.generalStatus.toString() | translate)\n              : (operation.status.toString() | translate)\n          }}\n        </span>\n      </li>\n      <li class=\"p-t-4 p-b-4 flex-row separator-bottom\" *ngIf=\"operation.failureReason\">\n        <label class=\"small m-0\" translate>Failure reason</label>\n        <span\n          class=\"inner-scroll p-l-16 m-b-0 flex-item-right text-truncate-wrap\"\n          title=\"{{ operation.failureReason | translate }}\"\n          >{{ operation.failureReason | translate }}</span\n        >\n      </li>\n      <li\n        class=\"p-t-4 p-b-4 flex-row separator-bottom\"\n        *ngIf=\"operation.result || operation.progress\"\n      >\n        <label class=\"small flex-item-top\" translate>Result</label>\n        <div\n          class=\"flex-row flex-grow flex-item-right flex-end\"\n          *ngIf=\"operation.progress; else singleOpResult\"\n        >\n          <div *ngIf=\"operation.progress.successful > 0\" class=\"icon-flex flex-item-left m-r-8\">\n            <i c8yIcon=\"check-circle\" class=\"text-success m-r-4\"></i>\n            <span\n              ngNonBindable\n              translate\n              [translateParams]=\"{\n                bulkOperationsCountSuccessful: operation.progress.successful\n              }\"\n            >\n              {{ bulkOperationsCountSuccessful }} successful\n            </span>\n          </div>\n          <div class=\"icon-flex flex-item-left m-r-8\" *ngIf=\"operation.progress.failed > 0\">\n            <i c8yIcon=\"warning\" class=\"text-danger m-r-4\"></i>\n            <span\n              ngNonBindable\n              translate\n              [translateParams]=\"{ bulkOperationsCountFailed: operation.progress.failed }\"\n            >\n              {{ bulkOperationsCountFailed }} failed\n            </span>\n          </div>\n          <div class=\"icon-flex flex-item-left m-r-8\" *ngIf=\"operation.progress.executing > 0\">\n            <i c8yIcon=\"refresh\" class=\"m-r-4 text-info\"></i>\n            <span\n              ngNonBindable\n              translate\n              [translateParams]=\"{ bulkOperationsCountExecuting: operation.progress.executing }\"\n            >\n              {{ bulkOperationsCountExecuting }} executing\n            </span>\n          </div>\n          <div class=\"icon-flex flex-item-left m-r-8\" *ngIf=\"operation.progress.pending > 0\">\n            <i c8yIcon=\"clock-o\" class=\"m-r-4 text-info\"></i>\n            <span\n              ngNonBindable\n              translate\n              [translateParams]=\"{ bulkOperationsCountPending: operation.progress.pending }\"\n            >\n              {{ bulkOperationsCountPending }} pending\n            </span>\n          </div>\n        </div>\n        <ng-template #singleOpResult>\n          <span\n            class=\"inner-scroll p-l-16 m-b-0 flex-item-right text-truncate-wrap\"\n            title=\"{{ operation.result }}\"\n            >{{ operation.result }}</span\n          >\n        </ng-template>\n      </li>\n      <li class=\"p-t-4 p-b-4 flex-row separator-bottom\" *ngIf=\"operation.failedParentId\">\n        <label class=\"small m-0\" translate>Retried from</label>\n        <span class=\"flex-item-right\">\n          <button\n            class=\"btn btn-default btn-xs\"\n            (click)=\"scrollToRetriedFromOperation(operation.failedParentId)\"\n            title=\"{{ 'Scroll to parent operation.' | translate }}\"\n          >\n            {{ operation.failedParentId }}\n          </button>\n        </span>\n      </li>\n      <li\n        class=\"p-t-4 p-b-4 flex-row separator-bottom\"\n        *ngIf=\"operation.note || operation.description\"\n      >\n        <label class=\"small flex-item-top\" translate>Description</label>\n        <span\n          title=\"{{ operation.note ? operation.note : (operation.description | translate) }}\"\n          class=\"inner-scroll p-l-16 m-b-0 flex-item-right text-truncate-wrap\"\n        >\n          {{ operation.note ? operation.note : (operation.description | translate) }}\n        </span>\n      </li>\n    </ul>\n  </div>\n  <div class=\"col-md-8 fit-h d-flex d-col\">\n    <div class=\"legend form-block\" translate>Operation</div>\n    <pre class=\"inner-scroll p-16 m-b-0 text-pre-wrap\">{{\n      operation.operationPrototype || operation | json\n    }}</pre>\n  </div>\n</div>\n"
            }]
    }], function () { return []; }, { showFailedOperation: [{
            type: Output
        }], operation: [{
            type: Input
        }], bulkOperationModalDetailsService: [{
            type: Input
        }] }); })();

/**
 * This module allows for displaying the list of single operations and their details.
 */
class OperationsListItemDetailsModule {
}
OperationsListItemDetailsModule.ɵfac = function OperationsListItemDetailsModule_Factory(t) { return new (t || OperationsListItemDetailsModule)(); };
OperationsListItemDetailsModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: OperationsListItemDetailsModule });
OperationsListItemDetailsModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CoreModule, TooltipModule, StatusFilterModule, BulkSingleOperationsListModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OperationsListItemDetailsModule, [{
        type: NgModule,
        args: [{
                imports: [CoreModule, TooltipModule, StatusFilterModule, BulkSingleOperationsListModule],
                declarations: [OperationDetailsComponent, OperationDetailsTabsComponent],
                exports: [OperationDetailsComponent, OperationDetailsTabsComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(OperationsListItemDetailsModule, { declarations: function () { return [OperationDetailsComponent, OperationDetailsTabsComponent]; }, imports: function () { return [CoreModule, TooltipModule, StatusFilterModule, BulkSingleOperationsListModule]; }, exports: function () { return [OperationDetailsComponent, OperationDetailsTabsComponent]; } }); })();

/**
 * Generated bundle index. Do not edit.
 */

export { OperationDetailsComponent, OperationDetailsTabsComponent, OperationsListItemDetailsModule };
