// @c8y/ngx-components/fesm2015/c8y-ngx-components-operations-stepper-frames.js
import { Directive, TemplateRef, Input, Component, ViewChild, ContentChildren, NgModule } from '@angular/core';
import { CoreModule } from '@c8y/ngx-components';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '@c8y/ngx-components';

const _c0 = ["list"];
function PreviewSelectionFrameComponent_p_6_strong_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "strong", 12);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("title", ctx_r4.name);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r4.name);
} }
function PreviewSelectionFrameComponent_p_6_small_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 12);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("title", ctx_r5.description);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r5.description);
} }
function PreviewSelectionFrameComponent_p_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p", 10);
    ɵngcc0.ɵɵtemplate(1, PreviewSelectionFrameComponent_p_6_strong_1_Template, 2, 2, "strong", 11);
    ɵngcc0.ɵɵelement(2, "br");
    ɵngcc0.ɵɵtemplate(3, PreviewSelectionFrameComponent_p_6_small_3_Template, 2, 2, "small", 11);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.name);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.description);
} }
function PreviewSelectionFrameComponent_p_7_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 16);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r6.type, " ");
} }
function PreviewSelectionFrameComponent_p_7_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 17);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Undefined"), " ");
} }
function PreviewSelectionFrameComponent_p_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p", 13);
    ɵngcc0.ɵɵtemplate(1, PreviewSelectionFrameComponent_p_7_span_1_Template, 2, 1, "span", 14);
    ɵngcc0.ɵɵtemplate(2, PreviewSelectionFrameComponent_p_7_ng_template_2_Template, 3, 3, "ng-template", null, 15, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _r7 = ɵngcc0.ɵɵreference(3);
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.type)("ngIfElse", _r7);
} }
function PreviewSelectionFrameComponent_div_8_ng_container_3_div_1_ng_container_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function PreviewSelectionFrameComponent_div_8_ng_container_3_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 21);
    ɵngcc0.ɵɵelementStart(1, "div", 22);
    ɵngcc0.ɵɵelementStart(2, "div", 23);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementStart(4, "span", 24);
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(7, PreviewSelectionFrameComponent_div_8_ng_container_3_div_1_ng_container_7_Template, 1, 0, "ng-container", 25);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const block_r10 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(3, 3, block_r10.label));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(6, 5, block_r10.label));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", block_r10.templateRef);
} }
function PreviewSelectionFrameComponent_div_8_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, PreviewSelectionFrameComponent_div_8_ng_container_3_div_1_Template, 8, 7, "div", 20);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const block_r10 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", block_r10.show);
} }
function PreviewSelectionFrameComponent_div_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 6);
    ɵngcc0.ɵɵelementStart(1, "div", 7);
    ɵngcc0.ɵɵelementStart(2, "div", 18);
    ɵngcc0.ɵɵtemplate(3, PreviewSelectionFrameComponent_div_8_ng_container_3_Template, 2, 1, "ng-container", 19);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r2.previewBlocks);
} }
const _c1 = [[["", "list", ""]]];
const _c2 = ["[list]"];
function SelectStepFrameComponent_c8y_search_result_empty_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-search-result-empty");
} }
const _c3 = [[["", "filters", ""]], [["", "header", ""]], [["", "list", ""]]];
const _c4 = ["[filters]", "[header]", "[list]"];
class PreviewBlockDirective {
    constructor(templateRef) {
        this.templateRef = templateRef;
        this.show = true;
    }
}
PreviewBlockDirective.ɵfac = function PreviewBlockDirective_Factory(t) { return new (t || PreviewBlockDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.TemplateRef)); };
PreviewBlockDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: PreviewBlockDirective, selectors: [["", "previewBlock", ""]], inputs: { show: ["previewBlockIf", "show"], label: ["previewBlock", "label"] } });
PreviewBlockDirective.ctorParameters = () => [
    { type: TemplateRef }
];
PreviewBlockDirective.propDecorators = {
    label: [{ type: Input, args: ['previewBlock',] }],
    show: [{ type: Input, args: ['previewBlockIf',] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(PreviewBlockDirective, [{
        type: Directive,
        args: [{
                selector: '[previewBlock]'
            }]
    }], function () { return [{ type: ɵngcc0.TemplateRef }]; }, { show: [{
            type: Input,
            args: ['previewBlockIf']
        }], label: [{
            type: Input,
            args: ['previewBlock']
        }] }); })();

class PreviewSelectionFrameComponent {
    constructor() {
        this.showType = false;
    }
    set _type(type) {
        this.type = type;
        this.showType = true;
    }
}
PreviewSelectionFrameComponent.ɵfac = function PreviewSelectionFrameComponent_Factory(t) { return new (t || PreviewSelectionFrameComponent)(); };
PreviewSelectionFrameComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: PreviewSelectionFrameComponent, selectors: [["c8y-preview-selection-frame"]], contentQueries: function PreviewSelectionFrameComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, PreviewBlockDirective, 4);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.previewBlocks = _t);
    } }, viewQuery: function PreviewSelectionFrameComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.listRef = _t.first);
    } }, inputs: { _type: ["type", "_type"], header: "header", name: "name", description: "description" }, ngContentSelectors: _c2, decls: 15, vars: 8, consts: [[1, "card-block", "p-t-0", "flex-no-shrink", "separator-bottom", "col-xs-12"], [1, "flex-row", "p-b-8", "p-t-4", "j-c-center"], [1, "text-center", "text-normal"], ["class", "text-center p-t-16 p-b-8 text-truncate", 4, "ngIf"], ["class", "text-center", 4, "ngIf"], ["class", "col-xs-12 flex-grow no-gutter", 4, "ngIf"], [1, "col-xs-12", "flex-grow", "no-gutter"], [1, "card-inner-scroll", "fit-h"], [1, "card-block", "p-t-0", "p-b-0"], ["list", ""], [1, "text-center", "p-t-16", "p-b-8", "text-truncate"], [3, "title", 4, "ngIf"], [3, "title"], [1, "text-center"], ["class", "label label-primary", 4, "ngIf", "ngIfElse"], ["noType", ""], [1, "label", "label-primary"], [1, "label", "label-default"], [1, "card-block"], [4, "ngFor", "ngForOf"], ["class", "flex-row p-b-8 j-c-center", 4, "ngIf"], [1, "flex-row", "p-b-8", "j-c-center"], [1, "col-xs-12", "col-sm-6"], [1, "text-truncate", 3, "title"], [1, "legend", "form-block"], [4, "ngTemplateOutlet"]], template: function PreviewSelectionFrameComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c1);
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div");
        ɵngcc0.ɵɵelementStart(3, "h4", 2);
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(6, PreviewSelectionFrameComponent_p_6_Template, 4, 2, "p", 3);
        ɵngcc0.ɵɵtemplate(7, PreviewSelectionFrameComponent_p_7_Template, 4, 2, "p", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(8, PreviewSelectionFrameComponent_div_8_Template, 4, 1, "div", 5);
        ɵngcc0.ɵɵelementStart(9, "div");
        ɵngcc0.ɵɵelementStart(10, "div", 6);
        ɵngcc0.ɵɵelementStart(11, "div", 7);
        ɵngcc0.ɵɵelementStart(12, "div", 8, 9);
        ɵngcc0.ɵɵprojection(14);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(5, 6, ctx.header));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.name || ctx.description);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showType);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.previewBlocks == null ? null : ctx.previewBlocks.length);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵstyleProp("display", (ctx.listRef == null ? null : ctx.listRef.nativeElement == null ? null : ctx.listRef.nativeElement.children == null ? null : ctx.listRef.nativeElement.children.length) ? "contents" : "none");
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgForOf, ɵngcc1.NgTemplateOutlet], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
PreviewSelectionFrameComponent.propDecorators = {
    header: [{ type: Input }],
    name: [{ type: Input }],
    description: [{ type: Input }],
    _type: [{ type: Input, args: ['type',] }],
    listRef: [{ type: ViewChild, args: ['list',] }],
    previewBlocks: [{ type: ContentChildren, args: [PreviewBlockDirective,] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(PreviewSelectionFrameComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-preview-selection-frame',
                template: "<div class=\"card-block p-t-0 flex-no-shrink separator-bottom col-xs-12\">\n  <div class=\"flex-row p-b-8 p-t-4 j-c-center\">\n    <div>\n      <h4 class=\"text-center text-normal\">{{ header | translate }}</h4>\n      <p *ngIf=\"name || description\" class=\"text-center p-t-16 p-b-8 text-truncate\">\n        <strong *ngIf=\"name\" [title]=\"name\">{{ name }}</strong\n        ><br />\n        <small *ngIf=\"description\" [title]=\"description\">{{ description }}</small>\n      </p>\n      <p class=\"text-center\" *ngIf=\"showType\">\n        <span class=\"label label-primary\" *ngIf=\"type; else noType\">\n          {{ type }}\n        </span>\n        <ng-template #noType>\n          <span class=\"label label-default\">\n            {{ 'Undefined' | translate }}\n          </span>\n        </ng-template>\n      </p>\n    </div>\n  </div>\n</div>\n<div *ngIf=\"previewBlocks?.length\" class=\"col-xs-12 flex-grow no-gutter\">\n  <div class=\"card-inner-scroll fit-h\">\n    <div class=\"card-block\">\n      <ng-container *ngFor=\"let block of previewBlocks\">\n        <div *ngIf=\"block.show\" class=\"flex-row p-b-8 j-c-center\">\n          <div class=\"col-xs-12 col-sm-6\">\n            <div class=\"text-truncate\" [title]=\"block.label | translate\">\n              <span class=\"legend form-block\">{{ block.label | translate }}</span>\n              <ng-container *ngTemplateOutlet=\"block.templateRef\"></ng-container>\n            </div>\n          </div>\n        </div>\n      </ng-container>\n    </div>\n  </div>\n</div>\n<div [style.display]=\"listRef?.nativeElement?.children?.length ? 'contents' : 'none'\">\n  <div class=\"col-xs-12 flex-grow no-gutter\">\n    <div class=\"card-inner-scroll fit-h\">\n      <div #list class=\"card-block p-t-0 p-b-0\">\n        <ng-content select=\"[list]\"> </ng-content>\n      </div>\n    </div>\n  </div>\n</div>\n"
            }]
    }], function () { return []; }, { _type: [{
            type: Input,
            args: ['type']
        }], header: [{
            type: Input
        }], name: [{
            type: Input
        }], description: [{
            type: Input
        }], listRef: [{
            type: ViewChild,
            args: ['list']
        }], previewBlocks: [{
            type: ContentChildren,
            args: [PreviewBlockDirective]
        }] }); })();

class SelectStepFrameComponent {
}
SelectStepFrameComponent.ɵfac = function SelectStepFrameComponent_Factory(t) { return new (t || SelectStepFrameComponent)(); };
SelectStepFrameComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SelectStepFrameComponent, selectors: [["c8y-select-step-frame"]], inputs: { header: "header", noResults: "noResults" }, ngContentSelectors: _c4, decls: 13, vars: 4, consts: [[1, "card-block", "overflow-visible", "flex-no-shrink", "col-xs-12", "p-b-0", "p-t-4", "separator-bottom"], [1, "flex-row", "j-c-center", "m-b-24"], [1, "col-xs-12", "col-sm-8"], [1, "text-normal", "text-center", "m-b-16"], [4, "ngIf"], [1, "col-xs-12", "flex-grow", "no-gutter"], [1, "card-inner-scroll", "fit-h"], [1, "card-block", "p-t-0", "p-b-0"]], template: function SelectStepFrameComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c3);
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelementStart(3, "h4", 3);
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵprojection(6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵprojection(7, 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(8, SelectStepFrameComponent_c8y_search_result_empty_8_Template, 1, 0, "c8y-search-result-empty", 4);
        ɵngcc0.ɵɵelementStart(9, "div", 5);
        ɵngcc0.ɵɵelementStart(10, "div", 6);
        ɵngcc0.ɵɵelementStart(11, "div", 7);
        ɵngcc0.ɵɵprojection(12, 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 2, ctx.header), " ");
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngIf", ctx.noResults);
    } }, directives: [ɵngcc1.NgIf, ɵngcc2.SearchResultEmptyComponent], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
SelectStepFrameComponent.propDecorators = {
    header: [{ type: Input }],
    noResults: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SelectStepFrameComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-select-step-frame',
                template: "<div class=\"card-block overflow-visible flex-no-shrink col-xs-12 p-b-0 p-t-4 separator-bottom\">\n  <div class=\"flex-row j-c-center m-b-24\">\n    <div class=\"col-xs-12 col-sm-8\">\n      <h4 class=\"text-normal text-center m-b-16\">\n        {{ header | translate }}\n      </h4>\n      <ng-content select=\"[filters]\"></ng-content>\n    </div>\n  </div>\n  <ng-content select=\"[header]\"></ng-content>\n</div>\n<c8y-search-result-empty *ngIf=\"noResults\"> </c8y-search-result-empty>\n<div class=\"col-xs-12 flex-grow no-gutter\">\n  <div class=\"card-inner-scroll fit-h\">\n    <div class=\"card-block p-t-0 p-b-0\">\n      <ng-content select=\"[list]\"></ng-content>\n    </div>\n  </div>\n</div>\n"
            }]
    }], null, { header: [{
            type: Input
        }], noResults: [{
            type: Input
        }] }); })();

class StepperFramesModule {
}
StepperFramesModule.ɵfac = function StepperFramesModule_Factory(t) { return new (t || StepperFramesModule)(); };
StepperFramesModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: StepperFramesModule });
StepperFramesModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CoreModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(StepperFramesModule, [{
        type: NgModule,
        args: [{
                imports: [CoreModule],
                declarations: [PreviewSelectionFrameComponent, SelectStepFrameComponent, PreviewBlockDirective],
                exports: [PreviewSelectionFrameComponent, SelectStepFrameComponent, PreviewBlockDirective]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(StepperFramesModule, { declarations: function () { return [PreviewSelectionFrameComponent, SelectStepFrameComponent, PreviewBlockDirective]; }, imports: function () { return [CoreModule]; }, exports: function () { return [PreviewSelectionFrameComponent, SelectStepFrameComponent, PreviewBlockDirective]; } }); })();

/**
 * Generated bundle index. Do not edit.
 */

export { PreviewBlockDirective, PreviewSelectionFrameComponent, SelectStepFrameComponent, StepperFramesModule };
