// @c8y/ngx-components/fesm2015/c8y-ngx-components-sigfox-device-registration.js
import * as i0 from '@angular/core';
import { Injectable, Component, NgModule } from '@angular/core';
import { BsModalRef, BsModalService } from 'ngx-bootstrap/modal';
import { __awaiter } from 'tslib';
import { gettext, TenantUiService, CoreModule, CommonModule } from '@c8y/ngx-components';
import { FormGroup } from '@angular/forms';
import { Subject, of, forkJoin, throwError, BehaviorSubject, defer, from } from 'rxjs';
import { FetchClient, InventoryService, TenantOptionsService, ApplicationService } from '@c8y/client';
import * as i4 from '@ngx-translate/core';
import { TranslateService } from '@ngx-translate/core';
import * as i1 from '@c8y/client';
import * as i2 from '@c8y/client';
import * as i3 from '@c8y/client';
import * as i5 from '@c8y/client';
import { catchError, switchMap, map, takeUntil, mergeMap, shareReplay } from 'rxjs/operators';
import { uniq, cloneDeep } from 'lodash-es';
import { HOOK_DEVICE_REGISTRATION } from '@c8y/ngx-components/register-device';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/client';
import * as ɵngcc2 from '@ngx-translate/core';
import * as ɵngcc3 from 'ngx-bootstrap/modal';
import * as ɵngcc4 from '@c8y/ngx-components';
import * as ɵngcc5 from '@angular/common';
import * as ɵngcc6 from '@angular/cdk/stepper';
import * as ɵngcc7 from '@ngx-formly/core';

function SigfoxDeviceRegistrationComponent_ng_container_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "div", 6);
    ɵngcc0.ɵɵelement(2, "c8y-loading");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} }
function SigfoxDeviceRegistrationComponent_ng_template_7_c8y_stepper_0_div_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 6);
    ɵngcc0.ɵɵelement(1, "c8y-loading");
    ɵngcc0.ɵɵelementEnd();
} }
function SigfoxDeviceRegistrationComponent_ng_template_7_c8y_stepper_0_c8y_operation_result_11_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-operation-result", 19);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("text", ɵngcc0.ɵɵpipeBind1(1, 3, "Device registered"));
    ɵngcc0.ɵɵproperty("size", 84)("vertical", true);
} }
const _c0 = function () { return { cancel: true, next: true }; };
const _c1 = function () { return { custom: true }; };
function SigfoxDeviceRegistrationComponent_ng_template_7_c8y_stepper_0_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-stepper", 8);
    ɵngcc0.ɵɵelementStart(1, "cdk-step", 9);
    ɵngcc0.ɵɵelementStart(2, "div", 10);
    ɵngcc0.ɵɵelementStart(3, "p", 11);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(6, "formly-form", 12);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "c8y-stepper-buttons", 13);
    ɵngcc0.ɵɵlistener("onNext", function SigfoxDeviceRegistrationComponent_ng_template_7_c8y_stepper_0_Template_c8y_stepper_buttons_onNext_7_listener($event) { ɵngcc0.ɵɵrestoreView(_r9); const ctx_r8 = ɵngcc0.ɵɵnextContext(2); return ctx_r8.create($event); })("onCancel", function SigfoxDeviceRegistrationComponent_ng_template_7_c8y_stepper_0_Template_c8y_stepper_buttons_onCancel_7_listener() { ɵngcc0.ɵɵrestoreView(_r9); const ctx_r10 = ɵngcc0.ɵɵnextContext(2); return ctx_r10.bsModalRef.hide(); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(8, "cdk-step", 14);
    ɵngcc0.ɵɵtemplate(9, SigfoxDeviceRegistrationComponent_ng_template_7_c8y_stepper_0_div_9_Template, 2, 0, "div", 15);
    ɵngcc0.ɵɵelementStart(10, "div", 16);
    ɵngcc0.ɵɵtemplate(11, SigfoxDeviceRegistrationComponent_ng_template_7_c8y_stepper_0_c8y_operation_result_11_Template, 2, 5, "c8y-operation-result", 17);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(12, "c8y-stepper-buttons", 18);
    ɵngcc0.ɵɵlistener("onCustom", function SigfoxDeviceRegistrationComponent_ng_template_7_c8y_stepper_0_Template_c8y_stepper_buttons_onCustom_12_listener() { ɵngcc0.ɵɵrestoreView(_r9); const ctx_r11 = ɵngcc0.ɵɵnextContext(2); return ctx_r11.bsModalRef.hide(); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("hideStepProgress", true);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("stepControl", ctx_r5.form);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 14, "Register a single Sigfox device"), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("form", ctx_r5.form)("fields", ctx_r5.fields)("model", ctx_r5.model);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("labels", ctx_r5.registrationStepLabels)("showButtons", ɵngcc0.ɵɵpureFunction0(16, _c0))("pending", ctx_r5.state === "registrationPending")("disabled", !ctx_r5.form.valid);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r5.state === "registrationPending");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r5.state === "registrationSuccess");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("showButtons", ɵngcc0.ɵɵpureFunction0(17, _c1))("labels", ctx_r5.finalStepLabels);
} }
function SigfoxDeviceRegistrationComponent_ng_template_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, SigfoxDeviceRegistrationComponent_ng_template_7_c8y_stepper_0_Template, 13, 18, "c8y-stepper", 7);
    ɵngcc0.ɵɵpipe(1, "async");
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    const _r3 = ɵngcc0.ɵɵreference(10);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(1, 2, ctx_r2.errorMessages$).length === 0)("ngIfElse", _r3);
} }
function SigfoxDeviceRegistrationComponent_ng_template_9_c8y_operation_result_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-operation-result", 24);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("text", ɵngcc0.ɵɵpipeBind1(1, 3, "Failed to register"));
    ɵngcc0.ɵɵproperty("size", 84)("vertical", true);
} }
const _c2 = function (a0, a1) { return { "text-center": a0, "alert alert-danger": a1 }; };
function SigfoxDeviceRegistrationComponent_ng_template_9_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 25);
    ɵngcc0.ɵɵelement(1, "span", 26);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const msg_r14 = ctx.$implicit;
    const ctx_r13 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(4, _c2, ctx_r13.state === "registrationError", ctx_r13.state === "loadError"));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("innerHTML", ɵngcc0.ɵɵpipeBind1(2, 2, msg_r14), ɵngcc0.ɵɵsanitizeHtml);
} }
function SigfoxDeviceRegistrationComponent_ng_template_9_Template(rf, ctx) { if (rf & 1) {
    const _r16 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 16);
    ɵngcc0.ɵɵtemplate(1, SigfoxDeviceRegistrationComponent_ng_template_9_c8y_operation_result_1_Template, 2, 5, "c8y-operation-result", 20);
    ɵngcc0.ɵɵtemplate(2, SigfoxDeviceRegistrationComponent_ng_template_9_div_2_Template, 3, 7, "div", 21);
    ɵngcc0.ɵɵpipe(3, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "div", 22);
    ɵngcc0.ɵɵelementStart(5, "button", 23);
    ɵngcc0.ɵɵlistener("click", function SigfoxDeviceRegistrationComponent_ng_template_9_Template_button_click_5_listener() { ɵngcc0.ɵɵrestoreView(_r16); const ctx_r15 = ɵngcc0.ɵɵnextContext(); return ctx_r15.bsModalRef.hide(); });
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵtext(7, " Close ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r4.state === "registrationError");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ɵngcc0.ɵɵpipeBind1(3, 3, ctx_r4.errorMessages$));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(6, 5, "Close"));
} }
var ErrorName;
(function (ErrorName) {
    ErrorName["NoDeviceProtocolsError"] = "NoDeviceProtocolsError";
    ErrorName["NoConnectivitySettingsError"] = "NoConnectivitySettingsError";
    ErrorName["ContractError"] = "ContractError";
    ErrorName["NoContractsError"] = "NoContractsError";
    ErrorName["RegistrationError"] = "RegistrationError";
    ErrorName["DeviceProtocolsFetchError"] = "DeviceProtocolsFetchError";
})(ErrorName || (ErrorName = {}));
class SigfoxProviderService {
    constructor(client, inventoryService, tenantOptions, translateService, applicationService) {
        this.client = client;
        this.inventoryService = inventoryService;
        this.tenantOptions = tenantOptions;
        this.translateService = translateService;
        this.applicationService = applicationService;
        this.baseUrl = '/service/sigfox-agent/';
        this.registrationUrl = `${this.baseUrl}newDeviceRequest`;
        this.contractsUrl = `${this.baseUrl}contract`;
        this.header = { 'Content-Type': 'application/json' };
    }
    getConnections() {
        return __awaiter(this, void 0, void 0, function* () {
            const options = {
                method: 'GET',
                headers: this.header
            };
            const res = yield this.client.fetch(`${this.baseUrl}lns-connection`, options);
            const data = yield res.json();
            if (res.status === 200) {
                if (data.length === 0) {
                    yield this.throwNoConnectivitySettingsError();
                }
            }
            else {
                yield this.throwNoConnectivitySettingsError();
            }
            return { res, data };
        });
    }
    /**
     * Gets contracts from Sigfox platform.
     * @param connectionName The name of connection for which contracts will be retrieved
     * @returns The result list with contract, or throws an error with exception.
     */
    getContracts(connectionName) {
        return __awaiter(this, void 0, void 0, function* () {
            const options = {
                method: 'GET',
                headers: this.header,
                params: {
                    sigfoxConnectionName: connectionName
                }
            };
            const res = yield this.client.fetch(this.contractsUrl, options);
            const data = yield res.json();
            if (res.status === 200) {
                if (data.length === 0) {
                    this.throwNoContractsError();
                }
            }
            else {
                this.throwContractError(data);
            }
            return { res, data };
        });
    }
    createDevice(device) {
        return __awaiter(this, void 0, void 0, function* () {
            const options = {
                method: 'POST',
                headers: this.header,
                body: JSON.stringify(device)
            };
            const res = yield this.client.fetch(this.registrationUrl, options);
            const data = yield res.json();
            if (res.status !== 201) {
                this.throwRegistrationError(data);
            }
            return { res, data };
        });
    }
    getAvailableProtocols(filter = { withTotalPages: true }) {
        return __awaiter(this, void 0, void 0, function* () {
            const query = {
                __filter: {
                    __and: [
                        { __has: 'c8y_IsDeviceType' },
                        {
                            type: { __in: ['c8y_SigfoxDeviceType', 'c8y_LpwanDeviceType'] }
                        }
                    ]
                },
                __orderby: [{ name: 1 }]
            };
            const { res, data } = yield this.inventoryService.listQuery(query, filter);
            if (res.status === 200) {
                if (data.length === 0) {
                    this.throwNoDeviceProtocolsError();
                }
            }
            else {
                this.throwDeviceProtocolsFetchError();
            }
            return { res, data };
        });
    }
    hasConnectivitySettings() {
        return __awaiter(this, void 0, void 0, function* () {
            const option = {
                category: 'sigfox-agent',
                key: 'provider.token'
            };
            try {
                yield this.tenantOptions.detail(option);
                return true;
            }
            catch (e) {
                yield this.throwNoConnectivitySettingsError();
            }
        });
    }
    throwNoConnectivitySettingsError() {
        return __awaiter(this, void 0, void 0, function* () {
            const error = new Error();
            error.name = ErrorName.NoConnectivitySettingsError;
            const hasAdminRight = (yield this.applicationService.isAvailable('administration')).data;
            if (hasAdminRight) {
                error.message = this.translateService.instant(gettext(`Connectivity settings are not configured. Configure them in the Administration app under <a href="{{ link }}">Settings</a>.`), {
                    link: '/apps/administration/index.html#/connectivitySettings/sigfox_provider_settings'
                });
            }
            else {
                error.message = gettext('Connectivity settings are not configured. Contact the administrator.');
            }
            throw error;
        });
    }
    throwRegistrationError(data) {
        const error = new Error();
        error.name = ErrorName.RegistrationError;
        error.message = data.message;
        throw error;
    }
    throwDeviceProtocolsFetchError() {
        const error = new Error();
        error.name = ErrorName.DeviceProtocolsFetchError;
        error.message = gettext('Could not load device protocols.');
        throw error;
    }
    throwNoDeviceProtocolsError() {
        const error = new Error();
        error.name = ErrorName.NoDeviceProtocolsError;
        error.message = this.translateService.instant(gettext(`No device protocols configured. Create a Sigfox device protocol in <a href="{{ link }}">Device protocols</a>.`), {
            link: '/apps/devicemanagement/#/deviceprotocols'
        });
        throw error;
    }
    throwContractError(data) {
        const error = new Error();
        error.name = ErrorName.ContractError;
        error.message = data.message;
        throw error;
    }
    throwNoContractsError() {
        const error = new Error();
        error.name = ErrorName.NoContractsError;
        error.message = gettext('No contracts found. New contracts must be created via the Sigfox platform.');
        throw error;
    }
}
SigfoxProviderService.ɵfac = function SigfoxProviderService_Factory(t) { return new (t || SigfoxProviderService)(ɵngcc0.ɵɵinject(ɵngcc1.FetchClient), ɵngcc0.ɵɵinject(ɵngcc1.InventoryService), ɵngcc0.ɵɵinject(ɵngcc1.TenantOptionsService), ɵngcc0.ɵɵinject(ɵngcc2.TranslateService), ɵngcc0.ɵɵinject(ɵngcc1.ApplicationService)); };
SigfoxProviderService.ɵprov = i0.ɵɵdefineInjectable({ factory: function SigfoxProviderService_Factory() { return new SigfoxProviderService(i0.ɵɵinject(i1.FetchClient), i0.ɵɵinject(i2.InventoryService), i0.ɵɵinject(i3.TenantOptionsService), i0.ɵɵinject(i4.TranslateService), i0.ɵɵinject(i5.ApplicationService)); }, token: SigfoxProviderService, providedIn: "root" });
SigfoxProviderService.ctorParameters = () => [
    { type: FetchClient },
    { type: InventoryService },
    { type: TenantOptionsService },
    { type: TranslateService },
    { type: ApplicationService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SigfoxProviderService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: ɵngcc1.FetchClient }, { type: ɵngcc1.InventoryService }, { type: ɵngcc1.TenantOptionsService }, { type: ɵngcc2.TranslateService }, { type: ɵngcc1.ApplicationService }]; }, null); })();

class SigfoxDeviceRegistrationComponent {
    constructor(bsModalRef, sigfoxService, translateService) {
        this.bsModalRef = bsModalRef;
        this.sigfoxService = sigfoxService;
        this.translateService = translateService;
        this.PAGING = {
            withTotalPages: true,
            pageSize: 10
        };
        this.form = new FormGroup({});
        this.model = {};
        this.protocols$ = this.getProtocols$();
        this.connections$ = this.getConnections$();
        this.unsubscribe$ = new Subject();
        this.load$ = this.connections$.pipe(catchError((error) => of(error)), switchMap(connections => {
            if (connections instanceof Error && connections.name === ErrorName.NoConnectivitySettingsError) {
                return of([connections]);
            }
            return forkJoin([
                of(connections),
                this.protocols$.pipe(catchError((error) => of(error)))
            ]);
        }), map(results => {
            return results.filter(result => {
                return result instanceof Error;
            });
        }), switchMap(errors => {
            return errors.length === 0 ? of([]) : throwError(errors);
        }));
        this.fields = [
            {
                key: 'id',
                type: 'string',
                templateOptions: {
                    placeholder: 'FED987',
                    label: gettext('ID'),
                    required: true,
                    pattern: '(0x){0,1}[0-9A-F]+(h){0,1}'
                },
                validation: {
                    messages: {
                        pattern: gettext('Must be a valid hexadecimal number.')
                    }
                }
            },
            {
                key: 'pac',
                type: 'string',
                templateOptions: {
                    placeholder: 'FEDCBA9876543210',
                    label: gettext('PAC'),
                    required: true,
                    pattern: '^([a-fA-F0-9]{16})$'
                },
                validation: {
                    messages: {
                        pattern: gettext('Must be a valid 16 digit hexadecimal number.')
                    }
                }
            },
            {
                key: 'connection',
                type: 'typeahead',
                templateOptions: {
                    label: gettext('Connection'),
                    required: true,
                    c8yForOptions: this.connections$,
                    displayProperty: 'name',
                    valueProperties: ['name']
                }
            },
            {
                key: 'contract',
                type: 'typeahead',
                templateOptions: {
                    label: gettext('Contract'),
                    required: true,
                    placeholder: 'Free contract_25',
                    displayProperty: 'name',
                    valueProperties: ['id'],
                    description: gettext('Only active contracts with free slots are displayed.')
                },
                hooks: {
                    onInit: field => {
                        const connectionControl = field.form.get('connection');
                        connectionControl.valueChanges
                            .pipe(takeUntil(this.unsubscribe$), mergeMap(() => this.getContracts$(this.form.get('connection').value.name)))
                            .subscribe(profiles => {
                            field.templateOptions.c8yForOptions = of(profiles);
                            field.formControl.setValue(null);
                        }, error => {
                            field.form.controls.contract.setErrors({ contract: true });
                            field.validators.contract.message = error.message;
                        });
                    }
                },
                validators: {
                    contract: {
                        expression: (control) => {
                            return control.status === 'VALID';
                        },
                        message: () => '',
                    },
                }
            },
            {
                key: 'deviceType',
                type: 'typeahead',
                templateOptions: {
                    label: gettext('Device protocol'),
                    required: true,
                    c8yForOptions: this.protocols$,
                    displayProperty: 'name',
                    valueProperties: ['id', 'name']
                }
            },
            {
                key: 'productCertificate',
                type: 'string',
                templateOptions: {
                    placeholder: 'P_001F_EDCB_01',
                    label: gettext('Product certificate key'),
                    pattern: 'P_[0-9A-F]{4}_[0-9A-F]{4}_[0-9A-F]{2}',
                    description: gettext('If no product certificate key is specified, the device is considered a prototype.')
                },
                validation: {
                    messages: {
                        pattern: (error, field) => this.translateService.instant(gettext('Must be a valid product certificate key, for example, {{ example }}'), { example: 'P_001F_EDCB_01' })
                    }
                }
            }
        ];
        this.registrationStepLabels = {
            next: gettext('Register')
        };
        this.finalStepLabels = {
            back: gettext('Close')
        };
        this.state = 'loadPending';
        this.errors$ = new BehaviorSubject([]);
        this.errorMessages$ = this.errors$.pipe(map(errors => errors.map(error => error.message)), map(messages => uniq(messages)));
        this.load$.subscribe(() => {
            this.state = 'loadSuccess';
        }, errors => {
            this.state = 'loadError';
            this.errors$.next(errors);
        });
    }
    create(event) {
        return __awaiter(this, void 0, void 0, function* () {
            this.state = 'registrationPending';
            const sigfoxDevice = this.getSigfoxDeviceToSend();
            try {
                yield this.sigfoxService.createDevice(sigfoxDevice);
                this.state = 'registrationSuccess';
            }
            catch (error) {
                this.state = 'registrationError';
                this.errors$.next([error]);
            }
            event.stepper.next();
        });
    }
    getSigfoxDeviceToSend() {
        const sigfoxDevice = cloneDeep(this.model);
        sigfoxDevice.lnsConnectionName = this.model.connection.name;
        sigfoxDevice.contractId = this.model.contract.id;
        sigfoxDevice.prototype = !sigfoxDevice.productCertificate;
        delete sigfoxDevice.contract;
        delete sigfoxDevice.connection;
        return sigfoxDevice;
    }
    getContracts$(name) {
        return defer(() => from(this.sigfoxService.getContracts(name))).pipe(shareReplay(1));
    }
    getProtocols$() {
        return defer(() => from(this.sigfoxService.getAvailableProtocols())).pipe(shareReplay(1));
    }
    getConnections$() {
        return defer(() => from(this.sigfoxService.getConnections())).pipe(shareReplay(1));
    }
    ngOnDestroy() {
        this.unsubscribe$.next();
        this.unsubscribe$.complete();
    }
}
SigfoxDeviceRegistrationComponent.ɵfac = function SigfoxDeviceRegistrationComponent_Factory(t) { return new (t || SigfoxDeviceRegistrationComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc3.BsModalRef), ɵngcc0.ɵɵdirectiveInject(SigfoxProviderService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.TranslateService)); };
SigfoxDeviceRegistrationComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SigfoxDeviceRegistrationComponent, selectors: [["c8y-sigfox-device-registration"]], decls: 11, vars: 8, consts: [[3, "headerClasses", "customFooter"], ["c8y-modal-title", ""], [3, "c8yIcon"], [4, "ngIf", "ngIfElse"], ["registrationForm", ""], ["errorMessagesPresent", ""], [1, "p-16", "text-center"], ["linear", "", 3, "hideStepProgress", 4, "ngIf", "ngIfElse"], ["linear", "", 3, "hideStepProgress"], [3, "stepControl"], [1, "p-b-16"], [1, "p-l-24", "p-r-24", "p-t-16", "p-b-16", "m-b-0", "sticky-top", "separator-bottom", "text-16", "text-medium", "text-center", "bg-component"], [1, "d-block", "p-l-24", "p-r-24", "p-t-16", 3, "form", "fields", "model"], [1, "modal-footer", "d-block", "sticky-bottom", "separator-top", "bg-white", 3, "labels", "showButtons", "pending", "disabled", "onNext", "onCancel"], ["state", "final"], ["class", "p-16 text-center", 4, "ngIf"], [1, "m-24"], ["type", "success", "class", "lead m-b-0", 3, "text", "size", "vertical", 4, "ngIf"], [1, "sticky-bottom", "d-block", "p-t-16", "p-b-16", "separator-top", "bg-white", 3, "showButtons", "labels", "onCustom"], ["type", "success", 1, "lead", "m-b-0", 3, "text", "size", "vertical"], ["type", "error", "class", "lead", 3, "text", "size", "vertical", 4, "ngIf"], ["class", "m-b-8", 3, "ngClass", 4, "ngFor", "ngForOf"], [1, "modal-footer"], ["type", "button", "translate", "", 1, "btn", "btn-default", 3, "title", "click"], ["type", "error", 1, "lead", 3, "text", "size", "vertical"], [1, "m-b-8", 3, "ngClass"], [3, "innerHTML"]], template: function SigfoxDeviceRegistrationComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-modal", 0);
        ɵngcc0.ɵɵelementContainerStart(1, 1);
        ɵngcc0.ɵɵelement(2, "span", 2);
        ɵngcc0.ɵɵelementStart(3, "h4");
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementContainerEnd();
        ɵngcc0.ɵɵtemplate(6, SigfoxDeviceRegistrationComponent_ng_container_6_Template, 3, 0, "ng-container", 3);
        ɵngcc0.ɵɵtemplate(7, SigfoxDeviceRegistrationComponent_ng_template_7_Template, 2, 4, "ng-template", null, 4, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(9, SigfoxDeviceRegistrationComponent_ng_template_9_Template, 8, 7, "ng-template", null, 5, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r1 = ɵngcc0.ɵɵreference(8);
        ɵngcc0.ɵɵproperty("headerClasses", "dialog-header")("customFooter", true);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yIcon", "c8y-device-connect");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(5, 6, "Sigfox registration"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.state === "loadPending")("ngIfElse", _r1);
    } }, directives: [ɵngcc4.ModalComponent, ɵngcc4.IconDirective, ɵngcc5.NgIf, ɵngcc4.LoadingComponent, ɵngcc4.C8yStepper, ɵngcc6.CdkStep, ɵngcc7.FormlyForm, ɵngcc4.C8yStepperButtons, ɵngcc4.OperationResultComponent, ɵngcc5.NgForOf, ɵngcc4.C8yTranslateDirective, ɵngcc5.NgClass], pipes: [ɵngcc4.C8yTranslatePipe, ɵngcc5.AsyncPipe], encapsulation: 2 });
SigfoxDeviceRegistrationComponent.ctorParameters = () => [
    { type: BsModalRef },
    { type: SigfoxProviderService },
    { type: TranslateService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SigfoxDeviceRegistrationComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-sigfox-device-registration',
                template: "<c8y-modal [headerClasses]=\"'dialog-header'\" [customFooter]=\"true\">\n  <ng-container c8y-modal-title>\n    <span [c8yIcon]=\"'c8y-device-connect'\"></span>\n    <h4>{{ 'Sigfox registration' | translate }}</h4>\n  </ng-container>\n  <ng-container *ngIf=\"state === 'loadPending'; else registrationForm\">\n    <div class=\"p-16 text-center\">\n      <c8y-loading></c8y-loading>\n    </div>\n  </ng-container>\n\n  <!--Formly schema is rendered-->\n  <ng-template #registrationForm>\n    <c8y-stepper\n      [hideStepProgress]=\"true\"\n      linear\n      *ngIf=\"(errorMessages$ | async).length === 0; else errorMessagesPresent\"\n    >\n      <cdk-step [stepControl]=\"form\">\n        <div class=\"p-b-16\">\n          <p\n            class=\"\n              p-l-24 p-r-24 p-t-16 p-b-16\n              m-b-0\n              sticky-top\n              separator-bottom\n              text-16 text-medium\n              text-center\n              bg-component\n            \"\n          >\n          {{ 'Register a single Sigfox device' | translate }}\n          </p>\n          <formly-form\n            [form]=\"form\"\n            [fields]=\"fields\"\n            [model]=\"model\"\n            class=\"d-block p-l-24 p-r-24 p-t-16\"\n          ></formly-form>\n        </div>\n        <c8y-stepper-buttons\n          [labels]=\"registrationStepLabels\"\n          (onNext)=\"create($event)\"\n          (onCancel)=\"bsModalRef.hide()\"\n          [showButtons]=\"{ cancel: true, next: true }\"\n          [pending]=\"state === 'registrationPending'\"\n          [disabled]=\"!form.valid\"\n          class=\"modal-footer d-block sticky-bottom separator-top bg-white\"\n        ></c8y-stepper-buttons>\n      </cdk-step>\n      <cdk-step state=\"final\">\n        <!--success scenario-->\n        <div class=\"p-16 text-center\" *ngIf=\"state === 'registrationPending'\">\n          <c8y-loading></c8y-loading>\n        </div>\n        <div class=\"m-24\">\n          <c8y-operation-result\n            *ngIf=\"state === 'registrationSuccess'\"\n            text=\"{{ 'Device registered' | translate }}\"\n            [size]=\"84\"\n            [vertical]=\"true\"\n            type=\"success\"\n            class=\"lead m-b-0\"\n          ></c8y-operation-result>\n        </div>\n\n        <c8y-stepper-buttons\n          class=\"sticky-bottom d-block p-t-16 p-b-16 separator-top bg-white\"\n          (onCustom)=\"bsModalRef.hide()\"\n          [showButtons]=\"{ custom: true }\"\n          [labels]=\"finalStepLabels\"\n        ></c8y-stepper-buttons>\n      </cdk-step>\n    </c8y-stepper>\n  </ng-template>\n\n  <!--Failure scenario-->\n  <ng-template #errorMessagesPresent>\n    <div class=\"m-24\">\n      <c8y-operation-result\n        *ngIf=\"state === 'registrationError'\"\n        text=\"{{ 'Failed to register' | translate }}\"\n        [size]=\"84\"\n        [vertical]=\"true\"\n        type=\"error\"\n        class=\"lead\"\n      ></c8y-operation-result>\n      <div\n        *ngFor=\"let msg of errorMessages$ | async\"\n        class=\"m-b-8\"\n        [ngClass]=\"{\n          'text-center': state === 'registrationError',\n          'alert alert-danger': state === 'loadError'\n        }\"\n      >\n        <span [innerHTML]=\"msg | translate\"></span>\n      </div>\n    </div>\n\n    <div class=\"modal-footer\">\n      <button\n        title=\"{{ 'Close' | translate }}\"\n        (click)=\"bsModalRef.hide()\"\n        type=\"button\"\n        class=\"btn btn-default\"\n        translate\n      >\n        Close\n      </button>\n    </div>\n  </ng-template>\n</c8y-modal>"
            }]
    }], function () { return [{ type: ɵngcc3.BsModalRef }, { type: SigfoxProviderService }, { type: ɵngcc2.TranslateService }]; }, null); })();

class SigfoxDeviceRegistrationButtonComponent {
    constructor(modalService) {
        this.modalService = modalService;
    }
    open() {
        this.modalService.show(SigfoxDeviceRegistrationComponent, {
            class: 'modal-sm',
            ignoreBackdropClick: true
        });
    }
}
SigfoxDeviceRegistrationButtonComponent.ɵfac = function SigfoxDeviceRegistrationButtonComponent_Factory(t) { return new (t || SigfoxDeviceRegistrationButtonComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc3.BsModalService)); };
SigfoxDeviceRegistrationButtonComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SigfoxDeviceRegistrationButtonComponent, selectors: [["c8y-sigfox-registration"]], decls: 4, vars: 3, consts: [[3, "click"], ["c8yIcon", "c8y-device-connect"]], template: function SigfoxDeviceRegistrationButtonComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "button", 0);
        ɵngcc0.ɵɵlistener("click", function SigfoxDeviceRegistrationButtonComponent_Template_button_click_0_listener() { return ctx.open(); });
        ɵngcc0.ɵɵelement(1, "i", 1);
        ɵngcc0.ɵɵtext(2);
        ɵngcc0.ɵɵpipe(3, "translate");
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 1, "Sigfox"), " ");
    } }, directives: [ɵngcc4.IconDirective], pipes: [ɵngcc4.C8yTranslatePipe], encapsulation: 2 });
SigfoxDeviceRegistrationButtonComponent.ctorParameters = () => [
    { type: BsModalService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SigfoxDeviceRegistrationButtonComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-sigfox-registration',
                template: "<button (click)=\"open()\"><i c8yIcon=\"c8y-device-connect\"></i> {{ 'Sigfox' | translate }} </button>"
            }]
    }], function () { return [{ type: ɵngcc3.BsModalService }]; }, null); })();

class SigfoxDeviceRegistrationFactory {
    constructor(tenantService) {
        this.tenantService = tenantService;
    }
    get() {
        const items = [];
        if (this.tenantService.isMicroserviceSubscribedInCurrentTenant('sigfox-agent')) {
            items.push({
                template: SigfoxDeviceRegistrationButtonComponent,
                priority: 98,
                category: 'single'
            });
        }
        return items;
    }
}
SigfoxDeviceRegistrationFactory.ɵfac = function SigfoxDeviceRegistrationFactory_Factory(t) { return new (t || SigfoxDeviceRegistrationFactory)(ɵngcc0.ɵɵinject(ɵngcc4.TenantUiService)); };
SigfoxDeviceRegistrationFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: SigfoxDeviceRegistrationFactory, factory: SigfoxDeviceRegistrationFactory.ɵfac });
SigfoxDeviceRegistrationFactory.ctorParameters = () => [
    { type: TenantUiService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SigfoxDeviceRegistrationFactory, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc4.TenantUiService }]; }, null); })();

class SigfoxDeviceRegistrationModule {
}
SigfoxDeviceRegistrationModule.ɵfac = function SigfoxDeviceRegistrationModule_Factory(t) { return new (t || SigfoxDeviceRegistrationModule)(); };
SigfoxDeviceRegistrationModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: SigfoxDeviceRegistrationModule });
SigfoxDeviceRegistrationModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        {
            provide: HOOK_DEVICE_REGISTRATION,
            useClass: SigfoxDeviceRegistrationFactory,
            multi: true
        },
        SigfoxProviderService
    ], imports: [[CoreModule, CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SigfoxDeviceRegistrationModule, [{
        type: NgModule,
        args: [{
                imports: [CoreModule, CommonModule],
                declarations: [SigfoxDeviceRegistrationButtonComponent, SigfoxDeviceRegistrationComponent],
                providers: [
                    {
                        provide: HOOK_DEVICE_REGISTRATION,
                        useClass: SigfoxDeviceRegistrationFactory,
                        multi: true
                    },
                    SigfoxProviderService
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(SigfoxDeviceRegistrationModule, { declarations: function () { return [SigfoxDeviceRegistrationButtonComponent, SigfoxDeviceRegistrationComponent]; }, imports: function () { return [CoreModule, CommonModule]; } }); })();

/**
 * Generated bundle index. Do not edit.
 */

export { SigfoxDeviceRegistrationButtonComponent, SigfoxDeviceRegistrationComponent, SigfoxDeviceRegistrationFactory, SigfoxDeviceRegistrationModule, SigfoxProviderService as ɵa };
