// @c8y/ngx-components/esm2015/core/realtime/managed-object-realtime.service.js
import { Injectable } from '@angular/core';
import { RealtimeService } from './realtime.service';
import { RealtimeSubjectService } from './realtime-subject.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './realtime-subject.service';
export class ManagedObjectRealtimeService extends RealtimeService {
    constructor(realtimeSubject) {
        super(realtimeSubject);
        this.realtimeSubject = realtimeSubject;
    }
    /**
     * Get an Observable of all CREATE realtime notifications.
     *
     * @returns An [[Observable]] of newly created entity objects.
     */
    onCreate$() {
        return super.onCreate$();
    }
    channel() {
        return '/managedobjects/*';
    }
}
ManagedObjectRealtimeService.ɵfac = function ManagedObjectRealtimeService_Factory(t) { return new (t || ManagedObjectRealtimeService)(ɵngcc0.ɵɵinject(ɵngcc1.RealtimeSubjectService)); };
ManagedObjectRealtimeService.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: ManagedObjectRealtimeService, factory: ManagedObjectRealtimeService.ɵfac });
ManagedObjectRealtimeService.ctorParameters = () => [
    { type: RealtimeSubjectService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ManagedObjectRealtimeService, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.RealtimeSubjectService }]; }, null); })();
