// @c8y/ngx-components/esm2015/core/realtime/operation-bulk-realtime.service.js
import { Injectable } from '@angular/core';
import { RealtimeSubjectService } from './realtime-subject.service';
import { RealtimeService } from './realtime.service';
/**
 * A service for handling bulk operation realtime notifications in RxJS fashion.
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './realtime-subject.service';
export class OperationBulkRealtimeService extends RealtimeService {
    constructor(realtimeSubject) {
        super(realtimeSubject);
        this.realtimeSubject = realtimeSubject;
    }
    /**
     * Get an Observable of all CREATE realtime notifications.
     *
     * @returns An [[Observable]] of newly created entity objects.
     */
    onCreate$() {
        return super.onCreate$();
    }
    channel() {
        return '/bulkOperations/*';
    }
}
OperationBulkRealtimeService.ɵfac = function OperationBulkRealtimeService_Factory(t) { return new (t || OperationBulkRealtimeService)(ɵngcc0.ɵɵinject(ɵngcc1.RealtimeSubjectService)); };
OperationBulkRealtimeService.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: OperationBulkRealtimeService, factory: OperationBulkRealtimeService.ɵfac });
OperationBulkRealtimeService.ctorParameters = () => [
    { type: RealtimeSubjectService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OperationBulkRealtimeService, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.RealtimeSubjectService }]; }, null); })();
