// @c8y/ngx-components/esm2015/core/realtime/realtime.service.js
import { coerceNumberProperty } from '@angular/cdk/coercion';
import { BehaviorSubject, NEVER } from 'rxjs';
import { filter, map, switchMap } from 'rxjs/operators';
/**
 * A wrapper class for handling realtime notifications in RxJS fashion.
 */
export class RealtimeService {
    constructor(realtimeSubject) {
        this.realtimeSubject = realtimeSubject;
        this.isActive = new BehaviorSubject(true);
    }
    /**
     * A flag displaying if realtime notifications are currently active.
     */
    get active() {
        return this.isActive.value;
    }
    /**
     * Get an Observable of all realtime notifications.
     *
     * @param {string | number | IIdentified} entityOrId Entity object or id
     *
     * @returns An [[Observable]] of notifications wrapped as [[RealtimeMessage]]
     */
    onAll$(entityOrId) {
        const subject$ = this.realtimeSubject.getObservableForChannel(this.getChannel(entityOrId));
        return this.isActive.pipe(switchMap(active => (active ? subject$ : NEVER)));
    }
    /**
     * Subscribes again all realtime channels with active observers.
     */
    start() {
        if (!this.active) {
            this.isActive.next(true);
        }
    }
    /**
     * Stops realtime notifications and unsubscribes all realtime channels.
     */
    stop() {
        if (this.active) {
            this.isActive.next(false);
        }
    }
    /**
     * Get an Observable of all CREATE realtime notifications.
     *
     * @param {string | number | IIdentified} entityOrId Entity object or id
     *
     * @returns An [[Observable]] of newly created entity objects.
     */
    onCreate$(entityOrId) {
        return this.onAll$(entityOrId).pipe(filter(msg => msg.realtimeAction === 'CREATE'), map(msg => msg.data));
    }
    /**
     * Get an Observable of all UPDATE realtime notifications.
     *
     * @param {string | number | IIdentified} entityOrId Entity object or id
     *
     * @returns An [[Observable]] of updated entity objects.
     */
    onUpdate$(entityOrId) {
        return this.onAll$(entityOrId).pipe(filter(msg => msg.realtimeAction === 'UPDATE'), map(msg => msg.data));
    }
    /**
     * Get an Observable of all DELETE realtime notifications.
     *
     * @param {string | number | IIdentified} entityOrId Entity object or id
     *
     * @returns An [[Observable]] of deleted entity objects.
     */
    onDelete$(entityOrId) {
        return this.onAll$(entityOrId).pipe(filter(msg => msg.realtimeAction === 'DELETE'), map(msg => coerceNumberProperty(msg.data)));
    }
    getIdString(reference) {
        let id;
        if (typeof reference === 'object') {
            id = reference.id;
        }
        else {
            id = reference;
        }
        return String(id);
    }
    getChannel(entityOrId) {
        return entityOrId ? this.channel().replace('*', this.getIdString(entityOrId)) : this.channel();
    }
}
