// @c8y/ngx-components/esm2015/core/realtime/realtime-subject.service.js
import { Injectable } from '@angular/core';
import { Realtime } from '@c8y/client';
import { Observable } from 'rxjs';
import { share } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/client/lib/src/realtime/Realtime";
/**
 * Service (providedIn root) that ensures to only create a single realtime subscription for each channel
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/client';
export class RealtimeSubjectService {
    constructor(realtime) {
        this.realtime = realtime;
        this.subjects$ = new Map();
    }
    getObservableForChannel(channel) {
        if (this.subjects$.has(channel)) {
            return this.subjects$.get(channel);
        }
        const observable$ = this.createObservableForChannel(channel, this.realtime);
        const sharedObservable$ = observable$.pipe(share());
        this.subjects$.set(channel, sharedObservable$);
        return sharedObservable$;
    }
    createObservableForChannel(channel, realtime) {
        return new Observable(observer => {
            const realtimeSubscription = realtime.subscribe(channel, msg => {
                const data = {
                    channel: msg.channel,
                    data: msg.data.data,
                    id: msg.id,
                    realtimeAction: msg.data.realtimeAction
                };
                observer.next(data);
            });
            return {
                unsubscribe() {
                    realtime.unsubscribe(realtimeSubscription);
                }
            };
        });
    }
}
RealtimeSubjectService.ɵfac = function RealtimeSubjectService_Factory(t) { return new (t || RealtimeSubjectService)(ɵngcc0.ɵɵinject(ɵngcc1.Realtime)); };
RealtimeSubjectService.ɵprov = i0.ɵɵdefineInjectable({ factory: function RealtimeSubjectService_Factory() { return new RealtimeSubjectService(i0.ɵɵinject(i1.Realtime)); }, token: RealtimeSubjectService, providedIn: "root" });
RealtimeSubjectService.ctorParameters = () => [
    { type: Realtime }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(RealtimeSubjectService, [{
        type: Injectable,
        args: [{ providedIn: 'root' }]
    }], function () { return [{ type: ɵngcc1.Realtime }]; }, null); })();
