// @c8y/ngx-components/esm2015/operations/stepper-bulk-type-firmware/version-or-patch.component.js
import { __decorate } from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { InventoryService } from '@c8y/client';
import { memoize } from '@c8y/ngx-components';
import { BulkOperationsService } from '@c8y/ngx-components/operations/bulk-operations-service';
import { RepositoryService } from '@c8y/ngx-components/repository/shared';
import { property } from 'lodash-es';
import { defer, of } from 'rxjs';
import { catchError, distinctUntilChanged, distinctUntilKeyChanged, map, shareReplay, switchMap, tap } from 'rxjs/operators';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components/repository/shared';
import * as ɵngcc2 from '@c8y/client';
import * as ɵngcc3 from '@c8y/ngx-components/operations/bulk-operations-service';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from '@c8y/ngx-components';

function VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 9);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Legacy"), " ");
} }
function VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_10_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵelementStart(1, "span", 10);
    ɵngcc0.ɵɵelement(2, "i", 11);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_10_span_3_ng_template_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 16);
    ɵngcc0.ɵɵelementStart(1, "span", 17);
    ɵngcc0.ɵɵtext(2, "No patches");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_10_span_3_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 10);
    ɵngcc0.ɵɵelementStart(1, "span", 17);
    ɵngcc0.ɵɵtext(2, "1 patch");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
const _c0 = function (a0) { return { count: a0 }; };
function VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_10_span_3_ng_template_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 10);
    ɵngcc0.ɵɵelementStart(1, "span", 18);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵtext(3, "{{ count }} patches");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const baseVersion_r2 = ɵngcc0.ɵɵnextContext(3).$implicit;
    const ctx_r11 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction1(3, _c0, ɵngcc0.ɵɵpipeBind1(2, 1, ctx_r11.getPatchVersionsCount$(baseVersion_r2))));
} }
function VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_10_span_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵelementStart(1, "span", 12);
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵtemplate(3, VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_10_span_3_ng_template_3_Template, 3, 0, "ng-template", 13);
    ɵngcc0.ɵɵtemplate(4, VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_10_span_3_ng_template_4_Template, 3, 0, "ng-template", 14);
    ɵngcc0.ɵɵtemplate(5, VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_10_span_3_ng_template_5_Template, 4, 5, "ng-template", 15);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const baseVersion_r2 = ɵngcc0.ɵɵnextContext(2).$implicit;
    const ctx_r8 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngPlural", ɵngcc0.ɵɵpipeBind1(2, 1, ctx_r8.getPatchVersionsCount$(baseVersion_r2)));
} }
function VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtemplate(1, VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_10_span_1_Template, 3, 0, "span", 8);
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵtemplate(3, VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_10_span_3_Template, 6, 3, "span", 8);
    ɵngcc0.ɵɵpipe(4, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const baseVersion_r2 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(2, 2, ctx_r5.getPatchVersionsCount$(baseVersion_r2)) === null);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(4, 4, ctx_r5.getPatchVersionsCount$(baseVersion_r2)) !== null);
} }
function VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_c8y_li_14_span_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 26);
    ɵngcc0.ɵɵtext(1, "patch");
    ɵngcc0.ɵɵelementEnd();
} }
function VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_c8y_li_14_ng_template_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 27);
    ɵngcc0.ɵɵtext(1, "version");
    ɵngcc0.ɵɵelementEnd();
} }
function VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_c8y_li_14_Template(rf, ctx) { if (rf & 1) {
    const _r21 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li");
    ɵngcc0.ɵɵelementStart(1, "c8y-li-radio", 19);
    ɵngcc0.ɵɵlistener("onSelect", function VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_c8y_li_14_Template_c8y_li_radio_onSelect_1_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r21); const patchVersion_r15 = restoredCtx.$implicit; const ctx_r20 = ɵngcc0.ɵɵnextContext(3); return ctx_r20.selectVersionOrPatch($event, patchVersion_r15); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(2, "c8y-li-icon");
    ɵngcc0.ɵɵelement(3, "i", 3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "c8y-li-body", 4);
    ɵngcc0.ɵɵelementStart(5, "div", 20);
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "div", 20);
    ɵngcc0.ɵɵtemplate(8, VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_c8y_li_14_span_8_Template, 2, 0, "span", 21);
    ɵngcc0.ɵɵtemplate(9, VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_c8y_li_14_ng_template_9_Template, 2, 0, "ng-template", null, 22, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(11, "div", 23);
    ɵngcc0.ɵɵelementStart(12, "span", 24);
    ɵngcc0.ɵɵtext(13, " File ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(14, "span", 25);
    ɵngcc0.ɵɵpipe(15, "async");
    ɵngcc0.ɵɵtext(16);
    ɵngcc0.ɵɵpipe(17, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const patchVersion_r15 = ctx.$implicit;
    const _r18 = ɵngcc0.ɵɵreference(10);
    const ctx_r6 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("selected", patchVersion_r15 === ctx_r6.selected);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate1(" ", patchVersion_r15.c8y_Firmware.version, " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", patchVersion_r15.c8y_Patch)("ngIfElse", _r18);
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(15, 6, ctx_r6.getBinaryName$(patchVersion_r15.c8y_Firmware.url)));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(17, 8, ctx_r6.getBinaryName$(patchVersion_r15.c8y_Firmware.url)), " ");
} }
function VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li");
    ɵngcc0.ɵɵelementStart(1, "c8y-li-icon");
    ɵngcc0.ɵɵelement(2, "i", 3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "c8y-li-body", 4);
    ɵngcc0.ɵɵelementStart(4, "div", 5);
    ɵngcc0.ɵɵelementStart(5, "p");
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "div", 6);
    ɵngcc0.ɵɵtemplate(8, VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_8_Template, 3, 3, "span", 7);
    ɵngcc0.ɵɵpipe(9, "async");
    ɵngcc0.ɵɵtemplate(10, VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_10_Template, 5, 6, "span", 8);
    ɵngcc0.ɵɵpipe(11, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(12, "c8y-li-collapse");
    ɵngcc0.ɵɵelementStart(13, "c8y-list-group");
    ɵngcc0.ɵɵtemplate(14, VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_c8y_li_14_Template, 18, 10, "c8y-li", 2);
    ɵngcc0.ɵɵpipe(15, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const baseVersion_r2 = ctx.$implicit;
    const ctx_r1 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵtextInterpolate(baseVersion_r2.c8y_Firmware.version);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(9, 5, ctx_r1.isLegacy$));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !ɵngcc0.ɵɵpipeBind1(11, 7, ctx_r1.isLegacy$));
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("c8yForOf", ɵngcc0.ɵɵpipeBind1(15, 9, ctx_r1.getBaseVersionAndPatches$(baseVersion_r2)))("c8yForLoadMore", "auto");
} }
const _c1 = function (a0) { return { "dd-low": a0 }; };
function VersionOrPatchComponent_c8y_list_group_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-list-group", 1);
    ɵngcc0.ɵɵtemplate(1, VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_Template, 16, 11, "c8y-li", 2);
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(5, _c1, ctx_r0.elementCount <= ctx_r0.DD_LOW_COUNT));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yForOf", ɵngcc0.ɵɵpipeBind1(2, 3, ctx_r0.baseVersions$))("c8yForLoadMore", "auto");
} }
export class VersionOrPatchComponent {
    constructor(repositoryService, inventoryService, bulkOpsService) {
        this.repositoryService = repositoryService;
        this.inventoryService = inventoryService;
        this.bulkOpsService = bulkOpsService;
        this.versionOrPatch = new EventEmitter();
        this.elementCount = 0;
        this.DD_LOW_COUNT = 10;
        this.firmware$ = this.bulkOpsService.firmwareId.pipe(distinctUntilChanged(), switchMap(id => defer(() => this.inventoryService.detail(id).then(result => result.data)).pipe(catchError(error => of(error)))), shareReplay(1));
        this.baseVersions$ = this.firmware$.pipe(distinctUntilKeyChanged('id'), switchMap(firmware => this.repositoryService.listBaseVersions(firmware)), tap(resp => {
            this.elementCount = resp.data ? resp.data.length : 0;
        }), shareReplay(1));
        this.isLegacy$ = this.firmware$.pipe(map(firmware => this.repositoryService.isLegacyEntry(firmware)), shareReplay(1));
        this.expanded = {};
        this.DD_LOW_COUNT = this.bulkOpsService.DD_LOW_COUNT;
    }
    getBinaryName$(binaryUrl) {
        return this.repositoryService.getBinaryName$(binaryUrl);
    }
    getBaseVersionAndPatches$(baseVersion) {
        return this.firmware$.pipe(distinctUntilKeyChanged('id'), switchMap(firmware => this.repositoryService.listBaseVersionAndPatches(firmware, baseVersion)), shareReplay(1));
    }
    getPatchVersionsCount$(baseVersion) {
        return this.firmware$.pipe(distinctUntilKeyChanged('id'), switchMap(() => this.firmware$), switchMap(firmware => this.repositoryService.getPatchVersionsCount$(firmware, baseVersion)), shareReplay(1));
    }
    selectVersionOrPatch(selected, versionOrPatch) {
        if (selected) {
            this.versionOrPatch.emit(versionOrPatch);
        }
    }
}
VersionOrPatchComponent.ɵfac = function VersionOrPatchComponent_Factory(t) { return new (t || VersionOrPatchComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.RepositoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.InventoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.BulkOperationsService)); };
VersionOrPatchComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: VersionOrPatchComponent, selectors: [["c8y-version-or-patch-step"]], inputs: { selected: "selected" }, outputs: { versionOrPatch: "versionOrPatch" }, decls: 2, vars: 3, consts: [["class", "m-b-16", 3, "ngClass", 4, "ngIf"], [1, "m-b-16", 3, "ngClass"], [4, "c8yFor", "c8yForOf", "c8yForLoadMore"], ["c8yIcon", "c8y-firmware"], [1, "content-flex-50"], [1, "col-10"], [1, "col-2"], ["class", "label label-warning", 4, "ngIf"], [4, "ngIf"], [1, "label", "label-warning"], [1, "label", "label-info"], ["c8yIcon", "circle-o-notch", 1, "icon-spin"], [3, "ngPlural"], ["ngPluralCase", "=0"], ["ngPluralCase", "=1"], ["ngPluralCase", "other"], [1, "label", "label-default"], ["translate", ""], ["translate", "", 3, "translateParams"], [3, "selected", "onSelect"], [1, "col-3"], ["translate", "", "class", "label label-info", 4, "ngIf", "ngIfElse"], ["version", ""], [1, "col-6", "text-truncate"], ["translate", "", 1, "text-label-small", "m-r-4"], [3, "title"], ["translate", "", 1, "label", "label-info"], ["translate", "", 1, "label", "label-primary"]], template: function VersionOrPatchComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, VersionOrPatchComponent_c8y_list_group_0_Template, 3, 7, "c8y-list-group", 0);
        ɵngcc0.ɵɵpipe(1, "async");
    } if (rf & 2) {
        let tmp_0_0;
        ɵngcc0.ɵɵproperty("ngIf", ((tmp_0_0 = ɵngcc0.ɵɵpipeBind1(1, 1, ctx.baseVersions$)) == null ? null : tmp_0_0.data.length) > 0);
    } }, directives: [ɵngcc4.NgIf, ɵngcc5.ListGroupComponent, ɵngcc4.NgClass, ɵngcc5.ForOfDirective, ɵngcc5.ListItemComponent, ɵngcc5.ListItemIconComponent, ɵngcc5.IconDirective, ɵngcc5.ListItemBodyComponent, ɵngcc5.ListItemCollapseComponent, ɵngcc4.NgPlural, ɵngcc4.NgPluralCase, ɵngcc5.C8yTranslateDirective, ɵngcc5.ListItemRadioComponent], pipes: [ɵngcc4.AsyncPipe, ɵngcc5.C8yTranslatePipe], encapsulation: 2 });
VersionOrPatchComponent.ctorParameters = () => [
    { type: RepositoryService },
    { type: InventoryService },
    { type: BulkOperationsService }
];
VersionOrPatchComponent.propDecorators = {
    selected: [{ type: Input }],
    versionOrPatch: [{ type: Output }]
};
__decorate([
    memoize()
], VersionOrPatchComponent.prototype, "getBinaryName$", null);
__decorate([
    memoize(property('id'))
], VersionOrPatchComponent.prototype, "getBaseVersionAndPatches$", null);
__decorate([
    memoize(property('id'))
], VersionOrPatchComponent.prototype, "getPatchVersionsCount$", null);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(VersionOrPatchComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-version-or-patch-step',
                template: "<c8y-list-group\n  class=\"m-b-16\"\n  [ngClass]=\"{ 'dd-low': elementCount <= DD_LOW_COUNT }\"\n  *ngIf=\"(baseVersions$ | async)?.data.length > 0\"\n>\n  <c8y-li *c8yFor=\"let baseVersion of baseVersions$ | async; let i = index; loadMore: 'auto'\">\n    <c8y-li-icon>\n      <i c8yIcon=\"c8y-firmware\"></i>\n    </c8y-li-icon>\n\n    <c8y-li-body class=\"content-flex-50\">\n      <div class=\"col-10\">\n        <p>{{ baseVersion.c8y_Firmware.version }}</p>\n      </div>\n      <div class=\"col-2\">\n        <span *ngIf=\"isLegacy$ | async\" class=\"label label-warning\">\n          {{ 'Legacy' | translate }}\n        </span>\n\n        <span *ngIf=\"!(isLegacy$ | async)\">\n          <span *ngIf=\"(getPatchVersionsCount$(baseVersion) | async) === null\">\n            <span class=\"label label-info\">\n              <i c8yIcon=\"circle-o-notch\" class=\"icon-spin\"></i>\n            </span>\n          </span>\n\n          <span *ngIf=\"(getPatchVersionsCount$(baseVersion) | async) !== null\">\n            <span [ngPlural]=\"getPatchVersionsCount$(baseVersion) | async\">\n              <ng-template ngPluralCase=\"=0\">\n                <span class=\"label label-default\"> <span translate>No patches</span></span>\n              </ng-template>\n              <ng-template ngPluralCase=\"=1\">\n                <span class=\"label label-info\">\n                  <span translate>1 patch</span>\n                </span>\n              </ng-template>\n              <ng-template ngPluralCase=\"other\">\n                <span class=\"label label-info\">\n                  <span\n                    ngNonBindable\n                    translate\n                    [translateParams]=\"{ count: getPatchVersionsCount$(baseVersion) | async }\"\n                    >{{ count }} patches</span\n                  ></span\n                >\n              </ng-template>\n            </span>\n          </span>\n        </span>\n      </div>\n    </c8y-li-body>\n\n    <c8y-li-collapse>\n      <c8y-list-group>\n        <c8y-li\n          *c8yFor=\"\n            let patchVersion of getBaseVersionAndPatches$(baseVersion) | async;\n            let i = index;\n            loadMore: 'auto'\n          \"\n        >\n          <c8y-li-radio\n            (onSelect)=\"selectVersionOrPatch($event, patchVersion)\"\n            [selected]=\"patchVersion === selected\"\n          ></c8y-li-radio>\n          <c8y-li-icon>\n            <i c8yIcon=\"c8y-firmware\"></i>\n          </c8y-li-icon>\n          <c8y-li-body class=\"content-flex-50\">\n            <div class=\"col-3\">\n              {{ patchVersion.c8y_Firmware.version }}\n            </div>\n            <div class=\"col-3\">\n              <span *ngIf=\"patchVersion.c8y_Patch; else version\" translate class=\"label label-info\"\n                >patch</span\n              >\n              <ng-template #version\n                ><span translate class=\"label label-primary\">version</span></ng-template\n              >\n            </div>\n            <div class=\"col-6 text-truncate\">\n              <span class=\"text-label-small m-r-4\" translate>\n                File\n              </span>\n              <span title=\"{{ getBinaryName$(patchVersion.c8y_Firmware.url) | async }}\">\n                {{ getBinaryName$(patchVersion.c8y_Firmware.url) | async }}\n              </span>\n            </div>\n          </c8y-li-body>\n        </c8y-li>\n      </c8y-list-group>\n    </c8y-li-collapse>\n  </c8y-li>\n</c8y-list-group>\n"
            }]
    }], function () { return [{ type: ɵngcc1.RepositoryService }, { type: ɵngcc2.InventoryService }, { type: ɵngcc3.BulkOperationsService }]; }, { versionOrPatch: [{
            type: Output
        }], selected: [{
            type: Input
        }] }); })();
