// @c8y/ngx-components/esm2015/search/search.module.js
import { NgModule } from '@angular/core';
import { CoreModule, HOOK_SEARCH, HOOK_ROUTE, SearchModule as CoreSearchModule } from '@c8y/ngx-components';
import { DeviceGridModule } from '@c8y/ngx-components/device-grid';
import { SearchResultsComponent } from './search-results.component';
import { SearchGridComponent } from './search-grid.component';
import { SearchActionComponent } from './search-action.component';
import { AssetSearchService } from './search.service';
import * as ɵngcc0 from '@angular/core';
const ɵ0 = [
    {
        path: 'assetsearch',
        component: SearchResultsComponent
    }
], ɵ1 = [
    {
        template: SearchActionComponent
    }
];
export class SearchModule {
}
SearchModule.ɵfac = function SearchModule_Factory(t) { return new (t || SearchModule)(); };
SearchModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: SearchModule });
SearchModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        AssetSearchService,
        {
            provide: HOOK_ROUTE,
            useValue: ɵ0,
            multi: true
        },
        {
            provide: HOOK_SEARCH,
            useValue: ɵ1,
            multi: true
        }
    ], imports: [[CoreModule, DeviceGridModule, CoreSearchModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SearchModule, [{
        type: NgModule,
        args: [{
                declarations: [SearchResultsComponent, SearchGridComponent, SearchActionComponent],
                imports: [CoreModule, DeviceGridModule, CoreSearchModule],
                entryComponents: [SearchResultsComponent, SearchActionComponent],
                exports: [SearchResultsComponent, SearchGridComponent],
                providers: [
                    AssetSearchService,
                    {
                        provide: HOOK_ROUTE,
                        useValue: ɵ0,
                        multi: true
                    },
                    {
                        provide: HOOK_SEARCH,
                        useValue: ɵ1,
                        multi: true
                    }
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(SearchModule, { declarations: function () { return [SearchResultsComponent, SearchGridComponent, SearchActionComponent]; }, imports: function () { return [CoreModule, DeviceGridModule, CoreSearchModule]; }, exports: function () { return [SearchResultsComponent, SearchGridComponent]; } }); })();
export { ɵ0, ɵ1 };
