// @c8y/ngx-components/esm2015/search/search-action.component.js
import { Component } from '@angular/core';
import { Router } from '@angular/router';
import { RouterService as C8yRouter } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from '@angular/router';
export class SearchActionComponent {
    constructor(c8yRouter, router) {
        this.c8yRouter = c8yRouter;
        this.router = router;
    }
    onSearch(on) {
        this.navigate(['/assetsearch'], {
            queryParams: { search: on },
            replaceUrl: true
        });
    }
    onFilter(on) {
        this.navigate(['/assetsearch'], {
            queryParams: { filter: on },
            replaceUrl: true
        });
    }
    onClick(mo) {
        this.router.navigateByUrl(this.c8yRouter.getHref(mo, '/'));
    }
    navigate(commands, extras) {
        this.router
            .navigateByUrl('/', { skipLocationChange: true })
            .then(() => this.router.navigate(commands, extras));
    }
}
SearchActionComponent.ɵfac = function SearchActionComponent_Factory(t) { return new (t || SearchActionComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.RouterService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.Router)); };
SearchActionComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SearchActionComponent, selectors: [["c8y-search-action"]], decls: 1, vars: 0, consts: [[3, "filter", "search", "onClick"]], template: function SearchActionComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-search-input", 0);
        ɵngcc0.ɵɵlistener("filter", function SearchActionComponent_Template_c8y_search_input_filter_0_listener($event) { return ctx.onFilter($event); })("search", function SearchActionComponent_Template_c8y_search_input_search_0_listener($event) { return ctx.onSearch($event); })("onClick", function SearchActionComponent_Template_c8y_search_input_onClick_0_listener($event) { return ctx.onClick($event); });
        ɵngcc0.ɵɵelementEnd();
    } }, directives: [ɵngcc1.SearchInputComponent], encapsulation: 2 });
SearchActionComponent.ctorParameters = () => [
    { type: C8yRouter },
    { type: Router }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SearchActionComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-search-action',
                template: "<c8y-search-input\n  (filter)=\"onFilter($event)\"\n  (search)=\"onSearch($event)\"\n  (onClick)=\"onClick($event)\"\n></c8y-search-input>\n"
            }]
    }], function () { return [{ type: ɵngcc1.RouterService }, { type: ɵngcc2.Router }]; }, null); })();
