// @c8y/ngx-components/esm2015/sub-assets/asset-properties-item.component.js
import { __awaiter } from "tslib";
import { Component, Input } from '@angular/core';
import { FormGroup } from '@angular/forms';
import { AlertService, C8yJSONSchema, gettext, FilesService } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from '@ngx-formly/core';

function AssetPropertiesItemComponent_ng_container_0_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "c8yDate");
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind2(2, 1, ctx_r2.value, "fullDate"), " ");
} }
function AssetPropertiesItemComponent_ng_container_0_ng_container_2_ng_container_1_img_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "img", 8);
} if (rf & 2) {
    const ctx_r8 = ɵngcc0.ɵɵnextContext(4);
    ɵngcc0.ɵɵproperty("src", ctx_r8.previewImage, ɵngcc0.ɵɵsanitizeUrl);
} }
function AssetPropertiesItemComponent_ng_container_0_ng_container_2_ng_container_1_button_2_Template(rf, ctx) { if (rf & 1) {
    const _r11 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 9);
    ɵngcc0.ɵɵlistener("click", function AssetPropertiesItemComponent_ng_container_0_ng_container_2_ng_container_1_button_2_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r11); const ctx_r10 = ɵngcc0.ɵɵnextContext(4); return ctx_r10.filesService.download(ctx_r10.file); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r9 = ɵngcc0.ɵɵnextContext(4);
    ɵngcc0.ɵɵpropertyInterpolate2("title", "", ɵngcc0.ɵɵpipeBind1(1, 3, "Download"), " ", ctx_r9.file.name, "");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r9.file.name, " ");
} }
function AssetPropertiesItemComponent_ng_container_0_ng_container_2_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, AssetPropertiesItemComponent_ng_container_0_ng_container_2_ng_container_1_img_1_Template, 1, 1, "img", 6);
    ɵngcc0.ɵɵtemplate(2, AssetPropertiesItemComponent_ng_container_0_ng_container_2_ng_container_1_button_2_Template, 3, 5, "button", 7);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r6.previewImage);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r6.previewImage);
} }
function AssetPropertiesItemComponent_ng_container_0_ng_container_2_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "No file attached."), " ");
} }
function AssetPropertiesItemComponent_ng_container_0_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, AssetPropertiesItemComponent_ng_container_0_ng_container_2_ng_container_1_Template, 3, 2, "ng-container", 5);
    ɵngcc0.ɵɵtemplate(2, AssetPropertiesItemComponent_ng_container_0_ng_container_2_ng_container_2_Template, 3, 3, "ng-container", 5);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.file);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r3.file);
} }
const _c0 = function (a0, a1) { return { "separator-top-bottom": a0, "separator-bottom": a1 }; };
const _c1 = function (a0) { return { "a-s-start": a0 }; };
function AssetPropertiesItemComponent_ng_container_0_ng_container_3_li_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 12);
    ɵngcc0.ɵɵelementStart(1, "label", 13);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "humanize");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "span", 14);
    ɵngcc0.ɵɵelement(5, "c8y-asset-properties-item", 15);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const prop_r13 = ctx.$implicit;
    const i_r14 = ctx.index;
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(12, _c0, i_r14 === 0, i_r14 > 0));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(15, _c1, prop_r13.file));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 10, prop_r13.label), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵstyleMapInterpolate1("max-width: ", prop_r13.file ? "50%" : "100%", "");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("file", prop_r13.file)("key", prop_r13.key)("type", prop_r13.type)("value", prop_r13.value);
} }
function AssetPropertiesItemComponent_ng_container_0_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "ul", 10);
    ɵngcc0.ɵɵtemplate(2, AssetPropertiesItemComponent_ng_container_0_ng_container_3_li_2_Template, 6, 17, "li", 11);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r4.complex);
} }
function AssetPropertiesItemComponent_ng_container_0_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "p");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(3, 1, ctx_r5.value) || ɵngcc0.ɵɵpipeBind1(4, 3, "Undefined"));
} }
function AssetPropertiesItemComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0, 2);
    ɵngcc0.ɵɵtemplate(1, AssetPropertiesItemComponent_ng_container_0_ng_container_1_Template, 3, 4, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(2, AssetPropertiesItemComponent_ng_container_0_ng_container_2_Template, 3, 2, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(3, AssetPropertiesItemComponent_ng_container_0_ng_container_3_Template, 3, 1, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(4, AssetPropertiesItemComponent_ng_container_0_ng_container_4_Template, 5, 5, "ng-container", 4);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngSwitch", ctx_r0.type);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "date");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "file");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "object");
} }
function AssetPropertiesItemComponent_formly_form_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "formly-form", 16);
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("form", ctx_r1.form)("fields", ctx_r1.fields)("model", ctx_r1.model);
} }
export class AssetPropertiesItemComponent {
    constructor(alert, c8yJsonSchemaService, filesService) {
        this.alert = alert;
        this.c8yJsonSchemaService = c8yJsonSchemaService;
        this.filesService = filesService;
    }
    ngOnChanges(changes) {
        return __awaiter(this, void 0, void 0, function* () {
            if (changes.isEdit) {
                this.resolveJsonSchema();
                yield this.resolveFile();
            }
        });
    }
    resolveFile() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.file) {
                try {
                    const imageFile = yield this.filesService.getFile(this.file);
                    this.previewImage = yield this.getPreviewIfImage(imageFile);
                }
                catch (ex) {
                    this.alert.danger(gettext('File could not be loaded.'));
                }
            }
        });
    }
    resolveJsonSchema() {
        if (this.jsonSchema) {
            const fieldConfig = this.c8yJsonSchemaService.toFieldConfig(this.jsonSchema, this.jsonSchema);
            this.form = new FormGroup({});
            this.fields = [fieldConfig];
            this.model = {
                [this.key]: this.value
            };
        }
    }
    getPreviewIfImage(imageFile) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.filesService.haveValidExtensions(imageFile, 'image')) {
                return this.filesService.toBase64(imageFile);
            }
        });
    }
}
AssetPropertiesItemComponent.ɵfac = function AssetPropertiesItemComponent_Factory(t) { return new (t || AssetPropertiesItemComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.C8yJSONSchema), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.FilesService)); };
AssetPropertiesItemComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AssetPropertiesItemComponent, selectors: [["c8y-asset-properties-item"]], inputs: { key: "key", value: "value", label: "label", type: "type", file: "file", complex: "complex", isEdit: "isEdit", jsonSchema: "jsonSchema" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], decls: 2, vars: 2, consts: [[3, "ngSwitch", 4, "ngIf"], [3, "form", "fields", "model", 4, "ngIf"], [3, "ngSwitch"], [4, "ngSwitchCase"], [4, "ngSwitchDefault"], [4, "ngIf"], ["class", "img-responsive", 3, "src", 4, "ngIf"], ["type", "button", "class", "btn btn-clean text-truncate p-0", 3, "title", "click", 4, "ngIf"], [1, "img-responsive", 3, "src"], ["type", "button", 1, "btn", "btn-clean", "text-truncate", "p-0", 3, "title", "click"], [1, "list-unstyled", "c8y-custom-properties"], ["class", "p-t-4 p-b-4 flex-row text-nowrap", 3, "ngClass", 4, "ngFor", "ngForOf"], [1, "p-t-4", "p-b-4", "flex-row", "text-nowrap", 3, "ngClass"], [1, "small", "m-b-0", "m-r-8", 3, "ngClass"], [1, "m-l-auto"], [3, "file", "key", "type", "value"], [3, "form", "fields", "model"]], template: function AssetPropertiesItemComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, AssetPropertiesItemComponent_ng_container_0_Template, 5, 4, "ng-container", 0);
        ɵngcc0.ɵɵtemplate(1, AssetPropertiesItemComponent_formly_form_1_Template, 1, 3, "formly-form", 1);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isEdit);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isEdit);
    } }, directives: [ɵngcc2.NgIf, ɵngcc2.NgSwitch, ɵngcc2.NgSwitchCase, ɵngcc2.NgSwitchDefault, ɵngcc2.NgForOf, ɵngcc2.NgClass, AssetPropertiesItemComponent, ɵngcc3.FormlyForm], pipes: [ɵngcc1.DatePipe, ɵngcc1.C8yTranslatePipe, ɵngcc1.HumanizePipe], encapsulation: 2 });
AssetPropertiesItemComponent.ctorParameters = () => [
    { type: AlertService },
    { type: C8yJSONSchema },
    { type: FilesService }
];
AssetPropertiesItemComponent.propDecorators = {
    key: [{ type: Input }],
    value: [{ type: Input }],
    label: [{ type: Input }],
    type: [{ type: Input }],
    file: [{ type: Input }],
    complex: [{ type: Input }],
    isEdit: [{ type: Input }],
    jsonSchema: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AssetPropertiesItemComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-asset-properties-item',
                template: "<ng-container [ngSwitch]=\"type\" *ngIf=\"!isEdit\">\n  <ng-container *ngSwitchCase=\"'date'\">\n    {{ value | c8yDate: 'fullDate' }}\n  </ng-container>\n  <ng-container *ngSwitchCase=\"'file'\">\n    <ng-container *ngIf=\"file\">\n      <img *ngIf=\"previewImage\" [src]=\"previewImage\" class=\"img-responsive\" />\n      <button\n        *ngIf=\"!previewImage\"\n        (click)=\"filesService.download(file)\"\n        type=\"button\"\n        title=\"{{ 'Download' | translate }} {{ file.name }}\"\n        class=\"btn btn-clean text-truncate p-0\"\n      >\n        {{ file.name }}\n      </button>\n    </ng-container>\n    <ng-container *ngIf=\"!file\">\n      {{ 'No file attached.' | translate }}\n    </ng-container>\n  </ng-container>\n  <ng-container *ngSwitchCase=\"'object'\">\n    <ul class=\"list-unstyled c8y-custom-properties\">\n      <li\n        *ngFor=\"let prop of complex; let i = index\"\n        [ngClass]=\"{ 'separator-top-bottom': i === 0, 'separator-bottom': i > 0 }\"\n        class=\"p-t-4 p-b-4 flex-row text-nowrap\"\n      >\n        <label class=\"small m-b-0 m-r-8\" [ngClass]=\"{ 'a-s-start': prop.file }\">\n          {{ prop.label | humanize }}\n        </label>\n        <span class=\"m-l-auto\" style=\"max-width: {{ prop.file ? '50%' : '100%' }}\">\n          <c8y-asset-properties-item\n            [file]=\"prop.file\"\n            [key]=\"prop.key\"\n            [type]=\"prop.type\"\n            [value]=\"prop.value\"\n          ></c8y-asset-properties-item>\n        </span>\n      </li>\n    </ul>\n  </ng-container>\n  <!--  \n  <ng-container *ngSwitchCase=\"'boolean'\">\n      <input type=\"checkbox\" [checked]=\"value\" [disabled]=\"true\" />\n    </ng-container>\n  -->\n  <ng-container *ngSwitchDefault>\n    <p>{{ (value | translate) || ('Undefined' | translate) }}</p>\n  </ng-container>\n</ng-container>\n<formly-form *ngIf=\"isEdit\" [form]=\"form\" [fields]=\"fields\" [model]=\"model\"></formly-form>\n"
            }]
    }], function () { return [{ type: ɵngcc1.AlertService }, { type: ɵngcc1.C8yJSONSchema }, { type: ɵngcc1.FilesService }]; }, { key: [{
            type: Input
        }], value: [{
            type: Input
        }], label: [{
            type: Input
        }], type: [{
            type: Input
        }], file: [{
            type: Input
        }], complex: [{
            type: Input
        }], isEdit: [{
            type: Input
        }], jsonSchema: [{
            type: Input
        }] }); })();
