// @c8y/ngx-components/esm2015/sub-assets/asset-properties.component.js
import { __awaiter } from "tslib";
import { Component, Input } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { InventoryBinaryService, InventoryService } from '@c8y/client';
import { AlertService, AssetTypesService, ContextRouteService, gettext } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from '@c8y/client';
import * as ɵngcc3 from '@angular/router';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from './asset-properties-item.component';

function AssetPropertiesComponent_div_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 5);
    ɵngcc0.ɵɵelement(1, "c8y-loading");
    ɵngcc0.ɵɵelementEnd();
} }
function AssetPropertiesComponent_ng_container_7_div_1_div_3_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 13);
    ɵngcc0.ɵɵelementStart(1, "p", 14);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "humanize");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "button", 15);
    ɵngcc0.ɵɵlistener("click", function AssetPropertiesComponent_ng_container_7_div_1_div_3_Template_button_click_4_listener() { ɵngcc0.ɵɵrestoreView(_r9); const prop_r3 = ɵngcc0.ɵɵnextContext().$implicit; const ctx_r7 = ɵngcc0.ɵɵnextContext(2); return ctx_r7.toggleEdit(prop_r3); });
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelement(6, "i", 16);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const prop_r3 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(3, 2, prop_r3 == null ? null : prop_r3.label));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(5, 4, "Edit"));
} }
function AssetPropertiesComponent_ng_container_7_div_1_div_6_Template(rf, ctx) { if (rf & 1) {
    const _r13 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 17);
    ɵngcc0.ɵɵelementStart(1, "button", 18);
    ɵngcc0.ɵɵlistener("click", function AssetPropertiesComponent_ng_container_7_div_1_div_6_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r13); const prop_r3 = ɵngcc0.ɵɵnextContext().$implicit; const ctx_r11 = ɵngcc0.ɵɵnextContext(2); return ctx_r11.toggleEdit(prop_r3); });
    ɵngcc0.ɵɵtext(2, " Cancel ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "button", 19);
    ɵngcc0.ɵɵlistener("click", function AssetPropertiesComponent_ng_container_7_div_1_div_6_Template_button_click_3_listener() { ɵngcc0.ɵɵrestoreView(_r13); const prop_r3 = ɵngcc0.ɵɵnextContext().$implicit; const _r5 = ɵngcc0.ɵɵreference(5); const ctx_r14 = ɵngcc0.ɵɵnextContext(2); return ctx_r14.save(_r5.model, prop_r3); });
    ɵngcc0.ɵɵtext(4, " Save ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵnextContext();
    const _r5 = ɵngcc0.ɵɵreference(5);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("disabled", !(_r5 == null ? null : _r5.form == null ? null : _r5.form.valid));
} }
const _c0 = function (a0) { return { "card-highlight": a0 }; };
const _c1 = function (a0) { return { "p-b-0": a0 }; };
function AssetPropertiesComponent_ng_container_7_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 7);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelementStart(2, "div", 8);
    ɵngcc0.ɵɵtemplate(3, AssetPropertiesComponent_ng_container_7_div_1_div_3_Template, 7, 6, "div", 9);
    ɵngcc0.ɵɵelement(4, "c8y-asset-properties-item", 10, 11);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(6, AssetPropertiesComponent_ng_container_7_div_1_div_6_Template, 5, 1, "div", 12);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const prop_r3 = ctx.$implicit;
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 12, prop_r3.description));
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(14, _c0, prop_r3.isEdit));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(16, _c1, prop_r3.isEdit));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !prop_r3.isEdit);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("file", prop_r3.file)("key", prop_r3.key)("type", prop_r3.type)("value", prop_r3.value)("complex", prop_r3.complex)("isEdit", prop_r3.isEdit)("jsonSchema", prop_r3.jsonSchema);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", prop_r3.isEdit);
} }
function AssetPropertiesComponent_ng_container_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, AssetPropertiesComponent_ng_container_7_div_1_Template, 7, 18, "div", 6);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r1.customProperties);
} }
const _c2 = function (a0) { return { label: a0 }; };
export class AssetPropertiesComponent {
    constructor(assetTypes, inventory, inventoryBinary, alert, contextRouteService, activatedRoute) {
        this.assetTypes = assetTypes;
        this.inventory = inventory;
        this.inventoryBinary = inventoryBinary;
        this.alert = alert;
        this.contextRouteService = contextRouteService;
        this.activatedRoute = activatedRoute;
        this.properties = [];
        this.customProperties = [];
        this.isEdit = false;
        this.isLoading = false;
    }
    ngOnChanges(changes) {
        if (changes.asset) {
            // Back button handling, as component is not destroyed
            this.assetType = undefined;
            this.customProperties = [];
            this.loadAsset();
        }
    }
    loadAsset() {
        return __awaiter(this, void 0, void 0, function* () {
            this.isLoading = true;
            this.assetType = this.assetTypes.getAssetTypeByName(this.asset.type);
            this.customProperties = yield this.resolveCustomProperties(this.properties);
            this.isLoading = false;
        });
    }
    resolveCustomProperties(managedObjects) {
        return __awaiter(this, void 0, void 0, function* () {
            const properties = [];
            for (const mo of managedObjects) {
                if (mo.c8y_JsonSchema) {
                    const items = yield this.parseItem(mo, mo.c8y_JsonSchema.properties, this.asset);
                    properties.push(items[0]);
                }
            }
            return properties;
        });
    }
    parseItem(mo, properties, asset) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            if (!asset) {
                return [];
            }
            const keys = Object.keys(properties);
            const items = [];
            for (const key of keys) {
                let value = asset[key];
                const type = properties[key].type;
                const title = properties[key].title;
                let file;
                if (type === 'file' && value) {
                    const fileId = typeof value === 'object' ? (_b = (_a = value[0]) === null || _a === void 0 ? void 0 : _a.file) === null || _b === void 0 ? void 0 : _b.id : value;
                    const fileData = yield this.getFileManagedObject(fileId);
                    file = fileData;
                    value = [fileData];
                }
                items.push({
                    key,
                    value,
                    label: title || mo.label,
                    type,
                    description: mo.description,
                    file,
                    complex: type === 'object'
                        ? yield this.parseItem(mo, properties[key].properties, asset[key])
                        : undefined,
                    isEdit: false,
                    jsonSchema: mo.c8y_JsonSchema
                });
            }
            return items;
        });
    }
    toggleEdit(prop) {
        prop.isEdit = !prop.isEdit;
    }
    getFileManagedObject(id) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const { data } = yield this.inventory.detail(id);
                return data;
            }
            catch (ex) {
                this.alert.addServerFailure(ex);
            }
        });
    }
    save(model, prop) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                if (prop.type === 'object') {
                    model[prop.key] = yield this.uploadFiles(model[prop.key], prop.value);
                }
                else {
                    model = yield this.uploadFiles(model, prop.value);
                }
                const updatedAsset = Object.assign({ id: this.asset.id }, model);
                const { data } = yield this.inventory.update(updatedAsset);
                this.toggleEdit(prop);
                this.asset = data;
                yield this.loadAsset();
                this.contextRouteService.refreshContext();
                this.alert.success(gettext('Asset properties updated.'));
                this.contextRouteService.setContext(this.activatedRoute, data);
            }
            catch (ex) {
                this.alert.addServerFailure(ex);
            }
        });
    }
    uploadFiles(model, moId) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const keys = Object.keys(model);
            for (const key of keys) {
                if (Array.isArray(model[key]) && ((_a = model[key][0]) === null || _a === void 0 ? void 0 : _a.file) instanceof File) {
                    try {
                        const upload = yield this.inventoryBinary.create(model[key][0].file);
                        try {
                            if (moId && moId[0]) {
                                yield this.inventory.childAdditionsRemove(moId[0], this.asset.id);
                            }
                        }
                        catch (ex) {
                            this.alert.addServerFailure(ex);
                        }
                        model[key] = upload.data.id;
                        yield this.inventory.childAdditionsAdd(upload.data.id, this.asset.id);
                    }
                    catch (ex) {
                        this.alert.addServerFailure(ex);
                    }
                }
            }
            return model;
        });
    }
}
AssetPropertiesComponent.ɵfac = function AssetPropertiesComponent_Factory(t) { return new (t || AssetPropertiesComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AssetTypesService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.InventoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.InventoryBinaryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ContextRouteService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.ActivatedRoute)); };
AssetPropertiesComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AssetPropertiesComponent, selectors: [["c8y-asset-properties"]], inputs: { properties: "properties", asset: "asset" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], decls: 8, vars: 7, consts: [[1, "card-header", "bg-inherit", "separator", "sticky-top"], ["translate", "", 1, "card-title", "p-t-4", "p-b-4", 3, "translateParams"], [1, "card-block"], ["class", "text-center", 4, "ngIf"], [4, "ngIf"], [1, "text-center"], ["class", "card m-b-8", 3, "ngClass", "title", 4, "ngFor", "ngForOf"], [1, "card", "m-b-8", 3, "ngClass", "title"], [1, "card-block", 3, "ngClass"], ["class", "d-flex p-b-8", 4, "ngIf"], [3, "file", "key", "type", "value", "complex", "isEdit", "jsonSchema"], ["assetProps", ""], ["class", "card-footer p-t-0", 4, "ngIf"], [1, "d-flex", "p-b-8"], [1, "text-medium"], [1, "btn", "btn-clean", "text-primary", "m-l-auto", "text-12", 3, "title", "click"], ["c8yIcon", "pencil"], [1, "card-footer", "p-t-0"], ["type", "button", "translate", "", 1, "btn", "btn-default", "btn-sm", 3, "click"], ["type", "button", "translate", "", 1, "btn", "btn-primary", "btn-sm", 3, "disabled", "click"]], template: function AssetPropertiesComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementContainerStart(0);
        ɵngcc0.ɵɵelementStart(1, "div", 0);
        ɵngcc0.ɵɵelementStart(2, "h4", 1);
        ɵngcc0.ɵɵdisableBindings();
        ɵngcc0.ɵɵpipe(3, "humanize");
        ɵngcc0.ɵɵtext(4, " {{ label }} properties ");
        ɵngcc0.ɵɵenableBindings();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "div", 2);
        ɵngcc0.ɵɵtemplate(6, AssetPropertiesComponent_div_6_Template, 2, 0, "div", 3);
        ɵngcc0.ɵɵtemplate(7, AssetPropertiesComponent_ng_container_7_Template, 2, 1, "ng-container", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementContainerEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction1(5, _c2, ɵngcc0.ɵɵpipeBind1(3, 3, (ctx.assetType == null ? null : ctx.assetType.label) || "")));
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isLoading);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isLoading);
    } }, directives: [ɵngcc1.C8yTranslateDirective, ɵngcc4.NgIf, ɵngcc1.LoadingComponent, ɵngcc4.NgForOf, ɵngcc4.NgClass, ɵngcc5.AssetPropertiesItemComponent, ɵngcc1.IconDirective], pipes: [ɵngcc1.HumanizePipe, ɵngcc1.C8yTranslatePipe], encapsulation: 2 });
AssetPropertiesComponent.ctorParameters = () => [
    { type: AssetTypesService },
    { type: InventoryService },
    { type: InventoryBinaryService },
    { type: AlertService },
    { type: ContextRouteService },
    { type: ActivatedRoute }
];
AssetPropertiesComponent.propDecorators = {
    asset: [{ type: Input }],
    properties: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AssetPropertiesComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-asset-properties',
                template: "<ng-container>\n  <div class=\"card-header bg-inherit separator sticky-top\">\n    <h4\n      class=\"card-title p-t-4 p-b-4\"\n      ngNonBindable\n      translate\n      [translateParams]=\"{ label: (assetType?.label || '') | humanize }\"\n    >\n      {{ label }} properties\n    </h4>\n  </div>\n  <div class=\"card-block\">\n    <div class=\"text-center\" *ngIf=\"isLoading\">\n      <c8y-loading></c8y-loading>\n    </div>\n\n    <ng-container *ngIf=\"!isLoading\">\n      <div\n        class=\"card m-b-8\"\n        *ngFor=\"let prop of customProperties\"\n        [ngClass]=\"{ 'card-highlight': prop.isEdit }\"\n        title=\"{{ prop.description | translate }}\"\n      >\n        <div class=\"card-block \" [ngClass]=\"{'p-b-0': prop.isEdit}\">\n          <div class=\"d-flex p-b-8\" *ngIf=\"!prop.isEdit\">\n            <p class=\"text-medium\">{{ prop?.label | humanize }}</p>\n            <button\n              class=\"btn btn-clean text-primary m-l-auto text-12\"\n              title=\"{{ 'Edit' | translate }}\"\n              (click)=\"toggleEdit(prop)\"\n            >\n              <i c8yIcon=\"pencil\"></i>\n            </button>\n          </div>\n          <c8y-asset-properties-item\n            #assetProps\n            [file]=\"prop.file\"\n            [key]=\"prop.key\"\n            [type]=\"prop.type\"\n            [value]=\"prop.value\"\n            [complex]=\"prop.complex\"\n            [isEdit]=\"prop.isEdit\"\n            [jsonSchema]=\"prop.jsonSchema\"\n          ></c8y-asset-properties-item>\n        </div>\n        <div class=\"card-footer p-t-0\" *ngIf=\"prop.isEdit\">\n          <button type=\"button\" class=\"btn btn-default btn-sm\" translate (click)=\"toggleEdit(prop)\">\n            Cancel\n          </button>\n          <button\n            type=\"button\"\n            class=\"btn btn-primary btn-sm\"\n            translate\n            [disabled]=\"!assetProps?.form?.valid\"\n            (click)=\"save(assetProps.model, prop)\"\n          >\n            Save\n          </button>\n        </div>\n      </div>\n    </ng-container>\n  </div>\n</ng-container>\n"
            }]
    }], function () { return [{ type: ɵngcc1.AssetTypesService }, { type: ɵngcc2.InventoryService }, { type: ɵngcc2.InventoryBinaryService }, { type: ɵngcc1.AlertService }, { type: ɵngcc1.ContextRouteService }, { type: ɵngcc3.ActivatedRoute }]; }, { properties: [{
            type: Input
        }], asset: [{
            type: Input
        }] }); })();
