// @c8y/ngx-components/esm2015/sub-assets/sub-assets.component.js
import { __awaiter } from "tslib";
import { Component, EventEmitter } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { SubAssetsService } from './sub-assets.service';
import { AssetNodeService } from '@c8y/ngx-components/assets-navigator';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from './sub-assets.service';
import * as ɵngcc3 from '@c8y/ngx-components/assets-navigator';
import * as ɵngcc4 from '@c8y/ngx-components';
import * as ɵngcc5 from '@angular/common';
import * as ɵngcc6 from './group-info.component';
import * as ɵngcc7 from './sub-assets-grid.component';
import * as ɵngcc8 from './asset-properties.component';
import * as ɵngcc9 from './add-group/add-group.component';
import * as ɵngcc10 from './assign-devices/assign-devices.component';
import * as ɵngcc11 from './assign-devices/assign-child-devices.component';

function SubAssetsComponent_c8y_action_bar_item_2_Template(rf, ctx) { if (rf & 1) {
    const _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-action-bar-item", 12);
    ɵngcc0.ɵɵelementStart(1, "button", 13);
    ɵngcc0.ɵɵlistener("click", function SubAssetsComponent_c8y_action_bar_item_2_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r7); const ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.showAddGroup = !ctx_r6.showAddGroup; });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 14);
    ɵngcc0.ɵɵelementStart(4, "span", 15);
    ɵngcc0.ɵɵtext(5, "Add group");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("placement", "right");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 3, "Add group"));
    ɵngcc0.ɵɵproperty("disabled", ctx_r0.shouldDisableAddGroup);
} }
function SubAssetsComponent_c8y_action_bar_item_3_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-action-bar-item", 12);
    ɵngcc0.ɵɵelementStart(1, "button", 13);
    ɵngcc0.ɵɵlistener("click", function SubAssetsComponent_c8y_action_bar_item_3_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r9); const ctx_r8 = ɵngcc0.ɵɵnextContext(); return ctx_r8.showAssignDevices = !ctx_r8.showAssignDevices; });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 14);
    ɵngcc0.ɵɵelementStart(4, "span", 15);
    ɵngcc0.ɵɵtext(5, "Assign devices");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("placement", "right");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 3, "Assign devices"));
    ɵngcc0.ɵɵproperty("disabled", ctx_r1.shouldDisableAssignDevices);
} }
function SubAssetsComponent_div_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 16);
    ɵngcc0.ɵɵelement(1, "c8y-asset-properties", 17);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("properties", ctx_r2.customProperties)("asset", ctx_r2.group);
} }
function SubAssetsComponent_c8y_add_group_10_Template(rf, ctx) { if (rf & 1) {
    const _r11 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-add-group", 18);
    ɵngcc0.ɵɵlistener("onCancel", function SubAssetsComponent_c8y_add_group_10_Template_c8y_add_group_onCancel_0_listener() { ɵngcc0.ɵɵrestoreView(_r11); const ctx_r10 = ɵngcc0.ɵɵnextContext(); return ctx_r10.showAddGroup = false; });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("currentGroupId", ctx_r3.currentGroupId)("refresh", ctx_r3.refresh);
} }
function SubAssetsComponent_c8y_assign_devices_14_Template(rf, ctx) { if (rf & 1) {
    const _r13 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-assign-devices", 19);
    ɵngcc0.ɵɵlistener("onCancel", function SubAssetsComponent_c8y_assign_devices_14_Template_c8y_assign_devices_onCancel_0_listener() { ɵngcc0.ɵɵrestoreView(_r13); const ctx_r12 = ɵngcc0.ɵɵnextContext(); return ctx_r12.showAssignDevices = false; })("onShowChildDevices", function SubAssetsComponent_c8y_assign_devices_14_Template_c8y_assign_devices_onShowChildDevices_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r13); const ctx_r14 = ɵngcc0.ɵɵnextContext(); return ctx_r14.showAssignChildDevices = $event; })("selectedDevice", function SubAssetsComponent_c8y_assign_devices_14_Template_c8y_assign_devices_selectedDevice_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r13); const ctx_r15 = ɵngcc0.ɵɵnextContext(); return ctx_r15.showChildrenForDevice = $event; });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("refresh", ctx_r4.refresh)("currentGroupId", ctx_r4.currentGroupId);
} }
function SubAssetsComponent_c8y_assign_child_devices_18_Template(rf, ctx) { if (rf & 1) {
    const _r17 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-assign-child-devices", 20);
    ɵngcc0.ɵɵlistener("onCancel", function SubAssetsComponent_c8y_assign_child_devices_18_Template_c8y_assign_child_devices_onCancel_0_listener() { ɵngcc0.ɵɵrestoreView(_r17); const ctx_r16 = ɵngcc0.ɵɵnextContext(); return ctx_r16.showAssignChildDevices = false; });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("refresh", ctx_r5.refresh)("currentGroupId", ctx_r5.currentGroupId)("parentDevice", ctx_r5.showChildrenForDevice);
} }
const _c0 = function (a0, a1) { return { "card--grid grid__col--8-4--md grid__row--fit-auto": a0, "d-flex d-col": a1 }; };
const _c1 = function (a0) { return { drawerOpen: a0 }; };
export class SubAssetsComponent {
    constructor(route, subAssetsService, assetNodeService) {
        this.route = route;
        this.subAssetsService = subAssetsService;
        this.assetNodeService = assetNodeService;
        this.toggle = false;
        this.showAddGroup = false;
        this.showAssignDevices = false;
        this.showAssignChildDevices = false;
        this.refresh = new EventEmitter();
        this.filterable = true;
        this.shouldDisableAddGroup = false;
        this.shouldDisableAssignDevices = false;
        this.shouldShowAssetsProperties = false;
        this.customProperties = [];
        this.dataSub = this.route.parent.data.subscribe(({ contextData }) => {
            this.init(contextData);
        });
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            this.displayOptions = {
                striped: true,
                bordered: false,
                gridHeader: true,
                filter: !this.isSmartGroup()
            };
        });
    }
    init(contextData) {
        return __awaiter(this, void 0, void 0, function* () {
            this.title = contextData.name;
            this.currentGroupId = contextData.id;
            this.group = contextData;
            this.filterable = !this.assetNodeService.isDynamicGroup(contextData);
            this.shouldDisableAddGroup = !this.subAssetsService.canCreateGroup();
            this.shouldDisableAssignDevices = !(yield this.subAssetsService.canAssignDevice(this.group));
            this.customProperties = yield this.subAssetsService.getCustomProperties(this.group);
            this.shouldShowAssetsProperties = this.customProperties.length > 0;
        });
    }
    groupChange(group) {
        /**
         * Group is cached in the context-route guard. To avoid loading old data (when switching tabs) we need to update
         * this object in the cache.
         */
        Object.assign(this.group, group);
        this.title = group.name;
        this.refresh.emit();
    }
    isSmartGroup() {
        return this.subAssetsService.isSmartGroup(this.group);
    }
    ngOnDestroy() {
        if (this.dataSub) {
            this.dataSub.unsubscribe();
        }
    }
}
SubAssetsComponent.ɵfac = function SubAssetsComponent_Factory(t) { return new (t || SubAssetsComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ActivatedRoute), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.SubAssetsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.AssetNodeService)); };
SubAssetsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SubAssetsComponent, selectors: [["c8y-sub-assets"]], decls: 19, vars: 25, consts: [[3, "placement", 4, "ngIf"], [1, "content-fullpage", "shadow1", 3, "ngClass"], [1, "grid__col--fullspan", 3, "group", "onGroupChange"], [3, "refresh", "parent-group", "filterable", "displayOptions"], ["class", "inner-scroll bg-gray-white", 4, "ngIf"], [3, "ngClass"], [1, "bottom-drawer"], [3, "currentGroupId", "refresh", "onCancel", 4, "ngIf"], [1, "d-flex", "d-col", "no-align-items", "fit-h"], ["class", "d-contents", 3, "refresh", "currentGroupId", "onCancel", "onShowChildDevices", "selectedDevice", 4, "ngIf"], [1, "bottom-drawer", "m-t-40"], ["class", "d-contents", 3, "refresh", "currentGroupId", "parentDevice", "onCancel", 4, "ngIf"], [3, "placement"], [1, "btn", "btn-link", 3, "title", "disabled", "click"], ["c8yIcon", "plus-circle", 1, "m-r-4"], ["translate", ""], [1, "inner-scroll", "bg-gray-white"], [1, "d-contents", 3, "properties", "asset"], [3, "currentGroupId", "refresh", "onCancel"], [1, "d-contents", 3, "refresh", "currentGroupId", "onCancel", "onShowChildDevices", "selectedDevice"], [1, "d-contents", 3, "refresh", "currentGroupId", "parentDevice", "onCancel"]], template: function SubAssetsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(2, SubAssetsComponent_c8y_action_bar_item_2_Template, 6, 5, "c8y-action-bar-item", 0);
        ɵngcc0.ɵɵtemplate(3, SubAssetsComponent_c8y_action_bar_item_3_Template, 6, 5, "c8y-action-bar-item", 0);
        ɵngcc0.ɵɵelementStart(4, "div", 1);
        ɵngcc0.ɵɵelementStart(5, "c8y-group-info", 2);
        ɵngcc0.ɵɵlistener("onGroupChange", function SubAssetsComponent_Template_c8y_group_info_onGroupChange_5_listener($event) { return ctx.groupChange($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(6, "c8y-sub-assets-grid", 3);
        ɵngcc0.ɵɵtemplate(7, SubAssetsComponent_div_7_Template, 2, 2, "div", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(8, "div", 5);
        ɵngcc0.ɵɵelementStart(9, "div", 6);
        ɵngcc0.ɵɵtemplate(10, SubAssetsComponent_c8y_add_group_10_Template, 1, 2, "c8y-add-group", 7);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "div", 5);
        ɵngcc0.ɵɵelementStart(12, "div", 6);
        ɵngcc0.ɵɵelementStart(13, "div", 8);
        ɵngcc0.ɵɵtemplate(14, SubAssetsComponent_c8y_assign_devices_14_Template, 1, 2, "c8y-assign-devices", 9);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(15, "div", 5);
        ɵngcc0.ɵɵelementStart(16, "div", 10);
        ɵngcc0.ɵɵelementStart(17, "div", 8);
        ɵngcc0.ɵɵtemplate(18, SubAssetsComponent_c8y_assign_child_devices_18_Template, 1, 3, "c8y-assign-child-devices", 11);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ctx.title, "\n");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isSmartGroup());
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isSmartGroup());
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(16, _c0, ctx.shouldShowAssetsProperties, !ctx.shouldShowAssetsProperties));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("group", ctx.group);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("refresh", ctx.refresh)("parent-group", ctx.group)("filterable", ctx.filterable)("displayOptions", ctx.displayOptions);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.shouldShowAssetsProperties);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(19, _c1, ctx.showAddGroup));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showAddGroup);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(21, _c1, ctx.showAssignDevices));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showAssignDevices);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(23, _c1, ctx.showAssignChildDevices));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showAssignChildDevices);
    } }, directives: [ɵngcc4.TitleComponent, ɵngcc5.NgIf, ɵngcc5.NgClass, ɵngcc6.GroupInfoComponent, ɵngcc7.SubAssetsGridComponent, ɵngcc4.ActionBarItemComponent, ɵngcc4.IconDirective, ɵngcc4.C8yTranslateDirective, ɵngcc8.AssetPropertiesComponent, ɵngcc9.AddGroupComponent, ɵngcc10.AssignDevicesComponent, ɵngcc11.AssignChildDevicesComponent], pipes: [ɵngcc4.C8yTranslatePipe], encapsulation: 2 });
SubAssetsComponent.ctorParameters = () => [
    { type: ActivatedRoute },
    { type: SubAssetsService },
    { type: AssetNodeService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SubAssetsComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-sub-assets',
                template: "<c8y-title>\n  {{ title }}\n</c8y-title>\n\n<c8y-action-bar-item [placement]=\"'right'\" *ngIf=\"!isSmartGroup()\">\n  <button\n    class=\"btn btn-link\"\n    title=\"{{ 'Add group' | translate }}\"\n    (click)=\"showAddGroup = !showAddGroup\"\n    [disabled]=\"shouldDisableAddGroup\"\n  >\n    <i c8yIcon=\"plus-circle\" class=\"m-r-4\"></i><span translate>Add group</span>\n  </button>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\" *ngIf=\"!isSmartGroup()\">\n  <button\n    class=\"btn btn-link\"\n    title=\"{{ 'Assign devices' | translate }}\"\n    (click)=\"showAssignDevices = !showAssignDevices\"\n    [disabled]=\"shouldDisableAssignDevices\"\n  >\n    <i c8yIcon=\"plus-circle\" class=\"m-r-4\"></i><span translate>Assign devices</span>\n  </button>\n</c8y-action-bar-item>\n\n<div\n  class=\"content-fullpage shadow1\"\n  [ngClass]=\"{\n    'card--grid grid__col--8-4--md grid__row--fit-auto': shouldShowAssetsProperties,\n    'd-flex d-col': !shouldShowAssetsProperties\n  }\"\n>\n  <c8y-group-info\n    [group]=\"group\"\n    (onGroupChange)=\"groupChange($event)\"\n    class=\"grid__col--fullspan\"\n  ></c8y-group-info>\n  <c8y-sub-assets-grid\n    [refresh]=\"refresh\"\n    [parent-group]=\"group\"\n    [filterable]=\"filterable\"\n    [displayOptions]=\"displayOptions\"\n  ></c8y-sub-assets-grid>\n  <div class=\"inner-scroll bg-gray-white\" *ngIf=\"shouldShowAssetsProperties\">\n    <c8y-asset-properties\n      [properties]=\"customProperties\"\n      [asset]=\"group\"\n      class=\"d-contents\"\n    ></c8y-asset-properties>\n  </div>\n</div>\n\n<div [ngClass]=\"{ drawerOpen: showAddGroup }\">\n  <div class=\"bottom-drawer\">\n    <c8y-add-group\n      [currentGroupId]=\"currentGroupId\"\n      [refresh]=\"refresh\"\n      (onCancel)=\"showAddGroup = false\"\n      *ngIf=\"showAddGroup\"\n    ></c8y-add-group>\n  </div>\n</div>\n\n<div [ngClass]=\"{ drawerOpen: showAssignDevices }\">\n  <div class=\"bottom-drawer\">\n    <div class=\"d-flex d-col no-align-items fit-h\">\n      <c8y-assign-devices\n        class=\"d-contents\"\n        (onCancel)=\"showAssignDevices = false\"\n        [refresh]=\"refresh\"\n        [currentGroupId]=\"currentGroupId\"\n        (onShowChildDevices)=\"showAssignChildDevices = $event\"\n        (selectedDevice)=\"showChildrenForDevice = $event\"\n        *ngIf=\"showAssignDevices\"\n      ></c8y-assign-devices>\n    </div>\n  </div>\n</div>\n\n<div [ngClass]=\"{ drawerOpen: showAssignChildDevices }\">\n  <div class=\"bottom-drawer m-t-40\">\n    <div class=\"d-flex d-col no-align-items fit-h\">\n      <c8y-assign-child-devices\n        *ngIf=\"showAssignChildDevices\"\n        class=\"d-contents\"\n        (onCancel)=\"showAssignChildDevices = false\"\n        [refresh]=\"refresh\"\n        [currentGroupId]=\"currentGroupId\"\n        [parentDevice]=\"showChildrenForDevice\"\n      ></c8y-assign-child-devices>\n    </div>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.ActivatedRoute }, { type: ɵngcc2.SubAssetsService }, { type: ɵngcc3.AssetNodeService }]; }, null); })();
