// @c8y/ngx-components/esm2015/operations/create-bulk-operation-details/create-bulk-operation-details.module.js
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { CoreModule, FormsModule } from '@c8y/ngx-components';
import { BulkOperationSchedulerModule } from '@c8y/ngx-components/operations/bulk-operation-scheduler';
import { CreateBulkOperationDetailsComponent } from './create-bulk-operation-details.component';
/**
 * This module provides a component for entering operation details.
 */
import * as ɵngcc0 from '@angular/core';
export class CreateBulkOperationDetailsModule {
}
CreateBulkOperationDetailsModule.ɵfac = function CreateBulkOperationDetailsModule_Factory(t) { return new (t || CreateBulkOperationDetailsModule)(); };
CreateBulkOperationDetailsModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: CreateBulkOperationDetailsModule });
CreateBulkOperationDetailsModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CoreModule, FormsModule, ReactiveFormsModule, BulkOperationSchedulerModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CreateBulkOperationDetailsModule, [{
        type: NgModule,
        args: [{
                imports: [CoreModule, FormsModule, ReactiveFormsModule, BulkOperationSchedulerModule],
                declarations: [CreateBulkOperationDetailsComponent],
                exports: [CreateBulkOperationDetailsComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(CreateBulkOperationDetailsModule, { declarations: function () { return [CreateBulkOperationDetailsComponent]; }, imports: function () { return [CoreModule, FormsModule, ReactiveFormsModule, BulkOperationSchedulerModule]; }, exports: function () { return [CreateBulkOperationDetailsComponent]; } }); })();
