// @c8y/ngx-components/esm2015/core/dynamic-forms/array/array.type.component.js
import { Component } from '@angular/core';
import { FieldArrayType } from '@ngx-formly/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '../../common/icon.directive';
import * as ɵngcc3 from 'ngx-bootstrap/popover';
import * as ɵngcc4 from '@ngx-formly/core';
import * as ɵngcc5 from '../../i18n/c8y-translate.directive';
import * as ɵngcc6 from '../../i18n/c8y-translate.pipe';
import * as ɵngcc7 from '../../common/humanize.pipe';

function ArrayTypeComponent_div_1_button_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "button", 11);
    ɵngcc0.ɵɵelement(1, "i", 12);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("popover", ctx_r5.to.description)("outsideClick", true);
} }
function ArrayTypeComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 9);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "humanize");
    ɵngcc0.ɵɵtemplate(3, ArrayTypeComponent_div_1_button_3_Template, 2, 2, "button", 10);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("id", ctx_r0.id + "-fieldset");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 3, ctx_r0.to.label), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.to.description);
} }
function ArrayTypeComponent_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 13);
    ɵngcc0.ɵɵelement(1, "formly-validation-message", 14);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("field", ctx_r1.field);
} }
const _c0 = function (a0) { return { "m-t-24": a0 }; };
function ArrayTypeComponent_div_3_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 15);
    ɵngcc0.ɵɵelement(1, "formly-field", 14);
    ɵngcc0.ɵɵelementStart(2, "div", 16);
    ɵngcc0.ɵɵelementStart(3, "button", 17);
    ɵngcc0.ɵɵlistener("click", function ArrayTypeComponent_div_3_Template_button_click_3_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r9); const i_r7 = restoredCtx.index; const ctx_r8 = ɵngcc0.ɵɵnextContext(); return ctx_r8.remove(i_r7); });
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelement(5, "i", 18);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ifield_r6 = ctx.$implicit;
    const i_r7 = ctx.index;
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("field", ifield_r6);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(6, _c0, i_r7 === 0));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(4, 4, "Delete"));
    ɵngcc0.ɵɵproperty("disabled", ctx_r2.to.disabled || ctx_r2.field.fieldGroup.length === 1);
} }
function ArrayTypeComponent_span_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 19);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(1, " Add {{ label }} ");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("translateParams", ctx_r3.field == null ? null : ctx_r3.field.fieldArray == null ? null : ctx_r3.field.fieldArray.templateOptions);
} }
function ArrayTypeComponent_span_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, ctx_r4.field.templateOptions == null ? null : ctx_r4.field.templateOptions.addText), " ");
} }
export class ArrayTypeComponent extends FieldArrayType {
    add() {
        this.formControl.markAsTouched();
        super.add();
    }
    remove(i) {
        this.formControl.markAsTouched();
        super.remove(i);
    }
    ngOnInit() {
        if (this.to.required && !this.field.fieldGroup.length) {
            this.add();
        }
    }
}
ArrayTypeComponent.ɵfac = /*@__PURE__*/ function () { let ɵArrayTypeComponent_BaseFactory; return function ArrayTypeComponent_Factory(t) { return (ɵArrayTypeComponent_BaseFactory || (ɵArrayTypeComponent_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(ArrayTypeComponent)))(t || ArrayTypeComponent); }; }();
ArrayTypeComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ArrayTypeComponent, selectors: [["c8y-array-type"]], features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 11, vars: 14, consts: [["role", "group"], ["class", "legend form-block", 3, "id", 4, "ngIf"], ["class", "form-control-feedback-message has-error", "role", "alert", 4, "ngIf"], ["class", "input-group-array", 4, "ngFor", "ngForOf"], [1, "p-t-8"], ["type", "button", 1, "btn", "btn-default", 3, "title", "disabled", "click"], ["c8yIcon", "plus-circle"], ["ngNonBindable", "", "translate", "", 3, "translateParams", 4, "ngIf"], [4, "ngIf"], [1, "legend", "form-block", 3, "id"], ["class", "btn-clean text-14 m-l-4 m-t-auto m-b-auto", "placement", "right", "type", "button", 3, "popover", "outsideClick", 4, "ngIf"], ["placement", "right", "type", "button", 1, "btn-clean", "text-14", "m-l-4", "m-t-auto", "m-b-auto", 3, "popover", "outsideClick"], ["c8yIcon", "question-circle-o", 1, "text-primary"], ["role", "alert", 1, "form-control-feedback-message", "has-error"], [3, "field"], [1, "input-group-array"], [1, "input-group-btn", 3, "ngClass"], ["type", "button", 1, "btn", "btn-clean", "m-l-auto", 3, "title", "disabled", "click"], ["c8yIcon", "minus-circle", 1, "text-danger"], ["translate", "", 3, "translateParams"]], template: function ArrayTypeComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, ArrayTypeComponent_div_1_Template, 4, 5, "div", 1);
        ɵngcc0.ɵɵtemplate(2, ArrayTypeComponent_div_2_Template, 2, 1, "div", 2);
        ɵngcc0.ɵɵtemplate(3, ArrayTypeComponent_div_3_Template, 6, 8, "div", 3);
        ɵngcc0.ɵɵelementStart(4, "div", 4);
        ɵngcc0.ɵɵelementStart(5, "button", 5);
        ɵngcc0.ɵɵlistener("click", function ArrayTypeComponent_Template_button_click_5_listener() { return ctx.add(); });
        ɵngcc0.ɵɵpipe(6, "translate");
        ɵngcc0.ɵɵpipe(7, "translate");
        ɵngcc0.ɵɵelement(8, "i", 6);
        ɵngcc0.ɵɵtemplate(9, ArrayTypeComponent_span_9_Template, 2, 1, "span", 7);
        ɵngcc0.ɵɵtemplate(10, ArrayTypeComponent_span_10_Template, 3, 3, "span", 8);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵattribute("aria-labelledby", ctx.id + "-fieldset");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.to.label || ctx.to.description);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showError && ctx.formControl.errors);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.field.fieldGroup);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind2(6, 9, ɵngcc0.ɵɵpipeBind1(7, 12, ctx.field.templateOptions == null ? null : ctx.field.templateOptions.addText) || "Add {{ label }}", ctx.field == null ? null : ctx.field.fieldArray == null ? null : ctx.field.fieldArray.templateOptions));
        ɵngcc0.ɵɵproperty("disabled", ctx.to.disabled);
        ɵngcc0.ɵɵattribute("data-cy", (ctx.field.templateOptions == null ? null : ctx.field.templateOptions.addTextDataCy) || "add-array-item");
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngIf", !(ctx.field.templateOptions == null ? null : ctx.field.templateOptions.addText));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.field.templateOptions == null ? null : ctx.field.templateOptions.addText);
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgForOf, ɵngcc2.IconDirective, ɵngcc3.PopoverDirective, ɵngcc4.ɵc, ɵngcc4.FormlyField, ɵngcc1.NgClass, ɵngcc5.C8yTranslateDirective], pipes: [ɵngcc6.C8yTranslatePipe, ɵngcc7.HumanizePipe], encapsulation: 2 });
ArrayTypeComponent.CONFIG = {
    types: [{ name: 'array', component: ArrayTypeComponent }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ArrayTypeComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-array-type',
                template: "<div role=\"group\" [attr.aria-labelledby]=\"id + '-fieldset'\">\n  <div id=\"{{ id + '-fieldset' }}\" class=\"legend form-block\" *ngIf=\"to.label || to.description\">\n    {{ to.label | humanize }}\n    <button\n      *ngIf=\"to.description\"\n      class=\"btn-clean text-14 m-l-4 m-t-auto m-b-auto\"\n      [popover]=\"to.description\"\n      placement=\"right\"\n      [outsideClick]=\"true\"\n      type=\"button\"\n    >\n      <i c8yIcon=\"question-circle-o\" class=\"text-primary\"></i>\n    </button>\n  </div>\n\n  <div\n    class=\"form-control-feedback-message has-error\"\n    role=\"alert\"\n    *ngIf=\"showError && formControl.errors\"\n  >\n    <formly-validation-message [field]=\"field\"></formly-validation-message>\n  </div>\n\n  <div *ngFor=\"let ifield of field.fieldGroup; let i = index\" class=\"input-group-array\">\n    <formly-field [field]=\"ifield\"></formly-field>\n    <div class=\"input-group-btn\" [ngClass]=\"{ 'm-t-24': i === 0 }\">\n      <button\n        type=\"button\"\n        class=\"btn btn-clean m-l-auto\"\n        title=\"{{ 'Delete' | translate }}\"\n        [disabled]=\"to.disabled || field.fieldGroup.length === 1\"\n        (click)=\"remove(i)\"\n      >\n        <i c8yIcon=\"minus-circle\" class=\"text-danger\"></i>\n      </button>\n    </div>\n  </div>\n\n  <div class=\"p-t-8\">\n    <button\n      type=\"button\"\n      title=\"{{\n        (field.templateOptions?.addText | translate) || 'Add \\{\\{ label \\}\\}'\n          | translate: field?.fieldArray?.templateOptions\n      }}\"\n      class=\"btn btn-default\"\n      [disabled]=\"to.disabled\"\n      (click)=\"add()\"\n      [attr.data-cy]=\"field.templateOptions?.addTextDataCy || 'add-array-item'\"\n    >\n      <i c8yIcon=\"plus-circle\"></i>\n      <span\n        *ngIf=\"!field.templateOptions?.addText\"\n        ngNonBindable\n        translate\n        [translateParams]=\"field?.fieldArray?.templateOptions\"\n      >\n        Add {{ label }}\n      </span>\n      <span *ngIf=\"field.templateOptions?.addText\">\n        {{ field.templateOptions?.addText | translate }}\n      </span>\n    </button>\n  </div>\n</div>\n"
            }]
    }], null, null); })();
