// @c8y/ngx-components/esm2015/core/drop-area/drop-area.component.js
import { __awaiter } from "tslib";
import { ChangeDetectorRef, Component, EventEmitter, HostListener, Input, Output, ViewChild } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { TranslateService } from '@ngx-translate/core';
import { get, map, some } from 'lodash-es';
import { BytesPipe } from '../common/bytes.pipe';
import { FilesService } from '../common/files.service';
import { gettext } from '../i18n/gettext';
/**
 * A drop-zone which is a file selector allowing users to select file(s) from their file system, either natively or by drag and drop.
 *
 * ## Example:
 *
 * ```html
 *  <div>
 *    <c8y-drop-area
 *      (dropped)="uploadFile($event)"
 *      [icon]="'upload'"
 *      [accept]="'.zip,.7z,video'">
 *    </c8y-drop-area>
 *  </div>
 * ```
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../common/files.service';
import * as ɵngcc2 from '@ngx-translate/core';
import * as ɵngcc3 from '../common/bytes.pipe';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from '../i18n/c8y-translate.directive';
import * as ɵngcc6 from '../common/icon.directive';
import * as ɵngcc7 from '../i18n/c8y-translate.pipe';

const _c0 = ["area"];
const _c1 = ["zone"];
const _c2 = ["picker"];
function DropAreaComponent_div_0_div_3_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 14);
    ɵngcc0.ɵɵelement(1, "div", 15);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵstyleProp("width", ctx_r7.progress + "%");
} }
function DropAreaComponent_div_0_div_3_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 16);
} }
function DropAreaComponent_div_0_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 10);
    ɵngcc0.ɵɵtemplate(1, DropAreaComponent_div_0_div_3_div_1_Template, 2, 2, "div", 11);
    ɵngcc0.ɵɵtemplate(2, DropAreaComponent_div_0_div_3_div_2_Template, 1, 0, "div", 12);
    ɵngcc0.ɵɵelementStart(3, "p", 13);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r5.progress !== -1);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r5.progress === -1);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 3, ctx_r5.loadingMessage), " ");
} }
function DropAreaComponent_div_0_div_4_p_2_span_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 21);
    ɵngcc0.ɵɵtext(1, "or click to browse your file system.");
    ɵngcc0.ɵɵelementEnd();
} }
function DropAreaComponent_div_0_div_4_p_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p");
    ɵngcc0.ɵɵelementStart(1, "b");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(4, "br");
    ɵngcc0.ɵɵtemplate(5, DropAreaComponent_div_0_div_4_p_2_span_5_Template, 2, 0, "span", 20);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r9 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(3, 2, ctx_r9.message));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r9.alwaysShow && ctx_r9.clickToOpen);
} }
function DropAreaComponent_div_0_div_4_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 22);
    ɵngcc0.ɵɵelementStart(1, "p", 23);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r10 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 1, ctx_r10.errorMessage), " ");
} }
function DropAreaComponent_div_0_div_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 17);
    ɵngcc0.ɵɵelement(1, "i");
    ɵngcc0.ɵɵtemplate(2, DropAreaComponent_div_0_div_4_p_2_Template, 6, 4, "p", 18);
    ɵngcc0.ɵɵtemplate(3, DropAreaComponent_div_0_div_4_div_3_Template, 4, 3, "div", 19);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵclassMapInterpolate1("dlt-c8y-icon-", ctx_r6.icon, "");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r6.errors);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r6.errors);
} }
const _c3 = function (a0) { return { "has-errors": a0 }; };
const _c4 = function (a0) { return { "drag-over": a0 }; };
function DropAreaComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    const _r13 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 5, 6);
    ɵngcc0.ɵɵlistener("dragleave", function DropAreaComponent_div_0_Template_div_dragleave_0_listener() { ɵngcc0.ɵɵrestoreView(_r13); const ctx_r12 = ɵngcc0.ɵɵnextContext(); return ctx_r12.stopDragging(); })("drop", function DropAreaComponent_div_0_Template_div_drop_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r13); const ctx_r14 = ɵngcc0.ɵɵnextContext(); return ctx_r14.onDrop($event); })("dragover", function DropAreaComponent_div_0_Template_div_dragover_0_listener() { ɵngcc0.ɵɵrestoreView(_r13); const ctx_r15 = ɵngcc0.ɵɵnextContext(); return ctx_r15.onOver(); })("click", function DropAreaComponent_div_0_Template_div_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r13); const ctx_r16 = ɵngcc0.ɵɵnextContext(); return ctx_r16.showPicker($event); });
    ɵngcc0.ɵɵelementStart(2, "div", 7);
    ɵngcc0.ɵɵtemplate(3, DropAreaComponent_div_0_div_3_Template, 6, 5, "div", 8);
    ɵngcc0.ɵɵtemplate(4, DropAreaComponent_div_0_div_4_Template, 4, 5, "div", 9);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵstyleProp("pointer-events", ctx_r0.loading ? "none" : "auto")("display", ctx_r0.isOver || ctx_r0.alwaysShow || ctx_r0.loading ? "block" : "none");
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(8, _c3, ctx_r0.errors));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(10, _c4, ctx_r0.isOver));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.loading);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r0.loading);
} }
function DropAreaComponent_div_1_div_1_div_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 31);
    ɵngcc0.ɵɵelement(1, "div", 15);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r19 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵstyleProp("width", ctx_r19.progress + "%");
} }
function DropAreaComponent_div_1_div_1_div_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 32);
    ɵngcc0.ɵɵelement(1, "div", 33);
    ɵngcc0.ɵɵelement(2, "div", 34);
    ɵngcc0.ɵɵelement(3, "div", 35);
    ɵngcc0.ɵɵelement(4, "div", 36);
    ɵngcc0.ɵɵelement(5, "div", 37);
    ɵngcc0.ɵɵelementEnd();
} }
function DropAreaComponent_div_1_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 27);
    ɵngcc0.ɵɵelementStart(1, "p", 28);
    ɵngcc0.ɵɵelementStart(2, "strong");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(5, DropAreaComponent_div_1_div_1_div_5_Template, 2, 2, "div", 29);
    ɵngcc0.ɵɵtemplate(6, DropAreaComponent_div_1_div_1_div_6_Template, 6, 0, "div", 30);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r17 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 3, ctx_r17.loadingMessage), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r17.progress !== -1);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r17.progress === -1);
} }
function DropAreaComponent_div_1_div_2_Template(rf, ctx) { if (rf & 1) {
    const _r22 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 38);
    ɵngcc0.ɵɵelementStart(1, "div", 39);
    ɵngcc0.ɵɵelement(2, "i", 40);
    ɵngcc0.ɵɵelementStart(3, "span", 41);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "button", 42);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementStart(7, "i", 43);
    ɵngcc0.ɵɵlistener("click", function DropAreaComponent_div_1_div_2_Template_i_click_7_listener() { ɵngcc0.ɵɵrestoreView(_r22); const ctx_r21 = ɵngcc0.ɵɵnextContext(2); return ctx_r21.onDelete(); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r18 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ctx_r18.filesNameString);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r18.filesNameString, " ");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(6, 3, "Remove"));
} }
function DropAreaComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 24);
    ɵngcc0.ɵɵtemplate(1, DropAreaComponent_div_1_div_1_Template, 7, 5, "div", 25);
    ɵngcc0.ɵɵtemplate(2, DropAreaComponent_div_1_div_2_Template, 8, 5, "div", 26);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵstyleProp("display", ctx_r1.isOver || ctx_r1.alwaysShow || ctx_r1.loading ? "block" : "none");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.loading);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r1.loading);
} }
function DropAreaComponent_input_2_Template(rf, ctx) { if (rf & 1) {
    const _r25 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "input", 44, 45);
    ɵngcc0.ɵɵlistener("change", function DropAreaComponent_input_2_Template_input_change_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r25); const ctx_r24 = ɵngcc0.ɵɵnextContext(); return ctx_r24.onPick($event); })("click", function DropAreaComponent_input_2_Template_input_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r25); const _r23 = ɵngcc0.ɵɵreference(1); return _r23.focus(); })("blur", function DropAreaComponent_input_2_Template_input_blur_0_listener() { ɵngcc0.ɵɵrestoreView(_r25); const ctx_r27 = ɵngcc0.ɵɵnextContext(); return ctx_r27.onTouched(); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("accept", ctx_r2.acceptedExts)("multiple", ctx_r2.maxAllowedFiles > 1);
} }
const _c5 = ["*"];
export class DropAreaComponent {
    constructor(cd, filesService, translate, bytes) {
        this.cd = cd;
        this.filesService = filesService;
        this.translate = translate;
        this.bytes = bytes;
        this.title = gettext('Upload file');
        this.message = gettext('Drop file here');
        this.icon = 'plus-square';
        this.loadingMessage = gettext('Uploading…');
        this.forceHideList = false;
        /** Affects displaying both the drop zone and the list of dropped files. */
        this.alwaysShow = false;
        this.clickToOpen = true;
        this.loading = false;
        /**
         * Current progress of the upload as a percentage. If not given a spinner will be displayed.
         */
        this.progress = -1; // -1 = spinner
        this.dropped = new EventEmitter();
        this.maxAllowedFiles = Infinity;
        this.isOver = false;
        this.errors = false;
        this.onChange = _ => undefined;
        this.onTouched = () => undefined;
    }
    onkeyup(event) {
        if (event.key === 'Enter') {
            this.picker.nativeElement.click();
        }
    }
    ngOnInit() {
        this.acceptedExts = this.filesService.extractFileExtensions(this.accept).map(t => `.${t}`);
        this.alwaysShow = this.alwaysShow || this.area.nativeElement.children.length === 0;
        if (this.files && this.isFilesAnObjectOrArray() && this.files.length > 0) {
            this.onFilesSelected(this.files);
        }
    }
    /**
     * Toggles the style of the drop zone element when a file is dragged over the component.
     */
    toggle($event) {
        this.zone.nativeElement.style.height = this.area.nativeElement.offsetHeight + 'px';
        this.onOver();
    }
    /**
     * Shows computer browser with files to drop into drop-area zone.
     */
    showPicker($event) {
        this.preventDefault($event);
        this.picker.nativeElement.value = '';
        this.picker.nativeElement.click();
    }
    /**
     * Triggered when file is on over drop area, but not dropped.
     */
    onOver() {
        if (!this.isOver) {
            this.isOver = true;
            document.addEventListener('dragover', this.preventDefault);
            document.addEventListener('drop', this.preventDefault);
        }
    }
    /**
     * Triggered when file is dropped.
     */
    onPick($event) {
        this.errors = false;
        this.preventDefault($event);
        this.onFilesSelected($event.target.files);
    }
    /**
     * Handle file when it is dropped into drop-area.
     */
    onDrop($event) {
        this.preventDefault($event);
        this.onFilesSelected($event.dataTransfer.files);
        this.stopDragging();
    }
    /**
     * Checks condition what should be displayed: drop-area zone or list of dropped files.
     */
    shouldShowFilesList() {
        return (this.isFilesAnObjectOrArray() &&
            !this.forceHideList &&
            this.alwaysShow &&
            !this.isFilesArrayEmpty() &&
            !this.hasEmptyFiles() &&
            !this.isTooManyFiles());
    }
    /**
     * Triggered when file is picked over web application.
     */
    stopDragging() {
        document.removeEventListener('dragover', this.preventDefault);
        document.removeEventListener('drop', this.preventDefault);
        this.isOver = false;
    }
    /**
     * Delete files already dropped files.
     */
    onDelete() {
        delete this.files;
        delete this.filesNameString;
        this.clearErrors();
        this.dropped.emit(null);
        this.onChange(null);
        this.cd.markForCheck();
    }
    writeValue(value) {
        this.files = value;
        if (!value) {
            this.onDelete();
        }
        this.cd.detectChanges();
    }
    registerOnChange(fn) {
        this.onChange = fn;
    }
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    onFilesSelected(files) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            const hasValidNameLength = this.filesService.checkMaxLength(files);
            if (!hasValidNameLength) {
                this.onFileInvalidNameLength();
                return;
            }
            const haveValidTypes = this.filesService.haveValidExtensions(files, this.accept);
            if (!haveValidTypes) {
                this.onFileInvalidType();
                return;
            }
            const haveValidSizes = yield this.filesService.haveValidSizes(files);
            if (!haveValidSizes) {
                yield this.onFileInvalidSize();
                return;
            }
            this.files = files;
            this.filesNameString = this.getFilesNamesAsString(files);
            this.errors = false;
            if (this.isTooManyFiles()) {
                this.errors = true;
                (_a = this.formControl) === null || _a === void 0 ? void 0 : _a.setErrors({ tooManyFiles: true });
                this.errorMessage = gettext('Too many files selected.');
                return;
            }
            if (this.hasEmptyFiles()) {
                this.errors = true;
                (_b = this.formControl) === null || _b === void 0 ? void 0 : _b.setErrors({ emptyFiles: true });
                this.errorMessage = gettext('File must not be empty, select another one.');
                return;
            }
            const droppedFiles = this.compose(files);
            this.dropped.emit(droppedFiles);
            this.onChange(droppedFiles);
            this.cd.markForCheck();
        });
    }
    onFileInvalidNameLength() {
        var _a;
        this.errors = true;
        (_a = this.formControl) === null || _a === void 0 ? void 0 : _a.setErrors({ invalidNameLength: true });
        this.errorMessage = gettext('The filename is too long.');
    }
    onFileInvalidType() {
        var _a;
        this.errors = true;
        (_a = this.formControl) === null || _a === void 0 ? void 0 : _a.setErrors({ invalidType: true });
        this.errorMessage = gettext('The selected file is not supported.');
    }
    onFileInvalidSize() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const msg = gettext('The selected file is too large. The size limit is {{ limit }}.');
            const limit = this.bytes.transform(yield this.filesService.loadBytesSizeLimit());
            this.errors = true;
            (_a = this.formControl) === null || _a === void 0 ? void 0 : _a.setErrors({ invalidSize: true });
            this.errorMessage = this.translate.instant(msg, { limit });
        });
    }
    getFilesNamesAsString(files) {
        return map(files, ({ name }) => name).join(', ');
    }
    isFilesArrayEmpty() {
        return get(this, 'files.length', 0) === 0;
    }
    isTooManyFiles() {
        return get(this, 'files.length', 0) > this.maxAllowedFiles;
    }
    isFilesAnObjectOrArray() {
        return typeof this.files === 'object';
    }
    hasEmptyFiles() {
        let result = true;
        if (!this.isFilesArrayEmpty()) {
            result = this.isAnyFileEmpty();
        }
        return result;
    }
    isAnyFileEmpty() {
        return some(Array.from(this.files), ['size', 0]);
    }
    clearErrors() {
        var _a;
        delete this.errorMessage;
        this.errors = false;
        (_a = this.formControl) === null || _a === void 0 ? void 0 : _a.setErrors(null);
    }
    preventDefault($event) {
        if ($event) {
            $event.preventDefault();
        }
    }
    compose(files) {
        return Array.from(files).map(file => ({
            file,
            readAsJson: () => __awaiter(this, void 0, void 0, function* () { return JSON.parse(yield this.read(file, ReadAsType.TEXT)); }),
            readAsText: () => __awaiter(this, void 0, void 0, function* () { return this.read(file, ReadAsType.TEXT); }),
            readAsArrayBuffer: () => __awaiter(this, void 0, void 0, function* () { return this.read(file, ReadAsType.ARRAY_BUFFER); }),
            readAsBinaryString: () => __awaiter(this, void 0, void 0, function* () { return this.read(file, ReadAsType.BINARY_STRING); }),
            readAsDataURL: () => __awaiter(this, void 0, void 0, function* () { return this.read(file, ReadAsType.DATA_URL); })
        }));
    }
    read(file, type) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => {
                const reader = new FileReader();
                switch (type) {
                    case ReadAsType.TEXT: {
                        reader.readAsText(file);
                        break;
                    }
                    case ReadAsType.ARRAY_BUFFER: {
                        reader.readAsArrayBuffer(file);
                        break;
                    }
                    case ReadAsType.BINARY_STRING: {
                        reader.readAsBinaryString(file);
                        break;
                    }
                    case ReadAsType.DATA_URL: {
                        reader.readAsDataURL(file);
                        break;
                    }
                }
                reader.onload = () => this.onLoad(reader, resolve, reject);
            });
        });
    }
    onLoad(reader, resolve, reject) {
        if (reader.readyState !== 2) {
            return;
        }
        if (reader.error) {
            reject(reader.error);
        }
        resolve(reader.result);
    }
}
DropAreaComponent.ɵfac = function DropAreaComponent_Factory(t) { return new (t || DropAreaComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.FilesService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.TranslateService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.BytesPipe)); };
DropAreaComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: DropAreaComponent, selectors: [["c8y-drop-area"]], viewQuery: function DropAreaComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 7);
        ɵngcc0.ɵɵviewQuery(_c1, 5);
        ɵngcc0.ɵɵviewQuery(_c2, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.area = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.zone = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.picker = _t.first);
    } }, hostBindings: function DropAreaComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("keyup", function DropAreaComponent_keyup_HostBindingHandler($event) { return ctx.onkeyup($event); });
    } }, inputs: { title: "title", message: "message", icon: "icon", loadingMessage: "loadingMessage", forceHideList: "forceHideList", alwaysShow: "alwaysShow", clickToOpen: "clickToOpen", loading: "loading", progress: "progress", maxAllowedFiles: "maxAllowedFiles", files: "files", formControl: "formControl", accept: "accept" }, outputs: { dropped: "dropped" }, features: [ɵngcc0.ɵɵProvidersFeature([{ provide: NG_VALUE_ACCESSOR, useExisting: DropAreaComponent, multi: true }])], ngContentSelectors: _c5, decls: 6, vars: 4, consts: [["class", "drop-zone", "tabindex", "0", 3, "ngClass", "pointerEvents", "display", "dragleave", "drop", "dragover", "click", 4, "ngIf"], ["class", "drop-zone", "tabindex", "0", 3, "display", 4, "ngIf"], ["type", "file", "style", "opacity: 0; filter: alpha(opacity = 0); height: 0px", 3, "accept", "multiple", "change", "click", "blur", 4, "ngIf"], [3, "hidden", "dragover"], ["area", ""], ["tabindex", "0", 1, "drop-zone", 3, "ngClass", "dragleave", "drop", "dragover", "click"], ["zone", ""], [1, "file-placeholder", 3, "ngClass"], ["class", "d-flex p-4 flex-center", 4, "ngIf"], ["class", "hint-placeholder pointer", 4, "ngIf"], [1, "d-flex", "p-4", "flex-center"], ["class", "progress progress-striped active m-0", "style", "min-width: 50%", 4, "ngIf"], ["class", "spinner-snake", 4, "ngIf"], [1, "flex-item-middle", "m-r-8"], [1, "progress", "progress-striped", "active", "m-0", 2, "min-width", "50%"], ["role", "progressbar", "aria-valuenow", "0", "aria-valuemin", "0", "aria-valuemax", "100", 1, "progress-bar"], [1, "spinner-snake"], [1, "hint-placeholder", "pointer"], [4, "ngIf"], ["class", "has-errors", 4, "ngIf"], ["translate", "", 4, "ngIf"], ["translate", ""], [1, "has-errors"], [1, "form-control-feedback-message"], ["tabindex", "0", 1, "drop-zone"], ["class", "p-absolute p-4 fit-w fit-h d-flex d-col j-c-center a-i-center", 4, "ngIf"], ["class", "file-placeholder p-4", 4, "ngIf"], [1, "p-absolute", "p-4", "fit-w", "fit-h", "d-flex", "d-col", "j-c-center", "a-i-center"], [1, "m-b-8"], ["class", "progress progress-striped active m-0", "style", "min-width: 80%", 4, "ngIf"], ["class", "spinner p-relative m-0", 4, "ngIf"], [1, "progress", "progress-striped", "active", "m-0", 2, "min-width", "80%"], [1, "spinner", "p-relative", "m-0"], [1, "rect1"], [1, "rect2"], [1, "rect3"], [1, "rect4"], [1, "rect5"], [1, "file-placeholder", "p-4"], [1, "flex-row", "p-4"], ["c8yIcon", "file-o", 1, "m-r-8"], [1, "text-truncate", 3, "title"], [1, "btn", "btn-clean", "showOnHover", "flex-item-right", 3, "title"], ["c8yIcon", "minus-circle", 1, "text-danger", 3, "click"], ["type", "file", 2, "opacity", "0", "filter", "alpha(opacity = 0)", "height", "0px", 3, "accept", "multiple", "change", "click", "blur"], ["picker", ""]], template: function DropAreaComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, DropAreaComponent_div_0_Template, 5, 12, "div", 0);
        ɵngcc0.ɵɵtemplate(1, DropAreaComponent_div_1_Template, 3, 4, "div", 1);
        ɵngcc0.ɵɵtemplate(2, DropAreaComponent_input_2_Template, 2, 2, "input", 2);
        ɵngcc0.ɵɵelementStart(3, "div", 3, 4);
        ɵngcc0.ɵɵlistener("dragover", function DropAreaComponent_Template_div_dragover_3_listener($event) { return ctx.toggle($event); });
        ɵngcc0.ɵɵprojection(5);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", !ctx.shouldShowFilesList());
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.shouldShowFilesList());
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.clickToOpen);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("hidden", ctx.isOver || ctx.loading);
    } }, directives: [ɵngcc4.NgIf, ɵngcc4.NgClass, ɵngcc5.C8yTranslateDirective, ɵngcc6.IconDirective], pipes: [ɵngcc7.C8yTranslatePipe], encapsulation: 2 });
DropAreaComponent.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: FilesService },
    { type: TranslateService },
    { type: BytesPipe }
];
DropAreaComponent.propDecorators = {
    formControl: [{ type: Input }],
    title: [{ type: Input }],
    message: [{ type: Input }],
    icon: [{ type: Input }],
    loadingMessage: [{ type: Input }],
    forceHideList: [{ type: Input }],
    alwaysShow: [{ type: Input }],
    clickToOpen: [{ type: Input }],
    loading: [{ type: Input }],
    progress: [{ type: Input }],
    dropped: [{ type: Output }],
    maxAllowedFiles: [{ type: Input }],
    files: [{ type: Input }],
    accept: [{ type: Input }],
    area: [{ type: ViewChild, args: ['area', { static: true },] }],
    zone: [{ type: ViewChild, args: ['zone', { static: false },] }],
    picker: [{ type: ViewChild, args: ['picker', { static: false },] }],
    onkeyup: [{ type: HostListener, args: ['keyup', ['$event'],] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DropAreaComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-drop-area',
                template: "<div\n  class=\"drop-zone\"\n  *ngIf=\"!shouldShowFilesList()\"\n  [ngClass]=\"{ 'has-errors': errors }\"\n  [style.pointerEvents]=\"loading ? 'none' : 'auto'\"\n  #zone\n  (dragleave)=\"stopDragging()\"\n  (drop)=\"onDrop($event)\"\n  (dragover)=\"onOver()\"\n  [style.display]=\"isOver || alwaysShow || loading ? 'block' : 'none'\"\n  (click)=\"showPicker($event)\"\n  tabindex=\"0\"\n>\n  <div class=\"file-placeholder\" [ngClass]=\"{ 'drag-over': isOver }\">\n    <div *ngIf=\"loading\" class=\"d-flex p-4 flex-center\">\n      <div\n        class=\"progress progress-striped active m-0\"\n        *ngIf=\"progress !== -1\"\n        style=\"min-width: 50%\"\n      >\n        <div\n          class=\"progress-bar\"\n          role=\"progressbar\"\n          aria-valuenow=\"0\"\n          aria-valuemin=\"0\"\n          aria-valuemax=\"100\"\n          [style.width]=\"progress + '%'\"\n        ></div>\n      </div>\n      <div class=\"spinner-snake\" *ngIf=\"progress === -1\"></div>\n      <p class=\"flex-item-middle m-r-8\">\n        {{ loadingMessage | translate }}\n      </p>\n    </div>\n\n    <div *ngIf=\"!loading\" class=\"hint-placeholder pointer\">\n      <i class=\"dlt-c8y-icon-{{ icon }}\"></i>\n      <p *ngIf=\"!errors\">\n        <b>{{ message | translate }}</b>\n        <br />\n        <span *ngIf=\"alwaysShow && clickToOpen\" translate\n          >or click to browse your file system.</span\n        >\n      </p>\n      <div *ngIf=\"errors\" class=\"has-errors\">\n        <p class=\"form-control-feedback-message\">\n          {{ errorMessage | translate }}\n        </p>\n      </div>\n    </div>\n  </div>\n</div>\n\n<div\n  class=\"drop-zone\"\n  *ngIf=\"shouldShowFilesList()\"\n  [style.display]=\"isOver || alwaysShow || loading ? 'block' : 'none'\"\n  tabindex=\"0\"\n>\n  <div *ngIf=\"loading\" class=\"p-absolute p-4 fit-w fit-h d-flex d-col j-c-center a-i-center\">\n    <p class=\"m-b-8\">\n      <strong>\n        {{ loadingMessage | translate }}\n      </strong>\n    </p>\n    <div\n      class=\"progress progress-striped active m-0\"\n      *ngIf=\"progress !== -1\"\n      style=\"min-width: 80%\"\n    >\n      <div\n        class=\"progress-bar\"\n        role=\"progressbar\"\n        aria-valuenow=\"0\"\n        aria-valuemin=\"0\"\n        aria-valuemax=\"100\"\n        [style.width]=\"progress + '%'\"\n      ></div>\n    </div>\n    <div class=\"spinner p-relative m-0\" *ngIf=\"progress === -1\">\n      <div class=\"rect1\"></div>\n      <div class=\"rect2\"></div>\n      <div class=\"rect3\"></div>\n      <div class=\"rect4\"></div>\n      <div class=\"rect5\"></div>\n    </div>\n  </div>\n  <div *ngIf=\"!loading\" class=\"file-placeholder p-4\">\n    <div class=\"flex-row p-4\">\n      <i c8yIcon=\"file-o\" class=\"m-r-8\"></i>\n      <span title=\"{{ filesNameString }}\" class=\"text-truncate\">\n        {{ filesNameString }}\n      </span>\n      <button title=\"{{ 'Remove' | translate }}\" class=\"btn btn-clean showOnHover flex-item-right\">\n        <i c8yIcon=\"minus-circle\" class=\"text-danger\" (click)=\"onDelete()\"></i>\n      </button>\n    </div>\n  </div>\n</div>\n\n<input\n  #picker\n  *ngIf=\"clickToOpen\"\n  (change)=\"onPick($event)\"\n  (click)=\"picker.focus()\"\n  (blur)=\"onTouched()\"\n  [accept]=\"acceptedExts\"\n  [multiple]=\"maxAllowedFiles > 1\"\n  type=\"file\"\n  style=\"opacity: 0; filter: alpha(opacity = 0); height: 0px\"\n/>\n<div #area [hidden]=\"isOver || loading\" (dragover)=\"toggle($event)\">\n  <ng-content></ng-content>\n</div>\n",
                providers: [{ provide: NG_VALUE_ACCESSOR, useExisting: DropAreaComponent, multi: true }]
            }]
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }, { type: ɵngcc1.FilesService }, { type: ɵngcc2.TranslateService }, { type: ɵngcc3.BytesPipe }]; }, { title: [{
            type: Input
        }], message: [{
            type: Input
        }], icon: [{
            type: Input
        }], loadingMessage: [{
            type: Input
        }], forceHideList: [{
            type: Input
        }], alwaysShow: [{
            type: Input
        }], clickToOpen: [{
            type: Input
        }], loading: [{
            type: Input
        }], progress: [{
            type: Input
        }], dropped: [{
            type: Output
        }], maxAllowedFiles: [{
            type: Input
        }], onkeyup: [{
            type: HostListener,
            args: ['keyup', ['$event']]
        }], files: [{
            type: Input
        }], formControl: [{
            type: Input
        }], accept: [{
            type: Input
        }], area: [{
            type: ViewChild,
            args: ['area', { static: true }]
        }], zone: [{
            type: ViewChild,
            args: ['zone', { static: false }]
        }], picker: [{
            type: ViewChild,
            args: ['picker', { static: false }]
        }] }); })();
var ReadAsType;
(function (ReadAsType) {
    ReadAsType[ReadAsType["TEXT"] = 0] = "TEXT";
    ReadAsType[ReadAsType["DATA_URL"] = 1] = "DATA_URL";
    ReadAsType[ReadAsType["ARRAY_BUFFER"] = 2] = "ARRAY_BUFFER";
    ReadAsType[ReadAsType["BINARY_STRING"] = 3] = "BINARY_STRING";
})(ReadAsType || (ReadAsType = {}));
