// @c8y/ngx-components/esm2015/ecosystem/shared/add-application.component.js
import { __awaiter } from "tslib";
import { Component, Input, ViewChild } from '@angular/core';
import { ApplicationService } from '@c8y/client';
import { AlertService, DropAreaComponent, WizardComponent } from '@c8y/ngx-components';
import { ERROR_MESSAGES } from './../ecosystem.constants';
import { EcosystemService } from './../ecosystem.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './../ecosystem.service';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from '@c8y/client';
import * as ɵngcc4 from '@angular/common';

function AddApplicationComponent_c8y_form_group_9_c8y_message_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-message");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, ctx_r7.errorMessage), " ");
} }
function AddApplicationComponent_c8y_form_group_9_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-form-group", 7);
    ɵngcc0.ɵɵelementStart(1, "c8y-drop-area", 8);
    ɵngcc0.ɵɵlistener("dropped", function AddApplicationComponent_c8y_form_group_9_Template_c8y_drop_area_dropped_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r9); const ctx_r8 = ɵngcc0.ɵɵnextContext(); return ctx_r8.onFileDroppedEvent($event); });
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "c8y-messages");
    ɵngcc0.ɵɵtemplate(4, AddApplicationComponent_c8y_form_group_9_c8y_message_4_Template, 3, 3, "c8y-message", 9);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("hasError", ctx_r0.errorMessage);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("accept", ".zip")("loading", ctx_r0.isLoading)("maxAllowedFiles", 1)("progress", ɵngcc0.ɵɵpipeBind1(2, 6, ctx_r0.progress));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.errorMessage);
} }
function AddApplicationComponent_ng_template_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 10);
    ɵngcc0.ɵɵelement(1, "c8y-operation-result", 11);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("text", ɵngcc0.ɵɵpipeBind1(2, 3, ctx_r2.successText));
    ɵngcc0.ɵɵproperty("vertical", true)("size", 84);
} }
function AddApplicationComponent_button_13_Template(rf, ctx) { if (rf & 1) {
    const _r11 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 12);
    ɵngcc0.ɵɵlistener("click", function AddApplicationComponent_button_13_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r11); const ctx_r10 = ɵngcc0.ɵɵnextContext(); return ctx_r10.back(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtext(2, " Back ");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 1, "Back"));
} }
function AddApplicationComponent_button_14_Template(rf, ctx) { if (rf & 1) {
    const _r13 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 12);
    ɵngcc0.ɵɵlistener("click", function AddApplicationComponent_button_14_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r13); const ctx_r12 = ɵngcc0.ɵɵnextContext(); return ctx_r12.cancel(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtext(2, " Cancel ");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 1, "Cancel"));
} }
function AddApplicationComponent_button_15_Template(rf, ctx) { if (rf & 1) {
    const _r15 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 12);
    ɵngcc0.ɵɵlistener("click", function AddApplicationComponent_button_15_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r15); const ctx_r14 = ɵngcc0.ɵɵnextContext(); return ctx_r14.done(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtext(2, " Done ");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 1, "Done"));
} }
function AddApplicationComponent_a_16_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "a", 13);
    ɵngcc0.ɵɵlistener("click", function AddApplicationComponent_a_16_Template_a_click_0_listener($event) { return $event.stopPropagation(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 14);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 3, "Open"));
    ɵngcc0.ɵɵproperty("href", ctx_r6.getHref(ctx_r6.createdApp), ɵngcc0.ɵɵsanitizeUrl);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 5, "Open"), " ");
} }
export class AddApplicationComponent {
    constructor(ecosystemService, alertService, applicationService, wizardComponent) {
        this.ecosystemService = ecosystemService;
        this.alertService = alertService;
        this.applicationService = applicationService;
        this.wizardComponent = wizardComponent;
        this.canGoBack = false;
        this.canOpenInBrowser = false;
        this.uploadCanceled = false;
    }
    get progress() {
        return this.ecosystemService.progress;
    }
    onFileDroppedEvent(event) {
        if (event && event.length > 0) {
            const file = event[0].file;
            this.onFile(file);
        }
    }
    onFile(file) {
        return __awaiter(this, void 0, void 0, function* () {
            this.isLoading = true;
            this.errorMessage = null;
            this.progress.next(0);
            try {
                this.createdApp = yield this.createApplicationHandler(file);
                yield this.uploadApplicationHandler(file, this.createdApp);
                this.canOpenInBrowser = this.ecosystemService.canOpenAppInBrowser(this.createdApp);
                this.isAppCreated = true;
            }
            catch (ex) {
                this.ecosystemService.cancelAppCreation(this.createdApp);
                this.createdApp = null;
                this.dropAreaComponent.onDelete();
                this.errorMessage = ERROR_MESSAGES[ex.message];
                if (!this.errorMessage && !this.uploadCanceled) {
                    this.alertService.addServerFailure(ex);
                }
            }
            this.progress.next(100);
            this.isLoading = false;
        });
    }
    getHref(app) {
        return this.applicationService.getHref(app);
    }
    cancel() {
        this.cancelFileUpload();
        this.wizardComponent.close();
    }
    done() {
        this.wizardComponent.close();
    }
    back() {
        this.wizardComponent.reset();
    }
    cancelFileUpload() {
        this.uploadCanceled = true;
        this.ecosystemService.cancelAppCreation(this.createdApp);
        this.createdApp = null;
    }
}
AddApplicationComponent.ɵfac = function AddApplicationComponent_Factory(t) { return new (t || AddApplicationComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.EcosystemService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.ApplicationService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.WizardComponent)); };
AddApplicationComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AddApplicationComponent, selectors: [["c8y-add-application"]], viewQuery: function AddApplicationComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(DropAreaComponent, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.dropAreaComponent = _t.first);
    } }, inputs: { canGoBack: "canGoBack", headerText: "headerText", headerIcon: "headerIcon", successText: "successText", createApplicationHandler: "createApplicationHandler", uploadApplicationHandler: "uploadApplicationHandler" }, decls: 17, vars: 13, consts: [[3, "c8yIcon"], [1, "modal-inner-scroll", "animated", "fadeIn"], [1, "p-16", "text-center", "text-medium", "separator-bottom", "sticky-top", "bg-component"], ["class", "m-t-16 m-l-auto m-r-auto", "style", "max-width: 285px", 3, "hasError", 4, "ngIf", "ngIfElse"], ["appCreated", ""], ["class", "btn btn-default", "translate", "", "type", "button", 3, "title", "click", 4, "ngIf"], ["target", "_blank", "class", "btn btn-primary", 3, "href", "title", "click", 4, "ngIf"], [1, "m-t-16", "m-l-auto", "m-r-auto", 2, "max-width", "285px", 3, "hasError"], [1, "drop-area", 3, "accept", "loading", "maxAllowedFiles", "progress", "dropped"], [4, "ngIf"], [1, "d-flex", "a-i-center", "j-c-center", 2, "min-height", "285px"], ["type", "success", 1, "lead", 3, "text", "vertical", "size"], ["translate", "", "type", "button", 1, "btn", "btn-default", 3, "title", "click"], ["target", "_blank", 1, "btn", "btn-primary", 3, "href", "title", "click"], ["c8yIcon", "external-link", 1, "m-r-4"]], template: function AddApplicationComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-wizard-header");
        ɵngcc0.ɵɵelement(1, "h1", 0);
        ɵngcc0.ɵɵelementStart(2, "h4");
        ɵngcc0.ɵɵtext(3);
        ɵngcc0.ɵɵpipe(4, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "div", 1);
        ɵngcc0.ɵɵelementStart(6, "p", 2);
        ɵngcc0.ɵɵtext(7);
        ɵngcc0.ɵɵpipe(8, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(9, AddApplicationComponent_c8y_form_group_9_Template, 5, 8, "c8y-form-group", 3);
        ɵngcc0.ɵɵtemplate(10, AddApplicationComponent_ng_template_10_Template, 3, 5, "ng-template", null, 4, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(12, "c8y-wizard-footer");
        ɵngcc0.ɵɵtemplate(13, AddApplicationComponent_button_13_Template, 3, 3, "button", 5);
        ɵngcc0.ɵɵtemplate(14, AddApplicationComponent_button_14_Template, 3, 3, "button", 5);
        ɵngcc0.ɵɵtemplate(15, AddApplicationComponent_button_15_Template, 3, 3, "button", 5);
        ɵngcc0.ɵɵtemplate(16, AddApplicationComponent_a_16_Template, 5, 7, "a", 6);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r1 = ɵngcc0.ɵɵreference(11);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("c8yIcon", ctx.headerIcon);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(4, 9, ctx.headerText));
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(8, 11, "Upload a *.zip file"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isAppCreated)("ngIfElse", _r1);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isAppCreated && ctx.canGoBack);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isAppCreated);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isAppCreated);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isAppCreated && ctx.canOpenInBrowser);
    } }, directives: [ɵngcc2.WizardHeaderComponent, ɵngcc2.IconDirective, ɵngcc4.NgIf, ɵngcc2.WizardFooterComponent, ɵngcc2.FormGroupComponent, ɵngcc2.DropAreaComponent, ɵngcc2.MessagesComponent, ɵngcc2.MessageDirective, ɵngcc2.OperationResultComponent, ɵngcc2.C8yTranslateDirective], pipes: [ɵngcc2.C8yTranslatePipe, ɵngcc4.AsyncPipe], encapsulation: 2 });
AddApplicationComponent.ctorParameters = () => [
    { type: EcosystemService },
    { type: AlertService },
    { type: ApplicationService },
    { type: WizardComponent }
];
AddApplicationComponent.propDecorators = {
    headerText: [{ type: Input }],
    headerIcon: [{ type: Input }],
    successText: [{ type: Input }],
    createApplicationHandler: [{ type: Input }],
    uploadApplicationHandler: [{ type: Input }],
    canGoBack: [{ type: Input }],
    dropAreaComponent: [{ type: ViewChild, args: [DropAreaComponent,] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AddApplicationComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-add-application',
                template: "<c8y-wizard-header>\n  <h1 [c8yIcon]=\"headerIcon\"></h1>\n  <h4>{{ headerText | translate }}</h4>\n</c8y-wizard-header>\n\n<div class=\"modal-inner-scroll animated fadeIn\">\n  <p class=\"p-16 text-center text-medium separator-bottom sticky-top bg-component\">\n    {{ 'Upload a *.zip file' | translate }}\n  </p>\n    <c8y-form-group\n      *ngIf=\"!isAppCreated; else appCreated\"\n      [hasError]=\"errorMessage\"\n      class=\"m-t-16 m-l-auto m-r-auto\"\n      style=\"max-width: 285px\"\n    >\n      <c8y-drop-area\n        (dropped)=\"onFileDroppedEvent($event)\"\n        [accept]=\"'.zip'\"\n        [loading]=\"isLoading\"\n        [maxAllowedFiles]=\"1\"\n        [progress]=\"progress | async\"\n        class=\"drop-area\"\n      >\n      </c8y-drop-area>\n      <c8y-messages>\n        <c8y-message *ngIf=\"errorMessage\">\n          {{ errorMessage | translate }}\n        </c8y-message>\n      </c8y-messages>\n    </c8y-form-group>\n    <ng-template #appCreated>\n      <div class=\"d-flex a-i-center j-c-center\" style=\"min-height: 285px\">\n        <c8y-operation-result\n          text=\"{{ successText | translate }}\"\n          [vertical]=\"true\"\n          [size]=\"84\"\n          class=\"lead\"\n          type=\"success\"\n        >\n        </c8y-operation-result>\n      </div>\n    </ng-template>\n  \n</div>\n<c8y-wizard-footer>\n  <button\n    (click)=\"back()\"\n    *ngIf=\"!isAppCreated && canGoBack\"\n    class=\"btn btn-default\"\n    title=\"{{ 'Back' | translate }}\"\n    translate\n    type=\"button\"\n  >\n    Back\n  </button>\n  <button\n    (click)=\"cancel()\"\n    *ngIf=\"!isAppCreated\"\n    class=\"btn btn-default\"\n    title=\"{{ 'Cancel' | translate }}\"\n    translate\n    type=\"button\"\n  >\n    Cancel\n  </button>\n  <button\n    (click)=\"done()\"\n    *ngIf=\"isAppCreated\"\n    class=\"btn btn-default\"\n    title=\"{{ 'Done' | translate }}\"\n    translate\n    type=\"button\"\n  >\n    Done\n  </button>\n  <a\n    (click)=\"$event.stopPropagation()\"\n    *ngIf=\"isAppCreated && canOpenInBrowser\"\n    [href]=\"getHref(createdApp)\"\n    target=\"_blank\"\n    class=\"btn btn-primary\"\n    title=\"{{ 'Open' | translate }}\"\n  >\n    <i c8yIcon=\"external-link\" class=\"m-r-4\"></i>\n    {{ 'Open' | translate }}\n  </a>\n</c8y-wizard-footer>\n"
            }]
    }], function () { return [{ type: ɵngcc1.EcosystemService }, { type: ɵngcc2.AlertService }, { type: ɵngcc3.ApplicationService }, { type: ɵngcc2.WizardComponent }]; }, { canGoBack: [{
            type: Input
        }], headerText: [{
            type: Input
        }], headerIcon: [{
            type: Input
        }], successText: [{
            type: Input
        }], createApplicationHandler: [{
            type: Input
        }], uploadApplicationHandler: [{
            type: Input
        }], dropAreaComponent: [{
            type: ViewChild,
            args: [DropAreaComponent]
        }] }); })();
