// @c8y/ngx-components/esm2015/core/tabs/tabs-outlet.component.js
import { Component, Input, HostBinding, Renderer2, ElementRef, ViewChild } from '@angular/core';
import { debounceTime } from 'rxjs/operators';
import { fromEventPattern, fromEvent, merge } from 'rxjs';
import { Router } from '@angular/router';
import { HeaderService } from '../header/header.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from '../header/header.service';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from '../common/icon.directive';
import * as ɵngcc5 from '../i18n/c8y-translate.pipe';

const _c0 = ["container"];
const _c1 = ["left"];
const _c2 = ["right"];
const _c3 = ["tabsContainer"];
function TabsOutletComponent_button_2_Template(rf, ctx) { if (rf & 1) {
    const _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 12, 13);
    ɵngcc0.ɵɵlistener("click", function TabsOutletComponent_button_2_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r10); const ctx_r9 = ɵngcc0.ɵɵnextContext(); return ctx_r9.scroll("left"); });
    ɵngcc0.ɵɵelement(2, "i", 14);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", "chevron-left");
} }
function TabsOutletComponent_button_3_Template(rf, ctx) { if (rf & 1) {
    const _r13 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 15, 16);
    ɵngcc0.ɵɵlistener("click", function TabsOutletComponent_button_3_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r13); const ctx_r12 = ɵngcc0.ɵɵnextContext(); return ctx_r12.scroll("right"); });
    ɵngcc0.ɵɵelement(2, "i", 14);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", "chevron-right");
} }
function TabsOutletComponent_ng_container_6_li_1_i_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 14);
} if (rf & 2) {
    const tab_r14 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵproperty("c8yIcon", tab_r14 == null ? null : tab_r14.icon);
} }
function TabsOutletComponent_ng_container_6_li_1_span_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 25);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const tab_r14 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(2, 1, tab_r14 == null ? null : tab_r14.label));
} }
function TabsOutletComponent_ng_container_6_li_1_ng_container_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
const _c4 = function (a0) { return { active: a0 }; };
function TabsOutletComponent_ng_container_6_li_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 19);
    ɵngcc0.ɵɵelementStart(1, "a", 20, 21);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵtemplate(4, TabsOutletComponent_ng_container_6_li_1_i_4_Template, 1, 1, "i", 22);
    ɵngcc0.ɵɵtemplate(5, TabsOutletComponent_ng_container_6_li_1_span_5_Template, 3, 3, "span", 23);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(6, TabsOutletComponent_ng_container_6_li_1_ng_container_6_Template, 1, 0, "ng-container", 24);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _r17 = ɵngcc0.ɵɵreference(2);
    const tab_r14 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(8, _c4, _r17 == null ? null : _r17.isActive));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(3, 6, tab_r14 == null ? null : tab_r14.label));
    ɵngcc0.ɵɵproperty("routerLink", tab_r14.path);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", !tab_r14.template && (tab_r14 == null ? null : tab_r14.icon));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !tab_r14.template && !tab_r14.component);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", tab_r14.template);
} }
function TabsOutletComponent_ng_container_6_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function TabsOutletComponent_ng_container_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, TabsOutletComponent_ng_container_6_li_1_Template, 7, 10, "li", 17);
    ɵngcc0.ɵɵtemplate(2, TabsOutletComponent_ng_container_6_ng_container_2_Template, 1, 0, "ng-container", 18);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const tab_r14 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !tab_r14.component);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngComponentOutlet", tab_r14.component);
} }
function TabsOutletComponent_ng_container_12_option_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "option", 27);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const tab_r24 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r25 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("value", tab_r24.path)("selected", ctx_r25.isActive(tab_r24));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 3, tab_r24 == null ? null : tab_r24.label), " ");
} }
function TabsOutletComponent_ng_container_12_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, TabsOutletComponent_ng_container_12_option_1_Template, 3, 5, "option", 26);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const tab_r24 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !tab_r24.component);
} }
function TabsOutletComponent_ng_container_13_ng_container_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function TabsOutletComponent_ng_container_13_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, TabsOutletComponent_ng_container_13_ng_container_1_ng_container_1_Template, 1, 0, "ng-container", 18);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const tab_r27 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngComponentOutlet", tab_r27.component);
} }
function TabsOutletComponent_ng_container_13_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, TabsOutletComponent_ng_container_13_ng_container_1_Template, 2, 1, "ng-container", 28);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const tab_r27 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", tab_r27.component);
} }
export class TabsOutletComponent {
    constructor(router, renderer, el, header) {
        this.router = router;
        this.renderer = renderer;
        this.el = el;
        this.header = header;
        this.tabs = [];
        this.orientation = 'vertical';
        this.navigatorOpen = false;
        this.showLeft = false;
        this.showRight = false;
    }
    get isNavigatorOpen() {
        return this.navigatorOpen;
    }
    get hasTabs() {
        let visibleTabs = [];
        if (this.tabs) {
            visibleTabs = this.tabs.filter(tab => tab.hide !== true);
        }
        return visibleTabs.length > 1;
    }
    get isHorizontal() {
        return this.orientation === 'horizontal';
    }
    get isVertical() {
        return this.orientation !== 'horizontal';
    }
    ngOnDestroy() {
        if (this.scrollSubscription) {
            this.scrollSubscription.unsubscribe();
        }
    }
    ngAfterContentInit() {
        if (this.orientation === 'horizontal') {
            this.scrollSubscribe();
        }
    }
    isActive(tab) {
        if (tab && tab.path) {
            return this.router.isActive(tab.path, false);
        }
        else {
            return false;
        }
    }
    navigateTo(path) {
        this.header.toggle();
        this.router.navigate([path]);
    }
    scroll(direction = 'right') {
        const { widthScroll, width, scrollLeft } = this.getDimensions();
        const multiplier = direction === 'left' ? -1 : 1;
        let targetScroll = scrollLeft + width * multiplier;
        if (targetScroll < 0) {
            targetScroll = 0;
        }
        if (targetScroll > widthScroll - width) {
            targetScroll = widthScroll - width;
        }
        // TODO: we should animate this
        this.container.nativeElement.scrollLeft = targetScroll;
    }
    scrollSubscribe() {
        let unlisten;
        this.scrollSubscription = merge(fromEvent(this.container.nativeElement, 'scroll'), fromEventPattern((handler) => {
            unlisten = this.renderer.listen('window', 'resize', handler);
        }, () => unlisten())).pipe(debounceTime(250)).subscribe(() => this.updateScroll());
    }
    updateScroll() {
        const { widthScroll, width, scrollLeft } = this.getDimensions();
        this.showLeft = !!scrollLeft;
        this.showRight = widthScroll > width + scrollLeft;
    }
    getDimensions() {
        const elTabs = this.tabsContainer.nativeElement;
        const el = this.container.nativeElement;
        const widthScroll = elTabs.scrollWidth;
        const width = el.clientWidth;
        const scrollLeft = el.scrollLeft;
        return { widthScroll, width, scrollLeft };
    }
}
TabsOutletComponent.ɵfac = function TabsOutletComponent_Factory(t) { return new (t || TabsOutletComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Router), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.HeaderService)); };
TabsOutletComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: TabsOutletComponent, selectors: [["c8y-tabs-outlet"], ["c8y-ui-tabs"]], viewQuery: function TabsOutletComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 7, ElementRef);
        ɵngcc0.ɵɵviewQuery(_c1, 5, ElementRef);
        ɵngcc0.ɵɵviewQuery(_c2, 5, ElementRef);
        ɵngcc0.ɵɵviewQuery(_c3, 5, ElementRef);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.container = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.leftBtn = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.rightBtn = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.tabsContainer = _t.first);
    } }, hostVars: 8, hostBindings: function TabsOutletComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("navigator-open", ctx.isNavigatorOpen)("page-tabs", ctx.hasTabs)("page-tabs-horizontal", ctx.isHorizontal)("page-tabs-vertical", ctx.isVertical);
    } }, inputs: { tabs: "tabs", orientation: "orientation", navigatorOpen: "navigatorOpen" }, decls: 14, vars: 5, consts: [[1, "tabContainer", "hidden-xs"], ["container", ""], ["class", "tabs-slider tabs-slider-left", 3, "click", 4, "ngIf"], ["class", "tabs-slider tabs-slider-right", 3, "click", 4, "ngIf"], [1, "nav", "nav-tabs", "nav-tabsc8y"], ["tabsContainer", ""], [4, "ngFor", "ngForOf"], [1, "visible-xs", "mobile-tabs"], [1, "list-unstyled", "d-flex", "m-b-0"], [1, "c8y-select-wrapper", "flex-grow"], [3, "change"], ["select", ""], [1, "tabs-slider", "tabs-slider-left", 3, "click"], ["left", ""], [3, "c8yIcon"], [1, "tabs-slider", "tabs-slider-right", 3, "click"], ["right", ""], [3, "ngClass", 4, "ngIf"], [4, "ngComponentOutlet"], [3, "ngClass"], ["routerLinkActive", "", 3, "routerLink", "title"], ["rla", "routerLinkActive"], [3, "c8yIcon", 4, "ngIf"], ["class", "txt", 4, "ngIf"], [4, "ngTemplateOutlet"], [1, "txt"], [3, "value", "selected", 4, "ngIf"], [3, "value", "selected"], [4, "ngIf"]], template: function TabsOutletComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0, 1);
        ɵngcc0.ɵɵtemplate(2, TabsOutletComponent_button_2_Template, 3, 1, "button", 2);
        ɵngcc0.ɵɵtemplate(3, TabsOutletComponent_button_3_Template, 3, 1, "button", 3);
        ɵngcc0.ɵɵelementStart(4, "ul", 4, 5);
        ɵngcc0.ɵɵtemplate(6, TabsOutletComponent_ng_container_6_Template, 3, 2, "ng-container", 6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(7, "div", 7);
        ɵngcc0.ɵɵelementStart(8, "ul", 8);
        ɵngcc0.ɵɵelementStart(9, "li", 9);
        ɵngcc0.ɵɵelementStart(10, "select", 10, 11);
        ɵngcc0.ɵɵlistener("change", function TabsOutletComponent_Template_select_change_10_listener($event) { return ctx.navigateTo($event.target.value); });
        ɵngcc0.ɵɵtemplate(12, TabsOutletComponent_ng_container_12_Template, 2, 1, "ng-container", 6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(13, TabsOutletComponent_ng_container_13_Template, 2, 1, "ng-container", 6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showLeft);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showRight);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.tabs);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.tabs);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.tabs);
    } }, directives: [ɵngcc3.NgIf, ɵngcc3.NgForOf, ɵngcc4.IconDirective, ɵngcc3.NgComponentOutlet, ɵngcc3.NgClass, ɵngcc1.RouterLinkWithHref, ɵngcc1.RouterLinkActive, ɵngcc3.NgTemplateOutlet], pipes: [ɵngcc5.C8yTranslatePipe], encapsulation: 2 });
TabsOutletComponent.ctorParameters = () => [
    { type: Router },
    { type: Renderer2 },
    { type: ElementRef },
    { type: HeaderService }
];
TabsOutletComponent.propDecorators = {
    tabs: [{ type: Input }],
    orientation: [{ type: Input }],
    navigatorOpen: [{ type: Input }],
    isNavigatorOpen: [{ type: HostBinding, args: ['class.navigator-open',] }],
    hasTabs: [{ type: HostBinding, args: ['class.page-tabs',] }],
    isHorizontal: [{ type: HostBinding, args: ['class.page-tabs-horizontal',] }],
    isVertical: [{ type: HostBinding, args: ['class.page-tabs-vertical',] }],
    container: [{ type: ViewChild, args: ['container', { read: ElementRef, static: true },] }],
    leftBtn: [{ type: ViewChild, args: ['left', { read: ElementRef, static: false },] }],
    rightBtn: [{ type: ViewChild, args: ['right', { read: ElementRef, static: false },] }],
    tabsContainer: [{ type: ViewChild, args: ['tabsContainer', { read: ElementRef, static: false },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TabsOutletComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-tabs-outlet,c8y-ui-tabs',
                template: "<div #container class=\"tabContainer hidden-xs\">\n  <button class=\"tabs-slider tabs-slider-left\" #left *ngIf=\"showLeft\" (click)=\"scroll('left')\">\n    <i [c8yIcon]=\"'chevron-left'\"></i>\n  </button>\n  <button class=\"tabs-slider tabs-slider-right\" #right *ngIf=\"showRight\" (click)=\"scroll('right')\">\n    <i [c8yIcon]=\"'chevron-right'\"></i>\n  </button>\n  <ul class=\"nav nav-tabs nav-tabsc8y\" #tabsContainer>\n    <ng-container *ngFor=\"let tab of tabs\">\n      <li [ngClass]=\"{ active: rla?.isActive }\" *ngIf=\"!tab.component\">\n        <a\n          [routerLink]=\"tab.path\"\n          routerLinkActive\n          #rla=\"routerLinkActive\"\n          title=\"{{ tab?.label | translate }}\"\n        >\n          <i [c8yIcon]=\"tab?.icon\" *ngIf=\"!tab.template && tab?.icon\"></i>\n          <span class=\"txt\" *ngIf=\"!tab.template && !tab.component\">{{\n            tab?.label | translate\n          }}</span>\n        </a>\n        <ng-container *ngTemplateOutlet=\"tab.template\"></ng-container>\n      </li>\n      <ng-container *ngComponentOutlet=\"tab.component\"></ng-container>\n    </ng-container>\n  </ul>\n</div>\n\n<div class=\"visible-xs mobile-tabs\">\n  <ul class=\"list-unstyled d-flex m-b-0\">\n    <li class=\"c8y-select-wrapper flex-grow\">\n      <select (change)=\"navigateTo($event.target.value)\" #select>\n        <ng-container *ngFor=\"let tab of tabs\">\n          <option *ngIf=\"!tab.component\" [value]=\"tab.path\" [selected]=\"isActive(tab)\">\n            {{ tab?.label | translate }}\n          </option>\n        </ng-container>\n      </select>\n    </li>\n    <ng-container *ngFor=\"let tab of tabs\">\n      <ng-container *ngIf=\"tab.component\">\n        <ng-container *ngComponentOutlet=\"tab.component\"></ng-container>\n      </ng-container>\n    </ng-container>\n  </ul>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.Router }, { type: ɵngcc0.Renderer2 }, { type: ɵngcc0.ElementRef }, { type: ɵngcc2.HeaderService }]; }, { tabs: [{
            type: Input
        }], orientation: [{
            type: Input
        }], navigatorOpen: [{
            type: Input
        }], isNavigatorOpen: [{
            type: HostBinding,
            args: ['class.navigator-open']
        }], hasTabs: [{
            type: HostBinding,
            args: ['class.page-tabs']
        }], isHorizontal: [{
            type: HostBinding,
            args: ['class.page-tabs-horizontal']
        }], isVertical: [{
            type: HostBinding,
            args: ['class.page-tabs-vertical']
        }], container: [{
            type: ViewChild,
            args: ['container', { read: ElementRef, static: true }]
        }], leftBtn: [{
            type: ViewChild,
            args: ['left', { read: ElementRef, static: false }]
        }], rightBtn: [{
            type: ViewChild,
            args: ['right', { read: ElementRef, static: false }]
        }], tabsContainer: [{
            type: ViewChild,
            args: ['tabsContainer', { read: ElementRef, static: false }]
        }] }); })();
