// @c8y/ngx-components/esm2015/ecosystem/packages/deploy-application/deploy-application.component.js
import { __awaiter } from "tslib";
import { Component, ViewChild } from '@angular/core';
import { ApplicationService } from '@c8y/client';
import { gettext, WizardComponent } from '@c8y/ngx-components';
import { TranslateService } from '@ngx-translate/core';
import { EcosystemService } from '../../ecosystem.service';
import { ApplicationPropertiesFormComponent } from '../../shared/application-properties-form.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../ecosystem.service';
import * as ɵngcc2 from '@c8y/client';
import * as ɵngcc3 from '@c8y/ngx-components';
import * as ɵngcc4 from '@ngx-translate/core';
import * as ɵngcc5 from '@angular/common';
import * as ɵngcc6 from '../../shared/application-properties-form.component';

function DeployApplicationComponent_ng_container_7_p_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p", 9);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, ctx_r3.headerText), " ");
} }
function DeployApplicationComponent_ng_container_7_c8y_application_properties_form_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-application-properties-form", 10);
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("application", ctx_r4.newAppConfig);
} }
function DeployApplicationComponent_ng_container_7_c8y_progress_bar_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-progress-bar", 11);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("message", ɵngcc0.ɵɵpipeBind1(1, 1, "Deploying\u2026"));
} }
function DeployApplicationComponent_ng_container_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "div", 5);
    ɵngcc0.ɵɵtemplate(2, DeployApplicationComponent_ng_container_7_p_2_Template, 3, 3, "p", 6);
    ɵngcc0.ɵɵtemplate(3, DeployApplicationComponent_ng_container_7_c8y_application_properties_form_3_Template, 1, 1, "c8y-application-properties-form", 7);
    ɵngcc0.ɵɵtemplate(4, DeployApplicationComponent_ng_container_7_c8y_progress_bar_4_Template, 2, 3, "c8y-progress-bar", 8);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r0.inProgress);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r0.inProgress);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.inProgress);
} }
const _c0 = function (a0) { return { packageName: a0 }; };
function DeployApplicationComponent_ng_container_8_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 14);
    ɵngcc0.ɵɵelementStart(1, "div", 15);
    ɵngcc0.ɵɵelement(2, "c8y-operation-result", 16);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("size", 84)("vertical", true)("text", ɵngcc0.ɵɵpipeBind2(3, 3, ctx_r6.successMessageTemplate, ɵngcc0.ɵɵpureFunction1(6, _c0, ctx_r6.package.name)));
} }
function DeployApplicationComponent_ng_container_8_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 17);
    ɵngcc0.ɵɵelement(1, "c8y-operation-result", 18);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("text", ɵngcc0.ɵɵpipeBind1(2, 3, "Application creation failed"));
    ɵngcc0.ɵɵproperty("size", 84)("vertical", true);
} }
function DeployApplicationComponent_ng_container_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, DeployApplicationComponent_ng_container_8_div_1_Template, 4, 8, "div", 12);
    ɵngcc0.ɵɵtemplate(2, DeployApplicationComponent_ng_container_8_ng_template_2_Template, 3, 5, "ng-template", null, 13, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const _r7 = ɵngcc0.ɵɵreference(3);
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.deployedWithSuccess)("ngIfElse", _r7);
} }
const _c1 = function (a0) { return { "btn-pending": a0 }; };
function DeployApplicationComponent_button_14_Template(rf, ctx) { if (rf & 1) {
    const _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 19);
    ɵngcc0.ɵɵlistener("click", function DeployApplicationComponent_button_14_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r10); const ctx_r9 = ɵngcc0.ɵɵnextContext(); return ctx_r9.deployApp(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 4, "Deploy"));
    ɵngcc0.ɵɵproperty("disabled", ctx_r2.inProgress)("ngClass", ɵngcc0.ɵɵpureFunction1(8, _c1, ctx_r2.inProgress));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 6, "Deploy"), " ");
} }
export class DeployApplicationComponent {
    constructor(ecosystemService, applicationService, wizardComponent, translate) {
        this.ecosystemService = ecosystemService;
        this.applicationService = applicationService;
        this.wizardComponent = wizardComponent;
        this.translate = translate;
        this.inProgress = true;
        this.isDeployed = false;
        this.deployedWithSuccess = false;
        this.descriptionTemplate = gettext('Deploy application using "{{ packageName }}" package');
        this.successMessageTemplate = gettext('Application "{{ packageName }}" created');
        this.doneLabel = gettext('Done');
        this.cancelLabel = gettext('Cancel');
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            this.package = this.wizardComponent.package;
            const apps = (yield this.ecosystemService.getApplications()).data;
            this.newAppConfig = this.ecosystemService.getUniqueAppConfig(this.package, apps);
            this.headerText = this.getHeaderText();
            this.inProgress = false;
        });
    }
    deployApp() {
        return __awaiter(this, void 0, void 0, function* () {
            const formGroupValue = this.applicationPropertiesForm.formGroup.getRawValue();
            this.inProgress = true;
            const { data: clonedPkg } = yield this.applicationService.clone(this.package);
            const config = this.ecosystemService.createConfig(clonedPkg, formGroupValue);
            try {
                const { data: updatedApp } = yield this.ecosystemService.updateApp(config, true);
                if (updatedApp) {
                    yield this.ecosystemService.updateAppManifest(updatedApp, this.package);
                }
                this.deployedWithSuccess = true;
            }
            catch (error) {
                this.markAsDeployed();
                yield this.applicationService.delete(clonedPkg.id);
            }
            this.markAsDeployed();
        });
    }
    cancel() {
        this.wizardComponent.close();
    }
    markAsDeployed() {
        this.isDeployed = true;
        this.inProgress = false;
    }
    getHeaderText() {
        return this.translate.instant(this.descriptionTemplate, {
            packageName: this.package.name
        });
    }
}
DeployApplicationComponent.ɵfac = function DeployApplicationComponent_Factory(t) { return new (t || DeployApplicationComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.EcosystemService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ApplicationService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.WizardComponent), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.TranslateService)); };
DeployApplicationComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: DeployApplicationComponent, selectors: [["c8y-deploy-application"]], viewQuery: function DeployApplicationComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(ApplicationPropertiesFormComponent, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.applicationPropertiesForm = _t.first);
    } }, decls: 15, vars: 12, consts: [[1, "modal-header", "dialog-header"], ["c8yIcon", "output"], [4, "ngIf"], ["type", "button", 1, "btn", "btn-default", 3, "title", "click"], ["class", "btn btn-primary", "type", "button", 3, "disabled", "ngClass", "title", "click", 4, "ngIf"], [1, "fadeIn", "animated", "d-flex", "a-i-center", "j-c-center", "d-col", 2, "min-height", "309px"], ["class", "bg-white fit-w p-16 text-center text-medium sticky-top bg-white separator-bottom", 4, "ngIf"], ["class", "d-block fit-w bg-gray-white", 3, "application", 4, "ngIf"], ["class", "text-center", 3, "message", 4, "ngIf"], [1, "bg-white", "fit-w", "p-16", "text-center", "text-medium", "sticky-top", "bg-white", "separator-bottom"], [1, "d-block", "fit-w", "bg-gray-white", 3, "application"], [1, "text-center", 3, "message"], ["class", "modal-body fadeIn animated", "style", "min-height: 309px", 4, "ngIf", "ngIfElse"], ["failedDeploy", ""], [1, "modal-body", "fadeIn", "animated", 2, "min-height", "309px"], [1, "d-flex", "a-i-center", "j-c-center", "d-col"], ["type", "success", 1, "lead", "d-block", "m-b-16", 3, "size", "vertical", "text"], [1, "modal-body", "fadeIn", "animated", "text-center", 2, "min-height", "257px"], ["type", "error", 1, "lead", 3, "size", "vertical", "text"], ["type", "button", 1, "btn", "btn-primary", 3, "disabled", "ngClass", "title", "click"]], template: function DeployApplicationComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-wizard-header");
        ɵngcc0.ɵɵelementStart(1, "div", 0);
        ɵngcc0.ɵɵelement(2, "h1", 1);
        ɵngcc0.ɵɵelementStart(3, "h4");
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "c8y-wizard-body");
        ɵngcc0.ɵɵtemplate(7, DeployApplicationComponent_ng_container_7_Template, 5, 3, "ng-container", 2);
        ɵngcc0.ɵɵtemplate(8, DeployApplicationComponent_ng_container_8_Template, 4, 2, "ng-container", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "c8y-wizard-footer");
        ɵngcc0.ɵɵelementStart(10, "button", 3);
        ɵngcc0.ɵɵlistener("click", function DeployApplicationComponent_Template_button_click_10_listener() { return ctx.cancel(); });
        ɵngcc0.ɵɵpipe(11, "translate");
        ɵngcc0.ɵɵtext(12);
        ɵngcc0.ɵɵpipe(13, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(14, DeployApplicationComponent_button_14_Template, 4, 10, "button", 4);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(5, 6, "Deploy application"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isDeployed);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isDeployed);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(11, 8, ctx.isDeployed && ctx.deployedWithSuccess ? ctx.doneLabel : ctx.cancelLabel));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(13, 10, ctx.isDeployed && ctx.deployedWithSuccess ? ctx.doneLabel : ctx.cancelLabel), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isDeployed);
    } }, directives: [ɵngcc3.WizardHeaderComponent, ɵngcc3.IconDirective, ɵngcc3.WizardBodyComponent, ɵngcc5.NgIf, ɵngcc3.WizardFooterComponent, ɵngcc6.ApplicationPropertiesFormComponent, ɵngcc3.ProgressBarComponent, ɵngcc3.OperationResultComponent, ɵngcc5.NgClass], pipes: [ɵngcc3.C8yTranslatePipe], encapsulation: 2 });
DeployApplicationComponent.ctorParameters = () => [
    { type: EcosystemService },
    { type: ApplicationService },
    { type: WizardComponent },
    { type: TranslateService }
];
DeployApplicationComponent.propDecorators = {
    applicationPropertiesForm: [{ type: ViewChild, args: [ApplicationPropertiesFormComponent,] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DeployApplicationComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-deploy-application',
                template: "<c8y-wizard-header>\n  <div class=\"modal-header dialog-header\">\n    <h1 c8yIcon=\"output\"></h1>\n    <h4>{{ 'Deploy application' | translate }}</h4>\n  </div>\n</c8y-wizard-header>\n\n<c8y-wizard-body>\n  <ng-container *ngIf=\"!isDeployed\">\n    <div class=\"fadeIn animated d-flex a-i-center j-c-center d-col\" style=\"min-height: 309px\">\n      <p\n        class=\"bg-white fit-w p-16 text-center text-medium sticky-top bg-white separator-bottom\"\n        *ngIf=\"!inProgress\"\n      >\n        {{ headerText | translate }}\n      </p>\n      <c8y-application-properties-form\n        *ngIf=\"!inProgress\"\n        [application]=\"newAppConfig\"\n        class=\"d-block fit-w bg-gray-white\"\n      ></c8y-application-properties-form>\n\n      <c8y-progress-bar\n        *ngIf=\"inProgress\"\n        [message]=\"'Deploying\u2026' | translate\"\n        class=\"text-center\"\n      ></c8y-progress-bar>\n    </div>\n  </ng-container>\n\n  <ng-container *ngIf=\"isDeployed\">\n    <div\n      *ngIf=\"deployedWithSuccess; else failedDeploy\"\n      class=\"modal-body fadeIn animated\"\n      style=\"min-height: 309px\"\n    >\n      <div class=\"d-flex a-i-center j-c-center d-col\">\n        <c8y-operation-result\n          type=\"success\"\n          [size]=\"84\"\n          [vertical]=\"true\"\n          [text]=\"successMessageTemplate | translate: { packageName: package.name }\"\n          class=\"lead d-block m-b-16\"\n        ></c8y-operation-result>\n      </div>\n    </div>\n    <ng-template #failedDeploy>\n      <div class=\"modal-body fadeIn animated text-center\" style=\"min-height: 257px\">\n        <c8y-operation-result\n          type=\"error\"\n          [size]=\"84\"\n          [vertical]=\"true\"\n          text=\"{{ 'Application creation failed' | translate }}\"\n          class=\"lead\"\n        ></c8y-operation-result>\n      </div>\n    </ng-template>\n  </ng-container>\n</c8y-wizard-body>\n\n<c8y-wizard-footer>\n  <button\n    (click)=\"cancel()\"\n    type=\"button\"\n    class=\"btn btn-default\"\n    title=\"{{ (isDeployed && deployedWithSuccess ? doneLabel : cancelLabel) | translate }}\"\n  >\n    {{ (isDeployed && deployedWithSuccess ? doneLabel : cancelLabel) | translate }}\n  </button>\n\n  <button\n    (click)=\"deployApp()\"\n    *ngIf=\"!isDeployed\"\n    [disabled]=\"inProgress\"\n    [ngClass]=\"{ 'btn-pending': inProgress }\"\n    class=\"btn btn-primary\"\n    type=\"button\"\n    title=\"{{ 'Deploy' | translate }}\"\n  >\n    {{ 'Deploy' | translate }}\n  </button>\n</c8y-wizard-footer>\n"
            }]
    }], function () { return [{ type: ɵngcc1.EcosystemService }, { type: ɵngcc2.ApplicationService }, { type: ɵngcc3.WizardComponent }, { type: ɵngcc4.TranslateService }]; }, { applicationPropertiesForm: [{
            type: ViewChild,
            args: [ApplicationPropertiesFormComponent]
        }] }); })();
