// @c8y/ngx-components/esm2015/core/authentication/authentication.module.js
import { NgModule } from '@angular/core';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { CommonModule } from '../common/common.module';
import { PasswordCheckListComponent } from './password-check-list.component';
import { PasswordConfirm } from './password-confirm.directive';
import { PasswordStrengthCheckerService } from './password-strength-checker.service';
import { PasswordStrengthComponent } from './password-strength.component';
import { NewPasswordComponent } from './new-password.component';
import { PasswordConfirmModalComponent } from './password-confirm-modal.component';
import { ModalModule } from '../modal/modal.module';
import { FormsModule } from '../forms/forms.module';
import { TotpChallengeComponent } from './totp-challenge.component';
import { SmsChallengeComponent } from './sms-challenge.component';
import { TotpSetupComponent } from './totp-setup.component';
import { PasswordService } from './password.service';
import { QRCodeModule } from 'angularx-qrcode';
import { ProvidePhoneNumberComponent } from './provide-phone-number.component';
import { CurrentPasswordModalComponent } from './current-password-modal.component';
/**
 * The angular module definition for new password.
 * @exports PasswordStrengthComponent
 * @exports PasswordCheckListComponent
 * @exports PasswordConfirm
 * @exports NewPasswordComponent
 * @exports PasswordConfirmModalComponent
 */
import * as ɵngcc0 from '@angular/core';
export class AuthenticationModule {
}
AuthenticationModule.ɵfac = function AuthenticationModule_Factory(t) { return new (t || AuthenticationModule)(); };
AuthenticationModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: AuthenticationModule });
AuthenticationModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [PasswordStrengthCheckerService, PasswordService], imports: [[FormsModule, TooltipModule, CommonModule, ModalModule, QRCodeModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AuthenticationModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    TotpChallengeComponent,
                    PasswordConfirm,
                    PasswordStrengthComponent,
                    PasswordCheckListComponent,
                    NewPasswordComponent,
                    PasswordConfirmModalComponent,
                    CurrentPasswordModalComponent,
                    SmsChallengeComponent,
                    TotpSetupComponent,
                    ProvidePhoneNumberComponent
                ],
                imports: [FormsModule, TooltipModule, CommonModule, ModalModule, QRCodeModule],
                exports: [
                    TotpChallengeComponent,
                    PasswordStrengthComponent,
                    PasswordCheckListComponent,
                    PasswordConfirm,
                    NewPasswordComponent,
                    PasswordConfirmModalComponent,
                    CurrentPasswordModalComponent,
                    SmsChallengeComponent,
                    TotpSetupComponent,
                    ProvidePhoneNumberComponent
                ],
                entryComponents: [PasswordConfirmModalComponent],
                providers: [PasswordStrengthCheckerService, PasswordService]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AuthenticationModule, { declarations: function () { return [TotpChallengeComponent,
        PasswordConfirm,
        PasswordStrengthComponent,
        PasswordCheckListComponent,
        NewPasswordComponent,
        PasswordConfirmModalComponent,
        CurrentPasswordModalComponent,
        SmsChallengeComponent,
        TotpSetupComponent,
        ProvidePhoneNumberComponent]; }, imports: function () { return [FormsModule, TooltipModule, CommonModule, ModalModule, QRCodeModule]; }, exports: function () { return [TotpChallengeComponent,
        PasswordStrengthComponent,
        PasswordCheckListComponent,
        PasswordConfirm,
        NewPasswordComponent,
        PasswordConfirmModalComponent,
        CurrentPasswordModalComponent,
        SmsChallengeComponent,
        TotpSetupComponent,
        ProvidePhoneNumberComponent]; } }); })();
