// @c8y/ngx-components/esm2015/core/authentication/current-password-modal.component.js
import { Component, Output, EventEmitter } from '@angular/core';
import { BsModalRef } from 'ngx-bootstrap/modal';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/modal';
import * as ɵngcc2 from '../modal/modal.component';
import * as ɵngcc3 from '@angular/forms';
import * as ɵngcc4 from '../forms/form-group.component';
import * as ɵngcc5 from '../i18n/c8y-translate.directive';
import * as ɵngcc6 from '../forms/required-input-placeholder.directive';
import * as ɵngcc7 from '../forms/messages.component';
import * as ɵngcc8 from '../forms/message.directive';
import * as ɵngcc9 from '../i18n/c8y-translate.pipe';
export class CurrentPasswordModalComponent {
    constructor(modal) {
        this.modal = modal;
        this.currentPasswordEmitter = new EventEmitter();
    }
    passwordConfirm() {
        this.currentPasswordEmitter.emit(this.password);
        this.modal.hide();
    }
    cancel() {
        this.currentPasswordEmitter.emit(null);
        this.modal.hide();
    }
}
CurrentPasswordModalComponent.ɵfac = function CurrentPasswordModalComponent_Factory(t) { return new (t || CurrentPasswordModalComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BsModalRef)); };
CurrentPasswordModalComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: CurrentPasswordModalComponent, selectors: [["c8y-current-password-modal"]], outputs: { currentPasswordEmitter: "currentPasswordEmitter" }, decls: 20, vars: 13, consts: [[3, "customFooter", "title"], ["modal", ""], [3, "ngSubmit"], ["confirmForm", "ngForm"], [1, "d-block", "p-24"], [3, "hasWarning"], ["translate", "", "for", "currentPassword"], ["id", "currentPassword", "type", "password", "name", "password", "required", "", 1, "form-control", 3, "ngModel", "ngModelChange"], ["translate", ""], [1, "modal-footer", "separator-top", "bg-white", "sticky-bottom"], ["type", "button", "translate", "", 1, "btn", "btn-default", 3, "title", "click"], ["type", "submit", "translate", "", 1, "btn", "btn-primary", 3, "title", "disabled"]], template: function CurrentPasswordModalComponent_Template(rf, ctx) { if (rf & 1) {
        const _r2 = ɵngcc0.ɵɵgetCurrentView();
        ɵngcc0.ɵɵelementStart(0, "c8y-modal", 0, 1);
        ɵngcc0.ɵɵpipe(2, "translate");
        ɵngcc0.ɵɵelementStart(3, "form", 2, 3);
        ɵngcc0.ɵɵlistener("ngSubmit", function CurrentPasswordModalComponent_Template_form_ngSubmit_3_listener() { ɵngcc0.ɵɵrestoreView(_r2); const _r1 = ɵngcc0.ɵɵreference(4); return _r1.form.valid && ctx.passwordConfirm(); });
        ɵngcc0.ɵɵelementStart(5, "div", 4);
        ɵngcc0.ɵɵelementStart(6, "c8y-form-group", 5);
        ɵngcc0.ɵɵelementStart(7, "label", 6);
        ɵngcc0.ɵɵtext(8, "Enter your password");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "input", 7);
        ɵngcc0.ɵɵlistener("ngModelChange", function CurrentPasswordModalComponent_Template_input_ngModelChange_9_listener($event) { return ctx.password = $event; });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(10, "c8y-messages");
        ɵngcc0.ɵɵelementStart(11, "c8y-message", 8);
        ɵngcc0.ɵɵtext(12, " The password of the user that you are currently logged in with. ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(13, "div", 9);
        ɵngcc0.ɵɵelementStart(14, "button", 10);
        ɵngcc0.ɵɵlistener("click", function CurrentPasswordModalComponent_Template_button_click_14_listener() { return ctx.cancel(); });
        ɵngcc0.ɵɵpipe(15, "translate");
        ɵngcc0.ɵɵtext(16, " Cancel ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(17, "button", 11);
        ɵngcc0.ɵɵpipe(18, "translate");
        ɵngcc0.ɵɵtext(19, " Confirm ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r1 = ɵngcc0.ɵɵreference(4);
        ɵngcc0.ɵɵproperty("customFooter", true)("title", ɵngcc0.ɵɵpipeBind1(2, 7, "Confirm your current password"));
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵproperty("hasWarning", true);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngModel", ctx.password);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(15, 9, "Cancel"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(18, 11, "Confirm"));
        ɵngcc0.ɵɵproperty("disabled", !_r1.form.valid);
    } }, directives: [ɵngcc2.ModalComponent, ɵngcc3.ɵNgNoValidate, ɵngcc3.NgControlStatusGroup, ɵngcc3.NgForm, ɵngcc4.FormGroupComponent, ɵngcc5.C8yTranslateDirective, ɵngcc6.RequiredInputPlaceholderDirective, ɵngcc3.DefaultValueAccessor, ɵngcc3.RequiredValidator, ɵngcc3.NgControlStatus, ɵngcc3.NgModel, ɵngcc7.MessagesComponent, ɵngcc8.MessageDirective], pipes: [ɵngcc9.C8yTranslatePipe], encapsulation: 2 });
CurrentPasswordModalComponent.ctorParameters = () => [
    { type: BsModalRef }
];
CurrentPasswordModalComponent.propDecorators = {
    currentPasswordEmitter: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CurrentPasswordModalComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-current-password-modal',
                template: "<c8y-modal [customFooter]=\"true\" [title]=\"'Confirm your current password' | translate\" #modal>\n  <form #confirmForm=\"ngForm\" (ngSubmit)=\"confirmForm.form.valid && passwordConfirm()\">\n    <div class=\"d-block p-24\">\n      <c8y-form-group [hasWarning]=\"true\">\n        <label translate for=\"currentPassword\">Enter your password</label>\n        <input\n          id=\"currentPassword\"\n          [(ngModel)]=\"password\"\n          type=\"password\"\n          name=\"password\"\n          class=\"form-control\"\n          required\n        />\n        <c8y-messages>\n          <c8y-message translate>\n            The password of the user that you are currently logged in with.\n          </c8y-message>\n        </c8y-messages>\n      </c8y-form-group>\n    </div>\n    <div class=\"modal-footer separator-top bg-white sticky-bottom\">\n      <button\n        title=\"{{ 'Cancel' | translate }}\"\n        class=\"btn btn-default\"\n        type=\"button\"\n        (click)=\"cancel()\"\n        translate\n      >\n        Cancel\n      </button>\n      <button\n        title=\"{{ 'Confirm' | translate }}\"\n        class=\"btn btn-primary\"\n        type=\"submit\"\n        [disabled]=\"!confirmForm.form.valid\"\n        translate\n      >\n        Confirm\n      </button>\n    </div>\n  </form>\n</c8y-modal>\n"
            }]
    }], function () { return [{ type: ɵngcc1.BsModalRef }]; }, { currentPasswordEmitter: [{
            type: Output
        }] }); })();
