// @c8y/ngx-components/esm2015/core/authentication/sms-challenge.component.js
import { __awaiter } from "tslib";
import { Component, Output, EventEmitter, Input } from '@angular/core';
import { UserService } from '@c8y/client';
import { LoginService } from '../login/login.service';
import { AlertService } from '../alert/alert.service';
import { gettext } from '../i18n/gettext';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../login/login.service';
import * as ɵngcc2 from '@c8y/client';
import * as ɵngcc3 from '../alert/alert.service';
import * as ɵngcc4 from '@angular/forms';
import * as ɵngcc5 from '../i18n/c8y-translate.directive';
import * as ɵngcc6 from '../forms/form-group.component';
import * as ɵngcc7 from '../forms/required-input-placeholder.directive';
import * as ɵngcc8 from '@angular/common';
import * as ɵngcc9 from '../i18n/c8y-translate.pipe';

function SmsChallengeComponent_p_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p", 11);
    ɵngcc0.ɵɵtext(1, " Insert the code received via SMS. ");
    ɵngcc0.ɵɵelementEnd();
} }
const _c0 = function (a0) { return { disabled: a0 }; };
export class SmsChallengeComponent {
    constructor(loginService, users, alert) {
        this.loginService = loginService;
        this.users = users;
        this.alert = alert;
        this.onCancel = new EventEmitter();
        this.model = {
            smsToken: ''
        };
        this.isLoading = false;
        this.resendTfa = '0';
    }
    verifyTFACode() {
        return __awaiter(this, void 0, void 0, function* () {
            this.isLoading = true;
            if (yield this.usesOAuthInternal()) {
                yield this.verifyCodeWithOauth();
            }
            else {
                yield this.verifyCodeWithBasicAuth();
            }
            this.isLoading = false;
        });
    }
    resendTFASms() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                this.isLoading = true;
                yield this.users.verifyTFACode(this.resendTfa);
            }
            catch (e) {
                if (e.res.status === 403) {
                    this.loginService.cleanMessages();
                    this.loginService.addSuccessMessage('resend_sms');
                }
                else {
                    this.alert.addServerFailure(e);
                }
            }
            finally {
                this.isLoading = false;
            }
        });
    }
    usesOAuthInternal() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.loginService.isPasswordGrantLogin(this.credentials);
        });
    }
    verifyCodeWithOauth() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const { credentials } = this;
                yield this.loginService.switchLoginMode(Object.assign(Object.assign({}, credentials), { tfa: this.model.smsToken }));
                yield this.loginService.verifyAppAccess();
                yield this.loginService.authFulfilled();
            }
            catch (e) {
                const resStatus = e.res && e.res.status;
                if (resStatus === 401) {
                    // it is assumed that the user and password are correct so it must be the tfa code
                    this.alert.danger(gettext('Invalid code'));
                }
                else {
                    this.alert.addServerFailure(e);
                }
            }
        });
    }
    verifyCodeWithBasicAuth() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const { res } = yield this.users.verifyTFACode(this.model.smsToken);
                const tfaToken = res.headers.get('tfatoken');
                this.credentials.tfa = tfaToken;
                yield this.loginWithTFA(tfaToken);
            }
            catch (e) {
                const resStatus = e.res && e.res.status;
                // BE returns 403 in case of invalid tfa code
                if (resStatus === 403) {
                    this.alert.danger(gettext('Invalid code'));
                }
                else {
                    this.alert.addServerFailure(e);
                }
            }
        });
    }
    loginWithTFA(tfaToken) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.loginService.login(this.loginService.useBasicAuth({ tfa: tfaToken }), this.credentials);
                this.loginService.saveTFAToken(tfaToken, sessionStorage);
                if (this.loginService.rememberMe) {
                    this.loginService.saveTFAToken(tfaToken, localStorage);
                }
            }
            catch (e) {
                this.alert.addServerFailure(e);
            }
        });
    }
}
SmsChallengeComponent.ɵfac = function SmsChallengeComponent_Factory(t) { return new (t || SmsChallengeComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.LoginService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.UserService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.AlertService)); };
SmsChallengeComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SmsChallengeComponent, selectors: [["c8y-sms-challenge"]], inputs: { credentials: "credentials" }, outputs: { onCancel: "onCancel" }, decls: 21, vars: 18, consts: [["novalidate", "", 1, "loginForm", 3, "ngSubmit"], ["twoFactorForm", "ngForm"], ["translate", "", 1, "legend", "form-block", "center"], ["translate", ""], ["type", "text", "name", "sms_token", "autofocus", "", "autocapitalize", "off", "autocorrect", "off", "required", "", 1, "form-control", 3, "ngModel", "placeholder", "ngModelChange"], ["sms_token", "ngModel"], ["class", "help-block", "translate", "", 4, "ngIf"], ["type", "submit", "translate", "", 1, "btn", "btn-primary", "btn-lg", "btn-block", "form-group", 3, "title", "disabled"], [1, "d-flex", "m-t-8"], ["translate", "", 1, "small", "pointer", "m-r-auto", 3, "title", "ngClass", "click"], ["translate", "", 1, "small", "pointer", "m-l-auto", 3, "title", "click"], ["translate", "", 1, "help-block"]], template: function SmsChallengeComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "form", 0, 1);
        ɵngcc0.ɵɵlistener("ngSubmit", function SmsChallengeComponent_Template_form_ngSubmit_0_listener() { return ctx.verifyTFACode(); });
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵtext(3, " Two-factor authentication ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(5, "label", 3);
        ɵngcc0.ɵɵtext(6, "Verification code");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(7, "input", 4, 5);
        ɵngcc0.ɵɵlistener("ngModelChange", function SmsChallengeComponent_Template_input_ngModelChange_7_listener($event) { return ctx.model.smsToken = $event; });
        ɵngcc0.ɵɵpipe(9, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(10, SmsChallengeComponent_p_10_Template, 2, 0, "p", 6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "button", 7);
        ɵngcc0.ɵɵpipe(12, "translate");
        ɵngcc0.ɵɵtext(13, " Verify ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(14, "div", 8);
        ɵngcc0.ɵɵelementStart(15, "a", 9);
        ɵngcc0.ɵɵlistener("click", function SmsChallengeComponent_Template_a_click_15_listener() { return ctx.resendTFASms(); });
        ɵngcc0.ɵɵpipe(16, "translate");
        ɵngcc0.ɵɵtext(17, " Send new code ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(18, "a", 10);
        ɵngcc0.ɵɵlistener("click", function SmsChallengeComponent_Template_a_click_18_listener() { return ctx.onCancel.emit(); });
        ɵngcc0.ɵɵpipe(19, "translate");
        ɵngcc0.ɵɵtext(20, " Log in ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r0 = ɵngcc0.ɵɵreference(1);
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(9, 8, "e.g."), " 624327");
        ɵngcc0.ɵɵproperty("ngModel", ctx.model.smsToken);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", !_r0.form.valid || ctx.isLoading);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(12, 10, "Verify"));
        ɵngcc0.ɵɵproperty("disabled", !_r0.form.valid || ctx.isLoading);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(16, 12, "Send new code"));
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(16, _c0, ctx.isLoading));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(19, 14, "Log in"));
    } }, directives: [ɵngcc4.ɵNgNoValidate, ɵngcc4.NgControlStatusGroup, ɵngcc4.NgForm, ɵngcc5.C8yTranslateDirective, ɵngcc6.FormGroupComponent, ɵngcc7.RequiredInputPlaceholderDirective, ɵngcc4.DefaultValueAccessor, ɵngcc4.RequiredValidator, ɵngcc4.NgControlStatus, ɵngcc4.NgModel, ɵngcc8.NgIf, ɵngcc8.NgClass], pipes: [ɵngcc9.C8yTranslatePipe], encapsulation: 2 });
SmsChallengeComponent.ctorParameters = () => [
    { type: LoginService },
    { type: UserService },
    { type: AlertService }
];
SmsChallengeComponent.propDecorators = {
    credentials: [{ type: Input }],
    onCancel: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SmsChallengeComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-sms-challenge',
                template: "<form #twoFactorForm=\"ngForm\" class=\"loginForm\" (ngSubmit)=\"verifyTFACode()\" novalidate>\n  <div class=\"legend form-block center\" translate>\n    Two-factor authentication\n  </div>\n\n  <c8y-form-group>\n    <label translate>Verification code</label>\n    <input\n      [(ngModel)]=\"model.smsToken\"\n      #sms_token=\"ngModel\"\n      type=\"text\"\n      name=\"sms_token\"\n      autofocus\n      autocapitalize=\"off\"\n      autocorrect=\"off\"\n      class=\"form-control\"\n      placeholder=\"{{ 'e.g.' | translate }} 624327\"\n      required\n    />\n    <p *ngIf=\"!twoFactorForm.form.valid || isLoading \" class=\"help-block\" translate>\n      Insert the code received via SMS.\n    </p>\n  </c8y-form-group>\n\n  <button\n    title=\"{{ 'Verify' | translate }}\"\n    [disabled]=\"!twoFactorForm.form.valid || isLoading\"\n    type=\"submit\"\n    class=\"btn btn-primary btn-lg btn-block form-group\"\n    translate\n  >\n    Verify\n  </button>\n\n  <div class=\"d-flex m-t-8\">\n    <a\n      title=\"{{ 'Send new code' | translate }}\"\n      [ngClass]=\"{ disabled: isLoading }\"\n      class=\"small pointer m-r-auto\"\n      (click)=\"resendTFASms()\"\n      translate\n    >\n      Send new code\n    </a>\n    <a\n      title=\"{{ 'Log in' | translate }}\"\n      class=\"small pointer m-l-auto\"\n      (click)=\"onCancel.emit()\"\n      translate\n    >\n      Log in\n    </a>\n  </div>\n</form>\n"
            }]
    }], function () { return [{ type: ɵngcc1.LoginService }, { type: ɵngcc2.UserService }, { type: ɵngcc3.AlertService }]; }, { onCancel: [{
            type: Output
        }], credentials: [{
            type: Input
        }] }); })();
