// @c8y/ngx-components/esm2015/child-devices/child-devices.component.js
import { __awaiter } from "tslib";
import { Component } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { ManagedObjectRealtimeService } from '@c8y/ngx-components';
import { DeviceGridService } from '@c8y/ngx-components/device-grid';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from '@c8y/ngx-components/device-grid';
import * as ɵngcc3 from '@c8y/ngx-components';
export class ChildDevicesComponent {
    constructor(route, deviceGridService) {
        this.route = route;
        this.deviceGridService = deviceGridService;
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.getAllDevices();
            this.columns = this.deviceGridService.getChildDeviceGridColumns();
        });
    }
    getAllDevices() {
        return __awaiter(this, void 0, void 0, function* () {
            this.id = this.route.snapshot.parent.data.contextData;
        });
    }
}
ChildDevicesComponent.ɵfac = function ChildDevicesComponent_Factory(t) { return new (t || ChildDevicesComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ActivatedRoute), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.DeviceGridService)); };
ChildDevicesComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ChildDevicesComponent, selectors: [["c8y-child-devices"]], features: [ɵngcc0.ɵɵProvidersFeature([ManagedObjectRealtimeService])], decls: 2, vars: 6, consts: [[1, "flex-grow", "col-xs-12", "no-gutter", 3, "title", "childDeviceGrid", "columns", "parentDeviceId"]], template: function ChildDevicesComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelement(0, "c8y-device-grid", 0);
        ɵngcc0.ɵɵpipe(1, "translate");
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(1, 4, "Child devices"))("childDeviceGrid", true)("columns", ctx.columns)("parentDeviceId", ctx.id);
    } }, directives: [ɵngcc2.DeviceGridComponent], pipes: [ɵngcc3.C8yTranslatePipe], encapsulation: 2 });
ChildDevicesComponent.ctorParameters = () => [
    { type: ActivatedRoute },
    { type: DeviceGridService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ChildDevicesComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-child-devices',
                template: "<c8y-device-grid\n  [title]=\"'Child devices' | translate\"\n  class=\"flex-grow col-xs-12 no-gutter\"\n  [childDeviceGrid]=\"true\"\n  [columns]=\"columns\"\n  [parentDeviceId]=\"id\"\n>\n</c8y-device-grid>\n",
                providers: [ManagedObjectRealtimeService]
            }]
    }], function () { return [{ type: ɵngcc1.ActivatedRoute }, { type: ɵngcc2.DeviceGridService }]; }, null); })();
