// @c8y/ngx-components/esm2015/auth-configuration/auth-configuration.component.js
import { __awaiter } from "tslib";
import { Component, EventEmitter } from '@angular/core';
import { AuthConfigurationService } from './auth-configuration.service';
import { AlertService, LoginService, ModalService } from '@c8y/ngx-components';
import { TenantLoginOptionsService } from '@c8y/client';
import { cloneDeep } from 'lodash-es';
import { BehaviorSubject } from 'rxjs';
import { tap, switchMap, shareReplay } from 'rxjs/operators';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './auth-configuration.service';
import * as ɵngcc2 from '@c8y/client';
import * as ɵngcc3 from '@c8y/ngx-components';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from '@angular/forms';
import * as ɵngcc6 from './login-settings.component';
import * as ɵngcc7 from './basic-auth-settings.component';
import * as ɵngcc8 from './session-configuration.component';
import * as ɵngcc9 from './tfa-settings.component';

function AuthConfigurationComponent_div_17_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 11);
    ɵngcc0.ɵɵelement(1, "c8y-login-settings", 12);
    ɵngcc0.ɵɵelement(2, "c8y-basic-auth-settings", 12);
    ɵngcc0.ɵɵelement(3, "c8y-session-configuration", 12);
    ɵngcc0.ɵɵelement(4, "c8y-auth-tfa", 12);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("authConfiguration", ctx_r1.authConfiguration);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("authConfiguration", ctx_r1.authConfiguration);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("authConfiguration", ctx_r1.authConfiguration);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("authConfiguration", ctx_r1.authConfiguration);
} }
const _c0 = function (a0) { return { "icon-spin": a0 }; };
export class AuthConfigurationComponent {
    constructor(authConfigurationService, tenantLoginOptionsService, modalService, loginService, alertService) {
        this.authConfigurationService = authConfigurationService;
        this.tenantLoginOptionsService = tenantLoginOptionsService;
        this.modalService = modalService;
        this.loginService = loginService;
        this.alertService = alertService;
        this.reloading$ = new BehaviorSubject(false);
        this.reload = new EventEmitter();
        this.authConfiguration$ = this.reload.pipe(tap(() => this.reloading$.next(true)), switchMap(() => this.authConfigurationService.getAuthConfiguration$()), tap(() => this.reloading$.next(false)), shareReplay(1));
    }
    ngOnInit() {
        this.authConfigurationSubscription = this.authConfiguration$.subscribe((authConfiguration) => {
            this.authConfiguration = authConfiguration;
            this.previousAuthConfiguration = cloneDeep(this.authConfiguration);
        });
        this.loadAuthConfig();
    }
    loadAuthConfig() {
        this.reload.next();
    }
    ngOnDestroy() {
        this.authConfigurationSubscription.unsubscribe();
    }
    save() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.modalService.confirmLogout();
                yield this.authConfigurationService.save(this.authConfiguration, this.previousAuthConfiguration);
                yield this.loginService.logout(true);
            }
            catch (ex) {
                if (ex) {
                    this.alertService.addServerFailure(ex);
                }
            }
        });
    }
}
AuthConfigurationComponent.ɵfac = function AuthConfigurationComponent_Factory(t) { return new (t || AuthConfigurationComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AuthConfigurationService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.TenantLoginOptionsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.ModalService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.LoginService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.AlertService)); };
AuthConfigurationComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AuthConfigurationComponent, selectors: [["c8y-auth-configuration"]], decls: 23, vars: 26, consts: [[3, "placement"], [1, "btn", "btn-link", 3, "title", "click"], ["c8yIcon", "refresh", 3, "ngClass"], ["src", "/users-guide/administration/#authentication"], ["role", "form", "novalidate", "", 1, "card", "card--fullpage"], ["authConfigurationForm", "ngForm"], [1, "card-header", "separator"], [1, "card-title"], ["class", "inner-scroll", 4, "ngIf"], [1, "card-footer", "separator"], [1, "btn", "btn-primary", 3, "title", "disabled", "click"], [1, "inner-scroll"], [3, "authConfiguration"]], template: function AuthConfigurationComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵpipe(2, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "c8y-action-bar-item", 0);
        ɵngcc0.ɵɵelementStart(4, "button", 1);
        ɵngcc0.ɵɵlistener("click", function AuthConfigurationComponent_Template_button_click_4_listener() { return ctx.loadAuthConfig(); });
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelement(6, "i", 2);
        ɵngcc0.ɵɵpipe(7, "async");
        ɵngcc0.ɵɵtext(8);
        ɵngcc0.ɵɵpipe(9, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(10, "c8y-help", 3);
        ɵngcc0.ɵɵelementStart(11, "form", 4, 5);
        ɵngcc0.ɵɵelementStart(13, "div", 6);
        ɵngcc0.ɵɵelementStart(14, "div", 7);
        ɵngcc0.ɵɵtext(15);
        ɵngcc0.ɵɵpipe(16, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(17, AuthConfigurationComponent_div_17_Template, 5, 4, "div", 8);
        ɵngcc0.ɵɵelementStart(18, "div", 9);
        ɵngcc0.ɵɵelementStart(19, "button", 10);
        ɵngcc0.ɵɵlistener("click", function AuthConfigurationComponent_Template_button_click_19_listener() { return ctx.save(); });
        ɵngcc0.ɵɵpipe(20, "translate");
        ɵngcc0.ɵɵtext(21);
        ɵngcc0.ɵɵpipe(22, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r0 = ɵngcc0.ɵɵreference(12);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(2, 10, "Authentication"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("placement", "right");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(5, 12, "Reload"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(24, _c0, ɵngcc0.ɵɵpipeBind1(7, 14, ctx.reloading$)));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(9, 16, "Reload"), " ");
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(16, 18, "Authentication"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.authConfiguration);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(20, 20, "Save"));
        ɵngcc0.ɵɵproperty("disabled", !_r0.form.valid || _r0.form.pristine);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(22, 22, "Save"), " ");
    } }, directives: [ɵngcc3.TitleComponent, ɵngcc3.ActionBarItemComponent, ɵngcc3.IconDirective, ɵngcc4.NgClass, ɵngcc3.HelpComponent, ɵngcc5.ɵNgNoValidate, ɵngcc5.NgControlStatusGroup, ɵngcc5.NgForm, ɵngcc4.NgIf, ɵngcc6.LoginSettingsComponent, ɵngcc7.BasicAuthSettingsComponent, ɵngcc8.SessionConfigurationComponent, ɵngcc9.TfaSettingsComponent], pipes: [ɵngcc3.C8yTranslatePipe, ɵngcc4.AsyncPipe], encapsulation: 2 });
AuthConfigurationComponent.ctorParameters = () => [
    { type: AuthConfigurationService },
    { type: TenantLoginOptionsService },
    { type: ModalService },
    { type: LoginService },
    { type: AlertService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AuthConfigurationComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-auth-configuration',
                template: "<c8y-title>{{ 'Authentication' | translate }}</c8y-title>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button title=\"{{ 'Reload' | translate }}\" class=\"btn btn-link\" (click)=\"loadAuthConfig()\">\n    <i c8yIcon=\"refresh\" [ngClass]=\"{ 'icon-spin': reloading$ | async }\"></i>\n    {{ 'Reload' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<c8y-help src=\"/users-guide/administration/#authentication\"></c8y-help>\n\n<form role=\"form\" class=\"card card--fullpage\" #authConfigurationForm=\"ngForm\" novalidate>\n  <div class=\"card-header separator\">\n    <div class=\"card-title\">\n      {{ 'Authentication' | translate }}\n    </div>\n  </div>\n  <div class=\"inner-scroll\" *ngIf=\"authConfiguration\">\n    <c8y-login-settings [authConfiguration]=\"authConfiguration\"></c8y-login-settings>\n    <c8y-basic-auth-settings [authConfiguration]=\"authConfiguration\"></c8y-basic-auth-settings>\n    <c8y-session-configuration [authConfiguration]=\"authConfiguration\"></c8y-session-configuration>\n    <c8y-auth-tfa [authConfiguration]=\"authConfiguration\"></c8y-auth-tfa>\n  </div>\n  <div class=\"card-footer separator\">\n    <button\n      title=\"{{ 'Save' | translate }}\"\n      class=\"btn btn-primary\"\n      (click)=\"save()\"\n      [disabled]=\"!authConfigurationForm.form.valid || authConfigurationForm.form.pristine\"\n    >\n      {{ 'Save' | translate }}\n    </button>\n  </div>\n</form>\n"
            }]
    }], function () { return [{ type: ɵngcc1.AuthConfigurationService }, { type: ɵngcc2.TenantLoginOptionsService }, { type: ɵngcc3.ModalService }, { type: ɵngcc3.LoginService }, { type: ɵngcc3.AlertService }]; }, null); })();
