// @c8y/ngx-components/esm2015/auth-configuration/auth-configuration.module.js
import { NgModule } from '@angular/core';
import { CoreModule, HOOK_NAVIGATOR_NODES, HOOK_ROUTE, HOOK_TABS } from '@c8y/ngx-components';
import { BasicAuthSettingsComponent } from './basic-auth-settings.component';
import { TabsFactory } from './factories/tabs.factory';
import { NavigationFactory } from './factories/navigation.factory';
import { AuthConfigurationGuard } from './auth-configuration.guard';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { AuthConfigurationService } from './auth-configuration.service';
import { AuthConfigurationComponent } from './auth-configuration.component';
import { TfaSettingsComponent } from './tfa-settings.component';
import { SessionConfigurationComponent } from './session-configuration.component';
import { LoginSettingsComponent } from './login-settings.component';
import { SsoConfigurationGuard } from './sso-configuration.guard';
import { SsoConfigurationService } from './sso-configuration.service';
import * as ɵngcc0 from '@angular/core';
const ɵ0 = [
    {
        path: 'auth-configuration',
        redirectTo: 'auth-configuration/basic_settings',
        pathMatch: 'full'
    },
    {
        path: 'auth-configuration/basic_settings',
        component: AuthConfigurationComponent,
        canActivate: [AuthConfigurationGuard]
    },
    {
        path: 'auth-configuration/single_sign-on',
        canActivate: [AuthConfigurationGuard, SsoConfigurationGuard],
        children: [] // This route exists to secure path by guards. Sso component exists only in Angular Js, so here children is used to avoid compilation error.
    }
];
export class AuthConfigurationModule {
}
AuthConfigurationModule.ɵfac = function AuthConfigurationModule_Factory(t) { return new (t || AuthConfigurationModule)(); };
AuthConfigurationModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: AuthConfigurationModule });
AuthConfigurationModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        AuthConfigurationService,
        AuthConfigurationGuard,
        SsoConfigurationService,
        SsoConfigurationGuard,
        { provide: HOOK_TABS, useClass: TabsFactory, multi: true },
        {
            provide: HOOK_NAVIGATOR_NODES,
            useClass: NavigationFactory,
            multi: true
        },
        {
            provide: HOOK_ROUTE,
            useValue: ɵ0,
            multi: true
        }
    ], imports: [[CoreModule, PopoverModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AuthConfigurationModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    SessionConfigurationComponent,
                    LoginSettingsComponent,
                    BasicAuthSettingsComponent,
                    AuthConfigurationComponent,
                    TfaSettingsComponent
                ],
                exports: [],
                imports: [CoreModule, PopoverModule],
                providers: [
                    AuthConfigurationService,
                    AuthConfigurationGuard,
                    SsoConfigurationService,
                    SsoConfigurationGuard,
                    { provide: HOOK_TABS, useClass: TabsFactory, multi: true },
                    {
                        provide: HOOK_NAVIGATOR_NODES,
                        useClass: NavigationFactory,
                        multi: true
                    },
                    {
                        provide: HOOK_ROUTE,
                        useValue: ɵ0,
                        multi: true
                    }
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AuthConfigurationModule, { declarations: function () { return [SessionConfigurationComponent,
        LoginSettingsComponent,
        BasicAuthSettingsComponent,
        AuthConfigurationComponent,
        TfaSettingsComponent]; }, imports: function () { return [CoreModule, PopoverModule]; } }); })();
export { ɵ0 };
