// @c8y/ngx-components/esm2015/auth-configuration/sso-configuration.service.js
import { Injectable } from '@angular/core';
import { TenantLoginOptionsService, TenantLoginOptionType } from '@c8y/client';
import { defer, of } from 'rxjs';
import { catchError, map, publishReplay, refCount, take } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/client/lib/src/tenant-login-options/TenantLoginOptionsService";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/client';
export class SsoConfigurationService {
    constructor(loginOptionsService) {
        this.loginOptionsService = loginOptionsService;
        this.isSsoConfigurationAvailable$ = defer(() => this.loginOptionsService.detail(TenantLoginOptionType.OAUTH2)).pipe(map(res => true), catchError(({ res }) => (res.status === 403 /* Forbidden */ ? of(false) : of(true))), 
        // Both guard and tab.factory ask for OAUTH2 - this observable will be send one request in 1 second.
        publishReplay(1, 1000), refCount(), take(1));
    }
    showSsoConfiguration$() {
        return this.isSsoConfigurationAvailable$;
    }
}
SsoConfigurationService.ɵfac = function SsoConfigurationService_Factory(t) { return new (t || SsoConfigurationService)(ɵngcc0.ɵɵinject(ɵngcc1.TenantLoginOptionsService)); };
SsoConfigurationService.ɵprov = i0.ɵɵdefineInjectable({ factory: function SsoConfigurationService_Factory() { return new SsoConfigurationService(i0.ɵɵinject(i1.TenantLoginOptionsService)); }, token: SsoConfigurationService, providedIn: "root" });
SsoConfigurationService.ctorParameters = () => [
    { type: TenantLoginOptionsService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SsoConfigurationService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: ɵngcc1.TenantLoginOptionsService }]; }, null); })();
