// @c8y/ngx-components/esm2015/protocol-lpwan/multiple-lns-connectors/connection-info-with-download-csv.component.js
import { __awaiter } from "tslib";
import { Component, Input } from '@angular/core';
import { BsModalRef } from 'ngx-bootstrap/modal';
import { MultipleLnsConnectorService } from './multiple-lns-connector.service';
import { saveAs } from 'file-saver';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/modal';
import * as ɵngcc2 from './multiple-lns-connector.service';
import * as ɵngcc3 from '@c8y/ngx-components';
export class ConnectionInfoWithDownloadCsvComponent {
    constructor(modal, connectorService) {
        this.modal = modal;
        this.connectorService = connectorService;
    }
    dismiss() {
        this.modal.hide();
    }
    download() {
        return __awaiter(this, void 0, void 0, function* () {
            const url = `/service/${this.appData.contextPath}${this.messageData.URL}`;
            const res = yield this.connectorService.download(url);
            const streamData = yield res.blob();
            saveAs(streamData, 'DeviceList.csv');
        });
    }
}
ConnectionInfoWithDownloadCsvComponent.ɵfac = function ConnectionInfoWithDownloadCsvComponent_Factory(t) { return new (t || ConnectionInfoWithDownloadCsvComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BsModalRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.MultipleLnsConnectorService)); };
ConnectionInfoWithDownloadCsvComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ConnectionInfoWithDownloadCsvComponent, selectors: [["connection-info-with-download-csv"]], inputs: { messageData: "messageData", appData: "appData", modalTitle: "modalTitle" }, decls: 19, vars: 15, consts: [[1, "modal-dialog"], [1, "modal-content"], [1, "c8y-prompt", "alert", "alert-danger"], [1, "m-b-16"], [1, "dlt-c8y-icon-exclamation-circle"], [1, "text-break-word"], [1, "btn-default", 3, "click"], [1, "alert-footer"], [1, "btn", "btn-default", 3, "title", "click"]], template: function ConnectionInfoWithDownloadCsvComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelementStart(3, "h3", 3);
        ɵngcc0.ɵɵelement(4, "i", 4);
        ɵngcc0.ɵɵelementStart(5, "span");
        ɵngcc0.ɵɵtext(6);
        ɵngcc0.ɵɵpipe(7, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(8, "p", 5);
        ɵngcc0.ɵɵtext(9);
        ɵngcc0.ɵɵpipe(10, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "span", 6);
        ɵngcc0.ɵɵlistener("click", function ConnectionInfoWithDownloadCsvComponent_Template_span_click_11_listener() { return ctx.download(); });
        ɵngcc0.ɵɵtext(12);
        ɵngcc0.ɵɵpipe(13, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(14, "div", 7);
        ɵngcc0.ɵɵelementStart(15, "button", 8);
        ɵngcc0.ɵɵlistener("click", function ConnectionInfoWithDownloadCsvComponent_Template_button_click_15_listener() { return ctx.dismiss(); });
        ɵngcc0.ɵɵpipe(16, "translate");
        ɵngcc0.ɵɵtext(17);
        ɵngcc0.ɵɵpipe(18, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(7, 5, ctx.modalTitle));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(10, 7, ctx.messageData.message), " ");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(13, 9, "Click to download link with affected devices"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(16, 11, "Cancel"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(18, 13, "Cancel"), " ");
    } }, pipes: [ɵngcc3.C8yTranslatePipe], encapsulation: 2 });
ConnectionInfoWithDownloadCsvComponent.ctorParameters = () => [
    { type: BsModalRef },
    { type: MultipleLnsConnectorService }
];
ConnectionInfoWithDownloadCsvComponent.propDecorators = {
    messageData: [{ type: Input }],
    appData: [{ type: Input }],
    modalTitle: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ConnectionInfoWithDownloadCsvComponent, [{
        type: Component,
        args: [{
                selector: 'connection-info-with-download-csv',
                template: "<div class=\"modal-dialog\">\n  <div class=\"modal-content\">\n    <div class=\"c8y-prompt alert alert-danger\">\n      <h3 class=\"m-b-16\">\n        <i class=\"dlt-c8y-icon-exclamation-circle\"></i>\n        <span>{{ modalTitle | translate }}</span>\n      </h3>\n      <p class=\"text-break-word\">\n        {{ messageData.message | translate }}\n      </p>\n      \n      <span class=\"btn-default\" (click)=\"download()\">{{\n        'Click to download link with affected devices' | translate\n      }}</span>\n      <div class=\"alert-footer\">\n        <button title=\"{{ 'Cancel' | translate }}\" class=\"btn btn-default\" (click)=\"dismiss()\">\n          {{ 'Cancel' | translate }}\n        </button>\n      </div>\n    </div>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.BsModalRef }, { type: ɵngcc2.MultipleLnsConnectorService }]; }, { messageData: [{
            type: Input
        }], appData: [{
            type: Input
        }], modalTitle: [{
            type: Input
        }] }); })();
