// @c8y/ngx-components/esm2015/protocol-lpwan/multiple-lns-connectors/multiple-lns-connectors.module.js
import { NgModule } from '@angular/core';
import { CommonModule, CoreModule, FormsModule } from '@c8y/ngx-components';
import { NoConnectionsFoundComponent } from './no-connections-found.component';
import { RouterModule } from '@angular/router';
import { SigfoxMultipleLnsConnectorComponent } from './sigfox-multiple-lns-connector.component';
import { ActilityMultipleLnsConnectorComponent } from './actility-multiple-lns-connector.component';
import { MultipleLnsConnectorService } from './multiple-lns-connector.service';
import { ConnectionInfoWithDownloadCsvComponent } from './connection-info-with-download-csv.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
const settingsRoutes = [
    {
        path: 'connectivitySettings/multiple_lns_connectors_sigfox',
        component: SigfoxMultipleLnsConnectorComponent
    },
    {
        path: 'connectivitySettings/multiple_lns_connectors_actility',
        component: ActilityMultipleLnsConnectorComponent
    }
];
export class MultipleLnsConnectorsModule {
}
MultipleLnsConnectorsModule.ɵfac = function MultipleLnsConnectorsModule_Factory(t) { return new (t || MultipleLnsConnectorsModule)(); };
MultipleLnsConnectorsModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: MultipleLnsConnectorsModule });
MultipleLnsConnectorsModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        MultipleLnsConnectorService
    ], imports: [[
            CoreModule,
            CommonModule,
            FormsModule,
            RouterModule.forChild(settingsRoutes)
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(MultipleLnsConnectorsModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    SigfoxMultipleLnsConnectorComponent,
                    NoConnectionsFoundComponent,
                    ActilityMultipleLnsConnectorComponent,
                    ConnectionInfoWithDownloadCsvComponent
                ],
                imports: [
                    CoreModule,
                    CommonModule,
                    FormsModule,
                    RouterModule.forChild(settingsRoutes)
                ],
                providers: [
                    MultipleLnsConnectorService
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(MultipleLnsConnectorsModule, { declarations: function () { return [SigfoxMultipleLnsConnectorComponent,
        NoConnectionsFoundComponent,
        ActilityMultipleLnsConnectorComponent,
        ConnectionInfoWithDownloadCsvComponent]; }, imports: function () { return [CoreModule,
        CommonModule,
        FormsModule, ɵngcc1.RouterModule]; } }); })();
