// @c8y/ngx-components/esm2015/core/dynamic-forms/checkbox/checkbox.type.component.js
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { FieldType } from '@ngx-formly/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/forms';
import * as ɵngcc2 from '@ngx-formly/core';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from '../../i18n/c8y-translate.directive';
import * as ɵngcc5 from 'ngx-bootstrap/popover';
import * as ɵngcc6 from '../../common/icon.directive';
import * as ɵngcc7 from '../../common/humanize.pipe';

function FieldCheckbox_span_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵelementStart(1, "em", 3);
    ɵngcc0.ɵɵtext(2, "(required)");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function FieldCheckbox_button_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "button", 4);
    ɵngcc0.ɵɵelement(1, "i", 5);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("popover", ctx_r1.to.description);
} }
export class FieldCheckbox extends FieldType {
    constructor() {
        super(...arguments);
        this.defaultOptions = {
            templateOptions: {
                indeterminate: true,
                formCheck: 'custom' // 'custom' | 'custom-inline' | 'custom-switch' | 'stacked' | 'inline' | 'nolabel'
            }
        };
    }
}
FieldCheckbox.ɵfac = /*@__PURE__*/ function () { let ɵFieldCheckbox_BaseFactory; return function FieldCheckbox_Factory(t) { return (ɵFieldCheckbox_BaseFactory || (ɵFieldCheckbox_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(FieldCheckbox)))(t || FieldCheckbox); }; }();
FieldCheckbox.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: FieldCheckbox, selectors: [["c8y-field-checkbox"]], features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 8, vars: 22, consts: [["type", "checkbox", 3, "indeterminate", "formControl", "formlyAttributes"], [4, "ngIf"], ["class", "btn-clean m-l-4 flex-item-middle", "placement", "auto", "outsideClick", "true", "type", "button", 3, "popover", 4, "ngIf"], ["translate", "", 1, "m-l-4"], ["placement", "auto", "outsideClick", "true", "type", "button", 1, "btn-clean", "m-l-4", "flex-item-middle", 3, "popover"], ["c8yIcon", "question-circle-o", 1, "text-primary"]], template: function FieldCheckbox_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "label");
        ɵngcc0.ɵɵelement(1, "input", 0);
        ɵngcc0.ɵɵelement(2, "span");
        ɵngcc0.ɵɵelementStart(3, "span");
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵpipe(5, "humanize");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(6, FieldCheckbox_span_6_Template, 3, 0, "span", 1);
        ɵngcc0.ɵɵtemplate(7, FieldCheckbox_button_7_Template, 2, 1, "button", 2);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassProp("c8y-checkbox", !ctx.to.switchMode)("c8y-switch", ctx.to.switchMode)("has-error", ctx.showError);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵclassProp("is-invalid", ctx.showError)("form-check-input", ctx.to.formCheck.indexOf("custom") === -1)("position-static", ctx.to.formCheck === "nolabel")("custom-control-input", ctx.to.formCheck.indexOf("custom") === 0);
        ɵngcc0.ɵɵproperty("indeterminate", ctx.to.indeterminate && ctx.formControl.value == null)("formControl", ctx.formControl)("formlyAttributes", ctx.field);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 20, ctx.to.label), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.to.required && ctx.to.hideRequiredMarker !== true);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !!ctx.to.description);
    } }, directives: [ɵngcc1.CheckboxControlValueAccessor, ɵngcc1.NgControlStatus, ɵngcc1.FormControlDirective, ɵngcc2.FormlyAttributes, ɵngcc3.NgIf, ɵngcc4.C8yTranslateDirective, ɵngcc5.PopoverDirective, ɵngcc6.IconDirective], pipes: [ɵngcc7.HumanizePipe], encapsulation: 2, changeDetection: 0 });
FieldCheckbox.CONFIG = {
    types: [
        {
            name: 'checkbox',
            component: FieldCheckbox
        },
        {
            name: 'boolean',
            extends: 'checkbox'
        },
        {
            name: 'switch',
            extends: 'checkbox',
            defaultOptions: {
                templateOptions: {
                    switchMode: true,
                    indeterminate: false,
                }
            },
        }
    ]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(FieldCheckbox, [{
        type: Component,
        args: [{
                selector: 'c8y-field-checkbox',
                template: "  <label\n    [class.c8y-checkbox]=\"!to.switchMode\"\n    [class.c8y-switch]=\"to.switchMode\"\n    [class.has-error]=\"showError\">\n    <input\n      type=\"checkbox\"\n      [class.is-invalid]=\"showError\"\n      [class.form-check-input]=\"to.formCheck.indexOf('custom') === -1\"\n      [class.position-static]=\"to.formCheck === 'nolabel'\"\n      [class.custom-control-input]=\"to.formCheck.indexOf('custom') === 0\"\n      [indeterminate]=\"to.indeterminate && formControl.value == null\"\n      [formControl]=\"formControl\"\n      [formlyAttributes]=\"field\"\n    /><span></span><span> {{ to.label | humanize}} </span>\n    <span *ngIf=\"to.required && to.hideRequiredMarker !== true\"><em class=\"m-l-4\" translate>(required)</em></span>\n    <button\n      *ngIf=\"!!to.description\"\n      class=\"btn-clean m-l-4 flex-item-middle\"\n      [popover]=\"to.description\"\n      placement=\"auto\"\n      outsideClick=\"true\"\n      type=\"button\"\n    >\n      <i c8yIcon=\"question-circle-o\" class=\"text-primary\"></i>\n  </button>\n  </label>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], null, null); })();
