// @c8y/ngx-components/esm2015/device-shell/shell/shell.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { CoreModule } from '@c8y/ngx-components';
import { OperationsTimelineModule } from '@c8y/ngx-components/operations/operations-timeline';
import { CommandTemplatesModule } from '../command-templates';
import { DeviceShellSharedModule } from '../shared';
import { DeviceShellComponent } from './shell.component';
import * as ɵngcc0 from '@angular/core';
export class ShellModule {
}
ShellModule.ɵfac = function ShellModule_Factory(t) { return new (t || ShellModule)(); };
ShellModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: ShellModule });
ShellModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            CoreModule,
            OperationsTimelineModule,
            CommandTemplatesModule,
            DeviceShellSharedModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ShellModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    CoreModule,
                    OperationsTimelineModule,
                    CommandTemplatesModule,
                    DeviceShellSharedModule
                ],
                declarations: [DeviceShellComponent],
                exports: [DeviceShellComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ShellModule, { declarations: function () { return [DeviceShellComponent]; }, imports: function () { return [CommonModule,
        CoreModule,
        OperationsTimelineModule,
        CommandTemplatesModule,
        DeviceShellSharedModule]; }, exports: function () { return [DeviceShellComponent]; } }); })();
