// @c8y/ngx-components/esm2015/device-shell/shell/shell.component.js
import { __awaiter } from "tslib";
import { Component } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { OperationService } from '@c8y/client';
import { AlertService, gettext, ManagedObjectRealtimeService, OperationRealtimeService } from '@c8y/ngx-components';
import { includes, isEmpty } from 'lodash-es';
import { BsModalService } from 'ngx-bootstrap/modal';
import { BehaviorSubject, of, pipe, Subject } from 'rxjs';
import { filter, map, startWith, takeUntil } from 'rxjs/operators';
import { CommandTemplatesComponent } from '../command-templates';
import { CommandDeliveryType, DeviceShellService } from '../shared';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../shared';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from '@c8y/client';
import * as ɵngcc4 from '@angular/router';
import * as ɵngcc5 from 'ngx-bootstrap/modal';
import * as ɵngcc6 from '@angular/forms';
import * as ɵngcc7 from '@angular/common';
import * as ɵngcc8 from '@c8y/ngx-components/operations/operations-timeline';

const _c0 = function (a0) { return { deliveryType: a0 }; };
function DeviceShellComponent_ng_container_20_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 20);
    ɵngcc0.ɵɵlistener("click", function DeviceShellComponent_ng_container_20_button_1_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r9); const deliveryType_r5 = ɵngcc0.ɵɵnextContext().$implicit; const ctx_r7 = ɵngcc0.ɵɵnextContext(); return ctx_r7.execute(deliveryType_r5.name); });
    ɵngcc0.ɵɵpipe(1, "async");
    ɵngcc0.ɵɵelementStart(2, "span", 21);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵpipe(7, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const deliveryType_r5 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r6 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("disabled", !(ctx_r6.command == null ? null : ctx_r6.command.text) || ɵngcc0.ɵɵpipeBind1(1, 3, ctx_r6.sendingCommand$));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("title", deliveryType_r5.default ? ɵngcc0.ɵɵpipeBind1(3, 5, "Execute") : ɵngcc0.ɵɵpipeBind2(4, 7, ctx_r6.executeViaLabel, ɵngcc0.ɵɵpureFunction1(15, _c0, deliveryType_r5.name)));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", deliveryType_r5.default ? ɵngcc0.ɵɵpipeBind1(6, 10, "Execute") : ɵngcc0.ɵɵpipeBind2(7, 12, ctx_r6.executeViaLabel, ɵngcc0.ɵɵpureFunction1(17, _c0, deliveryType_r5.name)), " ");
} }
function DeviceShellComponent_ng_container_20_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, DeviceShellComponent_ng_container_20_button_1_Template, 8, 19, "button", 19);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const deliveryType_r5 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", deliveryType_r5.isSupportedByCommand);
} }
function DeviceShellComponent_ng_template_28_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const operation_r11 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate((operation_r11.c8y_Command == null ? null : operation_r11.c8y_Command.text) || operation_r11.description);
} }
function DeviceShellComponent_ng_template_30_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "div", 23);
    ɵngcc0.ɵɵtext(2, "Command");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "pre", 24);
    ɵngcc0.ɵɵtext(4, "              ");
    ɵngcc0.ɵɵelementStart(5, "code");
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(7, "\n          ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const operation_r12 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵtextInterpolate(operation_r12.c8y_Command.text);
} }
function DeviceShellComponent_ng_template_30_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "div", 23);
    ɵngcc0.ɵɵtext(2, "Response");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "pre", 24);
    ɵngcc0.ɵɵtext(4, "              ");
    ɵngcc0.ɵɵelementStart(5, "code");
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(7, "\n          ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const operation_r12 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵtextInterpolate(operation_r12.c8y_Command.result);
} }
function DeviceShellComponent_ng_template_30_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, DeviceShellComponent_ng_template_30_div_0_Template, 8, 1, "div", 22);
    ɵngcc0.ɵɵtemplate(1, DeviceShellComponent_ng_template_30_div_1_Template, 8, 1, "div", 22);
} if (rf & 2) {
    const operation_r12 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("ngIf", operation_r12.c8y_Command == null ? null : operation_r12.c8y_Command.text);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", operation_r12.c8y_Command == null ? null : operation_r12.c8y_Command.result);
} }
const _c1 = function (a0) { return [a0]; };
const _c2 = function () { return ["c8y_Command"]; };
export class DeviceShellComponent {
    constructor(service, operationRealtime, moRealtime, operationService, route, modalService, alertService) {
        this.service = service;
        this.operationRealtime = operationRealtime;
        this.moRealtime = moRealtime;
        this.operationService = operationService;
        this.route = route;
        this.modalService = modalService;
        this.alertService = alertService;
        this.device = this.route.snapshot.parent.data.contextData;
        this.device$ = of({});
        this.deliveryTypes = (this.service.getDeliveryTypes() || []).map(deliveryType => (Object.assign(Object.assign({}, deliveryType), { isSupportedByCommand: true })));
        this.command = {};
        this.smsEnabled = false;
        this.sendingCommand$ = new BehaviorSubject(false);
        this.filterPipe = pipe(map((operations) => (operations || []).filter((operation) => !!operation.c8y_Command)));
        this.executeViaLabel = gettext('Execute via ({{deliveryType}})');
        this.destroyed$ = new Subject();
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            this.smsEnabled = yield this.service.canSendCommandsViaSMS();
            this.device$ = this.moRealtime.onUpdate$(this.device.id).pipe(startWith(this.device));
            this.operations = yield this.operationService.list({
                deviceId: this.device.id,
                fragmentType: 'c8y_Command',
                dateFrom: new Date(0).toISOString(),
                dateTo: new Date().toISOString(),
                pageSize: 50,
                withTotalPages: true,
                revert: true
            });
            this.operationRealtime
                .onCreate$(this.device.id)
                .pipe(takeUntil(this.destroyed$))
                .subscribe(() => this.alertService.success(gettext('Command sent.')));
            this.operationRealtime
                .onUpdate$(this.device.id)
                .pipe(filter(op => op.failureReason !== 'Operation cancelled by user.'), // avoid duplicate alerts
            takeUntil(this.destroyed$))
                .subscribe(() => this.alertService.success(gettext('Command status updated.')));
        });
    }
    getPredefinedCommand() {
        const modal = this.modalService.show(CommandTemplatesComponent);
        modal.content.onTemplateSelected.pipe(takeUntil(this.destroyed$)).subscribe(result => {
            this.command = Object.assign({}, result.commandTemplate);
            this.deliveryTypes = this.deliveryTypes.map(deliveryType => (Object.assign(Object.assign({}, deliveryType), { isSupportedByCommand: isEmpty(this.command.deliveryTypes) ||
                    includes(this.command.deliveryTypes, deliveryType.name) })));
        });
    }
    resetSupportedDeliveryTypes() {
        this.deliveryTypes = (this.service.getDeliveryTypes() || []).map(deliveryType => (Object.assign(Object.assign({}, deliveryType), { isSupportedByCommand: true })));
    }
    execute(commandDeliveryType) {
        return __awaiter(this, void 0, void 0, function* () {
            const useSMS = commandDeliveryType === CommandDeliveryType.SMS;
            if (useSMS && !this.smsEnabled) {
                this.alertService.warning(gettext('SMS transport is not configured.'));
                return;
            }
            this.sendingCommand$.next(true);
            yield this.service.createCommandOperation(this.device.id, this.command, commandDeliveryType);
            this.command.text = '';
            this.resetSupportedDeliveryTypes();
            this.sendingCommand$.next(false);
        });
    }
    ngOnDestroy() {
        this.destroyed$.next();
        this.destroyed$.complete();
    }
}
DeviceShellComponent.ɵfac = function DeviceShellComponent_Factory(t) { return new (t || DeviceShellComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.DeviceShellService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.OperationRealtimeService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ManagedObjectRealtimeService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.OperationService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.ActivatedRoute), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.BsModalService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.AlertService)); };
DeviceShellComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: DeviceShellComponent, selectors: [["c8y-device-shell"]], features: [ɵngcc0.ɵɵProvidersFeature([OperationRealtimeService, ManagedObjectRealtimeService])], decls: 32, vars: 28, consts: [[3, "placement"], [3, "service"], [1, "card", "content-fullpage", "d-grid", "grid__col--6-6--md"], [1, "inner-scroll", "d-flex", "d-col", "bg-white"], [1, "card-header", "large-padding", "separator", "sticky-top"], [1, "card-title"], [1, "card-block", "d-flex", "d-col", "flex-grow", "large-padding"], [1, "flex-row", "p-b-16"], ["type", "button", 1, "btn", "btn-default", "btn-sm", 3, "title", "click"], [1, "flex-item-right"], [3, "mo"], [1, "form-control", "inner-scroll", "flex-grow", "bg-gray-lighter", 3, "ngModel", "ngModelChange"], [1, "card-footer", "large-padding", "separator"], [4, "ngFor", "ngForOf"], [1, "inner-scroll", "bg-gray-white"], [1, "card-block", "large-padding"], [3, "operations", "sourceId", "filterPipe", "bodyTemplate", "footerTemplates", "propertiesToHide"], ["timelineItemBody", ""], ["timelineItemFooter", ""], ["class", "btn btn-primary", "type", "button", 3, "disabled", "click", 4, "ngIf"], ["type", "button", 1, "btn", "btn-primary", 3, "disabled", "click"], [3, "title"], [4, "ngIf"], ["translate", "", 1, "legend", "form-block"], [1, "text-pre-normal"]], template: function DeviceShellComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-action-bar-item", 0);
        ɵngcc0.ɵɵelement(1, "c8y-realtime-btn", 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelementStart(3, "div", 3);
        ɵngcc0.ɵɵelementStart(4, "div", 4);
        ɵngcc0.ɵɵelementStart(5, "h4", 5);
        ɵngcc0.ɵɵtext(6);
        ɵngcc0.ɵɵpipe(7, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(8, "div", 6);
        ɵngcc0.ɵɵelementStart(9, "div", 7);
        ɵngcc0.ɵɵelementStart(10, "button", 8);
        ɵngcc0.ɵɵlistener("click", function DeviceShellComponent_Template_button_click_10_listener() { return ctx.getPredefinedCommand(); });
        ɵngcc0.ɵɵpipe(11, "translate");
        ɵngcc0.ɵɵtext(12);
        ɵngcc0.ɵɵpipe(13, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(14, "div", 9);
        ɵngcc0.ɵɵelement(15, "device-status", 10);
        ɵngcc0.ɵɵpipe(16, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(17, "textarea", 11);
        ɵngcc0.ɵɵlistener("ngModelChange", function DeviceShellComponent_Template_textarea_ngModelChange_17_listener($event) { return ctx.command.text = $event; })("ngModelChange", function DeviceShellComponent_Template_textarea_ngModelChange_17_listener($event) { return $event || ctx.resetSupportedDeliveryTypes(); });
        ɵngcc0.ɵɵtext(18, "      ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(19, "div", 12);
        ɵngcc0.ɵɵtemplate(20, DeviceShellComponent_ng_container_20_Template, 2, 1, "ng-container", 13);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(21, "div", 14);
        ɵngcc0.ɵɵelementStart(22, "div", 4);
        ɵngcc0.ɵɵelementStart(23, "h4", 5);
        ɵngcc0.ɵɵtext(24);
        ɵngcc0.ɵɵpipe(25, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(26, "div", 15);
        ɵngcc0.ɵɵelement(27, "c8y-operations-timeline", 16);
        ɵngcc0.ɵɵtemplate(28, DeviceShellComponent_ng_template_28_Template, 2, 1, "ng-template", null, 17, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(30, DeviceShellComponent_ng_template_30_Template, 2, 2, "ng-template", null, 18, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r1 = ɵngcc0.ɵɵreference(29);
        const _r3 = ɵngcc0.ɵɵreference(31);
        ɵngcc0.ɵɵproperty("placement", "right");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("service", ctx.operationRealtime);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(7, 15, "Command"), " ");
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(11, 17, "Display a list of predefined commands"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(13, 19, "Predefined commands"), " ");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("mo", ɵngcc0.ɵɵpipeBind1(16, 21, ctx.device$));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngModel", ctx.command.text);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.deliveryTypes);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(25, 23, "Operations"), " ");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("operations", ctx.operations)("sourceId", ctx.device.id)("filterPipe", ctx.filterPipe)("bodyTemplate", _r1)("footerTemplates", ɵngcc0.ɵɵpureFunction1(25, _c1, _r3))("propertiesToHide", ɵngcc0.ɵɵpureFunction0(27, _c2));
    } }, directives: [ɵngcc2.ActionBarItemComponent, ɵngcc2.RealtimeButtonComponent, ɵngcc2.DeviceStatusComponent, ɵngcc6.DefaultValueAccessor, ɵngcc6.NgControlStatus, ɵngcc6.NgModel, ɵngcc7.NgForOf, ɵngcc8.OperationsTimelineComponent, ɵngcc7.NgIf, ɵngcc2.C8yTranslateDirective], pipes: [ɵngcc2.C8yTranslatePipe, ɵngcc7.AsyncPipe], encapsulation: 2 });
DeviceShellComponent.ctorParameters = () => [
    { type: DeviceShellService },
    { type: OperationRealtimeService },
    { type: ManagedObjectRealtimeService },
    { type: OperationService },
    { type: ActivatedRoute },
    { type: BsModalService },
    { type: AlertService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DeviceShellComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-device-shell',
                template: "<c8y-action-bar-item [placement]=\"'right'\">\n  <c8y-realtime-btn [service]=\"operationRealtime\"></c8y-realtime-btn>\n</c8y-action-bar-item>\n\n<div class=\"card content-fullpage d-grid grid__col--6-6--md\">\n  <div class=\"inner-scroll d-flex d-col bg-white\">\n    <div class=\"card-header large-padding separator sticky-top\">\n      <h4 class=\"card-title\">\n        {{ 'Command' | translate }}\n      </h4>\n    </div>\n\n    <div class=\"card-block d-flex d-col flex-grow large-padding\">\n      <div class=\"flex-row p-b-16\">\n        <button\n          class=\"btn btn-default btn-sm\"\n          type=\"button\"\n          (click)=\"getPredefinedCommand()\"\n          [title]=\"'Display a list of predefined commands' | translate\"\n        >\n          {{ 'Predefined commands' | translate }}\n        </button>\n\n        <div class=\"flex-item-right\">\n          <device-status [mo]=\"device$ | async\"></device-status>\n        </div>\n      </div>\n      <textarea\n        class=\"form-control inner-scroll flex-grow bg-gray-lighter\"\n        [(ngModel)]=\"command.text\"\n        (ngModelChange)=\"$event || resetSupportedDeliveryTypes()\"\n      >\n      </textarea>\n    </div>\n\n    <div class=\"card-footer large-padding separator\">\n      <ng-container *ngFor=\"let deliveryType of deliveryTypes\">\n        <button\n          class=\"btn btn-primary\"\n          type=\"button\"\n          *ngIf=\"deliveryType.isSupportedByCommand\"\n          [disabled]=\"!command?.text || (sendingCommand$ | async)\"\n          (click)=\"execute(deliveryType.name)\"\n        >\n          <span\n            [title]=\"\n              deliveryType.default\n                ? ('Execute' | translate)\n                : (executeViaLabel | translate: { deliveryType: deliveryType.name })\n            \"\n          >\n            {{\n              deliveryType.default\n                ? ('Execute' | translate)\n                : (executeViaLabel | translate: { deliveryType: deliveryType.name })\n            }}\n          </span>\n        </button>\n      </ng-container>\n    </div>\n  </div>\n  <div class=\"inner-scroll bg-gray-white\">\n    <div class=\"card-header large-padding separator sticky-top\">\n      <h4 class=\"card-title\">\n        {{ 'Operations' | translate }}\n      </h4>\n    </div>\n    <div class=\"card-block large-padding\">\n      <c8y-operations-timeline\n        [operations]=\"operations\"\n        [sourceId]=\"device.id\"\n        [filterPipe]=\"filterPipe\"\n        [bodyTemplate]=\"timelineItemBody\"\n        [footerTemplates]=\"[timelineItemFooter]\"\n        [propertiesToHide]=\"['c8y_Command']\"\n      ></c8y-operations-timeline>\n      <ng-template #timelineItemBody let-operation>\n        <small>{{ operation.c8y_Command?.text || operation.description }}</small>\n      </ng-template>\n      <ng-template #timelineItemFooter let-operation>\n        <div *ngIf=\"operation.c8y_Command?.text\">\n          <div class=\"legend form-block\" translate>Command</div>\n          <pre class=\"text-pre-normal\">\n              <code>{{operation.c8y_Command.text}}</code>\n          </pre>\n        </div>\n        <div *ngIf=\"operation.c8y_Command?.result\">\n          <div class=\"legend form-block\" translate>Response</div>\n          <pre class=\"text-pre-normal\">\n              <code>{{operation.c8y_Command.result}}</code>\n          </pre>\n        </div>\n      </ng-template>\n    </div>\n  </div>\n</div>\n",
                providers: [OperationRealtimeService, ManagedObjectRealtimeService]
            }]
    }], function () { return [{ type: ɵngcc1.DeviceShellService }, { type: ɵngcc2.OperationRealtimeService }, { type: ɵngcc2.ManagedObjectRealtimeService }, { type: ɵngcc3.OperationService }, { type: ɵngcc4.ActivatedRoute }, { type: ɵngcc5.BsModalService }, { type: ɵngcc2.AlertService }]; }, null); })();
