// @c8y/ngx-components/esm2015/register-device/general/general-device-registration.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component } from '@angular/core';
import { TenantUiService, gettext, memoize } from '@c8y/ngx-components';
import { FormGroup } from '@angular/forms';
import { from, Subject, defer } from 'rxjs';
import { filter, shareReplay, switchMap, takeUntil, tap } from 'rxjs/operators';
import { InventoryService, TenantService } from '@c8y/client';
import { RegisterDeviceService } from '../register-device.service';
import { BsModalRef } from 'ngx-bootstrap/modal';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from '@c8y/client';
import * as ɵngcc3 from '../register-device.service';
import * as ɵngcc4 from 'ngx-bootstrap/modal';
import * as ɵngcc5 from '@angular/cdk/stepper';
import * as ɵngcc6 from '@ngx-formly/core';
import * as ɵngcc7 from '@angular/common';

function GeneralDeviceRegistrationComponent_c8y_operation_result_16_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-operation-result", 19);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("text", ɵngcc0.ɵɵpipeBind1(1, 3, "Device registered"));
    ɵngcc0.ɵɵproperty("size", 84)("vertical", true);
} }
function GeneralDeviceRegistrationComponent_c8y_operation_result_17_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-operation-result", 20);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("text", ɵngcc0.ɵɵpipeBind1(1, 3, "Failed to register device"));
    ɵngcc0.ɵɵproperty("size", 84)("vertical", true);
} }
const _c0 = function (a0) { return { successfulDevicesCount: a0 }; };
function GeneralDeviceRegistrationComponent_ng_container_18_c8y_operation_result_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-operation-result", 19);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("text", ɵngcc0.ɵɵpipeBind2(1, 3, "{{ successfulDevicesCount }} devices registered", ɵngcc0.ɵɵpureFunction1(6, _c0, ctx_r7.success.length)))("size", 84)("vertical", true);
} }
const _c1 = function (a0) { return { failedDevicesCount: a0 }; };
function GeneralDeviceRegistrationComponent_ng_container_18_c8y_operation_result_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-operation-result", 20);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    const ctx_r8 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("text", ɵngcc0.ɵɵpipeBind2(1, 3, "{{ failedDevicesCount }} devices failed to register", ɵngcc0.ɵɵpureFunction1(6, _c1, ctx_r8.failed.length)))("size", 84)("vertical", true);
} }
function GeneralDeviceRegistrationComponent_ng_container_18_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, GeneralDeviceRegistrationComponent_ng_container_18_c8y_operation_result_1_Template, 2, 8, "c8y-operation-result", 11);
    ɵngcc0.ɵɵtemplate(2, GeneralDeviceRegistrationComponent_ng_container_18_c8y_operation_result_2_Template, 2, 8, "c8y-operation-result", 12);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.failed.length === 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.success.length === 0);
} }
const _c2 = function (a0, a1) { return { count: a0, total: a1 }; };
function GeneralDeviceRegistrationComponent_div_19_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 21);
    ɵngcc0.ɵɵelement(1, "c8y-operation-result", 20);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementStart(3, "p", 22);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(4, " Registration failed for {{ count }} devices out of {{ total }}. ");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("text", ɵngcc0.ɵɵpipeBind1(2, 4, "Several devices failed to register"));
    ɵngcc0.ɵɵproperty("size", 84)("vertical", true);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction2(6, _c2, ctx_r3.failed.length, ctx_r3.failed.length + ctx_r3.success.length));
} }
function GeneralDeviceRegistrationComponent_div_20_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 23);
    ɵngcc0.ɵɵtext(1, " Turn on the registered device(s) and wait for connection(s) to be established. Once a device is connected, its status will change to \"Pending acceptance\". You will need to approve it by clicking on the \"Accept\" button. ");
    ɵngcc0.ɵɵelementEnd();
} }
function GeneralDeviceRegistrationComponent_c8y_li_22_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li");
    ɵngcc0.ɵɵelement(1, "c8y-li-icon", 24);
    ɵngcc0.ɵɵelementStart(2, "p");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "small");
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "c8y-li-collapse");
    ɵngcc0.ɵɵelementStart(8, "pre");
    ɵngcc0.ɵɵelementStart(9, "code");
    ɵngcc0.ɵɵtext(10);
    ɵngcc0.ɵɵpipe(11, "json");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const fail_r9 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("icon", "ban");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(fail_r9 == null ? null : fail_r9.id);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(6, 4, fail_r9 == null ? null : fail_r9.message));
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(11, 6, fail_r9 == null ? null : fail_r9.details));
} }
function GeneralDeviceRegistrationComponent_c8y_li_23_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li");
    ɵngcc0.ɵɵelement(1, "c8y-li-icon", 25);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const s_r10 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("icon", "check-circle");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", s_r10 == null ? null : s_r10.id, " ");
} }
const _c3 = function () { return { cancel: true, next: true }; };
const _c4 = function (a0) { return { back: a0, custom: true }; };
const _c5 = function () { return { back: "Fix errors", custom: "Close" }; };
export class GeneralDeviceRegistrationComponent {
    constructor(tenantUIService, tenantService, registerDeviceService, inventoryService, cd, bsModalRef) {
        this.tenantUIService = tenantUIService;
        this.tenantService = tenantService;
        this.registerDeviceService = registerDeviceService;
        this.inventoryService = inventoryService;
        this.cd = cd;
        this.bsModalRef = bsModalRef;
        this.MANAGEMENT = 'management';
        this.FILTER = {
            withTotalPages: true,
            pageSize: 25
        };
        this.form = new FormGroup({});
        this.model = {
            devicesToCreate: [{}]
        };
        this.options = {
            formState: {
                canLoadTenants: true,
            }
        };
        this.success = [];
        this.failed = [];
        this.fields = [
            {
                type: 'array',
                key: 'devicesToCreate',
                templateOptions: {
                    addText: gettext('Add device'),
                    addTextDataCy: 'add-device'
                },
                fieldArray: {
                    fieldGroup: [{
                            key: 'id',
                            type: 'string',
                            focus: true,
                            templateOptions: {
                                placeholder: '0123ab32fcd',
                                label: gettext('Device ID'),
                                required: true
                            },
                            validators: {
                                unique: {
                                    expression: (control) => {
                                        const found = control.root.get('devicesToCreate').value.filter(el => el.id === control.value);
                                        return found.length === 0;
                                    },
                                    message: () => gettext('Device ID duplicates are not allowed'),
                                },
                            },
                        },
                        {
                            key: 'tenant',
                            type: 'typeahead',
                            hideExpression: (model, formState, field) => {
                                if (!(formState === null || formState === void 0 ? void 0 : formState.canLoadTenants)) {
                                    field.formControl.setValue(null);
                                }
                                return !(formState === null || formState === void 0 ? void 0 : formState.canLoadTenants) || false;
                            },
                            defaultValue: { id: this.MANAGEMENT },
                            templateOptions: {
                                label: gettext('Add to tenant'),
                                required: true,
                                c8yForOptions: this.canLoadTenants$().pipe(filter(canLoad => canLoad), switchMap(() => this.getTenants$())),
                                container: 'body',
                                displayProperty: 'id',
                                valueProperties: ['id']
                            },
                            hooks: {
                                onInit: field => this.canLoadTenants$().pipe(tap(canLoad => {
                                    this.options.formState.canLoadTenants = canLoad;
                                    this.cd.detectChanges();
                                }))
                            }
                        },
                        {
                            key: 'group',
                            type: 'typeahead',
                            expressionProperties: {
                                'templateOptions.disabled': (model, formState, field) => {
                                    var _a, _b, _c;
                                    if (formState === null || formState === void 0 ? void 0 : formState.canLoadTenants) {
                                        if (((_a = model === null || model === void 0 ? void 0 : model.tenant) === null || _a === void 0 ? void 0 : _a.id) !== this.MANAGEMENT) {
                                            field.formControl.setValue(null);
                                        }
                                        return !(((_b = model === null || model === void 0 ? void 0 : model.tenant) === null || _b === void 0 ? void 0 : _b.id) === this.MANAGEMENT);
                                    }
                                    (_c = field === null || field === void 0 ? void 0 : field.templateOptions) === null || _c === void 0 ? true : delete _c.description;
                                    return false;
                                }
                            },
                            templateOptions: {
                                disabled: false,
                                label: gettext('Add to group'),
                                description: gettext('You can add device to specific group for management tenant only.'),
                                container: 'body',
                                displayProperty: 'name',
                                valueProperties: ['id'],
                                c8yForOptions: this.getGroups$()
                            },
                            hooks: {
                                onInit: field => this.canLoadTenants$().pipe(tap(canLoad => {
                                    this.options.formState.canLoadTenants = canLoad;
                                    this.cd.detectChanges();
                                }))
                            }
                        }]
                }
            }
        ];
        this.destroy$ = new Subject();
        this.lastCreatedDevices = [];
        this.isLoading$ = this.registerDeviceService.loading$;
    }
    ngAfterViewInit() {
        this.cd.detectChanges();
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
    create(eventObject) {
        var _a, _b;
        if (((_b = (_a = this.model) === null || _a === void 0 ? void 0 : _a.devicesToCreate) === null || _b === void 0 ? void 0 : _b.length) > 0) {
            this.lastCreatedDevices = [...this.model.devicesToCreate];
            const dataToSend = this.model.devicesToCreate.map((el) => {
                const { id, tenant, group } = el;
                let data = { id };
                if (tenant === null || tenant === void 0 ? void 0 : tenant.id) {
                    data = Object.assign(Object.assign({}, data), { tenantId: tenant.id });
                }
                if (group === null || group === void 0 ? void 0 : group.id) {
                    data = Object.assign(Object.assign({}, data), { groupId: group.id });
                }
                return data;
            });
            this.registerDeviceService.createMultiple(dataToSend)
                .pipe(takeUntil(this.destroy$))
                .subscribe(requests => {
                this.success = requests.success;
                this.failed = requests.failed;
                if (eventObject) {
                    eventObject.stepper.next();
                }
            });
        }
    }
    fixErrors(event, failedRequests) {
        if (failedRequests && failedRequests.length > 0) {
            this.options.resetModel({ devicesToCreate: [...this.lastCreatedDevices.filter(el => failedRequests.map(data => data.id).includes(el.id))] });
            this.cd.detectChanges();
        }
        event === null || event === void 0 ? void 0 : event.stepper.previous();
    }
    canLoadTenants$() {
        return defer(() => from(this.tenantUIService.isManagementTenant())).pipe(shareReplay(1));
    }
    getTenants$() {
        return defer(() => from(this.tenantService.list(this.FILTER))).pipe(shareReplay(1));
    }
    getGroups$() {
        return defer(() => from(this.inventoryService.listQuery({ __filter: { __has: 'c8y_IsDeviceGroup' }, __orderby: [{ name: 1 }] }, Object.assign({}, this.FILTER)))).pipe(shareReplay(1));
    }
}
GeneralDeviceRegistrationComponent.ɵfac = function GeneralDeviceRegistrationComponent_Factory(t) { return new (t || GeneralDeviceRegistrationComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.TenantUiService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.TenantService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.RegisterDeviceService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.InventoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.BsModalRef)); };
GeneralDeviceRegistrationComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: GeneralDeviceRegistrationComponent, selectors: [["c8y-general-device-registration"]], decls: 25, vars: 30, consts: [[3, "headerClasses", "customFooter"], ["c8y-modal-title", ""], [3, "c8yIcon"], ["linear", "", 3, "hideStepProgress"], [3, "stepControl"], [1, "p-b-16"], ["translate", "", 1, "p-l-24", "p-r-24", "p-t-16", "p-b-16", "m-b-0", "sticky-top", "separator-bottom", "text-medium", "text-16", "text-center", "bg-component"], [1, "formly-group-array-cols", "d-block", "p-24", "min-height-fit", 3, "form", "fields", "model", "options"], [1, "sticky-bottom", "d-block", "p-t-16", "p-b-16", "separator-top", "bg-white", 3, "showButtons", "disabled", "pending", "onNext", "onCancel"], ["state", "final"], [1, "p-24", "min-height-fit"], ["type", "success", "class", "lead", 3, "text", "size", "vertical", 4, "ngIf"], ["type", "error", "class", "lead", 3, "text", "size", "vertical", 4, "ngIf"], [4, "ngIf"], ["class", "p-l-24 p-r-24 text-center", 4, "ngIf"], ["class", "m-b-8 p-l-24 p-r-24", "translate", "", 4, "ngIf"], [1, "separator-top", "m-t-16"], [4, "ngFor", "ngForOf"], [1, "sticky-bottom", "d-block", "p-t-16", "p-b-16", "separator-top", "bg-white", 3, "showButtons", "labels", "onCustom", "onBack"], ["type", "success", 1, "lead", 3, "text", "size", "vertical"], ["type", "error", 1, "lead", 3, "text", "size", "vertical"], [1, "p-l-24", "p-r-24", "text-center"], ["translate", "", 1, "p-b-16", "text-danger", 3, "translateParams"], ["translate", "", 1, "m-b-8", "p-l-24", "p-r-24"], [1, "text-danger", 3, "icon"], [1, "text-success", 3, "icon"]], template: function GeneralDeviceRegistrationComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-modal", 0);
        ɵngcc0.ɵɵelementContainerStart(1, 1);
        ɵngcc0.ɵɵelement(2, "span", 2);
        ɵngcc0.ɵɵelementStart(3, "h4");
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementContainerEnd();
        ɵngcc0.ɵɵelementStart(6, "c8y-stepper", 3);
        ɵngcc0.ɵɵelementStart(7, "cdk-step", 4);
        ɵngcc0.ɵɵelementStart(8, "div", 5);
        ɵngcc0.ɵɵelementStart(9, "p", 6);
        ɵngcc0.ɵɵtext(10, " Register a general device ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(11, "formly-form", 7);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(12, "c8y-stepper-buttons", 8);
        ɵngcc0.ɵɵlistener("onNext", function GeneralDeviceRegistrationComponent_Template_c8y_stepper_buttons_onNext_12_listener($event) { return ctx.create($event); })("onCancel", function GeneralDeviceRegistrationComponent_Template_c8y_stepper_buttons_onCancel_12_listener() { return ctx.bsModalRef.hide(); });
        ɵngcc0.ɵɵpipe(13, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(14, "cdk-step", 9);
        ɵngcc0.ɵɵelementStart(15, "div", 10);
        ɵngcc0.ɵɵtemplate(16, GeneralDeviceRegistrationComponent_c8y_operation_result_16_Template, 2, 5, "c8y-operation-result", 11);
        ɵngcc0.ɵɵtemplate(17, GeneralDeviceRegistrationComponent_c8y_operation_result_17_Template, 2, 5, "c8y-operation-result", 12);
        ɵngcc0.ɵɵtemplate(18, GeneralDeviceRegistrationComponent_ng_container_18_Template, 3, 2, "ng-container", 13);
        ɵngcc0.ɵɵtemplate(19, GeneralDeviceRegistrationComponent_div_19_Template, 5, 9, "div", 14);
        ɵngcc0.ɵɵtemplate(20, GeneralDeviceRegistrationComponent_div_20_Template, 2, 0, "div", 15);
        ɵngcc0.ɵɵelementStart(21, "c8y-list-group", 16);
        ɵngcc0.ɵɵtemplate(22, GeneralDeviceRegistrationComponent_c8y_li_22_Template, 12, 8, "c8y-li", 17);
        ɵngcc0.ɵɵtemplate(23, GeneralDeviceRegistrationComponent_c8y_li_23_Template, 3, 2, "c8y-li", 17);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(24, "c8y-stepper-buttons", 18);
        ɵngcc0.ɵɵlistener("onCustom", function GeneralDeviceRegistrationComponent_Template_c8y_stepper_buttons_onCustom_24_listener() { return ctx.bsModalRef.hide(); })("onBack", function GeneralDeviceRegistrationComponent_Template_c8y_stepper_buttons_onBack_24_listener($event) { return ctx.fixErrors($event, ctx.failed); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("headerClasses", "dialog-header")("customFooter", true);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yIcon", "c8y-device-connect");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(5, 22, "Register devices"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("hideStepProgress", true);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("stepControl", ctx.form);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("form", ctx.form)("fields", ctx.fields)("model", ctx.model)("options", ctx.options);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("showButtons", ɵngcc0.ɵɵpureFunction0(26, _c3))("disabled", !(ctx.form == null ? null : ctx.form.valid))("pending", ɵngcc0.ɵɵpipeBind1(13, 24, ctx.isLoading$));
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngIf", ctx.success.length === 1 && ctx.failed.length === 0);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.success.length === 0 && ctx.failed.length === 1);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.success.length > 1 || ctx.failed.length > 1);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.success.length > 0 && ctx.failed.length > 0);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.success.length > 0);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.failed);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.success);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("showButtons", ɵngcc0.ɵɵpureFunction1(27, _c4, ctx.failed.length > 0))("labels", ɵngcc0.ɵɵpureFunction0(29, _c5));
    } }, directives: [ɵngcc1.ModalComponent, ɵngcc1.IconDirective, ɵngcc1.C8yStepper, ɵngcc5.CdkStep, ɵngcc1.C8yTranslateDirective, ɵngcc6.FormlyForm, ɵngcc1.C8yStepperButtons, ɵngcc7.NgIf, ɵngcc1.ListGroupComponent, ɵngcc7.NgForOf, ɵngcc1.OperationResultComponent, ɵngcc1.ListItemComponent, ɵngcc1.ListItemIconComponent, ɵngcc1.ListItemCollapseComponent], pipes: [ɵngcc1.C8yTranslatePipe, ɵngcc7.AsyncPipe, ɵngcc7.JsonPipe], encapsulation: 2, changeDetection: 0 });
GeneralDeviceRegistrationComponent.ctorParameters = () => [
    { type: TenantUiService },
    { type: TenantService },
    { type: RegisterDeviceService },
    { type: InventoryService },
    { type: ChangeDetectorRef },
    { type: BsModalRef }
];
__decorate([
    memoize()
], GeneralDeviceRegistrationComponent.prototype, "canLoadTenants$", null);
__decorate([
    memoize()
], GeneralDeviceRegistrationComponent.prototype, "getTenants$", null);
__decorate([
    memoize()
], GeneralDeviceRegistrationComponent.prototype, "getGroups$", null);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(GeneralDeviceRegistrationComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-general-device-registration',
                template: "<c8y-modal [headerClasses]=\"'dialog-header'\" [customFooter]=\"true\">\n  <ng-container c8y-modal-title>\n    <span [c8yIcon]=\"'c8y-device-connect'\"></span>\n    <h4>{{ 'Register devices' | translate }}</h4>\n  </ng-container>\n  <c8y-stepper [hideStepProgress]=\"true\" linear>\n    <cdk-step [stepControl]=\"form\">\n      <div class=\"p-b-16\">\n        <p\n          class=\"\n            p-l-24 p-r-24 p-t-16 p-b-16\n            m-b-0\n            sticky-top\n            separator-bottom\n            text-medium text-16 text-center\n            bg-component\n          \"\n          translate\n        >\n          Register a general device\n        </p>\n        <formly-form\n          [form]=\"form\"\n          [fields]=\"fields\"\n          [model]=\"model\"\n          [options]=\"options\"\n          class=\"formly-group-array-cols d-block p-24 min-height-fit\"\n        ></formly-form>\n      </div>\n      <c8y-stepper-buttons\n        (onNext)=\"create($event)\"\n        (onCancel)=\"bsModalRef.hide()\"\n        [showButtons]=\"{ cancel: true, next: true }\"\n        [disabled]=\"!form?.valid\"\n        [pending]=\"isLoading$ | async\"\n        class=\"sticky-bottom d-block p-t-16 p-b-16 separator-top bg-white\"\n      ></c8y-stepper-buttons>\n    </cdk-step>\n    <cdk-step state=\"final\">\n      <div class=\"p-24 min-height-fit\">\n        <c8y-operation-result\n          *ngIf=\"success.length === 1 && failed.length === 0\"\n          text=\"{{ 'Device registered' | translate }}\"\n          [size]=\"84\"\n          [vertical]=\"true\"\n          type=\"success\"\n          class=\"lead\"\n        ></c8y-operation-result>\n        <c8y-operation-result\n          *ngIf=\"success.length === 0 && failed.length === 1\"\n          text=\"{{ 'Failed to register device' | translate }}\"\n          [size]=\"84\"\n          [vertical]=\"true\"\n          type=\"error\"\n          class=\"lead\"\n        ></c8y-operation-result>\n\n        <ng-container *ngIf=\"success.length > 1 || failed.length > 1\">\n          <c8y-operation-result\n            *ngIf=\"failed.length === 0\"\n            [text]=\"\n              '{{ successfulDevicesCount }} devices registered'\n                | translate: { successfulDevicesCount: success.length }\n            \"\n            [size]=\"84\"\n            [vertical]=\"true\"\n            type=\"success\"\n            class=\"lead\"\n          ></c8y-operation-result>\n          <c8y-operation-result\n            *ngIf=\"success.length === 0\"\n            [text]=\"\n              '{{ failedDevicesCount }} devices failed to register'\n                | translate: { failedDevicesCount: failed.length }\n            \"\n            [size]=\"84\"\n            [vertical]=\"true\"\n            type=\"error\"\n            class=\"lead\"\n          ></c8y-operation-result>\n        </ng-container>\n\n        <div *ngIf=\"success.length > 0 && failed.length > 0\" class=\"p-l-24 p-r-24 text-center\">\n          <c8y-operation-result\n            text=\"{{ 'Several devices failed to register' | translate }}\"\n            [size]=\"84\"\n            [vertical]=\"true\"\n            type=\"error\"\n            class=\"lead\"\n          ></c8y-operation-result>\n          <p\n            ngNonBindable\n            translate\n            [translateParams]=\"{ count: failed.length, total: failed.length + success.length }\"\n            class=\"p-b-16 text-danger\"\n          >\n            Registration failed for {{ count }} devices out of {{ total }}.\n          </p>\n        </div>\n\n        <div class=\"m-b-8 p-l-24 p-r-24\" *ngIf=\"success.length > 0\" translate>\n          Turn on the registered device(s) and wait for connection(s) to be established. Once a\n          device is connected, its status will change to \"Pending acceptance\". You will need to\n          approve it by clicking on the \"Accept\" button.\n        </div>\n\n        <c8y-list-group class=\"separator-top m-t-16\">\n          <c8y-li *ngFor=\"let fail of failed\">\n            <c8y-li-icon class=\"text-danger\" [icon]=\"'ban'\"></c8y-li-icon>\n            <p>{{ fail?.id }}</p>\n            <small>{{ fail?.message | translate }}</small>\n            <c8y-li-collapse>\n              <pre><code>{{ fail?.details | json }}</code></pre>\n            </c8y-li-collapse>\n          </c8y-li>\n\n          <c8y-li *ngFor=\"let s of success\">\n            <c8y-li-icon class=\"text-success\" [icon]=\"'check-circle'\"></c8y-li-icon>\n            {{ s?.id }}\n          </c8y-li>\n        </c8y-list-group>\n      </div>\n      <c8y-stepper-buttons\n        class=\"sticky-bottom d-block p-t-16 p-b-16 separator-top bg-white\"\n        (onCustom)=\"bsModalRef.hide()\"\n        (onBack)=\"fixErrors($event, failed)\"\n        [showButtons]=\"{ back: failed.length > 0, custom: true }\"\n        [labels]=\"{ back: 'Fix errors', custom: 'Close' }\"\n      ></c8y-stepper-buttons>\n    </cdk-step>\n  </c8y-stepper>\n</c8y-modal>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return [{ type: ɵngcc1.TenantUiService }, { type: ɵngcc2.TenantService }, { type: ɵngcc3.RegisterDeviceService }, { type: ɵngcc2.InventoryService }, { type: ɵngcc0.ChangeDetectorRef }, { type: ɵngcc4.BsModalRef }]; }, null); })();
