// @c8y/ngx-components/esm2015/register-device/general/general-device-registration-button.component.js
import { Component } from '@angular/core';
import { BsModalService } from 'ngx-bootstrap/modal';
import { GeneralDeviceRegistrationComponent } from './general-device-registration.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/modal';
import * as ɵngcc2 from '@c8y/ngx-components';
export class GeneralDeviceRegistrationButtonComponent {
    constructor(modalService) {
        this.modalService = modalService;
    }
    open() {
        this.modalService.show(GeneralDeviceRegistrationComponent, {
            class: 'modal-sm',
            ignoreBackdropClick: true
        });
    }
}
GeneralDeviceRegistrationButtonComponent.ɵfac = function GeneralDeviceRegistrationButtonComponent_Factory(t) { return new (t || GeneralDeviceRegistrationButtonComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BsModalService)); };
GeneralDeviceRegistrationButtonComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: GeneralDeviceRegistrationButtonComponent, selectors: [["c8y-general-device-registration-button"]], decls: 4, vars: 3, consts: [[3, "click"], ["c8yIcon", "c8y-device-connect"]], template: function GeneralDeviceRegistrationButtonComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "button", 0);
        ɵngcc0.ɵɵlistener("click", function GeneralDeviceRegistrationButtonComponent_Template_button_click_0_listener() { return ctx.open(); });
        ɵngcc0.ɵɵelement(1, "i", 1);
        ɵngcc0.ɵɵtext(2);
        ɵngcc0.ɵɵpipe(3, "translate");
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 1, "General"), " ");
    } }, directives: [ɵngcc2.IconDirective], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
GeneralDeviceRegistrationButtonComponent.ctorParameters = () => [
    { type: BsModalService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(GeneralDeviceRegistrationButtonComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-general-device-registration-button',
                template: "<button (click)=\"open()\"><i c8yIcon=\"c8y-device-connect\"></i> {{ 'General' | translate }} </button>\n"
            }]
    }], function () { return [{ type: ɵngcc1.BsModalService }]; }, null); })();
