// @c8y/ngx-components/esm2015/core/common/empty-state/empty-state.component.js
import { Component, Input } from '@angular/core';
/**
 * A component to show generic empty state.
 * Title and subtitle should be used with translate pipe to mark strings for translation.
 * Component can be displayed vertically (default) or horizontally.
 * Additional markup elements can be placed inside the tag.
 *
 * Example:
 * ```html
 * <c8y-ui-empty-state
 *   [icon]="'c8y-icon'"
 *   [title]="'Place your title here' | translate"
 *   [subtitle]="'Place your subtitle here' | translate"
 *   [horizontal]="true"
 * >
 *   (...)
 * </c8y-ui-empty-state>
 * ```
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '../icon.directive';
import * as ɵngcc3 from '../../i18n/c8y-translate.pipe';

function EmptyStateComponent_ng_container_2_p_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(2, 1, ctx_r4.subtitle));
} }
function EmptyStateComponent_ng_container_2_ng_container_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function EmptyStateComponent_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "div");
    ɵngcc0.ɵɵelementStart(2, "h3", 4);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(5, EmptyStateComponent_ng_container_2_p_5_Template, 3, 3, "p", 2);
    ɵngcc0.ɵɵelementStart(6, "div", 5);
    ɵngcc0.ɵɵtemplate(7, EmptyStateComponent_ng_container_2_ng_container_7_Template, 1, 0, "ng-container", 6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    const _r2 = ɵngcc0.ɵɵreference(5);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(4, 3, ctx_r0.title));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.subtitle);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r2);
} }
function EmptyStateComponent_ng_container_3_p_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p", 9);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(2, 1, ctx_r6.subtitle));
} }
function EmptyStateComponent_ng_container_3_ng_container_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function EmptyStateComponent_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "div");
    ɵngcc0.ɵɵelementStart(2, "p");
    ɵngcc0.ɵɵelementStart(3, "strong");
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(6, EmptyStateComponent_ng_container_3_p_6_Template, 3, 3, "p", 7);
    ɵngcc0.ɵɵelementStart(7, "div", 8);
    ɵngcc0.ɵɵtemplate(8, EmptyStateComponent_ng_container_3_ng_container_8_Template, 1, 0, "ng-container", 6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    const _r2 = ɵngcc0.ɵɵreference(5);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(5, 3, ctx_r1.title));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.subtitle);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r2);
} }
function EmptyStateComponent_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵprojection(0);
} }
const _c0 = function (a0) { return { "c8y-empty-state--horizontal": a0 }; };
const _c1 = ["*"];
export class EmptyStateComponent {
}
EmptyStateComponent.ɵfac = function EmptyStateComponent_Factory(t) { return new (t || EmptyStateComponent)(); };
EmptyStateComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: EmptyStateComponent, selectors: [["c8y-ui-empty-state"]], inputs: { icon: "icon", title: "title", subtitle: "subtitle", horizontal: "horizontal" }, ngContentSelectors: _c1, decls: 6, vars: 6, consts: [[1, "c8y-empty-state", 3, "ngClass"], [1, "c8y-icon-duocolor", 3, "c8yIcon"], [4, "ngIf"], ["content", ""], [1, "text-medium", "p-b-4"], [1, "p-t-16"], [4, "ngTemplateOutlet"], ["class", "small", 4, "ngIf"], [1, "small", "m-t-8"], [1, "small"]], template: function EmptyStateComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelement(1, "i", 1);
        ɵngcc0.ɵɵtemplate(2, EmptyStateComponent_ng_container_2_Template, 8, 5, "ng-container", 2);
        ɵngcc0.ɵɵtemplate(3, EmptyStateComponent_ng_container_3_Template, 9, 5, "ng-container", 2);
        ɵngcc0.ɵɵtemplate(4, EmptyStateComponent_ng_template_4_Template, 1, 0, "ng-template", null, 3, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(4, _c0, ctx.horizontal));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("c8yIcon", ctx.icon);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.horizontal);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.horizontal);
    } }, directives: [ɵngcc1.NgClass, ɵngcc2.IconDirective, ɵngcc1.NgIf, ɵngcc1.NgTemplateOutlet], pipes: [ɵngcc3.C8yTranslatePipe], encapsulation: 2 });
EmptyStateComponent.propDecorators = {
    icon: [{ type: Input }],
    title: [{ type: Input }],
    subtitle: [{ type: Input }],
    horizontal: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(EmptyStateComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-ui-empty-state',
                template: "<div class=\"c8y-empty-state\" [ngClass]=\"{ 'c8y-empty-state--horizontal': horizontal }\">\n  <i [c8yIcon]=\"icon\" class=\"c8y-icon-duocolor\"></i>\n\n  <ng-container *ngIf=\"!horizontal\">\n    <div>\n      <h3 class=\"text-medium p-b-4\">{{ title | translate }}</h3>\n      <p *ngIf=\"subtitle\">{{ subtitle | translate }}</p>\n      <div class=\"p-t-16\">\n        <ng-container *ngTemplateOutlet=\"content\"></ng-container>\n      </div>\n    </div>\n  </ng-container>\n\n  <ng-container *ngIf=\"horizontal\">\n    <div>\n      <p>\n        <strong>{{ title | translate }}</strong>\n      </p>\n      <p *ngIf=\"subtitle\" class=\"small\">{{ subtitle | translate }}</p>\n      <div class=\"small m-t-8\">\n        <ng-container *ngTemplateOutlet=\"content\"></ng-container>\n      </div>\n    </div>\n  </ng-container>\n\n  <ng-template #content>\n    <ng-content></ng-content>\n  </ng-template>\n</div>\n"
            }]
    }], null, { icon: [{
            type: Input
        }], title: [{
            type: Input
        }], subtitle: [{
            type: Input
        }], horizontal: [{
            type: Input
        }] }); })();
