// @c8y/ngx-components/esm2015/core/common/bytes.pipe.js
import { Pipe } from '@angular/core';
import { isNaN, isFinite } from 'lodash-es';
/*
 * Converts the file size to a readable format.
 * Usage:
 *
 * ```html
 *  {{ myData | bytes }} <!-- e.g. 100.1 MB -->
 *
 * ```
 * Default precision is set to 1 decimal place.
 * To change the precision, pass an additional argument to pipe.

 * ```html
 * {{ myData | bytes: 5 }} <!-- e.g. 100.12345 MB -->
 * ```
 *
 */
import * as ɵngcc0 from '@angular/core';
export class BytesPipe {
    static bytes(bytes, precision) {
        if (isNaN(bytes) || !isFinite(bytes)) {
            return '-';
        }
        if (bytes === 0) {
            return 0;
        }
        const units = ['bytes', 'kB', 'MB', 'GB', 'TB', 'PB'];
        const num = Math.floor(Math.log(bytes) / Math.log(1000));
        // eslint-disable-next-line no-mixed-operators
        const result = (bytes / Math.pow(1000, Math.floor(num))).toFixed(precision);
        return `${result} ${units[num]}`;
    }
    transform(value, precision = 1) {
        return BytesPipe.bytes(value, precision);
    }
}
BytesPipe.ɵfac = function BytesPipe_Factory(t) { return new (t || BytesPipe)(); };
BytesPipe.ɵpipe = /*@__PURE__*/ ɵngcc0.ɵɵdefinePipe({ name: "bytes", type: BytesPipe, pure: true });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BytesPipe, [{
        type: Pipe,
        args: [{ name: 'bytes' }]
    }], null, null); })();
