// @c8y/ngx-components/esm2015/core/common/load-more.component.js
import { __awaiter } from "tslib";
import { ChangeDetectorRef, Component, ElementRef, EventEmitter, HostBinding, Input, Output } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '../i18n/c8y-translate.directive';
import * as ɵngcc3 from './icon.directive';
import * as ɵngcc4 from '../i18n/c8y-translate.pipe';

function LoadMoreComponent_button_0_ng_container_2_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 6);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("innerHTML", ɵngcc0.ɵɵpipeBind1(1, 1, ctx_r7.loadNextLabel), ɵngcc0.ɵɵsanitizeHtml);
} }
const _c0 = function (a0) { return { pageNo: a0 }; };
function LoadMoreComponent_button_0_ng_container_2_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 7);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(1, " Load page {{ pageNo }}");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r9 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction1(1, _c0, ctx_r9.paging.currentPage + 1));
} }
function LoadMoreComponent_button_0_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, LoadMoreComponent_button_0_ng_container_2_span_1_Template, 2, 3, "span", 4);
    ɵngcc0.ɵɵtemplate(2, LoadMoreComponent_button_0_ng_container_2_ng_template_2_Template, 2, 3, "ng-template", null, 5, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const _r8 = ɵngcc0.ɵɵreference(3);
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r5.loadNextLabel)("ngIfElse", _r8);
} }
function LoadMoreComponent_button_0_ng_container_3_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 6);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    const ctx_r10 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("innerHTML", ɵngcc0.ɵɵpipeBind1(1, 1, ctx_r10.loadingLabel), ɵngcc0.ɵɵsanitizeHtml);
} }
function LoadMoreComponent_button_0_ng_container_3_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 7);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(1, " Page {{ pageNo }} is loading\u2026 ");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r12 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction1(1, _c0, ctx_r12.paging.currentPage + 1));
} }
function LoadMoreComponent_button_0_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, LoadMoreComponent_button_0_ng_container_3_span_1_Template, 2, 3, "span", 4);
    ɵngcc0.ɵɵtemplate(2, LoadMoreComponent_button_0_ng_container_3_ng_template_2_Template, 2, 3, "ng-template", null, 8, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const _r11 = ɵngcc0.ɵɵreference(3);
    const ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r6.loadingLabel)("ngIfElse", _r11);
} }
const _c1 = function (a0) { return { "btn-pending": a0 }; };
function LoadMoreComponent_button_0_Template(rf, ctx) { if (rf & 1) {
    const _r14 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 3);
    ɵngcc0.ɵɵlistener("click", function LoadMoreComponent_button_0_Template_button_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r14); const ctx_r13 = ɵngcc0.ɵɵnextContext(); return ctx_r13.loadMore($event); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtemplate(2, LoadMoreComponent_button_0_ng_container_2_Template, 4, 2, "ng-container", 1);
    ɵngcc0.ɵɵtemplate(3, LoadMoreComponent_button_0_ng_container_3_Template, 4, 2, "ng-container", 1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵstyleProp("visibility", ctx_r0.hidden ? "hidden" : "visible")("height", ctx_r0.hidden ? "1px" : null);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 8, "Load more"));
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(10, _c1, ctx_r0.isLoading));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r0.isLoading);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.isLoading);
} }
function LoadMoreComponent_ng_container_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function LoadMoreComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, LoadMoreComponent_ng_container_1_ng_container_1_Template, 1, 0, "ng-container", 9);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    const _r2 = ɵngcc0.ɵɵreference(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r1.noMoreDataHint || _r2);
} }
function LoadMoreComponent_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 10);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 11);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, "Last record"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", "circle");
} }
function LoadMoreComponent_ng_container_4_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function LoadMoreComponent_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, LoadMoreComponent_ng_container_4_ng_container_1_Template, 1, 0, "ng-container", 9);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r4.loadingTemplate);
} }
export class LoadMoreComponent {
    constructor(element, cdRef) {
        this.element = element;
        this.cdRef = cdRef;
        this.useIntersection = true;
        this.hidden = false;
        this.class = 'c8y-list__item p-0';
        this.maxIterations = 10;
        this.hideNoMoreDataHint = false;
        this.onLoad = new EventEmitter();
        this.isLoading = false;
        this.counter = 0;
        this.hasNoMoreData = false;
        this.LOAD_SAME_PAGE_THRESHOLD = 50;
        this.destroyed = false;
    }
    get hostClass() {
        return this.hidden || (!this.hasMore && !this.hasNoMoreData) ? '' : this.class;
    }
    get hasMore() {
        return (this.paging && (this.paging.totalPages > this.paging.currentPage || !!this.paging.nextPage));
    }
    ngAfterContentInit() {
        this.destroyed = false;
        if (this.useIntersection && 'IntersectionObserver' in window) {
            this.intersectionObserver = new IntersectionObserver(event => this.buttonInView(event[0]), {
                root: this.container ? this.container.nativeElement : null
            });
            this.intersectionObserver.observe(this.element.nativeElement);
        }
        this.hasNoMoreData = this.shouldShowNoMoreDataHint();
    }
    ngOnDestroy() {
        this.destroyed = true;
        if (this.intersectionObserver) {
            this.intersectionObserver.disconnect();
            this.intersectionObserver.unobserve(this.element.nativeElement);
            clearTimeout(this.loadUntilIntersected);
        }
    }
    loadMore(event) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.destroyed) {
                this.isLoading = true;
                this.cdRef.detectChanges();
                if (event) {
                    event.stopPropagation();
                }
                if (this.hasMore) {
                    const result = yield this.paging.next();
                    this.paging = result.paging;
                    this.onLoad.emit(result.data);
                    this.intersectionLoading();
                    this.hasNoMoreData = this.shouldShowNoMoreDataHint();
                }
                else {
                    this.counter = 0;
                    this.isLoading = false;
                }
                this.cdRef.detectChanges();
            }
        });
    }
    intersectionLoading() {
        if (this.useIntersection && this.hasMore && this.loadUntilIntersected !== null) {
            this.loadUntilIntersected = setTimeout(() => this.loadMore(), this.getLoadingThreshold());
            this.useIntersection = this.shouldSwitchMode();
        }
        else {
            this.isLoading = false;
            this.loadUntilIntersected = undefined;
            this.cdRef.detectChanges();
        }
    }
    getLoadingThreshold() {
        return this.LOAD_SAME_PAGE_THRESHOLD * this.counter++;
    }
    shouldShowNoMoreDataHint() {
        return (this.counter !== 0 || this.noMoreDataHint) && !this.hasMore && !this.hidden;
    }
    shouldSwitchMode() {
        return this.counter < this.maxIterations || this.hidden;
    }
    buttonInView(event) {
        if (event.isIntersecting) {
            this.loadMore();
        }
        else if (this.loadUntilIntersected) {
            clearTimeout(this.loadUntilIntersected);
            this.loadUntilIntersected = null;
            this.isLoading = false;
            this.cdRef.detectChanges();
        }
        else {
            // avoiding a race condition when timeout is faster
            // cleared then set
            this.loadUntilIntersected = null;
        }
    }
}
LoadMoreComponent.ɵfac = function LoadMoreComponent_Factory(t) { return new (t || LoadMoreComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef)); };
LoadMoreComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: LoadMoreComponent, selectors: [["c8y-load-more"]], hostVars: 2, hostBindings: function LoadMoreComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassMap(ctx.hostClass);
    } }, inputs: { useIntersection: "useIntersection", hidden: "hidden", class: "class", maxIterations: "maxIterations", hideNoMoreDataHint: "hideNoMoreDataHint", paging: "paging", container: "container", noMoreDataHint: "noMoreDataHint", loadingTemplate: "loadingTemplate", loadNextLabel: "loadNextLabel", loadingLabel: "loadingLabel" }, outputs: { onLoad: "onLoad" }, decls: 5, vars: 3, consts: [["class", "btn btn-default btn-block text-center", 3, "ngClass", "visibility", "height", "title", "click", 4, "ngIf"], [4, "ngIf"], ["finishHint", ""], [1, "btn", "btn-default", "btn-block", "text-center", 3, "ngClass", "title", "click"], [3, "innerHTML", 4, "ngIf", "ngIfElse"], ["loadPage", ""], [3, "innerHTML"], ["translate", "", 3, "translateParams"], ["loading", ""], [4, "ngTemplateOutlet"], [1, "legend", "form-block", "center", "last-record", 3, "title"], [3, "c8yIcon"]], template: function LoadMoreComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, LoadMoreComponent_button_0_Template, 4, 12, "button", 0);
        ɵngcc0.ɵɵtemplate(1, LoadMoreComponent_ng_container_1_Template, 2, 1, "ng-container", 1);
        ɵngcc0.ɵɵtemplate(2, LoadMoreComponent_ng_template_2_Template, 3, 4, "ng-template", null, 2, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(4, LoadMoreComponent_ng_container_4_Template, 2, 1, "ng-container", 1);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.hasMore && !(ctx.loadingTemplate && ctx.isLoading));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.hasNoMoreData && !ctx.hideNoMoreDataHint && !ctx.isLoading);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.loadingTemplate && ctx.isLoading);
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgClass, ɵngcc2.C8yTranslateDirective, ɵngcc1.NgTemplateOutlet, ɵngcc3.IconDirective], pipes: [ɵngcc4.C8yTranslatePipe], encapsulation: 2 });
LoadMoreComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: ChangeDetectorRef }
];
LoadMoreComponent.propDecorators = {
    paging: [{ type: Input }],
    useIntersection: [{ type: Input }],
    hidden: [{ type: Input }],
    container: [{ type: Input }],
    class: [{ type: Input }],
    maxIterations: [{ type: Input }],
    noMoreDataHint: [{ type: Input }],
    loadingTemplate: [{ type: Input }],
    hideNoMoreDataHint: [{ type: Input }],
    loadNextLabel: [{ type: Input }],
    loadingLabel: [{ type: Input }],
    onLoad: [{ type: Output }],
    hostClass: [{ type: HostBinding, args: ['class',] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(LoadMoreComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-load-more',
                template: "<button\n  class=\"btn btn-default btn-block text-center\"\n  (click)=\"loadMore($event)\"\n  [ngClass]=\"{ 'btn-pending': isLoading }\"\n  *ngIf=\"hasMore && !(loadingTemplate && isLoading)\"\n  [style.visibility]=\"hidden ? 'hidden' : 'visible'\"\n  [style.height]=\"hidden ? '1px' : null\"\n  title=\"{{ 'Load more' | translate }}\"\n>\n  <ng-container *ngIf=\"!isLoading\">\n    <span *ngIf=\"loadNextLabel; else loadPage\" [innerHTML]=\"loadNextLabel | translate\"></span>\n    <ng-template #loadPage>\n      <span translate ngNonBindable [translateParams]=\"{ pageNo: paging.currentPage + 1 }\">\n        Load page {{ pageNo }}</span\n      >\n    </ng-template>\n  </ng-container>\n  <ng-container *ngIf=\"isLoading\">\n    <span *ngIf=\"loadingLabel; else loading\" [innerHTML]=\"loadingLabel | translate\"></span>\n    <ng-template #loading>\n      <span translate ngNonBindable [translateParams]=\"{ pageNo: paging.currentPage + 1 }\">\n        Page {{ pageNo }} is loading\u2026\n      </span>\n    </ng-template>\n  </ng-container>\n</button>\n\n<ng-container *ngIf=\"hasNoMoreData && !hideNoMoreDataHint && !isLoading\">\n  <ng-container *ngTemplateOutlet=\"noMoreDataHint || finishHint\"></ng-container>\n</ng-container>\n\n<ng-template #finishHint>\n  <div class=\"legend form-block center last-record\" title=\"{{ 'Last record' | translate }}\">\n    <i [c8yIcon]=\"'circle'\"></i>\n  </div>\n</ng-template>\n\n<ng-container *ngIf=\"loadingTemplate && isLoading\">\n  <ng-container *ngTemplateOutlet=\"loadingTemplate\"></ng-container>\n</ng-container>\n"
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.ChangeDetectorRef }]; }, { useIntersection: [{
            type: Input
        }], hidden: [{
            type: Input
        }], class: [{
            type: Input
        }], maxIterations: [{
            type: Input
        }], hideNoMoreDataHint: [{
            type: Input
        }], onLoad: [{
            type: Output
        }], hostClass: [{
            type: HostBinding,
            args: ['class']
        }], paging: [{
            type: Input
        }], container: [{
            type: Input
        }], noMoreDataHint: [{
            type: Input
        }], loadingTemplate: [{
            type: Input
        }], loadNextLabel: [{
            type: Input
        }], loadingLabel: [{
            type: Input
        }] }); })();
