// @c8y/ngx-components/esm2015/protocol-lpwan/lpwan-set-connections.component.js
import { __awaiter } from "tslib";
import { Component, Input, ViewChild } from '@angular/core';
import { LpwanSetDeviceProtocolService } from './lpwan-set-device-protocol.service';
import { AlertService, gettext, PopoverConfirmComponent } from '@c8y/ngx-components';
import { pipe } from 'rxjs';
import { map } from 'rxjs/operators';
import { find } from 'lodash-es';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './lpwan-set-device-protocol.service';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from '@angular/forms';

function LpwanAssignLnsConnectionsComponent_p_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p", 8);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Connection is not set. Select connection for this device"), " ");
} }
function LpwanAssignLnsConnectionsComponent_p_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p", 9);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("title", ctx_r1.currentConnection.name);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r1.currentConnection.name, " ");
} }
function LpwanAssignLnsConnectionsComponent_c8y_typeahead_8_c8y_li_2_Template(rf, ctx) { if (rf & 1) {
    const _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 12);
    ɵngcc0.ɵɵlistener("click", function LpwanAssignLnsConnectionsComponent_c8y_typeahead_8_c8y_li_2_Template_c8y_li_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r5 = ɵngcc0.ɵɵnextContext(2); return ctx_r5.togglePopover(ctx_r5.popoverConfirm); });
    ɵngcc0.ɵɵelement(1, "c8y-highlight", 13);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const connection_r4 = ctx.$implicit;
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("text", connection_r4.name)("pattern", ctx_r3.pattern);
} }
function LpwanAssignLnsConnectionsComponent_c8y_typeahead_8_Template(rf, ctx) { if (rf & 1) {
    const _r8 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-typeahead", 10);
    ɵngcc0.ɵɵlistener("ngModelChange", function LpwanAssignLnsConnectionsComponent_c8y_typeahead_8_Template_c8y_typeahead_ngModelChange_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r7 = ɵngcc0.ɵɵnextContext(); return ctx_r7.newConnection = $event; })("onSearch", function LpwanAssignLnsConnectionsComponent_c8y_typeahead_8_Template_c8y_typeahead_onSearch_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r9 = ɵngcc0.ɵɵnextContext(); return ctx_r9.setConnectionPipe($event); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtemplate(2, LpwanAssignLnsConnectionsComponent_c8y_typeahead_8_c8y_li_2_Template, 2, 2, "c8y-li", 11);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(1, 6, "Select new connection"));
    ɵngcc0.ɵɵproperty("ngModel", ctx_r2.newConnection)("allowFreeEntries", false);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yForOf", ctx_r2.availableConnections)("c8yForLoadMore", "hidden")("c8yForPipe", ctx_r2.filterConnection);
} }
export class LpwanAssignLnsConnectionsComponent {
    constructor(lpwanService, alertService) {
        this.lpwanService = lpwanService;
        this.alertService = alertService;
        this.buttons = [
            { label: gettext('Cancel'), action: () => Promise.reject() },
            { label: gettext('Re-register'), action: () => this.deviceReregister(), status: 'default' }
        ];
        this.reRegisterMsg = gettext('To change provider connection, you need to re-register device first.');
        this.filterConnection = pipe();
        this.pattern = '';
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.reload();
            this.setConnectionPipe('');
        });
    }
    setConnectionPipe(filterStr) {
        this.pattern = filterStr;
        this.filterConnection = pipe(map((connections) => connections.filter(connection => (!this.currentConnection || this.currentConnection.name !== connection.name) &&
            (!filterStr || connection.name.toLowerCase().indexOf(filterStr.toLowerCase()) > -1))));
    }
    reload() {
        return __awaiter(this, void 0, void 0, function* () {
            this.loading = true;
            this.newConnection = null;
            try {
                this.availableConnections = yield this.lpwanService.getAvailableConnections(this.device);
                this.currentConnection = find(this.availableConnections.data, {
                    name: this.device.c8y_LpwanDevice.lnsConnectionName
                });
            }
            catch (ex) {
                this.alertService.addServerFailure(ex);
            }
            finally {
                this.loading = false;
            }
        });
    }
    togglePopover(popoverCfrm) {
        popoverCfrm.show();
        this.setConnectionPipe('');
    }
    deviceReregister() {
        return new Promise((resolve, reject) => {
            window.location.href = `#/deviceregistration`;
            resolve(true);
        });
    }
}
LpwanAssignLnsConnectionsComponent.ɵfac = function LpwanAssignLnsConnectionsComponent_Factory(t) { return new (t || LpwanAssignLnsConnectionsComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.LpwanSetDeviceProtocolService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.AlertService)); };
LpwanAssignLnsConnectionsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: LpwanAssignLnsConnectionsComponent, selectors: [["set-lns-connections"]], viewQuery: function LpwanAssignLnsConnectionsComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(PopoverConfirmComponent, 7);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.popoverConfirm = _t.first);
    } }, inputs: { device: "device" }, decls: 12, vars: 9, consts: [[1, "col-md-6"], [1, "form-group"], ["translate", ""], ["class", "form-control-static", 4, "ngIf"], ["class", "form-control-static  text-truncate", 3, "title", 4, "ngIf"], ["outsideClick", "true", "containerClass", "lns-connection-popover", 3, "message", "buttons", "placement"], ["name", "newConnection", 3, "ngModel", "placeholder", "allowFreeEntries", "ngModelChange", "onSearch", 4, "ngIf"], ["name", "notExisting", 3, "text"], [1, "form-control-static"], [1, "form-control-static", "text-truncate", 3, "title"], ["name", "newConnection", 3, "ngModel", "placeholder", "allowFreeEntries", "ngModelChange", "onSearch"], ["class", "p-l-8 p-r-8 c8y-list__item--link", 3, "click", 4, "c8yFor", "c8yForOf", "c8yForLoadMore", "c8yForPipe"], [1, "p-l-8", "p-r-8", "c8y-list__item--link", 3, "click"], [3, "text", "pattern"]], template: function LpwanAssignLnsConnectionsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "label", 2);
        ɵngcc0.ɵɵtext(3, "Current connection");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(4, LpwanAssignLnsConnectionsComponent_p_4_Template, 3, 3, "p", 3);
        ɵngcc0.ɵɵtemplate(5, LpwanAssignLnsConnectionsComponent_p_5_Template, 2, 2, "p", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "c8y-form-group");
        ɵngcc0.ɵɵelement(7, "c8y-popover-confirm", 5);
        ɵngcc0.ɵɵtemplate(8, LpwanAssignLnsConnectionsComponent_c8y_typeahead_8_Template, 3, 8, "c8y-typeahead", 6);
        ɵngcc0.ɵɵelementStart(9, "c8y-messages");
        ɵngcc0.ɵɵelement(10, "c8y-message", 7);
        ɵngcc0.ɵɵpipe(11, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.currentConnection);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.currentConnection);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("message", ctx.reRegisterMsg)("buttons", ctx.buttons)("placement", "bottom");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.loading);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("text", ɵngcc0.ɵɵpipeBind1(11, 7, "Select one of the connections."));
    } }, directives: [ɵngcc2.C8yTranslateDirective, ɵngcc3.NgIf, ɵngcc2.FormGroupComponent, ɵngcc2.PopoverConfirmComponent, ɵngcc2.MessagesComponent, ɵngcc2.MessageDirective, ɵngcc2.TypeaheadComponent, ɵngcc4.NgControlStatus, ɵngcc4.NgModel, ɵngcc2.ForOfDirective, ɵngcc2.ListItemComponent, ɵngcc2.HighlightComponent], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
LpwanAssignLnsConnectionsComponent.ctorParameters = () => [
    { type: LpwanSetDeviceProtocolService },
    { type: AlertService }
];
LpwanAssignLnsConnectionsComponent.propDecorators = {
    device: [{ type: Input }],
    popoverConfirm: [{ type: ViewChild, args: [PopoverConfirmComponent, { static: true },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(LpwanAssignLnsConnectionsComponent, [{
        type: Component,
        args: [{
                selector: 'set-lns-connections',
                template: "<div class=\"col-md-6\">\n  <div class=\"form-group\">\n    <label translate>Current connection</label>\n    <p class=\"form-control-static\" *ngIf=\"!currentConnection\">\n      {{ 'Connection is not set. Select connection for this device' | translate }}\n    </p>\n    <p class=\"form-control-static  text-truncate\" *ngIf=\"currentConnection\" title=\"{{ currentConnection.name }}\">\n      {{ currentConnection.name }}\n    </p>\n  </div>\n\n  <c8y-form-group>\n    <c8y-popover-confirm\n      outsideClick=\"true\"\n      containerClass=\"lns-connection-popover\"\n      [message]=\"reRegisterMsg\"\n      [buttons]=\"buttons\"\n      [placement]=\"'bottom'\"\n    ></c8y-popover-confirm>\n    <c8y-typeahead\n      *ngIf=\"!loading\"\n      [(ngModel)]=\"newConnection\"\n      placeholder=\"{{ 'Select new connection' | translate }}\"\n      name=\"newConnection\"\n      (onSearch)=\"setConnectionPipe($event)\"\n      [allowFreeEntries]=\"false\"\n    >\n      <c8y-li\n        *c8yFor=\"let connection of availableConnections; loadMore: 'hidden'; pipe: filterConnection\"\n        class=\"p-l-8 p-r-8 c8y-list__item--link\"\n        (click)=\"togglePopover(popoverConfirm)\"\n      >\n        <c8y-highlight [text]=\"connection.name\" [pattern]=\"pattern\"></c8y-highlight>\n      </c8y-li>\n    </c8y-typeahead>\n    <c8y-messages\n      ><c8y-message\n        name=\"notExisting\"\n        [text]=\"'Select one of the connections.' | translate\"\n      ></c8y-message>\n    </c8y-messages>\n  </c8y-form-group>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.LpwanSetDeviceProtocolService }, { type: ɵngcc2.AlertService }]; }, { device: [{
            type: Input
        }], popoverConfirm: [{
            type: ViewChild,
            args: [PopoverConfirmComponent, { static: true }]
        }] }); })();
