// @c8y/ngx-components/esm2015/repository/software/c8y-ngx-components-repository-software.js
/**
 * Generated bundle index. Do not edit.
 */
export * from './index';
export { DeviceSoftwareChangesComponent as ɵf } from './device-tab/device-software-changes.component';
export { DeviceSoftwareListComponent as ɵe } from './device-tab/device-software-list.component';
export { DeviceSoftwareService as ɵc } from './device-tab/device-software.service';
export { InstalledSoftwareComponent as ɵd } from './device-tab/installed-software.component';
export { SoftwareDeviceTabComponent as ɵb } from './device-tab/software-device-tab.component';
export { SoftwareDeviceTabGuard as ɵg } from './device-tab/software-device-tab.guard';
export { SoftwareRepositoryDeviceTabModule as ɵa } from './device-tab/software-repository-device-tab.module';
export { AddSoftwareModalComponent as ɵk } from './list/add-software-modal.component';
export { DeviceTypeCellRendererComponent as ɵo } from './list/columns/device-type.cell-renderer.component';
export { NameCellRendererComponent as ɵl } from './list/columns/name.cell-renderer.component';
export { SoftwareTypeCellRendererComponent as ɵm } from './list/columns/software-type.cell-renderer.component';
export { SoftwareTypeFilteringFormRendererComponent as ɵn } from './list/columns/software-type.filtering-form-renderer.component';
export { VersionsCellRendererComponent as ɵp } from './list/columns/versions.cell-renderer.component';
export { SoftwareDetailsComponent as ɵj } from './list/software-details.component';
export { SoftwareListComponent as ɵi } from './list/software-list.component';
export { SoftwareRepositoryListModule as ɵh } from './list/software-repository-list.module';
export { SoftwareRepositoryNavigationFactory as ɵq } from './list/software-repository-navigation-factory';
