// @c8y/ngx-components/esm2015/ecosystem/applications/add-external-applicaiton/add-external-application.component.js
import { __awaiter } from "tslib";
import { Component } from '@angular/core';
import { FormBuilder, Validators } from '@angular/forms';
import { ApplicationService, ApplicationType } from '@c8y/client';
import { AlertService, WizardComponent } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/client';
import * as ɵngcc2 from '@angular/forms';
import * as ɵngcc3 from '@c8y/ngx-components';
import * as ɵngcc4 from '@angular/common';

function AddExternalApplicationComponent_form_0_Template(rf, ctx) { if (rf & 1) {
    const _r4 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "form", 2);
    ɵngcc0.ɵɵelementStart(1, "c8y-wizard-header");
    ɵngcc0.ɵɵelement(2, "h1", 3);
    ɵngcc0.ɵɵelementStart(3, "h4", 4);
    ɵngcc0.ɵɵtext(4, "Add external application");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "c8y-wizard-body");
    ɵngcc0.ɵɵelementStart(6, "p", 5);
    ɵngcc0.ɵɵtext(7);
    ɵngcc0.ɵɵpipe(8, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(9, "div", 6);
    ɵngcc0.ɵɵelementStart(10, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(11, "label", 7);
    ɵngcc0.ɵɵtext(12);
    ɵngcc0.ɵɵpipe(13, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(14, "input", 8);
    ɵngcc0.ɵɵpipe(15, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(16, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(17, "label", 7);
    ɵngcc0.ɵɵtext(18);
    ɵngcc0.ɵɵpipe(19, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(20, "input", 9);
    ɵngcc0.ɵɵpipe(21, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(22, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(23, "label", 7);
    ɵngcc0.ɵɵtext(24);
    ɵngcc0.ɵɵpipe(25, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(26, "input", 10);
    ɵngcc0.ɵɵpipe(27, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(28, "c8y-wizard-footer");
    ɵngcc0.ɵɵelementStart(29, "button", 11);
    ɵngcc0.ɵɵlistener("click", function AddExternalApplicationComponent_form_0_Template_button_click_29_listener() { ɵngcc0.ɵɵrestoreView(_r4); const ctx_r3 = ɵngcc0.ɵɵnextContext(); return ctx_r3.back(); });
    ɵngcc0.ɵɵpipe(30, "translate");
    ɵngcc0.ɵɵtext(31, " Back ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(32, "button", 12);
    ɵngcc0.ɵɵlistener("click", function AddExternalApplicationComponent_form_0_Template_button_click_32_listener() { ɵngcc0.ɵɵrestoreView(_r4); const ctx_r5 = ɵngcc0.ɵɵnextContext(); return ctx_r5.cancel(); });
    ɵngcc0.ɵɵpipe(33, "translate");
    ɵngcc0.ɵɵtext(34, " Cancel ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(35, "button", 13);
    ɵngcc0.ɵɵlistener("click", function AddExternalApplicationComponent_form_0_Template_button_click_35_listener() { ɵngcc0.ɵɵrestoreView(_r4); const ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.save(ctx_r6.formGroup.value); });
    ɵngcc0.ɵɵpipe(36, "translate");
    ɵngcc0.ɵɵtext(37);
    ɵngcc0.ɵɵpipe(38, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("formGroup", ctx_r0.formGroup);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", "cloud-link");
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(8, 15, "Provide external application details"), " ");
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(13, 17, "Name"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(15, 19, "e.g. My application"));
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(19, 21, "Application key"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(21, 23, "e.g. my-external-application-key"));
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(25, 25, "External URL"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(27, 27, "e.g."), " https://www.example.com");
    ɵngcc0.ɵɵproperty("pattern", "^(?!javascript:).+");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(30, 29, "Back"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(33, 31, "Cancel"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(36, 33, "Add application"));
    ɵngcc0.ɵɵproperty("disabled", ctx_r0.formGroup.invalid || ctx_r0.formGroup.pristine);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(38, 35, "Add application"), " ");
} }
function AddExternalApplicationComponent_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    const _r8 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-wizard-body");
    ɵngcc0.ɵɵelementStart(1, "div", 14);
    ɵngcc0.ɵɵelement(2, "c8y-operation-result", 15);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "c8y-wizard-footer");
    ɵngcc0.ɵɵelementStart(5, "button", 12);
    ɵngcc0.ɵɵlistener("click", function AddExternalApplicationComponent_ng_template_1_Template_button_click_5_listener() { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r7 = ɵngcc0.ɵɵnextContext(); return ctx_r7.done(); });
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵtext(7, " Done ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(8, "a", 16);
    ɵngcc0.ɵɵlistener("click", function AddExternalApplicationComponent_ng_template_1_Template_a_click_8_listener($event) { return $event.stopPropagation(); });
    ɵngcc0.ɵɵpipe(9, "translate");
    ɵngcc0.ɵɵelement(10, "i", 17);
    ɵngcc0.ɵɵtext(11);
    ɵngcc0.ɵɵpipe(12, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("text", ɵngcc0.ɵɵpipeBind1(3, 7, "Application created"));
    ɵngcc0.ɵɵproperty("size", 84)("vertical", true);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(6, 9, "Done"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(9, 11, "Open"));
    ɵngcc0.ɵɵproperty("href", ctx_r2.formGroup.value.externalUrl, ɵngcc0.ɵɵsanitizeUrl);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(12, 13, "Open"), " ");
} }
export class AddExternalApplicationComponent {
    constructor(applicationService, formBuilder, alertService, wizardComponent) {
        this.applicationService = applicationService;
        this.formBuilder = formBuilder;
        this.alertService = alertService;
        this.wizardComponent = wizardComponent;
    }
    ngOnInit() {
        this.initForm();
    }
    back() {
        this.wizardComponent.reset();
    }
    cancel() {
        this.wizardComponent.close();
    }
    done() {
        this.wizardComponent.close();
    }
    save(app) {
        return __awaiter(this, void 0, void 0, function* () {
            const appToSave = {
                manifest: {},
                externalUrl: app.externalUrl,
                type: ApplicationType.EXTERNAL,
                name: app.name,
                key: app.key
            };
            try {
                this.createdApp = yield this.applicationService.create(appToSave);
            }
            catch (ex) {
                this.alertService.addServerFailure(ex);
            }
        });
    }
    initForm() {
        this.formGroup = this.formBuilder.group({
            name: ['', [Validators.required, Validators.maxLength(120)]],
            key: ['', [Validators.required, Validators.maxLength(120)]],
            externalUrl: ['', Validators.required]
        });
    }
}
AddExternalApplicationComponent.ɵfac = function AddExternalApplicationComponent_Factory(t) { return new (t || AddExternalApplicationComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ApplicationService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.FormBuilder), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.WizardComponent)); };
AddExternalApplicationComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AddExternalApplicationComponent, selectors: [["c8y-add-external-application"]], decls: 3, vars: 2, consts: [["novalidate", "", 3, "formGroup", 4, "ngIf", "ngIfElse"], ["appCreated", ""], ["novalidate", "", 3, "formGroup"], [3, "c8yIcon"], ["translate", ""], [1, "p-16", "text-medium", "text-center", "bg-white", "separator-bottom", "sticky-top"], [1, "modal-body", "p-t-8"], [2, "text-align", "left"], ["formControlName", "name", "name", "name", "required", "", "type", "text", 1, "form-control", 3, "placeholder"], ["formControlName", "key", "name", "key", "required", "", "type", "text", 1, "form-control", 3, "placeholder"], ["formControlName", "externalUrl", "name", "externalUrl", "required", "", "type", "url", 1, "form-control", 3, "pattern", "placeholder"], ["translate", "", "type", "button", 1, "btn", "btn-default", 3, "title", "click"], ["translate", "", 1, "btn", "btn-default", 3, "title", "click"], ["type", "button", 1, "btn", "btn-primary", "btn-form", 3, "disabled", "title", "click"], [1, "d-flex", "d-col", "j-c-center", "a-i-center", 2, "min-height", "324px"], ["type", "success", 1, "lead", 3, "text", "size", "vertical"], ["target", "_blank", 1, "btn", "btn-primary", 3, "href", "title", "click"], ["c8yIcon", "external-link", 1, "m-r-4"]], template: function AddExternalApplicationComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, AddExternalApplicationComponent_form_0_Template, 39, 37, "form", 0);
        ɵngcc0.ɵɵtemplate(1, AddExternalApplicationComponent_ng_template_1_Template, 13, 15, "ng-template", null, 1, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        const _r1 = ɵngcc0.ɵɵreference(2);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.createdApp)("ngIfElse", _r1);
    } }, directives: [ɵngcc4.NgIf, ɵngcc2.ɵNgNoValidate, ɵngcc2.NgControlStatusGroup, ɵngcc2.FormGroupDirective, ɵngcc3.WizardHeaderComponent, ɵngcc3.IconDirective, ɵngcc3.C8yTranslateDirective, ɵngcc3.WizardBodyComponent, ɵngcc3.FormGroupComponent, ɵngcc2.DefaultValueAccessor, ɵngcc3.RequiredInputPlaceholderDirective, ɵngcc2.NgControlStatus, ɵngcc2.FormControlName, ɵngcc2.RequiredValidator, ɵngcc2.PatternValidator, ɵngcc3.WizardFooterComponent, ɵngcc3.OperationResultComponent], pipes: [ɵngcc3.C8yTranslatePipe], encapsulation: 2 });
AddExternalApplicationComponent.ctorParameters = () => [
    { type: ApplicationService },
    { type: FormBuilder },
    { type: AlertService },
    { type: WizardComponent }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AddExternalApplicationComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-add-external-application',
                template: "<form *ngIf=\"!createdApp; else appCreated\" [formGroup]=\"formGroup\" novalidate>\n  <c8y-wizard-header>\n    <h1 [c8yIcon]=\"'cloud-link'\"></h1>\n    <h4 translate>Add external application</h4>\n  </c8y-wizard-header>\n\n  <c8y-wizard-body>\n    <p class=\"p-16 text-medium text-center bg-white separator-bottom sticky-top\">\n      {{ 'Provide external application details' | translate }}\n    </p>\n    <div class=\"modal-body p-t-8\">\n      <c8y-form-group>\n        <label style=\"text-align: left\">{{ 'Name' | translate }}</label>\n        <input\n          class=\"form-control\"\n          formControlName=\"name\"\n          name=\"name\"\n          placeholder=\"{{ 'e.g. My application' | translate }}\"\n          required\n          type=\"text\"\n        />\n      </c8y-form-group>\n\n      <c8y-form-group>\n        <label style=\"text-align: left\">{{ 'Application key' | translate }}</label>\n        <input\n          class=\"form-control\"\n          formControlName=\"key\"\n          name=\"key\"\n          placeholder=\"{{ 'e.g. my-external-application-key' | translate }}\"\n          required\n          type=\"text\"\n        />\n      </c8y-form-group>\n\n      <c8y-form-group>\n        <label style=\"text-align: left\">{{ 'External URL' | translate }}</label>\n        <input\n          [pattern]=\"'^(?!javascript:).+'\"\n          class=\"form-control\"\n          formControlName=\"externalUrl\"\n          name=\"externalUrl\"\n          placeholder=\"{{ 'e.g.' | translate }} https://www.example.com\"\n          required\n          type=\"url\"\n        />\n      </c8y-form-group>\n    </div>\n  </c8y-wizard-body>\n\n  <c8y-wizard-footer>\n    <button\n      (click)=\"back()\"\n      class=\"btn btn-default\"\n      title=\"{{ 'Back' | translate }}\"\n      translate\n      type=\"button\"\n    >\n      Back\n    </button>\n    <button (click)=\"cancel()\" class=\"btn btn-default\" title=\"{{ 'Cancel' | translate }}\" translate>\n      Cancel\n    </button>\n    <button\n      (click)=\"save(formGroup.value)\"\n      [disabled]=\"formGroup.invalid || formGroup.pristine\"\n      class=\"btn btn-primary btn-form\"\n      title=\"{{ 'Add application' | translate }}\"\n      type=\"button\"\n    >\n      {{ 'Add application' | translate }}\n    </button>\n  </c8y-wizard-footer>\n</form>\n<ng-template #appCreated>\n  <c8y-wizard-body>\n    <div class=\"d-flex d-col j-c-center a-i-center\" style=\"min-height: 324px\">\n      <c8y-operation-result\n        text=\"{{ 'Application created' | translate }}\"\n        [size]=\"84\"\n        [vertical]=\"true\"\n        type=\"success\"\n        class=\"lead\"\n      ></c8y-operation-result>\n    </div>\n  </c8y-wizard-body>\n  <c8y-wizard-footer>\n    <button (click)=\"done()\" class=\"btn btn-default\" title=\"{{ 'Done' | translate }}\" translate>\n      Done\n    </button>\n    <a\n      (click)=\"$event.stopPropagation()\"\n      [href]=\"formGroup.value.externalUrl\"\n      target=\"_blank\"\n      class=\"btn btn-primary\"\n      title=\"{{ 'Open' | translate }}\"\n    >\n      <i c8yIcon=\"external-link\" class=\"m-r-4\"></i>\n      {{ 'Open' | translate }}\n    </a>\n  </c8y-wizard-footer>\n</ng-template>\n"
            }]
    }], function () { return [{ type: ɵngcc1.ApplicationService }, { type: ɵngcc2.FormBuilder }, { type: ɵngcc3.AlertService }, { type: ɵngcc3.WizardComponent }]; }, null); })();
