// @c8y/ngx-components/esm2015/datapoint-library/datapoint-library-list.component.js
import { __awaiter } from "tslib";
import { Component } from '@angular/core';
import { Router } from '@angular/router';
import { InventoryService } from '@c8y/client';
import { AlertService, gettext, ModalService } from '@c8y/ngx-components';
import { TranslateService } from '@ngx-translate/core';
import { BehaviorSubject, pipe } from 'rxjs';
import { map, shareReplay, switchMap, tap } from 'rxjs/operators';
import { pathToDatapointLibrary } from './datapoint-library.model';
import { DatapointParserService } from '@c8y/ngx-components/datapoint-library/services';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/client';
import * as ɵngcc2 from '@angular/router';
import * as ɵngcc3 from '@c8y/ngx-components';
import * as ɵngcc4 from '@ngx-translate/core';
import * as ɵngcc5 from '@c8y/ngx-components/datapoint-library/services';
import * as ɵngcc6 from '@angular/common';
import * as ɵngcc7 from 'ngx-bootstrap/dropdown';

function DatapointLibraryListComponent_c8y_action_bar_item_2_Template(rf, ctx) { if (rf & 1) {
    const _r5 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-action-bar-item", 1);
    ɵngcc0.ɵɵelementStart(1, "button", 10);
    ɵngcc0.ɵɵlistener("click", function DatapointLibraryListComponent_c8y_action_bar_item_2_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r5); const ctx_r4 = ɵngcc0.ɵɵnextContext(); return ctx_r4.addDatapointEntry(); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 11);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("placement", "right")("priority", 1);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(2, 4, "Add data point"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 6, "Add data point"), " ");
} }
function DatapointLibraryListComponent_div_13_Template(rf, ctx) { if (rf & 1) {
    const _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 12);
    ɵngcc0.ɵɵelement(1, "h1", 13);
    ɵngcc0.ɵɵelementStart(2, "h3", 14);
    ɵngcc0.ɵɵtext(3, "No data points to display.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "p", 14);
    ɵngcc0.ɵɵtext(5, "Add a data point using the button below.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "p");
    ɵngcc0.ɵɵelementStart(7, "button", 15);
    ɵngcc0.ɵɵlistener("click", function DatapointLibraryListComponent_div_13_Template_button_click_7_listener() { ɵngcc0.ɵɵrestoreView(_r7); const ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.addDatapointEntry(); });
    ɵngcc0.ɵɵpipe(8, "translate");
    ɵngcc0.ɵɵelement(9, "i", 11);
    ɵngcc0.ɵɵtext(10);
    ɵngcc0.ɵɵpipe(11, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(7);
    ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(8, 2, "Add data point"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(11, 4, "Add data point"), " ");
} }
function DatapointLibraryListComponent_div_15_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 16);
    ɵngcc0.ɵɵelement(1, "c8y-loading");
    ɵngcc0.ɵɵelementEnd();
} }
function DatapointLibraryListComponent_div_16_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 20);
    ɵngcc0.ɵɵelementStart(1, "div", 21);
    ɵngcc0.ɵɵelementStart(2, "div", 22);
    ɵngcc0.ɵɵelement(3, "div", 23);
    ɵngcc0.ɵɵelementStart(4, "p", 14);
    ɵngcc0.ɵɵtext(5, "Label");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(6, "div", 24);
    ɵngcc0.ɵɵelementStart(7, "div", 25);
    ɵngcc0.ɵɵelementStart(8, "div", 26);
    ɵngcc0.ɵɵelementStart(9, "p", 14);
    ɵngcc0.ɵɵtext(10, "Fragment");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(11, "div", 27);
    ɵngcc0.ɵɵelementStart(12, "p", 14);
    ɵngcc0.ɵɵtext(13, "Series");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(14, "div", 28);
    ɵngcc0.ɵɵelementStart(15, "span", 14);
    ɵngcc0.ɵɵtext(16, "Unit");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(17, "div", 27);
    ɵngcc0.ɵɵelementStart(18, "span", 14);
    ɵngcc0.ɵɵtext(19, "Target");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(20, "div", 29);
    ɵngcc0.ɵɵelementStart(21, "span", 14);
    ɵngcc0.ɵɵtext(22, "Range");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function DatapointLibraryListComponent_div_16_div_2_ul_7_Template(rf, ctx) { if (rf & 1) {
    const _r17 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "ul", 51);
    ɵngcc0.ɵɵelementStart(1, "li");
    ɵngcc0.ɵɵelementStart(2, "button", 52);
    ɵngcc0.ɵɵlistener("click", function DatapointLibraryListComponent_div_16_div_2_ul_7_Template_button_click_2_listener() { ɵngcc0.ɵɵrestoreView(_r17); const datapoint_r11 = ɵngcc0.ɵɵnextContext().$implicit; const ctx_r15 = ɵngcc0.ɵɵnextContext(2); return ctx_r15.editDatapointEntry(datapoint_r11); });
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelement(4, "i", 53);
    ɵngcc0.ɵɵelementStart(5, "span", 14);
    ɵngcc0.ɵɵtext(6, "Edit");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "li");
    ɵngcc0.ɵɵelementStart(8, "button", 52);
    ɵngcc0.ɵɵlistener("click", function DatapointLibraryListComponent_div_16_div_2_ul_7_Template_button_click_8_listener() { ɵngcc0.ɵɵrestoreView(_r17); const datapoint_r11 = ɵngcc0.ɵɵnextContext().$implicit; const ctx_r18 = ɵngcc0.ɵɵnextContext(2); return ctx_r18.removeDatapointEntry(datapoint_r11); });
    ɵngcc0.ɵɵpipe(9, "translate");
    ɵngcc0.ɵɵelement(10, "i", 54);
    ɵngcc0.ɵɵelementStart(11, "span", 14);
    ɵngcc0.ɵɵtext(12, "Remove");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(3, 2, "Edit"));
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(9, 4, "Remove"));
} }
function DatapointLibraryListComponent_div_16_div_2_span_29_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵelementStart(1, "label", 45);
    ɵngcc0.ɵɵtext(2, "Unit");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "span", 46);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const datapoint_r11 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("title", datapoint_r11.c8y_Kpi == null ? null : datapoint_r11.c8y_Kpi.unit);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(datapoint_r11.c8y_Kpi == null ? null : datapoint_r11.c8y_Kpi.unit);
} }
function DatapointLibraryListComponent_div_16_div_2_span_31_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵelementStart(1, "label", 45);
    ɵngcc0.ɵɵtext(2, "Target");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "span", 46);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const datapoint_r11 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("title", datapoint_r11.c8y_Kpi == null ? null : datapoint_r11.c8y_Kpi.target);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(datapoint_r11.c8y_Kpi == null ? null : datapoint_r11.c8y_Kpi.target);
} }
function DatapointLibraryListComponent_div_16_div_2_Template(rf, ctx) { if (rf & 1) {
    const _r23 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 30);
    ɵngcc0.ɵɵelementStart(1, "div", 31);
    ɵngcc0.ɵɵlistener("click", function DatapointLibraryListComponent_div_16_div_2_Template_div_click_1_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r23); const datapoint_r11 = restoredCtx.$implicit; const ctx_r22 = ɵngcc0.ɵɵnextContext(2); return ctx_r22.editDatapointEntry(datapoint_r11); });
    ɵngcc0.ɵɵelementStart(2, "div", 32);
    ɵngcc0.ɵɵlistener("click", function DatapointLibraryListComponent_div_16_div_2_Template_div_click_2_listener($event) { return $event.stopPropagation(); });
    ɵngcc0.ɵɵelementStart(3, "div", 33);
    ɵngcc0.ɵɵelementStart(4, "button", 34);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelement(6, "i", 35);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(7, DatapointLibraryListComponent_div_16_div_2_ul_7_Template, 13, 6, "ul", 36);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(8, "div", 37);
    ɵngcc0.ɵɵelementStart(9, "div", 23);
    ɵngcc0.ɵɵelement(10, "i", 38);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(11, "div", 39);
    ɵngcc0.ɵɵtext(12);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(13, "div", 40);
    ɵngcc0.ɵɵelementStart(14, "div", 41);
    ɵngcc0.ɵɵelement(15, "c8y-range-display", 42);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(16, "div", 43);
    ɵngcc0.ɵɵelementStart(17, "div", 44);
    ɵngcc0.ɵɵelementStart(18, "span");
    ɵngcc0.ɵɵelementStart(19, "label", 45);
    ɵngcc0.ɵɵtext(20, "Fragment");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(21, "span", 46);
    ɵngcc0.ɵɵtext(22);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(23, "div", 47);
    ɵngcc0.ɵɵelementStart(24, "label", 45);
    ɵngcc0.ɵɵtext(25, "Series");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(26, "span", 46);
    ɵngcc0.ɵɵtext(27);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(28, "div", 48);
    ɵngcc0.ɵɵtemplate(29, DatapointLibraryListComponent_div_16_div_2_span_29_Template, 5, 2, "span", 49);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(30, "div", 47);
    ɵngcc0.ɵɵtemplate(31, DatapointLibraryListComponent_div_16_div_2_span_31_Template, 5, 2, "span", 49);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(32, "div", 50);
    ɵngcc0.ɵɵelement(33, "c8y-range-display", 42);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const datapoint_r11 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(5, 15, "Actions"));
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵstyleProp("color", datapoint_r11.c8y_Kpi == null ? null : datapoint_r11.c8y_Kpi.color);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("title", datapoint_r11.c8y_Kpi == null ? null : datapoint_r11.c8y_Kpi.label);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", datapoint_r11.c8y_Kpi == null ? null : datapoint_r11.c8y_Kpi.label, " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("config", datapoint_r11.c8y_Kpi)("display", "compact");
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵproperty("title", datapoint_r11.c8y_Kpi == null ? null : datapoint_r11.c8y_Kpi.fragment);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", datapoint_r11.c8y_Kpi == null ? null : datapoint_r11.c8y_Kpi.fragment, " ");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("title", datapoint_r11.c8y_Kpi == null ? null : datapoint_r11.c8y_Kpi.series);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(datapoint_r11.c8y_Kpi == null ? null : datapoint_r11.c8y_Kpi.series);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", datapoint_r11.c8y_Kpi == null ? null : datapoint_r11.c8y_Kpi.unit);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", datapoint_r11.c8y_Kpi == null ? null : datapoint_r11.c8y_Kpi.target);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("config", datapoint_r11.c8y_Kpi)("display", "inline");
} }
function DatapointLibraryListComponent_div_16_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 17);
    ɵngcc0.ɵɵtemplate(1, DatapointLibraryListComponent_div_16_div_1_Template, 23, 0, "div", 18);
    ɵngcc0.ɵɵtemplate(2, DatapointLibraryListComponent_div_16_div_2_Template, 34, 17, "div", 19);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const datapoints_r8 = ctx.ngIf;
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ctx_r3.listClass);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", datapoints_r8.data.length);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yForOf", datapoints_r8)("c8yForLoadMore", "hidden")("c8yForPipe", ctx_r3.parsePipe);
} }
const _c0 = function () { return ["ROLE_INVENTORY_ADMIN"]; };
const _c1 = function (a0) { return { "icon-spin": a0 }; };
export class DatapointLibraryListComponent {
    constructor(inventory, router, modalService, translateService, alertService, parser) {
        this.inventory = inventory;
        this.router = router;
        this.modalService = modalService;
        this.translateService = translateService;
        this.alertService = alertService;
        this.parser = parser;
        this.listClass = '';
        this.isLoading = true;
        this.reload$ = new BehaviorSubject(null);
        this.datapoints$ = this.reload$.pipe(tap(() => {
            this.isLoading = true;
        }), switchMap(() => this.getDatapoints()), tap(() => {
            this.isLoading = false;
        }), shareReplay());
        this.parsePipe = pipe(map((data) => this.parser.parseDatapoints(data)));
    }
    getDatapoints() {
        return this.inventory.list({ pageSize: 20, fragmentType: 'c8y_Kpi', withTotalPages: true });
    }
    addDatapointEntry() {
        this.router.navigate([pathToDatapointLibrary, 'new']);
    }
    editDatapointEntry(entry) {
        this.router.navigate([pathToDatapointLibrary, entry.id]);
    }
    removeDatapointEntry(entry) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.modalService.confirm(gettext('Delete data point'), this.translateService.instant(gettext('You are about to delete data point "{{ dpLabel }}". Do you want to proceed?'), { dpLabel: entry.c8y_Kpi.label }), 'danger');
            }
            catch (e) {
                return;
            }
            try {
                yield this.inventory.delete(entry.id);
                this.reload$.next();
            }
            catch (e) {
                this.alertService.addServerFailure(e);
            }
        });
    }
    reload() {
        this.reload$.next();
    }
}
DatapointLibraryListComponent.ɵfac = function DatapointLibraryListComponent_Factory(t) { return new (t || DatapointLibraryListComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.InventoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.Router), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.ModalService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.TranslateService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.DatapointParserService)); };
DatapointLibraryListComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: DatapointLibraryListComponent, selectors: [["c8y-datapoint-library-list"]], decls: 18, vars: 25, consts: [[3, "placement", "priority", 4, "c8yIfAllowed"], [3, "placement", "priority"], [1, "btn", "btn-link", 3, "disabled", "title", "click"], ["c8yIcon", "refresh", 3, "ngClass"], ["itemClass", "navbar-form hidden-xs", 3, "placement"], [3, "listLength", "onListClassChange"], ["src", "/users-guide/cockpit/#data-point-library"], ["class", "c8y-empty-state text-center", 4, "ngIf"], ["class", "d-flex j-c-center", 4, "ngIf"], ["class", "card-group", 3, "ngClass", 4, "ngIf"], [1, "btn", "btn-link", 3, "title", "click"], ["c8yIcon", "plus-circle"], [1, "c8y-empty-state", "text-center"], [1, "c8y-icon", "c8y-icon-modules", "c8y-icon-duocolor"], ["translate", ""], [1, "btn", "btn-primary", 3, "title", "click"], [1, "d-flex", "j-c-center"], [1, "card-group", 3, "ngClass"], ["class", "page-sticky-header hidden-xs", 4, "ngIf"], ["class", "col-sm-6 col-md-4 col-lg-3 col-xs-12", 4, "c8yFor", "c8yForOf", "c8yForLoadMore", "c8yForPipe"], [1, "page-sticky-header", "hidden-xs"], [1, "flex-row"], [1, "card-header"], [1, "card-icon"], [1, "card-actions"], [1, "card-block", "p-l-0"], [1, "col-sm-3", "p-l-16"], [1, "col-sm-2"], [1, "col-sm-1"], [1, "col-sm-4"], [1, "col-sm-6", "col-md-4", "col-lg-3", "col-xs-12"], [1, "card", "pointer", 3, "click"], [1, "card-actions", 3, "click"], ["dropdown", "", 1, "dropdown", "settings", "pull-right"], ["type", "button", "dropdownToggle", "", 1, "dropdown-toggle", "c8y-dropdown", 3, "title"], ["c8yIcon", "ellipsis-v"], ["class", "dropdown-menu", 4, "dropdownMenu"], [1, "card-header", "separator"], ["c8yIcon", "circle"], [1, "card-title", "text-truncate", 3, "title"], [1, "card-block", "bg-gray-white", "p-b-24", "card-hidden-list"], [1, "col-sm-3", "col-md-2"], [3, "config", "display"], [1, "card-block"], [1, "col-sm-3", "text-truncate"], ["translate", "", 1, "text-label-small", "m-r-4"], [3, "title"], [1, "col-sm-2", "text-truncate"], [1, "col-sm-1", "text-truncate", "card-hidden-grid"], [4, "ngIf"], [1, "col-sm-4", "card-hidden-grid"], [1, "dropdown-menu"], ["type", "button", 3, "title", "click"], ["c8yIcon", "pencil"], ["c8yIcon", "trash"]], template: function DatapointLibraryListComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵtext(1, "Data point library");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(2, DatapointLibraryListComponent_c8y_action_bar_item_2_Template, 6, 8, "c8y-action-bar-item", 0);
        ɵngcc0.ɵɵelementStart(3, "c8y-action-bar-item", 1);
        ɵngcc0.ɵɵelementStart(4, "button", 2);
        ɵngcc0.ɵɵlistener("click", function DatapointLibraryListComponent_Template_button_click_4_listener() { return ctx.reload(); });
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelement(6, "i", 3);
        ɵngcc0.ɵɵtext(7);
        ɵngcc0.ɵɵpipe(8, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "c8y-action-bar-item", 4);
        ɵngcc0.ɵɵelementStart(10, "c8y-list-display-switch", 5);
        ɵngcc0.ɵɵlistener("onListClassChange", function DatapointLibraryListComponent_Template_c8y_list_display_switch_onListClassChange_10_listener($event) { return ctx.listClass = $event; });
        ɵngcc0.ɵɵpipe(11, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(12, "c8y-help", 6);
        ɵngcc0.ɵɵtemplate(13, DatapointLibraryListComponent_div_13_Template, 12, 6, "div", 7);
        ɵngcc0.ɵɵpipe(14, "async");
        ɵngcc0.ɵɵtemplate(15, DatapointLibraryListComponent_div_15_Template, 2, 0, "div", 8);
        ɵngcc0.ɵɵtemplate(16, DatapointLibraryListComponent_div_16_Template, 3, 5, "div", 9);
        ɵngcc0.ɵɵpipe(17, "async");
    } if (rf & 2) {
        let tmp_8_0;
        let tmp_9_0;
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yIfAllowed", ɵngcc0.ɵɵpureFunction0(22, _c0));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("placement", "right")("priority", 0);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("disabled", ctx.isLoading)("title", ɵngcc0.ɵɵpipeBind1(5, 12, "Reload"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(23, _c1, ctx.isLoading));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(8, 14, "Reload"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("placement", "left");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("listLength", (tmp_8_0 = ɵngcc0.ɵɵpipeBind1(11, 16, ctx.datapoints$)) == null ? null : tmp_8_0.data == null ? null : tmp_8_0.data.length);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ((tmp_9_0 = ɵngcc0.ɵɵpipeBind1(14, 18, ctx.datapoints$)) == null ? null : tmp_9_0.data == null ? null : tmp_9_0.data.length) === 0);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isLoading);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isLoading && ɵngcc0.ɵɵpipeBind1(17, 20, ctx.datapoints$));
    } }, directives: [ɵngcc3.TitleComponent, ɵngcc3.IfAllowedDirective, ɵngcc3.ActionBarItemComponent, ɵngcc3.IconDirective, ɵngcc6.NgClass, ɵngcc3.ɵbh, ɵngcc3.HelpComponent, ɵngcc6.NgIf, ɵngcc3.C8yTranslateDirective, ɵngcc3.LoadingComponent, ɵngcc3.ForOfDirective, ɵngcc7.BsDropdownDirective, ɵngcc7.BsDropdownToggleDirective, ɵngcc7.BsDropdownMenuDirective, ɵngcc3.RangeDisplayComponent], pipes: [ɵngcc3.C8yTranslatePipe, ɵngcc6.AsyncPipe], encapsulation: 2 });
DatapointLibraryListComponent.ctorParameters = () => [
    { type: InventoryService },
    { type: Router },
    { type: ModalService },
    { type: TranslateService },
    { type: AlertService },
    { type: DatapointParserService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DatapointLibraryListComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-datapoint-library-list',
                template: "<c8y-title>Data point library</c8y-title>\n\n<c8y-action-bar-item [placement]=\"'right'\" *c8yIfAllowed=\"['ROLE_INVENTORY_ADMIN']\" [priority]=\"1\">\n  <button (click)=\"addDatapointEntry()\" class=\"btn btn-link\" [title]=\"'Add data point' | translate\">\n    <i c8yIcon=\"plus-circle\"></i>\n    {{ 'Add data point' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\" [priority]=\"0\">\n  <button\n    (click)=\"reload()\"\n    class=\"btn btn-link\"\n    [disabled]=\"isLoading\"\n    [title]=\"'Reload' | translate\"\n  >\n    <i c8yIcon=\"refresh\" [ngClass]=\"{ 'icon-spin': isLoading }\"></i>\n    {{ 'Reload' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'left'\" itemClass=\"navbar-form hidden-xs\">\n  <c8y-list-display-switch\n    (onListClassChange)=\"listClass = $event\"\n    [listLength]=\"(datapoints$ | async)?.data?.length\"\n  ></c8y-list-display-switch>\n</c8y-action-bar-item>\n\n<c8y-help src=\"/users-guide/cockpit/#data-point-library\"></c8y-help>\n\n<!-- empty state -->\n<div *ngIf=\"(datapoints$ | async)?.data?.length === 0\" class=\"c8y-empty-state text-center\">\n  <h1 class=\"c8y-icon c8y-icon-modules c8y-icon-duocolor\"></h1>\n  <h3 translate>No data points to display.</h3>\n  <p translate>Add a data point using the button below.</p>\n  <p>\n    <button\n      (click)=\"addDatapointEntry()\"\n      class=\"btn btn-primary\"\n      [title]=\"'Add data point' | translate\"\n    >\n      <i c8yIcon=\"plus-circle\"></i>\n      {{ 'Add data point' | translate }}\n    </button>\n  </p>\n</div>\n\n<!-- loading indicator -->\n<div *ngIf=\"isLoading\" class=\"d-flex j-c-center\">\n  <c8y-loading></c8y-loading>\n</div>\n\n<div\n  class=\"card-group\"\n  [ngClass]=\"listClass\"\n  *ngIf=\"!isLoading && (datapoints$ | async) as datapoints\"\n>\n  <div class=\"page-sticky-header hidden-xs\" *ngIf=\"datapoints.data.length\">\n    <div class=\"flex-row\">\n      <div class=\"card-header\">\n        <div class=\"card-icon\"></div>\n        <p translate>Label</p>\n      </div>\n      <div class=\"card-actions\"></div>\n      <div class=\"card-block p-l-0\">\n        <div class=\"col-sm-3 p-l-16\">\n          <p translate>Fragment</p>\n        </div>\n        <div class=\"col-sm-2\">\n          <p translate>Series</p>\n        </div>\n        <div class=\"col-sm-1\">\n          <span translate>Unit</span>\n        </div>\n        <div class=\"col-sm-2\">\n          <span translate>Target</span>\n        </div>\n        <div class=\"col-sm-4\">\n          <span translate>Range</span>\n        </div>\n      </div>\n    </div>\n  </div>\n  <div *c8yFor=\"let datapoint of datapoints; loadMore: 'hidden'; pipe: parsePipe\" class=\"col-sm-6 col-md-4 col-lg-3 col-xs-12\">\n    <div class=\"card pointer\" (click)=\"editDatapointEntry(datapoint)\">\n      <div class=\"card-actions\" (click)=\"$event.stopPropagation()\">\n        <div class=\"dropdown settings pull-right\" dropdown>\n          <button\n            [title]=\"'Actions' | translate\"\n            type=\"button\"\n            class=\"dropdown-toggle c8y-dropdown\"\n            dropdownToggle\n          >\n            <i c8yIcon=\"ellipsis-v\"></i>\n          </button>\n          <ul class=\"dropdown-menu\" *dropdownMenu>\n            <li>\n              <button\n                type=\"button\"\n                [title]=\"'Edit' | translate\"\n                (click)=\"editDatapointEntry(datapoint)\"\n              >\n                <i c8yIcon=\"pencil\"></i>\n                <span translate>Edit</span>\n              </button>\n            </li>\n            <li>\n              <button\n                type=\"button\"\n                [title]=\"'Remove' | translate\"\n                (click)=\"removeDatapointEntry(datapoint)\"\n              >\n                <i c8yIcon=\"trash\"></i>\n                <span translate>Remove</span>\n              </button>\n            </li>\n          </ul>\n        </div>\n      </div>\n\n      <div class=\"card-header separator\">\n        <div class=\"card-icon\">\n          <i c8yIcon=\"circle\" [style.color]=\"datapoint.c8y_Kpi?.color\"></i>\n        </div>\n        <div class=\"card-title text-truncate\" [title]=\"datapoint.c8y_Kpi?.label\">\n          {{ datapoint.c8y_Kpi?.label }}\n        </div>\n      </div>\n      <div class=\"card-block bg-gray-white p-b-24 card-hidden-list\">\n        <div class=\"col-sm-3 col-md-2 \">\n          <c8y-range-display [config]=\"datapoint.c8y_Kpi\" [display]=\"'compact'\"></c8y-range-display>\n        </div>\n        </div>\n        <div class=\"card-block\">\n        <div class=\"col-sm-3 text-truncate\">\n          <span>\n            <label class=\"text-label-small m-r-4\" translate>Fragment</label>\n            <span [title]=\"datapoint.c8y_Kpi?.fragment\">\n              {{ datapoint.c8y_Kpi?.fragment }}\n            </span>\n          </span>\n        </div>\n        <div class=\"col-sm-2 text-truncate\">\n          <label class=\"text-label-small m-r-4\" translate>Series</label>\n          <span [title]=\"datapoint.c8y_Kpi?.series\">{{ datapoint.c8y_Kpi?.series }}</span>\n        </div>\n        <div class=\"col-sm-1 text-truncate card-hidden-grid\">\n          <span *ngIf=\"datapoint.c8y_Kpi?.unit\">\n            <label class=\"text-label-small m-r-4\" translate>Unit</label>\n            <span [title]=\"datapoint.c8y_Kpi?.unit\">{{ datapoint.c8y_Kpi?.unit }}</span>\n          </span>\n        </div>\n        <div class=\"col-sm-2 text-truncate\">\n          <span *ngIf=\"datapoint.c8y_Kpi?.target\">\n            <label class=\"text-label-small m-r-4\" translate>Target</label>\n            <span [title]=\"datapoint.c8y_Kpi?.target\">{{ datapoint.c8y_Kpi?.target }}</span>\n          </span>\n        </div>\n        <div class=\"col-sm-4 card-hidden-grid\">\n          <c8y-range-display [config]=\"datapoint.c8y_Kpi\" [display]=\"'inline'\"></c8y-range-display>\n        </div>\n      </div>\n    </div>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.InventoryService }, { type: ɵngcc2.Router }, { type: ɵngcc3.ModalService }, { type: ɵngcc4.TranslateService }, { type: ɵngcc3.AlertService }, { type: ɵngcc5.DatapointParserService }]; }, null); })();
