// @c8y/ngx-components/esm2015/datapoint-library/datapoint-library.module.js
import { NgModule } from '@angular/core';
import { CoreModule, gettext, HOOK_NAVIGATOR_NODES, HOOK_ROUTE, NavigatorNode, ViewContext } from '@c8y/ngx-components';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { DatapointLibraryDetailsComponent } from './datapoint-library-details/datapoint-library-details.component';
import { DatapointLibraryListComponent } from './datapoint-library-list.component';
import { DATAPOINT_LIBRARY_CONFIG, pathToDatapointLibrary } from './datapoint-library.model';
import * as ɵngcc0 from '@angular/core';
export class DatapointLibraryModule {
    static forRoot(config = {}) {
        return {
            ngModule: DatapointLibraryModule,
            providers: [
                {
                    provide: HOOK_ROUTE,
                    useValue: [
                        {
                            path: pathToDatapointLibrary,
                            component: DatapointLibraryListComponent
                        },
                        {
                            context: ViewContext.DatapointLibrary,
                            path: 'details',
                            component: DatapointLibraryDetailsComponent,
                            label: gettext('Details'),
                            icon: 'palette',
                            priority: 500
                        }
                    ],
                    multi: true
                },
                {
                    provide: HOOK_NAVIGATOR_NODES,
                    useValue: [
                        new NavigatorNode({
                            parent: gettext('Configuration'),
                            label: gettext('Data point library'),
                            featureId: 'dataPointLibrary',
                            path: pathToDatapointLibrary,
                            icon: 'c8y-data-points',
                            routerLinkExact: false
                        })
                    ],
                    multi: true
                },
                {
                    provide: DATAPOINT_LIBRARY_CONFIG,
                    useValue: config
                }
            ]
        };
    }
}
DatapointLibraryModule.ɵfac = function DatapointLibraryModule_Factory(t) { return new (t || DatapointLibraryModule)(); };
DatapointLibraryModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DatapointLibraryModule });
DatapointLibraryModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CoreModule, BsDropdownModule, PopoverModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DatapointLibraryModule, [{
        type: NgModule,
        args: [{
                imports: [CoreModule, BsDropdownModule, PopoverModule],
                declarations: [DatapointLibraryListComponent, DatapointLibraryDetailsComponent],
                exports: [DatapointLibraryListComponent, DatapointLibraryDetailsComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DatapointLibraryModule, { declarations: function () { return [DatapointLibraryListComponent, DatapointLibraryDetailsComponent]; }, imports: function () { return [CoreModule, BsDropdownModule, PopoverModule]; }, exports: function () { return [DatapointLibraryListComponent, DatapointLibraryDetailsComponent]; } }); })();
