// @c8y/ngx-components/esm2015/datapoint-library/datapoint-library-details/datapoint-library-details.component.js
import { __awaiter } from "tslib";
import { Component, Inject, Optional } from '@angular/core';
import { FormBuilder, Validators } from '@angular/forms';
import { ActivatedRoute, Router } from '@angular/router';
import { InventoryService } from '@c8y/client';
import { AlertService, gettext, HumanizePipe } from '@c8y/ngx-components';
import { TranslateService } from '@ngx-translate/core';
import { get } from 'lodash-es';
import { DATAPOINT_LIBRARY_CONFIG, pathToDatapointLibrary } from '../datapoint-library.model';
import { C8yValidators, ColorService } from '@c8y/ngx-components';
import { DatapointParserService } from '@c8y/ngx-components/datapoint-library/services';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from '@c8y/client';
import * as ɵngcc3 from '@angular/forms';
import * as ɵngcc4 from '@c8y/ngx-components';
import * as ɵngcc5 from '@ngx-translate/core';
import * as ɵngcc6 from '@c8y/ngx-components/datapoint-library/services';
import * as ɵngcc7 from '@angular/common';
import * as ɵngcc8 from 'ngx-bootstrap/popover';

function DatapointLibraryDetailsComponent_c8y_title_0_small_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate2("", ctx_r3.formGroup.value == null ? null : ctx_r3.formGroup.value.fragment, " / ", ctx_r3.formGroup.value == null ? null : ctx_r3.formGroup.value.series, "");
} }
function DatapointLibraryDetailsComponent_c8y_title_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-title");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵtemplate(2, DatapointLibraryDetailsComponent_c8y_title_0_small_2_Template, 2, 2, "small", 0);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r0.formGroup.value == null ? null : ctx_r0.formGroup.value.label, " ");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", (ctx_r0.formGroup.value == null ? null : ctx_r0.formGroup.value.fragment) && (ctx_r0.formGroup.value == null ? null : ctx_r0.formGroup.value.series));
} }
function DatapointLibraryDetailsComponent_div_8_div_37_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 21);
    ɵngcc0.ɵɵelementStart(1, "div", 22);
    ɵngcc0.ɵɵelementStart(2, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(3, "label", 38);
    ɵngcc0.ɵɵelement(4, "input", 39);
    ɵngcc0.ɵɵelement(5, "span");
    ɵngcc0.ɵɵelementStart(6, "span", 13);
    ɵngcc0.ɵɵtext(7, "Globally available");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(8, "button", 40);
    ɵngcc0.ɵɵpipe(9, "translate");
    ɵngcc0.ɵɵelement(10, "i", 41);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(8);
    ɵngcc0.ɵɵproperty("popover", ɵngcc0.ɵɵpipeBind1(9, 3, "Will make this entry available to all users on the tenant if checked."))("outsideClick", true);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", "question-circle-o");
} }
const _c0 = function () { return { example: "c8y_Temperature" }; };
const _c1 = function () { return { example: "T" }; };
const _c2 = function (a0) { return { "has-error": a0 }; };
const _c3 = function () { return { example: 0 }; };
const _c4 = function () { return { example: 100 }; };
const _c5 = function () { return { example: "\u00BAC" }; };
const _c6 = function () { return { example: 25 }; };
const _c7 = function () { return { example: 50 }; };
const _c8 = function () { return { example: 75 }; };
function DatapointLibraryDetailsComponent_div_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 9);
    ɵngcc0.ɵɵelementStart(1, "div", 10);
    ɵngcc0.ɵɵelementStart(2, "div", 11);
    ɵngcc0.ɵɵelementStart(3, "div", 12);
    ɵngcc0.ɵɵelementStart(4, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(5, "label", 13);
    ɵngcc0.ɵɵtext(6, "Color");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "div", 14);
    ɵngcc0.ɵɵelementStart(8, "div", 15);
    ɵngcc0.ɵɵelement(9, "input", 16);
    ɵngcc0.ɵɵelement(10, "span");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(11, "c8y-form-group", 17);
    ɵngcc0.ɵɵelementStart(12, "label", 13);
    ɵngcc0.ɵɵtext(13, "Label");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(14, "input", 18);
    ɵngcc0.ɵɵpipe(15, "translate");
    ɵngcc0.ɵɵelement(16, "c8y-messages", 19);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(17, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(18, "label", 13);
    ɵngcc0.ɵɵtext(19, "Description");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(20, "textarea", 20);
    ɵngcc0.ɵɵpipe(21, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(22, "div", 21);
    ɵngcc0.ɵɵelementStart(23, "div", 22);
    ɵngcc0.ɵɵelementStart(24, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(25, "label", 13);
    ɵngcc0.ɵɵtext(26, "Fragment");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(27, "input", 23);
    ɵngcc0.ɵɵpipe(28, "translate");
    ɵngcc0.ɵɵelement(29, "c8y-messages", 19);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(30, "div", 22);
    ɵngcc0.ɵɵelementStart(31, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(32, "label", 13);
    ɵngcc0.ɵɵtext(33, "Series");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(34, "input", 24);
    ɵngcc0.ɵɵpipe(35, "translate");
    ɵngcc0.ɵɵelement(36, "c8y-messages", 19);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(37, DatapointLibraryDetailsComponent_div_8_div_37_Template, 11, 5, "div", 25);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(38, "div", 26);
    ɵngcc0.ɵɵelementStart(39, "div", 11);
    ɵngcc0.ɵɵelementStart(40, "fieldset", 27);
    ɵngcc0.ɵɵelementStart(41, "legend", 13);
    ɵngcc0.ɵɵtext(42, "Preview");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(43, "c8y-range-display", 28);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(44, "fieldset", 27);
    ɵngcc0.ɵɵelementStart(45, "legend", 13);
    ɵngcc0.ɵɵtext(46, "Range");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(47, "div", 29);
    ɵngcc0.ɵɵelementStart(48, "div", 30);
    ɵngcc0.ɵɵelementStart(49, "c8y-form-group", 31);
    ɵngcc0.ɵɵelementStart(50, "label", 13);
    ɵngcc0.ɵɵtext(51, "Min");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(52, "input", 32);
    ɵngcc0.ɵɵpipe(53, "translate");
    ɵngcc0.ɵɵelement(54, "c8y-messages", 19);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(55, "div", 30);
    ɵngcc0.ɵɵelementStart(56, "c8y-form-group", 31);
    ɵngcc0.ɵɵelementStart(57, "label", 13);
    ɵngcc0.ɵɵtext(58, "Max");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(59, "input", 33);
    ɵngcc0.ɵɵpipe(60, "translate");
    ɵngcc0.ɵɵelement(61, "c8y-messages", 19);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(62, "div", 21);
    ɵngcc0.ɵɵelementStart(63, "div", 30);
    ɵngcc0.ɵɵelementStart(64, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(65, "label", 13);
    ɵngcc0.ɵɵtext(66, "Unit");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(67, "input", 34);
    ɵngcc0.ɵɵpipe(68, "translate");
    ɵngcc0.ɵɵelement(69, "c8y-messages", 19);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(70, "div", 30);
    ɵngcc0.ɵɵelementStart(71, "c8y-form-group", 31);
    ɵngcc0.ɵɵelementStart(72, "label", 13);
    ɵngcc0.ɵɵtext(73, "Target");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(74, "input", 35);
    ɵngcc0.ɵɵpipe(75, "translate");
    ɵngcc0.ɵɵelement(76, "c8y-messages", 19);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(77, "fieldset", 36);
    ɵngcc0.ɵɵelementStart(78, "legend", 13);
    ɵngcc0.ɵɵtext(79, "Yellow range");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(80, "div", 21);
    ɵngcc0.ɵɵelementStart(81, "div", 30);
    ɵngcc0.ɵɵelementStart(82, "c8y-form-group", 31);
    ɵngcc0.ɵɵelementStart(83, "label", 13);
    ɵngcc0.ɵɵtext(84, "Min");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(85, "input", 32);
    ɵngcc0.ɵɵpipe(86, "translate");
    ɵngcc0.ɵɵelement(87, "c8y-messages", 19);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(88, "div", 30);
    ɵngcc0.ɵɵelementStart(89, "c8y-form-group", 31);
    ɵngcc0.ɵɵelementStart(90, "label", 13);
    ɵngcc0.ɵɵtext(91, "Max");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(92, "input", 33);
    ɵngcc0.ɵɵpipe(93, "translate");
    ɵngcc0.ɵɵelement(94, "c8y-messages", 19);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(95, "fieldset", 37);
    ɵngcc0.ɵɵelementStart(96, "legend", 13);
    ɵngcc0.ɵɵtext(97, "Red range");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(98, "div", 21);
    ɵngcc0.ɵɵelementStart(99, "div", 30);
    ɵngcc0.ɵɵelementStart(100, "c8y-form-group", 31);
    ɵngcc0.ɵɵelementStart(101, "label", 13);
    ɵngcc0.ɵɵtext(102, "Min");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(103, "input", 32);
    ɵngcc0.ɵɵpipe(104, "translate");
    ɵngcc0.ɵɵelement(105, "c8y-messages", 19);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(106, "div", 30);
    ɵngcc0.ɵɵelementStart(107, "c8y-form-group", 31);
    ɵngcc0.ɵɵelementStart(108, "label", 13);
    ɵngcc0.ɵɵtext(109, "Max");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(110, "input", 33);
    ɵngcc0.ɵɵpipe(111, "translate");
    ɵngcc0.ɵɵelement(112, "c8y-messages", 19);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(10);
    ɵngcc0.ɵɵstyleProp("background-color", ctx_r1.formGroup.value == null ? null : ctx_r1.formGroup.value.color);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("placeholder", ɵngcc0.ɵɵpipeBind1(15, 34, "e.g. Temperature"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("show", (ctx_r1.formGroup.controls == null ? null : ctx_r1.formGroup.controls.label == null ? null : ctx_r1.formGroup.controls.label.touched) && (ctx_r1.formGroup.controls == null ? null : ctx_r1.formGroup.controls.label == null ? null : ctx_r1.formGroup.controls.label.errors));
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("placeholder", ɵngcc0.ɵɵpipeBind1(21, 36, "e.g. Ambient Temperature in Celsius"));
    ɵngcc0.ɵɵadvance(7);
    ɵngcc0.ɵɵproperty("placeholder", ɵngcc0.ɵɵpipeBind2(28, 38, "e.g. {{ example }}", ɵngcc0.ɵɵpureFunction0(68, _c0)));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("show", (ctx_r1.formGroup.controls == null ? null : ctx_r1.formGroup.controls.fragment == null ? null : ctx_r1.formGroup.controls.fragment.touched) && (ctx_r1.formGroup.controls == null ? null : ctx_r1.formGroup.controls.fragment == null ? null : ctx_r1.formGroup.controls.fragment.errors));
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵproperty("placeholder", ɵngcc0.ɵɵpipeBind2(35, 41, "e.g. {{ example }}", ɵngcc0.ɵɵpureFunction0(69, _c1)));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("show", (ctx_r1.formGroup.controls == null ? null : ctx_r1.formGroup.controls.series == null ? null : ctx_r1.formGroup.controls.series.touched) && (ctx_r1.formGroup.controls == null ? null : ctx_r1.formGroup.controls.series == null ? null : ctx_r1.formGroup.controls.series.errors));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.config == null ? null : ctx_r1.config.showCheckboxForGlobalFragment);
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵproperty("config", ctx_r1.rangeConfig);
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(70, _c2, (ctx_r1.range == null ? null : ctx_r1.range.touched) && (ctx_r1.range == null ? null : ctx_r1.range.controls == null ? null : ctx_r1.range.controls.min == null ? null : ctx_r1.range.controls.min.errors)));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("placeholder", ɵngcc0.ɵɵpipeBind2(53, 44, "e.g. {{ example }}", ɵngcc0.ɵɵpureFunction0(72, _c3)));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("show", (ctx_r1.range == null ? null : ctx_r1.range.touched) && (ctx_r1.range.controls == null ? null : ctx_r1.range.controls.min == null ? null : ctx_r1.range.controls.min.errors));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(73, _c2, (ctx_r1.range == null ? null : ctx_r1.range.touched) && (ctx_r1.range == null ? null : ctx_r1.range.controls == null ? null : ctx_r1.range.controls.max == null ? null : ctx_r1.range.controls.max.errors)));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("placeholder", ɵngcc0.ɵɵpipeBind2(60, 47, "e.g. {{ example }}", ɵngcc0.ɵɵpureFunction0(75, _c4)));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("show", (ctx_r1.range == null ? null : ctx_r1.range.touched) && (ctx_r1.range.controls == null ? null : ctx_r1.range.controls.max == null ? null : ctx_r1.range.controls.max.errors));
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵproperty("placeholder", ɵngcc0.ɵɵpipeBind2(68, 50, "e.g. {{ example }}", ɵngcc0.ɵɵpureFunction0(76, _c5)));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("show", (ctx_r1.formGroup.controls == null ? null : ctx_r1.formGroup.controls.unit == null ? null : ctx_r1.formGroup.controls.unit.touched) && (ctx_r1.formGroup.controls == null ? null : ctx_r1.formGroup.controls.unit == null ? null : ctx_r1.formGroup.controls.unit.errors));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(77, _c2, ((ctx_r1.range == null ? null : ctx_r1.range.touched) || (ctx_r1.formGroup.controls == null ? null : ctx_r1.formGroup.controls.target == null ? null : ctx_r1.formGroup.controls.target.touched)) && (ctx_r1.formGroup.controls == null ? null : ctx_r1.formGroup.controls.target == null ? null : ctx_r1.formGroup.controls.target.errors)));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("placeholder", ɵngcc0.ɵɵpipeBind2(75, 53, "e.g. {{ example }}", ɵngcc0.ɵɵpureFunction0(79, _c6)));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("show", ((ctx_r1.range == null ? null : ctx_r1.range.touched) || (ctx_r1.formGroup.controls == null ? null : ctx_r1.formGroup.controls.target == null ? null : ctx_r1.formGroup.controls.target.touched)) && (ctx_r1.formGroup.controls == null ? null : ctx_r1.formGroup.controls.target == null ? null : ctx_r1.formGroup.controls.target.errors));
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(80, _c2, ((ctx_r1.range == null ? null : ctx_r1.range.touched) || (ctx_r1.yellowRange == null ? null : ctx_r1.yellowRange.touched)) && (ctx_r1.yellowRange == null ? null : ctx_r1.yellowRange.controls == null ? null : ctx_r1.yellowRange.controls.min == null ? null : ctx_r1.yellowRange.controls.min.errors)));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("placeholder", ɵngcc0.ɵɵpipeBind2(86, 56, "e.g. {{ example }}", ɵngcc0.ɵɵpureFunction0(82, _c7)));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("show", ((ctx_r1.range == null ? null : ctx_r1.range.touched) || (ctx_r1.yellowRange == null ? null : ctx_r1.yellowRange.touched)) && (ctx_r1.yellowRange.controls == null ? null : ctx_r1.yellowRange.controls.min == null ? null : ctx_r1.yellowRange.controls.min.errors));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(83, _c2, ((ctx_r1.range == null ? null : ctx_r1.range.touched) || (ctx_r1.yellowRange == null ? null : ctx_r1.yellowRange.touched)) && (ctx_r1.yellowRange == null ? null : ctx_r1.yellowRange.controls == null ? null : ctx_r1.yellowRange.controls.max == null ? null : ctx_r1.yellowRange.controls.max.errors)));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("placeholder", ɵngcc0.ɵɵpipeBind2(93, 59, "e.g. {{ example }}", ɵngcc0.ɵɵpureFunction0(85, _c8)));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("show", ((ctx_r1.range == null ? null : ctx_r1.range.touched) || (ctx_r1.yellowRange == null ? null : ctx_r1.yellowRange.touched)) && (ctx_r1.yellowRange.controls == null ? null : ctx_r1.yellowRange.controls.max == null ? null : ctx_r1.yellowRange.controls.max.errors));
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(86, _c2, ((ctx_r1.range == null ? null : ctx_r1.range.touched) || (ctx_r1.redRange == null ? null : ctx_r1.redRange.touched)) && (ctx_r1.redRange == null ? null : ctx_r1.redRange.controls == null ? null : ctx_r1.redRange.controls.min == null ? null : ctx_r1.redRange.controls.min.errors)));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("placeholder", ɵngcc0.ɵɵpipeBind2(104, 62, "e.g. {{ example }}", ɵngcc0.ɵɵpureFunction0(88, _c8)));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("show", ((ctx_r1.range == null ? null : ctx_r1.range.touched) || (ctx_r1.redRange == null ? null : ctx_r1.redRange.touched)) && (ctx_r1.redRange.controls == null ? null : ctx_r1.redRange.controls.min == null ? null : ctx_r1.redRange.controls.min.errors));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(89, _c2, ((ctx_r1.range == null ? null : ctx_r1.range.touched) || (ctx_r1.redRange == null ? null : ctx_r1.redRange.touched)) && (ctx_r1.redRange == null ? null : ctx_r1.redRange.controls == null ? null : ctx_r1.redRange.controls.max == null ? null : ctx_r1.redRange.controls.max.errors)));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("placeholder", ɵngcc0.ɵɵpipeBind2(111, 65, "e.g. {{ example }}", ɵngcc0.ɵɵpureFunction0(91, _c4)));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("show", ((ctx_r1.range == null ? null : ctx_r1.range.touched) || (ctx_r1.redRange == null ? null : ctx_r1.redRange.touched)) && (ctx_r1.redRange.controls == null ? null : ctx_r1.redRange.controls.max == null ? null : ctx_r1.redRange.controls.max.errors));
} }
function DatapointLibraryDetailsComponent_button_14_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "button", 42);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("disabled", ctx_r2.formGroup.invalid || ctx_r2.formGroup.pristine)("title", ɵngcc0.ɵɵpipeBind1(1, 3, "Save"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 5, "Save"), " ");
} }
const _c9 = function () { return ["ROLE_INVENTORY_ADMIN", "ROLE_INVENTORY_UPDATE"]; };
export class DatapointLibraryDetailsComponent {
    constructor(activatedRoute, inventory, formBuilder, alertService, router, translate, colorService, parser, config) {
        this.activatedRoute = activatedRoute;
        this.inventory = inventory;
        this.formBuilder = formBuilder;
        this.alertService = alertService;
        this.router = router;
        this.translate = translate;
        this.colorService = colorService;
        this.parser = parser;
        this.config = config;
        this.path = pathToDatapointLibrary;
        this.isLoading = true;
        this.rangeConfig = {};
        this.initForm();
    }
    ngOnInit() {
        this.routeSub = this.activatedRoute.parent.data.subscribe(data => {
            this.load(data);
        });
    }
    ngOnDestroy() {
        if (this.routeSub) {
            this.routeSub.unsubscribe();
        }
    }
    save(value) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            const kpiValues = this.extractKpiValuesFromFormValue(value);
            const kpi = Object.assign({}, ((_a = this.datapoint) === null || _a === void 0 ? void 0 : _a.c8y_Kpi) || {}, kpiValues);
            const cleanedUpKpi = this.removeNullValues(kpi);
            const mo = {
                c8y_Kpi: cleanedUpKpi,
                c8y_Global: value.c8y_Global ? {} : null
            };
            try {
                if ((_b = this.datapoint) === null || _b === void 0 ? void 0 : _b.id) {
                    mo.id = this.datapoint.id;
                    const res = yield this.inventory.update(mo);
                    this.datapointUpdate(res.data);
                    this.alertService.saveSuccess(gettext('Data point library entry'))();
                }
                else {
                    const res = yield this.inventory.create(mo);
                    this.datapointUpdate(res.data);
                    this.alertService.createSuccess(gettext('Data point library entry'))();
                }
                this.navigateBackToList();
            }
            catch (e) {
                this.alertService.addServerFailure(e);
            }
        });
    }
    cancel() {
        this.navigateBackToList();
    }
    formChange() {
        if (this.formGroup.invalid) {
            return;
        }
        const kpiValues = this.extractKpiValuesFromFormValue(this.formGroup.value);
        this.rangeConfig = Object.assign({ orientation: 'horizontal' }, kpiValues);
    }
    navigateBackToList() {
        return this.router.navigate([pathToDatapointLibrary]);
    }
    datapointUpdate(tmpDatapoint) {
        const datapoint = this.parser.parseDatapoint(tmpDatapoint);
        this.datapoint = datapoint;
        const { fragment, series, description, unit, target, label, color, yellowRangeMin, yellowRangeMax, redRangeMin, redRangeMax, min, max } = datapoint.c8y_Kpi;
        this.formGroup.patchValue({
            fragment,
            series,
            description,
            unit,
            target,
            label,
            color,
            range: { min, max },
            yellowRange: { min: yellowRangeMin, max: yellowRangeMax },
            redRange: { min: redRangeMin, max: redRangeMax },
            c8y_Global: !!datapoint.c8y_Global
        });
        this.formChange();
    }
    extractKpiValuesFromFormValue(formValue) {
        const { fragment, series, description, unit, target, label, color } = formValue;
        return {
            fragment,
            series,
            description,
            unit,
            target,
            label,
            color,
            yellowRangeMin: get(formValue, 'yellowRange.min'),
            yellowRangeMax: get(formValue, 'yellowRange.max'),
            redRangeMin: get(formValue, 'redRange.min'),
            redRangeMax: get(formValue, 'redRange.max'),
            min: get(formValue, 'range.min'),
            max: get(formValue, 'range.max')
        };
    }
    load(data) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            this.isLoading = true;
            this.formGroup.reset();
            const { fragment, series } = this.activatedRoute.snapshot.queryParams;
            const humanize = new HumanizePipe();
            const label = fragment && series
                ? humanize.transform(`${fragment} => ${series}`)
                : this.translate.instant(gettext('New data point template'));
            const currentDate = new Date().toISOString();
            const color = yield this.colorService.generateColor(currentDate);
            const contextData = (data && data.contextData) || {
                c8y_Kpi: {
                    label,
                    fragment,
                    series,
                    color
                },
                c8y_Global: !((_a = this.config) === null || _a === void 0 ? void 0 : _a.doNotAddGlobalFragmentByDefault) ? {} : null
            };
            const { id } = contextData;
            if (id) {
                this.datapointUpdate((yield this.inventory.detail(id)).data);
            }
            else {
                this.datapointUpdate(contextData);
            }
            this.isLoading = false;
        });
    }
    initForm() {
        var _a;
        this.range = this.formBuilder.group({
            min: [undefined, []],
            max: [undefined, []]
        }, { validators: [C8yValidators.minMaxValidator(), C8yValidators.requireBothMinAndMax()] });
        this.redRange = this.formBuilder.group({
            min: [undefined, []],
            max: [undefined, []]
        }, { validators: [C8yValidators.minMaxValidator(), C8yValidators.requireBothMinAndMax()] });
        this.yellowRange = this.formBuilder.group({
            min: [undefined, []],
            max: [undefined, []]
        }, { validators: [C8yValidators.minMaxValidator(), C8yValidators.requireBothMinAndMax()] });
        this.formGroup = this.formBuilder.group({
            color: ['', [Validators.required, Validators.minLength(4)]],
            label: ['', [Validators.required, Validators.minLength(1), Validators.maxLength(120)]],
            description: ['', []],
            fragment: [
                '',
                [
                    Validators.required,
                    Validators.minLength(1),
                    Validators.maxLength(120),
                    Validators.pattern(/^[^.]*$/)
                ]
            ],
            series: [
                '',
                [
                    Validators.required,
                    Validators.minLength(1),
                    Validators.maxLength(120),
                    Validators.pattern(/^[^.]*$/)
                ]
            ],
            range: this.range,
            unit: [undefined, []],
            target: [undefined, []],
            redRange: this.redRange,
            yellowRange: this.yellowRange,
            c8y_Global: [!((_a = this.config) === null || _a === void 0 ? void 0 : _a.doNotAddGlobalFragmentByDefault), []]
        }, {
            validators: [
                C8yValidators.withinScale('redRange.min'),
                C8yValidators.withinScale('redRange.max'),
                C8yValidators.withinScale('yellowRange.min'),
                C8yValidators.withinScale('yellowRange.max'),
                C8yValidators.withinScale('target')
            ]
        });
    }
    removeNullValues(value) {
        // remove null values before sending data to backend
        const cleanedupValues = Object.assign({}, value);
        Object.keys(cleanedupValues).forEach(key => {
            if (cleanedupValues[key] === null) {
                delete cleanedupValues[key];
            }
        });
        return cleanedupValues;
    }
}
DatapointLibraryDetailsComponent.ɵfac = function DatapointLibraryDetailsComponent_Factory(t) { return new (t || DatapointLibraryDetailsComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ActivatedRoute), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.InventoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.FormBuilder), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Router), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.TranslateService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.ColorService), ɵngcc0.ɵɵdirectiveInject(ɵngcc6.DatapointParserService), ɵngcc0.ɵɵdirectiveInject(DATAPOINT_LIBRARY_CONFIG, 8)); };
DatapointLibraryDetailsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: DatapointLibraryDetailsComponent, selectors: [["c8y-datapoint-library-details"]], decls: 15, vars: 18, consts: [[4, "ngIf"], [3, "icon", "label", "path"], ["novalidate", "", 1, "card", "content-fullpage", "card--grid", "grid__col--6-6--md", "grid__row--fit-auto", 3, "formGroup", "ngSubmit", "change"], [1, "card-header", "large-padding", "separator", "grid__col--fullspan"], [1, "card-title"], ["class", "d-contents", 4, "ngIf"], [1, "card-footer", "separator", "grid__col--fullspan"], ["type", "button", 1, "btn", "btn-default", 3, "title", "click"], ["class", "btn btn-primary btn-form", "type", "submit", 3, "disabled", "title", 4, "c8yIfAllowed", "c8yIfAllowedAllowAny"], [1, "d-contents"], [1, "inner-scroll", "bg-white", "flex-grow"], [1, "card-block", "large-padding"], [1, "d-flex"], ["translate", ""], [1, "data-point-color", "form-control"], [1, "c8y-colorpicker"], ["type", "color", "name", "color", "formControlName", "color"], [1, "flex-grow", "p-l-8"], ["formControlName", "label", "name", "label", "type", "text", 1, "form-control", 3, "placeholder"], [3, "show"], ["formControlName", "description", "name", "description", "rows", "3", 1, "form-control", 3, "placeholder"], [1, "row"], [1, "col-sm-6"], ["name", "fragment", "formControlName", "fragment", "type", "text", 1, "form-control", 3, "placeholder"], ["formControlName", "series", "name", "series", "type", "text", 1, "form-control", 3, "placeholder"], ["class", "row", 4, "ngIf"], [1, "inner-scroll", "bg-gray-white"], [1, "c8y-fieldset"], [1, "m-b-16", "d-block", 3, "config"], ["formGroupName", "range", 1, "row"], [1, "col-md-6"], [3, "ngClass"], ["type", "number", "name", "min", "formControlName", "min", 1, "form-control", 3, "placeholder"], ["type", "number", "name", "max", "formControlName", "max", 1, "form-control", 3, "placeholder"], ["name", "unit", "formControlName", "unit", 1, "form-control", 3, "placeholder"], ["type", "number", "name", "target", "formControlName", "target", 1, "form-control", 3, "placeholder"], ["formGroupName", "yellowRange", 1, "c8y-fieldset"], ["formGroupName", "redRange", 1, "c8y-fieldset"], [1, "c8y-checkbox"], ["name", "c8y_Global", "formControlName", "c8y_Global", "type", "checkbox"], ["container", "body", "type", "button", 1, "btn", "btn-clean", 3, "popover", "outsideClick"], [1, "text-info", 3, "c8yIcon"], ["type", "submit", 1, "btn", "btn-primary", "btn-form", 3, "disabled", "title"]], template: function DatapointLibraryDetailsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, DatapointLibraryDetailsComponent_c8y_title_0_Template, 3, 2, "c8y-title", 0);
        ɵngcc0.ɵɵelementStart(1, "c8y-breadcrumb");
        ɵngcc0.ɵɵelement(2, "c8y-breadcrumb-item", 1);
        ɵngcc0.ɵɵpipe(3, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "form", 2);
        ɵngcc0.ɵɵlistener("ngSubmit", function DatapointLibraryDetailsComponent_Template_form_ngSubmit_4_listener() { return ctx.formGroup.valid && ctx.save(ctx.formGroup.value); })("change", function DatapointLibraryDetailsComponent_Template_form_change_4_listener() { return ctx.formChange(); });
        ɵngcc0.ɵɵelementStart(5, "div", 3);
        ɵngcc0.ɵɵelementStart(6, "h4", 4);
        ɵngcc0.ɵɵtext(7);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(8, DatapointLibraryDetailsComponent_div_8_Template, 113, 92, "div", 5);
        ɵngcc0.ɵɵelementStart(9, "div", 6);
        ɵngcc0.ɵɵelementStart(10, "button", 7);
        ɵngcc0.ɵɵlistener("click", function DatapointLibraryDetailsComponent_Template_button_click_10_listener() { return ctx.cancel(); });
        ɵngcc0.ɵɵpipe(11, "translate");
        ɵngcc0.ɵɵtext(12);
        ɵngcc0.ɵɵpipe(13, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(14, DatapointLibraryDetailsComponent_button_14_Template, 4, 7, "button", 8);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isLoading);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("icon", "c8y-data-points")("label", ɵngcc0.ɵɵpipeBind1(3, 11, "Data point library"))("path", ctx.path);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("formGroup", ctx.formGroup);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate1(" ", ctx.formGroup.value == null ? null : ctx.formGroup.value.label, " ");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isLoading);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(11, 13, "Cancel"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(13, 15, "Cancel"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yIfAllowed", ɵngcc0.ɵɵpureFunction0(17, _c9))("c8yIfAllowedAllowAny", true);
    } }, directives: [ɵngcc7.NgIf, ɵngcc4.BreadcrumbComponent, ɵngcc4.BreadcrumbItemComponent, ɵngcc3.ɵNgNoValidate, ɵngcc3.NgControlStatusGroup, ɵngcc3.FormGroupDirective, ɵngcc4.IfAllowedDirective, ɵngcc4.TitleComponent, ɵngcc4.FormGroupComponent, ɵngcc4.C8yTranslateDirective, ɵngcc3.DefaultValueAccessor, ɵngcc4.RequiredInputPlaceholderDirective, ɵngcc3.NgControlStatus, ɵngcc3.FormControlName, ɵngcc4.MessagesComponent, ɵngcc4.RangeDisplayComponent, ɵngcc3.FormGroupName, ɵngcc7.NgClass, ɵngcc3.NumberValueAccessor, ɵngcc3.CheckboxControlValueAccessor, ɵngcc8.PopoverDirective, ɵngcc4.IconDirective], pipes: [ɵngcc4.C8yTranslatePipe], encapsulation: 2 });
DatapointLibraryDetailsComponent.ctorParameters = () => [
    { type: ActivatedRoute },
    { type: InventoryService },
    { type: FormBuilder },
    { type: AlertService },
    { type: Router },
    { type: TranslateService },
    { type: ColorService },
    { type: DatapointParserService },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [DATAPOINT_LIBRARY_CONFIG,] }] }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DatapointLibraryDetailsComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-datapoint-library-details',
                template: "<c8y-title *ngIf=\"!isLoading\">\n  {{ formGroup.value?.label }}\n  <small *ngIf=\"formGroup.value?.fragment && formGroup.value?.series\"\n    >{{ formGroup.value?.fragment }} / {{ formGroup.value?.series }}</small\n  >\n</c8y-title>\n\n<c8y-breadcrumb>\n  <c8y-breadcrumb-item\n    [icon]=\"'c8y-data-points'\"\n    [label]=\"'Data point library' | translate\"\n    [path]=\"path\"\n  ></c8y-breadcrumb-item>\n</c8y-breadcrumb>\n\n<form\n  (ngSubmit)=\"formGroup.valid && save(formGroup.value)\"\n  (change)=\"formChange()\"\n  [formGroup]=\"formGroup\"\n  class=\"card content-fullpage card--grid grid__col--6-6--md grid__row--fit-auto\"\n  novalidate\n>\n  <div class=\"card-header large-padding separator grid__col--fullspan\">\n    <h4 class=\"card-title\">\n      {{ formGroup.value?.label }}\n    </h4>\n  </div>\n  <div *ngIf=\"!isLoading\" class=\"d-contents\">\n    <div class=\"inner-scroll bg-white flex-grow\">\n      <div class=\"card-block large-padding\">\n        <div class=\"d-flex\">\n          <c8y-form-group>\n            <label translate>Color</label>\n            <div class=\"data-point-color form-control\">\n              <div class=\"c8y-colorpicker\">\n                <input type=\"color\" name=\"color\" formControlName=\"color\" />\n                <span [style.background-color]=\"formGroup.value?.color\"></span>\n              </div>\n            </div>\n          </c8y-form-group>\n          <c8y-form-group class=\"flex-grow p-l-8\">\n            <label translate>Label</label>\n            <input\n              class=\"form-control\"\n              formControlName=\"label\"\n              name=\"label\"\n              [placeholder]=\"'e.g. Temperature' | translate\"\n              type=\"text\"\n            />\n            <c8y-messages [show]=\"formGroup.controls?.label?.touched && formGroup.controls?.label?.errors\">\n            </c8y-messages>\n          </c8y-form-group>\n        </div>\n        <!-- TODO: add description to data point library templates -->\n        <c8y-form-group>\n          <label translate>Description</label>\n          <textarea\n            class=\"form-control\"\n            formControlName=\"description\"\n            name=\"description\"\n            [placeholder]=\"'e.g. Ambient Temperature in Celsius' | translate\"\n            rows=\"3\"\n          ></textarea>\n        </c8y-form-group>\n\n        <div class=\"row\">\n          <div class=\"col-sm-6\">\n            <c8y-form-group>\n              <label translate>Fragment</label>\n              <input\n                class=\"form-control\"\n                name=\"fragment\"\n                formControlName=\"fragment\"\n                [placeholder]=\"'e.g. {{ example }}' | translate: { example: 'c8y_Temperature' }\"\n                type=\"text\"\n              />\n              <c8y-messages [show]=\"formGroup.controls?.fragment?.touched && formGroup.controls?.fragment?.errors\">\n              </c8y-messages>\n            </c8y-form-group>\n          </div>\n\n          <div class=\"col-sm-6\">\n            <c8y-form-group>\n              <label translate>Series</label>\n              <input\n                class=\"form-control\"\n                formControlName=\"series\"\n                name=\"series\"\n                [placeholder]=\"'e.g. {{ example }}' | translate: { example: 'T' }\"\n                type=\"text\"\n              />\n              <c8y-messages [show]=\"formGroup.controls?.series?.touched && formGroup.controls?.series?.errors\">\n              </c8y-messages>\n            </c8y-form-group>\n          </div>\n        </div>\n        <div class=\"row\" *ngIf=\"config?.showCheckboxForGlobalFragment\">\n          <div class=\"col-sm-6\">\n            <c8y-form-group>\n              <label class=\"c8y-checkbox\">\n                <input\n                  name=\"c8y_Global\"\n                  formControlName=\"c8y_Global\"\n                  type=\"checkbox\"\n                />\n                <span></span>\n                <span translate>Globally available</span>\n                <button\n                  class=\"btn btn-clean\"\n                  [popover]=\"'Will make this entry available to all users on the tenant if checked.' | translate\"\n                  container=\"body\"\n                  type=\"button\"\n                  [outsideClick]=\"true\"\n                >\n                  <i [c8yIcon]=\"'question-circle-o'\" class=\"text-info\"></i>\n                </button>\n              </label>\n            </c8y-form-group>\n          </div>\n        </div>\n      </div>\n    </div>\n    <div class=\"inner-scroll bg-gray-white\">\n      <div class=\"card-block large-padding\">\n        <fieldset class=\"c8y-fieldset\">\n          <legend translate>Preview</legend>\n          <c8y-range-display [config]=\"rangeConfig\" class=\"m-b-16 d-block\"></c8y-range-display>\n        </fieldset>\n        <fieldset class=\"c8y-fieldset\">\n          <legend translate>Range</legend>\n          <div class=\"row\" formGroupName=\"range\">\n            <div class=\"col-md-6\">\n              <c8y-form-group\n                [ngClass]=\"{ 'has-error': range?.touched && range?.controls?.min?.errors }\"\n              >\n                <label translate>Min</label>\n                <input\n                  type=\"number\"\n                  class=\"form-control\"\n                  name=\"min\"\n                  formControlName=\"min\"\n                  [placeholder]=\"'e.g. {{ example }}' | translate: { example: 0 }\"\n                />\n                <c8y-messages [show]=\"range?.touched && range.controls?.min?.errors\">\n                </c8y-messages>\n              </c8y-form-group>\n            </div>\n            <div class=\"col-md-6\">\n              <c8y-form-group\n                [ngClass]=\"{ 'has-error': range?.touched && range?.controls?.max?.errors }\"\n              >\n                <label translate>Max</label>\n                <input\n                  type=\"number\"\n                  class=\"form-control\"\n                  name=\"max\"\n                  formControlName=\"max\"\n                  [placeholder]=\"'e.g. {{ example }}' | translate: { example: 100 }\"\n                />\n                <c8y-messages [show]=\"range?.touched && range.controls?.max?.errors\">\n                </c8y-messages>\n              </c8y-form-group>\n            </div>\n          </div>\n          <div class=\"row\">\n            <div class=\"col-md-6\">\n              <c8y-form-group>\n                <label translate>Unit</label>\n                <input\n                  class=\"form-control\"\n                  name=\"unit\"\n                  formControlName=\"unit\"\n                  [placeholder]=\"'e.g. {{ example }}' | translate: { example: '\u00BAC' }\"\n                />\n                <c8y-messages [show]=\"formGroup.controls?.unit?.touched && formGroup.controls?.unit?.errors\">\n                </c8y-messages>\n              </c8y-form-group>\n            </div>\n\n            <div class=\"col-md-6\">\n              <c8y-form-group\n                [ngClass]=\"{\n                  'has-error':\n                    (range?.touched || formGroup.controls?.target?.touched) &&\n                    formGroup.controls?.target?.errors\n                }\"\n              >\n                <label translate>Target</label>\n                <input\n                  type=\"number\"\n                  class=\"form-control\"\n                  name=\"target\"\n                  formControlName=\"target\"\n                  [placeholder]=\"'e.g. {{ example }}' | translate: { example: 25 }\"\n                />\n                <c8y-messages [show]=\"(range?.touched || formGroup.controls?.target?.touched) && formGroup.controls?.target?.errors\">\n                </c8y-messages>\n              </c8y-form-group>\n            </div>\n          </div>\n        </fieldset>\n\n        <fieldset class=\"c8y-fieldset\" formGroupName=\"yellowRange\">\n          <legend translate>Yellow range</legend>\n          <div class=\"row\">\n            <div class=\"col-md-6\">\n              <c8y-form-group\n                [ngClass]=\"{\n                  'has-error':\n                    (range?.touched || yellowRange?.touched) && yellowRange?.controls?.min?.errors\n                }\"\n              >\n                <label translate>Min</label>\n                <input\n                  type=\"number\"\n                  class=\"form-control\"\n                  name=\"min\"\n                  formControlName=\"min\"\n                  [placeholder]=\"'e.g. {{ example }}' | translate: { example: 50 }\"\n                />\n                <c8y-messages [show]=\"(range?.touched || yellowRange?.touched) && yellowRange.controls?.min?.errors\">\n                </c8y-messages>\n              </c8y-form-group>\n            </div>\n\n            <div class=\"col-md-6\">\n              <c8y-form-group\n                [ngClass]=\"{\n                  'has-error':\n                    (range?.touched || yellowRange?.touched) && yellowRange?.controls?.max?.errors\n                }\"\n              >\n                <label translate>Max</label>\n                <input\n                  type=\"number\"\n                  class=\"form-control\"\n                  name=\"max\"\n                  formControlName=\"max\"\n                  [placeholder]=\"'e.g. {{ example }}' | translate: { example: 75 }\"\n                />\n                <c8y-messages [show]=\"(range?.touched || yellowRange?.touched) && yellowRange.controls?.max?.errors\">\n                </c8y-messages>\n              </c8y-form-group>\n            </div>\n          </div>\n        </fieldset>\n\n        <fieldset class=\"c8y-fieldset\" formGroupName=\"redRange\">\n          <legend translate>Red range</legend>\n          <div class=\"row\">\n            <div class=\"col-md-6\">\n              <c8y-form-group\n                [ngClass]=\"{\n                  'has-error':\n                    (range?.touched || redRange?.touched) && redRange?.controls?.min?.errors\n                }\"\n              >\n                <label translate>Min</label>\n                <input\n                  type=\"number\"\n                  class=\"form-control\"\n                  name=\"min\"\n                  formControlName=\"min\"\n                  [placeholder]=\"'e.g. {{ example }}' | translate: { example: 75 }\"\n                />\n                <c8y-messages [show]=\"(range?.touched || redRange?.touched) && redRange.controls?.min?.errors\">\n                </c8y-messages>\n              </c8y-form-group>\n            </div>\n\n            <div class=\"col-md-6\">\n              <c8y-form-group\n                [ngClass]=\"{\n                  'has-error':\n                    (range?.touched || redRange?.touched) && redRange?.controls?.max?.errors\n                }\"\n              >\n                <label translate>Max</label>\n                <input\n                  type=\"number\"\n                  class=\"form-control\"\n                  name=\"max\"\n                  formControlName=\"max\"\n                  [placeholder]=\"'e.g. {{ example }}' | translate: { example: 100 }\"\n                />\n                <c8y-messages [show]=\"(range?.touched || redRange?.touched) && redRange.controls?.max?.errors\">\n                </c8y-messages>\n              </c8y-form-group>\n            </div>\n          </div>\n        </fieldset>\n      </div>\n    </div>\n  </div>\n  <div class=\"card-footer separator grid__col--fullspan\">\n    <button (click)=\"cancel()\" class=\"btn btn-default\" [title]=\"'Cancel' | translate\" type=\"button\">\n      {{ 'Cancel' | translate }}\n    </button>\n    <button\n      *c8yIfAllowed=\"['ROLE_INVENTORY_ADMIN', 'ROLE_INVENTORY_UPDATE']; allowAny: true\"\n      [disabled]=\"formGroup.invalid || formGroup.pristine\"\n      class=\"btn btn-primary btn-form\"\n      [title]=\"'Save' | translate\"\n      type=\"submit\"\n    >\n      {{ 'Save' | translate }}\n    </button>\n  </div>\n</form>\n"
            }]
    }], function () { return [{ type: ɵngcc1.ActivatedRoute }, { type: ɵngcc2.InventoryService }, { type: ɵngcc3.FormBuilder }, { type: ɵngcc4.AlertService }, { type: ɵngcc1.Router }, { type: ɵngcc5.TranslateService }, { type: ɵngcc4.ColorService }, { type: ɵngcc6.DatapointParserService }, { type: undefined, decorators: [{
                type: Optional
            }, {
                type: Inject,
                args: [DATAPOINT_LIBRARY_CONFIG]
            }] }]; }, null); })();
