// @c8y/ngx-components/esm2015/assets-navigator/asset-selector/asset-selector.module.js
import { NgModule } from '@angular/core';
import { CoreModule, CommonModule, SearchModule } from '@c8y/ngx-components';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { GroupNodeService } from './group-node.service';
import { AssetSelectorNodeComponent } from './asset-selector-node.component';
import { AssetSelectorComponent } from './asset-selector.component';
import { FormsModule } from '@angular/forms';
import { MillerViewComponent } from './miller-view.component';
import { AssetSelectorService } from './asset-selector.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from '@angular/forms';
export class AssetSelectorModule {
}
AssetSelectorModule.ɵfac = function AssetSelectorModule_Factory(t) { return new (t || AssetSelectorModule)(); };
AssetSelectorModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: AssetSelectorModule });
AssetSelectorModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [GroupNodeService, AssetSelectorService], imports: [[CoreModule, CommonModule, CollapseModule, BsDropdownModule, SearchModule, FormsModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AssetSelectorModule, [{
        type: NgModule,
        args: [{
                imports: [CoreModule, CommonModule, CollapseModule, BsDropdownModule, SearchModule, FormsModule],
                exports: [AssetSelectorNodeComponent, AssetSelectorComponent, MillerViewComponent],
                providers: [GroupNodeService, AssetSelectorService],
                declarations: [AssetSelectorNodeComponent, AssetSelectorComponent, MillerViewComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AssetSelectorModule, { declarations: function () { return [AssetSelectorNodeComponent, AssetSelectorComponent, MillerViewComponent]; }, imports: function () { return [CoreModule, CommonModule, CollapseModule, BsDropdownModule, SearchModule, FormsModule]; }, exports: function () { return [AssetSelectorNodeComponent, AssetSelectorComponent, MillerViewComponent]; } }); })();
ɵngcc0.ɵɵsetComponentScope(AssetSelectorComponent, [ɵngcc1.NgIf, ɵngcc1.NgClass, ɵngcc1.NgStyle, ɵngcc2.SearchInputComponent, ɵngcc2.IconDirective, ɵngcc3.DefaultValueAccessor, ɵngcc3.NgControlStatus, ɵngcc3.NgModel, AssetSelectorNodeComponent], [ɵngcc2.C8yTranslatePipe]);
