// @c8y/ngx-components/esm2015/assets-navigator/asset-selector/asset-selector.component.js
import { __awaiter } from "tslib";
import { Component, EventEmitter, Input, Output, forwardRef, ChangeDetectorRef } from '@angular/core';
import { NG_VALIDATORS, NG_VALUE_ACCESSOR } from '@angular/forms';
import { InventoryService } from '@c8y/client';
import { GroupNodeService } from './group-node.service';
import { AssetSelectorService } from './asset-selector.service';
import { CONFIG_OPTIONS_DEFAULT } from './asset-selector.model';
import { UnassignedDevicesNode } from './unassigned-devices-node';
import { AssetSelectorBase } from './asset-selector-base';
import { Subject } from 'rxjs';
/**
 * Displays a hierarchical tree selector for assets.
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './group-node.service';
import * as ɵngcc2 from '@c8y/client';
import * as ɵngcc3 from './asset-selector.service';

function AssetSelectorComponent_div_0_c8y_search_input_4_Template(rf, ctx) { if (rf & 1) {
    const _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-search-input", 7);
    ɵngcc0.ɵɵlistener("onClick", function AssetSelectorComponent_div_0_c8y_search_input_4_Template_c8y_search_input_onClick_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r5 = ɵngcc0.ɵɵnextContext(2); return ctx_r5.onSearchResultClick($event); })("reset", function AssetSelectorComponent_div_0_c8y_search_input_4_Template_c8y_search_input_reset_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r7 = ɵngcc0.ɵɵnextContext(2); return ctx_r7.onSearchResultReset($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("mode", "select")("container", ctx_r4.container);
} }
const _c0 = function (a0) { return { "p-l-16 p-absolute p-r-16": a0 }; };
const _c1 = function (a0) { return { "z-index": a0 }; };
function AssetSelectorComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 4);
    ɵngcc0.ɵɵelementStart(1, "p", 5);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(4, AssetSelectorComponent_div_0_c8y_search_input_4_Template, 1, 2, "c8y-search-input", 6);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(6, _c0, ctx_r0.config.view === "miller"))("ngStyle", ɵngcc0.ɵɵpureFunction1(8, _c1, ctx_r0.config.singleColumn ? "30" : "20"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 4, ctx_r0.config.label), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.config.search && ctx_r0.root);
} }
function AssetSelectorComponent_div_1_p_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p", 16);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Selected asset"), " ");
} }
function AssetSelectorComponent_div_1_ng_template_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p", 17);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(2, 1, "Selected assets"));
} }
function AssetSelectorComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    const _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 8);
    ɵngcc0.ɵɵelementStart(1, "div", 9);
    ɵngcc0.ɵɵtemplate(2, AssetSelectorComponent_div_1_p_2_Template, 3, 3, "p", 10);
    ɵngcc0.ɵɵtemplate(3, AssetSelectorComponent_div_1_ng_template_3_Template, 3, 3, "ng-template", null, 11, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementStart(5, "button", 12);
    ɵngcc0.ɵɵlistener("click", function AssetSelectorComponent_div_1_Template_button_click_5_listener() { ɵngcc0.ɵɵrestoreView(_r12); const ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.clearSelectedDevices(); });
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵtext(7);
    ɵngcc0.ɵɵpipe(8, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(9, "div", 13);
    ɵngcc0.ɵɵelement(10, "i", 14);
    ɵngcc0.ɵɵelement(11, "i", 15);
    ɵngcc0.ɵɵelementStart(12, "span");
    ɵngcc0.ɵɵtext(13);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _r9 = ɵngcc0.ɵɵreference(4);
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.selectedDevice)("ngIfElse", _r9);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(6, 7, "Edit"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(8, 9, "Change"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("c8yIcon", "check-circle");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("c8yIcon", ctx_r1.selectedDevice.type === "c8y_DeviceGroup" ? "c8y-group" : "exchange");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r1.selectedDevice.name);
} }
function AssetSelectorComponent_div_2_p_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p", 22);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 23);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r13 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, ctx_r13.rootNode.breadcrumb || ctx_r13.rootNode.label));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 4, ctx_r13.rootNode.breadcrumb || ctx_r13.rootNode.label), " ");
} }
function AssetSelectorComponent_div_2_div_2_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r20 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 28);
    ɵngcc0.ɵɵlistener("click", function AssetSelectorComponent_div_2_div_2_button_1_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r19 = ɵngcc0.ɵɵnextContext(3); return ctx_r19.back(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 29);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 1, "Back"));
} }
function AssetSelectorComponent_div_2_div_2_label_2_Template(rf, ctx) { if (rf & 1) {
    const _r22 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "label", 30);
    ɵngcc0.ɵɵelementStart(1, "input", 31);
    ɵngcc0.ɵɵlistener("change", function AssetSelectorComponent_div_2_div_2_label_2_Template_input_change_1_listener() { ɵngcc0.ɵɵrestoreView(_r22); const ctx_r21 = ɵngcc0.ɵɵnextContext(3); return ctx_r21.select(ctx_r21.rootNode.mo); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(3, "span");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r17 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 2, "Select group"));
    ɵngcc0.ɵɵproperty("checked", ctx_r17.isGroupSelected());
} }
function AssetSelectorComponent_div_2_div_2_p_3_i_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 34);
} }
const _c2 = function (a0, a1) { return { "text-label-small": a0, "text-medium": a1 }; };
function AssetSelectorComponent_div_2_div_2_p_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p", 32);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtemplate(2, AssetSelectorComponent_div_2_div_2_p_3_i_2_Template, 1, 0, "i", 33);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r18 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 4, ctx_r18.rootNode.label));
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(8, _c2, !ctx_r18.config.singleColumn, ctx_r18.config.singleColumn));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r18.config.singleColumn || !ctx_r18.rootNode.root && ctx_r18.index === 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 6, ctx_r18.rootNode.label), " ");
} }
function AssetSelectorComponent_div_2_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 24);
    ɵngcc0.ɵɵtemplate(1, AssetSelectorComponent_div_2_div_2_button_1_Template, 3, 3, "button", 25);
    ɵngcc0.ɵɵtemplate(2, AssetSelectorComponent_div_2_div_2_label_2_Template, 4, 4, "label", 26);
    ɵngcc0.ɵɵtemplate(3, AssetSelectorComponent_div_2_div_2_p_3_Template, 5, 11, "p", 27);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r14 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r14.config.singleColumn && !ctx_r14.rootNode.root && ctx_r14.index !== 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r14.rootNode.root && ctx_r14.index === 0 && (ctx_r14.config.groupsSelectable || !ctx_r14.rootNode.isGroup()));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r14.config.columnHeaders);
} }
function AssetSelectorComponent_div_2_div_3_button_8_Template(rf, ctx) { if (rf & 1) {
    const _r26 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 39);
    ɵngcc0.ɵɵlistener("click", function AssetSelectorComponent_div_2_div_3_button_8_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r26); const ctx_r25 = ɵngcc0.ɵɵnextContext(3); return ctx_r25.clearFilters(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 42);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate1("title", " ", ɵngcc0.ɵɵpipeBind1(1, 1, "Clear filters"), "");
} }
const _c3 = function (a0) { return { "p-t-4": a0 }; };
function AssetSelectorComponent_div_2_div_3_Template(rf, ctx) { if (rf & 1) {
    const _r28 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 35);
    ɵngcc0.ɵɵelementStart(1, "div", 36);
    ɵngcc0.ɵɵelementStart(2, "input", 37);
    ɵngcc0.ɵɵlistener("ngModelChange", function AssetSelectorComponent_div_2_div_3_Template_input_ngModelChange_2_listener($event) { ɵngcc0.ɵɵrestoreView(_r28); const ctx_r27 = ɵngcc0.ɵɵnextContext(2); return ctx_r27.filterText = $event; })("keyup.enter", function AssetSelectorComponent_div_2_div_3_Template_input_keyup_enter_2_listener() { ɵngcc0.ɵɵrestoreView(_r28); const ctx_r29 = ɵngcc0.ɵɵnextContext(2); return ctx_r29.applyFilter("*" + ctx_r29.filterText + "*"); });
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "span", 38);
    ɵngcc0.ɵɵelementStart(5, "button", 39);
    ɵngcc0.ɵɵlistener("click", function AssetSelectorComponent_div_2_div_3_Template_button_click_5_listener() { ɵngcc0.ɵɵrestoreView(_r28); const ctx_r30 = ɵngcc0.ɵɵnextContext(2); return ctx_r30.applyFilter("*" + ctx_r30.filterText + "*"); });
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelement(7, "i", 40);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(8, AssetSelectorComponent_div_2_div_3_button_8_Template, 3, 3, "button", 41);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r15 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(9, _c3, ctx_r15.config.columnHeaders));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(3, 5, "Filter this column\u2026"));
    ɵngcc0.ɵɵproperty("ngModel", ctx_r15.filterText);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(6, 7, "Apply filter"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r15.filterText.length);
} }
const _c4 = function (a0, a1) { return { "m-t-72 ": a0, "m-t-32": a1 }; };
const _c5 = function (a0) { return { top: a0 }; };
function AssetSelectorComponent_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 18);
    ɵngcc0.ɵɵtemplate(1, AssetSelectorComponent_div_2_p_1_Template, 5, 6, "p", 19);
    ɵngcc0.ɵɵtemplate(2, AssetSelectorComponent_div_2_div_2_Template, 4, 3, "div", 20);
    ɵngcc0.ɵɵtemplate(3, AssetSelectorComponent_div_2_div_3_Template, 9, 11, "div", 21);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(5, _c4, ctx_r2.config.search, !ctx_r2.config.search && ctx_r2.config.singleColumn))("ngStyle", ɵngcc0.ɵɵpureFunction1(8, _c5, ctx_r2.config.search && !ctx_r2.config.singleColumn ? "72px" : "0"));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.config.singleColumn && ctx_r2.index !== 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.config.columnHeaders || ctx_r2.config.singleColumn);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r2.rootNode.root && ctx_r2.config.showFilter);
} }
const _c6 = function (a0) { return { "collapse show": a0 }; };
function AssetSelectorComponent_c8y_asset_selector_node_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-asset-selector-node", 43);
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("node", ctx_r3.rootNode)("preselected", ctx_r3.selected)("multi", ctx_r3.config.multi)("view", ctx_r3.config.view)("index", ctx_r3.index)("selectedItems", ctx_r3.selectedItems)("active", ctx_r3.active)("ngClass", ɵngcc0.ɵɵpureFunction1(8, _c6, !ctx_r3.rootNode.root && !ctx_r3.rootNode.hidden && ctx_r3.config.view !== "miller"));
} }
export class AssetSelectorComponent extends AssetSelectorBase {
    /**
     * @ignore only DI
     */
    constructor(groupNodeService, inventory, assetSelectorService, cd) {
        super(groupNodeService, inventory, assetSelectorService);
        this.groupNodeService = groupNodeService;
        this.inventory = inventory;
        this.assetSelectorService = assetSelectorService;
        this.cd = cd;
        /**
         * Config object containing all options for the asset selector.
         */
        this.config = {};
        /**
         * Used only for miller-view, displays the column level for the current node.
         * E.g if the index is one, this will be second column.
         */
        this.index = 0;
        /**
         * The asset to use as root.
         */
        this.asset = undefined;
        /**
         * The selected device.
         */
        this.selectedDevice = undefined;
        /**
         * An array of predefined nodes.
         */
        this.selected = [];
        /**
         * All currently selected assets.
         */
        this.selectedItems = [];
        this.container = 'body';
        /**
         * Emits if one item was selected (all currently selected nodes).
         */
        this.onSelected = new EventEmitter();
        /**
         * Used only for miller view. Emit the selected node and use it as a 'rootNode' for the new column.
         */
        this.onRowSelected = new EventEmitter();
        /**
         * Emits the current loading state of the node.
         */
        this.onLoad = new EventEmitter();
        /**
         * The current filter applied.
         */
        this.filterText = '';
        /**
         * Displays the global search at all times if the miller view is used
         * on root group level.
         */
        this.root = false;
        /**
         * @ignore
         */
        this.unsubscribe$ = new Subject();
    }
    /**
     * Checks when a node was selected, if a new column needs to be added.
     * @param node The node that was clicked.
     * @param index The current index of this node.
     * @returns True if the click was handled and a new column was added.
     */
    handleNextMillerViewColumn(node, index) {
        const nodeCopy = Object.create(node);
        const isSameAsRoot = this.rootNode === node;
        const isUnassignedDevicesNode = node.toString() === UnassignedDevicesNode.NAME;
        const isNewColumnNeeded = !isSameAsRoot && (node.isGroup() || isUnassignedDevicesNode || node.hasChildDevices());
        if (isNewColumnNeeded) {
            this.onRowSelected.emit({ nodeCopy, index });
            return false;
        }
        return true;
    }
    /**
     * @ignore
     */
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.index === 0) {
                this.config = Object.assign(Object.assign({}, CONFIG_OPTIONS_DEFAULT), this.config);
            }
            const shouldResetNameFilter = this.rootNode && this.rootNode.mo && this.config.view === 'miller';
            if (shouldResetNameFilter) {
                this.applyFilter('');
            }
            const isOnlyAssetNodeDefined = this.asset;
            if (isOnlyAssetNodeDefined) {
                this.rootNode = yield this.createManagedObjectRootNode(this.asset);
            }
            const isAnyNodeDefined = !this.rootNode && (!this.asset || Object.keys(this.asset).length === 0);
            if (isAnyNodeDefined) {
                this.rootNode = this.createDefaultRootNode();
            }
            if (this.config.showUnassignedDevices &&
                (!this.asset || Object.keys(this.asset).length === 0) &&
                this.rootNode.root) {
                this.rootNode.add(new UnassignedDevicesNode(this.groupNodeService, {
                    showChildDevices: this.config.showChildDevices
                }));
            }
            this.root = this.rootNode.root;
        });
    }
    ngOnChanges(changes) {
        return __awaiter(this, void 0, void 0, function* () {
            if (changes.asset &&
                changes.asset.currentValue &&
                Object.keys(changes.asset.currentValue).length !== 0) {
                this.rootNode = yield this.createManagedObjectRootNode(changes.asset.currentValue);
            }
        });
    }
    /**
     * Changes the current root to one level back.
     */
    back() {
        const index = this.index - 1;
        const nodeCopy = Object.create(this.rootNode.parents[0]);
        this.onRowSelected.emit({ nodeCopy, index });
    }
    /**
     * Clears the current filter.
     */
    clearFilters() {
        this.filterText = '';
        this.applyFilter('');
    }
    clearSelectedDevices() {
        this.selectedDevice = undefined;
        this.applyFilter('');
    }
    isGroupSelected() {
        if (this.selectedItems.some(selectedNode => { var _a; return selectedNode.id === ((_a = this.rootNode.mo) === null || _a === void 0 ? void 0 : _a.id); })) {
            return true;
        }
        return false;
    }
    groupNameFilter(nameFilter, moId, showUnassignedDevices) {
        if (nameFilter) {
            const filterByName = showUnassignedDevices
                ? {
                    __filter: {
                        name: nameFilter
                    },
                    __orderby: [{ name: 1 }]
                }
                : {
                    __filter: {
                        name: nameFilter,
                        __bygroupid: moId
                    }
                };
            return this.assetSelectorService.queriesUtil.buildQuery(filterByName);
        }
        const defaultFilter = {
            __filter: {
                __bygroupid: moId
            },
            __orderby: [{ name: 1 }]
        };
        return this.assetSelectorService.queriesUtil.buildQuery(defaultFilter);
    }
    /**
     * Applies a filter.
     * @param filter The filter to apply.
     */
    applyFilter(filter) {
        return __awaiter(this, void 0, void 0, function* () {
            this.rootNode.filterQuery$.next(this.groupNameFilter(filter, this.rootNode.mo.id, this.isUnassignedDevicesNode() ? true : false));
            this.rootNode.refresh();
            this.onLoad.emit({
                loading: true,
                filterText: this.filterText,
                index: this.index,
                selectedDevice: this.selectedDevice
            });
            this.cd.detectChanges();
        });
    }
    /**
     * Check if the UnassignedDevice node is the rootNode.
     */
    isUnassignedDevicesNode() {
        return this.rootNode.toString() === UnassignedDevicesNode.NAME;
    }
    /**
     * Called when the user resets the search result.
     */
    onSearchResultReset(term) {
        if (term) {
            this.deselectAll();
            this.rootNode = this.createDefaultRootNode();
            if (this.config.showUnassignedDevices) {
                this.rootNode.add(new UnassignedDevicesNode(this.groupNodeService, {
                    showChildDevices: this.config.showChildDevices
                }));
            }
            if (this.config.view === 'miller') {
                this.onRowSelected.emit({ root: true });
            }
            this.rootNode.click({ open: true });
            if (this.config.view === 'tree') {
                this.rootNode.open = true;
            }
        }
    }
    /**
     * Called if the user clicks on a search result.
     */
    onSearchResultClick(mo) {
        return __awaiter(this, void 0, void 0, function* () {
            const canSelectNode = this.config.groupsSelectable || !this.groupNodeService.isGroup(mo);
            this.rootNode = yield this.createManagedObjectRootNode(mo);
            this.rootNode.click({ open: true });
            if (this.config.view === 'miller') {
                const nodeCopy = Object.create(this.rootNode);
                this.onRowSelected.emit({
                    nodeCopy,
                    index: 0,
                    selectedDevices: this.selectedDevice,
                    root: true
                });
            }
            this.deselectAll();
            if (canSelectNode) {
                this.select(this.rootNode.mo);
            }
        });
    }
}
AssetSelectorComponent.ɵfac = function AssetSelectorComponent_Factory(t) { return new (t || AssetSelectorComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.GroupNodeService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.InventoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.AssetSelectorService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef)); };
AssetSelectorComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AssetSelectorComponent, selectors: [["c8y-asset-selector"]], inputs: { config: "config", index: "index", asset: "asset", selectedDevice: "selectedDevice", selected: "selected", selectedItems: "selectedItems", container: "container", rootNode: "rootNode", active: "active" }, outputs: { onSelected: "onSelected", onRowSelected: "onRowSelected", onLoad: "onLoad" }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: NG_VALUE_ACCESSOR,
                multi: true,
                useExisting: forwardRef(() => AssetSelectorComponent)
            },
            {
                provide: NG_VALIDATORS,
                useExisting: forwardRef(() => AssetSelectorComponent),
                multi: true
            }
        ]), ɵngcc0.ɵɵInheritDefinitionFeature, ɵngcc0.ɵɵNgOnChangesFeature], decls: 4, vars: 4, consts: [["class", "p-b-8 bg-inherit", "style", "left: 0; right: 4px", 3, "ngClass", "ngStyle", 4, "ngIf"], ["class", "p-r-16", 4, "ngIf"], ["class", "miller-column__header sticky-top bg-inherit separator-top-bottom", 3, "ngClass", "ngStyle", 4, "ngIf"], ["class", "d-block bg-inherit p-relative", "style", "z-index: 9", 3, "node", "preselected", "multi", "view", "index", "selectedItems", "active", "ngClass", 4, "ngIf"], [1, "p-b-8", "bg-inherit", 2, "left", "0", "right", "4px", 3, "ngClass", "ngStyle"], [1, "text-medium", "m-b-4", "p-t-8"], [3, "mode", "container", "onClick", "reset", 4, "ngIf"], [3, "mode", "container", "onClick", "reset"], [1, "p-r-16"], [1, "d-flex", "p-b-4"], ["class", "text-medium p-t-8 m-r-8", 4, "ngIf", "ngIfElse"], ["multiAssets", ""], [1, "btn", "btn-default", "btn-xs", "a-s-center", "m-t-4", "m-l-auto", 3, "title", "click"], [1, "d-flex"], [1, "text-success", "p-l-0", "p-r-8", 3, "c8yIcon"], [1, "m-r-4", "icon-20", 3, "c8yIcon"], [1, "text-medium", "p-t-8", "m-r-8"], [1, "text-medium", "m-r-8"], [1, "miller-column__header", "sticky-top", "bg-inherit", "separator-top-bottom", 3, "ngClass", "ngStyle"], ["class", "text-12 text-muted text-truncate m-b-4", 3, "title", 4, "ngIf"], ["class", "d-flex a-i-center p-b-4", 4, "ngIf"], [3, "ngClass", 4, "ngIf"], [1, "text-12", "text-muted", "text-truncate", "m-b-4", 3, "title"], ["c8yIcon", "home", 1, "m-r-4"], [1, "d-flex", "a-i-center", "p-b-4"], ["class", "btn btn-default btn-xs m-r-8 p-t-0 p-b-0 p-l-4 p-r-4 l-h-1", 3, "title", "click", 4, "ngIf"], ["class", "c8y-radio checkbox-inline m-r-8", 4, "ngIf"], ["class", "text-truncate", 3, "title", "ngClass", 4, "ngIf"], [1, "btn", "btn-default", "btn-xs", "m-r-8", "p-t-0", "p-b-0", "p-l-4", "p-r-4", "l-h-1", 3, "title", "click"], ["c8yIcon", "angle-left"], [1, "c8y-radio", "checkbox-inline", "m-r-8"], ["type", "radio", 3, "checked", "title", "change"], [1, "text-truncate", 3, "title", "ngClass"], ["c8yIcon", "c8y-group-open", "class", "icon-20 c8y-icon-duocolor m-r-4", 4, "ngIf"], ["c8yIcon", "c8y-group-open", 1, "icon-20", "c8y-icon-duocolor", "m-r-4"], [3, "ngClass"], [1, "input-group", "input-group-sm", "input-group-search"], [1, "form-control", 3, "ngModel", "placeholder", "ngModelChange", "keyup.enter"], [1, "input-group-btn"], [1, "btn", "btn-clean", "p-r-8", "p-l-4", 3, "title", "click"], ["c8yIcon", "filter"], ["class", "btn btn-clean p-r-8 p-l-4", 3, "title", "click", 4, "ngIf"], ["c8yIcon", "times"], [1, "d-block", "bg-inherit", "p-relative", 2, "z-index", "9", 3, "node", "preselected", "multi", "view", "index", "selectedItems", "active", "ngClass"]], template: function AssetSelectorComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, AssetSelectorComponent_div_0_Template, 5, 10, "div", 0);
        ɵngcc0.ɵɵtemplate(1, AssetSelectorComponent_div_1_Template, 14, 11, "div", 1);
        ɵngcc0.ɵɵtemplate(2, AssetSelectorComponent_div_2_Template, 4, 10, "div", 2);
        ɵngcc0.ɵɵtemplate(3, AssetSelectorComponent_c8y_asset_selector_node_3_Template, 1, 10, "c8y-asset-selector-node", 3);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", !ctx.selectedDevice && ctx.index === 0 && (ctx.config.search || ctx.config.singleColumn));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.selectedDevice);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.config.view === "miller" && (ctx.config.showFilter && !ctx.rootNode.root || ctx.config.columnHeaders || ctx.config.singleColumn) && !ctx.selectedDevice);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.rootNode && !ctx.selectedDevice);
    } }, encapsulation: 2 });
AssetSelectorComponent.ctorParameters = () => [
    { type: GroupNodeService },
    { type: InventoryService },
    { type: AssetSelectorService },
    { type: ChangeDetectorRef }
];
AssetSelectorComponent.propDecorators = {
    config: [{ type: Input }],
    active: [{ type: Input }],
    index: [{ type: Input }],
    asset: [{ type: Input }],
    selectedDevice: [{ type: Input }],
    selected: [{ type: Input }],
    rootNode: [{ type: Input }],
    selectedItems: [{ type: Input }],
    container: [{ type: Input }],
    onSelected: [{ type: Output }],
    onRowSelected: [{ type: Output }],
    onLoad: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AssetSelectorComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-asset-selector',
                template: "<div\n  class=\"p-b-8 bg-inherit\"\n  [ngClass]=\"{ 'p-l-16 p-absolute p-r-16': config.view === 'miller' }\"\n  style=\"left: 0; right: 4px\"\n  [ngStyle]=\"{ 'z-index': config.singleColumn ? '30' : '20' }\"\n  *ngIf=\"!selectedDevice && index === 0 && (config.search || config.singleColumn)\"\n>\n  <p class=\"text-medium m-b-4 p-t-8\">\n    {{ config.label | translate }}\n  </p>\n  <c8y-search-input\n    *ngIf=\"config.search && root\"\n    (onClick)=\"onSearchResultClick($event)\"\n    (reset)=\"onSearchResultReset($event)\"\n    [mode]=\"'select'\"\n    [container]=\"container\"\n  ></c8y-search-input>\n</div>\n\n<div *ngIf=\"selectedDevice\" class=\"p-r-16\">\n  <div class=\"d-flex p-b-4\">\n    <p *ngIf=\"selectedDevice; else multiAssets\" class=\"text-medium p-t-8 m-r-8\">\n      {{ 'Selected asset' | translate }}\n    </p>\n    <ng-template #multiAssets>\n      <p class=\"text-medium m-r-8\">{{ 'Selected assets' | translate }}</p>\n    </ng-template>\n    <button\n      class=\"btn btn-default btn-xs a-s-center m-t-4 m-l-auto\"\n      (click)=\"clearSelectedDevices()\"\n      title=\"{{ 'Edit' | translate }}\"\n    >\n      {{ 'Change' | translate }}\n    </button>\n  </div>\n  <div class=\"d-flex\">\n    <i [c8yIcon]=\"'check-circle'\" class=\"text-success p-l-0 p-r-8\"></i>\n    <i\n      c8yIcon=\"{{ selectedDevice.type === 'c8y_DeviceGroup' ? 'c8y-group' : 'exchange' }}\"\n      class=\"m-r-4 icon-20\"\n    ></i>\n    <span>{{ selectedDevice.name }}</span>\n  </div>\n</div>\n\n<!-- miller columns header -->\n<div\n  *ngIf=\"\n    config.view === 'miller' &&\n    ((config.showFilter && !rootNode.root) || config.columnHeaders || config.singleColumn) &&\n    !selectedDevice\n  \"\n  class=\"miller-column__header sticky-top bg-inherit separator-top-bottom\"\n  [ngClass]=\"{ 'm-t-72 ': config.search, 'm-t-32': !config.search && config.singleColumn }\"\n  [ngStyle]=\"{ top: config.search && !config.singleColumn ? '72px' : '0' }\"\n>\n  <p\n    class=\"text-12 text-muted text-truncate m-b-4\"\n    *ngIf=\"config.singleColumn && index !== 0\"\n    title=\"{{ rootNode.breadcrumb || rootNode.label | translate }}\"\n  >\n    <i c8yIcon=\"home\" class=\"m-r-4\"></i>\n    {{ rootNode.breadcrumb || rootNode.label | translate }}\n  </p>\n  <div *ngIf=\"config.columnHeaders || config.singleColumn\" class=\"d-flex a-i-center p-b-4\">\n    <button\n      class=\"btn btn-default btn-xs m-r-8 p-t-0 p-b-0 p-l-4 p-r-4 l-h-1\"\n      *ngIf=\"config.singleColumn && !rootNode.root && index !== 0\"\n      title=\"{{ 'Back' | translate }}\"\n      (click)=\"back()\"\n    >\n      <i c8yIcon=\"angle-left\"></i>\n    </button>\n    <label\n      *ngIf=\"!rootNode.root && index === 0 && (config.groupsSelectable || !rootNode.isGroup())\"\n      class=\"c8y-radio checkbox-inline m-r-8\"\n    >\n      <input\n        type=\"radio\"\n        (change)=\"select(rootNode.mo)\"\n        [checked]=\"isGroupSelected()\"\n        title=\"{{ 'Select group' | translate }}\"\n      /><span></span>\n    </label>\n\n    <p\n      class=\"text-truncate\"\n      title=\"{{ rootNode.label | translate }}\"\n      *ngIf=\"config.columnHeaders\"\n      [ngClass]=\"{\n        'text-label-small': !config.singleColumn,\n        'text-medium': config.singleColumn\n      }\"\n    >\n      <i\n        *ngIf=\"config.singleColumn || (!rootNode.root && index === 0)\"\n        c8yIcon=\"c8y-group-open\"\n        class=\"icon-20 c8y-icon-duocolor m-r-4\"\n      ></i>\n      {{ rootNode.label | translate }}\n    </p>\n  </div>\n\n  <div *ngIf=\"!rootNode.root && config.showFilter\" [ngClass]=\"{ 'p-t-4': config.columnHeaders }\">\n    <div class=\"input-group input-group-sm input-group-search\">\n      <input\n        [(ngModel)]=\"filterText\"\n        placeholder=\"{{ 'Filter this column\u2026' | translate }}\"\n        class=\"form-control\"\n        (keyup.enter)=\"applyFilter('*' + filterText + '*')\"\n      />\n      <span class=\"input-group-btn\">\n        <button\n          title=\"{{ 'Apply filter' | translate }}\"\n          class=\"btn btn-clean p-r-8 p-l-4\"\n          (click)=\"applyFilter('*' + filterText + '*')\"\n        >\n          <i c8yIcon=\"filter\"></i>\n        </button>\n        <button\n          title=\" {{ 'Clear filters' | translate }}\"\n          class=\"btn btn-clean p-r-8 p-l-4\"\n          (click)=\"clearFilters()\"\n          *ngIf=\"filterText.length\"\n        >\n          <i c8yIcon=\"times\"></i>\n        </button>\n      </span>\n    </div>\n  </div>\n</div>\n\n<c8y-asset-selector-node\n  *ngIf=\"rootNode && !selectedDevice\"\n  [node]=\"rootNode\"\n  [preselected]=\"selected\"\n  [multi]=\"config.multi\"\n  [view]=\"config.view\"\n  [index]=\"index\"\n  [selectedItems]=\"selectedItems\"\n  [active]=\"active\"\n  class=\"d-block bg-inherit p-relative\"\n  [ngClass]=\"{ 'collapse show': !rootNode.root && !rootNode.hidden && config.view !== 'miller' }\"\n  style=\"z-index: 9\"\n></c8y-asset-selector-node>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        multi: true,
                        useExisting: forwardRef(() => AssetSelectorComponent)
                    },
                    {
                        provide: NG_VALIDATORS,
                        useExisting: forwardRef(() => AssetSelectorComponent),
                        multi: true
                    }
                ]
            }]
    }], function () { return [{ type: ɵngcc1.GroupNodeService }, { type: ɵngcc2.InventoryService }, { type: ɵngcc3.AssetSelectorService }, { type: ɵngcc0.ChangeDetectorRef }]; }, { config: [{
            type: Input
        }], index: [{
            type: Input
        }], asset: [{
            type: Input
        }], selectedDevice: [{
            type: Input
        }], selected: [{
            type: Input
        }], selectedItems: [{
            type: Input
        }], container: [{
            type: Input
        }], onSelected: [{
            type: Output
        }], onRowSelected: [{
            type: Output
        }], onLoad: [{
            type: Output
        }], rootNode: [{
            type: Input
        }], active: [{
            type: Input
        }] }); })();
