// @c8y/ngx-components/esm2015/assets-navigator/asset-selector/asset-selector.service.js
import { Injectable } from '@angular/core';
import { AssetNodeService } from '../asset-node.service';
import { isArray, isNumber, isObject, isString } from 'lodash';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/client/lib/src/inventory/InventoryService";
import * as i2 from "@c8y/ngx-components/api";
import * as i3 from "@c8y/ngx-components";
import * as i4 from "@c8y/client/lib/src/user/UserService";
import * as i5 from "../asset-node-config.model";
import * as i6 from "../group.service";
import * as ɵngcc0 from '@angular/core';
export class AssetSelectorService extends AssetNodeService {
    /**
     * Simplifies the object model based on the selected mode.
     * @param obj The selected asset.
     * @param mode The mode which will decide what type of model will be returned.
     */
    normalizeValue(obj, modelMode) {
        return this.simplifyModel(this.normalizeModelValue(obj), modelMode);
    }
    simplifyModel(model, mode) {
        const mapModel = modelObj => {
            const { id, name, c8y_DeviceQueryString } = modelObj;
            return Object.assign({ id, name }, (c8y_DeviceQueryString ? { c8y_DeviceQueryString } : {}));
        };
        if (mode === 'full') {
            return model;
        }
        if (!isArray(model)) {
            return mapModel(model);
        }
        return model.map(mapModel);
    }
    normalizeModelValue(value) {
        if (isNumber(value) || isString(value)) {
            return [{ id: value }];
        }
        if (isArray(value)) {
            return value;
        }
        if (isObject(value)) {
            return [value];
        }
        return [];
    }
}
AssetSelectorService.ɵfac = /*@__PURE__*/ function () { let ɵAssetSelectorService_BaseFactory; return function AssetSelectorService_Factory(t) { return (ɵAssetSelectorService_BaseFactory || (ɵAssetSelectorService_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(AssetSelectorService)))(t || AssetSelectorService); }; }();
AssetSelectorService.ɵprov = i0.ɵɵdefineInjectable({ factory: function AssetSelectorService_Factory() { return new AssetSelectorService(i0.ɵɵinject(i1.InventoryService), i0.ɵɵinject(i2.ApiService), i0.ɵɵinject(i3.ModalService), i0.ɵɵinject(i3.AlertService), i0.ɵɵinject(i3.BreadcrumbService), i0.ɵɵinject(i4.UserService), i0.ɵɵinject(i3.AppStateService), i0.ɵɵinject(i3.OptionsService), i0.ɵɵinject(i5.ASSET_NAVIGATOR_CONFIG, 8), i0.ɵɵinject(i6.DeviceGroupService)); }, token: AssetSelectorService, providedIn: "root" });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AssetSelectorService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], null, null); })();
