// @c8y/ngx-components/esm2015/cockpit-config/cockpit-configuration.component.js
import { __awaiter } from "tslib";
import { Component } from '@angular/core';
import { AlertService, gettext } from '@c8y/ngx-components';
import { DEFAULT_CONFIG } from './cockpit-config.model';
import { CockpitConfigService } from './cockpit-config.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './cockpit-config.service';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from '@angular/forms';
import * as ɵngcc4 from 'ngx-bootstrap/popover';
import * as ɵngcc5 from './feature-config.component';
import * as ɵngcc6 from './root-node-config.component';
import * as ɵngcc7 from './home-dashboard-config.component';
import * as ɵngcc8 from './misc-config.component';

const _c0 = function (a0) { return { config: a0 }; };
export class CockpitConfigurationComponent {
    constructor(cockpitConfigService, alertService) {
        this.cockpitConfigService = cockpitConfigService;
        this.alertService = alertService;
        /**
         * The currently used configuration.
         */
        this.config = DEFAULT_CONFIG;
    }
    /**
     * @ignore
     */
    ngOnInit() {
        this.config = this.cockpitConfigService.currentConfig;
    }
    /**
     * Stores the configuration and shows a success message.
     */
    save() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.cockpitConfigService.saveConfig(this.config);
                this.alertService.success(gettext('Cockpit configuration saved.'));
            }
            catch (ex) {
                this.alertService.addServerFailure(ex);
            }
        });
    }
    /**
     * Updates the features to directly reflect the results of the change.
     */
    updateFeatures() {
        this.cockpitConfigService.currentConfig = this.config;
        this.cockpitConfigService.refresh();
    }
    /**
     * Updates the root nodes to directly reflect the results of the change.
     */
    updateRootNodes() {
        this.cockpitConfigService.currentConfig = this.config;
        this.cockpitConfigService.setRootNodes();
    }
}
CockpitConfigurationComponent.ɵfac = function CockpitConfigurationComponent_Factory(t) { return new (t || CockpitConfigurationComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.CockpitConfigService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.AlertService)); };
CockpitConfigurationComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: CockpitConfigurationComponent, selectors: [["c8y-cockpit-configuration"]], decls: 53, vars: 27, consts: [[1, "row"], [1, "col-lg-12", "col-lg-max"], ["configForm", "ngForm"], [1, "card", "card--fullpage"], [1, "card-header", "separator"], ["translate", "", 1, "card-title"], [1, "inner-scroll"], [1, "card-block", "p-t-0", "p-b-0"], [1, "row", "separator-bottom", "p-t-24", "p-b-24"], [1, "col-xs-12", "col-sm-3", "col-md-2", "text-left-xs", "text-right-sm"], [1, "d-inline-flex", "m-b-16"], ["translate", "", 1, "text-medium", "m-r-4"], ["container", "body", "placement", "right", 1, "btn", "btn-clean", 3, "popover", "outsideClick"], [1, "text-info", 3, "c8yIcon"], [1, "col-xs-12", "col-sm-9", "col-md-8"], [3, "config", "onUpdate"], ["translate", "", 1, "text-medium", "m-r-8"], [1, "col-xs-12", "col-sm-10", "col-md-10"], ["placement", "right", "container", "body", 1, "btn", "btn-clean", 3, "popover", "outsideClick"], [3, "config"], ["translate", "", 1, "text-medium", "d-inline-block", "m-r-4"], [1, "card-footer", "separator"], ["type", "submit", "translate", "", "c8yProductExperience", "", 1, "btn", "btn-primary", 3, "disabled", "actionName", "actionData", "click"]], template: function CockpitConfigurationComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵpipe(2, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "div", 0);
        ɵngcc0.ɵɵelementStart(4, "div", 1);
        ɵngcc0.ɵɵelementStart(5, "form", null, 2);
        ɵngcc0.ɵɵelementStart(7, "div", 3);
        ɵngcc0.ɵɵelementStart(8, "div", 4);
        ɵngcc0.ɵɵelementStart(9, "div", 5);
        ɵngcc0.ɵɵtext(10, "Cockpit configuration");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "div", 6);
        ɵngcc0.ɵɵelementStart(12, "div", 7);
        ɵngcc0.ɵɵelementStart(13, "fieldset", 8);
        ɵngcc0.ɵɵelementStart(14, "div", 9);
        ɵngcc0.ɵɵelementStart(15, "div", 10);
        ɵngcc0.ɵɵelementStart(16, "h4", 11);
        ɵngcc0.ɵɵtext(17, "Features");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(18, "button", 12);
        ɵngcc0.ɵɵpipe(19, "translate");
        ɵngcc0.ɵɵelement(20, "i", 13);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(21, "div", 14);
        ɵngcc0.ɵɵelementStart(22, "c8y-feature-config", 15);
        ɵngcc0.ɵɵlistener("onUpdate", function CockpitConfigurationComponent_Template_c8y_feature_config_onUpdate_22_listener() { return ctx.updateFeatures(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(23, "fieldset", 8);
        ɵngcc0.ɵɵelementStart(24, "div", 9);
        ɵngcc0.ɵɵelementStart(25, "div", 10);
        ɵngcc0.ɵɵelementStart(26, "h4", 16);
        ɵngcc0.ɵɵtext(27, "Top level nodes");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(28, "button", 12);
        ɵngcc0.ɵɵpipe(29, "translate");
        ɵngcc0.ɵɵelement(30, "i", 13);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(31, "div", 17);
        ɵngcc0.ɵɵelementStart(32, "div", 0);
        ɵngcc0.ɵɵelementStart(33, "c8y-root-node-config", 15);
        ɵngcc0.ɵɵlistener("onUpdate", function CockpitConfigurationComponent_Template_c8y_root_node_config_onUpdate_33_listener() { return ctx.updateRootNodes(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(34, "fieldset", 8);
        ɵngcc0.ɵɵelementStart(35, "div", 9);
        ɵngcc0.ɵɵelementStart(36, "div", 10);
        ɵngcc0.ɵɵelementStart(37, "h4", 16);
        ɵngcc0.ɵɵtext(38, "Home dashboard");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(39, "button", 18);
        ɵngcc0.ɵɵpipe(40, "translate");
        ɵngcc0.ɵɵelement(41, "i", 13);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(42, "div", 14);
        ɵngcc0.ɵɵelement(43, "c8y-home-dashboard-config", 19);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(44, "fieldset", 8);
        ɵngcc0.ɵɵelementStart(45, "div", 9);
        ɵngcc0.ɵɵelementStart(46, "h4", 20);
        ɵngcc0.ɵɵtext(47, "Misc");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(48, "div", 14);
        ɵngcc0.ɵɵelement(49, "c8y-misc-config", 19);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(50, "div", 21);
        ɵngcc0.ɵɵelementStart(51, "button", 22);
        ɵngcc0.ɵɵlistener("click", function CockpitConfigurationComponent_Template_button_click_51_listener() { return ctx.save(); });
        ɵngcc0.ɵɵtext(52, " Save ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r0 = ɵngcc0.ɵɵreference(6);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(2, 17, "App configuration"));
        ɵngcc0.ɵɵadvance(17);
        ɵngcc0.ɵɵpropertyInterpolate("popover", ɵngcc0.ɵɵpipeBind1(19, 19, "Define which are the enabled features in the current application."));
        ɵngcc0.ɵɵproperty("outsideClick", true);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yIcon", "question-circle-o");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("config", ctx.config);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵpropertyInterpolate("popover", ɵngcc0.ɵɵpipeBind1(29, 21, "Select which nodes to display in the top level of the navigator menu. By default, only Groups is shown."));
        ɵngcc0.ɵɵproperty("outsideClick", true);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yIcon", "question-circle-o");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("config", ctx.config);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵpropertyInterpolate("popover", ɵngcc0.ɵɵpipeBind1(40, 23, "The landing page for this application. By default it is a customizable dashboard displaying data for the overall tenant."));
        ɵngcc0.ɵɵproperty("outsideClick", true);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yIcon", "question-circle-o");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("config", ctx.config);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵproperty("config", ctx.config);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("disabled", !_r0.form.valid)("actionName", "cockpitConfigurationSaved")("actionData", ɵngcc0.ɵɵpureFunction1(25, _c0, ctx.config));
    } }, directives: [ɵngcc2.TitleComponent, ɵngcc3.ɵNgNoValidate, ɵngcc3.NgControlStatusGroup, ɵngcc3.NgForm, ɵngcc2.C8yTranslateDirective, ɵngcc4.PopoverDirective, ɵngcc2.IconDirective, ɵngcc5.FeatureConfigComponent, ɵngcc6.RootNodeConfigComponent, ɵngcc7.HomeDashboardConfigComponent, ɵngcc8.MiscConfigComponent, ɵngcc2.ProductExperienceDirective], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
CockpitConfigurationComponent.ctorParameters = () => [
    { type: CockpitConfigService },
    { type: AlertService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CockpitConfigurationComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-cockpit-configuration',
                template: "<c8y-title>{{ 'App configuration' | translate }}</c8y-title>\n\n<div class=\"row\">\n  <div class=\"col-lg-12 col-lg-max\">\n    <form #configForm=\"ngForm\">\n      <div class=\"card card--fullpage\">\n        <div class=\"card-header separator\">\n          <div class=\"card-title\" translate>Cockpit configuration</div>\n        </div>\n\n        <div class=\"inner-scroll\">\n          <div class=\"card-block p-t-0 p-b-0\">\n            <fieldset class=\"row separator-bottom p-t-24 p-b-24\">\n              <div class=\"col-xs-12 col-sm-3 col-md-2 text-left-xs text-right-sm\">\n                <div class=\"d-inline-flex m-b-16\">\n                  <h4 class=\"text-medium m-r-4\" translate>Features</h4>\n                  <button\n                    class=\"btn btn-clean\"\n                    popover=\"{{\n                      'Define which are the enabled features in the current application.'\n                        | translate\n                    }}\"\n                    container=\"body\"\n                    placement=\"right\"\n                    [outsideClick]=\"true\"\n                  >\n                    <i [c8yIcon]=\"'question-circle-o'\" class=\"text-info\"></i>\n                  </button>\n                </div>\n              </div>\n              <div class=\"col-xs-12 col-sm-9 col-md-8\">\n                <c8y-feature-config\n                  [config]=\"config\"\n                  (onUpdate)=\"updateFeatures()\"\n                ></c8y-feature-config>\n              </div>\n            </fieldset>\n            <fieldset class=\"row separator-bottom p-t-24 p-b-24\">\n              <div class=\"col-xs-12 col-sm-3 col-md-2 text-left-xs text-right-sm\">\n                <div class=\"d-inline-flex m-b-16\">\n                  <h4 class=\"text-medium m-r-8\" translate>Top level nodes</h4>\n                  <button\n                    class=\"btn btn-clean\"\n                    popover=\"{{\n                      'Select which nodes to display in the top level of the navigator menu. By default, only Groups is shown.'\n                        | translate\n                    }}\"\n                    container=\"body\"\n                    placement=\"right\"\n                    [outsideClick]=\"true\"\n                  >\n                    <i [c8yIcon]=\"'question-circle-o'\" class=\"text-info\"></i>\n                  </button>\n                </div>\n              </div>\n              <div class=\"col-xs-12 col-sm-10 col-md-10\">\n                <div class=\"row\">\n                  <c8y-root-node-config\n                    [config]=\"config\"\n                    (onUpdate)=\"updateRootNodes()\"\n                  ></c8y-root-node-config>\n                </div>\n              </div>\n            </fieldset>\n\n            <fieldset class=\"row separator-bottom p-t-24 p-b-24\">\n              <div class=\"col-xs-12 col-sm-3 col-md-2 text-left-xs text-right-sm\">\n                <div class=\"d-inline-flex m-b-16\">\n                  <h4 class=\"text-medium m-r-8\" translate>Home dashboard</h4>\n                  <button\n                    class=\"btn btn-clean\"\n                    popover=\"{{\n                      'The landing page for this application. By default it is a customizable dashboard displaying data for the overall tenant.'\n                        | translate\n                    }}\"\n                    placement=\"right\"\n                    container=\"body\"\n                    [outsideClick]=\"true\"\n                  >\n                    <i [c8yIcon]=\"'question-circle-o'\" class=\"text-info\"></i>\n                  </button>\n                </div>\n              </div>\n              <div class=\"col-xs-12 col-sm-9 col-md-8\">\n                <c8y-home-dashboard-config [config]=\"config\"></c8y-home-dashboard-config>\n              </div>\n            </fieldset>\n\n            <fieldset class=\"row separator-bottom p-t-24 p-b-24\">\n              <div class=\"col-xs-12 col-sm-3 col-md-2 text-left-xs text-right-sm\">\n                <h4 class=\"text-medium d-inline-block m-r-4\" translate>Misc</h4>\n              </div>\n              <div class=\"col-xs-12 col-sm-9 col-md-8\">\n                <c8y-misc-config [config]=\"config\"></c8y-misc-config>\n              </div>\n            </fieldset>\n          </div>\n        </div>\n        <div class=\"card-footer separator\">\n          <button\n            type=\"submit\"\n            class=\"btn btn-primary\"\n            translate\n            [disabled]=\"!configForm.form.valid\"\n            (click)=\"save()\"\n            c8yProductExperience\n            [actionName]=\"'cockpitConfigurationSaved'\"\n            [actionData]=\"{ config: config }\"\n          >\n            Save\n          </button>\n        </div>\n      </div>\n    </form>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.CockpitConfigService }, { type: ɵngcc2.AlertService }]; }, null); })();
