// @c8y/ngx-components/esm2015/cockpit-config/cockpit-config.module.js
import { NgModule } from '@angular/core';
import { CoreModule, gettext, HOOK_NAVIGATOR_NODES, HOOK_ROUTE, HOOK_STEPPER, Steppers } from '@c8y/ngx-components';
import { AssetSelectorModule } from '@c8y/ngx-components/assets-navigator';
import { EcosystemModule } from '@c8y/ngx-components/ecosystem';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { CockpitConfigGuard } from './cockpit-config.guard';
import { COCKPIT_CONFIG_PATH } from './cockpit-config.model';
import { CockpitConfigService } from './cockpit-config.service';
import { CockpitConfigurationComponent } from './cockpit-configuration.component';
import { FeatureConfigComponent } from './feature-config.component';
import { HomeDashboardConfigComponent } from './home-dashboard-config.component';
import { MiscConfigComponent } from './misc-config.component';
import { RootNodeConfigComponent } from './root-node-config.component';
import { CockpitSetupStep1Component } from './setup/cockpit-setup-step1.component';
import { CockpitSetupStep2Component } from './setup/cockpit-setup-step2.component';
import { CockpitSetupStep3Component } from './setup/cockpit-setup-step3.component';
import { CockpitSetupStep4Component } from './setup/cockpit-setup-step4.component';
import { CockpitSetupStepperButtonsComponent } from './setup/cockpit-setup-stepper-buttons.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './cockpit-config.service';
const ɵ0 = [
    {
        path: COCKPIT_CONFIG_PATH,
        component: CockpitConfigurationComponent,
        canActivate: [CockpitConfigGuard]
    }
], ɵ1 = [
    {
        stepperId: Steppers.SETUP,
        component: CockpitSetupStep1Component,
        label: gettext('Features'),
        setupId: 'cockpitFeatures',
        priority: 0
    },
    {
        stepperId: Steppers.SETUP,
        component: CockpitSetupStep2Component,
        label: gettext('Top level nodes'),
        setupId: 'cockpitTopLevelNodes',
        priority: 10
    },
    {
        stepperId: Steppers.SETUP,
        component: CockpitSetupStep3Component,
        label: gettext('Home Dashboard'),
        setupId: 'cockpitHomeDashboard',
        priority: 20
    },
    {
        stepperId: Steppers.SETUP,
        component: CockpitSetupStep4Component,
        label: gettext('Misc'),
        setupId: 'cockpitMisc',
        priority: 30
    }
];
export class CockpitConfigModule {
    constructor(service) {
        // only for DI, not used but needed.
    }
}
CockpitConfigModule.ɵfac = function CockpitConfigModule_Factory(t) { return new (t || CockpitConfigModule)(ɵngcc0.ɵɵinject(ɵngcc1.CockpitConfigService)); };
CockpitConfigModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: CockpitConfigModule });
CockpitConfigModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        CockpitConfigGuard,
        {
            provide: HOOK_ROUTE,
            useValue: ɵ0,
            multi: true
        },
        {
            provide: HOOK_NAVIGATOR_NODES,
            useClass: CockpitConfigGuard,
            multi: true
        },
        {
            provide: HOOK_STEPPER,
            useValue: ɵ1,
            multi: true
        }
    ], imports: [[CoreModule, AssetSelectorModule, PopoverModule, EcosystemModule]] });
CockpitConfigModule.ctorParameters = () => [
    { type: CockpitConfigService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CockpitConfigModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    CockpitConfigurationComponent,
                    CockpitSetupStep1Component,
                    CockpitSetupStep2Component,
                    CockpitSetupStep3Component,
                    CockpitSetupStep4Component,
                    FeatureConfigComponent,
                    RootNodeConfigComponent,
                    HomeDashboardConfigComponent,
                    MiscConfigComponent,
                    CockpitSetupStepperButtonsComponent
                ],
                imports: [CoreModule, AssetSelectorModule, PopoverModule, EcosystemModule],
                providers: [
                    CockpitConfigGuard,
                    {
                        provide: HOOK_ROUTE,
                        useValue: ɵ0,
                        multi: true
                    },
                    {
                        provide: HOOK_NAVIGATOR_NODES,
                        useClass: CockpitConfigGuard,
                        multi: true
                    },
                    {
                        provide: HOOK_STEPPER,
                        useValue: ɵ1,
                        multi: true
                    }
                ]
            }]
    }], function () { return [{ type: ɵngcc1.CockpitConfigService }]; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(CockpitConfigModule, { declarations: function () { return [CockpitConfigurationComponent,
        CockpitSetupStep1Component,
        CockpitSetupStep2Component,
        CockpitSetupStep3Component,
        CockpitSetupStep4Component,
        FeatureConfigComponent,
        RootNodeConfigComponent,
        HomeDashboardConfigComponent,
        MiscConfigComponent,
        CockpitSetupStepperButtonsComponent]; }, imports: function () { return [CoreModule, AssetSelectorModule, PopoverModule, EcosystemModule]; } }); })();
export { ɵ0, ɵ1 };
