// @c8y/ngx-components/esm2015/cockpit-config/feature-config.component.js
import { Component, EventEmitter, Input, Output } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from '@angular/forms';
export class FeatureConfigComponent {
    constructor() {
        this.onUpdate = new EventEmitter();
    }
    updateFeatures() {
        this.onUpdate.emit();
    }
}
FeatureConfigComponent.ɵfac = function FeatureConfigComponent_Factory(t) { return new (t || FeatureConfigComponent)(); };
FeatureConfigComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: FeatureConfigComponent, selectors: [["c8y-feature-config"]], inputs: { config: "config" }, outputs: { onUpdate: "onUpdate" }, decls: 141, vars: 70, consts: [["icon", "search"], [1, "d-flex", "a-i-center"], ["translate", ""], [1, "c8y-switch", "c8y-switch--inline", "m-l-auto", 3, "title"], ["type", "checkbox", "name", "search", 3, "ngModel", "ngModelChange", "change"], ["icon", "c8y-group"], ["type", "checkbox", "name", "groups", 3, "ngModel", "ngModelChange", "change"], ["icon", "bell"], ["type", "checkbox", "name", "alarms", 3, "ngModel", "ngModelChange", "change"], ["icon", "c8y-data-explorer"], ["type", "checkbox", "name", "dataExplorer", 3, "ngModel", "ngModelChange", "change"], ["icon", "c8y-reports"], ["type", "checkbox", "name", "reports", 3, "ngModel", "ngModelChange", "change"], ["icon", "graph-report"], ["type", "checkbox", "name", "exports", 3, "ngModel", "ngModelChange", "change"], ["icon", "c8y-data-points"], ["type", "checkbox", "name", "dataPointLibrary", 3, "ngModel", "ngModelChange", "change"], ["icon", "c8y-smart-rules"], ["type", "checkbox", "name", "globalSmartRules", 3, "ngModel", "ngModelChange", "change"], ["icon", "c8y-group-open"], ["type", "checkbox", "name", "subassets", 3, "ngModel", "ngModelChange", "change"], ["type", "checkbox", "name", "smartRules", 3, "ngModel", "ngModelChange", "change"]], template: function FeatureConfigComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-list-group");
        ɵngcc0.ɵɵelementStart(1, "c8y-li");
        ɵngcc0.ɵɵelement(2, "c8y-li-icon", 0);
        ɵngcc0.ɵɵelementStart(3, "div", 1);
        ɵngcc0.ɵɵelementStart(4, "div");
        ɵngcc0.ɵɵelementStart(5, "p");
        ɵngcc0.ɵɵtext(6);
        ɵngcc0.ɵɵpipe(7, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(8, "p");
        ɵngcc0.ɵɵelementStart(9, "small", 2);
        ɵngcc0.ɵɵtext(10, "Display the global search in the main header.");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "label", 3);
        ɵngcc0.ɵɵpipe(12, "translate");
        ɵngcc0.ɵɵelementStart(13, "input", 4);
        ɵngcc0.ɵɵlistener("ngModelChange", function FeatureConfigComponent_Template_input_ngModelChange_13_listener($event) { return ctx.config.features.search = $event; })("change", function FeatureConfigComponent_Template_input_change_13_listener() { return ctx.updateFeatures(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(14, "span");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(15, "c8y-li");
        ɵngcc0.ɵɵelement(16, "c8y-li-icon", 5);
        ɵngcc0.ɵɵelementStart(17, "div", 1);
        ɵngcc0.ɵɵelementStart(18, "div");
        ɵngcc0.ɵɵelementStart(19, "p");
        ɵngcc0.ɵɵtext(20);
        ɵngcc0.ɵɵpipe(21, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(22, "p");
        ɵngcc0.ɵɵelementStart(23, "small", 2);
        ɵngcc0.ɵɵtext(24, "Display top level groups under the Groups navigator menu.");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(25, "label", 3);
        ɵngcc0.ɵɵpipe(26, "translate");
        ɵngcc0.ɵɵelementStart(27, "input", 6);
        ɵngcc0.ɵɵlistener("ngModelChange", function FeatureConfigComponent_Template_input_ngModelChange_27_listener($event) { return ctx.config.features.groups = $event; })("change", function FeatureConfigComponent_Template_input_change_27_listener() { return ctx.updateFeatures(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(28, "span");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(29, "c8y-li");
        ɵngcc0.ɵɵelement(30, "c8y-li-icon", 7);
        ɵngcc0.ɵɵelementStart(31, "div", 1);
        ɵngcc0.ɵɵelementStart(32, "div");
        ɵngcc0.ɵɵelementStart(33, "p");
        ɵngcc0.ɵɵtext(34);
        ɵngcc0.ɵɵpipe(35, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(36, "p");
        ɵngcc0.ɵɵelementStart(37, "small", 2);
        ɵngcc0.ɵɵtext(38, "Display a link to the global alarms list in the navigator menu.");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(39, "label", 3);
        ɵngcc0.ɵɵpipe(40, "translate");
        ɵngcc0.ɵɵelementStart(41, "input", 8);
        ɵngcc0.ɵɵlistener("ngModelChange", function FeatureConfigComponent_Template_input_ngModelChange_41_listener($event) { return ctx.config.features.alarms = $event; })("change", function FeatureConfigComponent_Template_input_change_41_listener() { return ctx.updateFeatures(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(42, "span");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(43, "c8y-li");
        ɵngcc0.ɵɵelement(44, "c8y-li-icon", 9);
        ɵngcc0.ɵɵelementStart(45, "div", 1);
        ɵngcc0.ɵɵelementStart(46, "div");
        ɵngcc0.ɵɵelementStart(47, "p");
        ɵngcc0.ɵɵtext(48);
        ɵngcc0.ɵɵpipe(49, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(50, "p");
        ɵngcc0.ɵɵelementStart(51, "small", 2);
        ɵngcc0.ɵɵtext(52, " Display the data explorer in the navigator menu and on the group tabs. ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(53, "label", 3);
        ɵngcc0.ɵɵpipe(54, "translate");
        ɵngcc0.ɵɵelementStart(55, "input", 10);
        ɵngcc0.ɵɵlistener("ngModelChange", function FeatureConfigComponent_Template_input_ngModelChange_55_listener($event) { return ctx.config.features.dataExplorer = $event; })("change", function FeatureConfigComponent_Template_input_change_55_listener() { return ctx.updateFeatures(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(56, "span");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(57, "c8y-li");
        ɵngcc0.ɵɵelement(58, "c8y-li-icon", 11);
        ɵngcc0.ɵɵelementStart(59, "div", 1);
        ɵngcc0.ɵɵelementStart(60, "div");
        ɵngcc0.ɵɵelementStart(61, "p");
        ɵngcc0.ɵɵtext(62);
        ɵngcc0.ɵɵpipe(63, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(64, "p");
        ɵngcc0.ɵɵelementStart(65, "small", 2);
        ɵngcc0.ɵɵtext(66, "Display a link to the Reports list in the navigator menu.");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(67, "label", 3);
        ɵngcc0.ɵɵpipe(68, "translate");
        ɵngcc0.ɵɵelementStart(69, "input", 12);
        ɵngcc0.ɵɵlistener("ngModelChange", function FeatureConfigComponent_Template_input_ngModelChange_69_listener($event) { return ctx.config.features.reports = $event; })("change", function FeatureConfigComponent_Template_input_change_69_listener() { return ctx.updateFeatures(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(70, "span");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(71, "c8y-li");
        ɵngcc0.ɵɵelement(72, "c8y-li-icon", 13);
        ɵngcc0.ɵɵelementStart(73, "div", 1);
        ɵngcc0.ɵɵelementStart(74, "div");
        ɵngcc0.ɵɵelementStart(75, "p");
        ɵngcc0.ɵɵtext(76);
        ɵngcc0.ɵɵpipe(77, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(78, "p");
        ɵngcc0.ɵɵelementStart(79, "small", 2);
        ɵngcc0.ɵɵtext(80, " Display a link to the Exports list under the Configuration navigator menu. ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(81, "label", 3);
        ɵngcc0.ɵɵpipe(82, "translate");
        ɵngcc0.ɵɵelementStart(83, "input", 14);
        ɵngcc0.ɵɵlistener("ngModelChange", function FeatureConfigComponent_Template_input_ngModelChange_83_listener($event) { return ctx.config.features.exports = $event; })("change", function FeatureConfigComponent_Template_input_change_83_listener() { return ctx.updateFeatures(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(84, "span");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(85, "c8y-li");
        ɵngcc0.ɵɵelement(86, "c8y-li-icon", 15);
        ɵngcc0.ɵɵelementStart(87, "div", 1);
        ɵngcc0.ɵɵelementStart(88, "div");
        ɵngcc0.ɵɵelementStart(89, "p");
        ɵngcc0.ɵɵtext(90);
        ɵngcc0.ɵɵpipe(91, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(92, "p");
        ɵngcc0.ɵɵelementStart(93, "small", 2);
        ɵngcc0.ɵɵtext(94, " Display a link to the Data point library under the Configuration navigator menu. ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(95, "label", 3);
        ɵngcc0.ɵɵpipe(96, "translate");
        ɵngcc0.ɵɵelementStart(97, "input", 16);
        ɵngcc0.ɵɵlistener("ngModelChange", function FeatureConfigComponent_Template_input_ngModelChange_97_listener($event) { return ctx.config.features.dataPointLibrary = $event; })("change", function FeatureConfigComponent_Template_input_change_97_listener() { return ctx.updateFeatures(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(98, "span");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(99, "c8y-li");
        ɵngcc0.ɵɵelement(100, "c8y-li-icon", 17);
        ɵngcc0.ɵɵelementStart(101, "div", 1);
        ɵngcc0.ɵɵelementStart(102, "div");
        ɵngcc0.ɵɵelementStart(103, "p");
        ɵngcc0.ɵɵtext(104);
        ɵngcc0.ɵɵpipe(105, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(106, "p");
        ɵngcc0.ɵɵelementStart(107, "small", 2);
        ɵngcc0.ɵɵtext(108, " Display a link to the Global smart rules under the Configuration navigator menu. ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(109, "label", 3);
        ɵngcc0.ɵɵpipe(110, "translate");
        ɵngcc0.ɵɵelementStart(111, "input", 18);
        ɵngcc0.ɵɵlistener("ngModelChange", function FeatureConfigComponent_Template_input_ngModelChange_111_listener($event) { return ctx.config.features.globalSmartRules = $event; })("change", function FeatureConfigComponent_Template_input_change_111_listener() { return ctx.updateFeatures(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(112, "span");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(113, "c8y-li");
        ɵngcc0.ɵɵelement(114, "c8y-li-icon", 19);
        ɵngcc0.ɵɵelementStart(115, "div", 1);
        ɵngcc0.ɵɵelementStart(116, "div");
        ɵngcc0.ɵɵelementStart(117, "p");
        ɵngcc0.ɵɵtext(118);
        ɵngcc0.ɵɵpipe(119, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(120, "p");
        ɵngcc0.ɵɵelementStart(121, "small", 2);
        ɵngcc0.ɵɵtext(122, "Display the sub assets tab on groups.");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(123, "label", 3);
        ɵngcc0.ɵɵpipe(124, "translate");
        ɵngcc0.ɵɵelementStart(125, "input", 20);
        ɵngcc0.ɵɵlistener("ngModelChange", function FeatureConfigComponent_Template_input_ngModelChange_125_listener($event) { return ctx.config.features.subassets = $event; })("change", function FeatureConfigComponent_Template_input_change_125_listener() { return ctx.updateFeatures(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(126, "span");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(127, "c8y-li");
        ɵngcc0.ɵɵelement(128, "c8y-li-icon", 17);
        ɵngcc0.ɵɵelementStart(129, "div", 1);
        ɵngcc0.ɵɵelementStart(130, "div");
        ɵngcc0.ɵɵelementStart(131, "p");
        ɵngcc0.ɵɵtext(132);
        ɵngcc0.ɵɵpipe(133, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(134, "p");
        ɵngcc0.ɵɵelementStart(135, "small", 2);
        ɵngcc0.ɵɵtext(136, "Display the smart rules tab on groups and devices.");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(137, "label", 3);
        ɵngcc0.ɵɵpipe(138, "translate");
        ɵngcc0.ɵɵelementStart(139, "input", 21);
        ɵngcc0.ɵɵlistener("ngModelChange", function FeatureConfigComponent_Template_input_ngModelChange_139_listener($event) { return ctx.config.features.smartRules = $event; })("change", function FeatureConfigComponent_Template_input_change_139_listener() { return ctx.updateFeatures(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(140, "span");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(7, 30, "Global search"));
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(12, 32, "Global search"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngModel", ctx.config.features.search);
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(21, 34, "Groups"));
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(26, 36, "Groups"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngModel", ctx.config.features.groups);
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(35, 38, "Alarms"));
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(40, 40, "Global Alarms view"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngModel", ctx.config.features.alarms);
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(49, 42, "Data explorer"));
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(54, 44, "Data explorer"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngModel", ctx.config.features.dataExplorer);
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(63, 46, "Reports"));
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(68, 48, "Reports"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngModel", ctx.config.features.reports);
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(77, 50, "Exports"));
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(82, 52, "Exports"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngModel", ctx.config.features.exports);
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(91, 54, "Data point library"));
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(96, 56, "Data point library"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngModel", ctx.config.features.dataPointLibrary);
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(105, 58, "Global smart rules"));
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(110, 60, "Global smart rules"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngModel", ctx.config.features.globalSmartRules);
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(119, 62, "Sub assets view"));
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(124, 64, "Sub assets view"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngModel", ctx.config.features.subassets);
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(133, 66, "Smart rules for devices and groups"));
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(138, 68, "Smart rules for devices and groups"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngModel", ctx.config.features.smartRules);
    } }, directives: [ɵngcc1.ListGroupComponent, ɵngcc1.ListItemComponent, ɵngcc1.ListItemIconComponent, ɵngcc1.C8yTranslateDirective, ɵngcc2.CheckboxControlValueAccessor, ɵngcc2.NgControlStatus, ɵngcc2.NgModel], pipes: [ɵngcc1.C8yTranslatePipe], encapsulation: 2 });
FeatureConfigComponent.propDecorators = {
    config: [{ type: Input }],
    onUpdate: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(FeatureConfigComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-feature-config',
                template: "<c8y-list-group>\n  <c8y-li>\n    <c8y-li-icon icon=\"search\"></c8y-li-icon>\n    <div class=\"d-flex a-i-center\">\n      <div>\n        <p>{{ 'Global search' | translate }}</p>\n        <p>\n          <small translate>Display the global search in the main header.</small>\n        </p>\n      </div>\n      <label\n        class=\"c8y-switch c8y-switch--inline m-l-auto\"\n        title=\"{{ 'Global search' | translate }}\"\n      >\n        <input\n          type=\"checkbox\"\n          [(ngModel)]=\"config.features.search\"\n          name=\"search\"\n          (change)=\"updateFeatures()\"\n        />\n        <span></span>\n      </label>\n    </div>\n  </c8y-li>\n  <c8y-li>\n    <c8y-li-icon icon=\"c8y-group\"></c8y-li-icon>\n    <div class=\"d-flex a-i-center\">\n      <div>\n        <p>{{ 'Groups' | translate }}</p>\n        <p>\n          <small translate>Display top level groups under the Groups navigator menu.</small>\n        </p>\n      </div>\n      <label class=\"c8y-switch c8y-switch--inline m-l-auto\" title=\"{{ 'Groups' | translate }}\">\n        <input\n          type=\"checkbox\"\n          [(ngModel)]=\"config.features.groups\"\n          name=\"groups\"\n          (change)=\"updateFeatures()\"\n        />\n        <span></span>\n      </label>\n    </div>\n  </c8y-li>\n  <c8y-li>\n    <c8y-li-icon icon=\"bell\"></c8y-li-icon>\n    <div class=\"d-flex a-i-center\">\n      <div>\n        <p>{{ 'Alarms' | translate }}</p>\n        <p>\n          <small translate>Display a link to the global alarms list in the navigator menu.</small>\n        </p>\n      </div>\n      <label\n        class=\"c8y-switch c8y-switch--inline m-l-auto\"\n        title=\"{{ 'Global Alarms view' | translate }}\"\n      >\n        <input\n          type=\"checkbox\"\n          [(ngModel)]=\"config.features.alarms\"\n          name=\"alarms\"\n          (change)=\"updateFeatures()\"\n        />\n        <span></span>\n      </label>\n    </div>\n  </c8y-li>\n  <c8y-li>\n    <c8y-li-icon icon=\"c8y-data-explorer\"></c8y-li-icon>\n    <div class=\"d-flex a-i-center\">\n      <div>\n        <p>{{ 'Data explorer' | translate }}</p>\n        <p>\n          <small translate>\n            Display the data explorer in the navigator menu and on the group tabs.\n          </small>\n        </p>\n      </div>\n      <label\n        class=\"c8y-switch c8y-switch--inline m-l-auto\"\n        title=\"{{ 'Data explorer' | translate }}\"\n      >\n        <input\n          type=\"checkbox\"\n          [(ngModel)]=\"config.features.dataExplorer\"\n          name=\"dataExplorer\"\n          (change)=\"updateFeatures()\"\n        />\n        <span></span>\n      </label>\n    </div>\n  </c8y-li>\n  <c8y-li>\n    <c8y-li-icon icon=\"c8y-reports\"></c8y-li-icon>\n    <div class=\"d-flex a-i-center\">\n      <div>\n        <p>{{ 'Reports' | translate }}</p>\n        <p>\n          <small translate>Display a link to the Reports list in the navigator menu.</small>\n        </p>\n      </div>\n      <label class=\"c8y-switch c8y-switch--inline m-l-auto\" title=\"{{ 'Reports' | translate }}\">\n        <input\n          type=\"checkbox\"\n          [(ngModel)]=\"config.features.reports\"\n          name=\"reports\"\n          (change)=\"updateFeatures()\"\n        />\n        <span></span>\n      </label>\n    </div>\n  </c8y-li>\n  <c8y-li>\n    <c8y-li-icon icon=\"graph-report\"></c8y-li-icon>\n    <div class=\"d-flex a-i-center\">\n      <div>\n        <p>{{ 'Exports' | translate }}</p>\n        <p>\n          <small translate>\n            Display a link to the Exports list under the Configuration navigator menu.\n          </small>\n        </p>\n      </div>\n      <label class=\"c8y-switch c8y-switch--inline m-l-auto\" title=\"{{ 'Exports' | translate }}\">\n        <input\n          type=\"checkbox\"\n          [(ngModel)]=\"config.features.exports\"\n          name=\"exports\"\n          (change)=\"updateFeatures()\"\n        />\n        <span></span>\n      </label>\n    </div>\n  </c8y-li>\n  <c8y-li>\n    <c8y-li-icon icon=\"c8y-data-points\"></c8y-li-icon>\n    <div class=\"d-flex a-i-center\">\n      <div>\n        <p>{{ 'Data point library' | translate }}</p>\n        <p>\n          <small translate>\n            Display a link to the Data point library under the Configuration navigator menu.\n          </small>\n        </p>\n      </div>\n      <label\n        class=\"c8y-switch c8y-switch--inline m-l-auto\"\n        title=\"{{ 'Data point library' | translate }}\"\n      >\n        <input\n          type=\"checkbox\"\n          [(ngModel)]=\"config.features.dataPointLibrary\"\n          name=\"dataPointLibrary\"\n          (change)=\"updateFeatures()\"\n        />\n        <span></span>\n      </label>\n    </div>\n  </c8y-li>\n  <c8y-li>\n    <c8y-li-icon icon=\"c8y-smart-rules\"></c8y-li-icon>\n    <div class=\"d-flex a-i-center\">\n      <div>\n        <p>{{ 'Global smart rules' | translate }}</p>\n        <p>\n          <small translate>\n            Display a link to the Global smart rules under the Configuration navigator menu.\n          </small>\n        </p>\n      </div>\n      <label\n        class=\"c8y-switch c8y-switch--inline m-l-auto\"\n        title=\"{{ 'Global smart rules' | translate }}\"\n      >\n        <input\n          type=\"checkbox\"\n          [(ngModel)]=\"config.features.globalSmartRules\"\n          name=\"globalSmartRules\"\n          (change)=\"updateFeatures()\"\n        />\n        <span></span>\n      </label>\n    </div>\n  </c8y-li>\n  <c8y-li>\n    <c8y-li-icon icon=\"c8y-group-open\"></c8y-li-icon>\n    <div class=\"d-flex a-i-center\">\n      <div>\n        <p>{{ 'Sub assets view' | translate }}</p>\n        <p><small translate>Display the sub assets tab on groups.</small></p>\n      </div>\n      <label\n        class=\"c8y-switch c8y-switch--inline m-l-auto\"\n        title=\"{{ 'Sub assets view' | translate }}\"\n      >\n        <input\n          type=\"checkbox\"\n          [(ngModel)]=\"config.features.subassets\"\n          name=\"subassets\"\n          (change)=\"updateFeatures()\"\n        />\n        <span></span>\n      </label>\n    </div>\n  </c8y-li>\n  <c8y-li>\n    <c8y-li-icon icon=\"c8y-smart-rules\"></c8y-li-icon>\n    <div class=\"d-flex a-i-center\">\n      <div>\n        <p>{{ 'Smart rules for devices and groups' | translate }}</p>\n        <p>\n          <small translate>Display the smart rules tab on groups and devices.</small>\n        </p>\n      </div>\n      <label\n        class=\"c8y-switch c8y-switch--inline m-l-auto\"\n        title=\"{{ 'Smart rules for devices and groups' | translate }}\"\n      >\n        <input\n          type=\"checkbox\"\n          [(ngModel)]=\"config.features.smartRules\"\n          name=\"smartRules\"\n          (change)=\"updateFeatures()\"\n        />\n        <span></span>\n      </label>\n    </div>\n  </c8y-li>\n</c8y-list-group>\n"
            }]
    }], function () { return []; }, { onUpdate: [{
            type: Output
        }], config: [{
            type: Input
        }] }); })();
