// @c8y/ngx-components/esm2015/cockpit-config/home-dashboard-config.component.js
import { Component, Input } from '@angular/core';
import { DEFAULT_HOME_DASHBOARD_NAME, HomeDashboardType, USER_HOME_DASHBOARD_NAME } from './cockpit-config.model';
import { CockpitConfigService } from './cockpit-config.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './cockpit-config.service';
import * as ɵngcc2 from '@c8y/ngx-components';
export class HomeDashboardConfigComponent {
    /**
     * @ignore
     */
    constructor(cockpitConfigService) {
        this.cockpitConfigService = cockpitConfigService;
        /**
         * The types of dashboard that can be configured.
         */
        this.homeDashboardTypes = HomeDashboardType;
    }
    /**
     * @ignore
     */
    dashboardChange(selected, type) {
        if (!selected) {
            return;
        }
        switch (type) {
            case this.homeDashboardTypes.DEFAULT: {
                this.config.homeDashboardName = DEFAULT_HOME_DASHBOARD_NAME;
                this.config.userSpecificHomeDashboard = false;
                break;
            }
            case this.homeDashboardTypes.APP: {
                this.config.homeDashboardName = this.cockpitConfigService.getAppDashboardName();
                this.config.userSpecificHomeDashboard = false;
                break;
            }
            case this.homeDashboardTypes.USER: {
                this.config.homeDashboardName = USER_HOME_DASHBOARD_NAME;
                this.config.userSpecificHomeDashboard = true;
                break;
            }
        }
    }
    /**
     * @ignore
     */
    verifySelected(type) {
        if (type === this.homeDashboardTypes.USER) {
            return this.config.userSpecificHomeDashboard;
        }
        if (type === this.homeDashboardTypes.DEFAULT) {
            return this.config.homeDashboardName === DEFAULT_HOME_DASHBOARD_NAME;
        }
        return this.config.homeDashboardName === this.cockpitConfigService.getAppDashboardName();
    }
}
HomeDashboardConfigComponent.ɵfac = function HomeDashboardConfigComponent_Factory(t) { return new (t || HomeDashboardConfigComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.CockpitConfigService)); };
HomeDashboardConfigComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: HomeDashboardConfigComponent, selectors: [["c8y-home-dashboard-config"]], inputs: { config: "config" }, decls: 19, vars: 3, consts: [[1, "separator-top"], [3, "selected", "onSelect"], ["translate", ""]], template: function HomeDashboardConfigComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-list-group", 0);
        ɵngcc0.ɵɵelementStart(1, "c8y-li");
        ɵngcc0.ɵɵelementStart(2, "c8y-li-radio", 1);
        ɵngcc0.ɵɵlistener("onSelect", function HomeDashboardConfigComponent_Template_c8y_li_radio_onSelect_2_listener($event) { return ctx.dashboardChange($event, ctx.homeDashboardTypes.DEFAULT); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "p", 2);
        ɵngcc0.ɵɵtext(4, "Default home dashboard");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "small", 2);
        ɵngcc0.ɵɵtext(6, "Changes done in the home dashboard are reflected across the platform.");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(7, "c8y-li");
        ɵngcc0.ɵɵelementStart(8, "c8y-li-radio", 1);
        ɵngcc0.ɵɵlistener("onSelect", function HomeDashboardConfigComponent_Template_c8y_li_radio_onSelect_8_listener($event) { return ctx.dashboardChange($event, ctx.homeDashboardTypes.APP); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "p", 2);
        ɵngcc0.ɵɵtext(10, "Custom home dashboard");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "small", 2);
        ɵngcc0.ɵɵtext(12, "Changes done to the home dashboard are reflected only in the current application.");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(13, "c8y-li");
        ɵngcc0.ɵɵelementStart(14, "c8y-li-radio", 1);
        ɵngcc0.ɵɵlistener("onSelect", function HomeDashboardConfigComponent_Template_c8y_li_radio_onSelect_14_listener($event) { return ctx.dashboardChange($event, ctx.homeDashboardTypes.USER); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(15, "p", 2);
        ɵngcc0.ɵɵtext(16, "User home dashboard");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(17, "small", 2);
        ɵngcc0.ɵɵtext(18, "Changes done to the home dashboard are reflected only for the current user. NOTE: This user needs to have inventory write permission.");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("selected", ctx.verifySelected(ctx.homeDashboardTypes.DEFAULT));
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵproperty("selected", ctx.verifySelected(ctx.homeDashboardTypes.APP));
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵproperty("selected", ctx.verifySelected(ctx.homeDashboardTypes.USER));
    } }, directives: [ɵngcc2.ListGroupComponent, ɵngcc2.ListItemComponent, ɵngcc2.ListItemRadioComponent, ɵngcc2.C8yTranslateDirective], encapsulation: 2 });
HomeDashboardConfigComponent.ctorParameters = () => [
    { type: CockpitConfigService }
];
HomeDashboardConfigComponent.propDecorators = {
    config: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(HomeDashboardConfigComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-home-dashboard-config',
                template: "<c8y-list-group class=\"separator-top\">\n  <c8y-li>\n    <c8y-li-radio\n      (onSelect)=\"dashboardChange($event, homeDashboardTypes.DEFAULT)\"\n      [selected]=\"verifySelected(homeDashboardTypes.DEFAULT)\"\n    >\n    </c8y-li-radio>\n    <p translate>Default home dashboard</p>\n    <small translate>Changes done in the home dashboard are reflected across the platform.</small>\n  </c8y-li>\n  <c8y-li>\n    <c8y-li-radio\n      (onSelect)=\"dashboardChange($event, homeDashboardTypes.APP)\"\n      [selected]=\"verifySelected(homeDashboardTypes.APP)\"\n    >\n    </c8y-li-radio>\n    <p translate>Custom home dashboard</p>\n    <small translate\n      >Changes done to the home dashboard are reflected only in the current application.</small\n    >\n  </c8y-li>\n  <c8y-li>\n    <c8y-li-radio\n      (onSelect)=\"dashboardChange($event, homeDashboardTypes.USER)\"\n      [selected]=\"verifySelected(homeDashboardTypes.USER)\"\n    >\n    </c8y-li-radio>\n    <p translate>User home dashboard</p>\n    <small translate\n      >Changes done to the home dashboard are reflected only for the current user. NOTE: This user\n      needs to have inventory write permission.</small\n    >\n  </c8y-li>\n</c8y-list-group>\n"
            }]
    }], function () { return [{ type: ɵngcc1.CockpitConfigService }]; }, { config: [{
            type: Input
        }] }); })();
