// @c8y/ngx-components/esm2015/cockpit-config/misc-config.component.js
import { Component, Input } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from '@angular/forms';
export class MiscConfigComponent {
}
MiscConfigComponent.ɵfac = function MiscConfigComponent_Factory(t) { return new (t || MiscConfigComponent)(); };
MiscConfigComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: MiscConfigComponent, selectors: [["c8y-misc-config"]], inputs: { config: "config" }, decls: 10, vars: 7, consts: [[1, "d-flex", "a-i-center"], [1, "c8y-switch", "c8y-switch--inline", "m-l-auto", 3, "title"], ["type", "checkbox", "name", "hideNavigator", 3, "ngModel", "ngModelChange"]], template: function MiscConfigComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-list-group");
        ɵngcc0.ɵɵelementStart(1, "c8y-li");
        ɵngcc0.ɵɵelementStart(2, "div", 0);
        ɵngcc0.ɵɵelementStart(3, "p");
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "label", 1);
        ɵngcc0.ɵɵpipe(7, "translate");
        ɵngcc0.ɵɵelementStart(8, "input", 2);
        ɵngcc0.ɵɵlistener("ngModelChange", function MiscConfigComponent_Template_input_ngModelChange_8_listener($event) { return ctx.config.hideNavigator = $event; });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(9, "span");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(5, 3, "Always collapse navigator on start up"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(7, 5, "Collapse navigator on start up"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngModel", ctx.config.hideNavigator);
    } }, directives: [ɵngcc1.ListGroupComponent, ɵngcc1.ListItemComponent, ɵngcc2.CheckboxControlValueAccessor, ɵngcc2.NgControlStatus, ɵngcc2.NgModel], pipes: [ɵngcc1.C8yTranslatePipe], encapsulation: 2 });
MiscConfigComponent.propDecorators = {
    config: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(MiscConfigComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-misc-config',
                template: "<c8y-list-group>\n  <c8y-li>\n    <div class=\"d-flex a-i-center\">\n      <p>{{ 'Always collapse navigator on start up' | translate }}</p>\n      <label\n        class=\"c8y-switch c8y-switch--inline m-l-auto\"\n        title=\"{{ 'Collapse navigator on start up' | translate }}\"\n      >\n        <input type=\"checkbox\" [(ngModel)]=\"config.hideNavigator\" name=\"hideNavigator\" />\n        <span></span>\n      </label>\n    </div>\n  </c8y-li>\n</c8y-list-group>\n"
            }]
    }], null, { config: [{
            type: Input
        }] }); })();
