// @c8y/ngx-components/esm2015/repository/software/device-tab/software-repository-device-tab.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CoreModule, gettext, HOOK_ROUTE, ViewContext } from '@c8y/ngx-components';
import { OperationDetailsModule } from '@c8y/ngx-components/operations/operation-details';
import { SharedRepositoryModule } from '@c8y/ngx-components/repository/shared';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { DeviceSoftwareChangesComponent } from './device-software-changes.component';
import { DeviceSoftwareListComponent } from './device-software-list.component';
import { InstalledSoftwareComponent } from './installed-software.component';
import { SoftwareDeviceTabComponent } from './software-device-tab.component';
import { SoftwareDeviceTabGuard } from './software-device-tab.guard';
export class SoftwareRepositoryDeviceTabModule {
    static forRoot() {
        return {
            ngModule: SoftwareRepositoryDeviceTabModule,
            providers: [
                SoftwareDeviceTabGuard,
                {
                    provide: HOOK_ROUTE,
                    useValue: [
                        {
                            context: ViewContext.Device,
                            path: 'software',
                            component: SoftwareDeviceTabComponent,
                            label: gettext('Software'),
                            icon: 'c8y-tools',
                            priority: 499,
                            canActivate: [SoftwareDeviceTabGuard]
                        }
                    ],
                    multi: true
                }
            ]
        };
    }
}
SoftwareRepositoryDeviceTabModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    CoreModule,
                    RouterModule,
                    SharedRepositoryModule,
                    OperationDetailsModule,
                    TooltipModule
                ],
                declarations: [
                    SoftwareDeviceTabComponent,
                    InstalledSoftwareComponent,
                    DeviceSoftwareListComponent,
                    DeviceSoftwareChangesComponent
                ],
                entryComponents: [
                    SoftwareDeviceTabComponent,
                    DeviceSoftwareListComponent,
                    DeviceSoftwareChangesComponent
                ]
            },] }
];
