// @c8y/ngx-components/bundles/c8y-ngx-components-operations-create-bulk-operation-details.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/forms'),require('@c8y/ngx-components'),require('@c8y/ngx-components/operations/bulk-operation-scheduler'),exports, require('@angular/core'), require('@angular/forms'), require('@c8y/ngx-components'), require('lodash-es'), require('@c8y/ngx-components/operations/bulk-operations-service'), require('@c8y/ngx-components/operations/bulk-operation-scheduler')) :
    typeof define === 'function' && define.amd ? define('@c8y/ngx-components/operations/create-bulk-operation-details', ['@angular/core','@angular/forms','@c8y/ngx-components','@c8y/ngx-components/operations/bulk-operation-scheduler','exports', '@angular/core', '@angular/forms', '@c8y/ngx-components', 'lodash-es', '@c8y/ngx-components/operations/bulk-operations-service', '@c8y/ngx-components/operations/bulk-operation-scheduler'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ng.forms,global.c8y.ngxComponents,global.c8y.ngxComponents.operations.bulkOperationScheduler,(global.c8y = global.c8y || {}, global.c8y["ngx-components"] = global.c8y["ngx-components"] || {}, global.c8y["ngx-components"].operations = global.c8y["ngx-components"].operations || {}, global.c8y["ngx-components"].operations["create-bulk-operation-details"] = {}), global.ng.core, global.ng.forms, global.c8y["ngx-components"], global.lodashEs, global.c8y["ngx-components"].operations["bulk-operations-service"], global.c8y["ngx-components"].operations["bulk-operation-scheduler"]));
})(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,exports, core, forms, ngxComponents, lodashEs, bulkOperationsService, bulkOperationScheduler) { 'use strict';

    var CreateBulkOperationDetailsComponent = /** @class */ (function () {
        function CreateBulkOperationDetailsComponent(formBuilder) {
            var _a;
            this.formBuilder = formBuilder;
            this.titlePlaceholder = ngxComponents.gettext('e.g. Reset factory settings');
            this.descriptionPlaceholder = ngxComponents.gettext('e.g. Reset factory settings of all devices of model XYZ');
            this.placeholders = (_a = {},
                _a[bulkOperationsService.BulkOperationType.SOFTWARE] = {
                    title: ngxComponents.gettext('e.g. Cloud connectivity software'),
                    description: ngxComponents.gettext('e.g. Cloud connectivity software applied to devices with type c8y_Linux')
                },
                _a[bulkOperationsService.BulkOperationType.FIRMWARE] = {
                    title: ngxComponents.gettext('e.g. Firmware for hardware revision b'),
                    description: ngxComponents.gettext('e.g. Firmware for hardware revision b applied to devices with type c8y_Linux')
                },
                _a[bulkOperationsService.BulkOperationType.CONFIGURATION] = {
                    title: ngxComponents.gettext('e.g. Host configuration c8y_Linux'),
                    description: ngxComponents.gettext('e.g. Host configuration applied to devices with type c8y_Linux')
                },
                _a[bulkOperationsService.BulkOperationType.DEVICE_PROFILE] = {
                    title: ngxComponents.gettext('e.g. Device profile for c8y_Linux devices'),
                    description: ngxComponents.gettext('e.g. Device profile applied to devices with type c8y_Linux')
                },
                _a);
        }
        Object.defineProperty(CreateBulkOperationDetailsComponent.prototype, "bulkOperationType", {
            set: function (bulkOperationType) {
                this.titlePlaceholder = lodashEs.get(this.placeholders[bulkOperationType], 'title');
                this.descriptionPlaceholder = lodashEs.get(this.placeholders[bulkOperationType], 'description');
            },
            enumerable: false,
            configurable: true
        });
        CreateBulkOperationDetailsComponent.prototype.ngOnInit = function () {
            var config = {
                description: ['', forms.Validators.required],
                note: [''],
                schedule: []
            };
            this.fgOperationDescription = this.formBuilder.group(config);
        };
CreateBulkOperationDetailsComponent.ɵfac = function CreateBulkOperationDetailsComponent_Factory(t) { return new (t || CreateBulkOperationDetailsComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.FormBuilder)); };
CreateBulkOperationDetailsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: CreateBulkOperationDetailsComponent, selectors: [["c8y-create-bulk-operation-details"]], inputs: { bulkOperationType: "bulkOperationType" }, decls: 12, vars: 7, consts: [[3, "formGroup"], ["translate", ""], ["formControlName", "description", "required", "", 1, "form-control", 3, "placeholder"], ["formControlName", "note", "rows", "3", 1, "form-control", 3, "placeholder"], ["formControlName", "schedule"]], template: function CreateBulkOperationDetailsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(2, "label", 1);
        ɵngcc0.ɵɵtext(3, "Title");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(4, "input", 2);
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(7, "label", 1);
        ɵngcc0.ɵɵtext(8, "Description");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(9, "textarea", 3);
        ɵngcc0.ɵɵpipe(10, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(11, "c8y-operation-scheduler", 4);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("formGroup", ctx.fgOperationDescription);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("placeholder", ɵngcc0.ɵɵpipeBind1(5, 3, ctx.titlePlaceholder));
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("placeholder", ɵngcc0.ɵɵpipeBind1(10, 5, ctx.descriptionPlaceholder));
    } }, directives: [ɵngcc1.NgControlStatusGroup, ɵngcc1.FormGroupDirective, ɵngcc2.FormGroupComponent, ɵngcc2.C8yTranslateDirective, ɵngcc1.DefaultValueAccessor, ɵngcc2.RequiredInputPlaceholderDirective, ɵngcc1.NgControlStatus, ɵngcc1.FormControlName, ɵngcc1.RequiredValidator, ɵngcc3.OperationSchedulerComponent], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CreateBulkOperationDetailsComponent, [{
        type: core.Component,
        args: [{
                selector: 'c8y-create-bulk-operation-details',
                template: "<div [formGroup]=\"fgOperationDescription\">\n  <c8y-form-group>\n    <label translate>Title</label>\n    <input\n      formControlName=\"description\"\n      class=\"form-control\"\n      [placeholder]=\"titlePlaceholder | translate\"\n      required\n    />\n  </c8y-form-group>\n  <c8y-form-group>\n    <label translate>Description</label>\n    <textarea\n      formControlName=\"note\"\n      class=\"form-control\"\n      [placeholder]=\"descriptionPlaceholder | translate\"\n      rows=\"3\"\n    ></textarea>\n  </c8y-form-group>\n  <c8y-operation-scheduler formControlName=\"schedule\"></c8y-operation-scheduler>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.FormBuilder }]; }, { bulkOperationType: [{
            type: core.Input
        }] }); })();
        return CreateBulkOperationDetailsComponent;
    }());
    CreateBulkOperationDetailsComponent.ctorParameters = function () { return [
        { type: forms.FormBuilder }
    ]; };
    CreateBulkOperationDetailsComponent.propDecorators = {
        bulkOperationType: [{ type: core.Input }]
    };

    /**
     * This module provides a component for entering operation details.
     */
    var CreateBulkOperationDetailsModule = /** @class */ (function () {
        function CreateBulkOperationDetailsModule() {
        }
CreateBulkOperationDetailsModule.ɵfac = function CreateBulkOperationDetailsModule_Factory(t) { return new (t || CreateBulkOperationDetailsModule)(); };
CreateBulkOperationDetailsModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: CreateBulkOperationDetailsModule });
CreateBulkOperationDetailsModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[ngxComponents.CoreModule, ngxComponents.FormsModule, forms.ReactiveFormsModule, bulkOperationScheduler.BulkOperationSchedulerModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CreateBulkOperationDetailsModule, [{
        type: core.NgModule,
        args: [{
                imports: [ngxComponents.CoreModule, ngxComponents.FormsModule, forms.ReactiveFormsModule, bulkOperationScheduler.BulkOperationSchedulerModule],
                declarations: [CreateBulkOperationDetailsComponent],
                exports: [CreateBulkOperationDetailsComponent]
            }]
    }], function () { return []; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(CreateBulkOperationDetailsModule, { declarations: [CreateBulkOperationDetailsComponent], imports: [ɵngcc2.CoreModule, ɵngcc2.FormsModule, ɵngcc1.ReactiveFormsModule, ɵngcc3.BulkOperationSchedulerModule], exports: [CreateBulkOperationDetailsComponent] }); })();
        return CreateBulkOperationDetailsModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.CreateBulkOperationDetailsComponent = CreateBulkOperationDetailsComponent;
    exports.CreateBulkOperationDetailsModule = CreateBulkOperationDetailsModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
