// @c8y/ngx-components/bundles/c8y-ngx-components-operations-stepper-bulk-type-firmware.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@c8y/ngx-components/operations/bulk-operations-service'),require('@ngx-translate/core'),require('@c8y/ngx-components/operations/bulk-operation-stepper'),require('@c8y/ngx-components/operations/stepper-frames'),require('@c8y/ngx-components'),require('@c8y/ngx-components/repository/shared'),require('@c8y/client'),require('@angular/common'),require('@angular/forms'),exports, require('@angular/core'), require('@c8y/ngx-components'), require('@ngx-translate/core'), require('lodash-es'), require('rxjs'), require('@c8y/ngx-components/operations/bulk-operation-stepper'), require('@c8y/ngx-components/operations/bulk-operations-service'), require('@c8y/client'), require('@c8y/ngx-components/repository/shared'), require('rxjs/operators'), require('@angular/forms'), require('@c8y/ngx-components/operations/stepper-frames')) :
    typeof define === 'function' && define.amd ? define('@c8y/ngx-components/operations/stepper-bulk-type-firmware', ['@angular/core','@c8y/ngx-components/operations/bulk-operations-service','@ngx-translate/core','@c8y/ngx-components/operations/bulk-operation-stepper','@c8y/ngx-components/operations/stepper-frames','@c8y/ngx-components','@c8y/ngx-components/repository/shared','@c8y/client','@angular/common','@angular/forms','exports', '@angular/core', '@c8y/ngx-components', '@ngx-translate/core', 'lodash-es', 'rxjs', '@c8y/ngx-components/operations/bulk-operation-stepper', '@c8y/ngx-components/operations/bulk-operations-service', '@c8y/client', '@c8y/ngx-components/repository/shared', 'rxjs/operators', '@angular/forms', '@c8y/ngx-components/operations/stepper-frames'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.c8y.ngxComponents.operations.bulkOperationsService,global.ngxTranslate.core,global.c8y.ngxComponents.operations.bulkOperationStepper,global.c8y.ngxComponents.operations.stepperFrames,global.c8y.ngxComponents,global.c8y.ngxComponents.repository.shared,global.c8y.client,global.ng.common,global.ng.forms,(global.c8y = global.c8y || {}, global.c8y["ngx-components"] = global.c8y["ngx-components"] || {}, global.c8y["ngx-components"].operations = global.c8y["ngx-components"].operations || {}, global.c8y["ngx-components"].operations["stepper-bulk-type-firmware"] = {}), global.ng.core, global.c8y["ngx-components"], global.core$1, global.lodashEs, global.rxjs, global.c8y["ngx-components"].operations["bulk-operation-stepper"], global.c8y["ngx-components"].operations["bulk-operations-service"], global.client, global.c8y["ngx-components"].repository.shared, global.rxjs.operators, global.ng.forms, global.c8y["ngx-components"].operations["stepper-frames"]));
})(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,ɵngcc4,ɵngcc5,ɵngcc6,ɵngcc7,ɵngcc8,ɵngcc9,exports, core, ngxComponents, core$1, lodashEs, rxjs, bulkOperationStepper, bulkOperationsService, client, shared, operators, forms, stepperFrames) { 
function StepperBulkTypeFirmwareComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    var _r3 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "c8y-select-firmware-step", 2);
    ɵngcc0.ɵɵlistener("firmware", function StepperBulkTypeFirmwareComponent_ng_container_1_Template_c8y_select_firmware_step_firmware_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r3); var ctx_r2 = ɵngcc0.ɵɵnextContext(); return ctx_r2.onFirmwareSelected($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("selected", ctx_r0.stepData == null ? null : ctx_r0.stepData.firmware);
} }
function StepperBulkTypeFirmwareComponent_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    var _r5 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "c8y-preview-selection-frame", 3);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementStart(3, "c8y-version-or-patch-step", 4);
    ɵngcc0.ɵɵlistener("versionOrPatch", function StepperBulkTypeFirmwareComponent_ng_container_3_Template_c8y_version_or_patch_step_versionOrPatch_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r5); var ctx_r4 = ɵngcc0.ɵɵnextContext(); return ctx_r4.onVersionOrPatchSelected($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("header", ɵngcc0.ɵɵpipeBind1(2, 4, "Select version or patch"))("name", ctx_r1.stepData == null ? null : ctx_r1.stepData.firmware == null ? null : ctx_r1.stepData.firmware.name)("description", ctx_r1.stepData == null ? null : ctx_r1.stepData.firmware == null ? null : ctx_r1.stepData.firmware.description);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("selected", ctx_r1.stepData == null ? null : ctx_r1.stepData.versionOrPatch);
} }
function VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 9);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Legacy"), " ");
} }
function VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_10_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵelementStart(1, "span", 10);
    ɵngcc0.ɵɵelement(2, "i", 11);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_10_span_3_ng_template_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 16);
    ɵngcc0.ɵɵelementStart(1, "span", 17);
    ɵngcc0.ɵɵtext(2, "No patches");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_10_span_3_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 10);
    ɵngcc0.ɵɵelementStart(1, "span", 17);
    ɵngcc0.ɵɵtext(2, "1 patch");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
var _c0 = function (a0) { return { count: a0 }; };
function VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_10_span_3_ng_template_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 10);
    ɵngcc0.ɵɵelementStart(1, "span", 18);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵtext(3, "{{ count }} patches");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var baseVersion_r2 = ɵngcc0.ɵɵnextContext(3).$implicit;
    var ctx_r11 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction1(3, _c0, ɵngcc0.ɵɵpipeBind1(2, 1, ctx_r11.getPatchVersionsCount$(baseVersion_r2))));
} }
function VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_10_span_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵelementStart(1, "span", 12);
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵtemplate(3, VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_10_span_3_ng_template_3_Template, 3, 0, "ng-template", 13);
    ɵngcc0.ɵɵtemplate(4, VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_10_span_3_ng_template_4_Template, 3, 0, "ng-template", 14);
    ɵngcc0.ɵɵtemplate(5, VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_10_span_3_ng_template_5_Template, 4, 5, "ng-template", 15);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var baseVersion_r2 = ɵngcc0.ɵɵnextContext(2).$implicit;
    var ctx_r8 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngPlural", ɵngcc0.ɵɵpipeBind1(2, 1, ctx_r8.getPatchVersionsCount$(baseVersion_r2)));
} }
function VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtemplate(1, VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_10_span_1_Template, 3, 0, "span", 8);
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵtemplate(3, VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_10_span_3_Template, 6, 3, "span", 8);
    ɵngcc0.ɵɵpipe(4, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var baseVersion_r2 = ɵngcc0.ɵɵnextContext().$implicit;
    var ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(2, 2, ctx_r5.getPatchVersionsCount$(baseVersion_r2)) === null);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(4, 4, ctx_r5.getPatchVersionsCount$(baseVersion_r2)) !== null);
} }
function VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_c8y_li_14_span_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 26);
    ɵngcc0.ɵɵtext(1, "patch");
    ɵngcc0.ɵɵelementEnd();
} }
function VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_c8y_li_14_ng_template_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 27);
    ɵngcc0.ɵɵtext(1, "version");
    ɵngcc0.ɵɵelementEnd();
} }
function VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_c8y_li_14_Template(rf, ctx) { if (rf & 1) {
    var _r21 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li");
    ɵngcc0.ɵɵelementStart(1, "c8y-li-radio", 19);
    ɵngcc0.ɵɵlistener("onSelect", function VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_c8y_li_14_Template_c8y_li_radio_onSelect_1_listener($event) { var restoredCtx = ɵngcc0.ɵɵrestoreView(_r21); var patchVersion_r15 = restoredCtx.$implicit; var ctx_r20 = ɵngcc0.ɵɵnextContext(3); return ctx_r20.selectVersionOrPatch($event, patchVersion_r15); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(2, "c8y-li-icon");
    ɵngcc0.ɵɵelement(3, "i", 3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "c8y-li-body", 4);
    ɵngcc0.ɵɵelementStart(5, "div", 20);
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "div", 20);
    ɵngcc0.ɵɵtemplate(8, VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_c8y_li_14_span_8_Template, 2, 0, "span", 21);
    ɵngcc0.ɵɵtemplate(9, VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_c8y_li_14_ng_template_9_Template, 2, 0, "ng-template", null, 22, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(11, "div", 23);
    ɵngcc0.ɵɵelementStart(12, "span", 24);
    ɵngcc0.ɵɵtext(13, " File ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(14, "span", 25);
    ɵngcc0.ɵɵpipe(15, "async");
    ɵngcc0.ɵɵtext(16);
    ɵngcc0.ɵɵpipe(17, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var patchVersion_r15 = ctx.$implicit;
    var _r18 = ɵngcc0.ɵɵreference(10);
    var ctx_r6 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("selected", patchVersion_r15 === ctx_r6.selected);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate1(" ", patchVersion_r15.c8y_Firmware.version, " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", patchVersion_r15.c8y_Patch)("ngIfElse", _r18);
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(15, 6, ctx_r6.getBinaryName$(patchVersion_r15.c8y_Firmware.url)));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(17, 8, ctx_r6.getBinaryName$(patchVersion_r15.c8y_Firmware.url)), " ");
} }
function VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li");
    ɵngcc0.ɵɵelementStart(1, "c8y-li-icon");
    ɵngcc0.ɵɵelement(2, "i", 3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "c8y-li-body", 4);
    ɵngcc0.ɵɵelementStart(4, "div", 5);
    ɵngcc0.ɵɵelementStart(5, "p");
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "div", 6);
    ɵngcc0.ɵɵtemplate(8, VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_8_Template, 3, 3, "span", 7);
    ɵngcc0.ɵɵpipe(9, "async");
    ɵngcc0.ɵɵtemplate(10, VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_span_10_Template, 5, 6, "span", 8);
    ɵngcc0.ɵɵpipe(11, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(12, "c8y-li-collapse");
    ɵngcc0.ɵɵelementStart(13, "c8y-list-group");
    ɵngcc0.ɵɵtemplate(14, VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_c8y_li_14_Template, 18, 10, "c8y-li", 2);
    ɵngcc0.ɵɵpipe(15, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var baseVersion_r2 = ctx.$implicit;
    var ctx_r1 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵtextInterpolate(baseVersion_r2.c8y_Firmware.version);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(9, 5, ctx_r1.isLegacy$));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !ɵngcc0.ɵɵpipeBind1(11, 7, ctx_r1.isLegacy$));
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("c8yForOf", ɵngcc0.ɵɵpipeBind1(15, 9, ctx_r1.getBaseVersionAndPatches$(baseVersion_r2)))("c8yForLoadMore", "auto");
} }
var _c1 = function (a0) { return { "dd-low": a0 }; };
function VersionOrPatchComponent_c8y_list_group_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-list-group", 1);
    ɵngcc0.ɵɵtemplate(1, VersionOrPatchComponent_c8y_list_group_0_c8y_li_1_Template, 16, 11, "c8y-li", 2);
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(5, _c1, ctx_r0.elementCount <= ctx_r0.DD_LOW_COUNT));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yForOf", ɵngcc0.ɵɵpipeBind1(2, 3, ctx_r0.baseVersions$))("c8yForLoadMore", "auto");
} }
function SelectFirmwareComponent_i_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 15);
} }
function SelectFirmwareComponent_i_11_Template(rf, ctx) { if (rf & 1) {
    var _r4 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "i", 16);
    ɵngcc0.ɵɵlistener("click", function SelectFirmwareComponent_i_11_Template_i_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r4); var ctx_r3 = ɵngcc0.ɵɵnextContext(); return ctx_r3.textFilter$.next(""); });
    ɵngcc0.ɵɵelementEnd();
} }
function SelectFirmwareComponent_c8y_li_38_small_13_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var firmware_r5 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", firmware_r5.description, " ");
} }
function SelectFirmwareComponent_c8y_li_38_ng_template_14_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small");
    ɵngcc0.ɵɵelementStart(1, "em", 28);
    ɵngcc0.ɵɵtext(2, "No description available.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function SelectFirmwareComponent_c8y_li_38_span_21_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var firmware_r5 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", firmware_r5.c8y_Filter == null ? null : firmware_r5.c8y_Filter.type, " ");
} }
function SelectFirmwareComponent_c8y_li_38_ng_template_22_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "em", 28);
    ɵngcc0.ɵɵtext(1, "Undefined`device type`");
    ɵngcc0.ɵɵelementEnd();
} }
function SelectFirmwareComponent_c8y_li_38_span_25_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 29);
    ɵngcc0.ɵɵelementStart(1, "span", 30);
    ɵngcc0.ɵɵtext(2, "Legacy");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function SelectFirmwareComponent_c8y_li_38_span_26_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵelementStart(1, "span", 31);
    ɵngcc0.ɵɵelement(2, "i", 32);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function SelectFirmwareComponent_c8y_li_38_span_26_span_3_ng_template_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 37);
    ɵngcc0.ɵɵelementStart(1, "span", 30);
    ɵngcc0.ɵɵtext(2, "No versions");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function SelectFirmwareComponent_c8y_li_38_span_26_span_3_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 22);
    ɵngcc0.ɵɵtext(1, " version ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(2, "span", 38);
    ɵngcc0.ɵɵtext(3, "1");
    ɵngcc0.ɵɵelementEnd();
} }
function SelectFirmwareComponent_c8y_li_38_span_26_span_3_ng_template_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 22);
    ɵngcc0.ɵɵtext(1, " versions ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(2, "span", 38);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var firmware_r5 = ɵngcc0.ɵɵnextContext(3).$implicit;
    var ctx_r21 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 1, ctx_r21.getBaseVersionsCount$(firmware_r5)), " ");
} }
function SelectFirmwareComponent_c8y_li_38_span_26_span_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵelementStart(1, "span", 33);
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵtemplate(3, SelectFirmwareComponent_c8y_li_38_span_26_span_3_ng_template_3_Template, 3, 0, "ng-template", 34);
    ɵngcc0.ɵɵtemplate(4, SelectFirmwareComponent_c8y_li_38_span_26_span_3_ng_template_4_Template, 4, 0, "ng-template", 35);
    ɵngcc0.ɵɵtemplate(5, SelectFirmwareComponent_c8y_li_38_span_26_span_3_ng_template_5_Template, 5, 3, "ng-template", 36);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var firmware_r5 = ɵngcc0.ɵɵnextContext(2).$implicit;
    var ctx_r18 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngPlural", ɵngcc0.ɵɵpipeBind1(2, 1, ctx_r18.getBaseVersionsCount$(firmware_r5)));
} }
function SelectFirmwareComponent_c8y_li_38_span_26_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtemplate(1, SelectFirmwareComponent_c8y_li_38_span_26_span_1_Template, 3, 0, "span", 27);
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵtemplate(3, SelectFirmwareComponent_c8y_li_38_span_26_span_3_Template, 6, 3, "span", 27);
    ɵngcc0.ɵɵpipe(4, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var firmware_r5 = ɵngcc0.ɵɵnextContext().$implicit;
    var ctx_r14 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(2, 2, ctx_r14.getBaseVersionsCount$(firmware_r5)) === null);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(4, 4, ctx_r14.getBaseVersionsCount$(firmware_r5)) !== null);
} }
function SelectFirmwareComponent_c8y_li_38_Template(rf, ctx) { if (rf & 1) {
    var _r26 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li");
    ɵngcc0.ɵɵelementStart(1, "c8y-li-radio", 17);
    ɵngcc0.ɵɵlistener("onSelect", function SelectFirmwareComponent_c8y_li_38_Template_c8y_li_radio_onSelect_1_listener($event) { var restoredCtx = ɵngcc0.ɵɵrestoreView(_r26); var firmware_r5 = restoredCtx.$implicit; var ctx_r25 = ɵngcc0.ɵɵnextContext(); return ctx_r25.selectFirmware($event, firmware_r5); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(2, "c8y-li-icon");
    ɵngcc0.ɵɵelement(3, "i", 18);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "c8y-li-body", 8);
    ɵngcc0.ɵɵelementStart(5, "div", 19);
    ɵngcc0.ɵɵelementStart(6, "div", 10);
    ɵngcc0.ɵɵelement(7, "c8y-highlight", 20);
    ɵngcc0.ɵɵpipe(8, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(9, "div", 21);
    ɵngcc0.ɵɵelementStart(10, "div", 10);
    ɵngcc0.ɵɵelementStart(11, "span", 22);
    ɵngcc0.ɵɵtext(12, " Description ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(13, SelectFirmwareComponent_c8y_li_38_small_13_Template, 2, 1, "small", 23);
    ɵngcc0.ɵɵtemplate(14, SelectFirmwareComponent_c8y_li_38_ng_template_14_Template, 3, 0, "ng-template", null, 24, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(16, "div", 19);
    ɵngcc0.ɵɵelementStart(17, "div", 10);
    ɵngcc0.ɵɵpipe(18, "translate");
    ɵngcc0.ɵɵelementStart(19, "span", 22);
    ɵngcc0.ɵɵtext(20, " Device type ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(21, SelectFirmwareComponent_c8y_li_38_span_21_Template, 2, 1, "span", 23);
    ɵngcc0.ɵɵtemplate(22, SelectFirmwareComponent_c8y_li_38_ng_template_22_Template, 2, 0, "ng-template", null, 25, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(24, "div", 12);
    ɵngcc0.ɵɵtemplate(25, SelectFirmwareComponent_c8y_li_38_span_25_Template, 3, 0, "span", 26);
    ɵngcc0.ɵɵtemplate(26, SelectFirmwareComponent_c8y_li_38_span_26_Template, 5, 6, "span", 27);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var firmware_r5 = ctx.$implicit;
    var _r8 = ɵngcc0.ɵɵreference(15);
    var _r11 = ɵngcc0.ɵɵreference(23);
    var ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("selected", firmware_r5 === ctx_r2.selected);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵpropertyInterpolate("title", firmware_r5.name);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("text", firmware_r5.name)("pattern", ɵngcc0.ɵɵpipeBind1(8, 13, ctx_r2.textFilter$));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ctx_r2.getDescriptionTitle(firmware_r5));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", firmware_r5.description)("ngIfElse", _r8);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵpropertyInterpolate2("title", "", ɵngcc0.ɵɵpipeBind1(18, 15, "Device type"), ": ", ctx_r2.getDeviceTypeTitle(firmware_r5), "");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngIf", firmware_r5.c8y_Filter == null ? null : firmware_r5.c8y_Filter.type)("ngIfElse", _r11);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.isLegacy(firmware_r5));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r2.isLegacy(firmware_r5));
} }
'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || Array.prototype.slice.call(from));
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var StepperBulkTypeFirmwareComponent = /** @class */ (function (_super) {
        __extends(StepperBulkTypeFirmwareComponent, _super);
        function StepperBulkTypeFirmwareComponent(bulkOperationService, translate) {
            var _this = _super.call(this) || this;
            _this.bulkOperationService = bulkOperationService;
            _this.translate = translate;
            _this.descriptionTemplate = ngxComponents.gettext('{{ name }} (version {{ version }})');
            _this.textFilter$ = new rxjs.BehaviorSubject('');
            _this.stepData = {
                firmware: undefined,
                versionOrPatch: undefined
            };
            return _this;
        }
        StepperBulkTypeFirmwareComponent.prototype.goToVersionOrPatchStep = function (_a) {
            var stepper = _a.stepper;
            var selectedFirmware = this.stepData.firmware;
            this.bulkOperationService.setFirmwareId(selectedFirmware.id);
            this.deviceTypes = (selectedFirmware.c8y_Filter || {}).type;
            stepper.next();
        };
        StepperBulkTypeFirmwareComponent.prototype.onVersionOrPatchSelected = function (versionOrPatch) {
            this.stepData.versionOrPatch = versionOrPatch;
        };
        StepperBulkTypeFirmwareComponent.prototype.onFirmwareSelected = function (fw) {
            this.stepData.firmware = fw;
        };
        StepperBulkTypeFirmwareComponent.prototype.ngOnDestroy = function () {
            if (this.subscription) {
                this.subscription.unsubscribe();
            }
        };
        StepperBulkTypeFirmwareComponent.prototype.retrieveOperationPrototype = function () {
            var selectedFirmware = this.stepData.firmware;
            var selectedVersionOrPatch = this.stepData.versionOrPatch;
            if (!selectedFirmware || !selectedVersionOrPatch) {
                return;
            }
            var firmware = {
                name: selectedFirmware.name,
                version: selectedVersionOrPatch.c8y_Firmware.version,
                url: selectedVersionOrPatch.c8y_Firmware.url
            };
            if (selectedVersionOrPatch.c8y_Patch) {
                firmware.isPatch = true;
                firmware.dependency = selectedVersionOrPatch.c8y_Patch.dependency;
            }
            var name = selectedFirmware.name;
            var version = selectedVersionOrPatch.c8y_Firmware.version;
            var description = "Update firmware to: " + name + " (version: " + version + ").";
            return {
                name: ngxComponents.gettext('Firmware update'),
                description: this.translate.instant(this.descriptionTemplate, {
                    name: lodashEs.get(this.stepData, 'firmware.name'),
                    version: lodashEs.get(this.stepData, 'versionOrPatch.c8y_Firmware.version')
                }),
                prototype: {
                    description: description,
                    c8y_Firmware: firmware
                }
            };
        };
        StepperBulkTypeFirmwareComponent.prototype.onSelectionChange = function (event) {
            this.textFilter$.next('');
            _super.prototype.onSelectionChange.call(this, event);
        };
StepperBulkTypeFirmwareComponent.ɵfac = function StepperBulkTypeFirmwareComponent_Factory(t) { return new (t || StepperBulkTypeFirmwareComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BulkOperationsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.TranslateService)); };
StepperBulkTypeFirmwareComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: StepperBulkTypeFirmwareComponent, selectors: [["c8y-stepper-firmware"]], features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 5, vars: 11, consts: [[4, "customStep", "customStepCompleted", "customStepButtonsDisabled", "customStepOnNext"], [4, "customStep", "customStepCompleted", "customStepButtonsDisabled"], [1, "d-contents", 3, "selected", "firmware"], [3, "header", "name", "description"], ["list", "", 3, "selected", "versionOrPatch"]], template: function StepperBulkTypeFirmwareComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-bulk-operation-stepper");
        ɵngcc0.ɵɵtemplate(1, StepperBulkTypeFirmwareComponent_ng_container_1_Template, 2, 1, "ng-container", 0);
        ɵngcc0.ɵɵpipe(2, "translate");
        ɵngcc0.ɵɵtemplate(3, StepperBulkTypeFirmwareComponent_ng_container_3_Template, 4, 6, "ng-container", 1);
        ɵngcc0.ɵɵpipe(4, "translate");
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("customStep", ɵngcc0.ɵɵpipeBind1(2, 7, "Select firmware"))("customStepCompleted", !!(ctx.stepData == null ? null : ctx.stepData.firmware))("customStepButtonsDisabled", !(ctx.stepData == null ? null : ctx.stepData.firmware))("customStepOnNext", ctx.goToVersionOrPatchStep.bind(ctx));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("customStep", ɵngcc0.ɵɵpipeBind1(4, 9, "Select version or patch"))("customStepCompleted", !!(ctx.stepData == null ? null : ctx.stepData.versionOrPatch))("customStepButtonsDisabled", !(ctx.stepData == null ? null : ctx.stepData.versionOrPatch));
    } }, directives: function () { return [ɵngcc3.BulkOperationStepper, ɵngcc3.CustomStep, SelectFirmwareComponent, ɵngcc4.PreviewSelectionFrameComponent, VersionOrPatchComponent]; }, pipes: function () { return [ɵngcc5.C8yTranslatePipe]; }, encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(StepperBulkTypeFirmwareComponent, [{
        type: core.Component,
        args: [{
                selector: 'c8y-stepper-firmware',
                template: "<c8y-bulk-operation-stepper>\n  <ng-container\n    *customStep=\"\n      'Select firmware' | translate;\n      completed: !!stepData?.firmware;\n      buttonsDisabled: !stepData?.firmware;\n      onNext: goToVersionOrPatchStep.bind(this)\n    \"\n  >\n    <c8y-select-firmware-step\n      [selected]=\"stepData?.firmware\"\n      (firmware)=\"onFirmwareSelected($event)\"\n      class=\"d-contents\"\n    ></c8y-select-firmware-step\n  ></ng-container>\n  <ng-container\n    *customStep=\"\n      'Select version or patch' | translate;\n      completed: !!stepData?.versionOrPatch;\n      buttonsDisabled: !stepData?.versionOrPatch\n    \"\n  >\n    <c8y-preview-selection-frame\n      [header]=\"'Select version or patch' | translate\"\n      [name]=\"stepData?.firmware?.name\"\n      [description]=\"stepData?.firmware?.description\"\n    >\n      <c8y-version-or-patch-step\n        list\n        [selected]=\"stepData?.versionOrPatch\"\n        (versionOrPatch)=\"onVersionOrPatchSelected($event)\"\n      >\n      </c8y-version-or-patch-step>\n    </c8y-preview-selection-frame>\n  </ng-container>\n</c8y-bulk-operation-stepper>\n"
            }]
    }], function () { return [{ type: ɵngcc1.BulkOperationsService }, { type: ɵngcc2.TranslateService }]; }, null); })();
        return StepperBulkTypeFirmwareComponent;
    }(bulkOperationStepper.BaseStepperComponent));
    StepperBulkTypeFirmwareComponent.ctorParameters = function () { return [
        { type: bulkOperationsService.BulkOperationsService },
        { type: core$1.TranslateService }
    ]; };

    var VersionOrPatchComponent = /** @class */ (function () {
        function VersionOrPatchComponent(repositoryService, inventoryService, bulkOpsService) {
            var _this = this;
            this.repositoryService = repositoryService;
            this.inventoryService = inventoryService;
            this.bulkOpsService = bulkOpsService;
            this.versionOrPatch = new core.EventEmitter();
            this.elementCount = 0;
            this.DD_LOW_COUNT = 10;
            this.firmware$ = this.bulkOpsService.firmwareId.pipe(operators.distinctUntilChanged(), operators.switchMap(function (id) { return rxjs.defer(function () { return _this.inventoryService.detail(id).then(function (result) { return result.data; }); }).pipe(operators.catchError(function (error) { return rxjs.of(error); })); }), operators.shareReplay(1));
            this.baseVersions$ = this.firmware$.pipe(operators.distinctUntilKeyChanged('id'), operators.switchMap(function (firmware) { return _this.repositoryService.listBaseVersions(firmware); }), operators.tap(function (resp) {
                _this.elementCount = resp.data ? resp.data.length : 0;
            }), operators.shareReplay(1));
            this.isLegacy$ = this.firmware$.pipe(operators.map(function (firmware) { return _this.repositoryService.isLegacyEntry(firmware); }), operators.shareReplay(1));
            this.expanded = {};
            this.DD_LOW_COUNT = this.bulkOpsService.DD_LOW_COUNT;
        }
        VersionOrPatchComponent.prototype.getBinaryName$ = function (binaryUrl) {
            return this.repositoryService.getBinaryName$(binaryUrl);
        };
        VersionOrPatchComponent.prototype.getBaseVersionAndPatches$ = function (baseVersion) {
            var _this = this;
            return this.firmware$.pipe(operators.distinctUntilKeyChanged('id'), operators.switchMap(function (firmware) { return _this.repositoryService.listBaseVersionAndPatches(firmware, baseVersion); }), operators.shareReplay(1));
        };
        VersionOrPatchComponent.prototype.getPatchVersionsCount$ = function (baseVersion) {
            var _this = this;
            return this.firmware$.pipe(operators.distinctUntilKeyChanged('id'), operators.switchMap(function () { return _this.firmware$; }), operators.switchMap(function (firmware) { return _this.repositoryService.getPatchVersionsCount$(firmware, baseVersion); }), operators.shareReplay(1));
        };
        VersionOrPatchComponent.prototype.selectVersionOrPatch = function (selected, versionOrPatch) {
            if (selected) {
                this.versionOrPatch.emit(versionOrPatch);
            }
        };
VersionOrPatchComponent.ɵfac = function VersionOrPatchComponent_Factory(t) { return new (t || VersionOrPatchComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc6.RepositoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc7.InventoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BulkOperationsService)); };
VersionOrPatchComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: VersionOrPatchComponent, selectors: [["c8y-version-or-patch-step"]], inputs: { selected: "selected" }, outputs: { versionOrPatch: "versionOrPatch" }, decls: 2, vars: 3, consts: [["class", "m-b-16", 3, "ngClass", 4, "ngIf"], [1, "m-b-16", 3, "ngClass"], [4, "c8yFor", "c8yForOf", "c8yForLoadMore"], ["c8yIcon", "c8y-firmware"], [1, "content-flex-50"], [1, "col-10"], [1, "col-2"], ["class", "label label-warning", 4, "ngIf"], [4, "ngIf"], [1, "label", "label-warning"], [1, "label", "label-info"], ["c8yIcon", "circle-o-notch", 1, "icon-spin"], [3, "ngPlural"], ["ngPluralCase", "=0"], ["ngPluralCase", "=1"], ["ngPluralCase", "other"], [1, "label", "label-default"], ["translate", ""], ["translate", "", 3, "translateParams"], [3, "selected", "onSelect"], [1, "col-3"], ["translate", "", "class", "label label-info", 4, "ngIf", "ngIfElse"], ["version", ""], [1, "col-6", "text-truncate"], ["translate", "", 1, "text-label-small", "m-r-4"], [3, "title"], ["translate", "", 1, "label", "label-info"], ["translate", "", 1, "label", "label-primary"]], template: function VersionOrPatchComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, VersionOrPatchComponent_c8y_list_group_0_Template, 3, 7, "c8y-list-group", 0);
        ɵngcc0.ɵɵpipe(1, "async");
    } if (rf & 2) {
        var tmp_0_0;
        ɵngcc0.ɵɵproperty("ngIf", ((tmp_0_0 = ɵngcc0.ɵɵpipeBind1(1, 1, ctx.baseVersions$)) == null ? null : tmp_0_0.data.length) > 0);
    } }, directives: [ɵngcc8.NgIf, ɵngcc5.ListGroupComponent, ɵngcc8.NgClass, ɵngcc5.ForOfDirective, ɵngcc5.ListItemComponent, ɵngcc5.ListItemIconComponent, ɵngcc5.IconDirective, ɵngcc5.ListItemBodyComponent, ɵngcc5.ListItemCollapseComponent, ɵngcc8.NgPlural, ɵngcc8.NgPluralCase, ɵngcc5.C8yTranslateDirective, ɵngcc5.ListItemRadioComponent], pipes: [ɵngcc8.AsyncPipe, ɵngcc5.C8yTranslatePipe], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(VersionOrPatchComponent, [{
        type: core.Component,
        args: [{
                selector: 'c8y-version-or-patch-step',
                template: "<c8y-list-group\n  class=\"m-b-16\"\n  [ngClass]=\"{ 'dd-low': elementCount <= DD_LOW_COUNT }\"\n  *ngIf=\"(baseVersions$ | async)?.data.length > 0\"\n>\n  <c8y-li *c8yFor=\"let baseVersion of baseVersions$ | async; let i = index; loadMore: 'auto'\">\n    <c8y-li-icon>\n      <i c8yIcon=\"c8y-firmware\"></i>\n    </c8y-li-icon>\n\n    <c8y-li-body class=\"content-flex-50\">\n      <div class=\"col-10\">\n        <p>{{ baseVersion.c8y_Firmware.version }}</p>\n      </div>\n      <div class=\"col-2\">\n        <span *ngIf=\"isLegacy$ | async\" class=\"label label-warning\">\n          {{ 'Legacy' | translate }}\n        </span>\n\n        <span *ngIf=\"!(isLegacy$ | async)\">\n          <span *ngIf=\"(getPatchVersionsCount$(baseVersion) | async) === null\">\n            <span class=\"label label-info\">\n              <i c8yIcon=\"circle-o-notch\" class=\"icon-spin\"></i>\n            </span>\n          </span>\n\n          <span *ngIf=\"(getPatchVersionsCount$(baseVersion) | async) !== null\">\n            <span [ngPlural]=\"getPatchVersionsCount$(baseVersion) | async\">\n              <ng-template ngPluralCase=\"=0\">\n                <span class=\"label label-default\"> <span translate>No patches</span></span>\n              </ng-template>\n              <ng-template ngPluralCase=\"=1\">\n                <span class=\"label label-info\">\n                  <span translate>1 patch</span>\n                </span>\n              </ng-template>\n              <ng-template ngPluralCase=\"other\">\n                <span class=\"label label-info\">\n                  <span\n                    ngNonBindable\n                    translate\n                    [translateParams]=\"{ count: getPatchVersionsCount$(baseVersion) | async }\"\n                    >{{ count }} patches</span\n                  ></span\n                >\n              </ng-template>\n            </span>\n          </span>\n        </span>\n      </div>\n    </c8y-li-body>\n\n    <c8y-li-collapse>\n      <c8y-list-group>\n        <c8y-li\n          *c8yFor=\"\n            let patchVersion of getBaseVersionAndPatches$(baseVersion) | async;\n            let i = index;\n            loadMore: 'auto'\n          \"\n        >\n          <c8y-li-radio\n            (onSelect)=\"selectVersionOrPatch($event, patchVersion)\"\n            [selected]=\"patchVersion === selected\"\n          ></c8y-li-radio>\n          <c8y-li-icon>\n            <i c8yIcon=\"c8y-firmware\"></i>\n          </c8y-li-icon>\n          <c8y-li-body class=\"content-flex-50\">\n            <div class=\"col-3\">\n              {{ patchVersion.c8y_Firmware.version }}\n            </div>\n            <div class=\"col-3\">\n              <span *ngIf=\"patchVersion.c8y_Patch; else version\" translate class=\"label label-info\"\n                >patch</span\n              >\n              <ng-template #version\n                ><span translate class=\"label label-primary\">version</span></ng-template\n              >\n            </div>\n            <div class=\"col-6 text-truncate\">\n              <span class=\"text-label-small m-r-4\" translate>\n                File\n              </span>\n              <span title=\"{{ getBinaryName$(patchVersion.c8y_Firmware.url) | async }}\">\n                {{ getBinaryName$(patchVersion.c8y_Firmware.url) | async }}\n              </span>\n            </div>\n          </c8y-li-body>\n        </c8y-li>\n      </c8y-list-group>\n    </c8y-li-collapse>\n  </c8y-li>\n</c8y-list-group>\n"
            }]
    }], function () { return [{ type: ɵngcc6.RepositoryService }, { type: ɵngcc7.InventoryService }, { type: ɵngcc1.BulkOperationsService }]; }, { versionOrPatch: [{
            type: core.Output
        }], selected: [{
            type: core.Input
        }] }); })();
        return VersionOrPatchComponent;
    }());
    VersionOrPatchComponent.ctorParameters = function () { return [
        { type: shared.RepositoryService },
        { type: client.InventoryService },
        { type: bulkOperationsService.BulkOperationsService }
    ]; };
    VersionOrPatchComponent.propDecorators = {
        selected: [{ type: core.Input }],
        versionOrPatch: [{ type: core.Output }]
    };
    __decorate([
        ngxComponents.memoize()
    ], VersionOrPatchComponent.prototype, "getBinaryName$", null);
    __decorate([
        ngxComponents.memoize(lodashEs.property('id'))
    ], VersionOrPatchComponent.prototype, "getBaseVersionAndPatches$", null);
    __decorate([
        ngxComponents.memoize(lodashEs.property('id'))
    ], VersionOrPatchComponent.prototype, "getPatchVersionsCount$", null);

    var SelectFirmwareComponent = /** @class */ (function () {
        function SelectFirmwareComponent(repositoryService, bulkOpsService, translate) {
            var _this = this;
            this.repositoryService = repositoryService;
            this.bulkOpsService = bulkOpsService;
            this.translate = translate;
            this.firmware = new core.EventEmitter();
            this.elementCount = 0;
            this.DD_LOW_COUNT = 10;
            this.textFilter$ = new rxjs.BehaviorSubject('');
            this.firmwares$ = this.textFilter$
                .pipe(operators.debounceTime(400), operators.distinctUntilChanged())
                .pipe(operators.switchMap(function (text) { return _this.getFirmwares(text); }), operators.tap(function (resp) {
                _this.elementCount = resp.data ? resp.data.length : 0;
            }), operators.shareReplay(1));
            this.isLegacy = this.repositoryService.isLegacyEntry.bind(this.repositoryService);
            this.DEVICE_TYPE_NOT_AVAILABLE = ngxComponents.gettext('No device type available.');
            this.DESCRIPTION_NOT_AVAILABLE = ngxComponents.gettext('No description available.');
        }
        SelectFirmwareComponent.prototype.ngOnInit = function () {
            this.DD_LOW_COUNT = this.bulkOpsService.DD_LOW_COUNT;
        };
        SelectFirmwareComponent.prototype.selectFirmware = function (selected, fw) {
            if (selected) {
                this.firmware.emit(fw);
            }
        };
        SelectFirmwareComponent.prototype.getBaseVersionsCount$ = function (firmware) {
            var _this = this;
            return this.firmwares$.pipe(operators.switchMap(function () { return _this.repositoryService.getBaseVersionsCount$(firmware); }), operators.shareReplay(1));
        };
        SelectFirmwareComponent.prototype.getDescriptionTitle = function (firmware) {
            return lodashEs.get(firmware, 'description', this.translate.instant(this.DESCRIPTION_NOT_AVAILABLE));
        };
        SelectFirmwareComponent.prototype.getDeviceTypeTitle = function (firmware) {
            return lodashEs.get(firmware, 'c8y_Filter.type', this.translate.instant(this.DEVICE_TYPE_NOT_AVAILABLE));
        };
        SelectFirmwareComponent.prototype.getFirmwares = function (partialName) {
            return this.repositoryService.listRepositoryEntries(shared.RepositoryType.FIRMWARE, { partialName: partialName });
        };
SelectFirmwareComponent.ɵfac = function SelectFirmwareComponent_Factory(t) { return new (t || SelectFirmwareComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc6.RepositoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BulkOperationsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.TranslateService)); };
SelectFirmwareComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SelectFirmwareComponent, selectors: [["c8y-select-firmware-step"]], inputs: { selected: "selected" }, outputs: { firmware: "firmware" }, decls: 40, vars: 52, consts: [[3, "header", "noResults"], ["filters", "", 1, "input-group", "input-group-search"], ["type", "search", 1, "form-control", 3, "title", "placeholder", "ngModel", "ngModelChange"], [1, "input-group-addon"], ["c8yIcon", "filter", 4, "ngIf"], ["c8yIcon", "times", "class", "text-muted", 3, "click", 4, "ngIf"], ["header", "", 1, "hidden-sm", "hidden-xs", "m-r-8", "m-l-8"], [1, "p-l-32", "p-r-40"], [1, "content-flex-40"], [1, "col-3"], [1, "text-truncate", 3, "title"], [1, "col-4"], [1, "col-2"], ["list", "", 3, "ngClass"], [4, "c8yFor", "c8yForOf", "c8yForLoadMore"], ["c8yIcon", "filter"], ["c8yIcon", "times", 1, "text-muted", 3, "click"], [3, "selected", "onSelect"], ["c8yIcon", "c8y-firmware"], [1, "col-3", "m-b-xs-8"], [3, "text", "pattern"], [1, "col-4", "m-b-xs-8"], ["translate", "", 1, "text-label-small", "m-r-8", "visible-xs", "visible-sm"], [4, "ngIf", "ngIfElse"], ["noDesc", ""], ["noType", ""], ["class", "label label-warning flex-item-right-sm", 4, "ngIf"], [4, "ngIf"], ["translate", "", 1, "text-muted"], [1, "label", "label-warning", "flex-item-right-sm"], ["translate", ""], [1, "label", "label-info"], ["c8yIcon", "circle-o-notch", 1, "icon-spin"], [3, "ngPlural"], ["ngPluralCase", "=0"], ["ngPluralCase", "=1"], ["ngPluralCase", "other"], [1, "label", "label-default"], [1, "badge", "badge-info", "flex-item-right-sm"]], template: function SelectFirmwareComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-select-step-frame", 0);
        ɵngcc0.ɵɵpipe(1, "translate");
        ɵngcc0.ɵɵpipe(2, "async");
        ɵngcc0.ɵɵelementStart(3, "div", 1);
        ɵngcc0.ɵɵelementStart(4, "input", 2);
        ɵngcc0.ɵɵlistener("ngModelChange", function SelectFirmwareComponent_Template_input_ngModelChange_4_listener($event) { return ctx.textFilter$.next($event); });
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵpipe(6, "translate");
        ɵngcc0.ɵɵpipe(7, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(8, "span", 3);
        ɵngcc0.ɵɵtemplate(9, SelectFirmwareComponent_i_9_Template, 1, 0, "i", 4);
        ɵngcc0.ɵɵpipe(10, "async");
        ɵngcc0.ɵɵtemplate(11, SelectFirmwareComponent_i_11_Template, 1, 0, "i", 5);
        ɵngcc0.ɵɵpipe(12, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(13, "c8y-li", 6);
        ɵngcc0.ɵɵelementStart(14, "c8y-li-icon");
        ɵngcc0.ɵɵelement(15, "i", 7);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(16, "c8y-li-body", 8);
        ɵngcc0.ɵɵelementStart(17, "div", 9);
        ɵngcc0.ɵɵelementStart(18, "span", 10);
        ɵngcc0.ɵɵpipe(19, "translate");
        ɵngcc0.ɵɵtext(20);
        ɵngcc0.ɵɵpipe(21, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(22, "div", 11);
        ɵngcc0.ɵɵelementStart(23, "span", 10);
        ɵngcc0.ɵɵpipe(24, "translate");
        ɵngcc0.ɵɵtext(25);
        ɵngcc0.ɵɵpipe(26, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(27, "div", 9);
        ɵngcc0.ɵɵelementStart(28, "span", 10);
        ɵngcc0.ɵɵpipe(29, "translate");
        ɵngcc0.ɵɵtext(30);
        ɵngcc0.ɵɵpipe(31, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(32, "div", 12);
        ɵngcc0.ɵɵelementStart(33, "span", 10);
        ɵngcc0.ɵɵpipe(34, "translate");
        ɵngcc0.ɵɵtext(35);
        ɵngcc0.ɵɵpipe(36, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(37, "c8y-list-group", 13);
        ɵngcc0.ɵɵtemplate(38, SelectFirmwareComponent_c8y_li_38_Template, 27, 17, "c8y-li", 14);
        ɵngcc0.ɵɵpipe(39, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("header", ɵngcc0.ɵɵpipeBind1(1, 18, "Select firmware"))("noResults", ɵngcc0.ɵɵpipeBind1(2, 20, ctx.textFilter$).length > 0 && ctx.elementCount === 0);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(5, 22, "Filter firmware\u2026"));
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(6, 24, "Filter firmware\u2026"));
        ɵngcc0.ɵɵproperty("ngModel", ɵngcc0.ɵɵpipeBind1(7, 26, ctx.textFilter$));
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(10, 28, ctx.textFilter$).length === 0);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(12, 30, ctx.textFilter$).length > 0);
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵpropertyInterpolate1("title", " ", ɵngcc0.ɵɵpipeBind1(19, 32, "Name"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(21, 34, "Name"), " ");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(24, 36, "Description"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(26, 38, "Description"), " ");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(29, 40, "Device type"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(31, 42, "Device type"), " ");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(34, 44, "Versions"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(36, 46, "Versions"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(50, _c1, ctx.elementCount < ctx.DD_LOW_COUNT));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("c8yForOf", ɵngcc0.ɵɵpipeBind1(39, 48, ctx.firmwares$))("c8yForLoadMore", "auto");
    } }, directives: [ɵngcc4.SelectStepFrameComponent, ɵngcc9.DefaultValueAccessor, ɵngcc9.NgControlStatus, ɵngcc9.NgModel, ɵngcc8.NgIf, ɵngcc5.ListItemComponent, ɵngcc5.ListItemIconComponent, ɵngcc5.ListItemBodyComponent, ɵngcc5.ListGroupComponent, ɵngcc8.NgClass, ɵngcc5.ForOfDirective, ɵngcc5.IconDirective, ɵngcc5.ListItemRadioComponent, ɵngcc5.HighlightComponent, ɵngcc5.C8yTranslateDirective, ɵngcc8.NgPlural, ɵngcc8.NgPluralCase], pipes: [ɵngcc5.C8yTranslatePipe, ɵngcc8.AsyncPipe], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SelectFirmwareComponent, [{
        type: core.Component,
        args: [{
                selector: 'c8y-select-firmware-step',
                template: "<c8y-select-step-frame\n  [header]=\"'Select firmware' | translate\"\n  [noResults]=\"(textFilter$ | async).length > 0 && elementCount === 0\"\n>\n  <div filters class=\"input-group input-group-search\">\n    <input\n      type=\"search\"\n      class=\"form-control\"\n      title=\"{{ 'Filter firmware\u2026' | translate }}\"\n      placeholder=\"{{ 'Filter firmware\u2026' | translate }}\"\n      [ngModel]=\"textFilter$ | async\"\n      (ngModelChange)=\"textFilter$.next($event)\"\n    />\n    <span class=\"input-group-addon\">\n      <i c8yIcon=\"filter\" *ngIf=\"(textFilter$ | async).length === 0\"></i>\n      <i\n        c8yIcon=\"times\"\n        class=\"text-muted\"\n        *ngIf=\"(textFilter$ | async).length > 0\"\n        (click)=\"textFilter$.next('')\"\n      ></i>\n    </span>\n  </div>\n  <c8y-li header class=\"hidden-sm hidden-xs m-r-8 m-l-8\">\n    <c8y-li-icon><i class=\"p-l-32 p-r-40\"></i></c8y-li-icon>\n    <c8y-li-body class=\"content-flex-40\">\n      <div class=\"col-3\">\n        <span class=\"text-truncate\" title=\" {{ 'Name' | translate }} \">\n          {{ 'Name' | translate }}\n        </span>\n      </div>\n      <div class=\"col-4\">\n        <span class=\"text-truncate\" title=\"{{ 'Description' | translate }}\">\n          {{ 'Description' | translate }}\n        </span>\n      </div>\n      <div class=\"col-3\">\n        <span class=\"text-truncate\" title=\"{{ 'Device type' | translate }}\">\n          {{ 'Device type' | translate }}\n        </span>\n      </div>\n      <div class=\"col-2\">\n        <span class=\"text-truncate\" title=\"{{ 'Versions' | translate }}\">\n          {{ 'Versions' | translate }}\n        </span>\n      </div>\n    </c8y-li-body>\n  </c8y-li>\n  <c8y-list-group list [ngClass]=\"{ 'dd-low': elementCount < DD_LOW_COUNT }\">\n    <c8y-li *c8yFor=\"let firmware of firmwares$ | async; let i = index; loadMore: 'auto'\">\n      <c8y-li-radio\n        (onSelect)=\"selectFirmware($event, firmware)\"\n        [selected]=\"firmware === selected\"\n      ></c8y-li-radio>\n      <c8y-li-icon>\n        <i c8yIcon=\"c8y-firmware\"></i>\n      </c8y-li-icon>\n\n      <c8y-li-body class=\"content-flex-40\">\n        <div class=\"col-3 m-b-xs-8\">\n          <div class=\"text-truncate\" title=\"{{ firmware.name }}\">\n            <c8y-highlight [text]=\"firmware.name\" [pattern]=\"textFilter$ | async\"></c8y-highlight>\n          </div>\n        </div>\n        <div class=\"col-4 m-b-xs-8\">\n          <div class=\"text-truncate\" title=\"{{ getDescriptionTitle(firmware) }}\">\n            <span class=\"text-label-small m-r-8 visible-xs visible-sm\" translate>\n              Description\n            </span>\n            <small *ngIf=\"firmware.description; else noDesc\">\n              {{ firmware.description }}\n            </small>\n            <ng-template #noDesc>\n              <small>\n                <em class=\"text-muted\" translate>No description available.</em>\n              </small>\n            </ng-template>\n          </div>\n        </div>\n        <div class=\"col-3 m-b-xs-8\">\n          <div\n            class=\"text-truncate\"\n            title=\"{{ 'Device type' | translate }}: {{ getDeviceTypeTitle(firmware) }}\"\n          >\n            <span class=\"text-label-small m-r-8 visible-xs visible-sm\" translate>\n              Device type\n            </span>\n            <span *ngIf=\"firmware.c8y_Filter?.type; else noType\">\n              {{ firmware.c8y_Filter?.type }}\n            </span>\n            <ng-template #noType>\n              <em class=\"text-muted\" translate>Undefined`device type`</em>\n            </ng-template>\n          </div>\n        </div>\n        <div class=\"col-2\">\n          <span *ngIf=\"isLegacy(firmware)\" class=\"label label-warning flex-item-right-sm\">\n            <span translate>Legacy</span>\n          </span>\n\n          <span *ngIf=\"!isLegacy(firmware)\">\n            <span *ngIf=\"(getBaseVersionsCount$(firmware) | async) === null\">\n              <span class=\"label label-info\">\n                <i c8yIcon=\"circle-o-notch\" class=\"icon-spin\"></i>\n              </span>\n            </span>\n            <span *ngIf=\"(getBaseVersionsCount$(firmware) | async) !== null\">\n              <span [ngPlural]=\"getBaseVersionsCount$(firmware) | async\">\n                <ng-template ngPluralCase=\"=0\">\n                  <span class=\"label label-default\">\n                    <span translate>No versions</span>\n                  </span>\n                </ng-template>\n                <ng-template ngPluralCase=\"=1\">\n                  <span class=\"text-label-small m-r-8 visible-xs visible-sm\" translate>\n                    version\n                  </span>\n                  <span class=\"badge badge-info flex-item-right-sm\">1</span>\n                </ng-template>\n                <ng-template ngPluralCase=\"other\">\n                  <span class=\"text-label-small m-r-8 visible-xs visible-sm\" translate>\n                    versions\n                  </span>\n                  <span class=\"badge badge-info flex-item-right-sm\">\n                    {{ getBaseVersionsCount$(firmware) | async }}\n                  </span>\n                </ng-template>\n              </span>\n            </span>\n          </span>\n        </div>\n      </c8y-li-body>\n    </c8y-li>\n  </c8y-list-group>\n</c8y-select-step-frame>\n"
            }]
    }], function () { return [{ type: ɵngcc6.RepositoryService }, { type: ɵngcc1.BulkOperationsService }, { type: ɵngcc2.TranslateService }]; }, { firmware: [{
            type: core.Output
        }], selected: [{
            type: core.Input
        }] }); })();
        return SelectFirmwareComponent;
    }());
    SelectFirmwareComponent.ctorParameters = function () { return [
        { type: shared.RepositoryService },
        { type: bulkOperationsService.BulkOperationsService },
        { type: core$1.TranslateService }
    ]; };
    SelectFirmwareComponent.propDecorators = {
        selected: [{ type: core.Input }],
        firmware: [{ type: core.Output }]
    };
    __decorate([
        ngxComponents.memoize(lodashEs.property('id'))
    ], SelectFirmwareComponent.prototype, "getBaseVersionsCount$", null);

    var ɵ0 = {
        type: bulkOperationsService.BulkOperationType.FIRMWARE,
        c8yIcon: 'c8y-firmware',
        name: ngxComponents.gettext('Firmware update'),
        path: bulkOperationsService.baseUrl + "firmware",
        component: StepperBulkTypeFirmwareComponent,
        fragments: ['c8y_Firmware'],
        selected: false
    };
    /** Module for the 'Firmware update' operation type stepper */
    var StepperBulkTypeFirmwareModule = /** @class */ (function () {
        function StepperBulkTypeFirmwareModule() {
        }
StepperBulkTypeFirmwareModule.ɵfac = function StepperBulkTypeFirmwareModule_Factory(t) { return new (t || StepperBulkTypeFirmwareModule)(); };
StepperBulkTypeFirmwareModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: StepperBulkTypeFirmwareModule });
StepperBulkTypeFirmwareModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        {
            provide: bulkOperationsService.HOOK_LIST_BULK_TYPE,
            useValue: ɵ0,
            multi: true
        }
    ], imports: [[
            ngxComponents.CoreModule,
            ngxComponents.FormsModule,
            forms.ReactiveFormsModule,
            bulkOperationStepper.BulkOperationStepperModule,
            stepperFrames.StepperFramesModule,
            shared.SharedRepositoryModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(StepperBulkTypeFirmwareModule, [{
        type: core.NgModule,
        args: [{
                declarations: [
                    StepperBulkTypeFirmwareComponent,
                    VersionOrPatchComponent,
                    SelectFirmwareComponent
                ],
                imports: [
                    ngxComponents.CoreModule,
                    ngxComponents.FormsModule,
                    forms.ReactiveFormsModule,
                    bulkOperationStepper.BulkOperationStepperModule,
                    stepperFrames.StepperFramesModule,
                    shared.SharedRepositoryModule
                ],
                providers: [
                    {
                        provide: bulkOperationsService.HOOK_LIST_BULK_TYPE,
                        useValue: ɵ0,
                        multi: true
                    }
                ],
                entryComponents: [StepperBulkTypeFirmwareComponent],
                exports: [StepperBulkTypeFirmwareComponent]
            }]
    }], function () { return []; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(StepperBulkTypeFirmwareModule, { declarations: [StepperBulkTypeFirmwareComponent, VersionOrPatchComponent, SelectFirmwareComponent], imports: [ɵngcc5.CoreModule, ɵngcc5.FormsModule, ɵngcc9.ReactiveFormsModule, ɵngcc3.BulkOperationStepperModule, ɵngcc4.StepperFramesModule, ɵngcc6.SharedRepositoryModule], exports: [StepperBulkTypeFirmwareComponent] }); })();
        return StepperBulkTypeFirmwareModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.SelectFirmwareComponent = SelectFirmwareComponent;
    exports.StepperBulkTypeFirmwareComponent = StepperBulkTypeFirmwareComponent;
    exports.StepperBulkTypeFirmwareModule = StepperBulkTypeFirmwareModule;
    exports.VersionOrPatchComponent = VersionOrPatchComponent;
    exports["ɵ0"] = ɵ0;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
