// @c8y/ngx-components/bundles/c8y-ngx-components-operations-operations-timeline.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@c8y/ngx-components'),require('@c8y/ngx-components/operations/shared'),require('@angular/common'),require('ngx-bootstrap/tooltip'),require('ngx-bootstrap/collapse'),exports, require('@angular/core'), require('@c8y/ngx-components'), require('@c8y/ngx-components/operations/shared'), require('rxjs'), require('rxjs/operators'), require('@angular/common'), require('ngx-bootstrap/collapse'), require('ngx-bootstrap/tooltip')) :
    typeof define === 'function' && define.amd ? define('@c8y/ngx-components/operations/operations-timeline', ['@angular/core','@c8y/ngx-components','@c8y/ngx-components/operations/shared','@angular/common','ngx-bootstrap/tooltip','ngx-bootstrap/collapse','exports', '@angular/core', '@c8y/ngx-components', '@c8y/ngx-components/operations/shared', 'rxjs', 'rxjs/operators', '@angular/common', 'ngx-bootstrap/collapse', 'ngx-bootstrap/tooltip'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.c8y.ngxComponents,global.c8y.ngxComponents.operations.shared,global.ng.common,global.ngxBootstrap.tooltip,global.ngxBootstrap.collapse,(global.c8y = global.c8y || {}, global.c8y["ngx-components"] = global.c8y["ngx-components"] || {}, global.c8y["ngx-components"].operations = global.c8y["ngx-components"].operations || {}, global.c8y["ngx-components"].operations["operations-timeline"] = {}), global.ng.core, global.c8y["ngx-components"], global.c8y["ngx-components"].operations.shared, global.rxjs, global.rxjs.operators, global.ng.common, global.collapse, global.tooltip));
})(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,ɵngcc4,ɵngcc5,exports, core, ngxComponents, shared, rxjs, operators, common, collapse, tooltip) { 
function OperationsTimelineComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 2);
    ɵngcc0.ɵɵelementStart(1, "div", 3);
    ɵngcc0.ɵɵelement(2, "h1", 4);
    ɵngcc0.ɵɵelementStart(3, "p");
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(5, 1, "No recent operations found."));
} }
var _c0 = function (a0) { return { $implicit: a0 }; };
function OperationsTimelineComponent_c8y_li_timeline_2_ng_container_14_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0, 17);
} if (rf & 2) {
    var op_r2 = ɵngcc0.ɵɵnextContext().$implicit;
    var ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r4.bodyTemplate)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(2, _c0, op_r2));
} }
function OperationsTimelineComponent_c8y_li_timeline_2_ng_template_15_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var op_r2 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(op_r2.description);
} }
function OperationsTimelineComponent_c8y_li_timeline_2_div_17_Template(rf, ctx) { if (rf & 1) {
    var _r15 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 18);
    ɵngcc0.ɵɵelementStart(1, "button", 19);
    ɵngcc0.ɵɵlistener("click", function OperationsTimelineComponent_c8y_li_timeline_2_div_17_Template_button_click_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r15); var op_r2 = ɵngcc0.ɵɵnextContext().$implicit; var ctx_r13 = ɵngcc0.ɵɵnextContext(); $event.stopPropagation(); return ctx_r13.operationsService.cancel(op_r2); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 20);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 1, "Cancel operation"));
} }
function OperationsTimelineComponent_c8y_li_timeline_2_li_22_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 21);
    ɵngcc0.ɵɵelementStart(1, "label", 22);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "span", 23);
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var prop_r16 = ctx.$implicit;
    var op_r2 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 2, prop_r16.value), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(6, 4, op_r2[prop_r16.key]), " ");
} }
function OperationsTimelineComponent_c8y_li_timeline_2_li_24_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 21);
    ɵngcc0.ɵɵelementStart(1, "label", 22);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵpipe(4, "humanize");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "span", 24);
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵpipe(7, "json");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var key_r18 = ctx.$implicit;
    var op_r2 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 2, ɵngcc0.ɵɵpipeBind1(4, 4, key_r18)), " ");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(7, 6, op_r2[key_r18]), " ");
} }
function OperationsTimelineComponent_c8y_li_timeline_2_ng_container_25_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0, 17);
} if (rf & 2) {
    var footer_r20 = ctx.$implicit;
    var op_r2 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", footer_r20)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(2, _c0, op_r2));
} }
function OperationsTimelineComponent_c8y_li_timeline_2_Template(rf, ctx) { if (rf & 1) {
    var _r23 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li-timeline");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "date");
    ɵngcc0.ɵɵelementStart(3, "c8y-li", 5, 6);
    ɵngcc0.ɵɵlistener("click", function OperationsTimelineComponent_c8y_li_timeline_2_Template_c8y_li_click_3_listener() { ɵngcc0.ɵɵrestoreView(_r23); var _r3 = ɵngcc0.ɵɵreference(4); return _r3.toggleCollapsed(); });
    ɵngcc0.ɵɵelementStart(5, "c8y-li-icon");
    ɵngcc0.ɵɵelement(6, "i", 7);
    ɵngcc0.ɵɵpipe(7, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(8, "c8y-li-body");
    ɵngcc0.ɵɵelementStart(9, "div", 8);
    ɵngcc0.ɵɵelementStart(10, "div");
    ɵngcc0.ɵɵtext(11);
    ɵngcc0.ɵɵpipe(12, "translate");
    ɵngcc0.ɵɵelement(13, "br", 9);
    ɵngcc0.ɵɵtemplate(14, OperationsTimelineComponent_c8y_li_timeline_2_ng_container_14_Template, 1, 4, "ng-container", 10);
    ɵngcc0.ɵɵtemplate(15, OperationsTimelineComponent_c8y_li_timeline_2_ng_template_15_Template, 2, 1, "ng-template", null, 11, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(17, OperationsTimelineComponent_c8y_li_timeline_2_div_17_Template, 4, 3, "div", 12);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(18, "c8y-li-collapse");
    ɵngcc0.ɵɵelementStart(19, "div", 13);
    ɵngcc0.ɵɵtext(20, "Details");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(21, "ul", 14);
    ɵngcc0.ɵɵtemplate(22, OperationsTimelineComponent_c8y_li_timeline_2_li_22_Template, 7, 6, "li", 15);
    ɵngcc0.ɵɵpipe(23, "keyvalue");
    ɵngcc0.ɵɵtemplate(24, OperationsTimelineComponent_c8y_li_timeline_2_li_24_Template, 8, 8, "li", 15);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(25, OperationsTimelineComponent_c8y_li_timeline_2_ng_container_25_Template, 1, 4, "ng-container", 16);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var op_r2 = ctx.$implicit;
    var _r5 = ɵngcc0.ɵɵreference(16);
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind2(2, 11, op_r2.creationTime, "medium"), " ");
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵproperty("c8yIcon", ctx_r1.OPERATION_STATUS_OPTIONS_MAP[op_r2.status].icon)("ngClass", ctx_r1.OPERATION_STATUS_OPTIONS_MAP[op_r2.status].styleClass)("tooltip", ɵngcc0.ɵɵpipeBind1(7, 14, op_r2.status));
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(12, 16, op_r2.status), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.bodyTemplate)("ngIfElse", _r5);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", op_r2.status === "PENDING");
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵproperty("ngForOf", ɵngcc0.ɵɵpipeBind1(23, 18, ctx_r1.operationsService.getStandardKeys(op_r2)));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r1.operationsService.getNonStandardKeys(op_r2, ctx_r1.propertiesToHide));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r1.footerTemplates);
} }
var _c1 = function (a0) { return { entityOrId: a0 }; };
'use strict';

    var OperationsTimelineComponent = /** @class */ (function () {
        function OperationsTimelineComponent(realtime, operationsService) {
            this.realtime = realtime;
            this.operationsService = operationsService;
            this.filterPipe = rxjs.pipe(operators.tap());
            this.footerTemplates = [];
            this.propertiesToHide = [];
            this.OPERATION_STATUS_OPTIONS_MAP = shared.OPERATION_STATUS_OPTIONS_MAP;
        }
OperationsTimelineComponent.ɵfac = function OperationsTimelineComponent_Factory(t) { return new (t || OperationsTimelineComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.OperationRealtimeService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.OperationsService)); };
OperationsTimelineComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: OperationsTimelineComponent, selectors: [["c8y-operations-timeline"]], inputs: { filterPipe: "filterPipe", footerTemplates: "footerTemplates", propertiesToHide: "propertiesToHide", operations: "operations", sourceId: "sourceId", bodyTemplate: "bodyTemplate" }, decls: 3, vars: 7, consts: [["class", "card", 4, "ngIf"], [4, "c8yFor", "c8yForOf", "c8yForPipe", "c8yForRealtime", "c8yForRealtimeOptions"], [1, "card"], [1, "c8y-empty-state", "text-left"], [1, "c8y-icon", "c8y-icon-energy", "c8y-icon-duocolor", "m-l-8"], [3, "click"], ["li", ""], [3, "c8yIcon", "ngClass", "tooltip"], [1, "flex-row"], [1, "visible-xs", "visible-sm"], [3, "ngTemplateOutlet", "ngTemplateOutletContext", 4, "ngIf", "ngIfElse"], ["defaultBody", ""], ["class", "flex-item-right showOnHover", 4, "ngIf"], ["translate", "", 1, "legend", "form-block"], [1, "list-unstyled", "small"], ["class", "p-t-4 p-b-4 flex-row separator-bottom", 4, "ngFor", "ngForOf"], [3, "ngTemplateOutlet", "ngTemplateOutletContext", 4, "ngFor", "ngForOf"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], [1, "flex-item-right", "showOnHover"], [1, "btn", "btn-clean", 3, "title", "click"], ["c8yIcon", "minus-circle", 1, "text-danger"], [1, "p-t-4", "p-b-4", "flex-row", "separator-bottom"], [1, "small", "m-b-0", "m-r-8", "flex-item-v-start"], [1, "flex-item-right"], [1, "flex-item-right", "text-code"]], template: function OperationsTimelineComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, OperationsTimelineComponent_div_0_Template, 6, 3, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "c8y-list-group");
        ɵngcc0.ɵɵtemplate(2, OperationsTimelineComponent_c8y_li_timeline_2_Template, 26, 20, "c8y-li-timeline", 1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", !(ctx.operations == null ? null : ctx.operations.data) || !(ctx.operations == null ? null : ctx.operations.data == null ? null : ctx.operations.data.length));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yForOf", ctx.operations)("c8yForPipe", ctx.filterPipe)("c8yForRealtime", ctx.realtime)("c8yForRealtimeOptions", ɵngcc0.ɵɵpureFunction1(5, _c1, ctx.sourceId));
    } }, directives: [ɵngcc3.NgIf, ɵngcc1.ListGroupComponent, ɵngcc1.ForOfDirective, ɵngcc1.ListItemTimelineComponent, ɵngcc1.ListItemComponent, ɵngcc1.ListItemIconComponent, ɵngcc1.IconDirective, ɵngcc3.NgClass, ɵngcc4.TooltipDirective, ɵngcc1.ListItemBodyComponent, ɵngcc1.ListItemCollapseComponent, ɵngcc1.C8yTranslateDirective, ɵngcc3.NgForOf, ɵngcc3.NgTemplateOutlet], pipes: [ɵngcc1.C8yTranslatePipe, ɵngcc3.DatePipe, ɵngcc3.KeyValuePipe, ɵngcc1.HumanizePipe, ɵngcc3.JsonPipe], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OperationsTimelineComponent, [{
        type: core.Component,
        args: [{
                selector: 'c8y-operations-timeline',
                template: "<div class=\"card\" *ngIf=\"!operations?.data || !operations?.data?.length\">\n  <div class=\"c8y-empty-state text-left\">\n    <h1 class=\"c8y-icon c8y-icon-energy c8y-icon-duocolor m-l-8\"></h1>\n    <p>{{ 'No recent operations found.' | translate }}</p>\n  </div>\n</div>\n\n<c8y-list-group>\n  <c8y-li-timeline\n    *c8yFor=\"\n      let op of operations;\n      pipe: filterPipe;\n      realtime: realtime;\n      realtimeOptions: { entityOrId: sourceId }\n    \"\n  >\n    {{ op.creationTime | date: 'medium' }}\n    <c8y-li #li (click)=\"li.toggleCollapsed()\">\n      <c8y-li-icon>\n        <i\n          [c8yIcon]=\"OPERATION_STATUS_OPTIONS_MAP[op.status].icon\"\n          [ngClass]=\"OPERATION_STATUS_OPTIONS_MAP[op.status].styleClass\"\n          [tooltip]=\"op.status | translate\"\n        ></i>\n      </c8y-li-icon>\n      <c8y-li-body>\n        <div class=\"flex-row\">\n          <div>\n            {{ op.status | translate }} <br class=\"visible-xs visible-sm\" />\n            <ng-container\n              *ngIf=\"bodyTemplate; else defaultBody\"\n              [ngTemplateOutlet]=\"bodyTemplate\"\n              [ngTemplateOutletContext]=\"{ $implicit: op }\"\n            ></ng-container>\n            <ng-template #defaultBody>\n              <small>{{ op.description }}</small>\n            </ng-template>\n          </div>\n          <div class=\"flex-item-right showOnHover\" *ngIf=\"op.status === 'PENDING'\">\n            <button\n              class=\"btn btn-clean\"\n              title=\"{{ 'Cancel operation' | translate }}\"\n              (click)=\"$event.stopPropagation(); operationsService.cancel(op)\"\n            >\n              <i c8yIcon=\"minus-circle\" class=\"text-danger\"></i>\n            </button>\n          </div>\n        </div>\n      </c8y-li-body>\n\n      <c8y-li-collapse>\n        <div class=\"legend form-block\" translate>Details</div>\n        <ul class=\"list-unstyled small\">\n          <li\n            class=\"p-t-4 p-b-4 flex-row separator-bottom\"\n            *ngFor=\"let prop of operationsService.getStandardKeys(op) | keyvalue\"\n          >\n            <label class=\"small m-b-0 m-r-8 flex-item-v-start\">\n              {{ prop.value | translate }}\n            </label>\n            <span class=\"flex-item-right\">\n              {{ op[prop.key] | translate }}\n            </span>\n          </li>\n          <li\n            class=\"p-t-4 p-b-4 flex-row separator-bottom\"\n            *ngFor=\"let key of operationsService.getNonStandardKeys(op, propertiesToHide)\"\n          >\n            <label class=\"small m-b-0 m-r-8 flex-item-v-start\">\n              {{ key | humanize | translate }}\n            </label>\n            <span class=\"flex-item-right text-code\">\n              {{ op[key] | json }}\n            </span>\n          </li>\n        </ul>\n        <ng-container\n          *ngFor=\"let footer of footerTemplates\"\n          [ngTemplateOutlet]=\"footer\"\n          [ngTemplateOutletContext]=\"{ $implicit: op }\"\n        ></ng-container>\n      </c8y-li-collapse>\n    </c8y-li>\n  </c8y-li-timeline>\n</c8y-list-group>\n"
            }]
    }], function () { return [{ type: ɵngcc1.OperationRealtimeService }, { type: ɵngcc2.OperationsService }]; }, { filterPipe: [{
            type: core.Input
        }], footerTemplates: [{
            type: core.Input
        }], propertiesToHide: [{
            type: core.Input
        }], operations: [{
            type: core.Input
        }], sourceId: [{
            type: core.Input
        }], bodyTemplate: [{
            type: core.Input
        }] }); })();
        return OperationsTimelineComponent;
    }());
    OperationsTimelineComponent.ctorParameters = function () { return [
        { type: ngxComponents.OperationRealtimeService },
        { type: shared.OperationsService }
    ]; };
    OperationsTimelineComponent.propDecorators = {
        operations: [{ type: core.Input }],
        sourceId: [{ type: core.Input }],
        filterPipe: [{ type: core.Input }],
        bodyTemplate: [{ type: core.Input }],
        footerTemplates: [{ type: core.Input }],
        propertiesToHide: [{ type: core.Input }]
    };

    var OperationsTimelineModule = /** @class */ (function () {
        function OperationsTimelineModule() {
        }
OperationsTimelineModule.ɵfac = function OperationsTimelineModule_Factory(t) { return new (t || OperationsTimelineModule)(); };
OperationsTimelineModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: OperationsTimelineModule });
OperationsTimelineModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[common.CommonModule, ngxComponents.CoreModule, collapse.CollapseModule, tooltip.TooltipModule, shared.OperationsSharedModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OperationsTimelineModule, [{
        type: core.NgModule,
        args: [{
                imports: [common.CommonModule, ngxComponents.CoreModule, collapse.CollapseModule, tooltip.TooltipModule, shared.OperationsSharedModule],
                declarations: [OperationsTimelineComponent],
                exports: [OperationsTimelineComponent]
            }]
    }], function () { return []; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(OperationsTimelineModule, { declarations: [OperationsTimelineComponent], imports: [ɵngcc3.CommonModule, ɵngcc1.CoreModule, ɵngcc5.CollapseModule, ɵngcc4.TooltipModule, ɵngcc2.OperationsSharedModule], exports: [OperationsTimelineComponent] }); })();
        return OperationsTimelineModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.OperationsTimelineComponent = OperationsTimelineComponent;
    exports.OperationsTimelineModule = OperationsTimelineModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
