// @c8y/ngx-components/bundles/c8y-ngx-components-widgets-kpi-widget.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/forms'),require('@c8y/ngx-components/icon-selector'),require('@c8y/ngx-components/context-dashboard'),require('@c8y/ngx-components/datapoint-selector'),require('@c8y/ngx-components'),require('ngx-bootstrap/popover'),require('@angular/common'),exports, require('@angular/core'), require('@c8y/ngx-components'), require('@c8y/ngx-components/datapoint-selector'), require('@c8y/ngx-components/icon-selector'), require('ngx-bootstrap/popover'), require('@angular/forms'), require('@c8y/ngx-components/context-dashboard'), require('rxjs'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('@c8y/ngx-components/widgets/kpi-widget', ['@angular/core','@angular/forms','@c8y/ngx-components/icon-selector','@c8y/ngx-components/context-dashboard','@c8y/ngx-components/datapoint-selector','@c8y/ngx-components','ngx-bootstrap/popover','@angular/common','exports', '@angular/core', '@c8y/ngx-components', '@c8y/ngx-components/datapoint-selector', '@c8y/ngx-components/icon-selector', 'ngx-bootstrap/popover', '@angular/forms', '@c8y/ngx-components/context-dashboard', 'rxjs', 'rxjs/operators'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ng.forms,global.c8y.ngxComponents.iconSelector,global.c8y.ngxComponents.contextDashboard,global.c8y.ngxComponents.datapointSelector,global.c8y.ngxComponents,global.ngxBootstrap.popover,global.ng.common,(global.c8y = global.c8y || {}, global.c8y["ngx-components"] = global.c8y["ngx-components"] || {}, global.c8y["ngx-components"].widgets = global.c8y["ngx-components"].widgets || {}, global.c8y["ngx-components"].widgets["kpi-widget"] = {}), global.ng.core, global.c8y["ngx-components"], global.c8y["ngx-components"]["datapoint-selector"], global.c8y["ngx-components"]["icon-selector"], global.popover, global.ng.forms, global.c8y["ngx-components"]["context-dashboard"], global.rxjs, global.rxjs.operators));
})(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,ɵngcc4,ɵngcc5,ɵngcc6,ɵngcc7,exports, core, ngxComponents, datapointSelector, iconSelector, popover, forms, contextDashboard, rxjs, operators) { 
var _c0 = function () { return { example: 1 }; };
var _c1 = function () { return { example: 36 }; };
function KpiWidgetViewComponent_div_0_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 11);
    ɵngcc0.ɵɵelement(1, "i", 12);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var lastState_r3 = ɵngcc0.ɵɵnextContext().ngIf;
    var ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", lastState_r3.colorClass);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yIcon", ctx_r4.config.icon);
} }
var _c2 = function (a0) { return { transform: a0 }; };
function KpiWidgetViewComponent_div_0_div_12_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 13);
    ɵngcc0.ɵɵelement(1, "i", 14);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵpipe(3, "number");
    ɵngcc0.ɵɵpipe(4, "date");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var previousValue_r8 = ctx.ngIf;
    var lastState_r3 = ɵngcc0.ɵɵnextContext().ngIf;
    var ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(2, 2, "Previous value") + ": " + ɵngcc0.ɵɵpipeBind2(3, 4, previousValue_r8.value, "1." + (ctx_r5.config.numberOfDecimalPlaces || "0") + "-" + (ctx_r5.config.numberOfDecimalPlaces || "0")) + " (" + ɵngcc0.ɵɵpipeBind2(4, 7, previousValue_r8.date, "medium") + ")")("ngStyle", ɵngcc0.ɵɵpureFunction1(10, _c2, "rotate(" + lastState_r3.trend + ")"));
} }
function KpiWidgetViewComponent_div_0_p_14_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p", 15);
    ɵngcc0.ɵɵelement(1, "i", 16);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "date");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var lastState_r3 = ɵngcc0.ɵɵnextContext().ngIf;
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind2(3, 1, lastState_r3.latestMeasurement.date, "medium"), " ");
} }
var _c3 = function (a0) { return { "font-size": a0 }; };
function KpiWidgetViewComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 2);
    ɵngcc0.ɵɵelementStart(1, "div", 3);
    ɵngcc0.ɵɵtemplate(2, KpiWidgetViewComponent_div_0_div_2_Template, 2, 2, "div", 4);
    ɵngcc0.ɵɵelementStart(3, "div", 5);
    ɵngcc0.ɵɵelementStart(4, "span", 6);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵpipe(7, "number");
    ɵngcc0.ɵɵtext(8);
    ɵngcc0.ɵɵpipe(9, "number");
    ɵngcc0.ɵɵelementStart(10, "small", 7);
    ɵngcc0.ɵɵtext(11);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(12, KpiWidgetViewComponent_div_0_div_12_Template, 5, 12, "div", 8);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(13, "div", 9);
    ɵngcc0.ɵɵtemplate(14, KpiWidgetViewComponent_div_0_p_14_Template, 4, 4, "p", 10);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var lastState_r3 = ctx.ngIf;
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.config.icon && ctx_r0.config.showIcon);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate3("title", "", lastState_r3.colorClass === "text-danger" ? ɵngcc0.ɵɵpipeBind1(5, 10, "Within red range:") : lastState_r3.colorClass === "text-warning" ? ɵngcc0.ɵɵpipeBind1(6, 12, "Within yellow range:") : "", " ", ɵngcc0.ɵɵpipeBind2(7, 14, lastState_r3.latestMeasurement.value, "1." + (ctx_r0.config.numberOfDecimalPlaces || "0") + "-" + (ctx_r0.config.numberOfDecimalPlaces || "0")), " ", lastState_r3.unit || "", "");
    ɵngcc0.ɵɵproperty("ngClass", lastState_r3.colorClass)("ngStyle", ɵngcc0.ɵɵpureFunction1(20, _c3, (ctx_r0.config.fontSize || "36") + "px"));
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind2(9, 17, lastState_r3.latestMeasurement.value, "1." + (ctx_r0.config.numberOfDecimalPlaces || "0") + "-" + (ctx_r0.config.numberOfDecimalPlaces || "0")), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(lastState_r3.unit || "");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", (ctx_r0.config == null ? null : ctx_r0.config.showTrend) && lastState_r3.previousValue);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.config == null ? null : ctx_r0.config.showTimestamp);
} }
function KpiWidgetViewComponent_ng_template_2_c8y_ui_empty_state_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-ui-empty-state", 20);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵpipe(2, "translate");
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("icon", "line-chart")("title", ɵngcc0.ɵɵpipeBind1(1, 4, "No measurement to display."))("subtitle", ɵngcc0.ɵɵpipeBind1(2, 6, "Waiting for measurements to be created."))("horizontal", true);
} }
function KpiWidgetViewComponent_ng_template_2_c8y_loading_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-loading");
} }
function KpiWidgetViewComponent_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 17);
    ɵngcc0.ɵɵtemplate(1, KpiWidgetViewComponent_ng_template_2_c8y_ui_empty_state_1_Template, 3, 8, "c8y-ui-empty-state", 18);
    ɵngcc0.ɵɵtemplate(2, KpiWidgetViewComponent_ng_template_2_c8y_loading_2_Template, 1, 0, "c8y-loading", 19);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.noDataInitiallyInDB);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r2.noDataInitiallyInDB);
} }
'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || Array.prototype.slice.call(from));
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    function exactlyASingleDatapointActive() {
        return function (control) {
            var datapoints = control.value;
            if (!datapoints || !datapoints.length) {
                return null;
            }
            var activeDatapoints = datapoints.filter(function (datapoint) { return datapoint.__active; });
            if (activeDatapoints.length === 1) {
                return null;
            }
            return { exactlyOneDatapointNeedsToBeActive: true };
        };
    }
    var KpiWidgetConfigComponent = /** @class */ (function () {
        function KpiWidgetConfigComponent(formBuilder, form, iconSelector, widgetConfig, datapointLibrary) {
            this.formBuilder = formBuilder;
            this.form = form;
            this.iconSelector = iconSelector;
            this.widgetConfig = widgetConfig;
            this.datapointLibrary = datapointLibrary;
            this.datapointSelectionConfig = {};
            this.defaultFormOptions = {
                showRedRange: true,
                showYellowRange: true
            };
            this.availableIcons = [];
            this.limits = {
                fontSizeMax: 72,
                fontSizeMin: 18,
                numberOfDecimalPlacesMax: 10,
                numberOfDecimalPlacesMin: 0
            };
        }
        KpiWidgetConfigComponent.prototype.onBeforeSave = function (config) {
            if (this.formGroup.valid) {
                Object.assign(config, this.formGroup.value);
                return true;
            }
            return false;
        };
        KpiWidgetConfigComponent.prototype.ngOnInit = function () {
            var _a, _b, _c, _d;
            return __awaiter(this, void 0, void 0, function () {
                var _e;
                return __generator(this, function (_f) {
                    switch (_f.label) {
                        case 0:
                            if ((_a = this.widgetConfig.context) === null || _a === void 0 ? void 0 : _a.id) {
                                this.datapointSelectionConfig.contextAsset = (_b = this.widgetConfig) === null || _b === void 0 ? void 0 : _b.context;
                            }
                            this.initForm();
                            if (!((_c = this.config) === null || _c === void 0 ? void 0 : _c.datapoints)) return [3 /*break*/, 2];
                            _e = this.config;
                            return [4 /*yield*/, this.datapointLibrary.updateDatapoints((_d = this.config) === null || _d === void 0 ? void 0 : _d.datapoints)];
                        case 1:
                            _e.datapoints = _f.sent();
                            this.formGroup.patchValue({ datapoints: this.config.datapoints });
                            _f.label = 2;
                        case 2: return [2 /*return*/];
                    }
                });
            });
        };
        KpiWidgetConfigComponent.prototype.openIconSelector = function () {
            return __awaiter(this, void 0, void 0, function () {
                var icon, _a_1;
                return __generator(this, function (_e) {
                    switch (_e.label) {
                        case 0:
                            _e.trys.push([0, 2, , 3]);
                            return [4 /*yield*/, this.iconSelector.selectIcon({
                                    currentSelection: this.formGroup.value.icon
                                })];
                        case 1:
                            icon = _e.sent();
                            this.formGroup.patchValue({ icon: icon });
                            return [3 /*break*/, 3];
                        case 2:
                            _a_1 = _e.sent();
                            return [3 /*break*/, 3];
                        case 3: return [2 /*return*/];
                    }
                });
            });
        };
        KpiWidgetConfigComponent.prototype.initForm = function () {
            this.formGroup = this.formBuilder.group({
                numberOfDecimalPlaces: [
                    1,
                    [
                        forms.Validators.required,
                        forms.Validators.min(this.limits.numberOfDecimalPlacesMin),
                        forms.Validators.max(this.limits.numberOfDecimalPlacesMax)
                    ]
                ],
                showTimestamp: [true, []],
                showTrend: [true, []],
                showIcon: [true, []],
                icon: ['water', [forms.Validators.required, forms.Validators.minLength(1)]],
                fontSize: [
                    36,
                    [
                        forms.Validators.required,
                        forms.Validators.min(this.limits.fontSizeMin),
                        forms.Validators.max(this.limits.fontSizeMax)
                    ]
                ],
                datapoints: [
                    [],
                    [forms.Validators.required, forms.Validators.minLength(1), exactlyASingleDatapointActive()]
                ]
            });
            this.form.form.addControl('config', this.formGroup);
            this.formGroup.patchValue(this.config);
        };
KpiWidgetConfigComponent.ɵfac = function KpiWidgetConfigComponent_Factory(t) { return new (t || KpiWidgetConfigComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.FormBuilder), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.NgForm), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.IconSelectorService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.WidgetConfigComponent), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.DatapointLibraryService)); };
KpiWidgetConfigComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: KpiWidgetConfigComponent, selectors: [["c8y-kpi-widget-config"]], inputs: { config: "config" }, features: [ɵngcc0.ɵɵProvidersFeature([], [{ provide: forms.ControlContainer, useExisting: forms.NgForm }])], decls: 56, vars: 44, consts: [[1, "row", 3, "formGroup"], [1, "col-sm-6"], ["formControlName", "datapoints", "name", "datapoints", 3, "defaultFormOptions", "config", "minActiveCount", "maxActiveCount"], ["translate", ""], [1, "d-flex", "a-i-center"], [1, "p-r-8", "icon-32", "text-muted"], [3, "c8yIcon"], [1, "btn", "btn-default", "btn-xs", 3, "click"], ["translate", "", 3, "title"], ["formControlName", "numberOfDecimalPlaces", "name", "numberOfDecimalPlaces", "type", "number", 1, "form-control", 3, "placeholder"], [3, "show"], [1, "d-flex", "gap-16", "flex-wrap"], [1, "c8y-checkbox", 3, "title"], ["type", "checkbox", "formControlName", "showTimestamp", "name", "showTimestamp"], ["type", "checkbox", "formControlName", "showIcon", "name", "showIcon"], ["type", "checkbox", "formControlName", "showTrend", "name", "showTrend"], ["container", "body", 1, "btn", "btn-clean", 3, "popover", "outsideClick"], [1, "text-info", 3, "c8yIcon"], ["formControlName", "fontSize", "name", "fontSize", "type", "number", 1, "form-control", 3, "placeholder"]], template: function KpiWidgetConfigComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "form", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelement(2, "c8y-datapoint-selection-list", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "div", 1);
        ɵngcc0.ɵɵelementStart(4, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(5, "label", 3);
        ɵngcc0.ɵɵtext(6, "Icon");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(7, "div", 4);
        ɵngcc0.ɵɵelementStart(8, "div", 5);
        ɵngcc0.ɵɵelement(9, "i", 6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(10, "button", 7);
        ɵngcc0.ɵɵlistener("click", function KpiWidgetConfigComponent_Template_button_click_10_listener() { return ctx.openIconSelector(); });
        ɵngcc0.ɵɵtext(11);
        ɵngcc0.ɵɵpipe(12, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(13, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(14, "label", 8);
        ɵngcc0.ɵɵpipe(15, "translate");
        ɵngcc0.ɵɵtext(16, " Number of decimal places ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(17, "input", 9);
        ɵngcc0.ɵɵpipe(18, "translate");
        ɵngcc0.ɵɵelement(19, "c8y-messages", 10);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(20, "div");
        ɵngcc0.ɵɵelementStart(21, "label");
        ɵngcc0.ɵɵtext(22);
        ɵngcc0.ɵɵpipe(23, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(24, "div", 11);
        ɵngcc0.ɵɵelementStart(25, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(26, "label", 12);
        ɵngcc0.ɵɵpipe(27, "translate");
        ɵngcc0.ɵɵelement(28, "input", 13);
        ɵngcc0.ɵɵelement(29, "span");
        ɵngcc0.ɵɵelementStart(30, "span", 3);
        ɵngcc0.ɵɵtext(31, "Show timestamp");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(32, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(33, "label", 12);
        ɵngcc0.ɵɵpipe(34, "translate");
        ɵngcc0.ɵɵelement(35, "input", 14);
        ɵngcc0.ɵɵelement(36, "span");
        ɵngcc0.ɵɵelementStart(37, "span", 3);
        ɵngcc0.ɵɵtext(38, "Show icon");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(39, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(40, "label", 12);
        ɵngcc0.ɵɵpipe(41, "translate");
        ɵngcc0.ɵɵelement(42, "input", 15);
        ɵngcc0.ɵɵelement(43, "span");
        ɵngcc0.ɵɵelementStart(44, "span", 3);
        ɵngcc0.ɵɵtext(45, "Show trend icon");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(46, "button", 16);
        ɵngcc0.ɵɵpipe(47, "translate");
        ɵngcc0.ɵɵelement(48, "i", 17);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(49, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(50, "label", 8);
        ɵngcc0.ɵɵpipe(51, "translate");
        ɵngcc0.ɵɵtext(52, " Font size of measurement value (px) ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(53, "input", 18);
        ɵngcc0.ɵɵpipe(54, "translate");
        ɵngcc0.ɵɵelement(55, "c8y-messages", 10);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("formGroup", ctx.formGroup);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("defaultFormOptions", ctx.defaultFormOptions)("config", ctx.datapointSelectionConfig)("minActiveCount", 1)("maxActiveCount", 1);
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵproperty("c8yIcon", ctx.formGroup.value.icon);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(12, 20, "Change"), " ");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(15, 22, "Number of decimal places"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("placeholder", ɵngcc0.ɵɵpipeBind2(18, 24, "e.g. {{ example }}", ɵngcc0.ɵɵpureFunction0(42, _c0)));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("show", (ctx.formGroup.controls == null ? null : ctx.formGroup.controls.numberOfDecimalPlaces == null ? null : ctx.formGroup.controls.numberOfDecimalPlaces.touched) && (ctx.formGroup == null ? null : ctx.formGroup.controls == null ? null : ctx.formGroup.controls.numberOfDecimalPlaces == null ? null : ctx.formGroup.controls.numberOfDecimalPlaces.errors));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(23, 27, "Display"), "");
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(27, 29, "Show timestamp"));
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(34, 31, "Show icon"));
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(41, 33, "Show trend icon"));
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵpropertyInterpolate("popover", ɵngcc0.ɵɵpipeBind1(47, 35, "Indicates the trend between the last two measurement values."));
        ɵngcc0.ɵɵproperty("outsideClick", true);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yIcon", "question-circle-o");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(51, 37, "Font size of measurement value (px)"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("placeholder", ɵngcc0.ɵɵpipeBind2(54, 39, "e.g. {{ example }}", ɵngcc0.ɵɵpureFunction0(43, _c1)));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("show", (ctx.formGroup.controls == null ? null : ctx.formGroup.controls.fontSize == null ? null : ctx.formGroup.controls.fontSize.touched) && (ctx.formGroup == null ? null : ctx.formGroup.controls == null ? null : ctx.formGroup.controls.fontSize == null ? null : ctx.formGroup.controls.fontSize.errors));
    } }, directives: [ɵngcc1.ɵNgNoValidate, ɵngcc1.NgControlStatusGroup, ɵngcc1.FormGroupDirective, ɵngcc4.DatapointSelectionListComponent, ɵngcc1.NgControlStatus, ɵngcc1.FormControlName, ɵngcc5.FormGroupComponent, ɵngcc5.C8yTranslateDirective, ɵngcc5.IconDirective, ɵngcc1.DefaultValueAccessor, ɵngcc5.RequiredInputPlaceholderDirective, ɵngcc1.NumberValueAccessor, ɵngcc5.MessagesComponent, ɵngcc1.CheckboxControlValueAccessor, ɵngcc6.PopoverDirective], pipes: [ɵngcc5.C8yTranslatePipe], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(KpiWidgetConfigComponent, [{
        type: core.Component,
        args: [{
                selector: 'c8y-kpi-widget-config',
                template: "<form [formGroup]=\"formGroup\" class=\"row\">\n  <div class=\"col-sm-6\">\n    <c8y-datapoint-selection-list\n      [defaultFormOptions]=\"defaultFormOptions\"\n      [config]=\"datapointSelectionConfig\"\n      [minActiveCount]=\"1\"\n      [maxActiveCount]=\"1\"\n      formControlName=\"datapoints\"\n      name=\"datapoints\"\n    >\n    </c8y-datapoint-selection-list>\n  </div>\n  <div class=\"col-sm-6\">\n    <c8y-form-group>\n      <label translate>Icon</label>\n      <div class=\"d-flex a-i-center\">\n        <div class=\"p-r-8 icon-32 text-muted\">\n          <i [c8yIcon]=\"formGroup.value.icon\"></i>\n        </div>\n        <button class=\"btn btn-default btn-xs\" (click)=\"openIconSelector()\">\n          {{ 'Change' | translate }}\n        </button>\n      </div>\n    </c8y-form-group>\n    <c8y-form-group>\n      <label [title]=\"'Number of decimal places' | translate\" translate>\n        Number of decimal places\n      </label>\n      <input\n        class=\"form-control\"\n        formControlName=\"numberOfDecimalPlaces\"\n        name=\"numberOfDecimalPlaces\"\n        type=\"number\"\n        [placeholder]=\"'e.g. {{ example }}' | translate: { example: 1 }\"\n      />\n      <c8y-messages [show]=\"formGroup.controls?.numberOfDecimalPlaces?.touched && formGroup?.controls?.numberOfDecimalPlaces?.errors\">\n      </c8y-messages>\n    </c8y-form-group>\n    <div>\n      <label> {{ 'Display' | translate }}</label>\n      <div class=\"d-flex gap-16 flex-wrap\">\n        <c8y-form-group>\n          <label [title]=\"'Show timestamp' | translate\" class=\"c8y-checkbox\">\n            <input type=\"checkbox\" formControlName=\"showTimestamp\" name=\"showTimestamp\" />\n            <span></span>\n            <span translate>Show timestamp</span>\n          </label>\n        </c8y-form-group>\n\n        <c8y-form-group>\n          <label [title]=\"'Show icon' | translate\" class=\"c8y-checkbox\">\n            <input type=\"checkbox\" formControlName=\"showIcon\" name=\"showIcon\" />\n            <span></span>\n            <span translate>Show icon</span>\n          </label>\n        </c8y-form-group>\n\n        <c8y-form-group>\n          <label [title]=\"'Show trend icon' | translate\" class=\"c8y-checkbox\">\n            <input type=\"checkbox\" formControlName=\"showTrend\" name=\"showTrend\" />\n            <span></span>\n            <span translate>Show trend icon</span>\n            <button\n              class=\"btn btn-clean\"\n              popover=\"{{\n                'Indicates the trend between the last two measurement values.' | translate\n              }}\"\n              container=\"body\"\n              [outsideClick]=\"true\"\n            >\n              <i [c8yIcon]=\"'question-circle-o'\" class=\"text-info\"></i>\n            </button>\n          </label>\n        </c8y-form-group>\n      </div>\n\n      <c8y-form-group>\n        <label [title]=\"'Font size of measurement value (px)' | translate\" translate>\n          Font size of measurement value (px)\n        </label>\n        <input\n          class=\"form-control\"\n          formControlName=\"fontSize\"\n          name=\"fontSize\"\n          type=\"number\"\n          [placeholder]=\"'e.g. {{ example }}' | translate: { example: 36 }\"\n        />\n        <c8y-messages [show]=\"formGroup.controls?.fontSize?.touched && formGroup?.controls?.fontSize?.errors\">\n        </c8y-messages>\n      </c8y-form-group>\n    </div>\n  </div>\n</form>\n",
                viewProviders: [{ provide: forms.ControlContainer, useExisting: forms.NgForm }]
            }]
    }], function () { return [{ type: ɵngcc1.FormBuilder }, { type: ɵngcc1.NgForm }, { type: ɵngcc2.IconSelectorService }, { type: ɵngcc3.WidgetConfigComponent }, { type: ɵngcc4.DatapointLibraryService }]; }, { config: [{
            type: core.Input
        }] }); })();
        return KpiWidgetConfigComponent;
    }());
    KpiWidgetConfigComponent.ctorParameters = function () { return [
        { type: forms.FormBuilder },
        { type: forms.NgForm },
        { type: iconSelector.IconSelectorService },
        { type: contextDashboard.WidgetConfigComponent },
        { type: datapointSelector.DatapointLibraryService }
    ]; };
    KpiWidgetConfigComponent.propDecorators = {
        config: [{ type: core.Input }]
    };

    var ColorClass;
    (function (ColorClass) {
        ColorClass["danger"] = "text-danger";
        ColorClass["warning"] = "text-warning";
        ColorClass["unknown"] = "";
    })(ColorClass || (ColorClass = {}));
    var KpiWidgetViewComponent = /** @class */ (function () {
        function KpiWidgetViewComponent(measurementRealtime, datapointLib, dashboard) {
            this.measurementRealtime = measurementRealtime;
            this.datapointLib = datapointLib;
            this.dashboard = dashboard;
            this.config = { datapoints: [] };
            this.state$ = rxjs.NEVER;
            // used to differentiate between loading state and empty state
            this.noDataInitiallyInDB = false;
        }
        KpiWidgetViewComponent.prototype.ngOnInit = function () {
            return __awaiter(this, void 0, void 0, function () {
                var datapoints, datapoint;
                return __generator(this, function (_c) {
                    datapoints = this.config.datapoints || [];
                    datapoint = datapoints.find(function (tmp) { return tmp.__active; });
                    if (!datapoint) {
                        return [2 /*return*/];
                    }
                    this.state$ = this.setupObservable(datapoint);
                    return [2 /*return*/];
                });
            });
        };
        KpiWidgetViewComponent.prototype.setupObservable = function (datapoint) {
            this.assignContextFromContextDashboard(datapoint);
            var latestMeasurement$ = this.getLatestMeasurement$(datapoint);
            var lastTwoValues$ = this.getLastTwoValuesOfObservable$(latestMeasurement$);
            var previousValue$ = lastTwoValues$.pipe(operators.map(function (res) { return res[0]; }), operators.startWith(undefined));
            var datapointUpdate$ = this.getDatapointLibraryEntry$(datapoint);
            var combineMeasurementAndDatapointLibEntry$ = rxjs.combineLatest([
                latestMeasurement$,
                datapointUpdate$
            ]);
            var unit$ = combineMeasurementAndDatapointLibEntry$.pipe(operators.map(function (_c) {
                var _d = __read(_c, 2), latestMeasurementValue = _d[0], currentDatapoint = _d[1];
                return currentDatapoint.unit || latestMeasurementValue.unit || '';
            }), operators.startWith(''), operators.distinctUntilChanged());
            return rxjs.combineLatest([
                latestMeasurement$,
                previousValue$,
                this.getTrendOfLatestMeasurements$(lastTwoValues$),
                unit$,
                this.getColorClass$(combineMeasurementAndDatapointLibEntry$)
            ]).pipe(operators.map(function (_c) {
                var _d = __read(_c, 5), latestMeasurement = _d[0], previousValue = _d[1], trend = _d[2], unit = _d[3], colorClass = _d[4];
                return {
                    latestMeasurement: latestMeasurement,
                    previousValue: previousValue,
                    trend: trend,
                    unit: unit,
                    colorClass: colorClass
                };
            }));
        };
        KpiWidgetViewComponent.prototype.getLatestMeasurement$ = function (datapoint) {
            var _this = this;
            return this.measurementRealtime
                .latestValueOfSpecificMeasurement$(datapoint.fragment, datapoint.series, datapoint.__target, 
            // we only need the last two values in case we want to show a trend
            this.config.showTrend ? 2 : 1, 
            // null will be emitted in case no measurement was found initially
            true)
                .pipe(operators.tap(function (measurement) {
                if (!measurement) {
                    _this.noDataInitiallyInDB = true;
                }
            }), operators.filter(function (measurement) { return !!measurement; }), operators.map(function (measurement) {
                return {
                    unit: measurement[datapoint.fragment][datapoint.series].unit,
                    value: measurement[datapoint.fragment][datapoint.series].value,
                    date: measurement.time
                };
            }));
        };
        KpiWidgetViewComponent.prototype.getDatapointLibraryEntry$ = function (datapoint) {
            return rxjs.from(this.datapointLib.updateDatapoints([datapoint], true)).pipe(operators.map(function (tmp) { return tmp[0]; }), operators.filter(function (tmp) { return !!tmp; }), operators.startWith(datapoint));
        };
        KpiWidgetViewComponent.prototype.getColorClass$ = function (measurementAndDatapointCombination$) {
            var _this = this;
            return measurementAndDatapointCombination$.pipe(operators.map(function (_c) {
                var _d = __read(_c, 2), latestMeasurementValue = _d[0], currentDatapoint = _d[1];
                if (_this.inRangeOf(currentDatapoint, latestMeasurementValue.value, 'redRangeMin', 'redRangeMax')) {
                    return ColorClass.danger;
                }
                if (_this.inRangeOf(currentDatapoint, latestMeasurementValue.value, 'yellowRangeMin', 'yellowRangeMax')) {
                    return ColorClass.warning;
                }
                return ColorClass.unknown;
            }), operators.startWith(ColorClass.unknown), operators.distinctUntilChanged());
        };
        KpiWidgetViewComponent.prototype.getLastTwoValuesOfObservable$ = function (input$) {
            return input$.pipe(operators.pairwise());
        };
        KpiWidgetViewComponent.prototype.getTrendOfLatestMeasurements$ = function (latestMeasurement$) {
            return latestMeasurement$.pipe(operators.map(function (res) {
                if (res.length === 2) {
                    var oldValue = res[0].value;
                    var newValue = res[1].value;
                    if (oldValue < newValue) {
                        return '45deg';
                    }
                    if (oldValue > newValue) {
                        return '135deg';
                    }
                }
                return '90deg';
            }), operators.startWith('90deg'), operators.distinctUntilChanged());
        };
        KpiWidgetViewComponent.prototype.inRangeOf = function (datapoint, measurementValue, minAttribute, maxAttribute) {
            if (typeof datapoint[minAttribute] === 'number' &&
                typeof datapoint[maxAttribute] === 'number') {
                if (measurementValue >= datapoint[minAttribute] &&
                    measurementValue < datapoint[maxAttribute]) {
                    return true;
                }
            }
            return false;
        };
        KpiWidgetViewComponent.prototype.assignContextFromContextDashboard = function (datapoint) {
            var _a, _b;
            if (!((_a = this.dashboard) === null || _a === void 0 ? void 0 : _a.isDeviceTypeDashboard)) {
                return;
            }
            var context = (_b = this.dashboard) === null || _b === void 0 ? void 0 : _b.context;
            if (context === null || context === void 0 ? void 0 : context.id) {
                var name = context.name, id = context.id;
                datapoint.__target = { name: name, id: id };
            }
        };
KpiWidgetViewComponent.ɵfac = function KpiWidgetViewComponent_Factory(t) { return new (t || KpiWidgetViewComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc5.MeasurementRealtimeService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.DatapointLibraryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.ContextDashboardComponent, 8)); };
KpiWidgetViewComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: KpiWidgetViewComponent, selectors: [["c8y-kpi-widget-view"]], inputs: { config: "config" }, features: [ɵngcc0.ɵɵProvidersFeature([ngxComponents.MeasurementRealtimeService])], decls: 4, vars: 4, consts: [["class", "kpi-widget__container d-flex d-col fit-h fit-w a-i-center j-c-center", 4, "ngIf", "ngIfElse"], ["noMeasurementFound", ""], [1, "kpi-widget__container", "d-flex", "d-col", "fit-h", "fit-w", "a-i-center", "j-c-center"], [1, "d-flex", "a-i-center", "j-c-center", "fit-w"], ["class", "m-r-16 flex-no-shrink text-muted", 3, "ngClass", 4, "ngIf"], [1, "text-truncate"], [1, "text-truncate", "text-medium", 3, "ngClass", "ngStyle", "title"], [1, "text-regular"], ["class", "dot dot-info dot-30 m-l-16 flex-no-shrink", 4, "ngIf"], [1, "d-flex", "j-c-center"], ["class", "icon-flex text-center text-muted small", 4, "ngIf"], [1, "m-r-16", "flex-no-shrink", "text-muted", 3, "ngClass"], [1, "icon-32", 3, "c8yIcon"], [1, "dot", "dot-info", "dot-30", "m-l-16", "flex-no-shrink"], ["c8yIcon", "arrow-dotted-up", 1, "icon-20", 3, "title", "ngStyle"], [1, "icon-flex", "text-center", "text-muted", "small"], ["c8yIcon", "calendar"], [1, "d-flex", "fit-h", "fit-w", "j-c-center", "a-i-center"], ["class", "fit-w", 3, "icon", "title", "subtitle", "horizontal", 4, "ngIf"], [4, "ngIf"], [1, "fit-w", 3, "icon", "title", "subtitle", "horizontal"]], template: function KpiWidgetViewComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, KpiWidgetViewComponent_div_0_Template, 15, 22, "div", 0);
        ɵngcc0.ɵɵpipe(1, "async");
        ɵngcc0.ɵɵtemplate(2, KpiWidgetViewComponent_ng_template_2_Template, 3, 2, "ng-template", null, 1, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        var _r1 = ɵngcc0.ɵɵreference(3);
        ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(1, 2, ctx.state$))("ngIfElse", _r1);
    } }, directives: [ɵngcc7.NgIf, ɵngcc7.NgClass, ɵngcc7.NgStyle, ɵngcc5.IconDirective, ɵngcc5.EmptyStateComponent, ɵngcc5.LoadingComponent], pipes: [ɵngcc7.AsyncPipe, ɵngcc5.C8yTranslatePipe, ɵngcc7.DecimalPipe, ɵngcc7.DatePipe], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(KpiWidgetViewComponent, [{
        type: core.Component,
        args: [{
                selector: 'c8y-kpi-widget-view',
                template: "<div\n  class=\"kpi-widget__container d-flex d-col fit-h fit-w a-i-center j-c-center\"\n  *ngIf=\"state$ | async as lastState; else noMeasurementFound\"\n>\n  <div class=\"d-flex a-i-center j-c-center fit-w\">\n    <div\n      class=\"m-r-16 flex-no-shrink text-muted\"\n      [ngClass]=\"lastState.colorClass\"\n      *ngIf=\"config.icon && config.showIcon\"\n    >\n      <i class=\"icon-32\" [c8yIcon]=\"config.icon\"></i>\n    </div>\n    <div class=\"text-truncate\">\n      <span\n        class=\"text-truncate text-medium\"\n        [ngClass]=\"lastState.colorClass\"\n        [ngStyle]=\"{ 'font-size': (config.fontSize || '36') + 'px' }\"\n        title=\"{{\n          lastState.colorClass === 'text-danger'\n            ? ('Within red range:' | translate)\n            : lastState.colorClass === 'text-warning'\n            ? ('Within yellow range:' | translate)\n            : ''\n        }} {{\n          lastState.latestMeasurement.value\n            | number\n              : '1.' +\n                  (config.numberOfDecimalPlaces || '0') +\n                  '-' +\n                  (config.numberOfDecimalPlaces || '0')\n        }} {{ lastState.unit || '' }}\"\n      >\n        {{\n          lastState.latestMeasurement.value\n            | number\n              : '1.' +\n                  (config.numberOfDecimalPlaces || '0') +\n                  '-' +\n                  (config.numberOfDecimalPlaces || '0')\n        }}\n        <small class=\"text-regular\">{{ lastState.unit || '' }}</small>\n      </span>\n    </div>\n    <div\n      class=\"dot dot-info dot-30 m-l-16 flex-no-shrink\"\n      *ngIf=\"config?.showTrend && lastState.previousValue as previousValue\"\n    >\n      <i\n        class=\"icon-20\"\n        [title]=\"\n          ('Previous value' | translate) +\n          ': ' +\n          (previousValue.value\n            | number\n              : '1.' +\n                  (config.numberOfDecimalPlaces || '0') +\n                  '-' +\n                  (config.numberOfDecimalPlaces || '0')) +\n          ' (' +\n          (previousValue.date | date: 'medium') +\n          ')'\n        \"\n        c8yIcon=\"arrow-dotted-up\"\n        [ngStyle]=\"{ transform: 'rotate(' + lastState.trend + ')' }\"\n      ></i>\n    </div>\n  </div>\n  <div class=\"d-flex j-c-center\">\n    <p *ngIf=\"config?.showTimestamp\" class=\"icon-flex text-center text-muted small\">\n      <i c8yIcon=\"calendar\"></i>\n      {{ lastState.latestMeasurement.date | date: 'medium' }}\n    </p>\n  </div>\n</div>\n\n<ng-template #noMeasurementFound>\n  <div class=\"d-flex fit-h fit-w j-c-center a-i-center\">\n    <c8y-ui-empty-state\n      *ngIf=\"noDataInitiallyInDB\"\n      class=\"fit-w\"\n      [icon]=\"'line-chart'\"\n      [title]=\"'No measurement to display.' | translate\"\n      [subtitle]=\"'Waiting for measurements to be created.' | translate\"\n      [horizontal]=\"true\"\n    ></c8y-ui-empty-state>\n    <c8y-loading *ngIf=\"!noDataInitiallyInDB\"></c8y-loading>\n  </div>\n</ng-template>\n",
                providers: [ngxComponents.MeasurementRealtimeService]
            }]
    }], function () { return [{ type: ɵngcc5.MeasurementRealtimeService }, { type: ɵngcc4.DatapointLibraryService }, { type: ɵngcc3.ContextDashboardComponent, decorators: [{
                type: core.Optional
            }] }]; }, { config: [{
            type: core.Input
        }] }); })();
        return KpiWidgetViewComponent;
    }());
    KpiWidgetViewComponent.ctorParameters = function () { return [
        { type: ngxComponents.MeasurementRealtimeService },
        { type: datapointSelector.DatapointLibraryService },
        { type: contextDashboard.ContextDashboardComponent, decorators: [{ type: core.Optional }] }
    ]; };
    KpiWidgetViewComponent.propDecorators = {
        config: [{ type: core.Input }]
    };

    // import for the preview img to get into the build.
    var previewImage = '';
    try {
        // tslint:disable-next-line: no-var-requires
        previewImage = require('@c8y/style/img/kpi-widget.png');
    }
    catch (ex) {
        // intended empty
    }
    var ɵ0 = [
        {
            id: 'kpi.widget',
            label: ngxComponents.gettext('KPI Widget'),
            description: ngxComponents.gettext('Displays the last measurement value of a device.'),
            component: KpiWidgetViewComponent,
            configComponent: KpiWidgetConfigComponent,
            previewImage: previewImage,
            data: {
                settings: {
                    noNewWidgets: false,
                    widgetDefaults: {
                        _width: 2,
                        _height: 2
                    },
                    ng1: {
                        options: {
                            noDeviceTarget: true,
                            groupsSelectable: false
                        }
                    }
                }
            }
        }
    ];
    var KpiWidgetModule = /** @class */ (function () {
        function KpiWidgetModule() {
        }
KpiWidgetModule.ɵfac = function KpiWidgetModule_Factory(t) { return new (t || KpiWidgetModule)(); };
KpiWidgetModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: KpiWidgetModule });
KpiWidgetModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        {
            provide: ngxComponents.HOOK_COMPONENTS,
            useValue: ɵ0,
            multi: true
        }
    ], imports: [[ngxComponents.CoreModule, datapointSelector.DatapointSelectorModule, iconSelector.IconSelectorModule, popover.PopoverModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(KpiWidgetModule, [{
        type: core.NgModule,
        args: [{
                imports: [ngxComponents.CoreModule, datapointSelector.DatapointSelectorModule, iconSelector.IconSelectorModule, popover.PopoverModule],
                declarations: [KpiWidgetViewComponent, KpiWidgetConfigComponent],
                providers: [
                    {
                        provide: ngxComponents.HOOK_COMPONENTS,
                        useValue: ɵ0,
                        multi: true
                    }
                ]
            }]
    }], function () { return []; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(KpiWidgetModule, { declarations: [KpiWidgetViewComponent, KpiWidgetConfigComponent], imports: [ɵngcc5.CoreModule, ɵngcc4.DatapointSelectorModule, ɵngcc2.IconSelectorModule, ɵngcc6.PopoverModule] }); })();
        return KpiWidgetModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.KpiWidgetModule = KpiWidgetModule;
    exports["ɵ0"] = ɵ0;
    exports["ɵa"] = KpiWidgetViewComponent;
    exports["ɵb"] = KpiWidgetConfigComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
