// @c8y/ngx-components/bundles/c8y-ngx-components-search.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@c8y/ngx-components'),require('ngx-bootstrap/modal'),require('@c8y/client'),require('@c8y/ngx-components/sub-assets'),require('@angular/common'),require('@angular/router'),require('@c8y/ngx-components/device-grid'),exports, require('@angular/core'), require('@c8y/ngx-components/sub-assets'), require('@ngx-translate/core'), require('@c8y/client'), require('@c8y/client'), require('@c8y/ngx-components'), require('@c8y/ngx-components/assets-navigator'), require('@c8y/client'), require('@c8y/client'), require('@c8y/ngx-components/device-grid'), require('@angular/router'), require('rxjs'), require('rxjs/operators'), require('@c8y/client'), require('ngx-bootstrap/modal')) :
    typeof define === 'function' && define.amd ? define('@c8y/ngx-components/search', ['@angular/core','@c8y/ngx-components','ngx-bootstrap/modal','@c8y/client','@c8y/ngx-components/sub-assets','@angular/common','@angular/router','@c8y/ngx-components/device-grid','exports', '@angular/core', '@c8y/ngx-components/sub-assets', '@ngx-translate/core', '@c8y/client', '@c8y/client', '@c8y/ngx-components', '@c8y/ngx-components/assets-navigator', '@c8y/client', '@c8y/client', '@c8y/ngx-components/device-grid', '@angular/router', 'rxjs', 'rxjs/operators', '@c8y/client', 'ngx-bootstrap/modal'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.c8y.ngxComponents,global.ngxBootstrap.modal,global.c8y.client,global.c8y.ngxComponents.subAssets,global.ng.common,global.ng.router,global.c8y.ngxComponents.deviceGrid,(global.c8y = global.c8y || {}, global.c8y["ngx-components"] = global.c8y["ngx-components"] || {}, global.c8y["ngx-components"].search = {}), global.ng.core, global.c8y["ngx-components"]["sub-assets"], global.i1, global.i2, global.i3, global.c8y["ngx-components"], global.i5, global.i6, global.i7, global.c8y["ngx-components"]["device-grid"], global.ng.router, global.rxjs, global.rxjs.operators, global.client, global.modal));
})(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,ɵngcc4,ɵngcc5,ɵngcc6,ɵngcc7,exports, i0, subAssets, i1, i2, i3, i4, i5, i6, i7, deviceGrid, router, rxjs, operators, client, modal) { 
function SearchGridComponent_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "c8y-column", 4);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var column_r1 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("name", column_r1.name);
} }
var _c0 = function (a0) { return { searchHint: a0 }; };
function SearchResultsComponent_small_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 3);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(1, "searching \"{{ searchHint }}\"");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction1(1, _c0, ctx_r0.searchText));
} }
var _c1 = function (a0) { return { filterHint: a0 }; };
function SearchResultsComponent_small_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 3);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(1, "filtered by \"{{ filterHint }}\"");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction1(1, _c1, ctx_r1.filter));
} }
'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);
    var i5__namespace = /*#__PURE__*/_interopNamespace(i5);
    var i6__namespace = /*#__PURE__*/_interopNamespace(i6);
    var i7__namespace = /*#__PURE__*/_interopNamespace(i7);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || Array.prototype.slice.call(from));
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var AssetSearchService = /** @class */ (function (_super) {
        __extends(AssetSearchService, _super);
        function AssetSearchService() {
            var _this = _super.apply(this, __spreadArray([], __read(arguments))) || this;
            _this.GRID_CONFIG_STORAGE_KEY = 'search-grid-config';
            _this.DEFAULT_PAGE_SIZE = 50;
            return _this;
        }
        AssetSearchService.prototype.buildCombinedRootQueryFilter = function (columns, pagination) {
            var rootQuery = {
                __filter: {
                    __and: { __not: { __has: "c8y_IsBinary" } }
                }
            };
            var userQuery = this.getQueryObj(columns, pagination);
            var queryPart = this.queriesUtil.addOrderbys(rootQuery, userQuery.__orderby, 'append');
            var fullQuery = this.queriesUtil.addAndFilter(queryPart, userQuery.__filter);
            return this.queriesUtil.buildQuery(fullQuery);
        };
        AssetSearchService.prototype.getData = function (columns, pagination) {
            return __awaiter(this, void 0, void 0, function () {
                var query;
                return __generator(this, function (_a) {
                    query = this.buildCombinedRootQueryFilter(columns, pagination);
                    return [2 /*return*/, this.assetNodeService.getAllInventories(Object.assign(Object.assign({}, pagination), { query: query }))];
                });
            });
        };
AssetSearchService.ɵfac = /*@__PURE__*/ function () { var ɵAssetSearchService_BaseFactory; return function AssetSearchService_Factory(t) { return (ɵAssetSearchService_BaseFactory || (ɵAssetSearchService_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(AssetSearchService)))(t || AssetSearchService); }; }();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AssetSearchService, [{
        type: i0.Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], null, null); })();
        return AssetSearchService;
    }(subAssets.SubAssetsService));
    AssetSearchService.ɵprov = i0__namespace.ɵɵdefineInjectable({ factory: function AssetSearchService_Factory() { return new AssetSearchService(i0__namespace.ɵɵinject(i1__namespace.TranslateService), i0__namespace.ɵɵinject(i2__namespace.InventoryService), i0__namespace.ɵɵinject(i3__namespace.UserService), i0__namespace.ɵɵinject(i4__namespace.AppStateService), i0__namespace.ɵɵinject(i3__namespace.UserService), i0__namespace.ɵɵinject(i5__namespace.AssetNodeService), i0__namespace.ɵɵinject(i6__namespace.SmartGroupsService), i0__namespace.ɵɵinject(i7__namespace.SmartRulesService), i0__namespace.ɵɵinject(i4__namespace.AlertService), i0__namespace.ɵɵinject(i4__namespace.Permissions), i0__namespace.ɵɵinject(i4__namespace.ModalService), i0__namespace.ɵɵinject(i4__namespace.AssetTypesService), i0__namespace.ɵɵinject(i4__namespace.UserPreferencesService), i0__namespace.ɵɵinject(i4__namespace.GainsightService, 8)); }, token: AssetSearchService, providedIn: "root" });

    var SearchGridComponent = /** @class */ (function () {
        function SearchGridComponent(coreSearchService, searchService, bsModalService, smartGroupsService, subAssetsGridService) {
            this.coreSearchService = coreSearchService;
            this.searchService = searchService;
            this.bsModalService = bsModalService;
            this.smartGroupsService = smartGroupsService;
            this.subAssetsGridService = subAssetsGridService;
            this.title = '';
            this.loadingItemsLabel = i4.gettext('Loading results…');
            this.selectable = false;
            this.onColumnsChange = new i0.EventEmitter();
            this.searchText = '';
            this.pagination = this.searchService.getDefaultPagination();
            this.bulkActionControls = this.searchService.getDefaultBulkActionControls();
            this.refresh = new i0.EventEmitter();
            this.sizeCount = 0;
        }
        Object.defineProperty(SearchGridComponent.prototype, "_columns", {
            set: function (value) {
                if (value) {
                    this.columns = value;
                }
                else {
                    this.columns = this.searchService.getDefaultColumns();
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(SearchGridComponent.prototype, "_pagination", {
            set: function (value) {
                if (value) {
                    this.pagination = value;
                }
                else {
                    this.pagination = this.searchService.getDefaultPagination();
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(SearchGridComponent.prototype, "_actionControls", {
            set: function (value) {
                if (value) {
                    this.actionControls = value;
                }
                else {
                    this.actionControls = this.searchService.getDefaultActionControls();
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(SearchGridComponent.prototype, "_bulkActionControls", {
            set: function (value) {
                if (value) {
                    this.bulkActionControls = value;
                }
                else {
                    this.bulkActionControls = this.searchService.getDefaultBulkActionControls();
                }
            },
            enumerable: false,
            configurable: true
        });
        SearchGridComponent.prototype.ngOnInit = function () {
            if (!this.filteringName) {
                this.columns = this.searchService.getDefaultColumns();
            }
            else {
                this.columns = [
                    new subAssets.AssetTypeGridColumn({ sortOrder: 'desc' }),
                    new deviceGrid.NameDeviceGridColumn({
                        sortOrder: 'asc',
                        filter: { externalFilterQuery: { names: [this.filteringName] } }
                    }),
                    new deviceGrid.ModelDeviceGridColumn(),
                    new deviceGrid.SerialNumberDeviceGridColumn({ visible: false }),
                    new deviceGrid.RegistrationDateDeviceGridColumn({ visible: false }),
                    new deviceGrid.SystemIdDeviceGridColumn({ visible: false }),
                    new deviceGrid.ImeiDeviceGridColumn({ visible: false }),
                    new deviceGrid.AlarmsDeviceGridColumn()
                ];
            }
            this.serverSideDataCallback = this.onDataSourceModifier.bind(this);
            this.setActionControls();
        };
        SearchGridComponent.prototype.trackByName = function (_index, column) {
            return column.name;
        };
        SearchGridComponent.prototype.onDataSourceModifier = function (dataSourceModifier) {
            return __awaiter(this, void 0, void 0, function () {
                var response, res, data, paging, filteredData;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!dataSourceModifier.searchText) return [3 /*break*/, 2];
                            return [4 /*yield*/, this.coreSearchService.search(dataSourceModifier.searchText, dataSourceModifier.pagination)];
                        case 1:
                            response = _a.sent();
                            return [3 /*break*/, 4];
                        case 2: return [4 /*yield*/, this.searchService.getData(dataSourceModifier.columns, dataSourceModifier.pagination)];
                        case 3:
                            response = _a.sent();
                            _a.label = 4;
                        case 4:
                            res = response.res, data = response.data, paging = response.paging;
                            filteredData = this.coreSearchService.filterOnlyAssets(data);
                            if (paging.currentPage === 1) {
                                this.sizeCount = 0;
                            }
                            this.sizeCount += filteredData.length;
                            this.onColumnsChange.emit(dataSourceModifier.columns);
                            return [2 /*return*/, {
                                    res: res,
                                    data: filteredData,
                                    paging: paging,
                                    filteredSize: this.sizeCount,
                                    size: undefined
                                }];
                    }
                });
            });
        };
        SearchGridComponent.prototype.setActionControls = function () {
            var _this = this;
            var actionControls = [];
            var deleteAction = {
                type: "DELETE" /* Delete */,
                callback: function (asset) { return _this.onDeleteAsset(asset, _this.parentGroup); }
            };
            actionControls.push(deleteAction);
            if (!this.actionControls) {
                this.actionControls = actionControls;
            }
        };
        SearchGridComponent.prototype.updateFiltering = function (columnNames, action) {
            var type = action.type;
            if (type === i4.FilteringActionType.ResetFilter) {
                this.dataGrid.clearFilters();
            }
            else {
                this.dataGrid.updateFiltering(columnNames, action, false);
            }
        };
        SearchGridComponent.prototype.configChange = function (config) {
            this.searchService.saveConfig(config);
        };
        SearchGridComponent.prototype.onDeleteAsset = function (asset, parentRef) {
            var _this = this;
            var initialState = {
                showWithDeviceUserCheckbox: this.subAssetsGridService.shouldShowWithDeviceUserCheckbox(asset),
                asset: asset,
                showWithCascadeCheckbox: !this.smartGroupsService.isSmartGroup(asset) &&
                    !this.smartGroupsService.isSmartGroupV2(asset)
            };
            var modalRef = this.bsModalService.show(subAssets.DeleteAssetsModalComponent, { initialState: initialState });
            modalRef.content.closeSubject.subscribe(function (result) { return __awaiter(_this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!result) return [3 /*break*/, 2];
                            return [4 /*yield*/, this.subAssetsGridService.deleteAsset(asset, parentRef, result)];
                        case 1:
                            _a.sent();
                            this.refresh.emit();
                            _a.label = 2;
                        case 2: return [2 /*return*/];
                    }
                });
            }); });
        };
SearchGridComponent.ɵfac = function SearchGridComponent_Factory(t) { return new (t || SearchGridComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.InventorySearchService), ɵngcc0.ɵɵdirectiveInject(AssetSearchService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.BsModalService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.SmartGroupsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.SubAssetsService)); };
SearchGridComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SearchGridComponent, selectors: [["c8y-search-grid"]], viewQuery: function SearchGridComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(i4.DataGridComponent, 7);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.dataGrid = _t.first);
    } }, inputs: { title: "title", loadingItemsLabel: "loadingItemsLabel", selectable: "selectable", searchText: "searchText", _columns: ["columns", "_columns"], _pagination: ["pagination", "_pagination"], _actionControls: ["actionControls", "_actionControls"], _bulkActionControls: ["bulkActionControls", "_bulkActionControls"], parentGroup: ["parent-group", "parentGroup"], filteringName: "filteringName" }, outputs: { onColumnsChange: "onColumnsChange" }, decls: 7, vars: 24, consts: [[1, "card--grid--fullpage"], [3, "title", "loadingItemsLabel", "columns", "pagination", "actionControls", "selectable", "bulkActionControls", "serverSideDataCallback", "infiniteScroll", "showSearch", "searchText", "refresh", "onConfigChange"], [4, "ngFor", "ngForOf", "ngForTrackBy"], [3, "icon", "title", "subtitle", "horizontal"], [3, "name"]], template: function SearchGridComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "c8y-data-grid", 1);
        ɵngcc0.ɵɵlistener("onConfigChange", function SearchGridComponent_Template_c8y_data_grid_onConfigChange_1_listener($event) { return ctx.configChange($event); });
        ɵngcc0.ɵɵpipe(2, "translate");
        ɵngcc0.ɵɵtemplate(3, SearchGridComponent_ng_container_3_Template, 2, 1, "ng-container", 2);
        ɵngcc0.ɵɵelement(4, "c8y-ui-empty-state", 3);
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵpipe(6, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(2, 18, "Search results"))("loadingItemsLabel", ctx.loadingItemsLabel)("columns", ctx.columns)("pagination", ctx.pagination)("actionControls", ctx.actionControls)("selectable", ctx.selectable)("bulkActionControls", ctx.bulkActionControls)("serverSideDataCallback", ctx.serverSideDataCallback)("infiniteScroll", "auto")("showSearch", true)("searchText", ctx.searchText)("refresh", ctx.refresh);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.columns)("ngForTrackBy", ctx.trackByName);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("icon", "search")("title", ɵngcc0.ɵɵpipeBind1(5, 20, "No results to display."))("subtitle", ɵngcc0.ɵɵpipeBind1(6, 22, "Refine your search terms or check your spelling."))("horizontal", true);
    } }, directives: [ɵngcc1.DataGridComponent, ɵngcc5.NgForOf, ɵngcc1.EmptyStateComponent, ɵngcc1.ColumnDirective], pipes: [ɵngcc1.C8yTranslatePipe], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SearchGridComponent, [{
        type: i0.Component,
        args: [{
                selector: 'c8y-search-grid',
                template: "<div class=\"card--grid--fullpage\">\n  <c8y-data-grid\n    [title]=\"'Search results' | translate\"\n    [loadingItemsLabel]=\"loadingItemsLabel\"\n    [columns]=\"columns\"\n    [pagination]=\"pagination\"\n    [actionControls]=\"actionControls\"\n    [selectable]=\"selectable\"\n    [bulkActionControls]=\"bulkActionControls\"\n    [serverSideDataCallback]=\"serverSideDataCallback\"\n    [infiniteScroll]=\"'auto'\"\n    [showSearch]=\"true\"\n    [searchText]=\"searchText\"\n    [refresh]=\"refresh\"\n    (onConfigChange)=\"configChange($event)\"\n  >\n    <ng-container *ngFor=\"let column of columns; trackBy: trackByName\">\n      <c8y-column [name]=\"column.name\"></c8y-column>\n    </ng-container>\n\n    <c8y-ui-empty-state\n      [icon]=\"'search'\"\n      [title]=\"'No results to display.' | translate\"\n      [subtitle]=\"'Refine your search terms or check your spelling.' | translate\"\n      [horizontal]=\"true\"\n    ></c8y-ui-empty-state>\n  </c8y-data-grid>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.InventorySearchService }, { type: AssetSearchService }, { type: ɵngcc2.BsModalService }, { type: ɵngcc3.SmartGroupsService }, { type: ɵngcc4.SubAssetsService }]; }, { title: [{
            type: i0.Input
        }], loadingItemsLabel: [{
            type: i0.Input
        }], selectable: [{
            type: i0.Input
        }], onColumnsChange: [{
            type: i0.Output
        }], searchText: [{
            type: i0.Input
        }], _columns: [{
            type: i0.Input,
            args: ['columns']
        }], _pagination: [{
            type: i0.Input,
            args: ['pagination']
        }], _actionControls: [{
            type: i0.Input,
            args: ['actionControls']
        }], _bulkActionControls: [{
            type: i0.Input,
            args: ['bulkActionControls']
        }], parentGroup: [{
            type: i0.Input,
            args: ['parent-group']
        }], filteringName: [{
            type: i0.Input
        }], dataGrid: [{
            type: i0.ViewChild,
            args: [i4.DataGridComponent, { static: true }]
        }] }); })();
        return SearchGridComponent;
    }());
    SearchGridComponent.ctorParameters = function () { return [
        { type: i4.InventorySearchService },
        { type: AssetSearchService },
        { type: modal.BsModalService },
        { type: client.SmartGroupsService },
        { type: subAssets.SubAssetsService }
    ]; };
    SearchGridComponent.propDecorators = {
        parentGroup: [{ type: i0.Input, args: ['parent-group',] }],
        title: [{ type: i0.Input }],
        loadingItemsLabel: [{ type: i0.Input }],
        _columns: [{ type: i0.Input, args: ['columns',] }],
        _pagination: [{ type: i0.Input, args: ['pagination',] }],
        _actionControls: [{ type: i0.Input, args: ['actionControls',] }],
        selectable: [{ type: i0.Input }],
        _bulkActionControls: [{ type: i0.Input, args: ['bulkActionControls',] }],
        onColumnsChange: [{ type: i0.Output }],
        searchText: [{ type: i0.Input }],
        filteringName: [{ type: i0.Input }],
        dataGrid: [{ type: i0.ViewChild, args: [i4.DataGridComponent, { static: true },] }]
    };

    var SearchResultsComponent = /** @class */ (function () {
        function SearchResultsComponent(route, alert) {
            this.route = route;
            this.alert = alert;
            this.filter = '';
            this.searchText = '';
            this.WARNING_TIMEOUT_TIME = 3000;
            this.unsubscribe$ = new rxjs.Subject();
        }
        SearchResultsComponent.prototype.ngOnInit = function () {
            var _this = this;
            this.route.queryParams.pipe(operators.takeUntil(this.unsubscribe$)).subscribe(function (params) {
                if (params.filter) {
                    _this.filteringName = params.filter;
                }
            });
        };
        SearchResultsComponent.prototype.ngAfterViewInit = function () {
            var _this = this;
            this.route.queryParams
                .pipe(operators.takeUntil(this.unsubscribe$))
                .subscribe(function (_a) {
                var filter = _a.filter, search = _a.search;
                return _this.onQueryParamsChange(filter, search);
            });
            this.searchGrid.dataGrid.onBeforeSearch.pipe(operators.takeUntil(this.unsubscribe$)).subscribe(function (text) {
                if (text) {
                    _this.resetFilter();
                }
            });
            this.searchGrid.dataGrid.onBeforeFilter
                .pipe(operators.takeUntil(this.unsubscribe$))
                .subscribe(function () { return _this.resetSearch(); });
        };
        SearchResultsComponent.prototype.resetSearch = function () {
            if (this.searchGrid.dataGrid.searchText) {
                this.alert.add({
                    text: i4.gettext('Search reset. Full text search does not support filtering.'),
                    type: i4.Status.WARNING,
                    timeout: this.WARNING_TIMEOUT_TIME
                });
                this.searchText = '';
                this.searchGrid.dataGrid.searchText = '';
            }
        };
        SearchResultsComponent.prototype.resetFilter = function () {
            this.filter = '';
            if (this.searchGrid.dataGrid.filteringApplied) {
                this.alert.add({
                    text: i4.gettext('Filter reset. Full text search does not support filtering.'),
                    type: i4.Status.WARNING,
                    timeout: this.WARNING_TIMEOUT_TIME
                });
                this.searchGrid.dataGrid.clearFilters(false);
            }
        };
        SearchResultsComponent.prototype.ngOnDestroy = function () {
            this.unsubscribe$.next();
            this.unsubscribe$.complete();
        };
        SearchResultsComponent.prototype.onQueryParamsChange = function (filter, searchTerm) {
            if (!this.shouldFilter(filter) && searchTerm) {
                this.searchText = searchTerm || '';
            }
        };
        SearchResultsComponent.prototype.shouldFilter = function (filter) {
            if (!filter) {
                return false;
            }
            this.resetSearch();
            this.filter = filter || '';
            this.searchGrid.updateFiltering(['name'], {
                type: i4.FilteringActionType.ApplyFilter,
                payload: {
                    filteringModifier: {
                        externalFilterQuery: {
                            names: [this.filter]
                        }
                    }
                }
            });
            return true;
        };
SearchResultsComponent.ɵfac = function SearchResultsComponent_Factory(t) { return new (t || SearchResultsComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc6.ActivatedRoute), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AlertService)); };
SearchResultsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SearchResultsComponent, selectors: [["c8y-search-results"]], viewQuery: function SearchResultsComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(SearchGridComponent, 7);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.searchGrid = _t.first);
    } }, decls: 6, vars: 4, consts: [["translate", "", 1, "p-r-4"], ["ngNonBindable", "", "translate", "", 3, "translateParams", 4, "ngIf"], [3, "searchText", "filteringName"], ["translate", "", 3, "translateParams"]], template: function SearchResultsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵelementStart(1, "span", 0);
        ɵngcc0.ɵɵtext(2, "Search");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(3, SearchResultsComponent_small_3_Template, 2, 3, "small", 1);
        ɵngcc0.ɵɵtemplate(4, SearchResultsComponent_small_4_Template, 2, 3, "small", 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(5, "c8y-search-grid", 2);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.searchText);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.filter);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("searchText", ctx.searchText)("filteringName", ctx.filteringName);
    } }, directives: [ɵngcc1.TitleComponent, ɵngcc1.C8yTranslateDirective, ɵngcc5.NgIf, SearchGridComponent], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SearchResultsComponent, [{
        type: i0.Component,
        args: [{
                selector: 'c8y-search-results',
                template: "<c8y-title>\n  <span translate class=\"p-r-4\">Search</span>\n  <small\n    ngNonBindable\n    translate\n    *ngIf=\"searchText\"\n    [translateParams]=\"{\n      searchHint: searchText\n    }\"\n    >searching \"{{ searchHint }}\"</small\n  >\n  <small\n    ngNonBindable\n    translate\n    *ngIf=\"filter\"\n    [translateParams]=\"{\n      filterHint: filter\n    }\"\n    >filtered by \"{{ filterHint }}\"</small\n  >\n</c8y-title>\n\n<c8y-search-grid [searchText]=\"searchText\" [filteringName]=\"filteringName\"></c8y-search-grid>\n"
            }]
    }], function () { return [{ type: ɵngcc6.ActivatedRoute }, { type: ɵngcc1.AlertService }]; }, { searchGrid: [{
            type: i0.ViewChild,
            args: [SearchGridComponent, { static: true }]
        }] }); })();
        return SearchResultsComponent;
    }());
    SearchResultsComponent.ctorParameters = function () { return [
        { type: router.ActivatedRoute },
        { type: i4.AlertService }
    ]; };
    SearchResultsComponent.propDecorators = {
        searchGrid: [{ type: i0.ViewChild, args: [SearchGridComponent, { static: true },] }]
    };

    var SearchActionComponent = /** @class */ (function () {
        function SearchActionComponent(c8yRouter, router) {
            this.c8yRouter = c8yRouter;
            this.router = router;
        }
        SearchActionComponent.prototype.onSearch = function (on) {
            this.navigate(['/assetsearch'], {
                queryParams: { search: on },
                replaceUrl: true
            });
        };
        SearchActionComponent.prototype.onFilter = function (on) {
            this.navigate(['/assetsearch'], {
                queryParams: { filter: on },
                replaceUrl: true
            });
        };
        SearchActionComponent.prototype.onClick = function (mo) {
            this.router.navigateByUrl(this.c8yRouter.getHref(mo, '/'));
        };
        SearchActionComponent.prototype.navigate = function (commands, extras) {
            var _this = this;
            this.router
                .navigateByUrl('/', { skipLocationChange: true })
                .then(function () { return _this.router.navigate(commands, extras); });
        };
SearchActionComponent.ɵfac = function SearchActionComponent_Factory(t) { return new (t || SearchActionComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.RouterService), ɵngcc0.ɵɵdirectiveInject(ɵngcc6.Router)); };
SearchActionComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SearchActionComponent, selectors: [["c8y-search-action"]], decls: 1, vars: 0, consts: [[3, "filter", "search", "onClick"]], template: function SearchActionComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-search-input", 0);
        ɵngcc0.ɵɵlistener("filter", function SearchActionComponent_Template_c8y_search_input_filter_0_listener($event) { return ctx.onFilter($event); })("search", function SearchActionComponent_Template_c8y_search_input_search_0_listener($event) { return ctx.onSearch($event); })("onClick", function SearchActionComponent_Template_c8y_search_input_onClick_0_listener($event) { return ctx.onClick($event); });
        ɵngcc0.ɵɵelementEnd();
    } }, directives: [ɵngcc1.SearchInputComponent], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SearchActionComponent, [{
        type: i0.Component,
        args: [{
                selector: 'c8y-search-action',
                template: "<c8y-search-input\n  (filter)=\"onFilter($event)\"\n  (search)=\"onSearch($event)\"\n  (onClick)=\"onClick($event)\"\n></c8y-search-input>\n"
            }]
    }], function () { return [{ type: ɵngcc1.RouterService }, { type: ɵngcc6.Router }]; }, null); })();
        return SearchActionComponent;
    }());
    SearchActionComponent.ctorParameters = function () { return [
        { type: i4.RouterService },
        { type: router.Router }
    ]; };

    var ɵ0 = [
        {
            path: 'assetsearch',
            component: SearchResultsComponent
        }
    ], ɵ1 = [
        {
            template: SearchActionComponent
        }
    ];
    var SearchModule = /** @class */ (function () {
        function SearchModule() {
        }
SearchModule.ɵfac = function SearchModule_Factory(t) { return new (t || SearchModule)(); };
SearchModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: SearchModule });
SearchModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        AssetSearchService,
        {
            provide: i4.HOOK_ROUTE,
            useValue: ɵ0,
            multi: true
        },
        {
            provide: i4.HOOK_SEARCH,
            useValue: ɵ1,
            multi: true
        }
    ], imports: [[i4.CoreModule, deviceGrid.DeviceGridModule, i4.SearchModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SearchModule, [{
        type: i0.NgModule,
        args: [{
                declarations: [SearchResultsComponent, SearchGridComponent, SearchActionComponent],
                imports: [i4.CoreModule, deviceGrid.DeviceGridModule, i4.SearchModule],
                entryComponents: [SearchResultsComponent, SearchActionComponent],
                exports: [SearchResultsComponent, SearchGridComponent],
                providers: [
                    AssetSearchService,
                    {
                        provide: i4.HOOK_ROUTE,
                        useValue: ɵ0,
                        multi: true
                    },
                    {
                        provide: i4.HOOK_SEARCH,
                        useValue: ɵ1,
                        multi: true
                    }
                ]
            }]
    }], function () { return []; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(SearchModule, { declarations: [SearchResultsComponent, SearchGridComponent, SearchActionComponent], imports: [ɵngcc1.CoreModule, ɵngcc7.DeviceGridModule, ɵngcc1.SearchModule], exports: [SearchResultsComponent, SearchGridComponent] }); })();
        return SearchModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.AssetSearchService = AssetSearchService;
    exports.SearchGridComponent = SearchGridComponent;
    exports.SearchModule = SearchModule;
    exports.SearchResultsComponent = SearchResultsComponent;
    exports["ɵ0"] = ɵ0;
    exports["ɵ1"] = ɵ1;
    exports["ɵa"] = SearchActionComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
