// @c8y/ngx-components/bundles/c8y-ngx-components-operations-operations-list.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/router'), require('ngx-bootstrap/tooltip'), require('@c8y/ngx-components'), require('@c8y/ngx-components/operations/status-filter'), require('@c8y/ngx-components/operations/operations-list-item-details'), require('@c8y/ngx-components/operations/bulk-operation-list-item'), require('@c8y/ngx-components/operations/shared'), require('lodash-es'), require('rxjs'), require('rxjs/operators'), require('@c8y/client'), require('@c8y/ngx-components/upgrade')) :
    typeof define === 'function' && define.amd ? define('@c8y/ngx-components/operations/operations-list', ['exports', '@angular/core', '@angular/router', 'ngx-bootstrap/tooltip', '@c8y/ngx-components', '@c8y/ngx-components/operations/status-filter', '@c8y/ngx-components/operations/operations-list-item-details', '@c8y/ngx-components/operations/bulk-operation-list-item', '@c8y/ngx-components/operations/shared', 'lodash-es', 'rxjs', 'rxjs/operators', '@c8y/client', '@c8y/ngx-components/upgrade'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.c8y = global.c8y || {}, global.c8y["ngx-components"] = global.c8y["ngx-components"] || {}, global.c8y["ngx-components"].operations = global.c8y["ngx-components"].operations || {}, global.c8y["ngx-components"].operations["operations-list"] = {}), global.ng.core, global.ng.router, global.tooltip, global.c8y["ngx-components"], global.c8y["ngx-components"].operations["status-filter"], global.c8y["ngx-components"].operations["operations-list-item-details"], global.c8y["ngx-components"].operations["bulk-operation-list-item"], global.c8y["ngx-components"].operations.shared, global.lodashEs, global.rxjs, global.rxjs.operators, global.client, global.c8y["ngx-components"].upgrade));
})(this, (function (exports, core, router, tooltip, ngxComponents, statusFilter, operationsListItemDetails, bulkOperationListItem, shared, lodashEs, rxjs, operators, client, upgrade) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || Array.prototype.slice.call(from));
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var OperationsListService = /** @class */ (function () {
        function OperationsListService(operationService, smartRulesService, ng1SmartRulesService) {
            this.operationService = operationService;
            this.smartRulesService = smartRulesService;
            this.ng1SmartRulesService = ng1SmartRulesService;
        }
        OperationsListService.prototype.getOperations = function (customFilter) {
            if (customFilter === void 0) { customFilter = {}; }
            var filter = Object.assign({ withTotalPages: true, withDeleted: true, pageSize: 50, dateFrom: new Date(0).toISOString(), revert: true }, customFilter);
            return this.operationService.list(filter);
        };
        OperationsListService.prototype.isSmartRulesMicroserviceAvailable = function () {
            return this.smartRulesService.isMicroserviceAvailable();
        };
        OperationsListService.prototype.getSmartRulesCreatePermissions = function () {
            if (this.ng1SmartRulesService && this.ng1SmartRulesService.permissionsCfgs) {
                return this.ng1SmartRulesService.permissionsCfgs.create;
            }
        };
        OperationsListService.prototype.createSmartRule = function (operation) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!this.ng1SmartRulesService) return [3 /*break*/, 2];
                            return [4 /*yield*/, this.ng1SmartRulesService.addNewForOutputOperationWithUI(operation)];
                        case 1: return [2 /*return*/, _a.sent()];
                        case 2: return [2 /*return*/];
                    }
                });
            });
        };
        return OperationsListService;
    }());
    OperationsListService.decorators = [
        { type: core.Injectable }
    ];
    OperationsListService.ctorParameters = function () { return [
        { type: client.OperationService },
        { type: client.SmartRulesService },
        { type: upgrade.Ng1SmartRulesService, decorators: [{ type: core.Optional }] }
    ]; };

    var OperationsListComponent = /** @class */ (function () {
        function OperationsListComponent(realtime, operationsListService, route) {
            var _this = this;
            var _a;
            this.realtime = realtime;
            this.operationsListService = operationsListService;
            this.route = route;
            this.deviceId = (_a = this.route.snapshot.parent.data.contextData) === null || _a === void 0 ? void 0 : _a.id;
            this.OPERATION_STATUS_OPTIONS_MAP = shared.OPERATION_STATUS_OPTIONS_MAP;
            this.refreshLoading = false;
            this.statusFilter$ = new rxjs.BehaviorSubject(null);
            this.reload$ = new rxjs.BehaviorSubject(null);
            this.operations$ = rxjs.combineLatest(this.statusFilter$, this.reload$).pipe(operators.tap(function () {
                _this.refreshLoading = true;
            }), operators.switchMap(function (_b) {
                var _c = __read(_b, 1), statusFilters = _c[0];
                return _this.filter(statusFilters);
            }), operators.tap(function () {
                _this.refreshLoading = false;
            }), operators.shareReplay(1));
            this.title = this.route.snapshot.parent.data.contextData
                ? this.route.snapshot.parent.data.contextData.name
                : ngxComponents.gettext('Device control');
        }
        OperationsListComponent.prototype.filter = function (statusFilters) {
            var status = statusFilters && statusFilters.length > 0
                ? {
                    status: lodashEs.flatten(statusFilters.map(function (statusFilter) { return statusFilter.status; }))
                }
                : {};
            return this.getOperations(status);
        };
        OperationsListComponent.prototype.getOperations = function (filter) {
            var queryFilter = this.deviceId ? Object.assign(Object.assign({}, filter), { deviceId: this.deviceId }) : filter;
            return this.operationsListService.getOperations(queryFilter);
        };
        OperationsListComponent.prototype.compareOperations = function (operationA, operationB) {
            return (new Date(operationA.creationTime).getTime() - new Date(operationB.creationTime).getTime());
        };
        return OperationsListComponent;
    }());
    OperationsListComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'c8y-operations-list',
                    template: "<c8y-title> {{ title | translate }}</c8y-title>\n<c8y-action-bar-item [placement]=\"'left'\" itemClass=\"navbar-form\">\n  <c8y-status-filter\n    #statusFilter\n    [options]=\"OPERATION_STATUS_OPTIONS_MAP\"\n    (onFilterChanged)=\"statusFilter$.next($event)\"\n  ></c8y-status-filter>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <c8y-realtime-btn [service]=\"realtime\"></c8y-realtime-btn>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button\n    class=\"btn btn-link d-flex a-i-center\"\n    title=\"{{ 'Reload' | translate }}\"\n    (click)=\"reload$.next()\"\n  >\n    <i c8yIcon=\"refresh\" [ngClass]=\"{ 'icon-spin': refreshLoading }\" class=\"m-r-4\"></i>\n    <span class=\"text-truncate\">\n      {{ 'Reload' | translate }}\n    </span>\n  </button>\n</c8y-action-bar-item>\n\n<!-- Empty state  -->\n<c8y-ui-empty-state\n  [icon]=\"'c8y-energy'\"\n  [title]=\"'No items to display.' | translate\"\n  [subtitle]=\"'Operations will be displayed here.' | translate\"\n  *ngIf=\"(operations$ | async)?.data.length === 0 && !(statusFilter$ | async)\"\n></c8y-ui-empty-state>\n<!-- No results empty state -->\n<c8y-ui-empty-state\n  [icon]=\"'search'\"\n  [title]=\"'No results to display.' | translate\"\n  [subtitle]=\"'Adjust or reset the filter.' | translate\"\n  *ngIf=\"(operations$ | async)?.data.length === 0 && (statusFilter$ | async)\"\n>\n  <button\n    type=\"button\"\n    title=\"{{ 'Reset filter' | translate }}\"\n    class=\"btn btn-primary\"\n    (click)=\"statusFilter.reset()\"\n    translate\n  >\n    Reset filter\n  </button>\n</c8y-ui-empty-state>\n\n<!-- Detailed list of operations + load more button -->\n<c8y-list-group class=\"m-b-24\" [ngClass]=\"{ 'dd-low': (operations$ | async)?.data.length < 10 }\">\n  <c8y-li\n    class=\"page-sticky-header c8y-list__item--empty-actions hidden-xs\"\n    *ngIf=\"(operations$ | async)?.data.length\"\n  >\n    <c8y-li-icon>\n      <i class=\"p-l-24\"></i>\n    </c8y-li-icon>\n    <c8y-li-body class=\"content-flex-57\">\n      <div class=\"col-5\">\n        {{ 'Operation' | translate }}\n      </div>\n      <div class=\"flex-grow\" *ngIf=\"!deviceId\">\n        {{ 'Device' | translate }}\n      </div>\n      <div class=\"col-4\">\n        {{ 'Date' | translate }}\n      </div>\n    </c8y-li-body>\n  </c8y-li>\n  <div\n    class=\"d-contents\"\n    *c8yFor=\"\n      let op of operations$ | async;\n      let i = index;\n      realtime: realtime;\n      realtimeOptions: { entityOrId: deviceId };\n      comparator: compareOperations.bind(this);\n      loadMore: 'auto'\n    \"\n  >\n    <c8y-operations-list-item [operation]=\"op\" (reload)=\"reload$.next()\" class=\"d-contents\">\n    </c8y-operations-list-item>\n  </div>\n</c8y-list-group>\n",
                    providers: [ngxComponents.OperationRealtimeService]
                },] }
    ];
    OperationsListComponent.ctorParameters = function () { return [
        { type: ngxComponents.OperationRealtimeService },
        { type: OperationsListService },
        { type: router.ActivatedRoute }
    ]; };
    OperationsListComponent.propDecorators = {
        statusFilter: [{ type: core.ViewChild, args: ['statusFilter', { static: true },] }]
    };

    var OperationsListItemComponent = /** @class */ (function () {
        function OperationsListItemComponent(bulkOperationListItemService, sharedService, operationsListService, route, router) {
            var _a, _b, _c, _d;
            this.bulkOperationListItemService = bulkOperationListItemService;
            this.sharedService = sharedService;
            this.operationsListService = operationsListService;
            this.route = route;
            this.router = router;
            this.collapsed = true;
            this.readOnly = false;
            this.noExpandToggle = false;
            this.refreshLoading = false;
            this.OPERATION_STATUS = client.OperationStatus;
            this.OPERATION_STATUS_OPTIONS_MAP = shared.OPERATION_STATUS_OPTIONS_MAP;
            this.deviceId = ((_b = (_a = this.route.snapshot.parent) === null || _a === void 0 ? void 0 : _a.data.contextData) === null || _b === void 0 ? void 0 : _b.id) ||
                ((_d = (_c = this.route.snapshot.firstChild) === null || _c === void 0 ? void 0 : _c.data.contextData) === null || _d === void 0 ? void 0 : _d.id);
            this.isSmartRulesMicroserviceAvailable = false;
        }
        OperationsListItemComponent.prototype.ngOnInit = function () {
            var _a;
            return __awaiter(this, void 0, void 0, function () {
                var _e, smartRulesCreatePermissions;
                return __generator(this, function (_f) {
                    switch (_f.label) {
                        case 0:
                            _e = this;
                            return [4 /*yield*/, this.operationsListService.isSmartRulesMicroserviceAvailable()];
                        case 1:
                            _e.isSmartRulesMicroserviceAvailable =
                                _f.sent();
                            this.operationTitle =
                                ((_a = this.operation.c8y_Command) === null || _a === void 0 ? void 0 : _a.text) ||
                                    this.operation.description ||
                                    ngxComponents.gettext('No command text or description available');
                            smartRulesCreatePermissions = this.operationsListService.getSmartRulesCreatePermissions();
                            if (smartRulesCreatePermissions) {
                                this.inventoryPermissions = smartRulesCreatePermissions[0].anyRole;
                                this.smartRuleAdminPermissions = smartRulesCreatePermissions[1].anyRole;
                            }
                            return [2 /*return*/];
                    }
                });
            });
        };
        OperationsListItemComponent.prototype.createCustomizedBulkOperation = function ($event) {
            if ($event) {
                $event.stopPropagation();
            }
            this.router.navigateByUrl("/devicecontrol/single/create-bulk/" + this.operation.id);
        };
        return OperationsListItemComponent;
    }());
    OperationsListItemComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'c8y-operations-list-item',
                    template: "<c8y-li\n  class=\"c8y-list__item--double-actions\"\n  [ngClass]=\"{ 'c8y-list__item--no-expand': noExpandToggle }\"\n  [collapsed]=\"collapsed\"\n  #listItem\n  id=\"{{ operation.id }}\"\n>\n  <c8y-li-icon>\n    <i\n      [c8yIcon]=\"OPERATION_STATUS_OPTIONS_MAP[operation.status].icon\"\n      [ngClass]=\"OPERATION_STATUS_OPTIONS_MAP[operation.status].styleClass\"\n      [tooltip]=\"OPERATION_STATUS_OPTIONS_MAP[operation.status].label | translate\"\n    ></i>\n  </c8y-li-icon>\n  <div\n    [ngClass]=\"{ 'content-flex-58': !readOnly, 'content-flex-50': readOnly }\"\n    [style.cursor]=\"!noExpandToggle ? 'pointer' : ''\"\n    (click)=\"!noExpandToggle && listItem.toggleCollapsed()\"\n  >\n    <div class=\"col-5\">\n      <span class=\"text-truncate\" title=\"{{ operationTitle | translate }}\">\n        {{ operationTitle | translate }}\n      </span>\n    </div>\n    <div class=\"col-3\" *ngIf=\"!deviceId\">\n      <a [routerLink]=\"['/device', operation.deviceId, 'device-info']\" routerLinkActive=\"active\">\n        {{ operation.deviceName }}\n      </a>\n    </div>\n    <div class=\"col-3\">\n      <small class=\"icon-flex text-muted\">\n        <i c8yIcon=\"calendar\" class=\"m-r-4\"></i>\n        {{ operation.creationTime | c8yDate }}\n      </small>\n    </div>\n  </div>\n  <ng-container *ngIf=\"!readOnly\">\n    <c8y-li-action\n      label=\"{{ 'Schedule as bulk operation' | translate }}\"\n      (click)=\"createCustomizedBulkOperation($event)\"\n      icon=\"c8y-icon c8y-icon-energy\"\n    >\n    </c8y-li-action>\n    <ng-container *c8yIfAllowed=\"inventoryPermissions; allowAny: true\">\n      <ng-container *c8yIfAllowed=\"smartRuleAdminPermissions; allowAny: true\">\n        <c8y-li-action\n          *ngIf=\"isSmartRulesMicroserviceAvailable\"\n          label=\"{{ 'Create smart rule' | translate }}\"\n          (click)=\"operationsListService.createSmartRule(operation)\"\n          icon=\"sliders\"\n        >\n        </c8y-li-action>\n      </ng-container>\n    </ng-container>\n    <c8y-li-action\n      *ngIf=\"operation.status === OPERATION_STATUS.PENDING\"\n      label=\"{{ 'Cancel operation' | translate }}\"\n      (click)=\"$event.stopPropagation(); sharedService.cancel(operation)\"\n      icon=\"times\"\n    >\n    </c8y-li-action>\n  </ng-container>\n  <c8y-li-collapse class=\"m-b-16\">\n    <c8y-operation-details-tabs\n      *ngIf=\"!listItem.collapsed\"\n      [operation]=\"operation\"\n      [bulkOperationModalDetailsService]=\"bulkOperationListItemService\"\n    ></c8y-operation-details-tabs>\n  </c8y-li-collapse>\n</c8y-li>\n"
                },] }
    ];
    OperationsListItemComponent.ctorParameters = function () { return [
        { type: bulkOperationListItem.BulkOperationListItemService },
        { type: shared.OperationsService },
        { type: OperationsListService },
        { type: router.ActivatedRoute },
        { type: router.Router }
    ]; };
    OperationsListItemComponent.propDecorators = {
        operation: [{ type: core.Input }],
        collapsed: [{ type: core.Input }],
        readOnly: [{ type: core.Input }],
        noExpandToggle: [{ type: core.Input }],
        listItem: [{ type: core.ViewChild, args: ['listItem', { static: true },] }]
    };

    var singleOperationListRoutes = [
        {
            path: 'devicecontrol/single',
            component: OperationsListComponent
        }
    ];
    var ɵ0 = [
        {
            context: ngxComponents.ViewContext.Device,
            path: 'control',
            component: OperationsListComponent,
            label: ngxComponents.gettext('Control'),
            icon: 'c8y-icon c8y-icon-device-control',
            priority: 600
        }
    ];
    /**
     * This module allows for displaying the list item details of single and bulk operations.
     */
    var OperationsListModule = /** @class */ (function () {
        function OperationsListModule() {
        }
        return OperationsListModule;
    }());
    OperationsListModule.decorators = [
        { type: core.NgModule, args: [{
                    imports: [
                        ngxComponents.CoreModule,
                        statusFilter.StatusFilterModule,
                        operationsListItemDetails.OperationsListItemDetailsModule,
                        shared.OperationsSharedModule,
                        tooltip.TooltipModule,
                        router.RouterModule.forChild(singleOperationListRoutes)
                    ],
                    declarations: [OperationsListComponent, OperationsListItemComponent],
                    exports: [OperationsListComponent, OperationsListItemComponent],
                    providers: [
                        OperationsListService,
                        bulkOperationListItem.BulkOperationListItemService,
                        {
                            provide: ngxComponents.HOOK_ROUTE,
                            useValue: ɵ0,
                            multi: true
                        }
                    ]
                },] }
    ];

    /**
     * Generated bundle index. Do not edit.
     */

    exports.OperationsListComponent = OperationsListComponent;
    exports.OperationsListItemComponent = OperationsListItemComponent;
    exports.OperationsListModule = OperationsListModule;
    exports.OperationsListService = OperationsListService;
    exports.singleOperationListRoutes = singleOperationListRoutes;
    exports["ɵ0"] = ɵ0;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
