// @c8y/ngx-components/bundles/c8y-ngx-components-device-list.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('ngx-bootstrap/modal'),require('@c8y/client'),require('@c8y/ngx-components/device-grid'),require('@c8y/ngx-components'),require('@ngx-translate/core'),require('@angular/forms'),require('@angular/common'),require('@angular/router'),exports, require('@angular/core'), require('@c8y/ngx-components/device-grid'), require('ngx-bootstrap/modal'), require('@c8y/client'), require('@c8y/ngx-components'), require('@ngx-translate/core'), require('rxjs'), require('rxjs/operators'), require('@angular/router')) :
    typeof define === 'function' && define.amd ? define('@c8y/ngx-components/device-list', ['@angular/core','ngx-bootstrap/modal','@c8y/client','@c8y/ngx-components/device-grid','@c8y/ngx-components','@ngx-translate/core','@angular/forms','@angular/common','@angular/router','exports', '@angular/core', '@c8y/ngx-components/device-grid', 'ngx-bootstrap/modal', '@c8y/client', '@c8y/ngx-components', '@ngx-translate/core', 'rxjs', 'rxjs/operators', '@angular/router'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ngxBootstrap.modal,global.c8y.client,global.c8y.ngxComponents.deviceGrid,global.c8y.ngxComponents,global.ngxTranslate.core,global.ng.forms,global.ng.common,global.ng.router,(global.c8y = global.c8y || {}, global.c8y["ngx-components"] = global.c8y["ngx-components"] || {}, global.c8y["ngx-components"]["device-list"] = {}), global.ng.core, global.c8y["ngx-components"]["device-grid"], global.modal, global.client, global.c8y["ngx-components"], global.core$1, global.rxjs, global.rxjs.operators, global.ng.router));
})(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,ɵngcc4,ɵngcc5,ɵngcc6,ɵngcc7,ɵngcc8,exports, core, deviceGrid, modal, client, ngxComponents, core$1, rxjs, operators, router) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || Array.prototype.slice.call(from));
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var AddSmartGroupComponent = /** @class */ (function () {
        function AddSmartGroupComponent(modal, smartGroupsService, deviceGridService, alert, translateService) {
            this.modal = modal;
            this.smartGroupsService = smartGroupsService;
            this.deviceGridService = deviceGridService;
            this.alert = alert;
            this.translateService = translateService;
            this.smartGroup = {
                name: '',
                type: 'c8y_DynamicGroup',
                c8y_IsDynamicGroup: {}
            };
        }
        AddSmartGroupComponent.prototype.create = function () {
            var _this = this;
            return this.deviceGridService
                .getConfig$(this.configKey)
                .pipe(operators.switchMap(function (config) {
                return _this.smartGroupsService.create({
                    smartGroup: _this.smartGroup,
                    deviceQueryString: _this.deviceQuery,
                    configurableColumns: config
                });
            }), operators.tap(function () {
                _this.alert.success(_this.translateService.instant(ngxComponents.gettext('Smart group {{ name }} created.'), {
                    name: _this.smartGroup.name
                }));
                _this.modal.hide();
            }), operators.catchError(function (ex) {
                _this.alert.addServerFailure(ex);
                return rxjs.of([]);
            }), operators.take(1))
                .subscribe();
        };
        AddSmartGroupComponent.prototype.close = function () {
            this.modal.hide();
        };
AddSmartGroupComponent.ɵfac = function AddSmartGroupComponent_Factory(t) { return new (t || AddSmartGroupComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BsModalRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.SmartGroupsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.DeviceGridService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.TranslateService)); };
AddSmartGroupComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AddSmartGroupComponent, selectors: [["c8y-add-smart-group"]], decls: 21, vars: 12, consts: [[1, "viewport-modal"], [1, "modal-header", "dialog-header"], [3, "c8yIcon"], ["translate", ""], [1, "modal-inner-scroll"], [1, "modal-body"], ["createSmartGroup", "ngForm"], ["translate", "", "for", "name"], ["id", "name", "type", "text", "autocomplete", "off", "name", "name", "required", "", 1, "form-control", 3, "ngModel", "placeholder", "ngModelChange"], [1, "modal-footer"], ["translate", "", 1, "btn", "btn-default", 3, "title", "click"], ["translate", "", 1, "btn", "btn-primary", 3, "title", "disabled", "click"]], template: function AddSmartGroupComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelement(2, "i", 2);
        ɵngcc0.ɵɵelementStart(3, "h4", 3);
        ɵngcc0.ɵɵtext(4, "Create smart group");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "div", 4);
        ɵngcc0.ɵɵelementStart(6, "div", 5);
        ɵngcc0.ɵɵelementStart(7, "form", null, 6);
        ɵngcc0.ɵɵelementStart(9, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(10, "label", 7);
        ɵngcc0.ɵɵtext(11, "Group name");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(12, "input", 8);
        ɵngcc0.ɵɵlistener("ngModelChange", function AddSmartGroupComponent_Template_input_ngModelChange_12_listener($event) { return ctx.smartGroup.name = $event; });
        ɵngcc0.ɵɵpipe(13, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(14, "div", 9);
        ɵngcc0.ɵɵelementStart(15, "button", 10);
        ɵngcc0.ɵɵlistener("click", function AddSmartGroupComponent_Template_button_click_15_listener() { return ctx.close(); });
        ɵngcc0.ɵɵpipe(16, "translate");
        ɵngcc0.ɵɵtext(17, " Cancel ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(18, "button", 11);
        ɵngcc0.ɵɵlistener("click", function AddSmartGroupComponent_Template_button_click_18_listener() { return ctx.create(); });
        ɵngcc0.ɵɵpipe(19, "translate");
        ɵngcc0.ɵɵtext(20, " Create ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        var _r0 = ɵngcc0.ɵɵreference(8);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yIcon", "c8y-group-smart");
        ɵngcc0.ɵɵadvance(10);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(13, 6, "e.g. My smart group"));
        ɵngcc0.ɵɵproperty("ngModel", ctx.smartGroup.name);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(16, 8, "Cancel"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(19, 10, "Create"));
        ɵngcc0.ɵɵproperty("disabled", _r0.form.invalid);
    } }, directives: [ɵngcc4.IconDirective, ɵngcc4.C8yTranslateDirective, ɵngcc6.ɵNgNoValidate, ɵngcc6.NgControlStatusGroup, ɵngcc6.NgForm, ɵngcc4.FormGroupComponent, ɵngcc4.RequiredInputPlaceholderDirective, ɵngcc6.DefaultValueAccessor, ɵngcc6.RequiredValidator, ɵngcc6.NgControlStatus, ɵngcc6.NgModel], pipes: [ɵngcc4.C8yTranslatePipe], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AddSmartGroupComponent, [{
        type: core.Component,
        args: [{
                selector: 'c8y-add-smart-group',
                template: "<div class=\"viewport-modal\">\n  <div class=\"modal-header dialog-header\">\n    <i [c8yIcon]=\"'c8y-group-smart'\"></i>\n    <h4 translate>Create smart group</h4>\n  </div>\n\n  <div class=\"modal-inner-scroll\">\n    <div class=\"modal-body\">\n      <form #createSmartGroup=\"ngForm\">\n        <c8y-form-group>\n          <label translate for=\"name\">Group name</label>\n          <input\n            id=\"name\"\n            type=\"text\"\n            class=\"form-control\"\n            autocomplete=\"off\"\n            name=\"name\"\n            [(ngModel)]=\"smartGroup.name\"\n            placeholder=\"{{ 'e.g. My smart group' | translate }}\"\n            required\n          />\n        </c8y-form-group>\n      </form>\n    </div>\n  </div>\n  <div class=\"modal-footer\">\n    <button title=\"{{ 'Cancel' | translate }}\" class=\"btn btn-default\" (click)=\"close()\" translate>\n      Cancel\n    </button>\n\n    <button\n      title=\"{{ 'Create' | translate }}\"\n      class=\"btn btn-primary\"\n      (click)=\"create()\"\n      [disabled]=\"createSmartGroup.form.invalid\"\n      translate\n    >\n      Create\n    </button>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.BsModalRef }, { type: ɵngcc2.SmartGroupsService }, { type: ɵngcc3.DeviceGridService }, { type: ɵngcc4.AlertService }, { type: ɵngcc5.TranslateService }]; }, null); })();
        return AddSmartGroupComponent;
    }());
    AddSmartGroupComponent.ctorParameters = function () { return [
        { type: modal.BsModalRef },
        { type: client.SmartGroupsService },
        { type: deviceGrid.DeviceGridService },
        { type: ngxComponents.AlertService },
        { type: core$1.TranslateService }
    ]; };

    var DeviceListComponent = /** @class */ (function () {
        function DeviceListComponent(deviceGridService, bsModalService) {
            this.deviceGridService = deviceGridService;
            this.bsModalService = bsModalService;
            this.GRID_CONFIG_KEY = 'device-grid-all';
            this.count$ = this.deviceGridService.getTotal({});
        }
        DeviceListComponent.prototype.createSmartGroup = function () {
            return __awaiter(this, void 0, void 0, function () {
                var initialState;
                return __generator(this, function (_a) {
                    initialState = {
                        configKey: this.GRID_CONFIG_KEY,
                        deviceQuery: this.deviceQueryStringOutput
                    };
                    this.bsModalService.show(AddSmartGroupComponent, {
                        class: 'modal-sm',
                        initialState: initialState
                    });
                    return [2 /*return*/];
                });
            });
        };
DeviceListComponent.ɵfac = function DeviceListComponent_Factory(t) { return new (t || DeviceListComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc3.DeviceGridService), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BsModalService)); };
DeviceListComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: DeviceListComponent, selectors: [["c8y-device-list"]], decls: 13, vars: 14, consts: [[3, "placement"], [1, "btn", "btn-link", 3, "title", "click"], ["c8yIcon", "c8y-group-smart"], ["src", "/users-guide/device-management/#viewing-devices"], [1, "content-fullpage"], [3, "infiniteScroll", "columnsConfigKey", "onDeviceQueryStringChange"]], template: function DeviceListComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵpipe(2, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "c8y-action-bar-item", 0);
        ɵngcc0.ɵɵelementStart(4, "button", 1);
        ɵngcc0.ɵɵlistener("click", function DeviceListComponent_Template_button_click_4_listener() { return ctx.createSmartGroup(); });
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelement(6, "i", 2);
        ɵngcc0.ɵɵtext(7);
        ɵngcc0.ɵɵpipe(8, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(9, "c8y-help", 3);
        ɵngcc0.ɵɵelementStart(10, "div", 4);
        ɵngcc0.ɵɵelementStart(11, "c8y-device-grid", 5);
        ɵngcc0.ɵɵlistener("onDeviceQueryStringChange", function DeviceListComponent_Template_c8y_device_grid_onDeviceQueryStringChange_11_listener($event) { return ctx.deviceQueryStringOutput = $event; });
        ɵngcc0.ɵɵpipe(12, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 6, "All devices"), "\n");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("placement", "right");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(5, 8, "Create smart group"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(8, 10, "Create smart group"), " ");
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("infiniteScroll", ɵngcc0.ɵɵpipeBind1(12, 12, ctx.count$) === undefined ? "auto" : undefined)("columnsConfigKey", ctx.GRID_CONFIG_KEY);
    } }, directives: [ɵngcc4.TitleComponent, ɵngcc4.ActionBarItemComponent, ɵngcc4.IconDirective, ɵngcc4.HelpComponent, ɵngcc3.DeviceGridComponent], pipes: [ɵngcc4.C8yTranslatePipe, ɵngcc7.AsyncPipe], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DeviceListComponent, [{
        type: core.Component,
        args: [{
                selector: 'c8y-device-list',
                template: "<c8y-title>\n  {{ 'All devices' | translate }}\n</c8y-title>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button\n    class=\"btn btn-link\"\n    title=\"{{ 'Create smart group' | translate }}\"\n    (click)=\"createSmartGroup()\"\n  >\n    <i c8yIcon=\"c8y-group-smart\"></i>\n    {{ 'Create smart group' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<c8y-help src=\"/users-guide/device-management/#viewing-devices\"></c8y-help>\n\n<div class=\"content-fullpage\">\n  <c8y-device-grid\n    [infiniteScroll]=\"(count$ | async) === undefined ? 'auto' : undefined\"\n    [columnsConfigKey]=\"GRID_CONFIG_KEY\"\n    (onDeviceQueryStringChange)=\"deviceQueryStringOutput = $event\"\n  ></c8y-device-grid>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc3.DeviceGridService }, { type: ɵngcc1.BsModalService }]; }, null); })();
        return DeviceListComponent;
    }());
    DeviceListComponent.ctorParameters = function () { return [
        { type: deviceGrid.DeviceGridService },
        { type: modal.BsModalService }
    ]; };

    var DeviceListNavigationFactory = /** @class */ (function () {
        function DeviceListNavigationFactory() {
            this.navs = [];
        }
        DeviceListNavigationFactory.prototype.get = function () {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    this.navs.push(new ngxComponents.NavigatorNode({
                        label: ngxComponents.gettext('All devices'),
                        path: 'device',
                        icon: 'c8y-device-management',
                        parent: {
                            label: ngxComponents.gettext('Devices'),
                            icon: 'exchange'
                        },
                        priority: 2000
                    }));
                    return [2 /*return*/, this.navs];
                });
            });
        };
DeviceListNavigationFactory.ɵfac = function DeviceListNavigationFactory_Factory(t) { return new (t || DeviceListNavigationFactory)(); };
DeviceListNavigationFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: DeviceListNavigationFactory, factory: function (t) { return DeviceListNavigationFactory.ɵfac(t); } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DeviceListNavigationFactory, [{
        type: core.Injectable
    }], function () { return []; }, null); })();
        return DeviceListNavigationFactory;
    }());

    var routes = [
        {
            path: 'device',
            component: DeviceListComponent
        }
    ];
    var DeviceListModule = /** @class */ (function () {
        function DeviceListModule() {
        }
DeviceListModule.ɵfac = function DeviceListModule_Factory(t) { return new (t || DeviceListModule)(); };
DeviceListModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DeviceListModule });
DeviceListModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        {
            provide: ngxComponents.HOOK_NAVIGATOR_NODES,
            useClass: DeviceListNavigationFactory,
            multi: true
        }
    ], imports: [[ngxComponents.CoreModule, ngxComponents.CommonModule, deviceGrid.DeviceGridModule, router.RouterModule.forChild(routes)]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DeviceListModule, [{
        type: core.NgModule,
        args: [{
                imports: [ngxComponents.CoreModule, ngxComponents.CommonModule, deviceGrid.DeviceGridModule, router.RouterModule.forChild(routes)],
                declarations: [DeviceListComponent, AddSmartGroupComponent],
                entryComponents: [DeviceListComponent],
                exports: [],
                providers: [
                    {
                        provide: ngxComponents.HOOK_NAVIGATOR_NODES,
                        useClass: DeviceListNavigationFactory,
                        multi: true
                    }
                ]
            }]
    }], function () { return []; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DeviceListModule, { declarations: [DeviceListComponent, AddSmartGroupComponent], imports: [ɵngcc4.CoreModule, ɵngcc4.CommonModule, ɵngcc3.DeviceGridModule, ɵngcc8.RouterModule] }); })();
        return DeviceListModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.AddSmartGroupComponent = AddSmartGroupComponent;
    exports.DeviceListComponent = DeviceListComponent;
    exports.DeviceListModule = DeviceListModule;
    exports.DeviceListNavigationFactory = DeviceListNavigationFactory;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
