// @c8y/ngx-components/esm2015/auth-configuration/factories/tabs.factory.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { gettext } from '@c8y/ngx-components';
import { of, EMPTY, merge } from 'rxjs';
import { switchMap, mergeAll, toArray } from 'rxjs/operators';
import { SsoConfigurationService } from '../sso-configuration.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from '../sso-configuration.service';
export class TabsFactory {
    constructor(router, ssoConfigurationService) {
        this.router = router;
        this.ssoConfigurationService = ssoConfigurationService;
    }
    get() {
        if (!this.router.url.match(/auth-configuration/g)) {
            return of([]);
        }
        const basicSettingsTab$ = of({
            path: 'auth-configuration/basic_settings',
            label: gettext('Basic settings'),
            icon: 'unlock-alt',
            priority: 1100,
            orientation: 'horizontal'
        });
        const ssoTab$ = this.ssoConfigurationService.showSsoConfiguration$().pipe(switchMap(showSsoConfiguration => showSsoConfiguration
            ? of({
                path: 'auth-configuration/single_sign-on',
                label: gettext('Single sign-on'),
                icon: 'sign-in',
                priority: 1050,
                orientation: 'horizontal'
            })
            : EMPTY));
        return merge([basicSettingsTab$, ssoTab$]).pipe(mergeAll(), toArray());
    }
}
TabsFactory.ɵfac = function TabsFactory_Factory(t) { return new (t || TabsFactory)(ɵngcc0.ɵɵinject(ɵngcc1.Router), ɵngcc0.ɵɵinject(ɵngcc2.SsoConfigurationService)); };
TabsFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: TabsFactory, factory: TabsFactory.ɵfac });
TabsFactory.ctorParameters = () => [
    { type: Router },
    { type: SsoConfigurationService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TabsFactory, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.Router }, { type: ɵngcc2.SsoConfigurationService }]; }, null); })();
