// @c8y/ngx-components/esm2015/core/dynamic-forms/date/date.type.component.js
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { FieldType } from '@ngx-formly/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from 'ngx-bootstrap/datepicker';
import * as ɵngcc3 from '@angular/forms';
import * as ɵngcc4 from '../../forms/required-input-placeholder.directive';
import * as ɵngcc5 from '@ngx-formly/core';
import * as ɵngcc6 from 'ngx-bootstrap/popover';
import * as ɵngcc7 from '../../common/icon.directive';
import * as ɵngcc8 from '../../i18n/c8y-translate.pipe';
import * as ɵngcc9 from '../../common/humanize.pipe';

function FieldDate_label_1_button_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "button", 6);
    ɵngcc0.ɵɵelement(1, "i", 7);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("popover", ctx_r1.to.description);
} }
function FieldDate_label_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "label", 4);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "humanize");
    ɵngcc0.ɵɵtemplate(3, FieldDate_label_1_button_3_Template, 2, 1, "button", 5);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("id", ctx_r0.id + "-fieldset");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 3, ctx_r0.to.label), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.to.description);
} }
const _c0 = function () { return { customTodayClass: "today", dateInputFormat: "YYYY-MM-DD" }; };
export class FieldDate extends FieldType {
}
FieldDate.ɵfac = /*@__PURE__*/ function () { let ɵFieldDate_BaseFactory; return function FieldDate_Factory(t) { return (ɵFieldDate_BaseFactory || (ɵFieldDate_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(FieldDate)))(t || FieldDate); }; }();
FieldDate.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: FieldDate, selectors: [["c8y-field-date"]], features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 5, vars: 10, consts: [[1, "form-group"], [3, "id", 4, "ngIf"], [1, "form-group", "datepicker"], ["placeholder", "Date", "placement", "top", "bsDatepicker", "", 1, "form-control", 3, "formControl", "formlyAttributes", "bsConfig", "placeholder", "required"], [3, "id"], ["class", "btn-clean text-14 m-l-4", "placement", "right", "outsideClick", "true", "type", "button", 3, "popover", 4, "ngIf"], ["placement", "right", "outsideClick", "true", "type", "button", 1, "btn-clean", "text-14", "m-l-4", 3, "popover"], ["c8yIcon", "question-circle-o", 1, "text-info"]], template: function FieldDate_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, FieldDate_label_1_Template, 4, 5, "label", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelement(3, "input", 3);
        ɵngcc0.ɵɵpipe(4, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵattribute("aria-labelledby", ctx.id + "-fieldset");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.to.label || ctx.to.description);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("formControl", ctx.formControl)("formlyAttributes", ctx.field)("bsConfig", ɵngcc0.ɵɵpureFunction0(9, _c0))("placeholder", ɵngcc0.ɵɵpipeBind1(4, 7, "Select a date\u2026"))("required", ctx.to.required);
    } }, directives: [ɵngcc1.NgIf, ɵngcc2.BsDatepickerInputDirective, ɵngcc3.DefaultValueAccessor, ɵngcc4.RequiredInputPlaceholderDirective, ɵngcc2.BsDatepickerDirective, ɵngcc3.NgControlStatus, ɵngcc3.FormControlDirective, ɵngcc5.FormlyAttributes, ɵngcc3.RequiredValidator, ɵngcc6.PopoverDirective, ɵngcc7.IconDirective], pipes: [ɵngcc8.C8yTranslatePipe, ɵngcc9.HumanizePipe], encapsulation: 2, changeDetection: 0 });
FieldDate.CONFIG = {
    types: [
        {
            name: 'date',
            component: FieldDate
        }
    ]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(FieldDate, [{
        type: Component,
        args: [{
                selector: 'c8y-field-date',
                template: "<div class=\"form-group\"  [attr.aria-labelledby]=\"id + '-fieldset'\">\n  <label id=\"{{ id + '-fieldset' }}\" *ngIf=\"to.label || to.description\">\n    {{ to.label | humanize}}\n    <button\n      *ngIf=\"to.description\"\n      class=\"btn-clean text-14 m-l-4\"\n      [popover]=\"to.description\"\n      placement=\"right\"\n      outsideClick=\"true\"\n      type=\"button\"\n    >\n      <i c8yIcon=\"question-circle-o\" class=\"text-info\"></i>\n    </button>\n  </label>\n  <div class=\"form-group datepicker\">\n    <input\n      [formControl]=\"formControl\"\n      [formlyAttributes]=\"field\"\n      class=\"form-control\"\n      placeholder=\"Date\"\n      [bsConfig]=\"{ customTodayClass: 'today', dateInputFormat: 'YYYY-MM-DD' }\"\n      [placeholder]=\"'Select a date\u2026' | translate\"\n      placement=\"top\"\n      bsDatepicker\n      [required]=\"to.required\"\n    />\n  </div>\n</div>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], null, null); })();
