// @c8y/ngx-components/esm2015/operations/operation-details/operation-details-modal.component.js
import { Component, Input } from '@angular/core';
import { BsModalRef } from 'ngx-bootstrap/modal';
export class OperationDetailsModalComponent {
    constructor(modalRef) {
        this.modalRef = modalRef;
        this.collapsed = false;
        this.readOnly = false;
    }
}
OperationDetailsModalComponent.decorators = [
    { type: Component, args: [{
                selector: 'c8y-operation-details-modal',
                template: "<div class=\"viewport-modal\">\n  <div class=\"modal-header\">\n    <h3>{{ 'Single operation details' | translate }}</h3>\n  </div>\n\n  <div class=\"modal-inner-scroll\">\n    <c8y-operations-list-item\n      [operation]=\"operation\"\n      [collapsed]=\"collapsed\"\n      [noExpandToggle]=\"true\"\n      [readOnly]=\"readOnly\"\n    >\n    </c8y-operations-list-item>\n  </div>\n\n  <div class=\"modal-footer\">\n    <button title=\"{{ 'Close' | translate }}\" class=\"btn btn-default\" (click)=\"modalRef.hide()\">\n      {{ 'Close' | translate }}\n    </button>\n  </div>\n</div>\n"
            },] }
];
OperationDetailsModalComponent.ctorParameters = () => [
    { type: BsModalRef }
];
OperationDetailsModalComponent.propDecorators = {
    operation: [{ type: Input }],
    collapsed: [{ type: Input }],
    readOnly: [{ type: Input }]
};
