// @c8y/ngx-components/esm2015/operations/operation-details/operation-details.module.js
import { NgModule } from '@angular/core';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { CoreModule } from '@c8y/ngx-components';
import { BulkOperationListItemService } from '@c8y/ngx-components/operations/bulk-operation-list-item';
import { OperationsListModule } from '@c8y/ngx-components/operations/operations-list';
import { OperationDetailsService } from './operation-details.service';
import { OperationDetailsComponent } from './operation-details.component';
import { OperationDetailsModalComponent } from './operation-details-modal.component';
/**
 * This module allows to display details of a single operation in a modal.
 */
export class OperationDetailsModule {
}
OperationDetailsModule.decorators = [
    { type: NgModule, args: [{
                imports: [CoreModule, PopoverModule, TooltipModule, CollapseModule, OperationsListModule],
                providers: [OperationDetailsService, BulkOperationListItemService],
                declarations: [OperationDetailsComponent, OperationDetailsModalComponent],
                entryComponents: [OperationDetailsComponent, OperationDetailsModalComponent],
                exports: [OperationDetailsComponent]
            },] }
];
