// @c8y/ngx-components/esm2015/core/date-picker/date-picker.component.js
import { Component, EventEmitter, Output } from '@angular/core';
import { FormControl, FormGroup } from '@angular/forms';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/dropdown';
import * as ɵngcc2 from '../common/icon.directive';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from '@angular/forms';
import * as ɵngcc5 from '../i18n/c8y-translate.directive';
import * as ɵngcc6 from '../forms/form-group.component';
import * as ɵngcc7 from '../forms/required-input-placeholder.directive';
import * as ɵngcc8 from 'ngx-bootstrap/datepicker';
import * as ɵngcc9 from '../i18n/c8y-translate.pipe';

function DatePickerComponent_span_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵelementStart(1, "span", 7);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "date");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(3, 2, "From`date`"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 4, ctx_r1.dateFrom), " ");
} }
function DatePickerComponent_span_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵelementStart(1, "span", 7);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "date");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(3, 2, "To`date`"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 4, ctx_r2.dateTo), " ");
} }
function DatePickerComponent_span_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Filter by date"), " ");
} }
const _c0 = function () { return { customTodayClass: "today", returnFocusToInput: true }; };
function DatePickerComponent_div_9_Template(rf, ctx) { if (rf & 1) {
    const _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 8);
    ɵngcc0.ɵɵelementStart(1, "form", 9);
    ɵngcc0.ɵɵelementStart(2, "span", 10);
    ɵngcc0.ɵɵtext(3, "Date from");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "c8y-form-group", 11);
    ɵngcc0.ɵɵelementStart(5, "input", 12);
    ɵngcc0.ɵɵlistener("bsValueChange", function DatePickerComponent_div_9_Template_input_bsValueChange_5_listener($event) { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r5 = ɵngcc0.ɵɵnextContext(); return ctx_r5.dateFrom = $event; });
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "span", 10);
    ɵngcc0.ɵɵtext(8, "Date to");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(9, "c8y-form-group", 13);
    ɵngcc0.ɵɵelementStart(10, "input", 14);
    ɵngcc0.ɵɵlistener("bsValueChange", function DatePickerComponent_div_9_Template_input_bsValueChange_10_listener($event) { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r7 = ɵngcc0.ɵɵnextContext(); return ctx_r7.dateTo = $event; });
    ɵngcc0.ɵɵpipe(11, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(12, "div", 15);
    ɵngcc0.ɵɵelementStart(13, "button", 16);
    ɵngcc0.ɵɵlistener("click", function DatePickerComponent_div_9_Template_button_click_13_listener() { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r8 = ɵngcc0.ɵɵnextContext(); const _r0 = ɵngcc0.ɵɵreference(1); ctx_r8.clearFilter(); return _r0.isOpen = !_r0.isOpen; });
    ɵngcc0.ɵɵpipe(14, "translate");
    ɵngcc0.ɵɵtext(15);
    ɵngcc0.ɵɵpipe(16, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(17, "button", 17);
    ɵngcc0.ɵɵlistener("click", function DatePickerComponent_div_9_Template_button_click_17_listener() { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r9 = ɵngcc0.ɵɵnextContext(); const _r0 = ɵngcc0.ɵɵreference(1); ctx_r9.filter(); return _r0.isOpen = !_r0.isOpen; });
    ɵngcc0.ɵɵpipe(18, "translate");
    ɵngcc0.ɵɵtext(19);
    ɵngcc0.ɵɵpipe(20, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("formGroup", ctx_r4.fgDatePicker);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(6, 12, "Date from"));
    ɵngcc0.ɵɵproperty("maxDate", ctx_r4.dateTo)("bsConfig", ɵngcc0.ɵɵpureFunction0(24, _c0));
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(11, 14, "Date to"));
    ɵngcc0.ɵɵproperty("minDate", ctx_r4.dateFrom)("bsConfig", ɵngcc0.ɵɵpureFunction0(25, _c0));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(14, 16, "Clear selection"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(16, 18, "Clear`selection`"), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(18, 20, "Apply selection"));
    ɵngcc0.ɵɵproperty("disabled", !ctx_r4.fgDatePicker.get("dateFrom").value && !ctx_r4.fgDatePicker.get("dateTo").value);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(20, 22, "Apply`selection`"), " ");
} }
export class DatePickerComponent {
    constructor() {
        this.onDateSelected = new EventEmitter();
    }
    ngOnInit() {
        this.fgDatePicker = new FormGroup({
            dateFrom: new FormControl(),
            dateTo: new FormControl()
        });
    }
    filter() {
        this.onDateSelected.emit(this.fgDatePicker.value);
        this.dateFrom = this.fgDatePicker.get('dateFrom').value;
        this.dateTo = this.fgDatePicker.get('dateTo').value;
    }
    clearFilter() {
        this.fgDatePicker.setValue({
            dateFrom: null,
            dateTo: null
        });
        this.onDateSelected.emit(null);
        this.dateFrom = null;
        this.dateTo = null;
    }
}
DatePickerComponent.ɵfac = function DatePickerComponent_Factory(t) { return new (t || DatePickerComponent)(); };
DatePickerComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: DatePickerComponent, selectors: [["c8y-date-picker"]], outputs: { onDateSelected: "onDateSelected" }, decls: 10, vars: 7, consts: [["dropdown", "", "placement", "bottom left", 1, "c8y-child-assets-selector", 3, "autoClose"], ["datefilter", "bs-dropdown"], ["id", "date-range", "dropdownToggle", "", "type", "button", "aria-controls", "dropdown-date-range", 1, "btn", "dropdown-toggle", "d-flex", "a-i-center", "c8y-dropdown", 3, "title"], ["c8yIcon", "calendar-o", 1, "m-r-4", "text-primary"], [1, "text-truncate"], [4, "ngIf"], ["id", "dropdown-date-range", "class", "dropdown-menu", "role", "menu", "aria-labelledby", "date-range", 4, "dropdownMenu"], [1, "text-label-small"], ["id", "dropdown-date-range", "role", "menu", "aria-labelledby", "date-range", 1, "dropdown-menu"], [1, "dropdown-form", 3, "formGroup"], ["translate", "", 1, "text-medium", "m-b-4"], [1, "datepicker", "d-block", "m-b-16", 2, "max-height", "32px"], ["formControlName", "dateFrom", "bsDatepicker", "", 1, "form-control", "fit-w", "text-left", 3, "placeholder", "maxDate", "bsConfig", "bsValueChange"], [1, "datepicker", "m-l-0", "d-block", 2, "max-height", "32px"], ["name", "dateTo", "formControlName", "dateTo", "bsDatepicker", "", 1, "form-control", "fit-w", "text-left", 3, "placeholder", "minDate", "bsConfig", "bsValueChange"], [1, "p-t-24"], ["type", "button", 1, "btn", "btn-default", "btn-sm", 3, "title", "click"], ["type", "submit", 1, "btn", "btn-primary", "btn-sm", 3, "disabled", "title", "click"]], template: function DatePickerComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0, 1);
        ɵngcc0.ɵɵelementStart(2, "button", 2);
        ɵngcc0.ɵɵpipe(3, "translate");
        ɵngcc0.ɵɵelement(4, "i", 3);
        ɵngcc0.ɵɵelementStart(5, "span", 4);
        ɵngcc0.ɵɵtemplate(6, DatePickerComponent_span_6_Template, 6, 6, "span", 5);
        ɵngcc0.ɵɵtemplate(7, DatePickerComponent_span_7_Template, 6, 6, "span", 5);
        ɵngcc0.ɵɵtemplate(8, DatePickerComponent_span_8_Template, 3, 3, "span", 5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(9, DatePickerComponent_div_9_Template, 21, 26, "div", 6);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("autoClose", false);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(3, 5, "Date filter"));
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngIf", ctx.dateFrom);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.dateTo);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.dateFrom && !ctx.dateTo);
    } }, directives: [ɵngcc1.BsDropdownDirective, ɵngcc1.BsDropdownToggleDirective, ɵngcc2.IconDirective, ɵngcc3.NgIf, ɵngcc1.BsDropdownMenuDirective, ɵngcc4.ɵNgNoValidate, ɵngcc4.NgControlStatusGroup, ɵngcc4.FormGroupDirective, ɵngcc5.C8yTranslateDirective, ɵngcc6.FormGroupComponent, ɵngcc4.DefaultValueAccessor, ɵngcc7.RequiredInputPlaceholderDirective, ɵngcc8.BsDatepickerInputDirective, ɵngcc4.NgControlStatus, ɵngcc4.FormControlName, ɵngcc8.BsDatepickerDirective], pipes: [ɵngcc9.C8yTranslatePipe, ɵngcc3.DatePipe], encapsulation: 2 });
DatePickerComponent.propDecorators = {
    onDateSelected: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DatePickerComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-date-picker',
                template: "<div\n  dropdown\n  class=\"c8y-child-assets-selector\"\n  #datefilter=\"bs-dropdown\"\n  placement=\"bottom left\"\n  [autoClose]=\"false\"\n>\n  <button\n    id=\"date-range\"\n    dropdownToggle\n    title=\"{{ 'Date filter' | translate }}\"\n    type=\"button\"\n    class=\"btn dropdown-toggle d-flex a-i-center c8y-dropdown\"\n    aria-controls=\"dropdown-date-range\"\n  >\n    <i c8yIcon=\"calendar-o\" class=\"m-r-4 text-primary\"></i>\n    <span class=\"text-truncate\">\n      <span *ngIf=\"dateFrom\">\n        <span class=\"text-label-small\">{{ 'From`date`' | translate }}</span>\n        {{ dateFrom | date }}\n      </span>\n      <span *ngIf=\"dateTo\">\n        <span class=\"text-label-small\">{{ 'To`date`' | translate }}</span>\n        {{ dateTo | date }}\n      </span>\n      <span *ngIf=\"!dateFrom && !dateTo\">\n        {{ 'Filter by date' | translate }}\n      </span>\n    </span>\n  </button>\n  <div\n    id=\"dropdown-date-range\"\n    *dropdownMenu\n    class=\"dropdown-menu\"\n    role=\"menu\"\n    aria-labelledby=\"date-range\"\n  >\n    <form [formGroup]=\"fgDatePicker\" class=\"dropdown-form\">\n      <span class=\"text-medium m-b-4\" translate>Date from</span>\n      <c8y-form-group class=\"datepicker d-block m-b-16\" style=\"max-height: 32px\">\n        <input\n          formControlName=\"dateFrom\"\n          class=\"form-control fit-w text-left\"\n          placeholder=\"{{ 'Date from' | translate }}\"\n          bsDatepicker\n          [maxDate]=\"dateTo\"\n          (bsValueChange)=\"dateFrom = $event\"\n          [bsConfig]=\"{ customTodayClass: 'today', returnFocusToInput: true }\"\n        />\n      </c8y-form-group>\n      <span class=\"text-medium m-b-4\" translate>Date to</span>\n      <c8y-form-group class=\"datepicker m-l-0 d-block\" style=\"max-height: 32px\">\n        <input\n          name=\"dateTo\"\n          formControlName=\"dateTo\"\n          class=\"form-control fit-w text-left\"\n          placeholder=\"{{ 'Date to' | translate }}\"\n          bsDatepicker\n          [minDate]=\"dateFrom\"\n          (bsValueChange)=\"dateTo = $event\"\n          [bsConfig]=\"{ customTodayClass: 'today', returnFocusToInput: true }\"\n        />\n      </c8y-form-group>\n      <div class=\"p-t-24\">\n        <button\n          title=\"{{ 'Clear selection' | translate }}\"\n          type=\"button\"\n          class=\"btn btn-default btn-sm\"\n          (click)=\"clearFilter(); datefilter.isOpen = !datefilter.isOpen\"\n        >\n          {{ 'Clear`selection`' | translate }}\n        </button>\n        <button\n          [disabled]=\"!fgDatePicker.get('dateFrom').value && !fgDatePicker.get('dateTo').value\"\n          title=\"{{ 'Apply selection' | translate }}\"\n          type=\"submit\"\n          class=\"btn btn-primary btn-sm\"\n          (click)=\"filter(); datefilter.isOpen = !datefilter.isOpen\"\n        >\n          {{ 'Apply`selection`' | translate }}\n        </button>\n      </div>\n    </form>\n  </div>\n</div>\n"
            }]
    }], function () { return []; }, { onDateSelected: [{
            type: Output
        }] }); })();
