// @c8y/ngx-components/esm2015/context-dashboard/add-dashboard.factory.js
import { __awaiter } from "tslib";
import { Injectable } from '@angular/core';
import { ContextRouteService, Permissions, ViewContext } from '@c8y/ngx-components';
import { BsModalService } from 'ngx-bootstrap/modal';
import { AddDashboardComponent } from './add-dashboard.component';
import { ContextDashboardService } from './context-dashboard.service';
import { DashboardDetailComponent } from './dashboard-detail.component';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
import * as i2 from "./context-dashboard.service";
import * as i3 from "ngx-bootstrap/modal";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from './context-dashboard.service';
import * as ɵngcc3 from 'ngx-bootstrap/modal';
export class AddDashboardFactory {
    constructor(permissions, contextDashboardService, bsModal, contextRoute) {
        this.permissions = permissions;
        this.contextDashboardService = contextDashboardService;
        this.bsModal = bsModal;
        this.contextRoute = contextRoute;
    }
    get(activatedRoute) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            this.currentContext = this.contextRoute.getContextData(activatedRoute);
            if ([ViewContext.Device, ViewContext.Group].includes((_a = this.currentContext) === null || _a === void 0 ? void 0 : _a.context) &&
                (yield this.hasPermission(this.currentContext.contextData))) {
                return [
                    {
                        label: 'Add dashboard',
                        component: AddDashboardComponent,
                        priority: -Infinity
                    }
                ];
            }
            return [];
        });
    }
    addDashboard() {
        return __awaiter(this, void 0, void 0, function* () {
            let initialState;
            if (this.currentContext.context === ViewContext.Device) {
                initialState = {
                    deviceType: this.currentContext.contextData.type
                };
            }
            const modal = this.bsModal.show(DashboardDetailComponent, {
                class: 'modal-lg',
                initialState,
                ignoreBackdropClick: true
            }).content;
            try {
                const dashboardCfg = yield modal.result;
                const dashboardMO = yield this.contextDashboardService.create(dashboardCfg, this.currentContext);
                yield this.contextDashboardService.navigateToDashboard(dashboardMO);
                modal.close();
            }
            catch (ex) {
                // intended empty
            }
        });
    }
    hasPermission(context) {
        return __awaiter(this, void 0, void 0, function* () {
            if (context === null || context === void 0 ? void 0 : context.id) {
                return yield this.permissions.canEdit(['ROLE_INVENTORY_ADMIN', 'ROLE_INVENTORY_CREATE'], context);
            }
            return (this.permissions.hasAnyRole(['ROLE_INVENTORY_ADMIN', 'ROLE_INVENTORY_CREATE']));
        });
    }
}
AddDashboardFactory.ɵfac = function AddDashboardFactory_Factory(t) { return new (t || AddDashboardFactory)(ɵngcc0.ɵɵinject(ɵngcc1.Permissions), ɵngcc0.ɵɵinject(ɵngcc2.ContextDashboardService), ɵngcc0.ɵɵinject(ɵngcc3.BsModalService), ɵngcc0.ɵɵinject(ɵngcc1.ContextRouteService)); };
AddDashboardFactory.ɵprov = i0.ɵɵdefineInjectable({ factory: function AddDashboardFactory_Factory() { return new AddDashboardFactory(i0.ɵɵinject(i1.Permissions), i0.ɵɵinject(i2.ContextDashboardService), i0.ɵɵinject(i3.BsModalService), i0.ɵɵinject(i1.ContextRouteService)); }, token: AddDashboardFactory, providedIn: "root" });
AddDashboardFactory.ctorParameters = () => [
    { type: Permissions },
    { type: ContextDashboardService },
    { type: BsModalService },
    { type: ContextRouteService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AddDashboardFactory, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: ɵngcc1.Permissions }, { type: ɵngcc2.ContextDashboardService }, { type: ɵngcc3.BsModalService }, { type: ɵngcc1.ContextRouteService }]; }, null); })();
