// @c8y/ngx-components/esm2015/context-dashboard/dashboard-action-bar.factory.js
import { Injectable } from '@angular/core';
import { ViewContext } from '@c8y/ngx-components';
import { ContextDashboardService } from './context-dashboard.service';
import { PasteDashboardActionComponent } from './paste-dashboard-action.component';
import * as i0 from "@angular/core";
import * as i1 from "./context-dashboard.service";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './context-dashboard.service';
export class DashboardActionBarFactory {
    constructor(contextDashboardService) {
        this.contextDashboardService = contextDashboardService;
    }
    get(activeRoute) {
        const actions = [];
        const data = !activeRoute.parent || activeRoute.snapshot.data.context
            ? activeRoute.snapshot.data
            : activeRoute.parent.snapshot.data;
        const { contextData } = data;
        const canPaste = contextData &&
            [ViewContext.Device, ViewContext.Group].includes(data.context) &&
            this.contextDashboardService.copyClipboard;
        if (canPaste) {
            actions.push({
                priority: -10,
                placement: 'more',
                template: PasteDashboardActionComponent
            });
        }
        return actions;
    }
}
DashboardActionBarFactory.ɵfac = function DashboardActionBarFactory_Factory(t) { return new (t || DashboardActionBarFactory)(ɵngcc0.ɵɵinject(ɵngcc1.ContextDashboardService)); };
DashboardActionBarFactory.ɵprov = i0.ɵɵdefineInjectable({ factory: function DashboardActionBarFactory_Factory() { return new DashboardActionBarFactory(i0.ɵɵinject(i1.ContextDashboardService)); }, token: DashboardActionBarFactory, providedIn: "root" });
DashboardActionBarFactory.ctorParameters = () => [
    { type: ContextDashboardService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DashboardActionBarFactory, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: ɵngcc1.ContextDashboardService }]; }, null); })();
