// @c8y/ngx-components/esm2015/context-dashboard/paste-dashboard-action.component.js
import { __awaiter } from "tslib";
import { Component, TemplateRef, ViewChild, ViewContainerRef } from '@angular/core';
import { Router } from '@angular/router';
import { InventoryService } from '@c8y/client';
import { ViewContext } from '@c8y/ngx-components';
import { ContextDashboardService } from './context-dashboard.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from './context-dashboard.service';
import * as ɵngcc3 from '@c8y/client';
import * as ɵngcc4 from '@c8y/ngx-components';

const _c0 = ["templatePaste"];
function PasteDashboardActionComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    const _r3 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li");
    ɵngcc0.ɵɵelementStart(1, "button", 1);
    ɵngcc0.ɵɵlistener("click", function PasteDashboardActionComponent_ng_template_0_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r3); const ctx_r2 = ɵngcc0.ɵɵnextContext(); return ctx_r2.pasteDashboard(); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 2);
    ɵngcc0.ɵɵelementStart(4, "span", 3);
    ɵngcc0.ɵɵtext(5, "Paste dashboard");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(6, "\u00A0");
    ɵngcc0.ɵɵelementStart(7, "span");
    ɵngcc0.ɵɵtext(8);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 2, "Paste dashboard"));
    ɵngcc0.ɵɵadvance(7);
    ɵngcc0.ɵɵtextInterpolate1("\"", ctx_r1.dashboardName, "\"");
} }
export class PasteDashboardActionComponent {
    constructor(vcRef, router, contextDashboardService, inventoryService) {
        this.vcRef = vcRef;
        this.router = router;
        this.contextDashboardService = contextDashboardService;
        this.inventoryService = inventoryService;
        this.dashboardName = '';
    }
    ngOnInit() {
        this.vcRef.createEmbeddedView(this.templatePaste);
        const clipboard = this.contextDashboardService.copyClipboard;
        if (clipboard) {
            this.dashboardName = clipboard.dashboard.name;
        }
    }
    pasteDashboard() {
        return __awaiter(this, void 0, void 0, function* () {
            let context;
            const currentUrl = this.router.url;
            const isDevice = new RegExp(/device\/\d+/).test(currentUrl);
            const isGroup = new RegExp(/group\/\d+/).test(currentUrl);
            const id = currentUrl.match(/\d+/)[0];
            if (isDevice) {
                context = {
                    context: ViewContext.Device,
                    contextData: yield this.getContextMo(id)
                };
            }
            else if (isGroup) {
                context = {
                    context: ViewContext.Group,
                    contextData: yield this.getContextMo(id)
                };
            }
            this.contextDashboardService.pasteDashboard(context);
        });
    }
    getContextMo(id) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.inventoryService.detail(id)).data;
        });
    }
}
PasteDashboardActionComponent.ɵfac = function PasteDashboardActionComponent_Factory(t) { return new (t || PasteDashboardActionComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Router), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ContextDashboardService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.InventoryService)); };
PasteDashboardActionComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: PasteDashboardActionComponent, selectors: [["c8y-paste-dashboard-action"]], viewQuery: function PasteDashboardActionComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 7, TemplateRef);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.templatePaste = _t.first);
    } }, decls: 2, vars: 0, consts: [["templatePaste", ""], [3, "title", "click"], ["c8yIcon", "clipboard", 1, "m-r-4"], ["translate", ""]], template: function PasteDashboardActionComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, PasteDashboardActionComponent_ng_template_0_Template, 9, 4, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, directives: [ɵngcc4.IconDirective, ɵngcc4.C8yTranslateDirective], pipes: [ɵngcc4.C8yTranslatePipe], encapsulation: 2 });
PasteDashboardActionComponent.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: Router },
    { type: ContextDashboardService },
    { type: InventoryService }
];
PasteDashboardActionComponent.propDecorators = {
    templatePaste: [{ type: ViewChild, args: ['templatePaste', { read: TemplateRef, static: true },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(PasteDashboardActionComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-paste-dashboard-action',
                template: `
    <ng-template #templatePaste>
      <li>
        <button title="{{ 'Paste dashboard' | translate }}" (click)="pasteDashboard()">
          <i class="m-r-4" c8yIcon="clipboard"></i><span translate>Paste dashboard</span>&nbsp;<span
            >"{{ dashboardName }}"</span
          >
        </button>
      </li>
    </ng-template>
  `
            }]
    }], function () { return [{ type: ɵngcc0.ViewContainerRef }, { type: ɵngcc1.Router }, { type: ɵngcc2.ContextDashboardService }, { type: ɵngcc3.InventoryService }]; }, { templatePaste: [{
            type: ViewChild,
            args: ['templatePaste', { read: TemplateRef, static: true }]
        }] }); })();
