// @c8y/ngx-components/esm2015/context-dashboard/widget-config.component.js
import { __awaiter } from "tslib";
import { Component, ViewChild } from '@angular/core';
import { clone, cloneDeep, escapeRegExp, get, omit } from 'lodash-es';
import { BsModalRef } from 'ngx-bootstrap/modal';
import { iif, Subject, timer } from 'rxjs';
import { switchMap } from 'rxjs/operators';
import { WIDGET_CONTENT_CLASSES, WIDGET_HEADER_CLASSES } from './context-dashboard.model';
import { WidgetService } from './widget.service';
import { ContextDashboardService } from './context-dashboard.service';
import { InventoryService } from '@c8y/client';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './widget.service';
import * as ɵngcc2 from 'ngx-bootstrap/modal';
import * as ɵngcc3 from '@c8y/client';
import * as ɵngcc4 from './context-dashboard.service';
import * as ɵngcc5 from '@angular/common';
import * as ɵngcc6 from '@angular/forms';
import * as ɵngcc7 from '@c8y/ngx-components';
import * as ɵngcc8 from '@c8y/ngx-components/assets-navigator';
import * as ɵngcc9 from './appearance-settings.component';
import * as ɵngcc10 from './widget-preview.component';

const _c0 = ["config"];
function WidgetConfigComponent_h4_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "h4", 21);
    ɵngcc0.ɵɵtext(1, "Add widget");
    ɵngcc0.ɵɵelementEnd();
} }
function WidgetConfigComponent_h4_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "h4", 21);
    ɵngcc0.ɵɵtext(1, "Edit widget");
    ɵngcc0.ɵɵelementEnd();
} }
const _c1 = function () { return { standalone: true }; };
function WidgetConfigComponent_div_24_Template(rf, ctx) { if (rf & 1) {
    const _r8 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 22);
    ɵngcc0.ɵɵelementStart(1, "div", 23);
    ɵngcc0.ɵɵelementStart(2, "div", 24);
    ɵngcc0.ɵɵelementStart(3, "div", 25);
    ɵngcc0.ɵɵelementStart(4, "input", 26);
    ɵngcc0.ɵɵlistener("ngModelChange", function WidgetConfigComponent_div_24_Template_input_ngModelChange_4_listener($event) { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r7 = ɵngcc0.ɵɵnextContext(); return ctx_r7.searchTerm = $event; })("keydown", function WidgetConfigComponent_div_24_Template_input_keydown_4_listener($event) { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r9 = ɵngcc0.ɵɵnextContext(); return ctx_r9.searchChange$.next($event); });
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "span", 27);
    ɵngcc0.ɵɵelementStart(7, "button", 28);
    ɵngcc0.ɵɵlistener("click", function WidgetConfigComponent_div_24_Template_button_click_7_listener() { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r10 = ɵngcc0.ɵɵnextContext(); return ctx_r10.resetSearch(); });
    ɵngcc0.ɵɵelement(8, "i", 29);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(5, 4, "Search\u2026"));
    ɵngcc0.ɵɵproperty("ngModel", ctx_r3.searchTerm)("ngModelOptions", ɵngcc0.ɵɵpureFunction0(6, _c1));
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("c8yIcon", ctx_r3.searchTerm.length === 0 ? "search" : "close");
} }
function WidgetConfigComponent_div_25_div_2_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "h1");
    ɵngcc0.ɵɵelement(2, "i", 41);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "small", 42);
    ɵngcc0.ɵɵtext(4, "Preview not available");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} }
function WidgetConfigComponent_div_25_div_2_ng_template_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "img", 43);
} if (rf & 2) {
    const cmp_r13 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("src", cmp_r13.previewImage, ɵngcc0.ɵɵsanitizeUrl);
} }
function WidgetConfigComponent_div_25_div_2_Template(rf, ctx) { if (rf & 1) {
    const _r19 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 34);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelementStart(2, "div", 35);
    ɵngcc0.ɵɵlistener("click", function WidgetConfigComponent_div_25_div_2_Template_div_click_2_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r19); const cmp_r13 = restoredCtx.$implicit; const ctx_r18 = ɵngcc0.ɵɵnextContext(2); return ctx_r18.select(cmp_r13); });
    ɵngcc0.ɵɵelementStart(3, "div", 36);
    ɵngcc0.ɵɵtemplate(4, WidgetConfigComponent_div_25_div_2_ng_container_4_Template, 5, 0, "ng-container", 37);
    ɵngcc0.ɵɵtemplate(5, WidgetConfigComponent_div_25_div_2_ng_template_5_Template, 1, 1, "ng-template", null, 38, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "p", 39);
    ɵngcc0.ɵɵelement(8, "c8y-highlight", 40);
    ɵngcc0.ɵɵpipe(9, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const cmp_r13 = ctx.$implicit;
    const _r15 = ɵngcc0.ɵɵreference(6);
    const ctx_r11 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 7, cmp_r13.description));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵclassProp("active", ctx_r11.selected === cmp_r13);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !cmp_r13.previewImage)("ngIfElse", _r15);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵpropertyInterpolate("text", ɵngcc0.ɵɵpipeBind1(9, 9, cmp_r13.label));
    ɵngcc0.ɵɵproperty("pattern", ctx_r11.searchTerm);
} }
function WidgetConfigComponent_div_25_div_3_Template(rf, ctx) { if (rf & 1) {
    const _r21 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 44);
    ɵngcc0.ɵɵelement(1, "h1", 45);
    ɵngcc0.ɵɵelementStart(2, "h3", 42);
    ɵngcc0.ɵɵtext(3, "No widgets found.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "div", 46);
    ɵngcc0.ɵɵelementStart(5, "p", 47);
    ɵngcc0.ɵɵtext(6, "Rephrase your search term.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "button", 48);
    ɵngcc0.ɵɵlistener("click", function WidgetConfigComponent_div_25_div_3_Template_button_click_7_listener() { ɵngcc0.ɵɵrestoreView(_r21); const ctx_r20 = ɵngcc0.ɵɵnextContext(2); return ctx_r20.resetSearch(); });
    ɵngcc0.ɵɵtext(8, "Reset search");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function WidgetConfigComponent_div_25_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 30);
    ɵngcc0.ɵɵelementStart(1, "div", 31);
    ɵngcc0.ɵɵtemplate(2, WidgetConfigComponent_div_25_div_2_Template, 10, 11, "div", 32);
    ɵngcc0.ɵɵtemplate(3, WidgetConfigComponent_div_25_div_3_Template, 9, 0, "div", 33);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r4.searchResult || ctx_r4.components);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r4.searchResult && ctx_r4.searchResult.length === 0);
} }
const _c2 = function (a0, a1) { return { "bg-gray-white col-sm-4 p-l-0 p-r-0": a0, "bg-white col-sm-12": a1 }; };
const _c3 = function (a0) { return { "p-r-16": a0 }; };
const _c4 = function (a1, a5, a7) { return { view: "miller", groupsSelectable: a1, showChildDevices: true, columnHeaders: true, showUnassignedDevices: true, search: a5, showFilter: true, singleColumn: a7 }; };
function WidgetConfigComponent_div_26_div_20_Template(rf, ctx) { if (rf & 1) {
    const _r25 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 63);
    ɵngcc0.ɵɵelementStart(1, "div", 64);
    ɵngcc0.ɵɵelementStart(2, "c8y-asset-selector-miller", 65);
    ɵngcc0.ɵɵlistener("onSelected", function WidgetConfigComponent_div_26_div_20_Template_c8y_asset_selector_miller_onSelected_2_listener($event) { ɵngcc0.ɵɵrestoreView(_r25); const ctx_r24 = ɵngcc0.ɵɵnextContext(2); return ctx_r24.selectionChanged($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r22 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(5, _c2, ctx_r22.hasConfig(), !ctx_r22.hasConfig()));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(8, _c3, !ctx_r22.hasConfig()));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("config", ɵngcc0.ɵɵpureFunction3(10, _c4, ctx_r22.widgetConfig.settings == null ? null : ctx_r22.widgetConfig.settings.groupsSelectable, !(ctx_r22.widgetConfig.settings.context == null ? null : ctx_r22.widgetConfig.settings.context.additionParents), !!ctx_r22.hasConfig()))("asset", ctx_r22.widgetConfig.settings == null ? null : ctx_r22.widgetConfig.settings.context)("selectedDevice", ctx_r22.selectedDevice);
} }
const _c5 = function (a0) { return { "d-flex m-l-0": a0 }; };
const _c6 = function (a0, a1) { return { "col-sm-8": a0, "col-sm-12 p-l-32 p-r-8": a1 }; };
function WidgetConfigComponent_div_26_Template(rf, ctx) { if (rf & 1) {
    const _r27 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 49);
    ɵngcc0.ɵɵelementStart(1, "div", 50);
    ɵngcc0.ɵɵelementStart(2, "div", 23);
    ɵngcc0.ɵɵelementStart(3, "div", 51);
    ɵngcc0.ɵɵelementStart(4, "div", 52);
    ɵngcc0.ɵɵelementStart(5, "h4", 53);
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵpipe(7, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(8, "p");
    ɵngcc0.ɵɵtext(9);
    ɵngcc0.ɵɵpipe(10, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(11, "div", 54);
    ɵngcc0.ɵɵelementStart(12, "div", 55);
    ɵngcc0.ɵɵelementStart(13, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(14, "label", 56);
    ɵngcc0.ɵɵtext(15, "Title");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(16, "input", 57);
    ɵngcc0.ɵɵlistener("ngModelChange", function WidgetConfigComponent_div_26_Template_input_ngModelChange_16_listener($event) { ɵngcc0.ɵɵrestoreView(_r27); const ctx_r26 = ɵngcc0.ɵɵnextContext(); return ctx_r26.selected.data.title = $event; });
    ɵngcc0.ɵɵpipe(17, "translate");
    ɵngcc0.ɵɵpipe(18, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(19, "div", 58);
    ɵngcc0.ɵɵtemplate(20, WidgetConfigComponent_div_26_div_20_Template, 3, 14, "div", 59);
    ɵngcc0.ɵɵelementStart(21, "div", 60);
    ɵngcc0.ɵɵelement(22, "c8y-dynamic-component", 61, 62);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("hidden", ctx_r5.mode !== "config");
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(7, 12, ctx_r5.selected.label), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(10, 14, ctx_r5.selected.description), " ");
    ɵngcc0.ɵɵadvance(7);
    ɵngcc0.ɵɵpropertyInterpolate2("placeholder", "", ɵngcc0.ɵɵpipeBind1(17, 16, "e.g."), " ", ɵngcc0.ɵɵpipeBind1(18, 18, ctx_r5.componentLabel), "");
    ɵngcc0.ɵɵproperty("ngModel", ctx_r5.selected.data.title);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(20, _c5, ctx_r5.widgetConfig.options || ctx_r5.hasConfig()));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !(ctx_r5.widgetConfig.settings == null ? null : ctx_r5.widgetConfig.settings.noDeviceTarget));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(22, _c6, !(ctx_r5.widgetConfig.settings == null ? null : ctx_r5.widgetConfig.settings.noDeviceTarget), ctx_r5.widgetConfig.settings == null ? null : ctx_r5.widgetConfig.settings.noDeviceTarget));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("componentId", ctx_r5.selected.id)("config", ctx_r5.widgetConfig)("notFoundError", false);
} }
function WidgetConfigComponent_div_27_Template(rf, ctx) { if (rf & 1) {
    const _r29 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 66);
    ɵngcc0.ɵɵelementStart(1, "div", 23);
    ɵngcc0.ɵɵelementStart(2, "div", 67);
    ɵngcc0.ɵɵelementStart(3, "c8y-appearance-settings", 68);
    ɵngcc0.ɵɵlistener("themeClassChange", function WidgetConfigComponent_div_27_Template_c8y_appearance_settings_themeClassChange_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r29); const ctx_r28 = ɵngcc0.ɵɵnextContext(); return ctx_r28.styling.contentClass = $event; })("headerClassChange", function WidgetConfigComponent_div_27_Template_c8y_appearance_settings_headerClassChange_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r29); const ctx_r30 = ɵngcc0.ɵɵnextContext(); return ctx_r30.styling.headerClass = $event; });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "div", 69);
    ɵngcc0.ɵɵelement(5, "c8y-widget-preview", 70);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("themeClass", ctx_r6.styling.contentClass)("headerClass", ctx_r6.styling.headerClass)("possibleStylingTheme", ctx_r6.possibleStyling.WIDGET_CONTENT_CLASSES)("possibleStylingHeader", ctx_r6.possibleStyling.WIDGET_HEADER_CLASSES)("defaultThemeClass", ctx_r6.defaultStyling.contentClass)("defaultHeaderClass", ctx_r6.defaultStyling.headerClass);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("previewClasses", ctx_r6.getStyle(true));
} }
const _c7 = function (a0) { return { widgetName: a0 }; };
export class WidgetConfigComponent {
    constructor(widgetService, modal, inventory, contextDashboardService) {
        this.widgetService = widgetService;
        this.modal = modal;
        this.inventory = inventory;
        this.contextDashboardService = contextDashboardService;
        this.mode = 'select';
        this.searchChange$ = new Subject();
        this.searchTerm = '';
        this.styling = {
            headerClass: 'panel-title-regular',
            contentClass: 'panel-content-light'
        };
        this.defaultStyling = {
            headerClass: 'panel-title-regular',
            contentClass: 'panel-content-light'
        };
        this.possibleStyling = { WIDGET_HEADER_CLASSES, WIDGET_CONTENT_CLASSES };
        this.isUpgrade = false;
        this.result = new Promise((resolve, reject) => {
            this._save = resolve;
            this._cancel = reject;
        });
    }
    get isEdit() {
        return !!this.current;
    }
    get isDeviceTypeDashboard() {
        var _a, _b;
        return !!((_a = this.mo.c8y_Dashboard) === null || _a === void 0 ? void 0 : _a.deviceType) && !!((_b = this.mo.c8y_Dashboard) === null || _b === void 0 ? void 0 : _b.deviceTypeValue);
    }
    ngAfterContentInit() {
        return __awaiter(this, void 0, void 0, function* () {
            this.components = this.widgetService.getWidgetDefinitions();
            if (this.selected) {
                this.current = clone(this.selected);
                this.select(this.selected, this.isEdit ? 'config' : 'select');
            }
            this.searchSub = this.searchChange$
                .pipe(switchMap((event) => iif(() => event.which !== 13, timer(200), timer(10))))
                .subscribe(() => {
                this.search();
            });
            if (this.mo.c8y_Dashboard.classes) {
                this.styling = this.setDefaultStyle(Object.assign(Object.assign(Object.assign({}, this.mo.c8y_Dashboard.classes), this.mo.c8y_Dashboard.widgetClasses), (this.isEdit ? this.current.data.classes : {})));
                this.defaultStyling = this.setDefaultStyle(Object.assign(Object.assign({}, this.mo.c8y_Dashboard.classes), this.mo.c8y_Dashboard.widgetClasses));
            }
            if (this.widgetConfig && this.widgetConfig.device) {
                const { data, res } = yield this.inventory.detail(this.widgetConfig.device.id);
                this.selectedDevice = data;
            }
        });
    }
    checkIfDeviceRequired() {
        return (!this.widgetConfig.settings.deviceTargetNotRequired &&
            !this.widgetConfig.device &&
            !this.widgetConfig.settings.noDeviceTarget);
    }
    selectionChanged(selection) {
        this.widgetConfig.device = selection.change.item;
    }
    save() {
        return __awaiter(this, void 0, void 0, function* () {
            const hookSuccess = yield this.dynamicComponent.callLifeCycleHooks().toPromise();
            if (!hookSuccess) {
                return;
            }
            const { _x, _y, _width, _height } = this.selected.data;
            if (this.widgetConfig && this.widgetConfig.device) {
                const { id, name } = this.widgetConfig.device;
                this.widgetConfig.device = { id, name };
            }
            const widget = Object.assign({ _x,
                _y,
                _width,
                _height, config: omit(this.widgetConfig, 'settings'), title: this.selected.data.title, componentId: this.selected.id, id: this.isEdit ? this.current.data.id : String(Math.random()).substr(2), classes: this.getStyle() }, (!this.isEdit ? this.widgetConfig.settings.widgetDefaults : {}));
            this._save(widget);
        });
    }
    select(cmp, mode = 'config') {
        cmp.data = cmp.data || {};
        this.selected = cmp;
        this.isUpgrade = !!get(cmp, 'data.settings.upgrade');
        this.contextDashboardService.formDisabled = this.isUpgrade;
        if (this.isEdit) {
            const { _x, _y, _width, _height, classes, title } = this.current.data;
            this.selected.data = Object.assign(Object.assign({}, this.selected.data), { _x, _y, _width, _height, classes, title });
        }
        this.widgetConfig = cloneDeep(this.composeWidgetConfig(this.selected, this.context));
        this.selected.data.title = this.selected.data.title || cmp.label;
        this.componentLabel = cmp.label;
        this.mode = mode;
    }
    search() {
        if (this.searchTerm.length > 0) {
            this.searchResult = this.components.filter(cmp => new RegExp(escapeRegExp(this.searchTerm.trim()), 'i').test(cmp.label));
        }
        else {
            this.resetSearch();
        }
    }
    resetSearch() {
        this.searchTerm = '';
        this.searchResult = undefined;
    }
    changeMode(mode) {
        this.mode = mode;
    }
    close() {
        this._cancel();
        this.modal.hide();
    }
    getStyle(isPreview = false) {
        const cssClasses = {};
        if (isPreview || !this.isDashboardDefaultStyle(this.styling.headerClass)) {
            cssClasses[this.styling.headerClass] = true;
        }
        if (isPreview || !this.isDashboardDefaultStyle(this.styling.contentClass)) {
            cssClasses[this.styling.contentClass] = true;
        }
        if (isPreview) {
            cssClasses[`dashboard-theme-${this.defaultStyling.contentClass.split('-').pop()}`] = true;
        }
        return cssClasses;
    }
    ngOnDestroy() {
        this.contextDashboardService.formDisabled = true;
        if (this.searchSub) {
            this.searchSub.unsubscribe();
        }
    }
    hasConfig() {
        var _a, _b, _c, _d;
        if ((_a = this.widgetConfig.settings) === null || _a === void 0 ? void 0 : _a.upgrade) {
            return (((_b = this.widgetConfig.settings) === null || _b === void 0 ? void 0 : _b.configComponent) || ((_c = this.widgetConfig.settings) === null || _c === void 0 ? void 0 : _c.configTemplateUrl));
        }
        return !((_d = this.dynamicComponent) === null || _d === void 0 ? void 0 : _d.error);
    }
    isDashboardDefaultStyle(className) {
        const allClasses = Object.assign(Object.assign({}, this.mo.c8y_Dashboard.classes), this.mo.c8y_Dashboard.widgetClasses);
        const styles = Object.keys(allClasses).map(cssClass => ({ class: cssClass }));
        const style = this.contextDashboardService.getStyling(styles, className.split('-').pop(), undefined);
        return !!style;
    }
    setDefaultStyle(setClasses) {
        let contentClass = this.styling.contentClass;
        let headerClass = this.styling.headerClass;
        const styles = this.contextDashboardService
            .getFilteredDashboardStyles(Object.keys(setClasses))
            .map(c => c.split('-').pop());
        styles.forEach(styleName => {
            contentClass = this.contextDashboardService.getStyling(WIDGET_CONTENT_CLASSES, styleName, contentClass);
            headerClass = this.contextDashboardService.getStyling(WIDGET_HEADER_CLASSES, styleName, headerClass);
        });
        return { headerClass, contentClass };
    }
    composeWidgetConfig(selectedComponent, context = {}) {
        const setting = Object.assign({ settings: Object.assign(Object.assign(Object.assign(Object.assign({}, selectedComponent.data.settings), get(selectedComponent.data.settings, 'ng1.options')), get(selectedComponent.data, 'ng1.options')), { context, dashboardMo: this.mo.c8y_Dashboard }) }, selectedComponent.data.config);
        return this.applyTargetIfDeviceDashboard(setting);
    }
    applyTargetIfDeviceDashboard(widgetConfig) {
        const isDeviceType = this.contextDashboardService.isDeviceType(this.mo);
        if (isDeviceType) {
            widgetConfig.settings.hideTarget = isDeviceType;
            const noDeviceTarget = widgetConfig.settings.ng1
                ? widgetConfig.settings.ng1.options.noDeviceTarget
                : widgetConfig.settings.noDeviceTarget;
            if (!noDeviceTarget) {
                widgetConfig.device = {
                    id: this.context.id,
                    name: this.context.name
                };
            }
        }
        return widgetConfig;
    }
}
WidgetConfigComponent.ɵfac = function WidgetConfigComponent_Factory(t) { return new (t || WidgetConfigComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.WidgetService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.BsModalRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.InventoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.ContextDashboardService)); };
WidgetConfigComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: WidgetConfigComponent, selectors: [["c8y-widget-config"]], viewQuery: function WidgetConfigComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.dynamicComponent = _t.first);
    } }, decls: 35, vars: 25, consts: [[1, "modal-header", "separator"], ["class", "text-medium", "translate", "", 4, "ngIf"], ["name", "form"], ["configForm", "ngForm"], [1, "c8y-modal-tabs"], [1, "tabContainer"], [1, "nav", "nav-tabs", "nav-tabsc8y", "p-l-24"], ["type", "button", 1, "btn", 3, "click"], ["c8yIcon", "th-large"], ["translate", "", 1, "txt"], ["type", "button", 1, "btn", 3, "disabled", "click"], ["c8yIcon", "cog"], ["c8yIcon", "paint-brush"], [1, "modal-inner-scroll"], ["class", "bg-white p-l-24 p-r-24 p-t-8 p-b-8 sticky-header-top-0", "style", "z-index: 2", 4, "ngIf"], ["class", "modal-body bg-gray-lighter", 4, "ngIf"], ["style", "max-height: calc(100vh - 290px)", 3, "hidden", 4, "ngIf"], ["class", "modal-body p-t-0", "style", "min-height: calc(100vh - 290px)", 4, "ngIf"], [1, "modal-footer"], ["type", "button", "translate", "", 1, "btn", "btn-default", 3, "title", "click"], ["translate", "", "c8yProductExperience", "", 1, "btn", "btn-primary", 3, "title", "disabled", "actionName", "actionData", "click"], ["translate", "", 1, "text-medium"], [1, "bg-white", "p-l-24", "p-r-24", "p-t-8", "p-b-8", "sticky-header-top-0", 2, "z-index", "2"], [1, "row"], [1, "col-sm-6"], [1, "input-group", "input-group-search"], ["type", "text", 1, "form-control", 3, "placeholder", "ngModel", "ngModelOptions", "ngModelChange", "keydown"], [1, "input-group-btn"], ["type", "button", 1, "btn", "btn-clean", 3, "click"], [3, "c8yIcon"], [1, "modal-body", "bg-gray-lighter"], [1, "card-group", "card-select", "m-b-0"], ["class", "col-md-3 col-sm-4 col-xs-6", 3, "title", 4, "ngFor", "ngForOf"], ["class", "c8y-empty-state text-center", 4, "ngIf"], [1, "col-md-3", "col-sm-4", "col-xs-6", 3, "title"], [1, "card", "p-8", 3, "click"], [1, "text-center", "p-8", "m-b-8", "flex-col", "flex-center", 2, "min-height", "170px", "background-color", "var(--body-background-color, #f2f3f4)"], [4, "ngIf", "ngIfElse"], ["previewImage", ""], [1, "card-title", "text-truncate"], [3, "text", "pattern"], ["c8yIcon", "file-image-o"], ["translate", ""], [1, "img-responsive", 3, "src"], [1, "c8y-empty-state", "text-center"], ["c8yIcon", "search"], [1, "d-flex"], ["translate", "", 1, "m-r-8"], ["translate", "", 1, "btn", "btn-primary", 3, "click"], [2, "max-height", "calc(100vh - 290px)", 3, "hidden"], [1, "p-t-16", "flex-no-grow"], [1, "col-sm-4"], [1, "p-l-24"], [1, "text-left", "text-medium"], [1, "col-sm-8"], [1, "p-r-24"], ["for", "widgetTitle", "translate", ""], ["id", "widgetTitle", "type", "text", "name", "title", "required", "", 1, "form-control", 3, "ngModel", "placeholder", "ngModelChange"], [1, "row", "flex-grow", "separator-top", 3, "ngClass"], ["class", "a-i-stretch", "style", "height: 360px", 3, "ngClass", 4, "ngIf"], [3, "ngClass"], ["mode", "config", 1, "d-block", "p-t-8", "p-r-24", 3, "componentId", "config", "notFoundError"], ["config", ""], [1, "a-i-stretch", 2, "height", "360px", 3, "ngClass"], [1, "p-l-16", "fit-h", "bg-inherit", 3, "ngClass"], [1, "bg-inherit", "p-relative", 3, "config", "asset", "selectedDevice", "onSelected"], [1, "modal-body", "p-t-0", 2, "min-height", "calc(100vh - 290px)"], [1, "col-xs-6"], [3, "themeClass", "headerClass", "possibleStylingTheme", "possibleStylingHeader", "defaultThemeClass", "defaultHeaderClass", "themeClassChange", "headerClassChange"], [1, "col-xs-6", "sticky-header-top-0"], [3, "previewClasses"]], template: function WidgetConfigComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, WidgetConfigComponent_h4_1_Template, 2, 0, "h4", 1);
        ɵngcc0.ɵɵtemplate(2, WidgetConfigComponent_h4_2_Template, 2, 0, "h4", 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "form", 2, 3);
        ɵngcc0.ɵɵelementStart(5, "div", 4);
        ɵngcc0.ɵɵelementStart(6, "div", 5);
        ɵngcc0.ɵɵelementStart(7, "ul", 6);
        ɵngcc0.ɵɵelementStart(8, "li");
        ɵngcc0.ɵɵelementStart(9, "button", 7);
        ɵngcc0.ɵɵlistener("click", function WidgetConfigComponent_Template_button_click_9_listener() { ctx.changeMode("select"); return false; });
        ɵngcc0.ɵɵelement(10, "i", 8);
        ɵngcc0.ɵɵelementStart(11, "span", 9);
        ɵngcc0.ɵɵtext(12, "Select widget");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(13, "li");
        ɵngcc0.ɵɵelementStart(14, "button", 10);
        ɵngcc0.ɵɵlistener("click", function WidgetConfigComponent_Template_button_click_14_listener() { ctx.changeMode("config"); return false; });
        ɵngcc0.ɵɵelement(15, "i", 11);
        ɵngcc0.ɵɵelementStart(16, "span", 9);
        ɵngcc0.ɵɵtext(17, "Configuration");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(18, "li");
        ɵngcc0.ɵɵelementStart(19, "button", 10);
        ɵngcc0.ɵɵlistener("click", function WidgetConfigComponent_Template_button_click_19_listener() { ctx.changeMode("style"); return false; });
        ɵngcc0.ɵɵelement(20, "i", 12);
        ɵngcc0.ɵɵelementStart(21, "span", 9);
        ɵngcc0.ɵɵtext(22, "Appearance");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(23, "div", 13);
        ɵngcc0.ɵɵtemplate(24, WidgetConfigComponent_div_24_Template, 9, 7, "div", 14);
        ɵngcc0.ɵɵtemplate(25, WidgetConfigComponent_div_25_Template, 4, 2, "div", 15);
        ɵngcc0.ɵɵtemplate(26, WidgetConfigComponent_div_26_Template, 24, 25, "div", 16);
        ɵngcc0.ɵɵtemplate(27, WidgetConfigComponent_div_27_Template, 6, 7, "div", 17);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(28, "div", 18);
        ɵngcc0.ɵɵelementStart(29, "button", 19);
        ɵngcc0.ɵɵlistener("click", function WidgetConfigComponent_Template_button_click_29_listener() { return ctx.close(); });
        ɵngcc0.ɵɵpipe(30, "translate");
        ɵngcc0.ɵɵtext(31, " Cancel ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(32, "button", 20);
        ɵngcc0.ɵɵlistener("click", function WidgetConfigComponent_Template_button_click_32_listener() { return ctx.save(); });
        ɵngcc0.ɵɵpipe(33, "translate");
        ɵngcc0.ɵɵtext(34, " Save ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r2 = ɵngcc0.ɵɵreference(4);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.current);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.current);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵclassProp("active", ctx.mode === "select");
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵclassProp("active", ctx.mode === "config");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("disabled", !ctx.selected);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵclassProp("active", ctx.mode === "style");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("disabled", !ctx.selected);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("ngIf", ctx.mode === "select");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.mode === "select");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.selected);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.mode === "style");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(30, 19, "Cancel"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(33, 21, "Save"));
        ɵngcc0.ɵɵproperty("disabled", ctx.contextDashboardService.formDisabled || _r2.invalid || ctx.checkIfDeviceRequired())("actionName", ctx.current ? "editWidget" : "createWidget")("actionData", ɵngcc0.ɵɵpureFunction1(23, _c7, ctx.selected && ctx.selected.id));
    } }, directives: [ɵngcc5.NgIf, ɵngcc6.ɵNgNoValidate, ɵngcc6.NgControlStatusGroup, ɵngcc6.NgForm, ɵngcc7.IconDirective, ɵngcc7.C8yTranslateDirective, ɵngcc7.ProductExperienceDirective, ɵngcc6.DefaultValueAccessor, ɵngcc6.NgControlStatus, ɵngcc6.NgModel, ɵngcc5.NgForOf, ɵngcc7.HighlightComponent, ɵngcc7.FormGroupComponent, ɵngcc7.RequiredInputPlaceholderDirective, ɵngcc6.RequiredValidator, ɵngcc5.NgClass, ɵngcc7.DynamicComponentComponent, ɵngcc8.MillerViewComponent, ɵngcc9.AppearanceSettingsComponent, ɵngcc10.WidgetPreviewComponent], pipes: [ɵngcc7.C8yTranslatePipe], encapsulation: 2 });
WidgetConfigComponent.ctorParameters = () => [
    { type: WidgetService },
    { type: BsModalRef },
    { type: InventoryService },
    { type: ContextDashboardService }
];
WidgetConfigComponent.propDecorators = {
    dynamicComponent: [{ type: ViewChild, args: ['config', { static: false },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(WidgetConfigComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-widget-config',
                template: "<div class=\"modal-header separator\">\n  <h4 class=\"text-medium\" *ngIf=\"!current\" translate>Add widget</h4>\n  <h4 class=\"text-medium\" *ngIf=\"current\" translate>Edit widget</h4>\n</div>\n<form #configForm=\"ngForm\" name=\"form\">\n  <div class=\"c8y-modal-tabs\">\n    <div class=\"tabContainer\">\n      <ul class=\"nav nav-tabs nav-tabsc8y p-l-24\">\n        <li [class.active]=\"mode === 'select'\">\n          <button type=\"button\" class=\"btn\" (click)=\"changeMode('select'); (false)\">\n            <i c8yIcon=\"th-large\"></i> <span class=\"txt\" translate>Select widget</span>\n          </button>\n        </li>\n        <li [class.active]=\"mode === 'config'\">\n          <button\n            type=\"button\"\n            class=\"btn\"\n            [disabled]=\"!selected\"\n            (click)=\"changeMode('config'); (false)\"\n          >\n            <i c8yIcon=\"cog\"></i> <span class=\"txt\" translate>Configuration</span>\n          </button>\n        </li>\n        <li [class.active]=\"mode === 'style'\">\n          <button\n            type=\"button\"\n            class=\"btn\"\n            [disabled]=\"!selected\"\n            (click)=\"changeMode('style'); (false)\"\n          >\n            <i c8yIcon=\"paint-brush\"></i> <span class=\"txt\" translate>Appearance</span>\n          </button>\n        </li>\n      </ul>\n    </div>\n  </div>\n\n  <div class=\"modal-inner-scroll\">\n    <div\n      *ngIf=\"mode === 'select'\"\n      class=\"bg-white p-l-24 p-r-24 p-t-8 p-b-8 sticky-header-top-0\"\n      style=\"z-index: 2\"\n    >\n      <div class=\"row\">\n        <div class=\"col-sm-6\">\n          <div class=\"input-group input-group-search\">\n            <input\n              class=\"form-control\"\n              placeholder=\"{{ 'Search\u2026' | translate }}\"\n              type=\"text\"\n              [(ngModel)]=\"searchTerm\"\n              [ngModelOptions]=\"{ standalone: true }\"\n              (keydown)=\"searchChange$.next($event)\"\n            />\n            <span class=\"input-group-btn\">\n              <button class=\"btn btn-clean\" (click)=\"resetSearch()\" type=\"button\">\n                <i [c8yIcon]=\"searchTerm.length === 0 ? 'search' : 'close'\"></i>\n              </button>\n            </span>\n          </div>\n        </div>\n      </div>\n    </div>\n    <div class=\"modal-body bg-gray-lighter\" *ngIf=\"mode === 'select'\">\n      <div class=\"card-group card-select m-b-0\">\n        <div\n          class=\"col-md-3 col-sm-4 col-xs-6\"\n          *ngFor=\"let cmp of searchResult || components\"\n          title=\"{{ cmp.description | translate }}\"\n        >\n          <div class=\"card p-8\" [class.active]=\"selected === cmp\" (click)=\"select(cmp)\">\n            <div\n              class=\"text-center p-8 m-b-8 flex-col flex-center\"\n              style=\"min-height: 170px; background-color: var(--body-background-color, #f2f3f4)\"\n            >\n              <ng-container *ngIf=\"!cmp.previewImage; else previewImage\">\n                <h1><i c8yIcon=\"file-image-o\"></i></h1>\n                <small translate>Preview not available</small>\n              </ng-container>\n              <ng-template #previewImage>\n                <img class=\"img-responsive\" [src]=\"cmp.previewImage\" />\n              </ng-template>\n            </div>\n            <p class=\"card-title text-truncate\">\n              <c8y-highlight\n                text=\"{{ cmp.label | translate }}\"\n                [pattern]=\"searchTerm\"\n              ></c8y-highlight>\n            </p>\n          </div>\n        </div>\n\n        <div class=\"c8y-empty-state text-center\" *ngIf=\"searchResult && searchResult.length === 0\">\n          <h1 c8yIcon=\"search\"></h1>\n          <h3 translate>No widgets found.</h3>\n          <div class=\"d-flex\">\n            <p translate class=\"m-r-8\">Rephrase your search term.</p>\n            <button class=\"btn btn-primary\" (click)=\"resetSearch()\" translate>Reset search</button>\n          </div>\n        </div>\n      </div>\n    </div>\n\n    <!-- The following is intentional set to hidden to allow the ViewChild ref in the controller -->\n    <div *ngIf=\"selected\" [hidden]=\"mode !== 'config'\" style=\"max-height: calc(100vh - 290px)\">\n      <div class=\"p-t-16 flex-no-grow\">\n        <div class=\"row\">\n          <div class=\"col-sm-4\">\n            <div class=\"p-l-24\">\n              <h4 class=\"text-left text-medium\">\n                {{ selected.label | translate }}\n              </h4>\n              <p>\n                {{ selected.description | translate }}\n              </p>\n            </div>\n          </div>\n          <div class=\"col-sm-8\">\n            <div class=\"p-r-24\">\n              <c8y-form-group>\n                <label for=\"widgetTitle\" translate>Title</label>\n                <input\n                  id=\"widgetTitle\"\n                  [(ngModel)]=\"selected.data.title\"\n                  type=\"text\"\n                  name=\"title\"\n                  class=\"form-control\"\n                  placeholder=\"{{ 'e.g.' | translate }} {{ componentLabel | translate }}\"\n                  required\n                />\n              </c8y-form-group>\n            </div>\n          </div>\n        </div>\n      </div>\n      <div\n        class=\"row flex-grow separator-top\"\n        [ngClass]=\"{ 'd-flex m-l-0': widgetConfig.options || hasConfig() }\"\n      >\n        <div\n          *ngIf=\"!widgetConfig.settings?.noDeviceTarget\"\n          class=\"a-i-stretch\"\n          [ngClass]=\"{\n            'bg-gray-white col-sm-4 p-l-0 p-r-0': hasConfig(),\n            'bg-white col-sm-12': !hasConfig()\n          }\"\n          style=\"height: 360px\"\n        >\n          <div class=\"p-l-16 fit-h bg-inherit\" [ngClass]=\"{ 'p-r-16': !hasConfig() }\">\n            <c8y-asset-selector-miller\n              (onSelected)=\"selectionChanged($event)\"\n              [config]=\"{\n                view: 'miller',\n                groupsSelectable: this.widgetConfig.settings?.groupsSelectable,\n                showChildDevices: true,\n                columnHeaders: true,\n                showUnassignedDevices: true,\n                search: !this.widgetConfig.settings.context?.additionParents,\n                showFilter: true,\n                singleColumn: !!this.hasConfig()\n              }\"\n              [asset]=\"widgetConfig.settings?.context\"\n              [selectedDevice]=\"selectedDevice\"\n              class=\"bg-inherit p-relative\"\n            ></c8y-asset-selector-miller>\n          </div>\n        </div>\n        <div\n          [ngClass]=\"{\n            'col-sm-8': !widgetConfig.settings?.noDeviceTarget,\n            'col-sm-12 p-l-32 p-r-8': widgetConfig.settings?.noDeviceTarget\n          }\"\n        >\n          <c8y-dynamic-component\n            class=\"d-block p-t-8 p-r-24\"\n            [componentId]=\"selected.id\"\n            mode=\"config\"\n            [config]=\"widgetConfig\"\n            [notFoundError]=\"false\"\n            #config\n          ></c8y-dynamic-component>\n        </div>\n      </div>\n    </div>\n\n    <div *ngIf=\"mode === 'style'\" class=\"modal-body p-t-0\" style=\"min-height: calc(100vh - 290px)\">\n      <div class=\"row\">\n        <div class=\"col-xs-6\">\n          <c8y-appearance-settings\n            [(themeClass)]=\"styling.contentClass\"\n            [(headerClass)]=\"styling.headerClass\"\n            [possibleStylingTheme]=\"possibleStyling.WIDGET_CONTENT_CLASSES\"\n            [possibleStylingHeader]=\"possibleStyling.WIDGET_HEADER_CLASSES\"\n            [defaultThemeClass]=\"defaultStyling.contentClass\"\n            [defaultHeaderClass]=\"defaultStyling.headerClass\"\n          >\n          </c8y-appearance-settings>\n        </div>\n        <div class=\"col-xs-6 sticky-header-top-0\">\n          <c8y-widget-preview [previewClasses]=\"getStyle(true)\"></c8y-widget-preview>\n        </div>\n      </div>\n    </div>\n  </div>\n\n  <div class=\"modal-footer\">\n    <button\n      type=\"button\"\n      title=\"{{ 'Cancel' | translate }}\"\n      class=\"btn btn-default\"\n      (click)=\"close()\"\n      translate\n    >\n      Cancel\n    </button>\n    <button\n      title=\"{{ 'Save' | translate }}\"\n      class=\"btn btn-primary\"\n      (click)=\"save()\"\n      translate\n      [disabled]=\"\n        contextDashboardService.formDisabled || configForm.invalid || checkIfDeviceRequired()\n      \"\n      c8yProductExperience\n      [actionName]=\"current ? 'editWidget' : 'createWidget'\"\n      [actionData]=\"{ widgetName: selected && selected.id }\"\n    >\n      Save\n    </button>\n  </div>\n</form>\n"
            }]
    }], function () { return [{ type: ɵngcc1.WidgetService }, { type: ɵngcc2.BsModalRef }, { type: ɵngcc3.InventoryService }, { type: ɵngcc4.ContextDashboardService }]; }, { dynamicComponent: [{
            type: ViewChild,
            args: ['config', { static: false }]
        }] }); })();
