// @c8y/ngx-components/esm2015/widgets/kpi-widget/kpi-widget.module.js
import { NgModule } from '@angular/core';
import { CoreModule, gettext, HOOK_COMPONENTS } from '@c8y/ngx-components';
import { DatapointSelectorModule } from '@c8y/ngx-components/datapoint-selector';
import { IconSelectorModule } from '@c8y/ngx-components/icon-selector';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { KpiWidgetConfigComponent } from './kpi-widget-config/kpi-widget-config.component';
import { KpiWidgetViewComponent } from './kpi-widget-view/kpi-widget-view.component';
// import for the preview img to get into the build.
import * as ɵngcc0 from '@angular/core';
let previewImage = '';
try {
    // tslint:disable-next-line: no-var-requires
    previewImage = require('@c8y/style/img/kpi-widget.png');
}
catch (ex) {
    // intended empty
}
const ɵ0 = [
    {
        id: 'kpi.widget',
        label: gettext('KPI Widget'),
        description: gettext('Displays the last measurement value of a device.'),
        component: KpiWidgetViewComponent,
        configComponent: KpiWidgetConfigComponent,
        previewImage,
        data: {
            settings: {
                noNewWidgets: false,
                widgetDefaults: {
                    _width: 2,
                    _height: 2
                },
                ng1: {
                    options: {
                        noDeviceTarget: true,
                        groupsSelectable: false
                    }
                }
            }
        }
    }
];
export class KpiWidgetModule {
}
KpiWidgetModule.ɵfac = function KpiWidgetModule_Factory(t) { return new (t || KpiWidgetModule)(); };
KpiWidgetModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: KpiWidgetModule });
KpiWidgetModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        {
            provide: HOOK_COMPONENTS,
            useValue: ɵ0,
            multi: true
        }
    ], imports: [[CoreModule, DatapointSelectorModule, IconSelectorModule, PopoverModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(KpiWidgetModule, [{
        type: NgModule,
        args: [{
                imports: [CoreModule, DatapointSelectorModule, IconSelectorModule, PopoverModule],
                declarations: [KpiWidgetViewComponent, KpiWidgetConfigComponent],
                providers: [
                    {
                        provide: HOOK_COMPONENTS,
                        useValue: ɵ0,
                        multi: true
                    }
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(KpiWidgetModule, { declarations: function () { return [KpiWidgetViewComponent, KpiWidgetConfigComponent]; }, imports: function () { return [CoreModule, DatapointSelectorModule, IconSelectorModule, PopoverModule]; } }); })();
export { ɵ0 };
