// @c8y/ngx-components/esm2015/core/login/change-password.component.js
import { __awaiter } from "tslib";
import { Component, Output, Input, EventEmitter } from '@angular/core';
import { LoginService } from './login.service';
import { PasswordService } from '../authentication/password.service';
import { UserService } from '@c8y/client';
import { OptionsService } from '../common/options.service';
import { AlertService } from '../alert/alert.service';
import { LoginViews } from './login.model';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './login.service';
import * as ɵngcc2 from '@c8y/client';
import * as ɵngcc3 from '../authentication/password.service';
import * as ɵngcc4 from '../common/options.service';
import * as ɵngcc5 from '../alert/alert.service';
import * as ɵngcc6 from '@angular/forms';
import * as ɵngcc7 from '../i18n/c8y-translate.directive';
import * as ɵngcc8 from '@angular/common';
import * as ɵngcc9 from '../forms/form-group.component';
import * as ɵngcc10 from '../forms/required-input-placeholder.directive';
import * as ɵngcc11 from './password-strength-validator.directive';
import * as ɵngcc12 from '../forms/messages.component';
import * as ɵngcc13 from '../forms/message.directive';
import * as ɵngcc14 from '../authentication/password-confirm.directive';
import * as ɵngcc15 from '../authentication/password-strength.component';
import * as ɵngcc16 from '../authentication/password-check-list.component';
import * as ɵngcc17 from '../i18n/c8y-translate.pipe';

function ChangePasswordComponent_c8y_form_group_4_Template(rf, ctx) { if (rf & 1) {
    const _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-form-group", 19);
    ɵngcc0.ɵɵelementStart(1, "label", 4);
    ɵngcc0.ɵɵtext(2, "Tenant ID");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "input", 20, 21);
    ɵngcc0.ɵɵlistener("ngModelChange", function ChangePasswordComponent_c8y_form_group_4_Template_input_ngModelChange_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r7); const ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.model.tenantId = $event; });
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(5, 2, "Tenant ID"));
    ɵngcc0.ɵɵproperty("ngModel", ctx_r1.model.tenantId);
} }
export class ChangePasswordComponent {
    constructor(loginService, users, passwordStrengthCheckerService, options, alert) {
        this.loginService = loginService;
        this.users = users;
        this.passwordStrengthCheckerService = passwordStrengthCheckerService;
        this.options = options;
        this.alert = alert;
        this.onChangeView = new EventEmitter();
        this.passwordPattern = /^[a-zA-Z0-9`~!@#$%^&*()_|+\-=?;:'",.<>{}[\]\\/]{8,32}$/;
        this.isLoading = false;
        this.model = {
            tenantId: '',
            email: '',
            newPassword: '',
            newPasswordConfirm: ''
        };
        this.emailReadOnly = false;
        this.passwordStrengthEnforced = false;
        this.TOKEN_PARAM = 'token';
        this.EMAIL_PARAM = 'email';
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            this.model.tenantId = this.loginService.getTenant();
            this.model.email = this.options.get(this.EMAIL_PARAM, '');
            this.emailReadOnly = !!this.model.email;
            this.passwordStrengthEnforced = yield this.loginService.getPasswordStrengthEnforced();
        });
    }
    changePassword() {
        return __awaiter(this, void 0, void 0, function* () {
            const passwordStrengthColor = this.passwordStrengthCheckerService.getStrengthColor(this.model.newPassword);
            const resetPassword = {
                token: this.credentials.token,
                email: this.model.email,
                newPassword: this.model.newPassword,
                passwordStrength: passwordStrengthColor.passwordStrength
            };
            try {
                this.isLoading = true;
                const { res } = yield this.users.resetPassword(resetPassword, this.model.tenantId);
                if (res.status === 200) {
                    this.loginService.addSuccessMessage('password_changed');
                    this.credentials.token = undefined;
                    this.options.set(this.TOKEN_PARAM, undefined);
                    if (this.loginService.showTenantSetup()) {
                        this.onChangeView.emit({ view: LoginViews.TenantIdSetup });
                    }
                    else {
                        this.onChangeView.emit({ view: LoginViews.Credentials });
                    }
                }
            }
            catch (e) {
                this.alert.addServerFailure(e);
            }
            finally {
                this.loginService.reset();
                this.isLoading = false;
            }
        });
    }
}
ChangePasswordComponent.ɵfac = function ChangePasswordComponent_Factory(t) { return new (t || ChangePasswordComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.LoginService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.UserService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.PasswordService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.OptionsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.AlertService)); };
ChangePasswordComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ChangePasswordComponent, selectors: [["c8y-change-password"]], inputs: { credentials: "credentials" }, outputs: { onChangeView: "onChangeView" }, decls: 35, vars: 26, consts: [["role", "form", "novalidate", "", 1, "loginForm", 3, "ngSubmit"], ["changePasswordForm", "ngForm"], ["translate", "", 1, "legend", "form-block", "center"], ["class", "tenantField", "id", "tenantField", 4, "ngIf"], ["translate", ""], ["type", "text", "name", "email", "autocapitalize", "off", "autocorrect", "off", "email", "", "required", "", 1, "form-control", 3, "ngModel", "placeholder", "readonly", "ngModelChange"], ["email", "ngModel"], [1, "row", "content-flex-50"], [1, "col-7"], ["type", "password", "name", "newPassword", "autocomplete", "new-password", "required", "", 1, "form-control", 3, "ngModel", "placeholder", "pattern", "passwordStrengthEnforced", "ngModelChange"], ["newPassword", "ngModel"], ["name", "pattern", 3, "text"], ["type", "password", "name", "newPasswordConfirm", "passwordConfirm", "newPassword", "autocomplete", "new-password", "required", "", 1, "form-control", 3, "ngModel", "placeholder", "ngModelChange"], ["newPasswordConfirm", "ngModel"], ["name", "passwordConfirm", 3, "text"], [1, "col-5"], [3, "password", "hidden"], [1, "m-t-32", 3, "password", "hidden"], ["type", "submit", "translate", "", 1, "btn", "btn-primary", "btn-lg", "btn-block", "form-group", 3, "title", "disabled"], ["id", "tenantField", 1, "tenantField"], ["type", "text", "name", "tenantId", "autocapitalize", "off", "autocorrect", "off", "required", "", 1, "form-control", 3, "ngModel", "placeholder", "ngModelChange"], ["tenantId", "ngModel"]], template: function ChangePasswordComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "form", 0, 1);
        ɵngcc0.ɵɵlistener("ngSubmit", function ChangePasswordComponent_Template_form_ngSubmit_0_listener() { return ctx.changePassword(); });
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵtext(3, " Change password ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(4, ChangePasswordComponent_c8y_form_group_4_Template, 6, 4, "c8y-form-group", 3);
        ɵngcc0.ɵɵelementStart(5, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(6, "label", 4);
        ɵngcc0.ɵɵtext(7, "Email address");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(8, "input", 5, 6);
        ɵngcc0.ɵɵlistener("ngModelChange", function ChangePasswordComponent_Template_input_ngModelChange_8_listener($event) { return ctx.model.email = $event; });
        ɵngcc0.ɵɵpipe(10, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "div", 7);
        ɵngcc0.ɵɵelementStart(12, "div", 8);
        ɵngcc0.ɵɵelementStart(13, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(14, "label", 4);
        ɵngcc0.ɵɵtext(15, "New password");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(16, "input", 9, 10);
        ɵngcc0.ɵɵlistener("ngModelChange", function ChangePasswordComponent_Template_input_ngModelChange_16_listener($event) { return ctx.model.newPassword = $event; });
        ɵngcc0.ɵɵpipe(18, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(19, "c8y-messages");
        ɵngcc0.ɵɵelement(20, "c8y-message", 11);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(21, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(22, "label", 4);
        ɵngcc0.ɵɵtext(23, "Confirm password");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(24, "input", 12, 13);
        ɵngcc0.ɵɵlistener("ngModelChange", function ChangePasswordComponent_Template_input_ngModelChange_24_listener($event) { return ctx.model.newPasswordConfirm = $event; });
        ɵngcc0.ɵɵpipe(26, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(27, "c8y-messages");
        ɵngcc0.ɵɵelement(28, "c8y-message", 14);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(29, "div", 15);
        ɵngcc0.ɵɵelement(30, "c8y-password-strength", 16);
        ɵngcc0.ɵɵelement(31, "c8y-password-check-list", 17);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(32, "button", 18);
        ɵngcc0.ɵɵpipe(33, "translate");
        ɵngcc0.ɵɵtext(34, " Set password ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r0 = ɵngcc0.ɵɵreference(1);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngIf", ctx.loginService.showTenant());
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(10, 18, "Email address"));
        ɵngcc0.ɵɵproperty("ngModel", ctx.model.email)("readonly", ctx.emailReadOnly);
        ɵngcc0.ɵɵadvance(8);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(18, 20, "New password"));
        ɵngcc0.ɵɵproperty("ngModel", ctx.model.newPassword)("pattern", ctx.passwordPattern)("passwordStrengthEnforced", ctx.passwordStrengthEnforced);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("text", ctx.loginService.ERROR_MESSAGES.pattern_newPassword);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(26, 22, "Confirm password"));
        ɵngcc0.ɵɵproperty("ngModel", ctx.model.newPasswordConfirm);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("text", ctx.loginService.ERROR_MESSAGES.passwordConfirm);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("password", ctx.model.newPassword)("hidden", !ctx.model.newPassword);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("password", ctx.model.newPassword)("hidden", !ctx.model.newPassword || !ctx.passwordStrengthEnforced);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(33, 24, "Set password"));
        ɵngcc0.ɵɵproperty("disabled", !_r0.form.valid || ctx.isLoading);
    } }, directives: [ɵngcc6.ɵNgNoValidate, ɵngcc6.NgControlStatusGroup, ɵngcc6.NgForm, ɵngcc7.C8yTranslateDirective, ɵngcc8.NgIf, ɵngcc9.FormGroupComponent, ɵngcc10.RequiredInputPlaceholderDirective, ɵngcc6.DefaultValueAccessor, ɵngcc6.EmailValidator, ɵngcc6.RequiredValidator, ɵngcc6.NgControlStatus, ɵngcc6.NgModel, ɵngcc6.PatternValidator, ɵngcc11.PasswordStrengthValidatorDirective, ɵngcc12.MessagesComponent, ɵngcc13.MessageDirective, ɵngcc14.PasswordConfirm, ɵngcc15.PasswordStrengthComponent, ɵngcc16.PasswordCheckListComponent], pipes: [ɵngcc17.C8yTranslatePipe], encapsulation: 2 });
ChangePasswordComponent.ctorParameters = () => [
    { type: LoginService },
    { type: UserService },
    { type: PasswordService },
    { type: OptionsService },
    { type: AlertService }
];
ChangePasswordComponent.propDecorators = {
    credentials: [{ type: Input }],
    onChangeView: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ChangePasswordComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-change-password',
                template: "<form\n  role=\"form\"\n  class=\"loginForm\"\n  (ngSubmit)=\"changePassword()\"\n  #changePasswordForm=\"ngForm\"\n  novalidate\n>\n\n  <div\n    class=\"legend form-block center\"\n    translate\n  >\n    Change password\n  </div>\n\n  <c8y-form-group class=\"tenantField\" id=\"tenantField\" *ngIf=\"loginService.showTenant()\">\n    <label translate>Tenant ID</label>\n    <input\n      [(ngModel)]=\"model.tenantId\"\n      #tenantId=\"ngModel\"\n      type=\"text\"\n      name=\"tenantId\"\n      autocapitalize=\"off\"\n      autocorrect=\"off\"\n      class=\"form-control\"\n      placeholder=\"{{ 'Tenant ID' | translate }}\"\n      required\n    />\n  </c8y-form-group>\n\n  <c8y-form-group>\n    <label translate>Email address</label>\n    <input\n      [(ngModel)]=\"model.email\"\n      #email=\"ngModel\"\n      type=\"text\"\n      name=\"email\"\n      autocapitalize=\"off\"\n      autocorrect=\"off\"\n      class=\"form-control\"\n      placeholder=\"{{ 'Email address' | translate }}\"\n      email\n      required\n      [readonly]=\"emailReadOnly\"\n    />\n  </c8y-form-group>\n\n  <div class=\"row content-flex-50\">\n    <div class=\"col-7\">\n      <c8y-form-group>\n        <label translate>New password</label>\n        <input\n          [(ngModel)]=\"model.newPassword\"\n          #newPassword=\"ngModel\"\n          type=\"password\"\n          name=\"newPassword\"\n          class=\"form-control\"\n          placeholder=\"{{ 'New password' | translate }}\"\n          [pattern]=\"passwordPattern\"\n          autocomplete=\"new-password\"\n          [passwordStrengthEnforced]='passwordStrengthEnforced'\n          required\n        />\n        <c8y-messages>\n          <c8y-message\n            name=\"pattern\"\n            [text]=\"loginService.ERROR_MESSAGES.pattern_newPassword\"\n          ></c8y-message>\n        </c8y-messages>\n      </c8y-form-group>\n\n      <c8y-form-group>\n        <label translate>Confirm password</label>\n        <input\n          [(ngModel)]=\"model.newPasswordConfirm\"\n          #newPasswordConfirm=\"ngModel\"\n          type=\"password\"\n          name=\"newPasswordConfirm\"\n          class=\"form-control\"\n          placeholder=\"{{ 'Confirm password' | translate }}\"\n          passwordConfirm=\"newPassword\"\n          autocomplete=\"new-password\"\n          required\n        />\n        <c8y-messages>\n          <c8y-message\n            name=\"passwordConfirm\"\n            [text]=\"loginService.ERROR_MESSAGES.passwordConfirm\"\n          ></c8y-message>\n        </c8y-messages>\n      </c8y-form-group>\n    </div>\n    <div class=\"col-5\">\n      <c8y-password-strength\n        [password]=\"model.newPassword\"\n        [hidden]=\"!model.newPassword\"\n      ></c8y-password-strength>\n      <c8y-password-check-list\n        class=\"m-t-32\"\n        [password]=\"model.newPassword\"\n        [hidden]=\"!model.newPassword || !passwordStrengthEnforced\"\n      ></c8y-password-check-list>\n    </div>\n  </div>\n\n  <button\n    title=\"{{ 'Set password' | translate }}\"\n    [disabled]=\"!changePasswordForm.form.valid || isLoading\"\n    type=\"submit\"\n    class=\"btn btn-primary btn-lg btn-block form-group\"\n    translate\n  >\n    Set password\n  </button>\n</form>\n"
            }]
    }], function () { return [{ type: ɵngcc1.LoginService }, { type: ɵngcc2.UserService }, { type: ɵngcc3.PasswordService }, { type: ɵngcc4.OptionsService }, { type: ɵngcc5.AlertService }]; }, { onChangeView: [{
            type: Output
        }], credentials: [{
            type: Input
        }] }); })();
