// @c8y/ngx-components/esm2015/core/login/login.component.js
import { __awaiter } from "tslib";
import { Component, Input, HostListener } from '@angular/core';
import { TenantLoginOptionType } from '@c8y/client';
import { LoginService } from './login.service';
import { OptionsService } from '../common/options.service';
import { LoginViews } from './login.model';
import { AlertService } from '../alert/alert.service';
import { gettext } from '../i18n/gettext';
import { CredentialsFromQueryParamsService } from './credentials-from-query-params.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './login.service';
import * as ɵngcc2 from '../common/options.service';
import * as ɵngcc3 from '../alert/alert.service';
import * as ɵngcc4 from './credentials-from-query-params.service';
import * as ɵngcc5 from '@angular/common';
import * as ɵngcc6 from '../alert/alert-outlet.component';
import * as ɵngcc7 from './credentials.component';
import * as ɵngcc8 from './recover-password.component';
import * as ɵngcc9 from './change-password.component';
import * as ɵngcc10 from './totp-auth.component';
import * as ɵngcc11 from '../authentication/sms-challenge.component';
import * as ɵngcc12 from '../authentication/provide-phone-number.component';
import * as ɵngcc13 from './tenant-id-setup.component';

function LoginComponent_div_0_c8y_credentials_3_Template(rf, ctx) { if (rf & 1) {
    const _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-credentials", 11);
    ɵngcc0.ɵɵlistener("onChangeView", function LoginComponent_div_0_c8y_credentials_3_Template_c8y_credentials_onChangeView_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r10); const ctx_r9 = ɵngcc0.ɵɵnextContext(2); return ctx_r9.handleLoginTemplate($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("loginViewParams", ctx_r1.loginViewParams);
} }
function LoginComponent_div_0_c8y_recover_password_4_Template(rf, ctx) { if (rf & 1) {
    const _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-recover-password", 12);
    ɵngcc0.ɵɵlistener("onChangeView", function LoginComponent_div_0_c8y_recover_password_4_Template_c8y_recover_password_onChangeView_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r12); const ctx_r11 = ɵngcc0.ɵɵnextContext(2); return ctx_r11.handleLoginTemplate($event); });
    ɵngcc0.ɵɵelementEnd();
} }
function LoginComponent_div_0_c8y_change_password_5_Template(rf, ctx) { if (rf & 1) {
    const _r14 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-change-password", 13);
    ɵngcc0.ɵɵlistener("onChangeView", function LoginComponent_div_0_c8y_change_password_5_Template_c8y_change_password_onChangeView_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r14); const ctx_r13 = ɵngcc0.ɵɵnextContext(2); return ctx_r13.handleLoginTemplate($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("credentials", ctx_r3.credentials);
} }
function LoginComponent_div_0_c8y_totp_auth_6_Template(rf, ctx) { if (rf & 1) {
    const _r16 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-totp-auth", 14);
    ɵngcc0.ɵɵlistener("onCancel", function LoginComponent_div_0_c8y_totp_auth_6_Template_c8y_totp_auth_onCancel_0_listener() { ɵngcc0.ɵɵrestoreView(_r16); const ctx_r15 = ɵngcc0.ɵɵnextContext(2); return ctx_r15.reset(); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("view", ctx_r4.currentView)("credentials", ctx_r4.credentials);
} }
function LoginComponent_div_0_c8y_totp_auth_7_Template(rf, ctx) { if (rf & 1) {
    const _r18 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-totp-auth", 14);
    ɵngcc0.ɵɵlistener("onCancel", function LoginComponent_div_0_c8y_totp_auth_7_Template_c8y_totp_auth_onCancel_0_listener() { ɵngcc0.ɵɵrestoreView(_r18); const ctx_r17 = ɵngcc0.ɵɵnextContext(2); return ctx_r17.reset(); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("view", ctx_r5.currentView)("credentials", ctx_r5.credentials);
} }
function LoginComponent_div_0_c8y_sms_challenge_8_Template(rf, ctx) { if (rf & 1) {
    const _r20 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-sms-challenge", 15);
    ɵngcc0.ɵɵlistener("onCancel", function LoginComponent_div_0_c8y_sms_challenge_8_Template_c8y_sms_challenge_onCancel_0_listener() { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r19 = ɵngcc0.ɵɵnextContext(2); return ctx_r19.reset(); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("credentials", ctx_r6.credentials);
} }
function LoginComponent_div_0_c8y_provide_phone_number_9_Template(rf, ctx) { if (rf & 1) {
    const _r22 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-provide-phone-number", 16);
    ɵngcc0.ɵɵlistener("onCancel", function LoginComponent_div_0_c8y_provide_phone_number_9_Template_c8y_provide_phone_number_onCancel_0_listener() { ɵngcc0.ɵɵrestoreView(_r22); const ctx_r21 = ɵngcc0.ɵɵnextContext(2); return ctx_r21.reset(); })("onChangeView", function LoginComponent_div_0_c8y_provide_phone_number_9_Template_c8y_provide_phone_number_onChangeView_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r22); const ctx_r23 = ɵngcc0.ɵɵnextContext(2); return ctx_r23.handleLoginTemplate($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("credentials", ctx_r7.credentials);
} }
function LoginComponent_div_0_c8y_tenant_id_setup_10_Template(rf, ctx) { if (rf & 1) {
    const _r25 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-tenant-id-setup", 12);
    ɵngcc0.ɵɵlistener("onChangeView", function LoginComponent_div_0_c8y_tenant_id_setup_10_Template_c8y_tenant_id_setup_onChangeView_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r25); const ctx_r24 = ɵngcc0.ɵɵnextContext(2); return ctx_r24.handleLoginTemplate($event); });
    ɵngcc0.ɵɵelementEnd();
} }
function LoginComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 1);
    ɵngcc0.ɵɵelementStart(1, "div", 2);
    ɵngcc0.ɵɵelement(2, "img", 3);
    ɵngcc0.ɵɵtemplate(3, LoginComponent_div_0_c8y_credentials_3_Template, 1, 1, "c8y-credentials", 4);
    ɵngcc0.ɵɵtemplate(4, LoginComponent_div_0_c8y_recover_password_4_Template, 1, 0, "c8y-recover-password", 5);
    ɵngcc0.ɵɵtemplate(5, LoginComponent_div_0_c8y_change_password_5_Template, 1, 1, "c8y-change-password", 6);
    ɵngcc0.ɵɵtemplate(6, LoginComponent_div_0_c8y_totp_auth_6_Template, 1, 2, "c8y-totp-auth", 7);
    ɵngcc0.ɵɵtemplate(7, LoginComponent_div_0_c8y_totp_auth_7_Template, 1, 2, "c8y-totp-auth", 7);
    ɵngcc0.ɵɵtemplate(8, LoginComponent_div_0_c8y_sms_challenge_8_Template, 1, 1, "c8y-sms-challenge", 8);
    ɵngcc0.ɵɵtemplate(9, LoginComponent_div_0_c8y_provide_phone_number_9_Template, 1, 1, "c8y-provide-phone-number", 9);
    ɵngcc0.ɵɵtemplate(10, LoginComponent_div_0_c8y_tenant_id_setup_10_Template, 1, 0, "c8y-tenant-id-setup", 5);
    ɵngcc0.ɵɵelement(11, "c8y-alert-outlet", 10);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngSwitch", ctx_r0.currentView);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngSwitchCase", ctx_r0.LOGIN_VIEWS.Credentials);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", ctx_r0.LOGIN_VIEWS.RecoverPassword);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", ctx_r0.LOGIN_VIEWS.ChangePassword);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", ctx_r0.LOGIN_VIEWS.TotpChallenge);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", ctx_r0.LOGIN_VIEWS.TotpSetup);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", ctx_r0.LOGIN_VIEWS.SmsChallenge);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", ctx_r0.LOGIN_VIEWS.ProvidePhoneNumber);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", ctx_r0.LOGIN_VIEWS.TenantIdSetup);
} }
export class LoginComponent {
    /**
     * Just DI.
     */
    constructor(loginService, options, alert, credentialsFromQueryParamsService) {
        this.loginService = loginService;
        this.options = options;
        this.alert = alert;
        this.credentialsFromQueryParamsService = credentialsFromQueryParamsService;
        this.currentView = LoginViews.None;
        this.LOGIN_VIEWS = LoginViews;
        this.disabled = false;
        this.credentials = {};
        this.loginViewParams = {};
        this.displayAlerts = false;
        this.TOKEN_PARAM = 'token';
    }
    ngOnInit() {
        const token = this.getResetPasswordToken();
        if (this.loginService.isFirstLogin) {
            if (!token) {
                this.loginAutomatically();
            }
            else {
                this.credentials.token = token;
                this.reset();
            }
        }
        this.loginService.isFirstLogin = false;
    }
    ngOnDestroy() {
        // make sure that we do not have any queryParameters related to credentials after logging in or even if we were already logged in.
        this.credentialsFromQueryParamsService.removeCredentialsFromQueryParams();
    }
    handleLoginTemplate(event) {
        this.currentView = event.view;
        this.credentials = event.credentials || {};
        this.loginViewParams = event.loginViewParams || {};
    }
    onkeyup(event) {
        if (event.key !== 'Enter') {
            this.loginService.cleanMessages();
        }
    }
    reset() {
        this.loginService.reset();
        this.setView();
        this.loginService.cleanMessages();
    }
    loginAutomatically() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.loginService.login();
            }
            catch (e) {
                const preferredLoginOptionType = this.loginService.loginMode.type;
                if (preferredLoginOptionType === TenantLoginOptionType.OAUTH2) {
                    this.loginService.redirectToOauth();
                }
                else {
                    this.reset();
                    if (preferredLoginOptionType === TenantLoginOptionType.OAUTH2_INTERNAL && window.location.protocol !== 'https:') {
                        this.alert.danger(gettext('Current login mode only supports HTTPS.'));
                    }
                    else if (e.res && e.res.status === 403) {
                        this.alert.addServerFailure(e);
                    }
                }
            }
        });
    }
    setView() {
        if (this.credentials && this.credentials.token) {
            this.handleLoginTemplate({ view: LoginViews.ChangePassword, credentials: this.credentials });
        }
        else if (this.loginService.showTenantSetup()) {
            this.handleLoginTemplate({ view: LoginViews.TenantIdSetup });
        }
        else {
            this.handleLoginTemplate({ view: LoginViews.Credentials });
        }
    }
    getResetPasswordToken() {
        const token = this.options.get(this.TOKEN_PARAM);
        if (token) {
            this.options.set(this.TOKEN_PARAM, undefined); // only use once
        }
        return token;
    }
}
LoginComponent.ɵfac = function LoginComponent_Factory(t) { return new (t || LoginComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.LoginService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.OptionsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.CredentialsFromQueryParamsService)); };
LoginComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: LoginComponent, selectors: [["c8y-login"]], hostBindings: function LoginComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("keyup", function LoginComponent_keyup_HostBindingHandler($event) { return ctx.onkeyup($event); });
    } }, inputs: { name: "name" }, decls: 1, vars: 1, consts: [["class", "loading card fadeInUp animated shadow5", 3, "ngSwitch", 4, "ngIf"], [1, "loading", "card", "fadeInUp", "animated", "shadow5", 3, "ngSwitch"], [1, "card-block", "p-b-0"], ["alt", "", 1, "mainlogo"], [3, "loginViewParams", "onChangeView", 4, "ngSwitchCase"], [3, "onChangeView", 4, "ngSwitchCase"], [3, "credentials", "onChangeView", 4, "ngSwitchCase"], [3, "view", "credentials", "onCancel", 4, "ngSwitchCase"], [3, "credentials", "onCancel", 4, "ngSwitchCase"], [3, "credentials", "onCancel", "onChangeView", 4, "ngSwitchCase"], ["position", "static"], [3, "loginViewParams", "onChangeView"], [3, "onChangeView"], [3, "credentials", "onChangeView"], [3, "view", "credentials", "onCancel"], [3, "credentials", "onCancel"], [3, "credentials", "onCancel", "onChangeView"]], template: function LoginComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, LoginComponent_div_0_Template, 12, 9, "div", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.currentView !== ctx.LOGIN_VIEWS.None);
    } }, directives: [ɵngcc5.NgIf, ɵngcc5.NgSwitch, ɵngcc5.NgSwitchCase, ɵngcc6.AlertOutletComponent, ɵngcc7.CredentialsComponent, ɵngcc8.RecoverPasswordComponent, ɵngcc9.ChangePasswordComponent, ɵngcc10.TotpAuthComponent, ɵngcc11.SmsChallengeComponent, ɵngcc12.ProvidePhoneNumberComponent, ɵngcc13.TenantIdSetupComponent], encapsulation: 2 });
LoginComponent.ctorParameters = () => [
    { type: LoginService },
    { type: OptionsService },
    { type: AlertService },
    { type: CredentialsFromQueryParamsService }
];
LoginComponent.propDecorators = {
    name: [{ type: Input }],
    onkeyup: [{ type: HostListener, args: ['keyup', ['$event'],] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(LoginComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-login',
                template: "<div\n  class=\"loading card fadeInUp animated shadow5\"\n  *ngIf=\"currentView !== LOGIN_VIEWS.None\"\n  [ngSwitch]=\"currentView\"\n>\n  <div class=\"card-block p-b-0\">\n    <img alt=\"\" class=\"mainlogo\">\n    <c8y-credentials\n      *ngSwitchCase=\"LOGIN_VIEWS.Credentials\"\n      (onChangeView)=\"handleLoginTemplate($event)\"\n      [loginViewParams]=\"loginViewParams\"\n    ></c8y-credentials>\n    <c8y-recover-password\n      *ngSwitchCase=\"LOGIN_VIEWS.RecoverPassword\"\n      (onChangeView)=\"handleLoginTemplate($event)\"\n    ></c8y-recover-password>\n    <c8y-change-password\n      *ngSwitchCase=\"LOGIN_VIEWS.ChangePassword\"\n      (onChangeView)=\"handleLoginTemplate($event)\"\n      [credentials]=\"credentials\"\n    ></c8y-change-password>\n    <c8y-totp-auth\n      *ngSwitchCase=\"LOGIN_VIEWS.TotpChallenge\"\n      (onCancel)=\"reset()\"\n      [view]=\"currentView\"\n      [credentials]=\"credentials\"\n    >\n    </c8y-totp-auth>\n    <c8y-totp-auth\n      *ngSwitchCase=\"LOGIN_VIEWS.TotpSetup\"\n      (onCancel)=\"reset()\"\n      [view]=\"currentView\"\n      [credentials]=\"credentials\"\n    >\n    </c8y-totp-auth>\n    <c8y-sms-challenge\n      *ngSwitchCase=\"LOGIN_VIEWS.SmsChallenge\"\n      (onCancel)=\"reset()\"\n      [credentials]=\"credentials\"\n    ></c8y-sms-challenge>\n\n    <c8y-provide-phone-number\n      *ngSwitchCase=\"LOGIN_VIEWS.ProvidePhoneNumber\"\n      (onCancel)=\"reset()\"\n      (onChangeView)=\"handleLoginTemplate($event)\"\n      [credentials]=\"credentials\"\n    ></c8y-provide-phone-number>\n    <c8y-tenant-id-setup\n      *ngSwitchCase=\"LOGIN_VIEWS.TenantIdSetup\"\n      (onChangeView)=\"handleLoginTemplate($event)\"\n    ></c8y-tenant-id-setup>\n\n    <c8y-alert-outlet position=\"static\"></c8y-alert-outlet>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.LoginService }, { type: ɵngcc2.OptionsService }, { type: ɵngcc3.AlertService }, { type: ɵngcc4.CredentialsFromQueryParamsService }]; }, { onkeyup: [{
            type: HostListener,
            args: ['keyup', ['$event']]
        }], name: [{
            type: Input
        }] }); })();
