// @c8y/ngx-components/esm2015/core/login/recover-password.component.js
import { __awaiter } from "tslib";
import { Component, Output, EventEmitter } from '@angular/core';
import { UserService } from '@c8y/client';
import { LoginService } from './login.service';
import { LoginViews } from './login.model';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/client';
import * as ɵngcc2 from './login.service';
import * as ɵngcc3 from '@angular/forms';
import * as ɵngcc4 from '../i18n/c8y-translate.directive';
import * as ɵngcc5 from '@angular/common';
import * as ɵngcc6 from '../forms/form-group.component';
import * as ɵngcc7 from '../forms/required-input-placeholder.directive';
import * as ɵngcc8 from '../i18n/c8y-translate.pipe';

function RecoverPasswordComponent_c8y_form_group_4_Template(rf, ctx) { if (rf & 1) {
    const _r5 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-form-group", 10);
    ɵngcc0.ɵɵelementStart(1, "label", 4);
    ɵngcc0.ɵɵtext(2, "Tenant ID");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "input", 11, 12);
    ɵngcc0.ɵɵlistener("ngModelChange", function RecoverPasswordComponent_c8y_form_group_4_Template_input_ngModelChange_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r5); const ctx_r4 = ɵngcc0.ɵɵnextContext(); return ctx_r4.model.tenantId = $event; });
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(5, 2, "Tenant ID"));
    ɵngcc0.ɵɵproperty("ngModel", ctx_r1.model.tenantId);
} }
export class RecoverPasswordComponent {
    constructor(users, loginService) {
        this.users = users;
        this.loginService = loginService;
        this.onChangeView = new EventEmitter();
        this.LOGIN_VIEWS = LoginViews;
        this.isLoading = false;
        this.model = {
            email: '',
            tenantId: ''
        };
    }
    ngOnInit() {
        this.model.tenantId = this.loginService.getTenant();
    }
    resetPassword() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                this.isLoading = true;
                const { res } = yield this.users.sendPasswordResetMail(this.model.email, this.model.tenantId);
                if (res.status === 200) {
                    this.loginService.addSuccessMessage('password_reset_requested');
                }
            }
            finally {
                this.loginService.reset();
                this.isLoading = false;
            }
        });
    }
}
RecoverPasswordComponent.ɵfac = function RecoverPasswordComponent_Factory(t) { return new (t || RecoverPasswordComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.UserService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.LoginService)); };
RecoverPasswordComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: RecoverPasswordComponent, selectors: [["c8y-recover-password"]], outputs: { onChangeView: "onChangeView" }, decls: 19, vars: 12, consts: [["novalidate", "", 1, "loginForm", 3, "ngSubmit"], ["resetForm", "ngForm"], ["translate", "", 1, "legend", "form-block", "center"], ["class", "tenantField", "id", "tenantField", 4, "ngIf"], ["translate", ""], ["type", "text", "name", "email", "autocapitalize", "off", "autocorrect", "off", "email", "", "required", "", 1, "form-control", 3, "ngModel", "placeholder", "ngModelChange"], ["email", "ngModel"], ["type", "submit", "translate", "", 1, "btn", "btn-primary", "btn-lg", "btn-block", "form-group", 3, "title", "disabled"], [1, "text-center", "m-t-8"], ["translate", "", 1, "btn", "btn-link", "btn-sm", 3, "title", "click"], ["id", "tenantField", 1, "tenantField"], ["type", "text", "name", "tenantId", "autocapitalize", "off", "autocorrect", "off", "required", "", 1, "form-control", 3, "ngModel", "placeholder", "ngModelChange"], ["tenantId", "ngModel"]], template: function RecoverPasswordComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "form", 0, 1);
        ɵngcc0.ɵɵlistener("ngSubmit", function RecoverPasswordComponent_Template_form_ngSubmit_0_listener() { return ctx.resetPassword(); });
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵtext(3, "Forgot password?");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(4, RecoverPasswordComponent_c8y_form_group_4_Template, 6, 4, "c8y-form-group", 3);
        ɵngcc0.ɵɵelementStart(5, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(6, "label", 4);
        ɵngcc0.ɵɵtext(7, "Email address");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(8, "input", 5, 6);
        ɵngcc0.ɵɵlistener("ngModelChange", function RecoverPasswordComponent_Template_input_ngModelChange_8_listener($event) { return ctx.model.email = $event; });
        ɵngcc0.ɵɵpipe(10, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "button", 7);
        ɵngcc0.ɵɵpipe(12, "translate");
        ɵngcc0.ɵɵtext(13, " Reset password ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(14, "div", 8);
        ɵngcc0.ɵɵelementStart(15, "p");
        ɵngcc0.ɵɵelementStart(16, "button", 9);
        ɵngcc0.ɵɵlistener("click", function RecoverPasswordComponent_Template_button_click_16_listener() { return ctx.onChangeView.emit({ view: ctx.LOGIN_VIEWS.Credentials }); });
        ɵngcc0.ɵɵpipe(17, "translate");
        ɵngcc0.ɵɵtext(18, " Login ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r0 = ɵngcc0.ɵɵreference(1);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngIf", ctx.loginService.showTenant());
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(10, 6, "Email address"));
        ɵngcc0.ɵɵproperty("ngModel", ctx.model.email);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(12, 8, "Reset password"));
        ɵngcc0.ɵɵproperty("disabled", !_r0.form.valid || ctx.isLoading);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(17, 10, "Login"));
    } }, directives: [ɵngcc3.ɵNgNoValidate, ɵngcc3.NgControlStatusGroup, ɵngcc3.NgForm, ɵngcc4.C8yTranslateDirective, ɵngcc5.NgIf, ɵngcc6.FormGroupComponent, ɵngcc7.RequiredInputPlaceholderDirective, ɵngcc3.DefaultValueAccessor, ɵngcc3.EmailValidator, ɵngcc3.RequiredValidator, ɵngcc3.NgControlStatus, ɵngcc3.NgModel], pipes: [ɵngcc8.C8yTranslatePipe], encapsulation: 2 });
RecoverPasswordComponent.ctorParameters = () => [
    { type: UserService },
    { type: LoginService }
];
RecoverPasswordComponent.propDecorators = {
    onChangeView: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(RecoverPasswordComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-recover-password',
                template: "<form #resetForm=\"ngForm\" class=\"loginForm\" (ngSubmit)=\"resetPassword()\" novalidate>\n  <div class=\"legend form-block center\" translate>Forgot password?</div>\n  <c8y-form-group class=\"tenantField\" id=\"tenantField\" *ngIf=\"loginService.showTenant()\">\n    <label translate>Tenant ID</label>\n    <input\n      [(ngModel)]=\"model.tenantId\"\n      #tenantId=\"ngModel\"\n      type=\"text\"\n      name=\"tenantId\"\n      autocapitalize=\"off\"\n      autocorrect=\"off\"\n      class=\"form-control\"\n      placeholder=\"{{ 'Tenant ID' | translate }}\"\n      required\n    />\n  </c8y-form-group>\n\n  <c8y-form-group>\n    <label translate>Email address</label>\n    <input\n      [(ngModel)]=\"model.email\"\n      #email=\"ngModel\"\n      type=\"text\"\n      name=\"email\"\n      autocapitalize=\"off\"\n      autocorrect=\"off\"\n      class=\"form-control\"\n      placeholder=\"{{ 'Email address' | translate }}\"\n      email\n      required\n    />\n  </c8y-form-group>\n\n  <button\n    title=\"{{ 'Reset password' | translate }}\"\n    [disabled]=\"!resetForm.form.valid || isLoading\"\n    type=\"submit\"\n    class=\"btn btn-primary btn-lg btn-block form-group\"\n    translate\n  >\n    Reset password\n  </button>\n  <div class=\"text-center m-t-8\">\n    <p>\n      <button\n        title=\"{{ 'Login' | translate }}\"\n        class=\"btn btn-link btn-sm\"\n        (click)=\"onChangeView.emit({ view: LOGIN_VIEWS.Credentials })\"\n        translate\n      >\n        Login\n      </button>\n    </p>\n  </div>\n</form>\n"
            }]
    }], function () { return [{ type: ɵngcc1.UserService }, { type: ɵngcc2.LoginService }]; }, { onChangeView: [{
            type: Output
        }] }); })();
