// @c8y/ngx-components/esm2015/core/login/tenant-id-setup.component.js
import { __awaiter } from "tslib";
import { Component, Output, EventEmitter } from '@angular/core';
import { LoginViews } from './login.model';
import { FetchClient } from '@c8y/client';
import { AppStateService } from '../common/ui-state.service';
import { LoginService } from './login.service';
import { AlertService } from '../alert/alert.service';
import { TranslateService } from '@ngx-translate/core';
import { gettext } from '../i18n/gettext';
/**
 * `TenantIdSetupComponent` is intended to be shown when tenant's id cannot be determined based on the current URL.
 * It asks the user to provide target tenant's id and then it fetches login options for this tenant.
 * In case of OAI-Secure login mode, login options will contain `domain` property set by backend.
 * The component will redirect user to this domain, preserving URL path and params.
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/client';
import * as ɵngcc2 from '../common/ui-state.service';
import * as ɵngcc3 from './login.service';
import * as ɵngcc4 from '../alert/alert.service';
import * as ɵngcc5 from '@ngx-translate/core';
import * as ɵngcc6 from '@angular/forms';
import * as ɵngcc7 from '../i18n/c8y-translate.directive';
import * as ɵngcc8 from '../forms/form-group.component';
import * as ɵngcc9 from '../forms/required-input-placeholder.directive';
import * as ɵngcc10 from '../i18n/c8y-translate.pipe';
export class TenantIdSetupComponent {
    constructor(client, ui, loginService, alert, translateService) {
        this.client = client;
        this.ui = ui;
        this.loginService = loginService;
        this.alert = alert;
        this.translateService = translateService;
        this.onChangeView = new EventEmitter();
        this.LOGIN_VIEWS = LoginViews;
        this.model = {
            tenant: ''
        };
    }
    /**
     * Sets up login mode for particular tenant. In case of OAI-Secure will redirect user to tenant domain.
     */
    setupLoginMode() {
        return __awaiter(this, void 0, void 0, function* () {
            this.client.tenant = this.model.tenant;
            try {
                yield this.ui.refreshLoginOptions();
                this.loginService.initLoginOptions();
                this.redirectToCorrectDomain();
            }
            catch (e) {
                if (e.res && e.res.status === 401) {
                    this.alert.danger(this.translateService.instant(gettext('Could not find tenant with ID "{{ tenantId }}".'), { tenantId: this.model.tenant }));
                }
                else {
                    this.alert.addServerFailure(e);
                }
            }
        });
    }
    /**
     * Redirects to tenant domain when login mode contains domain.
     */
    redirectToCorrectDomain() {
        const loginRedirectDomain = this.loginService.loginMode.loginRedirectDomain;
        if (loginRedirectDomain) {
            const alreadyOnCorrectDomain = window.location.href.includes(loginRedirectDomain);
            if (!alreadyOnCorrectDomain) {
                this.loginService.redirectToDomain(loginRedirectDomain);
            }
            else {
                this.onChangeView.emit({
                    view: LoginViews.Credentials,
                    loginViewParams: { showTenant: true, disableTenant: true }
                });
            }
        }
        else {
            this.onChangeView.emit({ view: LoginViews.Credentials });
        }
    }
}
TenantIdSetupComponent.ɵfac = function TenantIdSetupComponent_Factory(t) { return new (t || TenantIdSetupComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.FetchClient), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.AppStateService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.LoginService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.TranslateService)); };
TenantIdSetupComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: TenantIdSetupComponent, selectors: [["c8y-tenant-id-setup"]], outputs: { onChangeView: "onChangeView" }, decls: 13, vars: 8, consts: [["role", "form", "novalidate", "", 1, "loginForm", 3, "ngSubmit"], ["tenantIdSetupForm", "ngForm"], ["translate", "", 1, "legend", "form-block", "center"], ["id", "tenantField", 1, "tenantField"], ["for", "tenant", "translate", ""], ["type", "text", "name", "tenant", "id", "tenant", "autocapitalize", "off", "autocorrect", "off", "placeholder-no-required-hint", "", "required", "", 1, "form-control", 3, "ngModel", "placeholder", "ngModelChange"], ["tenant", "ngModel"], ["type", "submit", "translate", "", 1, "btn", "btn-primary", "btn-lg", "btn-block", "form-group", 3, "title", "disabled"]], template: function TenantIdSetupComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "form", 0, 1);
        ɵngcc0.ɵɵlistener("ngSubmit", function TenantIdSetupComponent_Template_form_ngSubmit_0_listener() { return ctx.setupLoginMode(); });
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵtext(3, " Tenant setup ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "c8y-form-group", 3);
        ɵngcc0.ɵɵelementStart(5, "label", 4);
        ɵngcc0.ɵɵtext(6, "Tenant ID");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(7, "input", 5, 6);
        ɵngcc0.ɵɵlistener("ngModelChange", function TenantIdSetupComponent_Template_input_ngModelChange_7_listener($event) { return ctx.model.tenant = $event; });
        ɵngcc0.ɵɵpipe(9, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(10, "button", 7);
        ɵngcc0.ɵɵpipe(11, "translate");
        ɵngcc0.ɵɵtext(12, " Apply ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r0 = ɵngcc0.ɵɵreference(1);
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(9, 4, "e.g."), " t12345");
        ɵngcc0.ɵɵproperty("ngModel", ctx.model.tenant);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(11, 6, "Apply"));
        ɵngcc0.ɵɵproperty("disabled", !_r0.form.valid);
    } }, directives: [ɵngcc6.ɵNgNoValidate, ɵngcc6.NgControlStatusGroup, ɵngcc6.NgForm, ɵngcc7.C8yTranslateDirective, ɵngcc8.FormGroupComponent, ɵngcc9.RequiredInputPlaceholderDirective, ɵngcc6.DefaultValueAccessor, ɵngcc6.RequiredValidator, ɵngcc6.NgControlStatus, ɵngcc6.NgModel], pipes: [ɵngcc10.C8yTranslatePipe], encapsulation: 2 });
TenantIdSetupComponent.ctorParameters = () => [
    { type: FetchClient },
    { type: AppStateService },
    { type: LoginService },
    { type: AlertService },
    { type: TranslateService }
];
TenantIdSetupComponent.propDecorators = {
    onChangeView: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TenantIdSetupComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-tenant-id-setup',
                template: "<form #tenantIdSetupForm=\"ngForm\" role=\"form\" class=\"loginForm\" (ngSubmit)=\"setupLoginMode()\" novalidate>\n  <div class=\"legend form-block center\" translate>\n    Tenant setup\n  </div>\n  <c8y-form-group class=\"tenantField\" id=\"tenantField\">\n    <label for=\"tenant\" translate>Tenant ID</label>\n    <input\n      [(ngModel)]=\"model.tenant\"\n      #tenant=\"ngModel\"\n      type=\"text\"\n      name=\"tenant\"\n      id=\"tenant\"\n      autocapitalize=\"off\"\n      autocorrect=\"off\"\n      class=\"form-control\"\n      placeholder=\"{{ 'e.g.' | translate }} t12345\"\n      placeholder-no-required-hint\n      required\n    />\n  </c8y-form-group>\n\n  <button\n    title=\"{{ 'Apply' | translate }}\"\n    type=\"submit\"\n    class=\"btn btn-primary btn-lg btn-block form-group\"\n    [disabled]=\"!tenantIdSetupForm.form.valid\"\n    translate\n  >\n    Apply\n  </button>\n</form>\n"
            }]
    }], function () { return [{ type: ɵngcc1.FetchClient }, { type: ɵngcc2.AppStateService }, { type: ɵngcc3.LoginService }, { type: ɵngcc4.AlertService }, { type: ɵngcc5.TranslateService }]; }, { onChangeView: [{
            type: Output
        }] }); })();
