// @c8y/ngx-components/esm2015/actility-device-registration/actility-device-registration-button.component.js
import { Component } from '@angular/core';
import { BsModalService } from 'ngx-bootstrap/modal';
import { ActilityDeviceRegistrationComponent } from './actility-device-registration.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/modal';
import * as ɵngcc2 from '@c8y/ngx-components';
export class ActilityDeviceRegistrationButtonComponent {
    constructor(modalService) {
        this.modalService = modalService;
    }
    open() {
        this.modalService.show(ActilityDeviceRegistrationComponent, {
            class: 'modal-sm',
            ignoreBackdropClick: true
        });
    }
}
ActilityDeviceRegistrationButtonComponent.ɵfac = function ActilityDeviceRegistrationButtonComponent_Factory(t) { return new (t || ActilityDeviceRegistrationButtonComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BsModalService)); };
ActilityDeviceRegistrationButtonComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ActilityDeviceRegistrationButtonComponent, selectors: [["c8y-actility-registration-button"]], decls: 4, vars: 3, consts: [[3, "click"], ["c8yIcon", "c8y-device-connect"]], template: function ActilityDeviceRegistrationButtonComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "button", 0);
        ɵngcc0.ɵɵlistener("click", function ActilityDeviceRegistrationButtonComponent_Template_button_click_0_listener() { return ctx.open(); });
        ɵngcc0.ɵɵelement(1, "i", 1);
        ɵngcc0.ɵɵtext(2);
        ɵngcc0.ɵɵpipe(3, "translate");
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 1, "Actility LoRa"), "\n");
    } }, directives: [ɵngcc2.IconDirective], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
ActilityDeviceRegistrationButtonComponent.ctorParameters = () => [
    { type: BsModalService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ActilityDeviceRegistrationButtonComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-actility-registration-button',
                template: "<button (click)=\"open()\">\n  <i c8yIcon=\"c8y-device-connect\"></i>\n  {{ 'Actility LoRa' | translate }}\n</button>\n"
            }]
    }], function () { return [{ type: ɵngcc1.BsModalService }]; }, null); })();
