// @c8y/ngx-components/esm2015/actility-device-registration/actility-device-registration.module.js
import { NgModule } from '@angular/core';
import { CommonModule, CoreModule } from '@c8y/ngx-components';
import { HOOK_DEVICE_REGISTRATION } from '@c8y/ngx-components/register-device';
import { ActilityDeviceRegistrationButtonComponent } from './actility-device-registration-button.component';
import { ActilityDeviceRegistrationFactory } from './actility-device-registration.factory';
import { ActilityDeviceRegistrationComponent } from './actility-device-registration.component';
import { ActilityDeviceRegistrationService } from './actility-device-registration.service';
import * as ɵngcc0 from '@angular/core';
export class ActilityDeviceRegistrationModule {
}
ActilityDeviceRegistrationModule.ɵfac = function ActilityDeviceRegistrationModule_Factory(t) { return new (t || ActilityDeviceRegistrationModule)(); };
ActilityDeviceRegistrationModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: ActilityDeviceRegistrationModule });
ActilityDeviceRegistrationModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        ActilityDeviceRegistrationService,
        {
            provide: HOOK_DEVICE_REGISTRATION,
            useClass: ActilityDeviceRegistrationFactory,
            multi: true
        }
    ], imports: [[CoreModule, CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ActilityDeviceRegistrationModule, [{
        type: NgModule,
        args: [{
                imports: [CoreModule, CommonModule],
                declarations: [ActilityDeviceRegistrationButtonComponent, ActilityDeviceRegistrationComponent],
                providers: [
                    ActilityDeviceRegistrationService,
                    {
                        provide: HOOK_DEVICE_REGISTRATION,
                        useClass: ActilityDeviceRegistrationFactory,
                        multi: true
                    }
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ActilityDeviceRegistrationModule, { declarations: function () { return [ActilityDeviceRegistrationButtonComponent, ActilityDeviceRegistrationComponent]; }, imports: function () { return [CoreModule, CommonModule]; } }); })();
