// @c8y/ngx-components/esm2015/actility-device-registration/actility-device-registration.factory.js
import { Injectable } from '@angular/core';
import { TenantUiService } from '@c8y/ngx-components';
import { ActilityDeviceRegistrationButtonComponent } from './actility-device-registration-button.component';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
export class ActilityDeviceRegistrationFactory {
    constructor(tenantService) {
        this.tenantService = tenantService;
    }
    get() {
        const actions = [];
        if (this.tenantService.isMicroserviceSubscribedInCurrentTenant('actility')) {
            actions.push({
                template: ActilityDeviceRegistrationButtonComponent,
                priority: 99,
                category: 'single'
            });
        }
        return actions;
    }
}
ActilityDeviceRegistrationFactory.ɵfac = function ActilityDeviceRegistrationFactory_Factory(t) { return new (t || ActilityDeviceRegistrationFactory)(ɵngcc0.ɵɵinject(ɵngcc1.TenantUiService)); };
ActilityDeviceRegistrationFactory.ɵprov = i0.ɵɵdefineInjectable({ factory: function ActilityDeviceRegistrationFactory_Factory() { return new ActilityDeviceRegistrationFactory(i0.ɵɵinject(i1.TenantUiService)); }, token: ActilityDeviceRegistrationFactory, providedIn: "root" });
ActilityDeviceRegistrationFactory.ctorParameters = () => [
    { type: TenantUiService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ActilityDeviceRegistrationFactory, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: ɵngcc1.TenantUiService }]; }, null); })();
