// @c8y/ngx-components/esm2015/repository/firmware/list/firmware-repository-list.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FirmwareListComponent } from './firmware-list.component';
import { FirmwareDetailsComponent } from './firmware-details.component';
import { AddFirmwareModalComponent } from './add-firmware-modal.component';
import { AddFirmwarePatchModalComponent } from './add-firmware-patch-modal.component';
import { CoreModule, FormsModule, HOOK_NAVIGATOR_NODES, HOOK_ROUTE } from '@c8y/ngx-components';
import { FirmwareRepositoryNavigationFactory } from './firmware-repository-navigation-factory';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { SharedRepositoryModule } from '@c8y/ngx-components/repository/shared';
export class FirmwareRepositoryListModule {
    static forRoot() {
        return {
            ngModule: FirmwareRepositoryListModule,
            providers: [
                {
                    provide: HOOK_NAVIGATOR_NODES,
                    useClass: FirmwareRepositoryNavigationFactory,
                    multi: true
                },
                {
                    provide: HOOK_ROUTE,
                    useValue: [
                        {
                            path: 'firmware',
                            component: FirmwareListComponent
                        },
                        {
                            path: 'firmware/:id',
                            component: FirmwareDetailsComponent
                        }
                    ],
                    multi: true
                }
            ]
        };
    }
}
FirmwareRepositoryListModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    CoreModule,
                    FormsModule,
                    PopoverModule,
                    SharedRepositoryModule
                ],
                declarations: [
                    FirmwareListComponent,
                    FirmwareDetailsComponent,
                    AddFirmwareModalComponent,
                    AddFirmwarePatchModalComponent,
                ],
                entryComponents: [
                    FirmwareListComponent,
                    FirmwareDetailsComponent,
                    AddFirmwareModalComponent,
                    AddFirmwarePatchModalComponent,
                ]
            },] }
];
