// @c8y/ngx-components/esm2015/operations/bulk-single-operations-list/bulk-single-operations-list.module.js
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { CoreModule } from '@c8y/ngx-components';
import { StatusFilterModule } from '@c8y/ngx-components/operations/status-filter';
import { SingleOperationItemComponent } from './single-operation-item.component';
import { SingleOperationsListComponent } from './single-operations-list.component';
/**
 * This module allows for displaying the list of single operations in a bulk operations details.
 */
import * as ɵngcc0 from '@angular/core';
export class BulkSingleOperationsListModule {
}
BulkSingleOperationsListModule.ɵfac = function BulkSingleOperationsListModule_Factory(t) { return new (t || BulkSingleOperationsListModule)(); };
BulkSingleOperationsListModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: BulkSingleOperationsListModule });
BulkSingleOperationsListModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CoreModule, PopoverModule, RouterModule, TooltipModule, StatusFilterModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BulkSingleOperationsListModule, [{
        type: NgModule,
        args: [{
                imports: [CoreModule, PopoverModule, RouterModule, TooltipModule, StatusFilterModule],
                declarations: [SingleOperationsListComponent, SingleOperationItemComponent],
                entryComponents: [SingleOperationsListComponent, SingleOperationItemComponent],
                exports: [SingleOperationsListComponent, SingleOperationItemComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(BulkSingleOperationsListModule, { declarations: function () { return [SingleOperationsListComponent, SingleOperationItemComponent]; }, imports: function () { return [CoreModule, PopoverModule, RouterModule, TooltipModule, StatusFilterModule]; }, exports: function () { return [SingleOperationsListComponent, SingleOperationItemComponent]; } }); })();
