// @c8y/ngx-components/esm2015/operations/bulk-single-operations-list/single-operation-item.component.js
import { __awaiter } from "tslib";
import { Component, Input, ViewChild } from '@angular/core';
import { InventoryService, OperationService } from '@c8y/client';
import { AlertService, gettext } from '@c8y/ngx-components';
import { OPERATION_STATUS_LABELS, OPERATION_STATUS_OPTIONS_MAP } from '@c8y/ngx-components/operations/shared';
import { omit } from 'lodash-es';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from '@c8y/client';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from 'ngx-bootstrap/tooltip';
import * as ɵngcc5 from '@angular/router';
import * as ɵngcc6 from 'ngx-bootstrap/popover';

const _c0 = ["failureReason"];
function SingleOperationItemComponent_a_21_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "a", 16);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 17);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("popover", ɵngcc0.ɵɵpipeBind1(1, 2, ctx_r1.operation.failureReason));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", "info-circle text-primary");
} }
function SingleOperationItemComponent_div_22_Template(rf, ctx) { if (rf & 1) {
    const _r4 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 18);
    ɵngcc0.ɵɵelementStart(1, "a", 19);
    ɵngcc0.ɵɵlistener("click", function SingleOperationItemComponent_div_22_Template_a_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r4); const ctx_r3 = ɵngcc0.ɵɵnextContext(); return ctx_r3.retrySingleOperation(ctx_r3.operation); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 17);
    ɵngcc0.ɵɵelementStart(4, "span", 20);
    ɵngcc0.ɵɵtext(5, "Retry operation");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 2, "Retry operation"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", "repeat m-r-4");
} }
const _c1 = function (a0, a1) { return { "content-flex-80": a0, "content-flex-50": a1 }; };
const _c2 = function (a1) { return ["/device", a1, "device-info"]; };
export class SingleOperationItemComponent {
    constructor(alertService, inventoryService, operationService) {
        this.alertService = alertService;
        this.inventoryService = inventoryService;
        this.operationService = operationService;
        this.device = {};
        this.OPERATION_STATUS_LABELS = OPERATION_STATUS_LABELS;
        this.OPERATION_STATUS_OPTIONS_MAP = OPERATION_STATUS_OPTIONS_MAP;
        this.readOnly = false;
    }
    ngOnInit() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            this.device = (_a = (yield this.inventoryService.detail(this.operation.deviceId))) === null || _a === void 0 ? void 0 : _a.data;
        });
    }
    retrySingleOperation(singleOperation) {
        return __awaiter(this, void 0, void 0, function* () {
            const operationToCreate = omit(singleOperation, [
                'id',
                'self',
                'status',
                'failureReason',
                'creationTime',
                'delivery'
            ]);
            try {
                yield this.operationService.create(operationToCreate);
                this.alertService.success(gettext('Operation created.'));
            }
            catch (er) {
                this.alertService.addServerFailure(er);
            }
        });
    }
}
SingleOperationItemComponent.ɵfac = function SingleOperationItemComponent_Factory(t) { return new (t || SingleOperationItemComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.InventoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.OperationService)); };
SingleOperationItemComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SingleOperationItemComponent, selectors: [["c8y-single-operation-item"]], viewQuery: function SingleOperationItemComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 7);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.failureReasonRef = _t.first);
    } }, inputs: { readOnly: "readOnly", operation: "operation" }, decls: 23, vars: 22, consts: [[1, "c8y-list__item--double-actions"], [3, "c8yIcon", "ngClass", "tooltip"], [3, "ngClass"], [1, "col-4"], [1, "m-r-4"], [3, "mo"], ["routerLinkActive", "active", 3, "routerLink"], [1, "col-3"], [1, "icon-flex", "text-muted"], ["c8yIcon", "calendar", 1, "m-r-4"], [1, "col-5"], [1, "flex-row"], [1, "text-truncate"], ["failureReason", ""], ["class", "m-l-8", "placement", "left", "outsideClick", "true", 3, "popover", 4, "ngIf"], ["class", "showOnHover p-l-8 p-r-8 fit-h-20 flex-item-right", 4, "ngIf"], ["placement", "left", "outsideClick", "true", 1, "m-l-8", 3, "popover"], [3, "c8yIcon"], [1, "showOnHover", "p-l-8", "p-r-8", "fit-h-20", "flex-item-right"], [1, "btn", "btn-default", "btn-xs", 3, "title", "click"], ["translate", ""]], template: function SingleOperationItemComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-li", 0);
        ɵngcc0.ɵɵelementStart(1, "c8y-li-icon");
        ɵngcc0.ɵɵelement(2, "i", 1);
        ɵngcc0.ɵɵpipe(3, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "div", 2);
        ɵngcc0.ɵɵelementStart(5, "div", 3);
        ɵngcc0.ɵɵelementStart(6, "span", 4);
        ɵngcc0.ɵɵelement(7, "device-status", 5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(8, "a", 6);
        ɵngcc0.ɵɵtext(9);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(10, "div", 7);
        ɵngcc0.ɵɵelementStart(11, "small", 8);
        ɵngcc0.ɵɵelement(12, "i", 9);
        ɵngcc0.ɵɵtext(13);
        ɵngcc0.ɵɵpipe(14, "c8yDate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(15, "div", 10);
        ɵngcc0.ɵɵelementStart(16, "div", 11);
        ɵngcc0.ɵɵelementStart(17, "span", 12, 13);
        ɵngcc0.ɵɵtext(19);
        ɵngcc0.ɵɵpipe(20, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(21, SingleOperationItemComponent_a_21_Template, 3, 4, "a", 14);
        ɵngcc0.ɵɵtemplate(22, SingleOperationItemComponent_div_22_Template, 6, 4, "div", 15);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r0 = ɵngcc0.ɵɵreference(18);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yIcon", ctx.OPERATION_STATUS_OPTIONS_MAP[ctx.operation.status].icon)("ngClass", ctx.OPERATION_STATUS_OPTIONS_MAP[ctx.operation.status].styleClass)("tooltip", ɵngcc0.ɵɵpipeBind1(3, 11, ctx.operation.status.toString()));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(17, _c1, !ctx.readOnly, ctx.readOnly));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("mo", ctx.device);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("routerLink", ɵngcc0.ɵɵpureFunction1(20, _c2, ctx.operation.deviceId));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ctx.operation.deviceName, " ");
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(14, 13, ctx.operation.creationTime), " ");
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(20, 15, ctx.operation.failureReason), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", _r0.offsetWidth < _r0.scrollWidth);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.readOnly && ctx.operation.status.toString() == ctx.OPERATION_STATUS_OPTIONS_MAP.FAILED.status.toString());
    } }, directives: [ɵngcc1.ListItemComponent, ɵngcc1.ListItemIconComponent, ɵngcc1.IconDirective, ɵngcc3.NgClass, ɵngcc4.TooltipDirective, ɵngcc1.DeviceStatusComponent, ɵngcc5.RouterLinkWithHref, ɵngcc5.RouterLinkActive, ɵngcc3.NgIf, ɵngcc6.PopoverDirective, ɵngcc1.C8yTranslateDirective], pipes: [ɵngcc1.C8yTranslatePipe, ɵngcc1.DatePipe], encapsulation: 2 });
SingleOperationItemComponent.ctorParameters = () => [
    { type: AlertService },
    { type: InventoryService },
    { type: OperationService }
];
SingleOperationItemComponent.propDecorators = {
    operation: [{ type: Input }],
    readOnly: [{ type: Input }],
    failureReasonRef: [{ type: ViewChild, args: ['failureReason', { static: true },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SingleOperationItemComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-single-operation-item',
                template: "<c8y-li class=\"c8y-list__item--double-actions\">\n  <c8y-li-icon>\n    <i\n      [c8yIcon]=\"OPERATION_STATUS_OPTIONS_MAP[operation.status].icon\"\n      [ngClass]=\"OPERATION_STATUS_OPTIONS_MAP[operation.status].styleClass\"\n      [tooltip]=\"operation.status.toString() | translate\"\n    ></i>\n  </c8y-li-icon>\n  <div [ngClass]=\"{ 'content-flex-80': !readOnly, 'content-flex-50': readOnly }\">\n    <div class=\"col-4\">\n      <span class=\"m-r-4\">\n        <device-status [mo]=\"device\"></device-status>\n      </span>\n      <a [routerLink]=\"['/device', operation.deviceId, 'device-info']\" routerLinkActive=\"active\">\n        {{ operation.deviceName }}\n      </a>\n    </div>\n    <div class=\"col-3\">\n      <small class=\"icon-flex text-muted\">\n        <i c8yIcon=\"calendar\" class=\"m-r-4\"></i>\n        {{ operation.creationTime | c8yDate }}\n      </small>\n    </div>\n    <div class=\"col-5\">\n      <div class=\"flex-row\">\n        <span class=\"text-truncate\" #failureReason>\n          {{ operation.failureReason | translate }}\n        </span>\n        <a\n          *ngIf=\"failureReason.offsetWidth < failureReason.scrollWidth\"\n          class=\"m-l-8\"\n          popover=\"{{ operation.failureReason | translate }}\"\n          placement=\"left\"\n          outsideClick=\"true\"\n        >\n          <i [c8yIcon]=\"'info-circle text-primary'\"></i>\n        </a>\n        <div\n          class=\"showOnHover p-l-8 p-r-8 fit-h-20 flex-item-right\"\n          *ngIf=\"\n            !readOnly &&\n            operation.status.toString() == OPERATION_STATUS_OPTIONS_MAP.FAILED.status.toString()\n          \"\n        >\n          <a\n            class=\"btn btn-default btn-xs\"\n            (click)=\"retrySingleOperation(operation)\"\n            title=\"{{ 'Retry operation' | translate }}\"\n          >\n            <i [c8yIcon]=\"'repeat m-r-4'\"></i>\n            <span translate>Retry operation</span>\n          </a>\n        </div>\n      </div>\n    </div>\n  </div>\n</c8y-li>\n"
            }]
    }], function () { return [{ type: ɵngcc1.AlertService }, { type: ɵngcc2.InventoryService }, { type: ɵngcc2.OperationService }]; }, { readOnly: [{
            type: Input
        }], operation: [{
            type: Input
        }], failureReasonRef: [{
            type: ViewChild,
            args: ['failureReason', { static: true }]
        }] }); })();
