// @c8y/ngx-components/esm2015/core/dynamic-forms/file/file.type.component.js
import { ChangeDetectionStrategy, Component, ViewChild } from '@angular/core';
import { FieldType } from '@ngx-formly/core';
import { get } from 'lodash';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '../../drop-area/drop-area.component';
import * as ɵngcc3 from '@angular/forms';
import * as ɵngcc4 from 'ngx-bootstrap/popover';
import * as ɵngcc5 from '../../common/icon.directive';
import * as ɵngcc6 from '../../common/humanize.pipe';

const _c0 = ["dropArea"];
function FieldFile_label_1_button_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "button", 6);
    ɵngcc0.ɵɵelement(1, "i", 7);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("popover", ctx_r2.to.description)("outsideClick", true);
} }
function FieldFile_label_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "label", 4);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "humanize");
    ɵngcc0.ɵɵtemplate(3, FieldFile_label_1_button_3_Template, 2, 2, "button", 5);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("id", ctx_r0.id + "-fieldset");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 3, ctx_r0.to.label), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.to.description);
} }
export class FieldFile extends FieldType {
    constructor() {
        super(...arguments);
        this.dropAreaInputs = [
            'title',
            'message',
            'icon',
            'loadingMessage',
            'alwaysShow',
            'clickToOpen',
            'loading',
            'progress',
            'maxAllowedFiles',
            'files',
            'accept'
        ];
    }
    ngOnInit() {
        const keys = Array.isArray(this.key) ? this.key : [this.key];
        const path = keys.join('.');
        this.dropArea.files = get(this.model, path);
        this.dropAreaInputs
            .filter(input => this.to[input] !== undefined)
            .forEach(input => (this.dropArea[input] = this.to[input]));
    }
}
FieldFile.ɵfac = /*@__PURE__*/ function () { let ɵFieldFile_BaseFactory; return function FieldFile_Factory(t) { return (ɵFieldFile_BaseFactory || (ɵFieldFile_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(FieldFile)))(t || FieldFile); }; }();
FieldFile.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: FieldFile, selectors: [["c8y-field-file"]], viewQuery: function FieldFile_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 7);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.dropArea = _t.first);
    } }, features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 4, vars: 3, consts: [["role", "group", 1, "form-group"], [3, "id", 4, "ngIf"], [1, "drop-area-sm", 3, "formControl"], ["dropArea", ""], [3, "id"], ["class", "btn-clean text-14 m-l-4", "placement", "right", "type", "button", 3, "popover", "outsideClick", 4, "ngIf"], ["placement", "right", "type", "button", 1, "btn-clean", "text-14", "m-l-4", 3, "popover", "outsideClick"], ["c8yIcon", "question-circle-o", 1, "text-info"]], template: function FieldFile_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, FieldFile_label_1_Template, 4, 5, "label", 1);
        ɵngcc0.ɵɵelement(2, "c8y-drop-area", 2, 3);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵattribute("aria-labelledby", ctx.id + "-fieldset");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.to.label || ctx.to.description);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("formControl", ctx.formControl);
    } }, directives: [ɵngcc1.NgIf, ɵngcc2.DropAreaComponent, ɵngcc3.NgControlStatus, ɵngcc3.FormControlDirective, ɵngcc4.PopoverDirective, ɵngcc5.IconDirective], pipes: [ɵngcc6.HumanizePipe], encapsulation: 2, changeDetection: 0 });
FieldFile.CONFIG = {
    types: [
        {
            name: 'file',
            component: FieldFile
        }
    ]
};
FieldFile.propDecorators = {
    dropArea: [{ type: ViewChild, args: ['dropArea', { static: true },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(FieldFile, [{
        type: Component,
        args: [{
                selector: 'c8y-field-file',
                template: "<div class=\"form-group\" role=\"group\" [attr.aria-labelledby]=\"id + '-fieldset'\">\n  <label id=\"{{ id + '-fieldset' }}\" *ngIf=\"to.label || to.description\">\n    {{ to.label | humanize }}\n    <button\n      *ngIf=\"to.description\"\n      class=\"btn-clean text-14 m-l-4\"\n      [popover]=\"to.description\"\n      placement=\"right\"\n      [outsideClick]=\"true\"\n      type=\"button\"\n    >\n      <i c8yIcon=\"question-circle-o\" class=\"text-info\"></i>\n    </button>\n  </label>\n\n  <c8y-drop-area #dropArea class=\"drop-area-sm\" [formControl]=\"formControl\"> </c8y-drop-area>\n</div>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], null, { dropArea: [{
            type: ViewChild,
            args: ['dropArea', { static: true }]
        }] }); })();
