// @c8y/ngx-components/esm2015/core/setup/setup.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '../common/common.module';
import { HeaderModule } from '../header/header.module';
import { gettext } from '../i18n/gettext';
import { PluginsModule } from '../plugins/plugins.module';
import { PropertiesListModule } from '../properties-list/properties-list.module';
import { Steppers } from '../stepper/stepper.model';
import { StepperModule } from '../stepper/stepper.module';
import { HOOK_STEPPER } from '../stepper/stepper.service';
import { SetupCompletedComponent } from './setup-completed.component';
import { SetupComponent } from './setup.component';
import { SETUP_FINISHED_STEP_ID } from './setup.model';
import * as ɵngcc0 from '@angular/core';
const ɵ0 = {
    setupId: SETUP_FINISHED_STEP_ID,
    stepperId: Steppers.SETUP,
    component: SetupCompletedComponent,
    label: gettext('Summary'),
    priority: Infinity
};
export class SetupModule {
}
SetupModule.ɵfac = function SetupModule_Factory(t) { return new (t || SetupModule)(); };
SetupModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: SetupModule });
SetupModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        {
            provide: HOOK_STEPPER,
            useValue: ɵ0,
            multi: true
        }
    ], imports: [[CommonModule, HeaderModule, StepperModule, PluginsModule, PropertiesListModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SetupModule, [{
        type: NgModule,
        args: [{
                declarations: [SetupComponent, SetupCompletedComponent],
                imports: [CommonModule, HeaderModule, StepperModule, PluginsModule, PropertiesListModule],
                exports: [SetupComponent, SetupCompletedComponent],
                entryComponents: [SetupCompletedComponent],
                providers: [
                    {
                        provide: HOOK_STEPPER,
                        useValue: ɵ0,
                        multi: true
                    }
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(SetupModule, { declarations: function () { return [SetupComponent, SetupCompletedComponent]; }, imports: function () { return [CommonModule, HeaderModule, StepperModule, PluginsModule, PropertiesListModule]; }, exports: function () { return [SetupComponent, SetupCompletedComponent]; } }); })();
export { ɵ0 };
