// @c8y/ngx-components/esm2015/core/setup/setup-completed.component.js
import { CdkStep } from '@angular/cdk/stepper';
import { Component } from '@angular/core';
import { C8yStepper } from '../stepper/stepper.component';
import { SetupComponent } from './setup.component';
import { gettext } from '../i18n/gettext';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../stepper/stepper.component';
import * as ɵngcc2 from '@angular/cdk/stepper';
import * as ɵngcc3 from './setup.component';
import * as ɵngcc4 from '../i18n/c8y-translate.directive';
import * as ɵngcc5 from '@angular/common';
import * as ɵngcc6 from '../common/operation-result.component';
import * as ɵngcc7 from '../common/icon.directive';
import * as ɵngcc8 from '../i18n/c8y-translate.pipe';

function SetupCompletedComponent_c8y_operation_result_12_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-operation-result", 19);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("text", ɵngcc0.ɵɵpipeBind1(1, 3, "Setup completed"));
    ɵngcc0.ɵɵproperty("size", 120)("vertical", true);
} }
function SetupCompletedComponent_c8y_operation_result_13_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-operation-result", 20);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("text", ɵngcc0.ɵɵpipeBind1(1, 3, "Setup completed with warnings"));
    ɵngcc0.ɵɵproperty("size", 120)("vertical", true);
} }
function SetupCompletedComponent_c8y_operation_result_14_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-operation-result", 21);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("text", ɵngcc0.ɵɵpipeBind1(1, 3, "Setup not completed"));
    ɵngcc0.ɵɵproperty("size", 120)("vertical", true);
} }
const _c0 = function (a0, a1, a2) { return { "dot-success": a0, "dot-error": a1, "dot-warning": a2 }; };
function SetupCompletedComponent_li_16_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵelementStart(1, "span", 28);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const step_r8 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r9 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction3(5, _c0, step_r8.completed, !step_r8.completed && step_r8.required, !step_r8.completed && !step_r8.required));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", step_r8.index + 1, " ");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", step_r8.label || ɵngcc0.ɵɵpipeBind1(4, 3, ctx_r9.STEP_LABEL) + " ", " ");
} }
function SetupCompletedComponent_li_16_span_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵelementStart(1, "span", 29);
    ɵngcc0.ɵɵtext(2, "Step");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const step_r8 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1("", " " + (step_r8.index + 1), " ");
} }
function SetupCompletedComponent_li_16_i_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 30);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("c8yIcon", "success")("title", ɵngcc0.ɵɵpipeBind1(1, 2, "Completed"));
} }
function SetupCompletedComponent_li_16_i_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 31);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("c8yIcon", "error")("title", ɵngcc0.ɵɵpipeBind1(1, 2, "Not completed and required. Please go to the step and fulfil the setup."));
} }
function SetupCompletedComponent_li_16_i_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 32);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("c8yIcon", "warning")("title", ɵngcc0.ɵɵpipeBind1(1, 2, "This step was not completed but it is also not required."));
} }
function SetupCompletedComponent_li_16_Template(rf, ctx) { if (rf & 1) {
    const _r17 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 22);
    ɵngcc0.ɵɵlistener("click", function SetupCompletedComponent_li_16_Template_li_click_0_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r17); const step_r8 = restoredCtx.$implicit; const ctx_r16 = ɵngcc0.ɵɵnextContext(); return ctx_r16.gotoStep(step_r8.index); });
    ɵngcc0.ɵɵtemplate(1, SetupCompletedComponent_li_16_span_1_Template, 5, 9, "span", 23);
    ɵngcc0.ɵɵtemplate(2, SetupCompletedComponent_li_16_span_2_Template, 4, 1, "span", 23);
    ɵngcc0.ɵɵelementStart(3, "span", 24);
    ɵngcc0.ɵɵtemplate(4, SetupCompletedComponent_li_16_i_4_Template, 2, 4, "i", 25);
    ɵngcc0.ɵɵtemplate(5, SetupCompletedComponent_li_16_i_5_Template, 2, 4, "i", 26);
    ɵngcc0.ɵɵtemplate(6, SetupCompletedComponent_li_16_i_6_Template, 2, 4, "i", 27);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const step_r8 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", step_r8.label);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !step_r8.label);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", step_r8.completed);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !step_r8.completed && step_r8.required);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !step_r8.completed && !step_r8.required);
} }
function SetupCompletedComponent_button_18_Template(rf, ctx) { if (rf & 1) {
    const _r19 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 33);
    ɵngcc0.ɵɵlistener("click", function SetupCompletedComponent_button_18_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r19); const ctx_r18 = ɵngcc0.ɵɵnextContext(); return ctx_r18.back(); });
    ɵngcc0.ɵɵtext(1, " Back ");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("disabled", ctx_r4.setup.currentSetupState === ctx_r4.setup.setupState.FINALIZING || ctx_r4.setup.currentSetupState === ctx_r4.setup.setupState.RELOADING);
} }
function SetupCompletedComponent_span_20_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 29);
    ɵngcc0.ɵɵtext(1, "Save and reload");
    ɵngcc0.ɵɵelementEnd();
} }
function SetupCompletedComponent_span_21_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 29);
    ɵngcc0.ɵɵtext(1, "Getting things ready\u2026");
    ɵngcc0.ɵɵelementEnd();
} }
function SetupCompletedComponent_span_22_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 29);
    ɵngcc0.ɵɵtext(1, "Done. Reloading the application.");
    ɵngcc0.ɵɵelementEnd();
} }
const _c1 = function (a0) { return { "btn-pending": a0 }; };
export class SetupCompletedComponent {
    constructor(stepper, step, setup) {
        this.stepper = stepper;
        this.step = step;
        this.setup = setup;
        this.status = 'done';
        this.STEP_LABEL = gettext('Step');
        this.isVerified = false;
    }
    ngOnInit() {
        this.stepperSelectionChangeSubscription = this.stepper.selectionChange.subscribe(({ selectedIndex }) => {
            this.verify(selectedIndex);
        });
    }
    ngOnDestroy() {
        this.stepperSelectionChangeSubscription.unsubscribe();
    }
    verify(selectedIndex) {
        const isLastStep = this.stepper.steps.length - 1 === selectedIndex;
        if (isLastStep || this.isVerified) {
            // set the last step to completed
            this.getOverallStatus();
            this.setup.verify();
            this.isVerified = true;
        }
    }
    start() {
        this.setup.finish();
    }
    back() {
        this.stepper.previous();
    }
    gotoStep(index) {
        this.stepper.selectedIndex = index;
    }
    getOverallStatus() {
        const validSteps = this.setup.steps.slice(0, this.setup.steps.length - 1);
        this.status = 'done';
        if (validSteps.find(step => !step.completed && step.required)) {
            this.status = 'error';
        }
        else if (validSteps.find(step => !step.completed && !step.required)) {
            this.status = 'warning';
        }
        this.step.state = this.status;
    }
}
SetupCompletedComponent.ɵfac = function SetupCompletedComponent_Factory(t) { return new (t || SetupCompletedComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.C8yStepper), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.CdkStep), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.SetupComponent)); };
SetupCompletedComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SetupCompletedComponent, selectors: [["c8y-setup-completed"]], hostAttrs: [1, "d-contents"], decls: 23, vars: 13, consts: [[1, "container-fluid", "flex-no-shrink", "fit-w"], [1, "row", "separator-bottom"], [1, "col-md-8", "col-md-offset-2", "col-lg-6", "col-lg-offset-3", "p-t-24", "p-l-16", "p-r-16"], ["translate", "", 1, "text-medium", "l-h-base"], ["translate", "", 1, "lead", "text-normal", "p-t-8", "p-b-8"], [1, "inner-scroll", "flex-grow"], [1, "container-fluid", "fit-w", "p-t-16"], [1, "row"], [1, "col-md-8", "col-md-offset-2", "col-lg-6", "col-lg-offset-3"], [1, "text-center", 3, "ngSwitch"], ["type", "success", 3, "text", "size", "vertical", 4, "ngSwitchCase"], ["type", "warning", 3, "text", "size", "vertical", 4, "ngSwitchCase"], ["type", "error", 3, "text", "size", "vertical", 4, "ngSwitchCase"], [1, "list-unstyled", "m-t-16", "separator-top"], ["class", "p-t-4 p-b-4 d-flex a-i-center separator-bottom text-nowrap l-h-1", "style", "cursor: pointer", 3, "click", 4, "ngFor", "ngForOf"], [1, "card-footer", "separator", "d-flex", "j-c-center"], ["translate", "", "class", "btn btn-default", "type", "button", 3, "disabled", "click", 4, "ngIf"], ["type", "button", 1, "btn", "btn-primary", 3, "disabled", "ngClass", "click"], ["translate", "", 4, "ngIf"], ["type", "success", 3, "text", "size", "vertical"], ["type", "warning", 3, "text", "size", "vertical"], ["type", "error", 3, "text", "size", "vertical"], [1, "p-t-4", "p-b-4", "d-flex", "a-i-center", "separator-bottom", "text-nowrap", "l-h-1", 2, "cursor", "pointer", 3, "click"], [4, "ngIf"], [1, "m-l-auto", "icon-20", "l-h-1"], ["class", "text-success", 3, "c8yIcon", "title", 4, "ngIf"], ["class", "text-danger", 3, "c8yIcon", "title", 4, "ngIf"], ["class", "text-warning", 3, "c8yIcon", "title", 4, "ngIf"], [1, "dot", "small", "m-r-4", "l-h-1", 3, "ngClass"], ["translate", ""], [1, "text-success", 3, "c8yIcon", "title"], [1, "text-danger", 3, "c8yIcon", "title"], [1, "text-warning", 3, "c8yIcon", "title"], ["translate", "", "type", "button", 1, "btn", "btn-default", 3, "disabled", "click"]], template: function SetupCompletedComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelementStart(3, "h3", 3);
        ɵngcc0.ɵɵtext(4, "Summary");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "p", 4);
        ɵngcc0.ɵɵtext(6, " Please verify the setup result. If any step is not finished but required, you need to ensure that you have finished the step. ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(7, "div", 5);
        ɵngcc0.ɵɵelementStart(8, "div", 6);
        ɵngcc0.ɵɵelementStart(9, "div", 7);
        ɵngcc0.ɵɵelementStart(10, "div", 8);
        ɵngcc0.ɵɵelementStart(11, "div", 9);
        ɵngcc0.ɵɵtemplate(12, SetupCompletedComponent_c8y_operation_result_12_Template, 2, 5, "c8y-operation-result", 10);
        ɵngcc0.ɵɵtemplate(13, SetupCompletedComponent_c8y_operation_result_13_Template, 2, 5, "c8y-operation-result", 11);
        ɵngcc0.ɵɵtemplate(14, SetupCompletedComponent_c8y_operation_result_14_Template, 2, 5, "c8y-operation-result", 12);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(15, "ul", 13);
        ɵngcc0.ɵɵtemplate(16, SetupCompletedComponent_li_16_Template, 7, 5, "li", 14);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(17, "div", 15);
        ɵngcc0.ɵɵtemplate(18, SetupCompletedComponent_button_18_Template, 2, 1, "button", 16);
        ɵngcc0.ɵɵelementStart(19, "button", 17);
        ɵngcc0.ɵɵlistener("click", function SetupCompletedComponent_Template_button_click_19_listener() { return ctx.start(); });
        ɵngcc0.ɵɵtemplate(20, SetupCompletedComponent_span_20_Template, 2, 0, "span", 18);
        ɵngcc0.ɵɵtemplate(21, SetupCompletedComponent_span_21_Template, 2, 0, "span", 18);
        ɵngcc0.ɵɵtemplate(22, SetupCompletedComponent_span_22_Template, 2, 0, "span", 18);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(11);
        ɵngcc0.ɵɵproperty("ngSwitch", ctx.status);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngSwitchCase", "done");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngSwitchCase", "warning");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngSwitchCase", "error");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.setup.steps.slice(0, ctx.setup.steps.length - 1));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.stepper.selectedIndex !== 0);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("disabled", ctx.status === "error" || ctx.setup.currentSetupState === ctx.setup.setupState.FINALIZING || ctx.setup.currentSetupState === ctx.setup.setupState.RELOADING)("ngClass", ɵngcc0.ɵɵpureFunction1(11, _c1, ctx.setup.currentSetupState === ctx.setup.setupState.FINALIZING || ctx.setup.currentSetupState === ctx.setup.setupState.RELOADING));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.setup.currentSetupState === ctx.setup.setupState.WIZARD);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.setup.currentSetupState === ctx.setup.setupState.FINALIZING);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.setup.currentSetupState === ctx.setup.setupState.RELOADING);
    } }, directives: [ɵngcc4.C8yTranslateDirective, ɵngcc5.NgSwitch, ɵngcc5.NgSwitchCase, ɵngcc5.NgForOf, ɵngcc5.NgIf, ɵngcc5.NgClass, ɵngcc6.OperationResultComponent, ɵngcc7.IconDirective], pipes: [ɵngcc8.C8yTranslatePipe], encapsulation: 2 });
SetupCompletedComponent.ctorParameters = () => [
    { type: C8yStepper },
    { type: CdkStep },
    { type: SetupComponent }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SetupCompletedComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-setup-completed',
                template: "<div class=\"container-fluid flex-no-shrink fit-w\">\n  <div class=\"row separator-bottom\">\n    <div class=\"col-md-8 col-md-offset-2 col-lg-6 col-lg-offset-3  p-t-24 p-l-16 p-r-16\">\n      <h3 translate class=\"text-medium l-h-base\">Summary</h3>\n      <p translate class=\"lead text-normal p-t-8 p-b-8\">\n        Please verify the setup result. If any step is not finished but required, you need to ensure\n        that you have finished the step.\n      </p>\n    </div>\n  </div>\n</div>\n<div class=\"inner-scroll flex-grow\">\n  <div class=\"container-fluid fit-w p-t-16\">\n    <div class=\"row\">\n      <div class=\"col-md-8 col-md-offset-2 col-lg-6 col-lg-offset-3\">\n        <div class=\"text-center\" [ngSwitch]=\"status\">\n          <c8y-operation-result\n            *ngSwitchCase=\"'done'\"\n            text=\"{{ 'Setup completed' | translate }}\"\n            [size]=\"120\"\n            [vertical]=\"true\"\n            type=\"success\"\n          ></c8y-operation-result>\n          <c8y-operation-result\n            *ngSwitchCase=\"'warning'\"\n            text=\"{{ 'Setup completed with warnings' | translate }}\"\n            [size]=\"120\"\n            [vertical]=\"true\"\n            type=\"warning\"\n          ></c8y-operation-result>\n          <c8y-operation-result\n            *ngSwitchCase=\"'error'\"\n            text=\"{{ 'Setup not completed' | translate }}\"\n            [size]=\"120\"\n            [vertical]=\"true\"\n            type=\"error\"\n          ></c8y-operation-result>\n        </div>\n        <ul class=\"list-unstyled m-t-16 separator-top\">\n          <li\n            class=\"p-t-4 p-b-4 d-flex a-i-center separator-bottom text-nowrap l-h-1\"\n            *ngFor=\"let step of setup.steps.slice(0, setup.steps.length - 1)\"\n            (click)=\"gotoStep(step.index)\"\n            style=\"cursor: pointer\"\n          >\n            <span *ngIf=\"step.label\">\n              <span class=\"dot small m-r-4 l-h-1\"\n                [ngClass]=\"{ 'dot-success': step.completed, \n                            'dot-error' : !step.completed && step.required,\n                            'dot-warning' : !step.completed && !step.required }\"\n              >\n                {{step.index + 1}}\n              </span>\n              {{ step.label || (STEP_LABEL | translate) + ' ' }}\n            </span>\n            <span *ngIf=\"!step.label\">\n              <span translate>Step</span>{{ ' ' + (step.index + 1) }}\n            </span>\n            <span class=\"m-l-auto icon-20 l-h-1\">\n              <i\n                *ngIf=\"step.completed\"\n                [c8yIcon]=\"'success'\"\n                class=\"text-success\"\n                [title]=\"'Completed' | translate\"\n              ></i>\n              <i\n                *ngIf=\"!step.completed && step.required\"\n                [c8yIcon]=\"'error'\"\n                class=\"text-danger\"\n                [title]=\"\n                  'Not completed and required. Please go to the step and fulfil the setup.'\n                    | translate\n                \"\n              ></i>\n              <i\n                *ngIf=\"!step.completed && !step.required\"\n                [c8yIcon]=\"'warning'\"\n                class=\"text-warning\"\n                [title]=\"'This step was not completed but it is also not required.' | translate\"\n              ></i>\n            </span>\n          </li>\n        </ul>\n      </div>\n    </div>\n  </div>\n</div>\n<div class=\"card-footer separator d-flex j-c-center\">\n  <button\n    translate\n    class=\"btn btn-default\"\n    type=\"button\"\n    (click)=\"back()\"\n    *ngIf=\"stepper.selectedIndex !== 0\"\n    [disabled]=\"\n      setup.currentSetupState === setup.setupState.FINALIZING ||\n      setup.currentSetupState === setup.setupState.RELOADING\n    \"\n  >\n    Back\n  </button>\n  <button\n    class=\"btn btn-primary\"\n    type=\"button\"\n    (click)=\"start()\"\n    [disabled]=\"\n      status === 'error' ||\n      setup.currentSetupState === setup.setupState.FINALIZING ||\n      setup.currentSetupState === setup.setupState.RELOADING\n    \"\n    [ngClass]=\"{\n      'btn-pending':\n        setup.currentSetupState === setup.setupState.FINALIZING ||\n        setup.currentSetupState === setup.setupState.RELOADING\n    }\"\n  >\n    <span *ngIf=\"setup.currentSetupState === setup.setupState.WIZARD\" translate\n      >Save and reload</span\n    >\n    <span *ngIf=\"setup.currentSetupState === setup.setupState.FINALIZING\" translate\n      >Getting things ready\u2026</span\n    >\n    <span *ngIf=\"setup.currentSetupState === setup.setupState.RELOADING\" translate\n      >Done. Reloading the application.</span\n    >\n  </button>\n</div>\n",
                host: { class: 'd-contents' }
            }]
    }], function () { return [{ type: ɵngcc1.C8yStepper }, { type: ɵngcc2.CdkStep }, { type: ɵngcc3.SetupComponent }]; }, null); })();
