// @c8y/ngx-components/esm2015/core/dynamic-forms/date-time/date-time.type.component.js
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { FieldType } from '@ngx-formly/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../date-time-picker/date-time-picker.component';
import * as ɵngcc2 from '@angular/forms';
import * as ɵngcc3 from '../../i18n/c8y-translate.pipe';
export class FieldDateTime extends FieldType {
}
FieldDateTime.ɵfac = /*@__PURE__*/ function () { let ɵFieldDateTime_BaseFactory; return function FieldDateTime_Factory(t) { return (ɵFieldDateTime_BaseFactory || (ɵFieldDateTime_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(FieldDateTime)))(t || FieldDateTime); }; }();
FieldDateTime.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: FieldDateTime, selectors: [["c8y-field-date-time"]], features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 2, vars: 6, consts: [[3, "minDate", "maxDate", "placeholder", "formControl"]], template: function FieldDateTime_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelement(0, "c8y-date-time-picker", 0);
        ɵngcc0.ɵɵpipe(1, "translate");
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("minDate", ctx.to.minDate)("maxDate", ctx.to.maxDate)("placeholder", ɵngcc0.ɵɵpipeBind1(1, 4, ctx.to.placeholder))("formControl", ctx.formControl);
    } }, directives: [ɵngcc1.DateTimePickerComponent, ɵngcc2.NgControlStatus, ɵngcc2.FormControlDirective], pipes: [ɵngcc3.C8yTranslatePipe], encapsulation: 2, changeDetection: 0 });
FieldDateTime.CONFIG = {
    types: [
        {
            name: 'date-time',
            component: FieldDateTime,
            wrappers: ['c8y-form-field']
        }
    ]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(FieldDateTime, [{
        type: Component,
        args: [{
                selector: 'c8y-field-date-time',
                template: "<c8y-date-time-picker\n  [minDate]=\"to.minDate\"\n  [maxDate]=\"to.maxDate\"\n  [placeholder]=\"(to.placeholder | translate)\"\n  [formControl]=\"formControl\"\n></c8y-date-time-picker>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], null, null); })();
