// @c8y/ngx-components/esm2015/operations/bulk-operations-list/bulk-operations-list.module.js
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { CoreModule } from '@c8y/ngx-components';
import { StatusFilterModule } from '@c8y/ngx-components/operations/status-filter';
import { BulkOperationSchedulerModule } from '@c8y/ngx-components/operations/bulk-operation-scheduler';
import { BulkOperationListItemModule } from '@c8y/ngx-components/operations/bulk-operation-list-item';
import { BulkOperationsListComponent } from './bulk-operations-list.component';
import { BulkOperationModalsService } from './modals/bulk-operation-modals.service';
import { BulkOperationsModalComponent } from './modals/bulk-operations-modal.component';
import { BulkOperationTypeListComponent } from './type-list/bulk-operation-type-list.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from '@c8y/ngx-components';
export const bulkOperationListRoutes = [
    {
        path: 'devicecontrol/bulk',
        component: BulkOperationsListComponent
    }
];
/**
 * This module allows for displaying the list of bulk operations and their details.
 */
export class BulkOperationsListModule {
}
BulkOperationsListModule.ɵfac = function BulkOperationsListModule_Factory(t) { return new (t || BulkOperationsListModule)(); };
BulkOperationsListModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: BulkOperationsListModule });
BulkOperationsListModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [BulkOperationModalsService], imports: [[
            CoreModule,
            PopoverModule,
            TooltipModule,
            StatusFilterModule,
            BulkOperationSchedulerModule,
            BulkOperationListItemModule,
            RouterModule.forChild(bulkOperationListRoutes)
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BulkOperationsListModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CoreModule,
                    PopoverModule,
                    TooltipModule,
                    StatusFilterModule,
                    BulkOperationSchedulerModule,
                    BulkOperationListItemModule,
                    RouterModule.forChild(bulkOperationListRoutes)
                ],
                providers: [BulkOperationModalsService],
                declarations: [
                    BulkOperationsListComponent,
                    BulkOperationTypeListComponent,
                    BulkOperationsModalComponent
                ],
                entryComponents: [
                    BulkOperationsListComponent,
                    BulkOperationTypeListComponent,
                    BulkOperationsModalComponent
                ],
                exports: [
                    BulkOperationsListComponent,
                    BulkOperationTypeListComponent,
                    BulkOperationsModalComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(BulkOperationsListModule, { declarations: function () { return [BulkOperationsListComponent,
        BulkOperationTypeListComponent,
        BulkOperationsModalComponent]; }, imports: function () { return [CoreModule,
        PopoverModule,
        TooltipModule,
        StatusFilterModule,
        BulkOperationSchedulerModule,
        BulkOperationListItemModule, ɵngcc1.RouterModule]; }, exports: function () { return [BulkOperationsListComponent,
        BulkOperationTypeListComponent,
        BulkOperationsModalComponent]; } }); })();
ɵngcc0.ɵɵsetComponentScope(BulkOperationsModalComponent, [ɵngcc2.C8yTranslateDirective, ɵngcc2.GuideDocsComponent, ɵngcc2.GuideHrefDirective, BulkOperationTypeListComponent], [ɵngcc2.C8yTranslatePipe]);
