// @c8y/ngx-components/esm2015/connectivity/sigfox/sigfox-tab.factory.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { gettext } from '@c8y/ngx-components';
import { AbstractConnectivityTabFactory } from '../abstract-connectivity-tab.factory';
import { SigfoxGuard } from './sigfox.guard';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from './sigfox.guard';
export class SigfoxTabFactory extends AbstractConnectivityTabFactory {
    constructor(router, guard) {
        super(router, guard);
        this.router = router;
        this.guard = guard;
        this.tabConfig = {
            label: gettext('Sigfox'),
            path: '/connectivitySettings/multiple_lns_connectors_sigfox',
            icon: 'wall-socket'
        };
    }
    config() {
        return this.tabConfig;
    }
}
SigfoxTabFactory.ɵfac = function SigfoxTabFactory_Factory(t) { return new (t || SigfoxTabFactory)(ɵngcc0.ɵɵinject(ɵngcc1.Router), ɵngcc0.ɵɵinject(ɵngcc2.SigfoxGuard)); };
SigfoxTabFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: SigfoxTabFactory, factory: SigfoxTabFactory.ɵfac });
SigfoxTabFactory.ctorParameters = () => [
    { type: Router },
    { type: SigfoxGuard }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SigfoxTabFactory, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.Router }, { type: ɵngcc2.SigfoxGuard }]; }, null); })();
