// @c8y/ngx-components/esm2015/operations/bulk-operation-list-item/bulk-operation-list-item.model.js
import { OperationBulkGeneralStatus } from '@c8y/client';
import { gettext } from '@c8y/ngx-components';
export const BULK_OPERATION_STATUS_LABELS = {
    SCHEDULED: gettext('Scheduled'),
    EXECUTING: gettext('Executing'),
    CANCELED: gettext('Canceled'),
    FAILED: gettext('Completed with failures'),
    SUCCESSFUL: gettext('Completed successfully'),
    DELETED: gettext('Deleted')
};
export const BULK_OPERATION_STATUS_OPTIONS = {
    SCHEDULED: {
        label: BULK_OPERATION_STATUS_LABELS.SCHEDULED,
        generalStatuses: [OperationBulkGeneralStatus.SCHEDULED],
        icon: 'calendar',
        styleClass: 'text-info'
    },
    EXECUTING: {
        label: BULK_OPERATION_STATUS_LABELS.EXECUTING,
        generalStatuses: [
            OperationBulkGeneralStatus.EXECUTING,
            OperationBulkGeneralStatus.EXECUTING_WITH_ERROR
        ],
        icon: 'refresh',
        styleClass: 'text-primary'
    },
    DELETED: {
        label: BULK_OPERATION_STATUS_LABELS.DELETED,
        generalStatuses: [OperationBulkGeneralStatus.DELETED],
        icon: 'delete-document',
        styleClass: 'text-muted'
    },
    CANCELED: {
        label: BULK_OPERATION_STATUS_LABELS.CANCELED,
        generalStatuses: [OperationBulkGeneralStatus.CANCELED],
        icon: 'ban',
        styleClass: 'text-primary'
    },
    FAILED: {
        label: BULK_OPERATION_STATUS_LABELS.FAILED,
        generalStatuses: [OperationBulkGeneralStatus.FAILED],
        icon: 'exclamation-triangle',
        styleClass: 'text-danger'
    },
    SUCCESSFUL: {
        label: BULK_OPERATION_STATUS_LABELS.SUCCESSFUL,
        generalStatuses: [OperationBulkGeneralStatus.SUCCESSFUL],
        icon: 'check-circle',
        styleClass: 'text-success'
    }
};
