// @c8y/ngx-components/esm2015/operations/bulk-operation-list-item/bulk-operation-list-item.module.js
import { NgModule } from '@angular/core';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { CoreModule } from '@c8y/ngx-components';
import { OperationSummaryModule } from '@c8y/ngx-components/operations/operation-summary';
import { BulkOperationSchedulerModule } from '@c8y/ngx-components/operations/bulk-operation-scheduler';
import { OperationsListItemDetailsModule } from '@c8y/ngx-components/operations/operations-list-item-details';
import { BulkOperationListItemComponent } from './bulk-operation-list-item.component';
import { BulkOperationsRescheduleModalComponent } from './modals/bulk-operations-reschedule-modal.component';
import { BulkOperationListItemService } from './bulk-operation-list-item.service';
import { BulkOperationWrapperComponent } from './bulk-operation-wrapper.component';
/**
 * This module allows for displaying the list item of a bulk operation with its details.
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
export class BulkOperationListItemModule {
}
BulkOperationListItemModule.ɵfac = function BulkOperationListItemModule_Factory(t) { return new (t || BulkOperationListItemModule)(); };
BulkOperationListItemModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: BulkOperationListItemModule });
BulkOperationListItemModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [BulkOperationListItemService], imports: [[
            CoreModule,
            PopoverModule,
            TooltipModule,
            OperationSummaryModule,
            BulkOperationSchedulerModule,
            OperationsListItemDetailsModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BulkOperationListItemModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CoreModule,
                    PopoverModule,
                    TooltipModule,
                    OperationSummaryModule,
                    BulkOperationSchedulerModule,
                    OperationsListItemDetailsModule
                ],
                providers: [BulkOperationListItemService],
                declarations: [
                    BulkOperationListItemComponent,
                    BulkOperationsRescheduleModalComponent,
                    BulkOperationWrapperComponent
                ],
                entryComponents: [
                    BulkOperationListItemComponent,
                    BulkOperationsRescheduleModalComponent,
                    BulkOperationWrapperComponent
                ],
                exports: [
                    BulkOperationListItemComponent,
                    BulkOperationsRescheduleModalComponent,
                    BulkOperationWrapperComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(BulkOperationListItemModule, { declarations: function () { return [BulkOperationListItemComponent,
        BulkOperationsRescheduleModalComponent,
        BulkOperationWrapperComponent]; }, imports: function () { return [CoreModule,
        PopoverModule,
        TooltipModule,
        OperationSummaryModule,
        BulkOperationSchedulerModule,
        OperationsListItemDetailsModule]; }, exports: function () { return [BulkOperationListItemComponent,
        BulkOperationsRescheduleModalComponent,
        BulkOperationWrapperComponent]; } }); })();
ɵngcc0.ɵɵsetComponentScope(BulkOperationWrapperComponent, [BulkOperationListItemComponent], [ɵngcc1.C8yTranslatePipe]);
