// @c8y/ngx-components/esm2015/register-device/extensible/bulk/extensible-bulk-device-registration-modal.component.js
import { __awaiter } from "tslib";
import { Component, ViewChild } from '@angular/core';
import { C8yJSONSchema, C8yStepper, gettext } from '@c8y/ngx-components';
import { FormGroup } from '@angular/forms';
import { BulkExtensibleDeviceRegistrationService } from './extensible-bulk-device-registration.service';
import { RegisterDeviceService } from '../../register-device.service';
import { BsModalRef } from 'ngx-bootstrap/modal';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from './extensible-bulk-device-registration.service';
import * as ɵngcc3 from '../../register-device.service';
import * as ɵngcc4 from 'ngx-bootstrap/modal';
import * as ɵngcc5 from '@angular/common';
import * as ɵngcc6 from '@angular/cdk/stepper';
import * as ɵngcc7 from '@ngx-formly/core';

function ExtensibleBulkDeviceRegistrationModalComponent_ng_container_7_div_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 17);
    ɵngcc0.ɵɵelementStart(1, "p", 18);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "p", 19);
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "a", 20);
    ɵngcc0.ɵɵpipe(8, "translate");
    ɵngcc0.ɵɵpipe(9, "translate");
    ɵngcc0.ɵɵelement(10, "i", 21);
    ɵngcc0.ɵɵelementStart(11, "span", 22);
    ɵngcc0.ɵɵtext(12, "Download template");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const example_r5 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 5, example_r5.title), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(6, 7, example_r5.description));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(8, 9, "Download template"));
    ɵngcc0.ɵɵpropertyInterpolate("download", ɵngcc0.ɵɵpipeBind1(9, 11, "Example bulk registration - template.csv"));
    ɵngcc0.ɵɵproperty("href", example_r5.url, ɵngcc0.ɵɵsanitizeUrl);
} }
const _c0 = function () { return { cancel: true, next: true }; };
const _c1 = function () { return { next: "Upload" }; };
function ExtensibleBulkDeviceRegistrationModalComponent_ng_container_7_Template(rf, ctx) { if (rf & 1) {
    const _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "cdk-step", 10);
    ɵngcc0.ɵɵelementStart(2, "div", 11);
    ɵngcc0.ɵɵelementStart(3, "p", 12);
    ɵngcc0.ɵɵtext(4, " Register devices in bulk ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "c8y-form-group", 13);
    ɵngcc0.ɵɵelement(6, "formly-form", 14);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(7, ExtensibleBulkDeviceRegistrationModalComponent_ng_container_7_div_7_Template, 13, 13, "div", 15);
    ɵngcc0.ɵɵelementStart(8, "c8y-stepper-buttons", 16);
    ɵngcc0.ɵɵlistener("onCancel", function ExtensibleBulkDeviceRegistrationModalComponent_ng_container_7_Template_c8y_stepper_buttons_onCancel_8_listener() { ɵngcc0.ɵɵrestoreView(_r7); const ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.cancel(); })("onNext", function ExtensibleBulkDeviceRegistrationModalComponent_ng_container_7_Template_c8y_stepper_buttons_onNext_8_listener() { ɵngcc0.ɵɵrestoreView(_r7); const ctx_r8 = ɵngcc0.ɵɵnextContext(); return ctx_r8.upload(); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("stepControl", ctx_r0.form);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵproperty("form", ctx_r0.form)("fields", ctx_r0.template)("model", ctx_r0.model);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r0.examples);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("showButtons", ɵngcc0.ɵɵpureFunction0(9, _c0))("disabled", ctx_r0.form.invalid)("pending", ctx_r0.pending)("labels", ɵngcc0.ɵɵpureFunction0(10, _c1));
} }
function ExtensibleBulkDeviceRegistrationModalComponent_div_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelement(1, "c8y-operation-result", 23);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementStart(3, "c8y-list-group", 24);
    ɵngcc0.ɵɵelementStart(4, "c8y-li");
    ɵngcc0.ɵɵelement(5, "c8y-li-icon", 25);
    ɵngcc0.ɵɵelementStart(6, "p", 22);
    ɵngcc0.ɵɵtext(7, "File successfully uploaded.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(8, "c8y-li-collapse");
    ɵngcc0.ɵɵelementStart(9, "pre");
    ɵngcc0.ɵɵelementStart(10, "code");
    ɵngcc0.ɵɵtext(11);
    ɵngcc0.ɵɵpipe(12, "json");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("text", ɵngcc0.ɵɵpipeBind1(2, 5, ctx_r1.message));
    ɵngcc0.ɵɵproperty("size", 84)("vertical", true);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("icon", "check-circle");
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(12, 7, ctx_r1.result));
} }
function ExtensibleBulkDeviceRegistrationModalComponent_ng_template_11_c8y_list_group_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-list-group", 24);
    ɵngcc0.ɵɵelementStart(1, "c8y-li");
    ɵngcc0.ɵɵelement(2, "c8y-li-icon", 28);
    ɵngcc0.ɵɵelementStart(3, "p", 22);
    ɵngcc0.ɵɵtext(4, "Bulk operation failed.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "c8y-li-collapse");
    ɵngcc0.ɵɵelementStart(6, "pre");
    ɵngcc0.ɵɵelementStart(7, "code");
    ɵngcc0.ɵɵtext(8);
    ɵngcc0.ɵɵpipe(9, "json");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r9 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("icon", "ban");
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(9, 2, ctx_r9.result));
} }
function ExtensibleBulkDeviceRegistrationModalComponent_ng_template_11_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-operation-result", 26);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtemplate(2, ExtensibleBulkDeviceRegistrationModalComponent_ng_template_11_c8y_list_group_2_Template, 10, 4, "c8y-list-group", 27);
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("text", ɵngcc0.ɵɵpipeBind1(1, 4, ctx_r3.message));
    ɵngcc0.ɵɵproperty("size", 84)("vertical", true);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.result);
} }
const _c2 = function () { return { next: true }; };
const _c3 = function (a0) { return { next: a0 }; };
const bulkSchema = {
    $schema: 'https://json-schema.org/draft/2019-09/schema',
    type: 'object',
    properties: {
        csvBulkFile: {
            type: 'string',
            title: gettext('CSV file upload'),
            description: gettext('You can use file upload component to let users send files. This input accepts only a single CSV file.'),
            contentEncoding: 'base64',
            contentMediaType: 'csv'
        }
    },
    required: ['csvBulkFile'],
    additionalProperties: false
};
export class ExtensibleBulkDeviceRegistrationModalComponent {
    constructor(jsonschema, bulkExtensibleDeviceRegistrationService, registerDeviceService, bsModalRef) {
        this.jsonschema = jsonschema;
        this.bulkExtensibleDeviceRegistrationService = bulkExtensibleDeviceRegistrationService;
        this.registerDeviceService = registerDeviceService;
        this.bsModalRef = bsModalRef;
        this.loadingError = false;
        this.pending = false;
        this.form = new FormGroup({});
        this.model = {};
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            const { title, exampleFileUrls } = yield this.bulkExtensibleDeviceRegistrationService
                .getBulkDescription(this.contextPath)
                .catch(error => {
                this.message = error.message;
                this.loadingError = true;
                return {
                    title: gettext('Error'),
                    exampleFileUrls: []
                };
            });
            this.title = title;
            this.template = [this.jsonschema.toFieldConfig(bulkSchema)];
            this.examples = exampleFileUrls;
        });
    }
    upload() {
        return __awaiter(this, void 0, void 0, function* () {
            this.pending = true;
            const { res, data } = yield this.bulkExtensibleDeviceRegistrationService
                .upload(this.getFile(this.model), this.contextPath)
                .catch(err => {
                return {
                    res: undefined,
                    data: undefined
                };
            })
                .finally(() => {
                this.pending = false;
                this.stepper.next();
            });
            this.result = data;
            this.success = (res === null || res === void 0 ? void 0 : res.status) < 400;
            if (this.success) {
                this.message = gettext('Device registration created.');
                this.model = {};
            }
            if (res) {
                if (res.status >= 400 && res.status < 500) {
                    this.message = gettext('Device registration failed.');
                }
                else if (res.status >= 500) {
                    this.message = gettext('Error occurred while processing the uploaded file.');
                }
            }
            else {
                this.message = gettext('No internet connection. Unable to reach the microservice.');
            }
        });
    }
    getFile(model) {
        var _a, _b;
        const csvBulkFile = (_a = model) === null || _a === void 0 ? void 0 : _a.csvBulkFile;
        return csvBulkFile ? (_b = csvBulkFile[0]) === null || _b === void 0 ? void 0 : _b.file : undefined;
    }
    complete() {
        this.registerDeviceService.list();
        this.bsModalRef.hide();
    }
    cancel() {
        this.bsModalRef.hide();
    }
}
ExtensibleBulkDeviceRegistrationModalComponent.ɵfac = function ExtensibleBulkDeviceRegistrationModalComponent_Factory(t) { return new (t || ExtensibleBulkDeviceRegistrationModalComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.C8yJSONSchema), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.BulkExtensibleDeviceRegistrationService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.RegisterDeviceService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.BsModalRef)); };
ExtensibleBulkDeviceRegistrationModalComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ExtensibleBulkDeviceRegistrationModalComponent, selectors: [["extensible-device-registration-modal-component"]], viewQuery: function ExtensibleBulkDeviceRegistrationModalComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(C8yStepper, 7);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.stepper = _t.first);
    } }, decls: 14, vars: 14, consts: [[3, "headerClasses", "customFooter"], ["c8y-modal-title", ""], [1, "c8y-icon", "c8y-icon-device-connect"], ["linear", "", 1, "p-t-16", 3, "hideStepProgress"], [4, "ngIf"], ["state", "final"], [1, "p-24"], [4, "ngIf", "ngIfElse"], ["fail", ""], [1, "sticky-bottom", "d-block", "p-t-16", "p-b-16", "separator-top", "bg-white", 3, "showButtons", "labels", "onNext"], [3, "stepControl"], [1, "sticky-top", "bg-component", "separator-bottom"], ["translate", "", 1, "p-l-24", "p-r-24", "p-t-16", "p-b-16", "m-b-0", "separator-bottom", "text-16", "text-medium", "text-center", "bg-component"], [1, "p-24", "p-t-16", "p-b-8", "m-b-0"], [3, "form", "fields", "model"], ["class", "bg-gray-white separator-bottom p-t-16 p-b-16 p-l-24 p-r-24", 4, "ngFor", "ngForOf"], [1, "sticky-bottom", "d-block", "p-t-16", "p-b-16", "separator-top", "bg-white", 3, "showButtons", "disabled", "pending", "labels", "onCancel", "onNext"], [1, "bg-gray-white", "separator-bottom", "p-t-16", "p-b-16", "p-l-24", "p-r-24"], [1, "m-b-8", "text-medium"], [1, "small", "m-b-8"], ["target", "_self", 1, "btn", "btn-default", "btn-sm", 3, "title", "href", "download"], ["c8yIcon", "download"], ["translate", ""], ["type", "success", 1, "lead", 3, "text", "size", "vertical"], [1, "separator-top", "m-t-16"], [1, "text-success", 3, "icon"], ["type", "error", 1, "lead", 3, "text", "size", "vertical"], ["class", "separator-top m-t-16", 4, "ngIf"], [1, "text-danger", 3, "icon"]], template: function ExtensibleBulkDeviceRegistrationModalComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-modal", 0);
        ɵngcc0.ɵɵelementContainerStart(1, 1);
        ɵngcc0.ɵɵelement(2, "span", 2);
        ɵngcc0.ɵɵelementStart(3, "h4");
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementContainerEnd();
        ɵngcc0.ɵɵelementStart(6, "c8y-stepper", 3);
        ɵngcc0.ɵɵtemplate(7, ExtensibleBulkDeviceRegistrationModalComponent_ng_container_7_Template, 9, 11, "ng-container", 4);
        ɵngcc0.ɵɵelementStart(8, "cdk-step", 5);
        ɵngcc0.ɵɵelementStart(9, "div", 6);
        ɵngcc0.ɵɵtemplate(10, ExtensibleBulkDeviceRegistrationModalComponent_div_10_Template, 13, 9, "div", 7);
        ɵngcc0.ɵɵtemplate(11, ExtensibleBulkDeviceRegistrationModalComponent_ng_template_11_Template, 3, 6, "ng-template", null, 8, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(13, "c8y-stepper-buttons", 9);
        ɵngcc0.ɵɵlistener("onNext", function ExtensibleBulkDeviceRegistrationModalComponent_Template_c8y_stepper_buttons_onNext_13_listener() { return ctx.complete(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r2 = ɵngcc0.ɵɵreference(12);
        ɵngcc0.ɵɵproperty("headerClasses", "dialog-header")("customFooter", true);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(5, 9, ctx.title));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("hideStepProgress", true);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.loadingError);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.success)("ngIfElse", _r2);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("showButtons", ɵngcc0.ɵɵpureFunction0(11, _c2))("labels", ɵngcc0.ɵɵpureFunction1(12, _c3, ctx.success ? "Complete" : "Cancel"));
    } }, directives: [ɵngcc1.ModalComponent, ɵngcc1.C8yStepper, ɵngcc5.NgIf, ɵngcc6.CdkStep, ɵngcc1.C8yStepperButtons, ɵngcc1.C8yTranslateDirective, ɵngcc1.FormGroupComponent, ɵngcc7.FormlyForm, ɵngcc5.NgForOf, ɵngcc1.IconDirective, ɵngcc1.OperationResultComponent, ɵngcc1.ListGroupComponent, ɵngcc1.ListItemComponent, ɵngcc1.ListItemIconComponent, ɵngcc1.ListItemCollapseComponent], pipes: [ɵngcc1.C8yTranslatePipe, ɵngcc5.JsonPipe], encapsulation: 2 });
ExtensibleBulkDeviceRegistrationModalComponent.ctorParameters = () => [
    { type: C8yJSONSchema },
    { type: BulkExtensibleDeviceRegistrationService },
    { type: RegisterDeviceService },
    { type: BsModalRef }
];
ExtensibleBulkDeviceRegistrationModalComponent.propDecorators = {
    stepper: [{ type: ViewChild, args: [C8yStepper, { static: true },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ExtensibleBulkDeviceRegistrationModalComponent, [{
        type: Component,
        args: [{
                selector: 'extensible-device-registration-modal-component',
                template: "<c8y-modal [headerClasses]=\"'dialog-header'\" [customFooter]=\"true\">\n  <ng-container c8y-modal-title>\n    <span class=\"c8y-icon c8y-icon-device-connect\"></span>\n    <h4>{{ title | translate }}</h4>\n  </ng-container>\n  <c8y-stepper [hideStepProgress]=\"true\" linear class=\"p-t-16\">\n    <ng-container *ngIf=\"!loadingError\">\n      <cdk-step [stepControl]=\"form\">\n        <div class=\"sticky-top bg-component separator-bottom\">\n          <p\n            class=\"\n              p-l-24 p-r-24 p-t-16 p-b-16\n              m-b-0\n              separator-bottom\n              text-16 text-medium text-center\n              bg-component\n            \"\n            translate\n          >\n            Register devices in bulk\n          </p>\n          <c8y-form-group class=\"p-24 p-t-16 p-b-8 m-b-0\">\n            <formly-form [form]=\"form\" [fields]=\"template\" [model]=\"model\"></formly-form>\n          </c8y-form-group>\n        </div>\n        <div\n          *ngFor=\"let example of examples\"\n          class=\"bg-gray-white separator-bottom p-t-16 p-b-16 p-l-24 p-r-24\"\n        >\n          <p class=\"m-b-8 text-medium\">\n            {{ example.title | translate }}\n          </p>\n          <p class=\"small m-b-8\">{{ example.description | translate }}</p>\n          <a\n            title=\"{{ 'Download template' | translate }}\"\n            class=\"btn btn-default btn-sm\"\n            target=\"_self\"\n            [href]=\"example.url\"\n            download=\"{{ 'Example bulk registration - template.csv' | translate }}\"\n          >\n            <i c8yIcon=\"download\"></i>\n            <span translate>Download template</span>\n          </a>\n        </div>\n        <c8y-stepper-buttons\n          class=\"sticky-bottom d-block p-t-16 p-b-16 separator-top bg-white\"\n          [showButtons]=\"{ cancel: true, next: true }\"\n          [disabled]=\"form.invalid\"\n          [pending]=\"pending\"\n          (onCancel)=\"cancel()\"\n          (onNext)=\"upload()\"\n          [labels]=\"{ next: 'Upload' }\"\n        ></c8y-stepper-buttons>\n      </cdk-step>\n    </ng-container>\n    <cdk-step state=\"final\">\n      <div class=\"p-24\">\n        <div *ngIf=\"success; else fail\">\n          <c8y-operation-result\n            text=\"{{ message | translate }}\"\n            [size]=\"84\"\n            [vertical]=\"true\"\n            type=\"success\"\n            class=\"lead\"\n          ></c8y-operation-result>\n          <c8y-list-group class=\"separator-top m-t-16\">\n            <c8y-li>\n              <c8y-li-icon class=\"text-success\" [icon]=\"'check-circle'\"></c8y-li-icon>\n              <p translate>File successfully uploaded.</p>\n              <c8y-li-collapse>\n                <pre><code>{{ result | json }}</code></pre>\n              </c8y-li-collapse>\n            </c8y-li>\n          </c8y-list-group>\n        </div>\n        <ng-template #fail>\n          <c8y-operation-result\n            text=\"{{ message | translate }}\"\n            [size]=\"84\"\n            [vertical]=\"true\"\n            type=\"error\"\n            class=\"lead\"\n          ></c8y-operation-result>\n          <c8y-list-group *ngIf=\"result\" class=\"separator-top m-t-16\">\n            <c8y-li>\n              <c8y-li-icon class=\"text-danger\" [icon]=\"'ban'\"></c8y-li-icon>\n              <p translate>Bulk operation failed.</p>\n              <c8y-li-collapse>\n                <pre><code>{{ result | json }}</code></pre>\n              </c8y-li-collapse>\n            </c8y-li>\n          </c8y-list-group>\n        </ng-template>\n      </div>\n      <c8y-stepper-buttons\n        class=\"sticky-bottom d-block p-t-16 p-b-16 separator-top bg-white\"\n        [showButtons]=\"{ next: true }\"\n        (onNext)=\"complete()\"\n        [labels]=\"{ next: success ? 'Complete' : 'Cancel' }\"\n      ></c8y-stepper-buttons>\n    </cdk-step>\n  </c8y-stepper>\n</c8y-modal>\n"
            }]
    }], function () { return [{ type: ɵngcc1.C8yJSONSchema }, { type: ɵngcc2.BulkExtensibleDeviceRegistrationService }, { type: ɵngcc3.RegisterDeviceService }, { type: ɵngcc4.BsModalRef }]; }, { stepper: [{
            type: ViewChild,
            args: [C8yStepper, { static: true }]
        }] }); })();
