// @c8y/ngx-components/esm2015/register-device/extensible/bulk/extensible-bulk-device-registration-button.component.js
import { Component } from '@angular/core';
import { BulkExtensibleDeviceRegistrationService } from './extensible-bulk-device-registration.service';
import { BsModalService } from 'ngx-bootstrap/modal';
import { ExtensibleBulkDeviceRegistrationModalComponent } from './extensible-bulk-device-registration-modal.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './extensible-bulk-device-registration.service';
import * as ɵngcc2 from 'ngx-bootstrap/modal';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from '@c8y/ngx-components';

function ExtensibleBulkDeviceRegistrationButtonComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    const _r3 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "button", 1);
    ɵngcc0.ɵɵlistener("click", function ExtensibleBulkDeviceRegistrationButtonComponent_div_0_Template_button_click_1_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r3); const provider_r1 = restoredCtx.$implicit; const ctx_r2 = ɵngcc0.ɵɵnextContext(); return ctx_r2.click(provider_r1.contextPath); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 2);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const provider_r1 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 2, provider_r1.name));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 4, provider_r1.name), " ");
} }
export class ExtensibleBulkDeviceRegistrationButtonComponent {
    constructor(bulkExtensibleDeviceRegistrationService, modalService) {
        this.bulkExtensibleDeviceRegistrationService = bulkExtensibleDeviceRegistrationService;
        this.modalService = modalService;
    }
    ngOnInit() {
        this.bulkProvidersMetaData = this.bulkExtensibleDeviceRegistrationService.getDeviceRegistrationProviders();
    }
    click(contextPath) {
        this.modalService.show(ExtensibleBulkDeviceRegistrationModalComponent, {
            initialState: { contextPath },
            class: 'modal-sm',
            ignoreBackdropClick: true
        });
    }
}
ExtensibleBulkDeviceRegistrationButtonComponent.ɵfac = function ExtensibleBulkDeviceRegistrationButtonComponent_Factory(t) { return new (t || ExtensibleBulkDeviceRegistrationButtonComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BulkExtensibleDeviceRegistrationService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.BsModalService)); };
ExtensibleBulkDeviceRegistrationButtonComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ExtensibleBulkDeviceRegistrationButtonComponent, selectors: [["extensible-device-registration-component"]], decls: 2, vars: 3, consts: [[4, "ngFor", "ngForOf"], ["type", "button", 3, "title", "click"], ["c8yIcon", "upload"]], template: function ExtensibleBulkDeviceRegistrationButtonComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ExtensibleBulkDeviceRegistrationButtonComponent_div_0_Template, 6, 6, "div", 0);
        ɵngcc0.ɵɵpipe(1, "async");
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngForOf", ɵngcc0.ɵɵpipeBind1(1, 1, ctx.bulkProvidersMetaData));
    } }, directives: [ɵngcc3.NgForOf, ɵngcc4.IconDirective], pipes: [ɵngcc3.AsyncPipe, ɵngcc4.C8yTranslatePipe], encapsulation: 2 });
ExtensibleBulkDeviceRegistrationButtonComponent.ctorParameters = () => [
    { type: BulkExtensibleDeviceRegistrationService },
    { type: BsModalService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ExtensibleBulkDeviceRegistrationButtonComponent, [{
        type: Component,
        args: [{
                selector: 'extensible-device-registration-component',
                template: "<div *ngFor=\"let provider of bulkProvidersMetaData | async\">\n  <button\n    title=\"{{ provider.name | translate }}\"\n    type=\"button\"\n    (click)=\"click(provider.contextPath)\"\n  >\n    <i c8yIcon=\"upload\"></i>\n    {{ provider.name | translate }}\n  </button>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.BulkExtensibleDeviceRegistrationService }, { type: ɵngcc2.BsModalService }]; }, null); })();
