// @c8y/ngx-components/esm2015/repository/configuration/list/configuration-repository-list.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ConfigurationListComponent } from './configuration-list.component';
import { ConfigurationDetailComponent } from './configuration-detail.component';
import { CoreModule, HOOK_NAVIGATOR_NODES, HOOK_ROUTE } from '@c8y/ngx-components';
import { ConfigurationRepositoryNavigationFactory } from './configuration-repository-navigation-factory';
import { SharedRepositoryModule } from '@c8y/ngx-components/repository/shared';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
export class ConfigurationRepositoryListModule {
    static forRoot() {
        return {
            ngModule: ConfigurationRepositoryListModule,
            providers: [
                {
                    provide: HOOK_NAVIGATOR_NODES,
                    useClass: ConfigurationRepositoryNavigationFactory,
                    multi: true
                },
                {
                    provide: HOOK_ROUTE,
                    useValue: [
                        {
                            path: 'configuration',
                            component: ConfigurationListComponent
                        },
                    ],
                    multi: true
                }
            ]
        };
    }
}
ConfigurationRepositoryListModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    CoreModule,
                    SharedRepositoryModule,
                    TooltipModule
                ],
                declarations: [
                    ConfigurationListComponent,
                    ConfigurationDetailComponent,
                ],
                entryComponents: [
                    ConfigurationListComponent,
                    ConfigurationDetailComponent,
                ]
            },] }
];
