// @c8y/ngx-components/esm2015/register-device/register-device.module.js
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CommonModule, CoreModule, DeviceBootstrapRealtimeService, DynamicFormsModule, HOOK_NAVIGATOR_NODES, StepperModule } from '@c8y/ngx-components';
import { DeviceRegistrationViewComponent } from './device-registration-view.component';
import { GeneralDeviceRegistrationComponent } from './general/general-device-registration.component';
import { GeneralDeviceRegistrationButtonComponent } from './general/general-device-registration-button.component';
import { RegisterDeviceNavigationFactory } from './register-device-navigation.factory';
import { RegisterDeviceService } from './register-device.service';
import { ExtensibleDeviceRegistrationButtonComponent } from './extensible/single/extensible-device-registration-button.component';
import { ExtensibleDeviceRegistrationModalComponent } from './extensible/single/extensible-device-registration-modal.component';
import { ExtensibleDeviceRegistrationService } from './extensible/single/extensible-device-registration.service';
import { ExtensibleDeviceRegistrationStepperComponent } from './extensible/single/extensible-device-registration-stepper.component';
import { ExtensibleBulkDeviceRegistrationModalComponent } from './extensible/bulk/extensible-bulk-device-registration-modal.component';
import { BulkExtensibleDeviceRegistrationService } from './extensible/bulk/extensible-bulk-device-registration.service';
import { BulkDeviceRegistrationModalComponent } from './bulk/bulk-device-registration-modal.component';
import { ExtensibleBulkDeviceRegistrationButtonComponent } from './extensible/bulk/extensible-bulk-device-registration-button.component';
import { BulkDeviceRegistrationButtonComponent } from './bulk/bulk-device-registration-button.component';
import { RegisterDeviceDropdownComponent } from './dropdown/register-device-dropdown.component';
import { HOOK_DEVICE_REGISTRATION } from './dropdown/register-device-extension.service';
import { DeviceRegistrationFactory } from './register-device.factory';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { PopoverModule } from 'ngx-bootstrap/popover';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
const routes = [
    {
        path: 'deviceregistration',
        component: DeviceRegistrationViewComponent,
    }
];
export class RegisterDeviceModule {
}
RegisterDeviceModule.ɵfac = function RegisterDeviceModule_Factory(t) { return new (t || RegisterDeviceModule)(); };
RegisterDeviceModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: RegisterDeviceModule });
RegisterDeviceModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [{
            provide: HOOK_DEVICE_REGISTRATION,
            useClass: DeviceRegistrationFactory,
            multi: true
        },
        {
            provide: HOOK_NAVIGATOR_NODES,
            useClass: RegisterDeviceNavigationFactory,
            multi: true
        },
        DeviceBootstrapRealtimeService,
        RegisterDeviceService,
        ExtensibleDeviceRegistrationService,
        BulkExtensibleDeviceRegistrationService,
    ], imports: [[BsDropdownModule, PopoverModule, CommonModule, CoreModule, StepperModule, DynamicFormsModule, RouterModule.forChild(routes)]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(RegisterDeviceModule, [{
        type: NgModule,
        args: [{
                imports: [BsDropdownModule, PopoverModule, CommonModule, CoreModule, StepperModule, DynamicFormsModule, RouterModule.forChild(routes)],
                declarations: [
                    DeviceRegistrationViewComponent,
                    GeneralDeviceRegistrationComponent,
                    GeneralDeviceRegistrationButtonComponent,
                    BulkDeviceRegistrationModalComponent,
                    BulkDeviceRegistrationButtonComponent,
                    ExtensibleDeviceRegistrationButtonComponent,
                    ExtensibleDeviceRegistrationModalComponent,
                    ExtensibleDeviceRegistrationStepperComponent,
                    ExtensibleBulkDeviceRegistrationModalComponent,
                    ExtensibleBulkDeviceRegistrationButtonComponent,
                    RegisterDeviceDropdownComponent
                ],
                providers: [{
                        provide: HOOK_DEVICE_REGISTRATION,
                        useClass: DeviceRegistrationFactory,
                        multi: true
                    },
                    {
                        provide: HOOK_NAVIGATOR_NODES,
                        useClass: RegisterDeviceNavigationFactory,
                        multi: true
                    },
                    DeviceBootstrapRealtimeService,
                    RegisterDeviceService,
                    ExtensibleDeviceRegistrationService,
                    BulkExtensibleDeviceRegistrationService,
                ],
                entryComponents: [
                    GeneralDeviceRegistrationComponent,
                    ExtensibleBulkDeviceRegistrationButtonComponent,
                    BulkDeviceRegistrationModalComponent,
                    ExtensibleDeviceRegistrationModalComponent,
                    RegisterDeviceDropdownComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(RegisterDeviceModule, { declarations: function () { return [DeviceRegistrationViewComponent,
        GeneralDeviceRegistrationComponent,
        GeneralDeviceRegistrationButtonComponent,
        BulkDeviceRegistrationModalComponent,
        BulkDeviceRegistrationButtonComponent,
        ExtensibleDeviceRegistrationButtonComponent,
        ExtensibleDeviceRegistrationModalComponent,
        ExtensibleDeviceRegistrationStepperComponent,
        ExtensibleBulkDeviceRegistrationModalComponent,
        ExtensibleBulkDeviceRegistrationButtonComponent,
        RegisterDeviceDropdownComponent]; }, imports: function () { return [BsDropdownModule, PopoverModule, CommonModule, CoreModule, StepperModule, DynamicFormsModule, ɵngcc1.RouterModule]; } }); })();
