// @c8y/ngx-components/esm2015/register-device/register-device-navigation.factory.js
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode, Permissions } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
export class RegisterDeviceNavigationFactory {
    constructor(permissions) {
        this.permissions = permissions;
    }
    get() {
        const userHasPermission = this.permissions.hasAnyRole([
            'ROLE_DEVICE_CONTROL_ADMIN',
            'ROLE_DEVICE_CONTROL_READ'
        ]);
        const nav = [];
        if (userHasPermission) {
            nav.push(new NavigatorNode({
                label: gettext('Registration'),
                icon: 'c8y-device-connect',
                path: '/deviceregistration',
                parent: gettext('Devices'),
                preventDuplicates: true,
                priority: 3000
            }));
        }
        return nav;
    }
}
RegisterDeviceNavigationFactory.ɵfac = function RegisterDeviceNavigationFactory_Factory(t) { return new (t || RegisterDeviceNavigationFactory)(ɵngcc0.ɵɵinject(ɵngcc1.Permissions)); };
RegisterDeviceNavigationFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: RegisterDeviceNavigationFactory, factory: RegisterDeviceNavigationFactory.ɵfac });
RegisterDeviceNavigationFactory.ctorParameters = () => [
    { type: Permissions }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(RegisterDeviceNavigationFactory, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.Permissions }]; }, null); })();
