// @c8y/ngx-components/esm2015/register-device/register-device.factory.js
import { Injectable } from '@angular/core';
import { GeneralDeviceRegistrationButtonComponent } from './general/general-device-registration-button.component';
import { ExtensibleDeviceRegistrationButtonComponent } from './extensible/single/extensible-device-registration-button.component';
import { ExtensibleBulkDeviceRegistrationButtonComponent } from './extensible/bulk/extensible-bulk-device-registration-button.component';
import { BulkDeviceRegistrationButtonComponent } from './bulk/bulk-device-registration-button.component';
import * as ɵngcc0 from '@angular/core';
export class DeviceRegistrationFactory {
    get() {
        return [{
                template: GeneralDeviceRegistrationButtonComponent,
                priority: 100,
                category: 'single'
            }, {
                template: ExtensibleDeviceRegistrationButtonComponent,
                priority: 50,
                category: 'single'
            }, {
                template: ExtensibleBulkDeviceRegistrationButtonComponent,
                priority: 50,
                category: 'bulk'
            }, {
                template: BulkDeviceRegistrationButtonComponent,
                priority: 100,
                category: 'bulk'
            }];
    }
}
DeviceRegistrationFactory.ɵfac = function DeviceRegistrationFactory_Factory(t) { return new (t || DeviceRegistrationFactory)(); };
DeviceRegistrationFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: DeviceRegistrationFactory, factory: DeviceRegistrationFactory.ɵfac });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DeviceRegistrationFactory, [{
        type: Injectable
    }], null, null); })();
