// @c8y/ngx-components/esm2015/core/stepper/stepper.module.js
import { NgModule } from '@angular/core';
import { CdkStepperModule, CdkStep, STEPPER_GLOBAL_OPTIONS } from '@angular/cdk/stepper';
import { C8yStepper } from './stepper.component';
import { CommonModule } from '../common/common.module';
import { C8yStepperIcon } from './stepper-icon.directive';
import { C8yStepperProgress } from './stepper-progress.component';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { C8yStepperButtons } from './stepper-buttons.component';
import { StepperService } from './stepper.service';
import { StepperOutletComponent } from './stepper-outlet.component';
import { FormsModule } from '@angular/forms';
import * as ɵngcc0 from '@angular/core';
const ɵ0 = { displayDefaultIndicatorType: false };
/**
 * The angular module definition for c8y-stepper (Wizard).
 * @exports C8yStepper
 * @exports C8yStepperIcon
 * @exports C8yStepperButtons
 * @exports CdkStep
 */
export class StepperModule {
}
StepperModule.ɵfac = function StepperModule_Factory(t) { return new (t || StepperModule)(); };
StepperModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: StepperModule });
StepperModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        StepperService,
        {
            /**
             * This enables additional states, like: edit, done and error
             * These states are used to display custom icons, if needed.
             * Any change in here will most probably result in not showing custom icons anymore
             */
            provide: STEPPER_GLOBAL_OPTIONS,
            useValue: ɵ0
        }
    ], imports: [[CommonModule, CdkStepperModule, TooltipModule, FormsModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(StepperModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, CdkStepperModule, TooltipModule, FormsModule],
                declarations: [
                    C8yStepper,
                    C8yStepperIcon,
                    C8yStepperProgress,
                    C8yStepperButtons,
                    StepperOutletComponent
                ],
                exports: [C8yStepper, CdkStep, C8yStepperIcon, C8yStepperButtons, StepperOutletComponent],
                providers: [
                    StepperService,
                    {
                        /**
                         * This enables additional states, like: edit, done and error
                         * These states are used to display custom icons, if needed.
                         * Any change in here will most probably result in not showing custom icons anymore
                         */
                        provide: STEPPER_GLOBAL_OPTIONS,
                        useValue: ɵ0
                    }
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(StepperModule, { declarations: function () { return [C8yStepper,
        C8yStepperIcon,
        C8yStepperProgress,
        C8yStepperButtons,
        StepperOutletComponent]; }, imports: function () { return [CommonModule, CdkStepperModule, TooltipModule, FormsModule]; }, exports: function () { return [C8yStepper, CdkStep, C8yStepperIcon, C8yStepperButtons, StepperOutletComponent]; } }); })();
export { ɵ0 };
