// @c8y/ngx-components/esm2015/core/stepper/stepper-outlet.component.js
import { Component, Input } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './stepper.component';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from '@angular/cdk/stepper';
import * as ɵngcc4 from './stepper-icon.directive';
import * as ɵngcc5 from './stepper-buttons.component';
import * as ɵngcc6 from '../common/icon.directive';

function StepperOutletComponent_cdk_step_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function StepperOutletComponent_cdk_step_1_c8y_stepper_buttons_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-stepper-buttons");
} }
function StepperOutletComponent_cdk_step_1_ng_template_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 8);
} }
function StepperOutletComponent_cdk_step_1_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 9);
} }
function StepperOutletComponent_cdk_step_1_ng_template_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 10);
} }
function StepperOutletComponent_cdk_step_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "cdk-step", 2);
    ɵngcc0.ɵɵtemplate(1, StepperOutletComponent_cdk_step_1_ng_container_1_Template, 1, 0, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(2, StepperOutletComponent_cdk_step_1_c8y_stepper_buttons_2_Template, 1, 0, "c8y-stepper-buttons", 4);
    ɵngcc0.ɵɵtemplate(3, StepperOutletComponent_cdk_step_1_ng_template_3_Template, 1, 0, "ng-template", 5);
    ɵngcc0.ɵɵtemplate(4, StepperOutletComponent_cdk_step_1_ng_template_4_Template, 1, 0, "ng-template", 6);
    ɵngcc0.ɵɵtemplate(5, StepperOutletComponent_cdk_step_1_ng_template_5_Template, 1, 0, "ng-template", 7);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const step_r1 = ctx.$implicit;
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("label", step_r1.label)("state", step_r1.state)("completed", step_r1.completed)("optional", !step_r1.required);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngComponentOutlet", step_r1.component);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.showDefaultButtons);
} }
const _c0 = function () { return ["p-t-32", "p-b-32", "m-l-48", "m-r-48"]; };
const _c1 = function () { return { edit: true, done: false }; };
export class StepperOutletComponent {
    constructor() {
        this.showDefaultButtons = true;
    }
}
StepperOutletComponent.ɵfac = function StepperOutletComponent_Factory(t) { return new (t || StepperOutletComponent)(); };
StepperOutletComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: StepperOutletComponent, selectors: [["c8y-stepper-outlet"]], inputs: { showDefaultButtons: "showDefaultButtons", steps: "steps" }, decls: 2, vars: 5, consts: [[1, "d-flex", "d-col", "fit-h", 3, "customClasses", "disableDefaultIcons"], [3, "label", "state", "completed", "optional", 4, "ngFor", "ngForOf"], [3, "label", "state", "completed", "optional"], [4, "ngComponentOutlet"], [4, "ngIf"], ["c8yStepperIcon", "error"], ["c8yStepperIcon", "warning"], ["c8yStepperIcon", "done"], ["c8yIcon", "error"], ["c8yIcon", "warning"], ["c8yIcon", "check"]], template: function StepperOutletComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-stepper", 0);
        ɵngcc0.ɵɵtemplate(1, StepperOutletComponent_cdk_step_1_Template, 6, 6, "cdk-step", 1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("customClasses", ɵngcc0.ɵɵpureFunction0(3, _c0))("disableDefaultIcons", ɵngcc0.ɵɵpureFunction0(4, _c1));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.steps);
    } }, directives: [ɵngcc1.C8yStepper, ɵngcc2.NgForOf, ɵngcc3.CdkStep, ɵngcc2.NgComponentOutlet, ɵngcc2.NgIf, ɵngcc4.C8yStepperIcon, ɵngcc5.C8yStepperButtons, ɵngcc6.IconDirective], encapsulation: 2 });
StepperOutletComponent.propDecorators = {
    steps: [{ type: Input }],
    showDefaultButtons: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(StepperOutletComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-stepper-outlet',
                template: "<c8y-stepper \n  [customClasses]=\"['p-t-32', 'p-b-32', 'm-l-48', 'm-r-48']\" \n  class=\"d-flex d-col fit-h\"\n  [disableDefaultIcons]=\"{ edit: true, done: false }\"\n>\n  <cdk-step\n    *ngFor=\"let step of steps\"\n    [label]=\"step.label\"\n    [state]=\"step.state\"\n    [completed]=\"step.completed\"\n    [optional]=\"!step.required\"\n  >\n    <ng-container *ngComponentOutlet=\"step.component\"></ng-container>\n    <c8y-stepper-buttons *ngIf=\"showDefaultButtons\"></c8y-stepper-buttons>\n    <ng-template c8yStepperIcon=\"error\">\n      <span c8yIcon=\"error\"></span>\n    </ng-template>\n    <ng-template c8yStepperIcon=\"warning\">\n      <span c8yIcon=\"warning\"></span>\n    </ng-template>\n    <ng-template c8yStepperIcon=\"done\">\n      <span c8yIcon=\"check\"></span>\n    </ng-template>\n  </cdk-step>\n</c8y-stepper>\n"
            }]
    }], function () { return []; }, { showDefaultButtons: [{
            type: Input
        }], steps: [{
            type: Input
        }] }); })();
