// @c8y/ngx-components/esm2015/device-shell/command-templates/command-templates.directive.js
import { Directive, ElementRef, Injector, Output } from '@angular/core';
import { UpgradeComponent } from '@angular/upgrade/static';
import * as ɵngcc0 from '@angular/core';
export class CommandTemplates extends UpgradeComponent {
    constructor(elementRef, injector) {
        super('c8yCommandTemplates', elementRef, injector);
    }
}
CommandTemplates.ɵfac = function CommandTemplates_Factory(t) { return new (t || CommandTemplates)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Injector)); };
CommandTemplates.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: CommandTemplates, selectors: [["c8y-command-templates-directive"]], outputs: { close: "close", dismiss: "dismiss" }, features: [ɵngcc0.ɵɵInheritDefinitionFeature] });
CommandTemplates.ctorParameters = () => [
    { type: ElementRef },
    { type: Injector }
];
CommandTemplates.propDecorators = {
    close: [{ type: Output }],
    dismiss: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CommandTemplates, [{
        type: Directive,
        args: [{
                selector: 'c8y-command-templates-directive'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.Injector }]; }, { close: [{
            type: Output
        }], dismiss: [{
            type: Output
        }] }); })();
