// @c8y/ngx-components/esm2015/device-shell/command-templates/command-templates.module.js
import { NgModule } from '@angular/core';
import { CommandTemplatesComponent } from './command-templates.component';
import { CommandTemplates } from './command-templates.directive';
import * as ɵngcc0 from '@angular/core';
export class CommandTemplatesModule {
}
CommandTemplatesModule.ɵfac = function CommandTemplatesModule_Factory(t) { return new (t || CommandTemplatesModule)(); };
CommandTemplatesModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: CommandTemplatesModule });
CommandTemplatesModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({});
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CommandTemplatesModule, [{
        type: NgModule,
        args: [{
                declarations: [CommandTemplates, CommandTemplatesComponent],
                exports: [CommandTemplatesComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(CommandTemplatesModule, { declarations: function () { return [CommandTemplates, CommandTemplatesComponent]; }, exports: function () { return [CommandTemplatesComponent]; } }); })();
