// @c8y/ngx-components/esm2015/device-shell/command-templates/command-templates.component.js
import { Component, EventEmitter } from '@angular/core';
import { BsModalRef } from 'ngx-bootstrap/modal';
/* https://stackoverflow.com/a/65290658 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/modal';
import * as ɵngcc2 from './command-templates.directive';
export function rootScopeFactory($injector) {
    return $injector.get('$rootScope').$new();
}
export class CommandTemplatesComponent {
    constructor(modalRef) {
        this.modalRef = modalRef;
        this.onTemplateSelected = new EventEmitter();
    }
}
CommandTemplatesComponent.ɵfac = function CommandTemplatesComponent_Factory(t) { return new (t || CommandTemplatesComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BsModalRef)); };
CommandTemplatesComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: CommandTemplatesComponent, selectors: [["c8y-command-templates"]], features: [ɵngcc0.ɵɵProvidersFeature([
            {
                deps: ['$injector'],
                provide: '$scope',
                useFactory: rootScopeFactory
            }
        ])], decls: 1, vars: 0, consts: [[3, "close", "dismiss"]], template: function CommandTemplatesComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-command-templates-directive", 0);
        ɵngcc0.ɵɵlistener("close", function CommandTemplatesComponent_Template_c8y_command_templates_directive_close_0_listener($event) { ctx.modalRef.hide(); return ctx.onTemplateSelected.next($event); })("dismiss", function CommandTemplatesComponent_Template_c8y_command_templates_directive_dismiss_0_listener() { return ctx.modalRef.hide(); });
        ɵngcc0.ɵɵelementEnd();
    } }, directives: [ɵngcc2.CommandTemplates], encapsulation: 2 });
CommandTemplatesComponent.ctorParameters = () => [
    { type: BsModalRef }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CommandTemplatesComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-command-templates',
                template: `<c8y-command-templates-directive
    (close)="modalRef.hide(); onTemplateSelected.next($event)"
    (dismiss)="modalRef.hide()"
  ></c8y-command-templates-directive>`,
                providers: [
                    {
                        deps: ['$injector'],
                        provide: '$scope',
                        useFactory: rootScopeFactory
                    }
                ]
            }]
    }], function () { return [{ type: ɵngcc1.BsModalRef }]; }, null); })();
