// @c8y/ngx-components/esm2015/datapoint-selector/datapoint-selection-list/datapoint-selection-list.component.js
import { moveItemInArray } from '@angular/cdk/drag-drop';
import { Component, forwardRef, Input, Optional, Output } from '@angular/core';
import { FormBuilder, NG_VALIDATORS, NG_VALUE_ACCESSOR } from '@angular/forms';
import { C8yValidators } from '@c8y/ngx-components';
import { WidgetConfigComponent } from '@c8y/ngx-components/context-dashboard';
import { from } from 'rxjs';
import { map, shareReplay, take, tap } from 'rxjs/operators';
import { DatapointLibraryService } from '../datapoint-library.service';
import { DatapointSelectorService } from '../datapoint-selector.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../datapoint-selector.service';
import * as ɵngcc2 from '../datapoint-library.service';
import * as ɵngcc3 from '@angular/forms';
import * as ɵngcc4 from '@c8y/ngx-components/context-dashboard';
import * as ɵngcc5 from '@c8y/ngx-components';
import * as ɵngcc6 from '@angular/cdk/drag-drop';
import * as ɵngcc7 from '@angular/common';
import * as ɵngcc8 from '../datapoint-selector-list-item/datapoint-selector-list-item.component';

function DatapointSelectionListComponent_div_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 10);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(1, " At least {{ minActive }} active data points must be selected. ");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("translateParams", ctx_r0.formArray.errors == null ? null : ctx_r0.formArray.errors.minActiveCount);
} }
function DatapointSelectionListComponent_div_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 10);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(1, " At maximum {{ maxActive }} active data points are allowed to be selected. ");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("translateParams", ctx_r1.formArray.errors == null ? null : ctx_r1.formArray.errors.maxActiveCount);
} }
function DatapointSelectionListComponent_div_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 11);
    ɵngcc0.ɵɵelement(1, "h1", 12);
    ɵngcc0.ɵɵelementStart(2, "div");
    ɵngcc0.ɵɵelementStart(3, "p");
    ɵngcc0.ɵɵelementStart(4, "strong", 9);
    ɵngcc0.ɵɵtext(5, "No data points to display.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "small", 9);
    ɵngcc0.ɵɵtext(7, "Add your first data point.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function DatapointSelectionListComponent_div_8_Template(rf, ctx) { if (rf & 1) {
    const _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 13);
    ɵngcc0.ɵɵelementStart(1, "c8y-datapoint-selector-list-item", 14);
    ɵngcc0.ɵɵlistener("removed", function DatapointSelectionListComponent_div_8_Template_c8y_datapoint_selector_list_item_removed_1_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r7); const index_r5 = restoredCtx.index; const ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.onItemRemoved(index_r5); });
    ɵngcc0.ɵɵelementStart(2, "c8y-li-drag-handle", 15);
    ɵngcc0.ɵɵelement(3, "i", 16);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const dpForm_r4 = ctx.$implicit;
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("formGroup", dpForm_r4);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("defaultFormOptions", ctx_r3.defaultFormOptions)("activeToggleDisabled", ctx_r3.maxActiveCountReached)("showActiveToggle", true)("showAddRemoveButton", false)("showOptions", true)("editable", true)("colorPickerDisabled", false)("actions", ctx_r3.actions)("optionToRemove", true)("datapointLibraryEntries", ctx_r3.datapointLibraryEntries)("hasUnlinkTemplateOption", true);
} }
const _c0 = [[["", 8, "alert"]]];
const _c1 = [".alert"];
export class DatapointSelectionListComponent {
    constructor(datapointSelector, datapointLibrary, formBuilder, widgetComponent) {
        this.datapointSelector = datapointSelector;
        this.datapointLibrary = datapointLibrary;
        this.formBuilder = formBuilder;
        this.widgetComponent = widgetComponent;
        this.actions = [];
        this.allowDragAndDrop = true;
        this.config = {};
        this.defaultFormOptions = {};
        this.minActiveCount = 1;
        this.resolveContext = true;
        this.maxActiveCountReached = false;
        this.usedValidators = {};
        this.formArray = this.formBuilder.array([]);
        this.isValid = this.formArray.statusChanges.pipe(map(status => status === 'VALID'));
        this.datapointLibraryEntries = from(this.datapointLibrary.getFirstDatapointLibraryPage()).pipe(shareReplay());
        this.change = this.formArray.valueChanges.pipe(map(res => this.transformValue(res)));
    }
    ngOnChanges(changes) {
        if (!changes.maxActiveCount && !changes.minActiveCount) {
            return;
        }
        if (changes.maxActiveCount) {
            this.usedValidators.maxActiveCount = C8yValidators.maxActiveCount(this.maxActiveCount);
        }
        if (changes.minActiveCount) {
            this.usedValidators.minActiveCount = C8yValidators.minActiveCount(this.minActiveCount);
        }
        const validators = Object.values(this.usedValidators);
        this.formArray.setValidators(validators);
    }
    registerOnTouched(fn) {
        this.formArray.valueChanges.pipe(take(1)).subscribe(fn);
    }
    validate(control) {
        return this.formArray.valid ? null : { formInvalid: {} };
    }
    ngOnInit() {
        var _a;
        const context = (_a = this.widgetComponent) === null || _a === void 0 ? void 0 : _a.context;
        if ((context === null || context === void 0 ? void 0 : context.id) && this.resolveContext) {
            const { name, id, c8y_IsDevice } = context;
            this.config.contextAsset = { name, id, c8y_IsDevice };
        }
    }
    writeValue(obj) {
        this.formArray.clear();
        if (obj === null || obj === void 0 ? void 0 : obj.length) {
            obj.forEach(val => {
                const formgroup = this.formBuilder.group({ details: [] });
                formgroup.patchValue({ details: val });
                this.formArray.push(formgroup);
            });
        }
        this.calculateMaxActiveCount();
    }
    registerOnChange(fn) {
        this.formArray.valueChanges
            .pipe(map(res => this.transformValue(res)), 
        // check maxActiveCount
        tap(res => {
            this.calculateMaxActiveCount();
        }))
            .subscribe(fn);
    }
    add() {
        var _a, _b, _c;
        const allowChangingContext = !((_a = this.widgetComponent) === null || _a === void 0 ? void 0 : _a.isDeviceTypeDashboard) && ((_b = this.config) === null || _b === void 0 ? void 0 : _b.allowChangingContext) !== false;
        this.datapointSelector
            .selectDataPoints(Object.assign(Object.assign({}, (this.config || {})), { selectedDatapoints: this.transformValue(this.formArray.value), defaultActiveState: true, allowChangingContext, allowSearch: !((_c = this.config) === null || _c === void 0 ? void 0 : _c.contextAsset) }))
            .then(result => {
            this.writeValue(result);
        }, error => {
            // nothing to do, modal was closed
        });
    }
    onItemRemoved(index) {
        this.formArray.removeAt(index);
    }
    drop(event) {
        const currentSorting = this.formArray.value;
        moveItemInArray(currentSorting, event.previousIndex, event.currentIndex);
        this.formArray.setValue(currentSorting);
    }
    transformValue(formArrayValue) {
        if (!formArrayValue) {
            return [];
        }
        return formArrayValue.map(tmp => Object.assign({}, ...Object.values(tmp)));
    }
    calculateMaxActiveCount() {
        if (this.maxActiveCount) {
            const currentlyActive = this.formArray.value.filter(tmp => { var _a; return (_a = tmp.details) === null || _a === void 0 ? void 0 : _a.__active; }).length;
            this.maxActiveCountReached = currentlyActive >= this.maxActiveCount;
        }
        this.maxActiveCountReached = false;
    }
}
DatapointSelectionListComponent.ɵfac = function DatapointSelectionListComponent_Factory(t) { return new (t || DatapointSelectionListComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.DatapointSelectorService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.DatapointLibraryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.FormBuilder), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.WidgetConfigComponent, 8)); };
DatapointSelectionListComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: DatapointSelectionListComponent, selectors: [["c8y-datapoint-selection-list"]], inputs: { actions: "actions", allowDragAndDrop: "allowDragAndDrop", config: "config", defaultFormOptions: "defaultFormOptions", minActiveCount: "minActiveCount", resolveContext: "resolveContext", maxActiveCount: "maxActiveCount" }, outputs: { isValid: "isValid", change: "change" }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: NG_VALUE_ACCESSOR,
                multi: true,
                useExisting: forwardRef(() => DatapointSelectionListComponent)
            },
            {
                provide: NG_VALIDATORS,
                useExisting: forwardRef(() => DatapointSelectionListComponent),
                multi: true
            }
        ]), ɵngcc0.ɵɵNgOnChangesFeature], ngContentSelectors: _c1, decls: 15, vars: 8, consts: [[1, "card-header", "separator", "sticky-top", "bg-component"], ["translate", "", 1, "card-title", "h4"], ["cdkDropList", "", 1, "flex-grow", "ff-scroll-fix", "cdk-droplist", 3, "cdkDropListDisabled", "cdkDropListDropped"], ["class", "alert alert-info", "role", "alert", "ngNonBindable", "", "translate", "", 3, "translateParams", 4, "ngIf"], ["class", "c8y-empty-state", 4, "ngIf"], [3, "formGroup", 4, "ngFor", "ngForOf"], [1, "card-footer", "bg-component", "sticky-bottom", "separator"], [1, "btn", "btn-default", "btn-sm", 3, "title", "click"], ["c8yIcon", "plus-circle"], ["translate", ""], ["role", "alert", "translate", "", 1, "alert", "alert-info", 3, "translateParams"], [1, "c8y-empty-state"], [1, "c8y-icon", "c8y-icon-data-points", "c8y-icon-duocolor"], [3, "formGroup"], ["formControlName", "details", "cdkDrag", "", 1, "d-block", 3, "defaultFormOptions", "activeToggleDisabled", "showActiveToggle", "showAddRemoveButton", "showOptions", "editable", "colorPickerDisabled", "actions", "optionToRemove", "datapointLibraryEntries", "hasUnlinkTemplateOption", "removed"], ["cdkDragHandle", ""], ["c8yIcon", "drag-reorder"]], template: function DatapointSelectionListComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c0);
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "span", 1);
        ɵngcc0.ɵɵtext(2, "Data points");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "c8y-list-group", 2);
        ɵngcc0.ɵɵlistener("cdkDropListDropped", function DatapointSelectionListComponent_Template_c8y_list_group_cdkDropListDropped_3_listener($event) { return ctx.drop($event); });
        ɵngcc0.ɵɵtemplate(4, DatapointSelectionListComponent_div_4_Template, 2, 1, "div", 3);
        ɵngcc0.ɵɵtemplate(5, DatapointSelectionListComponent_div_5_Template, 2, 1, "div", 3);
        ɵngcc0.ɵɵprojection(6);
        ɵngcc0.ɵɵtemplate(7, DatapointSelectionListComponent_div_7_Template, 8, 0, "div", 4);
        ɵngcc0.ɵɵtemplate(8, DatapointSelectionListComponent_div_8_Template, 4, 12, "div", 5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "div", 6);
        ɵngcc0.ɵɵelementStart(10, "button", 7);
        ɵngcc0.ɵɵlistener("click", function DatapointSelectionListComponent_Template_button_click_10_listener() { return ctx.add(); });
        ɵngcc0.ɵɵpipe(11, "translate");
        ɵngcc0.ɵɵelement(12, "i", 8);
        ɵngcc0.ɵɵelementStart(13, "span", 9);
        ɵngcc0.ɵɵtext(14, "Add data point");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("cdkDropListDisabled", !ctx.allowDragAndDrop || (ctx.formArray.controls == null ? null : ctx.formArray.controls.length) < 2);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.formArray.errors == null ? null : ctx.formArray.errors.minActiveCount);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.formArray.errors == null ? null : ctx.formArray.errors.maxActiveCount);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", !(ctx.formArray.controls == null ? null : ctx.formArray.controls.length));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.formArray.controls);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(11, 6, "Add data point"));
    } }, directives: [ɵngcc5.C8yTranslateDirective, ɵngcc5.ListGroupComponent, ɵngcc6.CdkDropList, ɵngcc7.NgIf, ɵngcc7.NgForOf, ɵngcc5.IconDirective, ɵngcc3.NgControlStatusGroup, ɵngcc3.FormGroupDirective, ɵngcc8.DatapointSelectorListItemComponent, ɵngcc3.NgControlStatus, ɵngcc3.FormControlName, ɵngcc6.CdkDrag, ɵngcc5.ListItemDragHandleComponent, ɵngcc6.CdkDragHandle], pipes: [ɵngcc5.C8yTranslatePipe], encapsulation: 2 });
DatapointSelectionListComponent.ctorParameters = () => [
    { type: DatapointSelectorService },
    { type: DatapointLibraryService },
    { type: FormBuilder },
    { type: WidgetConfigComponent, decorators: [{ type: Optional }] }
];
DatapointSelectionListComponent.propDecorators = {
    actions: [{ type: Input }],
    allowDragAndDrop: [{ type: Input }],
    config: [{ type: Input }],
    defaultFormOptions: [{ type: Input }],
    maxActiveCount: [{ type: Input }],
    minActiveCount: [{ type: Input }],
    resolveContext: [{ type: Input }],
    isValid: [{ type: Output }],
    change: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DatapointSelectionListComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-datapoint-selection-list',
                template: "<div class=\"card-header separator sticky-top bg-component\">\n  <span class=\"card-title h4\" translate>Data points</span>\n</div>\n\n<c8y-list-group\n  class=\"flex-grow ff-scroll-fix cdk-droplist\"\n  cdkDropList\n  (cdkDropListDropped)=\"drop($event)\"\n  [cdkDropListDisabled]=\"!allowDragAndDrop || formArray.controls?.length < 2\"\n>\n  <div\n    class=\"alert alert-info\"\n    role=\"alert\"\n    ngNonBindable\n    *ngIf=\"formArray.errors?.minActiveCount\"\n    translate\n    [translateParams]=\"formArray.errors?.minActiveCount\"\n  >\n    At least {{ minActive }} active data points must be selected.\n  </div>\n\n  <div\n    class=\"alert alert-info\"\n    role=\"alert\"\n    ngNonBindable\n    *ngIf=\"formArray.errors?.maxActiveCount\"\n    translate\n    [translateParams]=\"formArray.errors?.maxActiveCount\"\n  >\n    At maximum {{ maxActive }} active data points are allowed to be selected.\n  </div>\n\n  <ng-content select=\".alert\"></ng-content>\n\n  <div class=\"c8y-empty-state\" *ngIf=\"!formArray.controls?.length\">\n    <h1 class=\"c8y-icon c8y-icon-data-points c8y-icon-duocolor\"></h1>\n    <div>\n      <p>\n        <strong translate>No data points to display.</strong>\n      </p>\n      <small translate>Add your first data point.</small>\n    </div>\n  </div>\n  <div [formGroup]=\"dpForm\" *ngFor=\"let dpForm of formArray.controls; let index = index\">\n    <c8y-datapoint-selector-list-item\n      class=\"d-block\"\n      [defaultFormOptions]=\"defaultFormOptions\"\n      [activeToggleDisabled]=\"maxActiveCountReached\"\n      [showActiveToggle]=\"true\"\n      [showAddRemoveButton]=\"false\"\n      [showOptions]=\"true\"\n      [editable]=\"true\"\n      [colorPickerDisabled]=\"false\"\n      [actions]=\"actions\"\n      [optionToRemove]=\"true\"\n      [datapointLibraryEntries]=\"datapointLibraryEntries\"\n      [hasUnlinkTemplateOption]=\"true\"\n      formControlName=\"details\"\n      (removed)=\"onItemRemoved(index)\"\n      cdkDrag\n    >\n      <c8y-li-drag-handle cdkDragHandle>\n        <i c8yIcon=\"drag-reorder\"></i>\n      </c8y-li-drag-handle>\n    </c8y-datapoint-selector-list-item>\n  </div>\n</c8y-list-group>\n\n<div class=\"card-footer bg-component sticky-bottom separator\">\n  <button [title]=\"'Add data point' | translate\" class=\"btn btn-default btn-sm\" (click)=\"add()\">\n    <i c8yIcon=\"plus-circle\"></i>\n    <span translate>Add data point</span>\n  </button>\n</div>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        multi: true,
                        useExisting: forwardRef(() => DatapointSelectionListComponent)
                    },
                    {
                        provide: NG_VALIDATORS,
                        useExisting: forwardRef(() => DatapointSelectionListComponent),
                        multi: true
                    }
                ]
            }]
    }], function () { return [{ type: ɵngcc1.DatapointSelectorService }, { type: ɵngcc2.DatapointLibraryService }, { type: ɵngcc3.FormBuilder }, { type: ɵngcc4.WidgetConfigComponent, decorators: [{
                type: Optional
            }] }]; }, { actions: [{
            type: Input
        }], allowDragAndDrop: [{
            type: Input
        }], config: [{
            type: Input
        }], defaultFormOptions: [{
            type: Input
        }], minActiveCount: [{
            type: Input
        }], resolveContext: [{
            type: Input
        }], isValid: [{
            type: Output
        }], change: [{
            type: Output
        }], maxActiveCount: [{
            type: Input
        }] }); })();
