// @c8y/ngx-components/esm2015/register-device/extensible/base-extensible-device-registration.service.js
import { __awaiter } from "tslib";
import { ApplicationType } from '@c8y/client';
import { gettext } from '@c8y/ngx-components';
import { flatMap } from 'lodash-es';
export class BaseExtensibleDeviceRegistrationService {
    constructor(applicationService, stateService) {
        this.applicationService = applicationService;
        this.stateService = stateService;
    }
    getMetadataUrl(contextPath) {
        return `${this.getServiceBaseUrl(contextPath)}/metadata`;
    }
    getServiceBaseUrl(contextPath) {
        if (!contextPath || contextPath.length === 0) {
            throw new Error(gettext('The URL of the microservice is misconfigured. A context path is missing.'));
        }
        return `/service/${contextPath}/deviceRegistration`;
    }
    getDeviceRegistrationProviders(bulk) {
        return __awaiter(this, void 0, void 0, function* () {
            let providers = [];
            const microservices = yield this.listMicroservicesByUser();
            if (microservices && microservices.length > 0) {
                providers = this.getProvidersFromManifestExtensions(microservices, bulk).concat(this.getProvidersFromExtensions(microservices, bulk));
            }
            return providers;
        });
    }
    listMicroservicesByUser() {
        var _a, _b, _c;
        return __awaiter(this, void 0, void 0, function* () {
            const { data } = yield this.applicationService.listByUser((_c = (_b = (_a = this.stateService) === null || _a === void 0 ? void 0 : _a.currentUser) === null || _b === void 0 ? void 0 : _b.value) === null || _c === void 0 ? void 0 : _c.userName, {
                dropOverwrittenApps: true,
                noPaging: true,
                pageSize: 100,
                withTotalPages: true
            });
            return data === null || data === void 0 ? void 0 : data.filter(({ type }) => type === ApplicationType.MICROSERVICE);
        });
    }
    getProvidersFromManifestExtensions(applications, bulk) {
        return flatMap(applications, application => {
            var _a, _b;
            return (_b = (_a = application === null || application === void 0 ? void 0 : application.manifest) === null || _a === void 0 ? void 0 : _a.extensions) === null || _b === void 0 ? void 0 : _b.map(extension => {
                return Object.assign(Object.assign({}, extension), { contextPath: application.contextPath });
            });
        }).filter(provider => this.appSupportingExtensibleDeviceRegistrationForm(provider, bulk));
    }
    getProvidersFromExtensions(applications, bulk) {
        return flatMap(applications, application => {
            var _a;
            return (_a = application.extensions) === null || _a === void 0 ? void 0 : _a.map(extension => {
                return Object.assign(Object.assign({}, extension), { contextPath: application.contextPath });
            });
        }).filter(provider => this.appSupportingExtensibleDeviceRegistrationForm(provider, bulk));
    }
    appSupportingExtensibleDeviceRegistrationForm(extension, bulk) {
        if (bulk) {
            return (extension === null || extension === void 0 ? void 0 : extension.type) === 'extensibleBulkDeviceRegistration';
        }
        else {
            return (extension === null || extension === void 0 ? void 0 : extension.type) === 'extensibleDeviceRegistration';
        }
    }
}
