// @c8y/ngx-components/esm2015/core/core.module.js
import { NgModule } from '@angular/core';
import { ActionBarModule } from './action-bar/action-bar.module';
import { ActionModule } from './action/action.module';
import { AlertModule } from './alert/alert.module';
import { AuditLogModule } from './audit-log/audit-log.module';
import { AuthenticationModule } from './authentication/authentication.module';
import { BootstrapModule } from './bootstrap/bootstrap.module';
import { BreadcrumbModule } from './breadcrumb/breadcrumb.module';
import { CommonModule } from './common/common.module';
import { DashboardModule } from './dashboard/dashboard.module';
import { DataGridModule } from './data-grid/data-grid.module';
import { DeviceStatusModule } from './device-status/device-status.module';
import { DocsModule } from './docs/docs.module';
import { DropAreaModule } from './drop-area/drop-area.module';
import { DynamicComponentModule } from './dynamic-component/dynamic-component.module';
import { FilePickerModule } from './file-picker/file-picker.module';
import { DatePickerModule } from './date-picker/date-picker.module';
import { FormsModule as c8yFormsModule } from './forms/forms.module';
import { HeaderModule } from './header/header.module';
import { HelpModule } from './help/help.module';
import { ListGroupModule } from './list-group/list-group.module';
import { LoginModule } from './login/login.module';
import { ModalModule } from './modal/modal.module';
import { NavigatorModule } from './navigator/navigator.module';
import { ProductExperienceModule } from './product-experience/product-experience.module';
import { RealtimeModule } from './realtime/realtime.module';
import { SearchModule } from './search/search.module';
import { SelectModalModule } from './select-modal/select-modal.module';
import { SelectModule } from './select/select.module';
import { StepperModule as c8yStepperModule } from './stepper/stepper.module';
import { TabsModule } from './tabs/tabs.module';
import { UserModule } from './user/user.module';
import { PluginsModule } from './plugins/plugins.module';
import { RangeDisplayModule } from './range-display/range-display.module';
import { ListDisplaySwitchModule } from './list-display-switch/list-display-switch.module';
import { DynamicFormsModule } from './dynamic-forms/dynamic-forms.module';
import { WizardModule } from './wizard/wizard.module';
import { PropertiesListModule } from './properties-list/properties-list.module';
import { DateTimePickerModule } from './date-time-picker/date-time-picker.module';
import { AssetTypesModule } from './asset-types/asset-types.module';
import * as ɵngcc0 from '@angular/core';
export class CoreModule {
    static forRoot() {
        return {
            ngModule: CoreModule,
            providers: [...CommonModule.providers(), ...LoginModule.providers()]
        };
    }
}
CoreModule.ɵfac = function CoreModule_Factory(t) { return new (t || CoreModule)(); };
CoreModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: CoreModule });
CoreModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [ActionBarModule,
        ActionModule,
        AlertModule,
        AuditLogModule,
        BootstrapModule,
        BreadcrumbModule,
        CommonModule,
        DataGridModule,
        DropAreaModule,
        HeaderModule,
        LoginModule,
        ModalModule,
        NavigatorModule,
        AuthenticationModule,
        SearchModule,
        SelectModule,
        TabsModule,
        UserModule,
        c8yFormsModule,
        DeviceStatusModule,
        DocsModule,
        c8yStepperModule,
        DynamicComponentModule,
        DashboardModule,
        ListGroupModule,
        SelectModalModule,
        FilePickerModule,
        DatePickerModule,
        ProductExperienceModule,
        HelpModule,
        RealtimeModule,
        RangeDisplayModule,
        ListDisplaySwitchModule,
        WizardModule,
        PluginsModule,
        DynamicFormsModule,
        PropertiesListModule,
        AssetTypesModule,
        DateTimePickerModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CoreModule, [{
        type: NgModule,
        args: [{
                exports: [
                    ActionBarModule,
                    ActionModule,
                    AlertModule,
                    AuditLogModule,
                    BootstrapModule,
                    BreadcrumbModule,
                    CommonModule,
                    DataGridModule,
                    DropAreaModule,
                    HeaderModule,
                    LoginModule,
                    ModalModule,
                    NavigatorModule,
                    AuthenticationModule,
                    SearchModule,
                    SelectModule,
                    TabsModule,
                    UserModule,
                    c8yFormsModule,
                    DeviceStatusModule,
                    DocsModule,
                    c8yStepperModule,
                    DynamicComponentModule,
                    DashboardModule,
                    ListGroupModule,
                    SelectModalModule,
                    FilePickerModule,
                    DatePickerModule,
                    ProductExperienceModule,
                    HelpModule,
                    RealtimeModule,
                    RangeDisplayModule,
                    ListDisplaySwitchModule,
                    WizardModule,
                    PluginsModule,
                    DynamicFormsModule,
                    PropertiesListModule,
                    AssetTypesModule,
                    DateTimePickerModule
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(CoreModule, { exports: function () { return [ActionBarModule,
        ActionModule,
        AlertModule,
        AuditLogModule,
        BootstrapModule,
        BreadcrumbModule,
        CommonModule,
        DataGridModule,
        DropAreaModule,
        HeaderModule,
        LoginModule,
        ModalModule,
        NavigatorModule,
        AuthenticationModule,
        SearchModule,
        SelectModule,
        TabsModule,
        UserModule,
        c8yFormsModule,
        DeviceStatusModule,
        DocsModule,
        c8yStepperModule,
        DynamicComponentModule,
        DashboardModule,
        ListGroupModule,
        SelectModalModule,
        FilePickerModule,
        DatePickerModule,
        ProductExperienceModule,
        HelpModule,
        RealtimeModule,
        RangeDisplayModule,
        ListDisplaySwitchModule,
        WizardModule,
        PluginsModule,
        DynamicFormsModule,
        PropertiesListModule,
        AssetTypesModule,
        DateTimePickerModule]; } }); })();
