// @c8y/ngx-components/esm2015/tenants/tenant-list/tenant-list.component.js
import { __awaiter } from "tslib";
import { Component } from '@angular/core';
import { Location } from '@angular/common';
import { TranslateService } from '@ngx-translate/core';
import { BehaviorSubject, from } from 'rxjs';
import { expand, reduce, shareReplay, takeWhile } from 'rxjs/operators';
import { saveAs } from 'file-saver';
import { TenantService, TenantStatus, UserService } from '@c8y/client';
import { AlertService, AppStateService, gettext, ModalService, Status, TenantUiService, PasswordService } from '@c8y/ngx-components';
import { StatusFilteringFormRendererComponent } from './status.filtering-form-renderer.component';
import { CreationTimeFilteringFormRendererComponent } from './creation-time.filtering-form-renderer.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from '@ngx-translate/core';
import * as ɵngcc3 from '@c8y/client';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from '@angular/router';

function TenantListComponent_c8y_action_bar_item_3_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-action-bar-item", 1);
    ɵngcc0.ɵɵelementStart(1, "button", 2);
    ɵngcc0.ɵɵlistener("click", function TenantListComponent_c8y_action_bar_item_3_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r9); const ctx_r8 = ɵngcc0.ɵɵnextContext(); return ctx_r8.downloadNewsletterEmails(); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 13);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("placement", "right");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 3, "Downloads the list of emails of users subscribed for newsletter on the current tenant and its subtenants."));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 5, "Email addresses"), " ");
} }
function TenantListComponent_ng_container_15_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "c8y-loading");
    ɵngcc0.ɵɵelementContainerEnd();
} }
function TenantListComponent_ng_container_17_Template(rf, ctx) { if (rf & 1) {
    const _r11 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "div", 14);
    ɵngcc0.ɵɵelement(2, "h1", 15);
    ɵngcc0.ɵɵelementStart(3, "h3", 16);
    ɵngcc0.ɵɵtext(4, "There are no tenants defined.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "p", 16);
    ɵngcc0.ɵɵtext(6, "Create the first tenant.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "div");
    ɵngcc0.ɵɵelementStart(8, "button", 17);
    ɵngcc0.ɵɵlistener("click", function TenantListComponent_ng_container_17_Template_button_click_8_listener() { ɵngcc0.ɵɵrestoreView(_r11); const ctx_r10 = ɵngcc0.ɵɵnextContext(); return ctx_r10.createTenant(); });
    ɵngcc0.ɵɵpipe(9, "translate");
    ɵngcc0.ɵɵtext(10);
    ɵngcc0.ɵɵpipe(11, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(12, "p", 18);
    ɵngcc0.ɵɵelementStart(13, "small", 16);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(14, " Find out more in the ");
    ɵngcc0.ɵɵelementStart(15, "a", 19);
    ɵngcc0.ɵɵtext(16, " User guide`KEEP_ORIGINAL` ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(17, ". ");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(8);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(9, 2, "Create tenant"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(11, 4, "Create tenant"), " ");
} }
function TenantListComponent_ng_container_19_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "div", 14);
    ɵngcc0.ɵɵelement(2, "h1", 20);
    ɵngcc0.ɵɵelementStart(3, "h3", 16);
    ɵngcc0.ɵɵtext(4, "No tenants to display.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "p", 16);
    ɵngcc0.ɵɵtext(6, "Refine your search terms or check your spelling.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} }
const _c0 = function (a1) { return ["/tenants", a1]; };
function TenantListComponent_ng_container_22_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "span", 21);
    ɵngcc0.ɵɵelementStart(2, "a", 22);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const context_r12 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", context_r12.value);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("routerLink", ɵngcc0.ɵɵpureFunction1(3, _c0, context_r12.item.id));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(context_r12.value);
} }
function TenantListComponent_ng_container_24_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "span", 21);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const context_r13 = ctx.$implicit;
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", context_r13.value || ctx_r5.currentTenant.name);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", context_r13.value || ctx_r5.currentTenant.name, " ");
} }
function TenantListComponent_ng_container_26_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "span", 21);
    ɵngcc0.ɵɵpipe(2, "c8yDate");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "c8yDate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const context_r14 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 2, context_r14.value));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 4, context_r14.value), " ");
} }
function TenantListComponent_ng_container_28_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 21);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 24);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 1, "Active`tenant`"));
} }
function TenantListComponent_ng_container_28_span_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 21);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 25);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 1, "Suspended`tenant`"));
} }
function TenantListComponent_ng_container_28_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, TenantListComponent_ng_container_28_span_1_Template, 3, 3, "span", 23);
    ɵngcc0.ɵɵtemplate(2, TenantListComponent_ng_container_28_span_2_Template, 3, 3, "span", 23);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const context_r15 = ctx.$implicit;
    const ctx_r7 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", context_r15.item.status === ctx_r7.TenantStatus.ACTIVE);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", context_r15.item.status === ctx_r7.TenantStatus.SUSPENDED);
} }
export class TenantListComponent {
    constructor(appState, alertService, modalService, translateService, tenantService, tenantUiService, location, passwordService, userService) {
        this.appState = appState;
        this.alertService = alertService;
        this.modalService = modalService;
        this.translateService = translateService;
        this.tenantService = tenantService;
        this.tenantUiService = tenantUiService;
        this.location = location;
        this.passwordService = passwordService;
        this.userService = userService;
        this.tenants$ = new BehaviorSubject(undefined);
        this.TOP_TENANT_NAME = 'management';
        this.title = null;
        this.loadMoreItemsLabel = gettext('Load more tenants');
        this.loadingItemsLabel = gettext('Loading tenants…');
        this.displayOptions = {
            bordered: false,
            striped: true,
            filter: true,
            gridHeader: true
        };
        this.columns = this.getColumns();
        this.pagination = this.getPagination();
        this.showSearch = true;
        this.actionControls = this.getActionControls();
        this.TenantStatus = TenantStatus;
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            this.currentTenant = this.appState.currentTenant.value;
            this.isManagementTenant = yield this.tenantUiService.isManagementTenant();
            this.loadTenants();
        });
    }
    loadTenants() {
        this.tenants$.next(undefined);
        from(this.tenantService.list({ pageSize: 2000, withTotalPages: true, withApps: false }))
            .pipe(expand(resultList => resultList.paging.nextPage !== null && resultList.paging.next()), takeWhile(resultList => resultList.paging.nextPage !== null, true), reduce((tenants, resultList) => [
            ...tenants,
            ...resultList.data
        ], []), shareReplay(1))
            .subscribe(tenants => this.tenants$.next(tenants));
    }
    getColumns() {
        return [
            {
                name: 'company',
                header: gettext('Tenant'),
                path: 'company',
                filterable: true,
                sortable: true,
                sortOrder: 'asc'
            },
            {
                name: 'id',
                header: gettext('ID'),
                path: 'id',
                filterable: true,
                sortable: true
            },
            {
                name: 'domain',
                header: gettext('Domain'),
                path: 'domain',
                filterable: true,
                sortable: true
            },
            {
                name: 'parent',
                header: gettext('Parent tenant'),
                path: 'parent',
                filterable: true,
                sortable: true
            },
            {
                name: 'contactName',
                header: gettext('Contact name'),
                path: 'contactName',
                filterable: true,
                sortable: true
            },
            {
                name: 'creationTime',
                header: gettext('Created'),
                path: 'creationTime',
                filterable: true,
                filteringFormRendererComponent: CreationTimeFilteringFormRendererComponent,
                sortable: true
            },
            {
                name: 'externalReference',
                header: gettext('External reference'),
                path: 'customProperties.externalReference',
                filterable: true,
                sortable: true
            },
            {
                name: 'status',
                header: gettext('Status'),
                path: 'status',
                filterable: true,
                filteringFormRendererComponent: StatusFilteringFormRendererComponent,
                sortable: true,
                resizable: false
            }
        ];
    }
    getPagination() {
        return {
            pageSize: 10,
            currentPage: 1
        };
    }
    getActionControls() {
        return [
            {
                type: "EDIT" /* Edit */,
                text: gettext('Edit`tenant`'),
                callback: tenant => this.goToDetails(tenant)
            },
            {
                type: 'activateTenantAction',
                icon: 'power-off',
                text: gettext('Activate`tenant`'),
                callback: (tenant) => this.activateTenant(tenant),
                showIf: (tenant) => this.isSuspended(tenant)
            },
            {
                type: 'suspendTenantAction',
                icon: 'power-off',
                text: gettext('Suspend`tenant`'),
                callback: (tenant) => this.suspendTenant(tenant),
                showIf: (tenant) => this.isActive(tenant)
            },
            {
                type: "DELETE" /* Delete */,
                text: gettext('Delete`tenant`'),
                callback: tenant => this.delete(tenant),
                showIf: () => this.isManagementTenant
            }
        ];
    }
    createTenant() {
        this.location.go('/tenants/new');
    }
    goToDetails(tenant) {
        this.location.go(`/tenants/${tenant.id}`);
    }
    activateTenant(tenant) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const { data: savedTenant } = yield this.tenantService.update({
                    id: tenant.id,
                    status: TenantStatus.ACTIVE
                });
                Object.assign(tenant, savedTenant);
                this.alertService.success(gettext('Tenant activated.'));
            }
            catch (ex) {
                if (ex) {
                    this.alertService.addServerFailure(ex);
                }
            }
        });
    }
    suspendTenant(tenant) {
        return __awaiter(this, void 0, void 0, function* () {
            const title = gettext('Suspend tenant');
            const confirmationText = gettext('You are about to suspend tenant "{{ company }}" (ID "{{ id }}").');
            const proceed = gettext('Do you want to proceed?');
            const body = [
                this.translateService.instant(confirmationText, {
                    company: tenant.company,
                    id: tenant.id
                }),
                this.translateService.instant(proceed)
            ].join(' ');
            const labels = {
                ok: gettext('Suspend`tenant`')
            };
            try {
                yield this.modalService.confirm(title, body, Status.DANGER, labels);
                const confirmed = yield this.passwordService.confirmPassword().toPromise();
                if (confirmed === true) {
                    const { data: savedTenant } = yield this.tenantService.update({
                        id: tenant.id,
                        status: TenantStatus.SUSPENDED
                    });
                    Object.assign(tenant, savedTenant);
                    this.alertService.success(gettext('Tenant suspended.'));
                }
            }
            catch (ex) {
                if (ex) {
                    this.alertService.addServerFailure(ex);
                }
            }
        });
    }
    delete(tenant) {
        return __awaiter(this, void 0, void 0, function* () {
            const title = gettext('Delete tenant');
            const confirmationText = gettext('You are about to delete tenant "{{ company }}" (ID "{{ id }}").');
            const hint = gettext('This operation is irreversible.');
            const proceed = gettext('Do you want to proceed?');
            const body = [
                this.translateService.instant(confirmationText, {
                    company: tenant.company,
                    id: tenant.id
                }),
                this.translateService.instant(hint),
                this.translateService.instant(proceed)
            ].join(' ');
            const labels = {
                ok: gettext('Delete`tenant`')
            };
            try {
                yield this.modalService.confirm(title, body, Status.DANGER, labels);
                const confirmed = yield this.passwordService.confirmPassword().toPromise();
                if (confirmed === true) {
                    yield this.tenantService.delete(tenant);
                    const tenantsWithoutRemovedOne = this.tenants$.value.filter(t => t !== tenant);
                    this.tenants$.next(tenantsWithoutRemovedOne);
                    this.alertService.success(gettext('Tenant is being deleted in the background. This might take a while…'));
                }
            }
            catch (ex) {
                if (ex) {
                    this.alertService.addServerFailure(ex);
                }
            }
        });
    }
    isActive(tenant) {
        return tenant.status === TenantStatus.ACTIVE;
    }
    isSuspended(tenant) {
        return tenant.status === TenantStatus.SUSPENDED;
    }
    downloadNewsletterEmails() {
        return __awaiter(this, void 0, void 0, function* () {
            const { res, data } = yield this.userService.getNewsletterEmails();
            const contentType = res.headers.get('content-type');
            const contentDisposition = res.headers.get('content-disposition');
            const filename = /filename="(.*)"/.exec(contentDisposition)[1];
            const blob = new Blob([data], { type: contentType });
            saveAs(blob, filename);
        });
    }
}
TenantListComponent.ɵfac = function TenantListComponent_Factory(t) { return new (t || TenantListComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AppStateService), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ModalService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.TranslateService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.TenantService), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.TenantUiService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.Location), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.PasswordService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.UserService)); };
TenantListComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: TenantListComponent, selectors: [["c8y-tenant-list"]], decls: 29, vars: 33, consts: [[3, "placement", 4, "ngIf"], [3, "placement"], [1, "btn", "btn-link", 3, "title", "click"], ["c8yIcon", "plus-circle"], [1, "content-fullpage", "d-flex", "d-col", "shadow1"], [3, "title", "loadMoreItemsLabel", "loadingItemsLabel", "displayOptions", "columns", "rows", "pagination", "showSearch", "actionControls", "onReload"], [1, "c8y-empty-state", "j-c-center"], [4, "ngIf"], ["name", "company"], [4, "c8yCellRendererDef"], ["name", "parent"], ["name", "creationTime"], ["name", "status"], ["c8yIcon", "download"], [1, "text-center"], ["c8yIcon", "gears"], ["translate", ""], [1, "btn", "btn-primary", 3, "title", "click"], ["c8y-guide-docs", ""], ["c8y-guide-href", "users-guide/enterprise-edition/#managing-tenants"], ["c8yIcon", "search", 1, "c8y-icon-duocolor"], [3, "title"], [3, "routerLink"], [3, "title", 4, "ngIf"], ["c8yIcon", "check-circle", 1, "text-success"], ["c8yIcon", "ban", 1, "text-danger"]], template: function TenantListComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵpipe(2, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(3, TenantListComponent_c8y_action_bar_item_3_Template, 6, 7, "c8y-action-bar-item", 0);
        ɵngcc0.ɵɵpipe(4, "async");
        ɵngcc0.ɵɵelementStart(5, "c8y-action-bar-item", 1);
        ɵngcc0.ɵɵelementStart(6, "button", 2);
        ɵngcc0.ɵɵlistener("click", function TenantListComponent_Template_button_click_6_listener() { return ctx.createTenant(); });
        ɵngcc0.ɵɵpipe(7, "translate");
        ɵngcc0.ɵɵelement(8, "i", 3);
        ɵngcc0.ɵɵtext(9);
        ɵngcc0.ɵɵpipe(10, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "div", 4);
        ɵngcc0.ɵɵelementStart(12, "c8y-data-grid", 5);
        ɵngcc0.ɵɵlistener("onReload", function TenantListComponent_Template_c8y_data_grid_onReload_12_listener() { return ctx.loadTenants(); });
        ɵngcc0.ɵɵpipe(13, "async");
        ɵngcc0.ɵɵelementStart(14, "div", 6);
        ɵngcc0.ɵɵtemplate(15, TenantListComponent_ng_container_15_Template, 2, 0, "ng-container", 7);
        ɵngcc0.ɵɵpipe(16, "async");
        ɵngcc0.ɵɵtemplate(17, TenantListComponent_ng_container_17_Template, 18, 6, "ng-container", 7);
        ɵngcc0.ɵɵpipe(18, "async");
        ɵngcc0.ɵɵtemplate(19, TenantListComponent_ng_container_19_Template, 7, 0, "ng-container", 7);
        ɵngcc0.ɵɵpipe(20, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(21, "c8y-column", 8);
        ɵngcc0.ɵɵtemplate(22, TenantListComponent_ng_container_22_Template, 4, 5, "ng-container", 9);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(23, "c8y-column", 10);
        ɵngcc0.ɵɵtemplate(24, TenantListComponent_ng_container_24_Template, 3, 2, "ng-container", 9);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(25, "c8y-column", 11);
        ɵngcc0.ɵɵtemplate(26, TenantListComponent_ng_container_26_Template, 5, 6, "ng-container", 9);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(27, "c8y-column", 12);
        ɵngcc0.ɵɵtemplate(28, TenantListComponent_ng_container_28_Template, 3, 2, "ng-container", 9);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        let tmp_15_0;
        let tmp_16_0;
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 17, "Tenants"), "\n");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", !!ɵngcc0.ɵɵpipeBind1(4, 19, ctx.appState.state$).newsletter);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("placement", "right");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(7, 21, "Create tenant"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(10, 23, "Create tenant"), " ");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("title", ctx.title)("loadMoreItemsLabel", ctx.loadMoreItemsLabel)("loadingItemsLabel", ctx.loadingItemsLabel)("displayOptions", ctx.displayOptions)("columns", ctx.columns)("rows", ɵngcc0.ɵɵpipeBind1(13, 25, ctx.tenants$))("pagination", ctx.pagination)("showSearch", ctx.showSearch)("actionControls", ctx.actionControls);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", !ɵngcc0.ɵɵpipeBind1(16, 27, ctx.tenants$));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ((tmp_15_0 = ɵngcc0.ɵɵpipeBind1(18, 29, ctx.tenants$)) == null ? null : tmp_15_0.length) === 0);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ((tmp_16_0 = ɵngcc0.ɵɵpipeBind1(20, 31, ctx.tenants$)) == null ? null : tmp_16_0.length) > 0);
    } }, directives: [ɵngcc1.TitleComponent, ɵngcc4.NgIf, ɵngcc1.ActionBarItemComponent, ɵngcc1.IconDirective, ɵngcc1.DataGridComponent, ɵngcc1.ColumnDirective, ɵngcc1.CellRendererDefDirective, ɵngcc1.LoadingComponent, ɵngcc1.C8yTranslateDirective, ɵngcc1.GuideDocsComponent, ɵngcc1.GuideHrefDirective, ɵngcc5.RouterLinkWithHref], pipes: [ɵngcc1.C8yTranslatePipe, ɵngcc4.AsyncPipe, ɵngcc1.DatePipe], encapsulation: 2 });
TenantListComponent.ctorParameters = () => [
    { type: AppStateService },
    { type: AlertService },
    { type: ModalService },
    { type: TranslateService },
    { type: TenantService },
    { type: TenantUiService },
    { type: Location },
    { type: PasswordService },
    { type: UserService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TenantListComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-tenant-list',
                template: "<c8y-title>\n  {{ 'Tenants' | translate }}\n</c8y-title>\n\n<c8y-action-bar-item *ngIf=\"!!(appState.state$ | async).newsletter\" [placement]=\"'right'\">\n  <button\n    class=\"btn btn-link\"\n    title=\"{{\n      'Downloads the list of emails of users subscribed for newsletter on the current tenant and its subtenants.'\n        | translate\n    }}\"\n    (click)=\"downloadNewsletterEmails()\"\n  >\n    <i c8yIcon=\"download\"></i>\n    {{ 'Email addresses' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button class=\"btn btn-link\" title=\"{{ 'Create tenant' | translate }}\" (click)=\"createTenant()\">\n    <i c8yIcon=\"plus-circle\"></i>\n    {{ 'Create tenant' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<!-- <c8y-help src=\"/users-guide/enterprise-tenant/#managing-tenants\"></c8y-help> -->\n\n<div class=\"content-fullpage d-flex d-col shadow1\">\n  <c8y-data-grid\n    [title]=\"title\"\n    [loadMoreItemsLabel]=\"loadMoreItemsLabel\"\n    [loadingItemsLabel]=\"loadingItemsLabel\"\n    [displayOptions]=\"displayOptions\"\n    [columns]=\"columns\"\n    [rows]=\"tenants$ | async\"\n    [pagination]=\"pagination\"\n    [showSearch]=\"showSearch\"\n    [actionControls]=\"actionControls\"\n    (onReload)=\"loadTenants()\"\n  >\n    <div class=\"c8y-empty-state j-c-center\">\n      <ng-container *ngIf=\"!(tenants$ | async)\">\n        <c8y-loading></c8y-loading>\n      </ng-container>\n\n      <ng-container *ngIf=\"(tenants$ | async)?.length === 0\">\n        <div class=\"text-center\">\n          <h1 c8yIcon=\"gears\"></h1>\n          <h3 translate>There are no tenants defined.</h3>\n          <p translate>Create the first tenant.</p>\n          <div>\n            <button\n              class=\"btn btn-primary\"\n              title=\"{{ 'Create tenant' | translate }}\"\n              (click)=\"createTenant()\"\n            >\n              {{ 'Create tenant' | translate }}\n            </button>\n          </div>\n          <p c8y-guide-docs>\n            <small translate ngNonBindable>\n              Find out more in the\n              <a c8y-guide-href=\"users-guide/enterprise-edition/#managing-tenants\">\n                User guide`KEEP_ORIGINAL` </a\n              >.\n            </small>\n          </p>\n        </div>\n      </ng-container>\n\n      <ng-container *ngIf=\"(tenants$ | async)?.length > 0\">\n        <div class=\"text-center\">\n          <h1 class=\"c8y-icon-duocolor\" c8yIcon=\"search\"></h1>\n          <h3 translate>No tenants to display.</h3>\n          <p translate>Refine your search terms or check your spelling.</p>\n        </div>\n      </ng-container>\n    </div>\n\n    <c8y-column name=\"company\">\n      <ng-container *c8yCellRendererDef=\"let context\">\n        <span title=\"{{ context.value }}\">\n          <a [routerLink]=\"['/tenants', context.item.id]\">{{ context.value }}</a>\n        </span>\n      </ng-container>\n    </c8y-column>\n\n    <c8y-column name=\"parent\">\n      <ng-container *c8yCellRendererDef=\"let context\">\n        <span title=\"{{ context.value || currentTenant.name }}\">\n          {{ context.value || currentTenant.name }}\n        </span>\n      </ng-container>\n    </c8y-column>\n\n    <c8y-column name=\"creationTime\">\n      <ng-container *c8yCellRendererDef=\"let context\">\n        <span title=\"{{ context.value | c8yDate }}\">\n          {{ context.value | c8yDate }}\n        </span>\n      </ng-container>\n    </c8y-column>\n\n    <c8y-column name=\"status\">\n      <ng-container *c8yCellRendererDef=\"let context\">\n        <span\n          *ngIf=\"context.item.status === TenantStatus.ACTIVE\"\n          title=\"{{ 'Active`tenant`' | translate }}\"\n        >\n          <i c8yIcon=\"check-circle\" class=\"text-success\"></i>\n        </span>\n        <span\n          *ngIf=\"context.item.status === TenantStatus.SUSPENDED\"\n          title=\"{{ 'Suspended`tenant`' | translate }}\"\n        >\n          <i c8yIcon=\"ban\" class=\"text-danger\"></i>\n        </span>\n      </ng-container>\n    </c8y-column>\n  </c8y-data-grid>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.AppStateService }, { type: ɵngcc1.AlertService }, { type: ɵngcc1.ModalService }, { type: ɵngcc2.TranslateService }, { type: ɵngcc3.TenantService }, { type: ɵngcc1.TenantUiService }, { type: ɵngcc4.Location }, { type: ɵngcc1.PasswordService }, { type: ɵngcc3.UserService }]; }, null); })();
