// @c8y/ngx-components/esm2015/operations/stepper-bulk-type-device-profile/stepper-bulk-type-device-profile.module.js
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { CoreModule, FormsModule, gettext } from '@c8y/ngx-components';
import { DeviceProfileModule } from '@c8y/ngx-components/device-profile';
import { BulkOperationStepperModule } from '@c8y/ngx-components/operations/bulk-operation-stepper';
import { HOOK_LIST_BULK_TYPE, baseUrl, BulkOperationType } from '@c8y/ngx-components/operations/bulk-operations-service';
import { StepperFramesModule } from '@c8y/ngx-components/operations/stepper-frames';
import { ConfirmDeviceProfileSelectionStepComponent } from './confirm-device-profile-selection-step.component';
import { SelectDeviceProfileStepComponent } from './select-device-profile-step.component';
import { StepperBulkTypeDeviceProfileComponent } from './stepper-bulk-type-device-profile.component';
const ɵ0 = {
    type: BulkOperationType.DEVICE_PROFILE,
    c8yIcon: 'c8y-device-profile',
    name: gettext('Apply device profile'),
    path: `${baseUrl}device-profile`,
    component: StepperBulkTypeDeviceProfileComponent,
    fragments: ['c8y_DeviceProfile'],
    selected: false
};
/** Module for the 'Apply device profile' operation type stepper */
export class StepperBulkTypeDeviceProfileModule {
}
StepperBulkTypeDeviceProfileModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    StepperBulkTypeDeviceProfileComponent,
                    SelectDeviceProfileStepComponent,
                    ConfirmDeviceProfileSelectionStepComponent
                ],
                imports: [
                    CoreModule,
                    FormsModule,
                    ReactiveFormsModule,
                    BulkOperationStepperModule,
                    DeviceProfileModule,
                    StepperFramesModule
                ],
                providers: [
                    {
                        provide: HOOK_LIST_BULK_TYPE,
                        useValue: ɵ0,
                        multi: true
                    }
                ],
                entryComponents: [StepperBulkTypeDeviceProfileComponent],
                exports: [StepperBulkTypeDeviceProfileComponent]
            },] }
];
export { ɵ0 };
