// @c8y/ngx-components/esm2015/ecosystem/packages/package-details/package-details.component.js
import { __awaiter } from "tslib";
import { Component } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { FetchClient } from '@c8y/client';
import { WizardService } from '@c8y/ngx-components';
import { lexer as markedLexer, parser as markedParser } from 'marked';
import { BehaviorSubject } from 'rxjs';
import { packageProperties } from '../../ecosystem.constants';
import { EcosystemService } from '../../ecosystem.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from '@c8y/client';
import * as ɵngcc3 from '@c8y/ngx-components';
import * as ɵngcc4 from '../../ecosystem.service';
import * as ɵngcc5 from '@angular/common';
import * as ɵngcc6 from '../../applications/application-plugins/plugin-list.component';

function PackageDetailsComponent_p_33_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r0.description);
} }
function PackageDetailsComponent_p_34_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p", 28);
    ɵngcc0.ɵɵelementStart(1, "em");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(3, 1, "No description available."));
} }
function PackageDetailsComponent_div_35_Template(rf, ctx) { if (rf & 1) {
    const _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 29);
    ɵngcc0.ɵɵelementStart(1, "button", 30);
    ɵngcc0.ɵɵlistener("click", function PackageDetailsComponent_div_35_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r5 = ɵngcc0.ɵɵnextContext(); return ctx_r5.deploy(); });
    ɵngcc0.ɵɵelement(2, "i", 31);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 1, "Deploy application"), " ");
} }
function PackageDetailsComponent_div_45_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 32);
    ɵngcc0.ɵɵelement(1, "h1", 33);
    ɵngcc0.ɵɵelementStart(2, "div");
    ɵngcc0.ɵɵelementStart(3, "h3", 34);
    ɵngcc0.ɵɵtext(4, "No README.md found.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "p", 35);
    ɵngcc0.ɵɵtext(6, " To view the contents of \"README\", add the file \"README.md\" to the package. ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function PackageDetailsComponent_div_54_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 32);
    ɵngcc0.ɵɵelement(1, "h1", 36);
    ɵngcc0.ɵɵelementStart(2, "div");
    ɵngcc0.ɵɵelementStart(3, "h3", 34);
    ɵngcc0.ɵɵtext(4, "No plugins to display.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "p", 35);
    ɵngcc0.ɵɵtext(6, "This package doesn't contain plugins.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
export class PackageDetailsComponent {
    constructor(activatedRoute, client, wizard, ecosystemService) {
        this.activatedRoute = activatedRoute;
        this.client = client;
        this.wizard = wizard;
        this.ecosystemService = ecosystemService;
        this.package = {};
        this.plugins = new BehaviorSubject([]);
        this.packageProperties = packageProperties;
        this.headers = { 'Content-Type': 'text/markdown', responseType: 'blob' };
        this.NOT_FOUND_ERROR_CODE = 404;
    }
    ngOnInit() {
        var _a, _b, _c;
        return __awaiter(this, void 0, void 0, function* () {
            this.package = this.getPackage();
            this.name = (_a = this.package) === null || _a === void 0 ? void 0 : _a.name;
            this.description = (_c = (_b = this.package) === null || _b === void 0 ? void 0 : _b.manifest) === null || _c === void 0 ? void 0 : _c.description;
            this.readme = yield this.getReadmeFileContentAsHtml();
            this.appState = this.ecosystemService.getAppState(this.package);
            this.isPackageBlueprint = this.ecosystemService.isPackageBlueprint(this.package);
            this.plugins.next(Object.assign([], this.getPlugins()));
        });
    }
    deploy() {
        const initialState = {
            wizardConfig: {},
            id: 'deployPackage',
            package: this.package
        };
        const modalOptions = { initialState };
        this.wizard.show(modalOptions);
    }
    getPlugins() {
        const exports = (this.package.manifest && this.package.manifest.exports) || [];
        return exports;
    }
    getPackage() {
        var _a, _b, _c, _d;
        return (_d = (_c = (_b = (_a = this.activatedRoute) === null || _a === void 0 ? void 0 : _a.snapshot) === null || _b === void 0 ? void 0 : _b.parent) === null || _c === void 0 ? void 0 : _c.data) === null || _d === void 0 ? void 0 : _d.contextData;
    }
    getReadmeFileContentAsHtml() {
        return __awaiter(this, void 0, void 0, function* () {
            const readmeFile = yield this.getReadmeFile();
            const readmeContent = yield readmeFile.text();
            if (readmeFile.status === 200) {
                const tokens = markedLexer(readmeContent);
                const html = markedParser(tokens);
                return html;
            }
        });
    }
    getReadmeFile() {
        return __awaiter(this, void 0, void 0, function* () {
            let result;
            const options = {
                method: 'GET',
                headers: this.headers
            };
            result = yield this.client.fetch(`/apps/${this.package.contextPath}/readme.md`, options);
            if (result && result.status === this.NOT_FOUND_ERROR_CODE) {
                result = yield this.client.fetch(`/apps/${this.package.contextPath}/README.md`, options);
            }
            return result;
        });
    }
}
PackageDetailsComponent.ɵfac = function PackageDetailsComponent_Factory(t) { return new (t || PackageDetailsComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ActivatedRoute), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.FetchClient), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.WizardService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.EcosystemService)); };
PackageDetailsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: PackageDetailsComponent, selectors: [["c8y-package-details"]], decls: 58, vars: 59, consts: [[3, "icon", "label"], [3, "icon", "label", "path"], [3, "label"], [1, "card", "content-fullpage", "d-grid", "grid__col--8-4--md", "grid__row--fit-auto"], [1, "bg-gray-white", "grid__col--fullspan", "separator-bottom"], [1, "card-block", "p-t-24", "p-b-24", "large-padding"], [1, "content-flex-70"], [1, "text-center"], ["c8yIcon", "big-parcel", 1, "c8y-icon-duocolor", "icon-48"], [1, "label", 3, "ngClass"], [1, "flex-grow", "col-10"], [1, "content-flex-80"], [1, "col-5"], [1, "card-title", "text-bold", "m-b-8"], [4, "ngIf"], ["class", "text-muted", 4, "ngIf"], ["class", "col-3 text-right-md", 4, "ngIf"], [1, "flex-grow"], [3, "data", "emptyLabel", "icon", "properties", "title"], [1, "inner-scroll"], [1, "card-header", "separator", "sticky-top"], [1, "card-title"], [1, "card-block", "p-l-16", "p-r-16"], ["class", "c8y-empty-state text-center", 4, "ngIf"], [1, "markdown-content", 3, "innerHTML"], [1, "separator-bottom", "visible-sm", "visible-xs"], [1, "card-block"], [3, "plugins$", "selectable"], [1, "text-muted"], [1, "col-3", "text-right-md"], [1, "btn", "btn-primary", "btn-xs", 3, "click"], ["c8yIcon", "output", 1, "m-r-4"], [1, "c8y-empty-state", "text-center"], ["c8yIcon", "user-manual", 1, "c8y-icon-duocolor"], ["translate", ""], ["translate", "", 1, "m-r-8"], ["c8yIcon", "plugin"]], template: function PackageDetailsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵpipe(2, "async");
        ɵngcc0.ɵɵpipe(3, "humanizeAppName");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "c8y-breadcrumb");
        ɵngcc0.ɵɵelement(5, "c8y-breadcrumb-item", 0);
        ɵngcc0.ɵɵpipe(6, "translate");
        ɵngcc0.ɵɵelement(7, "c8y-breadcrumb-item", 1);
        ɵngcc0.ɵɵpipe(8, "translate");
        ɵngcc0.ɵɵelement(9, "c8y-breadcrumb-item", 1);
        ɵngcc0.ɵɵpipe(10, "translate");
        ɵngcc0.ɵɵelement(11, "c8y-breadcrumb-item", 2);
        ɵngcc0.ɵɵpipe(12, "async");
        ɵngcc0.ɵɵpipe(13, "humanizeAppName");
        ɵngcc0.ɵɵelement(14, "c8y-breadcrumb-item", 2);
        ɵngcc0.ɵɵpipe(15, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(16, "div", 3);
        ɵngcc0.ɵɵelementStart(17, "div", 4);
        ɵngcc0.ɵɵelementStart(18, "div", 5);
        ɵngcc0.ɵɵelementStart(19, "div", 6);
        ɵngcc0.ɵɵelementStart(20, "div", 7);
        ɵngcc0.ɵɵelement(21, "i", 8);
        ɵngcc0.ɵɵelementStart(22, "p");
        ɵngcc0.ɵɵelementStart(23, "span", 9);
        ɵngcc0.ɵɵtext(24);
        ɵngcc0.ɵɵpipe(25, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(26, "div", 10);
        ɵngcc0.ɵɵelementStart(27, "div", 11);
        ɵngcc0.ɵɵelementStart(28, "div", 12);
        ɵngcc0.ɵɵelementStart(29, "h4", 13);
        ɵngcc0.ɵɵtext(30);
        ɵngcc0.ɵɵpipe(31, "async");
        ɵngcc0.ɵɵpipe(32, "humanizeAppName");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(33, PackageDetailsComponent_p_33_Template, 2, 1, "p", 14);
        ɵngcc0.ɵɵtemplate(34, PackageDetailsComponent_p_34_Template, 4, 3, "p", 15);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(35, PackageDetailsComponent_div_35_Template, 5, 3, "div", 16);
        ɵngcc0.ɵɵelementStart(36, "div", 17);
        ɵngcc0.ɵɵelement(37, "c8y-properties-list", 18);
        ɵngcc0.ɵɵpipe(38, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(39, "div", 19);
        ɵngcc0.ɵɵelementStart(40, "div", 20);
        ɵngcc0.ɵɵelementStart(41, "h4", 21);
        ɵngcc0.ɵɵtext(42);
        ɵngcc0.ɵɵpipe(43, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(44, "div", 22);
        ɵngcc0.ɵɵtemplate(45, PackageDetailsComponent_div_45_Template, 7, 0, "div", 23);
        ɵngcc0.ɵɵelement(46, "div", 24);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(47, "div", 25);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(48, "div", 19);
        ɵngcc0.ɵɵelementStart(49, "div", 20);
        ɵngcc0.ɵɵelementStart(50, "h4", 21);
        ɵngcc0.ɵɵtext(51);
        ɵngcc0.ɵɵpipe(52, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(53, "div", 26);
        ɵngcc0.ɵɵtemplate(54, PackageDetailsComponent_div_54_Template, 7, 0, "div", 23);
        ɵngcc0.ɵɵpipe(55, "async");
        ɵngcc0.ɵɵelementStart(56, "div");
        ɵngcc0.ɵɵelement(57, "c8y-plugin-list", 27);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(2, 29, ɵngcc0.ɵɵpipeBind1(3, 31, ctx.name)));
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("icon", "c8y-atom")("label", ɵngcc0.ɵɵpipeBind1(6, 33, "Ecosystem"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("icon", "c8y-modules")("label", ɵngcc0.ɵɵpipeBind1(8, 35, "Applications"))("path", "ecosystem/applications");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("icon", "big-parcel")("label", ɵngcc0.ɵɵpipeBind1(10, 37, "Packages"))("path", "ecosystem/package-list");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("label", ɵngcc0.ɵɵpipeBind1(12, 39, ɵngcc0.ɵɵpipeBind1(13, 41, ctx.name)));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("label", ɵngcc0.ɵɵpipeBind1(15, 43, "Info"));
        ɵngcc0.ɵɵadvance(9);
        ɵngcc0.ɵɵproperty("ngClass", ctx.appState == null ? null : ctx.appState.class);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(25, 45, ctx.appState == null ? null : ctx.appState.label));
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(31, 47, ɵngcc0.ɵɵpipeBind1(32, 49, ctx.name)));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.description);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.description);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isPackageBlueprint);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("data", ctx.package.manifest)("emptyLabel", "--")("icon", "info-circle")("properties", ctx.packageProperties)("title", ɵngcc0.ɵɵpipeBind1(38, 51, "Package info"));
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(43, 53, "Package overview"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.readme);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("innerHTML", ctx.readme, ɵngcc0.ɵɵsanitizeHtml);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(52, 55, "Package plugins"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(55, 57, ctx.plugins).length === 0);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("plugins$", ctx.plugins)("selectable", false);
    } }, directives: [ɵngcc3.TitleComponent, ɵngcc3.BreadcrumbComponent, ɵngcc3.BreadcrumbItemComponent, ɵngcc3.IconDirective, ɵngcc5.NgClass, ɵngcc5.NgIf, ɵngcc3.PropertiesListComponent, ɵngcc6.PluginListComponent, ɵngcc3.C8yTranslateDirective], pipes: [ɵngcc5.AsyncPipe, ɵngcc3.HumanizeAppNamePipe, ɵngcc3.C8yTranslatePipe], encapsulation: 2 });
PackageDetailsComponent.ctorParameters = () => [
    { type: ActivatedRoute },
    { type: FetchClient },
    { type: WizardService },
    { type: EcosystemService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(PackageDetailsComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-package-details',
                template: "<c8y-title>{{ name | humanizeAppName | async }}</c8y-title>\n\n<c8y-breadcrumb>\n  <c8y-breadcrumb-item [icon]=\"'c8y-atom'\" [label]=\"'Ecosystem' | translate\"></c8y-breadcrumb-item>\n  <c8y-breadcrumb-item\n    [icon]=\"'c8y-modules'\"\n    [label]=\"'Applications' | translate\"\n    [path]=\"'ecosystem/applications'\"\n  ></c8y-breadcrumb-item>\n  <c8y-breadcrumb-item\n    [icon]=\"'big-parcel'\"\n    [label]=\"'Packages' | translate\"\n    [path]=\"'ecosystem/package-list'\"\n  ></c8y-breadcrumb-item>\n  <c8y-breadcrumb-item [label]=\"name | humanizeAppName | async\"></c8y-breadcrumb-item>\n  <c8y-breadcrumb-item [label]=\"'Info' | translate\"></c8y-breadcrumb-item>\n</c8y-breadcrumb>\n\n<div class=\"card content-fullpage d-grid grid__col--8-4--md grid__row--fit-auto\">\n  <div class=\"bg-gray-white grid__col--fullspan separator-bottom\">\n    <div class=\"card-block p-t-24 p-b-24 large-padding\">\n      <div class=\"content-flex-70\">\n        <div class=\"text-center\">\n          <i c8yIcon=\"big-parcel\" class=\"c8y-icon-duocolor icon-48\"></i>\n          <p>\n            <span [ngClass]=\"appState?.class\" class=\"label\">{{ appState?.label | translate }}</span>\n          </p>\n        </div>\n\n        <div class=\"flex-grow col-10\">\n          <div class=\"content-flex-80\">\n            <div class=\"col-5\">\n              <h4 class=\"card-title text-bold m-b-8\">{{ name | humanizeAppName | async }}</h4>\n              <p *ngIf=\"description\">{{ description }}</p>\n              <p *ngIf=\"!description\" class=\"text-muted\">\n                <em>{{ 'No description available.' | translate }}</em>\n              </p>\n            </div>\n            <div *ngIf=\"isPackageBlueprint\" class=\"col-3 text-right-md\">\n              <button (click)=\"deploy()\" class=\"btn btn-primary btn-xs\">\n                <i c8yIcon=\"output\" class=\"m-r-4\"></i>\n                {{ 'Deploy application' | translate }}\n              </button>\n            </div>\n            <div class=\"flex-grow\">\n              <c8y-properties-list\n                [data]=\"package.manifest\"\n                [emptyLabel]=\"'--'\"\n                [icon]=\"'info-circle'\"\n                [properties]=\"packageProperties\"\n                [title]=\"'Package info' | translate\"\n              ></c8y-properties-list>\n            </div>\n          </div>\n        </div>\n      </div>\n    </div>\n  </div>\n  <div class=\"inner-scroll\">\n    <div class=\"card-header separator sticky-top\">\n      <h4 class=\"card-title\">{{ 'Package overview' | translate }}</h4>\n    </div>\n    <div class=\"card-block p-l-16 p-r-16\">\n      <div *ngIf=\"!readme\" class=\"c8y-empty-state text-center\">\n        <h1 c8yIcon=\"user-manual\" class=\"c8y-icon-duocolor\"></h1>\n        <div>\n          <h3 translate>No README.md found.</h3>\n          <p class=\"m-r-8\" translate>\n            To view the contents of \"README\", add the file \"README.md\" to the package.\n          </p>\n        </div>\n      </div>\n\n      <div [innerHTML]=\"readme\" class=\"markdown-content\"></div>\n    </div>\n    <div class=\"separator-bottom visible-sm visible-xs\"></div>\n  </div>\n\n  <div class=\"inner-scroll\">\n    <div class=\"card-header separator sticky-top\">\n      <h4 class=\"card-title\">{{ 'Package plugins' | translate }}</h4>\n    </div>\n    <div class=\"card-block\">\n      <!-- empty state -->\n      <div *ngIf=\"(plugins | async).length === 0\" class=\"c8y-empty-state text-center\">\n        <h1 c8yIcon=\"plugin\"></h1>\n        <div>\n          <h3 translate>No plugins to display.</h3>\n          <p class=\"m-r-8\" translate>This package doesn't contain plugins.</p>\n        </div>\n      </div>\n\n      <div>\n        <c8y-plugin-list [plugins$]=\"plugins\" [selectable]=\"false\"></c8y-plugin-list>\n      </div>\n    </div>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.ActivatedRoute }, { type: ɵngcc2.FetchClient }, { type: ɵngcc3.WizardService }, { type: ɵngcc4.EcosystemService }]; }, null); })();
