// @c8y/ngx-components/esm2015/services/services-device-tab/services.breadcrumb-factory.js
import { __awaiter } from "tslib";
import { Injectable } from '@angular/core';
import { InventoryService } from '@c8y/client';
import { gettext, ViewContext } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/client';
export class ServicesBreadcrumbFactory {
    constructor(inventoryService) {
        this.inventoryService = inventoryService;
    }
    get(route) {
        var _a, _b, _c, _d;
        return __awaiter(this, void 0, void 0, function* () {
            const { context, contextData } = route.snapshot.data;
            const { routeConfig } = route.snapshot;
            this.contextData = contextData || this.contextData;
            if ((context === ViewContext.Service || ((_a = routeConfig) === null || _a === void 0 ? void 0 : _a.context) === ViewContext.Service) &&
                !!this.contextData) {
                const serviceMO = (_b = (yield this.inventoryService.detail(this.contextData, { withParents: true }))) === null || _b === void 0 ? void 0 : _b.data;
                const breadcrumb = { items: [] };
                const breadcrumbItems = [];
                const device = (_d = (_c = serviceMO === null || serviceMO === void 0 ? void 0 : serviceMO.additionParents) === null || _c === void 0 ? void 0 : _c.references[0]) === null || _d === void 0 ? void 0 : _d.managedObject;
                if (device) {
                    breadcrumbItems.push({
                        label: device.name,
                        path: `/device/${device.id}`
                    });
                    breadcrumbItems.push({
                        label: gettext('Services'),
                        path: `/device/${device.id}/services`
                    });
                    breadcrumbItems.push({
                        label: serviceMO.name
                    });
                }
                breadcrumb.items = breadcrumbItems;
                return breadcrumb;
            }
        });
    }
}
ServicesBreadcrumbFactory.ɵfac = function ServicesBreadcrumbFactory_Factory(t) { return new (t || ServicesBreadcrumbFactory)(ɵngcc0.ɵɵinject(ɵngcc1.InventoryService)); };
ServicesBreadcrumbFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: ServicesBreadcrumbFactory, factory: ServicesBreadcrumbFactory.ɵfac });
ServicesBreadcrumbFactory.ctorParameters = () => [
    { type: InventoryService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ServicesBreadcrumbFactory, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.InventoryService }]; }, null); })();
