// @c8y/ngx-components/esm2015/services/services-device-tab/services-device-tab.module.js
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CoreModule, gettext, HOOK_BREADCRUMB, HOOK_ROUTE, ViewContext } from '@c8y/ngx-components';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { LastUpdatedDateCellRendererComponent, NameCellRendererComponent, StatusCellRendererComponent } from './columns';
import { ServicesDeviceTabComponent } from './services-device-tab.component';
import { ServicesDeviceTabGuard } from './services-device-tab.guard';
import { ServicesBreadcrumbFactory } from './services.breadcrumb-factory';
import * as ɵngcc0 from '@angular/core';
const ɵ0 = [
    {
        context: ViewContext.Device,
        path: 'services',
        component: ServicesDeviceTabComponent,
        label: gettext('Services'),
        icon: 'statistics',
        priority: 500,
        canActivate: [ServicesDeviceTabGuard]
    }
];
export class ServicesDeviceTabModule {
}
ServicesDeviceTabModule.ɵfac = function ServicesDeviceTabModule_Factory(t) { return new (t || ServicesDeviceTabModule)(); };
ServicesDeviceTabModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: ServicesDeviceTabModule });
ServicesDeviceTabModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        ServicesDeviceTabGuard,
        ServicesBreadcrumbFactory,
        {
            provide: HOOK_ROUTE,
            useValue: ɵ0,
            multi: true
        },
        { provide: HOOK_BREADCRUMB, useClass: ServicesBreadcrumbFactory, multi: true }
    ], imports: [[CoreModule, RouterModule, TooltipModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ServicesDeviceTabModule, [{
        type: NgModule,
        args: [{
                imports: [CoreModule, RouterModule, TooltipModule],
                declarations: [
                    ServicesDeviceTabComponent,
                    LastUpdatedDateCellRendererComponent,
                    NameCellRendererComponent,
                    StatusCellRendererComponent
                ],
                providers: [
                    ServicesDeviceTabGuard,
                    ServicesBreadcrumbFactory,
                    {
                        provide: HOOK_ROUTE,
                        useValue: ɵ0,
                        multi: true
                    },
                    { provide: HOOK_BREADCRUMB, useClass: ServicesBreadcrumbFactory, multi: true }
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ServicesDeviceTabModule, { declarations: function () { return [ServicesDeviceTabComponent,
        LastUpdatedDateCellRendererComponent,
        NameCellRendererComponent,
        StatusCellRendererComponent]; }, imports: function () { return [CoreModule, RouterModule, TooltipModule]; } }); })();
export { ɵ0 };
