// @c8y/ngx-components/esm2015/core/data-grid/configure-custom-column/configure-custom-column.component.js
import { Component, EventEmitter, Output, ViewChild } from '@angular/core';
import { FormBuilder } from '@angular/forms';
import { camelCase } from 'lodash-es';
import { BsModalRef } from 'ngx-bootstrap/modal';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/forms';
import * as ɵngcc2 from 'ngx-bootstrap/modal';
import * as ɵngcc3 from '../../i18n/c8y-translate.directive';
import * as ɵngcc4 from '../../forms/form-group.component';
import * as ɵngcc5 from '../../forms/required-input-placeholder.directive';
import * as ɵngcc6 from '../../forms/unique-in-collection-by-path-validation.directive';
import * as ɵngcc7 from '../../forms/messages.component';
import * as ɵngcc8 from '../../forms/message.directive';
import * as ɵngcc9 from '../../i18n/c8y-translate.pipe';

const _c0 = ["columnConfigForm"];
const _c1 = function () { return { example: "c8y_Agent.name" }; };
export class ConfigureCustomColumnComponent {
    constructor(fb, modalRef) {
        this.fb = fb;
        this.modalRef = modalRef;
        this.SIMPLE_JSON_PATH_REGEX = /^[^.\s]+(\.\S+)*$/;
        this.onAddCustomColumn = new EventEmitter();
        this.formGroup = this.fb.group({
            header: [''],
            path: [''],
            addAnother: [false]
        });
    }
    comparator(val1, val2) {
        return camelCase(val1) === camelCase(val2);
    }
    addColumn() {
        const { header, path } = this.formGroup.value;
        this.onAddCustomColumn.emit({
            name: camelCase(header),
            header,
            path,
            custom: true,
            visible: true
        });
        if (this.formGroup.value.addAnother) {
            this.form.resetForm({ addAnother: true });
        }
        else {
            this.modalRef.hide();
        }
    }
}
ConfigureCustomColumnComponent.ɵfac = function ConfigureCustomColumnComponent_Factory(t) { return new (t || ConfigureCustomColumnComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.FormBuilder), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.BsModalRef)); };
ConfigureCustomColumnComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ConfigureCustomColumnComponent, selectors: [["c8y-configure-custom-column"]], viewQuery: function ConfigureCustomColumnComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.form = _t.first);
    } }, outputs: { onAddCustomColumn: "onAddCustomColumn" }, decls: 39, vars: 36, consts: [[1, "viewport-modal"], [1, "modal-header", "separator-bottom"], ["translate", ""], [1, "d-contents", 3, "formGroup"], ["columnConfigForm", "ngForm"], [1, "modal-inner-scroll"], [1, "modal-body"], ["for", "header", "translate", ""], ["id", "header", "formControlName", "header", "autocomplete", "off", "uniqueByPath", "", "path", "header", 1, "form-control", 3, "placeholder", "required", "collection", "comparator"], ["name", "uniqueByPath", 3, "text"], ["for", "fragmentPath", "translate", ""], ["id", "fragmentPath", "formControlName", "path", "autocomplete", "off", 1, "form-control", 3, "placeholder", "required", "pattern"], ["name", "pattern", 3, "text"], [1, "c8y-switch", 3, "title"], ["type", "checkbox", "formControlName", "addAnother", 1, "form-control"], [1, "modal-footer"], [1, "btn", "btn-default", 3, "title", "click"], [1, "btn", "btn-primary", 3, "title", "disabled", "click"]], template: function ConfigureCustomColumnComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "h3", 2);
        ɵngcc0.ɵɵtext(3, "Configure custom column");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "form", 3, 4);
        ɵngcc0.ɵɵelementStart(6, "div", 5);
        ɵngcc0.ɵɵelementStart(7, "div", 6);
        ɵngcc0.ɵɵelementStart(8, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(9, "label", 7);
        ɵngcc0.ɵɵtext(10, "Header");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(11, "input", 8);
        ɵngcc0.ɵɵpipe(12, "translate");
        ɵngcc0.ɵɵelementStart(13, "c8y-messages");
        ɵngcc0.ɵɵelement(14, "c8y-message", 9);
        ɵngcc0.ɵɵpipe(15, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(16, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(17, "label", 10);
        ɵngcc0.ɵɵtext(18, "Fragment path");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(19, "input", 11);
        ɵngcc0.ɵɵpipe(20, "translate");
        ɵngcc0.ɵɵelementStart(21, "c8y-messages");
        ɵngcc0.ɵɵelement(22, "c8y-message", 12);
        ɵngcc0.ɵɵpipe(23, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(24, "label", 13);
        ɵngcc0.ɵɵpipe(25, "translate");
        ɵngcc0.ɵɵelement(26, "input", 14);
        ɵngcc0.ɵɵelement(27, "span");
        ɵngcc0.ɵɵelementStart(28, "span", 2);
        ɵngcc0.ɵɵtext(29, "Add another column after saving this one");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(30, "div", 15);
        ɵngcc0.ɵɵelementStart(31, "button", 16);
        ɵngcc0.ɵɵlistener("click", function ConfigureCustomColumnComponent_Template_button_click_31_listener() { return ctx.modalRef.hide(); });
        ɵngcc0.ɵɵpipe(32, "translate");
        ɵngcc0.ɵɵtext(33);
        ɵngcc0.ɵɵpipe(34, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(35, "button", 17);
        ɵngcc0.ɵɵlistener("click", function ConfigureCustomColumnComponent_Template_button_click_35_listener() { return ctx.addColumn(); });
        ɵngcc0.ɵɵpipe(36, "translate");
        ɵngcc0.ɵɵtext(37);
        ɵngcc0.ɵɵpipe(38, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r0 = ɵngcc0.ɵɵreference(5);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("formGroup", ctx.formGroup);
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵproperty("placeholder", ɵngcc0.ɵɵpipeBind1(12, 16, "e.g. Agent name"))("required", true)("collection", ctx.columns)("comparator", ctx.comparator);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("text", ɵngcc0.ɵɵpipeBind1(15, 18, "Column's header name must be unique"));
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("placeholder", ɵngcc0.ɵɵpipeBind2(20, 20, "e.g. {{ example }}", ɵngcc0.ɵɵpureFunction0(35, _c1)))("required", true)("pattern", ctx.SIMPLE_JSON_PATH_REGEX);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("text", ɵngcc0.ɵɵpipeBind1(23, 23, "Must be a valid JSON path"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(25, 25, "Add another column after saving this one"));
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(32, 27, "Cancel"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(34, 29, "Cancel"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(36, 31, "Save"));
        ɵngcc0.ɵɵproperty("disabled", _r0.form.pristine || !_r0.form.valid);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(38, 33, "Save"), " ");
    } }, directives: [ɵngcc3.C8yTranslateDirective, ɵngcc1.ɵNgNoValidate, ɵngcc1.NgControlStatusGroup, ɵngcc1.FormGroupDirective, ɵngcc4.FormGroupComponent, ɵngcc1.DefaultValueAccessor, ɵngcc5.RequiredInputPlaceholderDirective, ɵngcc1.NgControlStatus, ɵngcc1.FormControlName, ɵngcc6.UniqueInCollectionByPathValidationDirective, ɵngcc1.RequiredValidator, ɵngcc7.MessagesComponent, ɵngcc8.MessageDirective, ɵngcc1.PatternValidator, ɵngcc1.CheckboxControlValueAccessor], pipes: [ɵngcc9.C8yTranslatePipe], encapsulation: 2 });
ConfigureCustomColumnComponent.ctorParameters = () => [
    { type: FormBuilder },
    { type: BsModalRef }
];
ConfigureCustomColumnComponent.propDecorators = {
    form: [{ type: ViewChild, args: ['columnConfigForm', { static: false },] }],
    onAddCustomColumn: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ConfigureCustomColumnComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-configure-custom-column',
                template: "<div class=\"viewport-modal\">\n  <div class=\"modal-header separator-bottom\">\n    <h3 translate>Configure custom column</h3>\n  </div>\n\n  <form class=\"d-contents\" #columnConfigForm=\"ngForm\" [formGroup]=\"formGroup\">\n    <div class=\"modal-inner-scroll\">\n      <div class=\"modal-body\">\n        <c8y-form-group>\n          <label for=\"header\" translate>Header</label>\n          <input\n            id=\"header\"\n            formControlName=\"header\"\n            class=\"form-control\"\n            autocomplete=\"off\"\n            [placeholder]=\"'e.g. Agent name' | translate\"\n            [required]=\"true\"\n            uniqueByPath\n            [collection]=\"columns\"\n            path=\"header\"\n            [comparator]=\"comparator\"\n          />\n          <c8y-messages>\n            <c8y-message\n              name=\"uniqueByPath\"\n              [text]=\"'Column\\'s header name must be unique' | translate\"\n            ></c8y-message>\n          </c8y-messages>\n        </c8y-form-group>\n        <c8y-form-group>\n          <label for=\"fragmentPath\" translate>Fragment path</label>\n          <input\n            id=\"fragmentPath\"\n            formControlName=\"path\"\n            class=\"form-control\"\n            autocomplete=\"off\"\n            [placeholder]=\"'e.g. {{ example }}' | translate: { example: 'c8y_Agent.name' }\"\n            [required]=\"true\"\n            [pattern]=\"SIMPLE_JSON_PATH_REGEX\"\n          />\n          <c8y-messages>\n            <c8y-message\n              name=\"pattern\"\n              [text]=\"'Must be a valid JSON path' | translate\"\n            ></c8y-message>\n          </c8y-messages>\n        </c8y-form-group>\n        <label class=\"c8y-switch\" [title]=\"'Add another column after saving this one' | translate\">\n          <input type=\"checkbox\" formControlName=\"addAnother\" class=\"form-control\" />\n          <span></span>\n          <span translate>Add another column after saving this one</span>\n        </label>\n      </div>\n    </div>\n\n    <div class=\"modal-footer\">\n      <button title=\"{{ 'Cancel' | translate }}\" class=\"btn btn-default\" (click)=\"modalRef.hide()\">\n        {{ 'Cancel' | translate }}\n      </button>\n      <button\n        title=\"{{ 'Save' | translate }}\"\n        class=\"btn btn-primary\"\n        [disabled]=\"columnConfigForm.form.pristine || !columnConfigForm.form.valid\"\n        (click)=\"addColumn()\"\n      >\n        {{ 'Save' | translate }}\n      </button>\n    </div>\n  </form>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.FormBuilder }, { type: ɵngcc2.BsModalRef }]; }, { onAddCustomColumn: [{
            type: Output
        }], form: [{
            type: ViewChild,
            args: ['columnConfigForm', { static: false }]
        }] }); })();
