// @c8y/ngx-components/esm2015/device-list/device-list.module.js
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CoreModule, CommonModule, HOOK_NAVIGATOR_NODES } from '@c8y/ngx-components';
import { DeviceGridModule } from '@c8y/ngx-components/device-grid';
import { DeviceListNavigationFactory } from './device-list-navigation.factory';
import { DeviceListComponent } from './device-list.component';
import { AddSmartGroupComponent } from './add-smart-group.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
const routes = [
    {
        path: 'device',
        component: DeviceListComponent
    }
];
export class DeviceListModule {
}
DeviceListModule.ɵfac = function DeviceListModule_Factory(t) { return new (t || DeviceListModule)(); };
DeviceListModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DeviceListModule });
DeviceListModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        {
            provide: HOOK_NAVIGATOR_NODES,
            useClass: DeviceListNavigationFactory,
            multi: true
        }
    ], imports: [[CoreModule, CommonModule, DeviceGridModule, RouterModule.forChild(routes)]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DeviceListModule, [{
        type: NgModule,
        args: [{
                imports: [CoreModule, CommonModule, DeviceGridModule, RouterModule.forChild(routes)],
                declarations: [DeviceListComponent, AddSmartGroupComponent],
                entryComponents: [DeviceListComponent],
                exports: [],
                providers: [
                    {
                        provide: HOOK_NAVIGATOR_NODES,
                        useClass: DeviceListNavigationFactory,
                        multi: true
                    }
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DeviceListModule, { declarations: function () { return [DeviceListComponent, AddSmartGroupComponent]; }, imports: function () { return [CoreModule, CommonModule, DeviceGridModule, ɵngcc1.RouterModule]; } }); })();
