// @c8y/ngx-components/esm2015/device-list/device-list-navigation.factory.js
import { __awaiter } from "tslib";
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
export class DeviceListNavigationFactory {
    constructor() {
        this.navs = [];
    }
    get() {
        return __awaiter(this, void 0, void 0, function* () {
            this.navs.push(new NavigatorNode({
                label: gettext('All devices'),
                path: 'device',
                icon: 'c8y-device-management',
                parent: {
                    label: gettext('Devices'),
                    icon: 'exchange'
                },
                priority: 2000
            }));
            return this.navs;
        });
    }
}
DeviceListNavigationFactory.ɵfac = function DeviceListNavigationFactory_Factory(t) { return new (t || DeviceListNavigationFactory)(); };
DeviceListNavigationFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: DeviceListNavigationFactory, factory: DeviceListNavigationFactory.ɵfac });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DeviceListNavigationFactory, [{
        type: Injectable
    }], function () { return []; }, null); })();
