// @c8y/ngx-components/esm2015/operations/bulk-operation-stepper/bulk-operation-stepper.module.js
import { NgModule } from '@angular/core';
import { CoreModule } from '@c8y/ngx-components';
import { CreateBulkOperationDetailsModule } from '@c8y/ngx-components/operations/create-bulk-operation-details';
import { DeviceSelectorModule } from '@c8y/ngx-components/operations/device-selector';
import { OperationSummaryModule } from '@c8y/ngx-components/operations/operation-summary';
import { BulkOperationStepper } from './bulk-operation-stepper.component';
import { CustomStep } from './custom-step.directive';
/**
 * This module provides base stepper class and stepper wrapper component.
 */
import * as ɵngcc0 from '@angular/core';
export class BulkOperationStepperModule {
}
BulkOperationStepperModule.ɵfac = function BulkOperationStepperModule_Factory(t) { return new (t || BulkOperationStepperModule)(); };
BulkOperationStepperModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: BulkOperationStepperModule });
BulkOperationStepperModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            CoreModule,
            DeviceSelectorModule,
            CreateBulkOperationDetailsModule,
            OperationSummaryModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BulkOperationStepperModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CoreModule,
                    DeviceSelectorModule,
                    CreateBulkOperationDetailsModule,
                    OperationSummaryModule
                ],
                declarations: [BulkOperationStepper, CustomStep],
                exports: [BulkOperationStepper, CustomStep]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(BulkOperationStepperModule, { declarations: function () { return [BulkOperationStepper, CustomStep]; }, imports: function () { return [CoreModule,
        DeviceSelectorModule,
        CreateBulkOperationDetailsModule,
        OperationSummaryModule]; }, exports: function () { return [BulkOperationStepper, CustomStep]; } }); })();
