// @c8y/ngx-components/esm2015/core/date-time-picker/date-time-picker.module.js
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { CommonModule as C8yCommonModule } from '../common/common.module';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { FormsModule as C8yFormsModule } from '../forms/index';
import { BsDatepickerModule } from 'ngx-bootstrap/datepicker';
import { DateTimePickerComponent } from './date-time-picker.component';
import { TimePickerModule } from '../time-picker/time-picker.module';
/**
 * The angular module definition for a date-time picker - it provides FormControl that handles date and time.
 * @exports DateTimePickerComponent
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/datepicker';
export class DateTimePickerModule {
}
DateTimePickerModule.ɵfac = function DateTimePickerModule_Factory(t) { return new (t || DateTimePickerModule)(); };
DateTimePickerModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DateTimePickerModule });
DateTimePickerModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            C8yFormsModule,
            CommonModule,
            FormsModule,
            ReactiveFormsModule,
            BsDatepickerModule.forRoot(),
            C8yCommonModule,
            TimePickerModule,
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DateTimePickerModule, [{
        type: NgModule,
        args: [{
                imports: [
                    C8yFormsModule,
                    CommonModule,
                    FormsModule,
                    ReactiveFormsModule,
                    BsDatepickerModule.forRoot(),
                    C8yCommonModule,
                    TimePickerModule,
                ],
                exports: [DateTimePickerComponent],
                declarations: [DateTimePickerComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DateTimePickerModule, { declarations: function () { return [DateTimePickerComponent]; }, imports: function () { return [C8yFormsModule,
        CommonModule,
        FormsModule,
        ReactiveFormsModule, ɵngcc1.BsDatepickerModule, C8yCommonModule,
        TimePickerModule]; }, exports: function () { return [DateTimePickerComponent]; } }); })();
