// @c8y/ngx-components/esm2015/upgrade/dashboard/dashboard-upgrade.module.js
import { CommonModule } from '@angular/common';
import { NgModule, NgZone, Optional } from '@angular/core';
import { Router } from '@angular/router';
import { HOOK_COMPONENTS } from '@c8y/ngx-components';
import { ContextDashboardService } from '@c8y/ngx-components/context-dashboard';
import { Ng1ComponentFactoryService } from './component.factory';
import { DashboardBridgeService } from './dashboard-bridge.service';
import { WidgetComponent } from './widget.component';
import { DeviceSelectorComponent } from './device-selector.component';
import { C8yTranslateModule } from '@c8y/ngx-components';
import { AssetSelectorModule } from '@c8y/ngx-components/assets-navigator';
export class DashboardUpgradeModule {
}
DashboardUpgradeModule.decorators = [
    { type: NgModule, args: [{
                declarations: [DeviceSelectorComponent, WidgetComponent],
                entryComponents: [DeviceSelectorComponent, WidgetComponent],
                imports: [CommonModule, C8yTranslateModule, AssetSelectorModule],
                providers: [
                    {
                        provide: DashboardBridgeService,
                        useFactory: dashboardBridgeFactory,
                        deps: [
                            '$injector',
                            NgZone,
                            Router,
                            [new Optional(), ContextDashboardService]
                        ]
                    },
                    { provide: HOOK_COMPONENTS, useClass: Ng1ComponentFactoryService, multi: true }
                ]
            },] }
];
export function dashboardBridgeFactory(injector, ngZone, router, contextDashboardService) {
    return new DashboardBridgeService(injector, ngZone, router, contextDashboardService);
}
