// @c8y/ngx-components/esm2015/operations/operations-list-item-details/operations-list-item-details.module.js
import { NgModule } from '@angular/core';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { CoreModule } from '@c8y/ngx-components';
import { BulkSingleOperationsListModule } from '@c8y/ngx-components/operations/bulk-single-operations-list';
import { StatusFilterModule } from '@c8y/ngx-components/operations/status-filter';
import { OperationDetailsTabsComponent } from './operation-details-tabs.component';
import { OperationDetailsComponent } from './operation-details.component';
/**
 * This module allows for displaying the list of single operations and their details.
 */
import * as ɵngcc0 from '@angular/core';
export class OperationsListItemDetailsModule {
}
OperationsListItemDetailsModule.ɵfac = function OperationsListItemDetailsModule_Factory(t) { return new (t || OperationsListItemDetailsModule)(); };
OperationsListItemDetailsModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: OperationsListItemDetailsModule });
OperationsListItemDetailsModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CoreModule, TooltipModule, StatusFilterModule, BulkSingleOperationsListModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OperationsListItemDetailsModule, [{
        type: NgModule,
        args: [{
                imports: [CoreModule, TooltipModule, StatusFilterModule, BulkSingleOperationsListModule],
                declarations: [OperationDetailsComponent, OperationDetailsTabsComponent],
                exports: [OperationDetailsComponent, OperationDetailsTabsComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(OperationsListItemDetailsModule, { declarations: function () { return [OperationDetailsComponent, OperationDetailsTabsComponent]; }, imports: function () { return [CoreModule, TooltipModule, StatusFilterModule, BulkSingleOperationsListModule]; }, exports: function () { return [OperationDetailsComponent, OperationDetailsTabsComponent]; } }); })();
