// @c8y/ngx-components/esm2015/operations/operations-list-item-details/operation-details-tabs.component.js
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { gettext } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from './operation-details.component';
import * as ɵngcc4 from '@c8y/ngx-components/operations/bulk-single-operations-list';

const _c0 = ["details"];
const _c1 = ["audit"];
const _c2 = function (a0) { return { active: a0 }; };
function OperationDetailsTabsComponent_li_3_Template(rf, ctx) { if (rf & 1) {
    const _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 8);
    ɵngcc0.ɵɵelementStart(1, "button", 9);
    ɵngcc0.ɵɵlistener("click", function OperationDetailsTabsComponent_li_3_Template_button_click_1_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r10); const tab_r8 = restoredCtx.$implicit; const ctx_r9 = ɵngcc0.ɵɵnextContext(); return ctx_r9.template = tab_r8.template; });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 10);
    ɵngcc0.ɵɵelementStart(4, "span", 11);
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const tab_r8 = ctx.$implicit;
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(8, _c2, tab_r8.template === ctx_r1.template));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(2, 4, tab_r8.label));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", tab_r8.icon);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(6, 6, tab_r8.label));
} }
function OperationDetailsTabsComponent_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function OperationDetailsTabsComponent_ng_template_5_Template(rf, ctx) { if (rf & 1) {
    const _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-operation-details", 12);
    ɵngcc0.ɵɵlistener("showFailedOperation", function OperationDetailsTabsComponent_ng_template_5_Template_c8y_operation_details_showFailedOperation_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r12); const ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.showFailedOperation.emit($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("operation", ctx_r4.operation)("bulkOperationModalDetailsService", ctx_r4.bulkOperationModalDetailsService);
} }
function OperationDetailsTabsComponent_ng_template_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-audit-log", 13);
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("source", ctx_r6.operation.id);
} }
function OperationDetailsTabsComponent_c8y_single_operations_list_9_Template(rf, ctx) { if (rf & 1) {
    const _r14 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-single-operations-list", 14);
    ɵngcc0.ɵɵlistener("onRetryFailedOperations", function OperationDetailsTabsComponent_c8y_single_operations_list_9_Template_c8y_single_operations_list_onRetryFailedOperations_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r14); const ctx_r13 = ɵngcc0.ɵɵnextContext(); return ctx_r13.onRetryFailedOperations.emit($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("bulkOperation", ctx_r7.operation)("readOnly", ctx_r7.readOnly);
} }
export class OperationDetailsTabsComponent {
    constructor() {
        this.readOnly = false;
        this.onRetryFailedOperations = new EventEmitter();
        this.showFailedOperation = new EventEmitter();
        this.tabs = [];
    }
    ngOnInit() {
        this.template = this.detailsTemplate;
        this.tabs = [
            {
                label: gettext('Details'),
                path: null,
                icon: 'asterisk',
                template: this.detailsTemplate
            },
            {
                label: gettext('History of changes'),
                path: null,
                icon: 'archive',
                template: this.auditTemplate
            }
        ];
    }
}
OperationDetailsTabsComponent.ɵfac = function OperationDetailsTabsComponent_Factory(t) { return new (t || OperationDetailsTabsComponent)(); };
OperationDetailsTabsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: OperationDetailsTabsComponent, selectors: [["c8y-operation-details-tabs"]], viewQuery: function OperationDetailsTabsComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 7);
        ɵngcc0.ɵɵviewQuery(_c1, 7);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.detailsTemplate = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.auditTemplate = _t.first);
    } }, inputs: { readOnly: "readOnly", operation: "operation", bulkOperationModalDetailsService: "bulkOperationModalDetailsService" }, outputs: { onRetryFailedOperations: "onRetryFailedOperations", showFailedOperation: "showFailedOperation" }, decls: 10, vars: 3, consts: [[1, "tabContainer"], ["container", ""], [1, "nav", "nav-tabs", "nav-tabsc8y"], [3, "ngClass", 4, "ngFor", "ngForOf"], [4, "ngTemplateOutlet"], ["details", ""], ["audit", ""], ["class", "d-flex d-col", 3, "bulkOperation", "readOnly", "onRetryFailedOperations", 4, "ngIf"], [3, "ngClass"], [1, "btn", 3, "title", "click"], [3, "c8yIcon"], [1, "txt"], [3, "operation", "bulkOperationModalDetailsService", "showFailedOperation"], [3, "source"], [1, "d-flex", "d-col", 3, "bulkOperation", "readOnly", "onRetryFailedOperations"]], template: function OperationDetailsTabsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0, 1);
        ɵngcc0.ɵɵelementStart(2, "ul", 2);
        ɵngcc0.ɵɵtemplate(3, OperationDetailsTabsComponent_li_3_Template, 7, 10, "li", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(4, OperationDetailsTabsComponent_ng_container_4_Template, 1, 0, "ng-container", 4);
        ɵngcc0.ɵɵtemplate(5, OperationDetailsTabsComponent_ng_template_5_Template, 1, 2, "ng-template", null, 5, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(7, OperationDetailsTabsComponent_ng_template_7_Template, 1, 1, "ng-template", null, 6, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(9, OperationDetailsTabsComponent_c8y_single_operations_list_9_Template, 1, 2, "c8y-single-operations-list", 7);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.tabs);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.template);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("ngIf", ctx.operation.operationPrototype);
    } }, directives: [ɵngcc1.NgForOf, ɵngcc1.NgTemplateOutlet, ɵngcc1.NgIf, ɵngcc1.NgClass, ɵngcc2.IconDirective, ɵngcc3.OperationDetailsComponent, ɵngcc2.ɵbd, ɵngcc4.SingleOperationsListComponent], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
OperationDetailsTabsComponent.propDecorators = {
    operation: [{ type: Input }],
    readOnly: [{ type: Input }],
    bulkOperationModalDetailsService: [{ type: Input }],
    onRetryFailedOperations: [{ type: Output }],
    showFailedOperation: [{ type: Output }],
    detailsTemplate: [{ type: ViewChild, args: ['details', { static: true },] }],
    auditTemplate: [{ type: ViewChild, args: ['audit', { static: true },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OperationDetailsTabsComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-operation-details-tabs',
                template: "<div #container class=\"tabContainer\">\n  <ul class=\"nav nav-tabs nav-tabsc8y\">\n    <li [ngClass]=\"{ active: tab.template === template }\" *ngFor=\"let tab of tabs\">\n      <button [title]=\"tab.label | translate\" class=\"btn\" (click)=\"template = tab.template\">\n        <i [c8yIcon]=\"tab.icon\"></i>\n        <span class=\"txt\">{{ tab.label | translate }}</span>\n      </button>\n    </li>\n  </ul>\n</div>\n\n<ng-container *ngTemplateOutlet=\"template\"></ng-container>\n\n<ng-template #details>\n  <c8y-operation-details\n    [operation]=\"operation\"\n    (showFailedOperation)=\"showFailedOperation.emit($event)\"\n    [bulkOperationModalDetailsService]=\"bulkOperationModalDetailsService\"\n  >\n  </c8y-operation-details>\n</ng-template>\n\n<ng-template #audit>\n  <c8y-audit-log [source]=\"operation.id\"></c8y-audit-log>\n</ng-template>\n\n<c8y-single-operations-list\n  *ngIf=\"operation.operationPrototype\"\n  [bulkOperation]=\"operation\"\n  [readOnly]=\"readOnly\"\n  (onRetryFailedOperations)=\"onRetryFailedOperations.emit($event)\"\n  class=\"d-flex d-col\"\n></c8y-single-operations-list>\n"
            }]
    }], function () { return []; }, { readOnly: [{
            type: Input
        }], onRetryFailedOperations: [{
            type: Output
        }], showFailedOperation: [{
            type: Output
        }], operation: [{
            type: Input
        }], bulkOperationModalDetailsService: [{
            type: Input
        }], detailsTemplate: [{
            type: ViewChild,
            args: ['details', { static: true }]
        }], auditTemplate: [{
            type: ViewChild,
            args: ['audit', { static: true }]
        }] }); })();
