// @c8y/ngx-components/esm2015/context-dashboard/cockpit-dashboard/cockpit-dashboard.component.js
import { Component, Inject, Optional } from '@angular/core';
import { AppStateService, gettext } from '@c8y/ngx-components';
import { COCKPIT_HOME_DASHBOARD_CONFIG, DEFAULT_COCKPIT_HOME_WIDGETS } from './cockpit-dashboard.model';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from '../context-dashboard.component';
export class CockpitDashboardComponent {
    constructor(appState, homeDashboardConfig) {
        var _a, _b;
        this.appState = appState;
        this.homeDashboardConfig = homeDashboardConfig;
        this.defaultWidgets = DEFAULT_COCKPIT_HOME_WIDGETS;
        this.dashboardName = 'home-cockpit1';
        this.pageTitle = gettext('Home');
        const config = Object.assign(Object.assign({}, this.homeDashboardConfig), { dashboardName: (_a = appState.currentApplication.value.config) === null || _a === void 0 ? void 0 : _a.homeDashboardName, userSpecific: (_b = appState.currentApplication.value.config) === null || _b === void 0 ? void 0 : _b.userSpecificHomeDashboard });
        this.applyDashboardConfig(config);
    }
    applyDashboardConfig(config) {
        if (!config) {
            return;
        }
        if (config.dashboardName) {
            this.dashboardName = config.dashboardName;
        }
        if (config.defaultWidgets) {
            this.defaultWidgets = config.defaultWidgets;
        }
        if (config.userSpecific && this.appState.currentUser.value) {
            this.dashboardName = `${this.appState.currentUser.value.userName.replace(/\./g, '__')}_${this.dashboardName}`;
        }
        if (config.pageTitle) {
            this.pageTitle = config.pageTitle;
        }
    }
}
CockpitDashboardComponent.ɵfac = function CockpitDashboardComponent_Factory(t) { return new (t || CockpitDashboardComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AppStateService), ɵngcc0.ɵɵdirectiveInject(COCKPIT_HOME_DASHBOARD_CONFIG, 8)); };
CockpitDashboardComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: CockpitDashboardComponent, selectors: [["c8y-cockpit-dashboard"]], decls: 5, vars: 6, consts: [["src", "/users-guide/cockpit/#dashboards"], [3, "name", "defaultWidgets", "canDelete"]], template: function CockpitDashboardComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵpipe(2, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(3, "c8y-help", 0);
        ɵngcc0.ɵɵelement(4, "c8y-context-dashboard", 1);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 4, ctx.pageTitle), "\n");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("name", ctx.dashboardName)("defaultWidgets", ctx.defaultWidgets)("canDelete", false);
    } }, directives: [ɵngcc1.TitleComponent, ɵngcc1.HelpComponent, ɵngcc2.ContextDashboardComponent], pipes: [ɵngcc1.C8yTranslatePipe], encapsulation: 2 });
CockpitDashboardComponent.ctorParameters = () => [
    { type: AppStateService },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [COCKPIT_HOME_DASHBOARD_CONFIG,] }] }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CockpitDashboardComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-cockpit-dashboard',
                template: "<c8y-title>\n  {{ pageTitle | translate }}\n</c8y-title>\n\n<c8y-help src=\"/users-guide/cockpit/#dashboards\"></c8y-help>\n\n<c8y-context-dashboard\n  [name]=\"dashboardName\"\n  [defaultWidgets]=\"defaultWidgets\"\n  [canDelete]=\"false\"\n></c8y-context-dashboard>\n"
            }]
    }], function () { return [{ type: ɵngcc1.AppStateService }, { type: undefined, decorators: [{
                type: Optional
            }, {
                type: Inject,
                args: [COCKPIT_HOME_DASHBOARD_CONFIG]
            }] }]; }, null); })();
