// @c8y/ngx-components/esm2015/core/header/application/app-switcher.component.js
import { __awaiter } from "tslib";
import { Component } from '@angular/core';
import { ApplicationService } from '@c8y/client';
import { get, includes, orderBy } from 'lodash-es';
import { combineLatest } from 'rxjs';
import { first } from 'rxjs/operators';
import { AppStateService } from '../../common/ui-state.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../common/ui-state.service';
import * as ɵngcc2 from '@c8y/client';
import * as ɵngcc3 from 'ngx-bootstrap/dropdown';
import * as ɵngcc4 from '../../common/icon.directive';
import * as ɵngcc5 from '@angular/common';
import * as ɵngcc6 from '../../product-experience/product-experience.directive';
import * as ɵngcc7 from './app-icon.component';
import * as ɵngcc8 from '../../i18n/c8y-translate.pipe';
import * as ɵngcc9 from '../../common/humanize-app-name.pipe';

const _c0 = function (a0) { return { applicationName: a0 }; };
function AppSwitcherComponent_div_4_div_1_a_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "a", 10);
    ɵngcc0.ɵɵpipe(1, "async");
    ɵngcc0.ɵɵpipe(2, "humanizeAppName");
    ɵngcc0.ɵɵelement(3, "c8y-app-icon", 11);
    ɵngcc0.ɵɵelementStart(4, "span", 12);
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "async");
    ɵngcc0.ɵɵpipe(7, "humanizeAppName");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const app_r4 = ctx.$implicit;
    const ctx_r3 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 8, ɵngcc0.ɵɵpipeBind1(2, 10, app_r4.name)));
    ɵngcc0.ɵɵproperty("href", ctx_r3.getHref(app_r4), ɵngcc0.ɵɵsanitizeUrl)("actionName", "appSwitched")("actionData", ɵngcc0.ɵɵpureFunction1(16, _c0, app_r4 && app_r4.name));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("app", app_r4)("name", app_r4.name)("contextPath", app_r4.contextPath);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(6, 12, ɵngcc0.ɵɵpipeBind1(7, 14, app_r4)));
} }
function AppSwitcherComponent_div_4_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 8);
    ɵngcc0.ɵɵtemplate(1, AppSwitcherComponent_div_4_div_1_a_1_Template, 8, 18, "a", 9);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r1.oneCloudApps);
} }
function AppSwitcherComponent_div_4_a_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "a", 13);
    ɵngcc0.ɵɵpipe(1, "async");
    ɵngcc0.ɵɵpipe(2, "humanizeAppName");
    ɵngcc0.ɵɵelement(3, "c8y-app-icon", 11);
    ɵngcc0.ɵɵelementStart(4, "span", 12);
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "async");
    ɵngcc0.ɵɵpipe(7, "humanizeAppName");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const app_r5 = ctx.$implicit;
    const ctx_r2 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 9, ɵngcc0.ɵɵpipeBind1(2, 11, app_r5.name)));
    ɵngcc0.ɵɵproperty("ngClass", ctx_r2.isActive(app_r5) ? "active" : "")("href", ctx_r2.getHref(app_r5), ɵngcc0.ɵɵsanitizeUrl)("actionName", "appSwitched")("actionData", ɵngcc0.ɵɵpureFunction1(17, _c0, app_r5 && app_r5.name));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("app", app_r5)("name", app_r5.name)("contextPath", app_r5.contextPath);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(6, 13, ɵngcc0.ɵɵpipeBind1(7, 15, app_r5)));
} }
function AppSwitcherComponent_div_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 4);
    ɵngcc0.ɵɵtemplate(1, AppSwitcherComponent_div_4_div_1_Template, 2, 1, "div", 5);
    ɵngcc0.ɵɵelementStart(2, "div", 6);
    ɵngcc0.ɵɵtemplate(3, AppSwitcherComponent_div_4_a_3_Template, 8, 19, "a", 7);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.oneCloudApps && ctx_r0.oneCloudApps.length > 0);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r0.apps);
} }
export class AppSwitcherComponent {
    constructor(ui, applications) {
        this.ui = ui;
        this.applications = applications;
        const { currentUser, currentTenant } = this.ui;
        combineLatest(currentUser, currentTenant)
            .pipe(first())
            .subscribe(([user, tenant]) => {
            this.tenant = tenant;
            this.listAppsForUser(user);
        });
    }
    get hideCloudApps() {
        return false; // this should be configurable from outside
    }
    isActive(app) {
        return app.contextPath === this.ui.state.app.contextPath;
    }
    getHref(app) {
        return this.applications.getHref(app);
    }
    listAppsForUser(user) {
        return __awaiter(this, void 0, void 0, function* () {
            if (user) {
                const { data } = yield this.applications.listByUser(user, { pageSize: 1000 });
                this.apps = this.orderApps(this.filterApps(data));
            }
        });
    }
    filterApps(apps) {
        apps = this.filterVisible(apps);
        apps = this.filterCloud(apps);
        apps = this.filterDuplicates(apps);
        return apps;
    }
    filterVisible(apps) {
        const visibleApplicationTypes = ['HOSTED', 'EXTERNAL'];
        return apps.filter(app => includes(visibleApplicationTypes, app.type) &&
            !app.noAppSwitcher &&
            !get(app, 'manifest.noAppSwitcher'));
    }
    filterCloud(apps) {
        const { hideCloudApps } = this;
        const filterFn = app => !!app.cloud;
        if (!hideCloudApps) {
            this.oneCloudApps = this.orderApps(apps.filter(filterFn));
        }
        return apps.filter(app => !filterFn(app));
    }
    filterDuplicates(apps) {
        // Filter out apps that have duplicate contextpaths and are not own owned but the current tenant;
        const filterFn = app => !apps.some(otherApp => app !== otherApp &&
            (app.contextPath && app.contextPath === otherApp.contextPath) &&
            app.owner.tenant.id !== this.tenant.name);
        return apps.filter(filterFn);
    }
    orderApps(apps) {
        return orderBy(apps, ({ name }) => name.toLowerCase());
    }
}
AppSwitcherComponent.ɵfac = function AppSwitcherComponent_Factory(t) { return new (t || AppSwitcherComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AppStateService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ApplicationService)); };
AppSwitcherComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AppSwitcherComponent, selectors: [["c8y-app-switcher"]], decls: 5, vars: 3, consts: [["dropdown", "", 1, "app-switcher-dropdown"], ["id", "appSwitcherDropdown", "dropdownToggle", "", 1, "main-header-button", "c8y-dropdown", "dropdown-toggle", 3, "title"], ["c8yIcon", "th", 1, "icon-2x"], ["class", "app-switcher-dropdown-menu dropdown-menu", "aria-labelledby", "appSwitcherDropdown", 4, "dropdownMenu"], ["aria-labelledby", "appSwitcherDropdown", 1, "app-switcher-dropdown-menu", "dropdown-menu"], ["class", "app-switcher-sagcloud", 4, "ngIf"], [1, "app-switcher"], ["class", "appLink", "c8yProductExperience", "", 3, "title", "ngClass", "href", "actionName", "actionData", 4, "ngFor", "ngForOf"], [1, "app-switcher-sagcloud"], ["class", "appLink", "c8yProductExperience", "", 3, "title", "href", "actionName", "actionData", 4, "ngFor", "ngForOf"], ["c8yProductExperience", "", 1, "appLink", 3, "title", "href", "actionName", "actionData"], [3, "app", "name", "contextPath"], [1, "text-truncate-wrap"], ["c8yProductExperience", "", 1, "appLink", 3, "title", "ngClass", "href", "actionName", "actionData"]], template: function AppSwitcherComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "button", 1);
        ɵngcc0.ɵɵpipe(2, "translate");
        ɵngcc0.ɵɵelement(3, "i", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(4, AppSwitcherComponent_div_4_Template, 4, 2, "div", 3);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 1, "Application switcher"));
    } }, directives: [ɵngcc3.BsDropdownDirective, ɵngcc3.BsDropdownToggleDirective, ɵngcc4.IconDirective, ɵngcc3.BsDropdownMenuDirective, ɵngcc5.NgIf, ɵngcc5.NgForOf, ɵngcc6.ProductExperienceDirective, ɵngcc7.AppIconComponent, ɵngcc5.NgClass], pipes: [ɵngcc8.C8yTranslatePipe, ɵngcc5.AsyncPipe, ɵngcc9.HumanizeAppNamePipe], encapsulation: 2 });
AppSwitcherComponent.ctorParameters = () => [
    { type: AppStateService },
    { type: ApplicationService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AppSwitcherComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-app-switcher',
                template: "<div class=\"app-switcher-dropdown\" dropdown>\n  <button\n    class=\"main-header-button c8y-dropdown dropdown-toggle\"\n    id=\"appSwitcherDropdown\"\n    dropdownToggle\n    title=\"{{ 'Application switcher' | translate }}\"\n  >\n    <i c8yIcon=\"th\" class=\"icon-2x\"></i>\n  </button>\n  <div\n    *dropdownMenu\n    class=\"app-switcher-dropdown-menu dropdown-menu\"\n    aria-labelledby=\"appSwitcherDropdown\"\n  >\n    <div class=\"app-switcher-sagcloud\" *ngIf=\"oneCloudApps && oneCloudApps.length > 0\">\n      <a\n        title=\"{{ app.name | humanizeAppName | async }}\"\n        class=\"appLink\"\n        *ngFor=\"let app of oneCloudApps\"\n        [href]=\"getHref(app)\"\n        c8yProductExperience\n        [actionName]=\"'appSwitched'\"\n        [actionData]=\"{ applicationName: app && app.name }\"\n      >\n        <c8y-app-icon [app]=\"app\" [name]=\"app.name\" [contextPath]=\"app.contextPath\"></c8y-app-icon>\n        <span class=\"text-truncate-wrap\">{{ app | humanizeAppName | async }}</span>\n      </a>\n    </div>\n    <div class=\"app-switcher\">\n      <a\n        title=\"{{ app.name | humanizeAppName | async }}\"\n        class=\"appLink\"\n        [ngClass]=\"isActive(app) ? 'active' : ''\"\n        *ngFor=\"let app of apps\"\n        [href]=\"getHref(app)\"\n        c8yProductExperience\n        [actionName]=\"'appSwitched'\"\n        [actionData]=\"{ applicationName: app && app.name }\"\n      >\n        <c8y-app-icon [app]=\"app\" [name]=\"app.name\" [contextPath]=\"app.contextPath\"></c8y-app-icon>\n        <span class=\"text-truncate-wrap\">{{ app | humanizeAppName | async }}</span>\n      </a>\n    </div>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.AppStateService }, { type: ɵngcc2.ApplicationService }]; }, null); })();
