// @c8y/ngx-components/esm2015/sigfox-device-registration/sigfox-device-registration.module.js
import { NgModule } from '@angular/core';
import { CommonModule, CoreModule } from '@c8y/ngx-components';
import { HOOK_DEVICE_REGISTRATION } from '@c8y/ngx-components/register-device';
import { SigfoxDeviceRegistrationButtonComponent } from './sigfox-device-registration-button.component';
import { SigfoxDeviceRegistrationComponent } from './sigfox-device-registration.component';
import { SigfoxDeviceRegistrationFactory } from './sigfox-device-registration.factory';
import { SigfoxProviderService } from './sigfox-provider.service';
import * as ɵngcc0 from '@angular/core';
export class SigfoxDeviceRegistrationModule {
}
SigfoxDeviceRegistrationModule.ɵfac = function SigfoxDeviceRegistrationModule_Factory(t) { return new (t || SigfoxDeviceRegistrationModule)(); };
SigfoxDeviceRegistrationModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: SigfoxDeviceRegistrationModule });
SigfoxDeviceRegistrationModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        {
            provide: HOOK_DEVICE_REGISTRATION,
            useClass: SigfoxDeviceRegistrationFactory,
            multi: true
        },
        SigfoxProviderService
    ], imports: [[CoreModule, CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SigfoxDeviceRegistrationModule, [{
        type: NgModule,
        args: [{
                imports: [CoreModule, CommonModule],
                declarations: [SigfoxDeviceRegistrationButtonComponent, SigfoxDeviceRegistrationComponent],
                providers: [
                    {
                        provide: HOOK_DEVICE_REGISTRATION,
                        useClass: SigfoxDeviceRegistrationFactory,
                        multi: true
                    },
                    SigfoxProviderService
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(SigfoxDeviceRegistrationModule, { declarations: function () { return [SigfoxDeviceRegistrationButtonComponent, SigfoxDeviceRegistrationComponent]; }, imports: function () { return [CoreModule, CommonModule]; } }); })();
