// @c8y/ngx-components/esm2015/sigfox-device-registration/sigfox-device-registration.factory.js
import { Injectable } from '@angular/core';
import { TenantUiService } from '@c8y/ngx-components';
import { SigfoxDeviceRegistrationButtonComponent } from './sigfox-device-registration-button.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
export class SigfoxDeviceRegistrationFactory {
    constructor(tenantService) {
        this.tenantService = tenantService;
    }
    get() {
        const items = [];
        if (this.tenantService.isMicroserviceSubscribedInCurrentTenant('sigfox-agent')) {
            items.push({
                template: SigfoxDeviceRegistrationButtonComponent,
                priority: 98,
                category: 'single'
            });
        }
        return items;
    }
}
SigfoxDeviceRegistrationFactory.ɵfac = function SigfoxDeviceRegistrationFactory_Factory(t) { return new (t || SigfoxDeviceRegistrationFactory)(ɵngcc0.ɵɵinject(ɵngcc1.TenantUiService)); };
SigfoxDeviceRegistrationFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: SigfoxDeviceRegistrationFactory, factory: SigfoxDeviceRegistrationFactory.ɵfac });
SigfoxDeviceRegistrationFactory.ctorParameters = () => [
    { type: TenantUiService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SigfoxDeviceRegistrationFactory, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.TenantUiService }]; }, null); })();
