// @c8y/ngx-components/esm2015/sigfox-device-registration/sigfox-device-registration-button.component.js
import { Component } from '@angular/core';
import { BsModalService } from 'ngx-bootstrap/modal';
import { SigfoxDeviceRegistrationComponent } from './sigfox-device-registration.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/modal';
import * as ɵngcc2 from '@c8y/ngx-components';
export class SigfoxDeviceRegistrationButtonComponent {
    constructor(modalService) {
        this.modalService = modalService;
    }
    open() {
        this.modalService.show(SigfoxDeviceRegistrationComponent, {
            class: 'modal-sm',
            ignoreBackdropClick: true
        });
    }
}
SigfoxDeviceRegistrationButtonComponent.ɵfac = function SigfoxDeviceRegistrationButtonComponent_Factory(t) { return new (t || SigfoxDeviceRegistrationButtonComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BsModalService)); };
SigfoxDeviceRegistrationButtonComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SigfoxDeviceRegistrationButtonComponent, selectors: [["c8y-sigfox-registration"]], decls: 4, vars: 3, consts: [[3, "click"], ["c8yIcon", "c8y-device-connect"]], template: function SigfoxDeviceRegistrationButtonComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "button", 0);
        ɵngcc0.ɵɵlistener("click", function SigfoxDeviceRegistrationButtonComponent_Template_button_click_0_listener() { return ctx.open(); });
        ɵngcc0.ɵɵelement(1, "i", 1);
        ɵngcc0.ɵɵtext(2);
        ɵngcc0.ɵɵpipe(3, "translate");
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 1, "Sigfox"), " ");
    } }, directives: [ɵngcc2.IconDirective], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
SigfoxDeviceRegistrationButtonComponent.ctorParameters = () => [
    { type: BsModalService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SigfoxDeviceRegistrationButtonComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-sigfox-registration',
                template: "<button (click)=\"open()\"><i c8yIcon=\"c8y-device-connect\"></i> {{ 'Sigfox' | translate }} </button>"
            }]
    }], function () { return [{ type: ɵngcc1.BsModalService }]; }, null); })();
