// @c8y/ngx-components/esm2015/repository/configuration/device-tab/configuration-repository-device-tab.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CoreModule, ViewContext, HOOK_ROUTE, gettext } from '@c8y/ngx-components';
import { SharedRepositoryModule } from '@c8y/ngx-components/repository/shared';
import { DeviceConfigurationGuard } from './device-configuration.guard';
import { DeviceConfigurationComponent } from './device-configuration.component';
import { ConfigurationPreviewComponent } from './configuration-preview.component';
import { ConfigurationFilterPipe } from './configuration-filter.pipe';
import { DeviceConfigurationService } from './device-configuration.service';
import { DeviceConfigurationListComponent } from './device-configuration-list.component';
import { TextBasedConfigurationComponent } from './text-based-configuration.component';
import { SaveToRepositoryComponent } from './save-to-repository.component';
import { SourceCodePreviewComponent } from './source-code-preview.component';
import { TabsModule } from 'ngx-bootstrap/tabs';
import { OperationDetailsModule } from '@c8y/ngx-components/operations/operation-details';
export class ConfigurationRepositoryDeviceTabModule {
    static forRoot() {
        return {
            ngModule: ConfigurationRepositoryDeviceTabModule,
            providers: [
                DeviceConfigurationGuard,
                {
                    provide: HOOK_ROUTE,
                    useValue: [
                        {
                            context: ViewContext.Device,
                            path: 'device-configuration',
                            component: DeviceConfigurationComponent,
                            label: gettext('Configuration'),
                            icon: 'gears',
                            priority: 600,
                            canActivate: [DeviceConfigurationGuard]
                        }
                    ],
                    multi: true
                }
            ]
        };
    }
}
ConfigurationRepositoryDeviceTabModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    CoreModule,
                    SharedRepositoryModule,
                    OperationDetailsModule,
                    TabsModule.forRoot()
                ],
                declarations: [
                    DeviceConfigurationComponent,
                    DeviceConfigurationListComponent,
                    ConfigurationPreviewComponent,
                    ConfigurationFilterPipe,
                    SaveToRepositoryComponent,
                    SourceCodePreviewComponent,
                    TextBasedConfigurationComponent
                ],
                entryComponents: [
                    DeviceConfigurationComponent,
                    SaveToRepositoryComponent,
                    TextBasedConfigurationComponent
                ],
                providers: [DeviceConfigurationService]
            },] }
];
