// @c8y/ngx-components/esm2015/ecosystem/applications/application-plugins/plugin-list-item.component.js
import { Component, EventEmitter, Input, Output } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from '@angular/common';

function PluginListItemComponent_c8y_li_checkbox_1_Template(rf, ctx) { if (rf & 1) {
    const _r4 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li-checkbox", 10);
    ɵngcc0.ɵɵlistener("change", function PluginListItemComponent_c8y_li_checkbox_1_Template_c8y_li_checkbox_change_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r4); const ctx_r3 = ɵngcc0.ɵɵnextContext(); return ctx_r3.onChange($event.target.checked); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("disabled", ctx_r0.plugin.installed);
} }
function PluginListItemComponent_span_13_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵelement(1, "i", 11);
    ɵngcc0.ɵɵelementStart(2, "em", 12);
    ɵngcc0.ɵɵtext(3, "Installed`plugins`");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yIcon", "check-circle");
} }
function PluginListItemComponent_span_16_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 13);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r2.plugin.contextPath);
} }
const _c0 = function (a0) { return { disabled: a0 }; };
export class PluginListItemComponent {
    constructor() {
        this.isItemSelected = new EventEmitter();
    }
    onChange(event) {
        this.isItemSelected.next(event);
    }
}
PluginListItemComponent.ɵfac = function PluginListItemComponent_Factory(t) { return new (t || PluginListItemComponent)(); };
PluginListItemComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: PluginListItemComponent, selectors: [["c8y-plugin-list-item"]], inputs: { plugin: "plugin", selectable: "selectable" }, outputs: { isItemSelected: "isItemSelected" }, decls: 17, vars: 10, consts: [[1, "bg-inherit", 3, "ngClass"], ["class", "p-r-0 p-l-0", 3, "disabled", "change", 4, "ngIf"], [1, "c8y-plugin-icon"], [1, "d-flex"], [1, "p-r-8"], [1, "text-medium"], [1, "text-muted", "small", "m-l-8"], [4, "ngIf"], [1, "small", "l-h-tight"], ["class", "label label-info m-l-auto a-s-start", 4, "ngIf"], [1, "p-r-0", "p-l-0", 3, "disabled", "change"], [1, "text-success", 3, "c8yIcon"], ["translate", "", 1, "text-muted", "small"], [1, "label", "label-info", "m-l-auto", "a-s-start"]], template: function PluginListItemComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-li", 0);
        ɵngcc0.ɵɵtemplate(1, PluginListItemComponent_c8y_li_checkbox_1_Template, 1, 1, "c8y-li-checkbox", 1);
        ɵngcc0.ɵɵelementStart(2, "c8y-li-icon");
        ɵngcc0.ɵɵelementStart(3, "i", 2);
        ɵngcc0.ɵɵelementStart(4, "span");
        ɵngcc0.ɵɵtext(5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "c8y-li-body", 3);
        ɵngcc0.ɵɵelementStart(7, "div", 4);
        ɵngcc0.ɵɵelementStart(8, "p");
        ɵngcc0.ɵɵelementStart(9, "span", 5);
        ɵngcc0.ɵɵtext(10);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "em", 6);
        ɵngcc0.ɵɵtext(12);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(13, PluginListItemComponent_span_13_Template, 4, 1, "span", 7);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(14, "p", 8);
        ɵngcc0.ɵɵtext(15);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(16, PluginListItemComponent_span_16_Template, 2, 1, "span", 9);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(8, _c0, ctx.plugin.installed));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.selectable);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate(ctx.plugin.name == null ? null : ctx.plugin.name.substr(0, 2));
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵtextInterpolate(ctx.plugin.name);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate(ctx.plugin.version);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.plugin.installed);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate(ctx.plugin.description);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.selectable);
    } }, directives: [ɵngcc1.ListItemComponent, ɵngcc2.NgClass, ɵngcc2.NgIf, ɵngcc1.ListItemIconComponent, ɵngcc1.ListItemBodyComponent, ɵngcc1.ListItemCheckboxComponent, ɵngcc1.IconDirective, ɵngcc1.C8yTranslateDirective], encapsulation: 2 });
PluginListItemComponent.propDecorators = {
    plugin: [{ type: Input }],
    selectable: [{ type: Input }],
    isItemSelected: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(PluginListItemComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-plugin-list-item',
                template: "<c8y-li [ngClass]=\"{ disabled: plugin.installed }\" class=\"bg-inherit\">\n  <c8y-li-checkbox\n    (change)=\"onChange($event.target.checked)\"\n    *ngIf=\"selectable\"\n    [disabled]=\"plugin.installed\"\n    class=\"p-r-0 p-l-0\"\n  ></c8y-li-checkbox>\n  <c8y-li-icon>\n    <i class=\"c8y-plugin-icon\">\n      <span>{{ plugin.name?.substr(0, 2) }}</span>\n    </i>\n  </c8y-li-icon>\n  <c8y-li-body class=\"d-flex\">\n    <div class=\"p-r-8\">\n      <p>\n        <span class=\"text-medium\">{{ plugin.name }}</span>\n        <em class=\"text-muted small m-l-8\">{{ plugin.version }}</em>\n        <span *ngIf=\"plugin.installed\">\n          <i [c8yIcon]=\"'check-circle'\" class=\"text-success\"></i>\n          <em class=\"text-muted small\" translate>Installed`plugins`</em>\n        </span>\n      </p>\n      <p class=\"small l-h-tight\">{{ plugin.description }}</p>\n    </div>\n    <span *ngIf=\"selectable\" class=\"label label-info m-l-auto a-s-start\">{{\n      plugin.contextPath\n    }}</span>\n  </c8y-li-body>\n</c8y-li>\n"
            }]
    }], function () { return []; }, { isItemSelected: [{
            type: Output
        }], plugin: [{
            type: Input
        }], selectable: [{
            type: Input
        }] }); })();
