// @c8y/ngx-components/esm2015/register-device/extensible/single/extensible-device-registration-modal.component.js
import { __awaiter } from "tslib";
import { Component } from '@angular/core';
import { C8yJSONSchema, gettext } from '@c8y/ngx-components';
import { ExtensibleDeviceRegistrationService } from './extensible-device-registration.service';
import { BsModalRef } from 'ngx-bootstrap/modal';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/modal';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from './extensible-device-registration.service';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from './extensible-device-registration-stepper.component';

function ExtensibleDeviceRegistrationModalComponent_ng_container_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "c8y-extensible-device-registration-stepper", 5);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const steps_r3 = ctx.ngIf;
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("steps", steps_r3)("labels", ctx_r0.labels)("contextPath", ctx_r0.contextPath);
} }
function ExtensibleDeviceRegistrationModalComponent_ng_template_7_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    const _r8 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "c8y-operation-result", 7);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementStart(3, "div", 8);
    ɵngcc0.ɵɵelementStart(4, "button", 9);
    ɵngcc0.ɵɵlistener("click", function ExtensibleDeviceRegistrationModalComponent_ng_template_7_ng_container_0_Template_button_click_4_listener() { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r7 = ɵngcc0.ɵɵnextContext(2); return ctx_r7.bsModalRef.hide(); });
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵtext(6, " Close ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("text", ɵngcc0.ɵɵpipeBind1(2, 4, ctx_r4.loadingErrorMessage));
    ɵngcc0.ɵɵproperty("size", 84)("vertical", true);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(5, 6, "Close"));
} }
function ExtensibleDeviceRegistrationModalComponent_ng_template_7_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 10);
    ɵngcc0.ɵɵelement(1, "c8y-progress-bar", 11);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("message", ɵngcc0.ɵɵpipeBind1(2, 1, "Loading the registration form definition\u2026"));
} }
function ExtensibleDeviceRegistrationModalComponent_ng_template_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ExtensibleDeviceRegistrationModalComponent_ng_template_7_ng_container_0_Template, 7, 8, "ng-container", 3);
    ɵngcc0.ɵɵtemplate(1, ExtensibleDeviceRegistrationModalComponent_ng_template_7_ng_template_1_Template, 3, 3, "ng-template", null, 6, ɵngcc0.ɵɵtemplateRefExtractor);
} if (rf & 2) {
    const _r5 = ɵngcc0.ɵɵreference(2);
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.loadingErrorMessage)("ngIfElse", _r5);
} }
export class ExtensibleDeviceRegistrationModalComponent {
    constructor(bsModalRef, jsonschema, registrationService) {
        this.bsModalRef = bsModalRef;
        this.jsonschema = jsonschema;
        this.registrationService = registrationService;
        this.labels = [];
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            const { title, pages } = yield this.registrationService.getSchema(this.contextPath)
                .catch(error => {
                this.loadingErrorMessage = (error === null || error === void 0 ? void 0 : error.message) || gettext('Cannot fetch a form definition from the microservice.');
                return {
                    title: gettext('Error')
                };
            });
            this.title = title;
            this.pages = pages === null || pages === void 0 ? void 0 : pages.map((step) => {
                this.labels.push(step.title);
                return [this.jsonschema.toFieldConfig(step)];
            });
        });
    }
}
ExtensibleDeviceRegistrationModalComponent.ɵfac = function ExtensibleDeviceRegistrationModalComponent_Factory(t) { return new (t || ExtensibleDeviceRegistrationModalComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BsModalRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.C8yJSONSchema), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.ExtensibleDeviceRegistrationService)); };
ExtensibleDeviceRegistrationModalComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ExtensibleDeviceRegistrationModalComponent, selectors: [["extensible-device-registration-modal"]], decls: 9, vars: 7, consts: [[3, "headerClasses", "customFooter"], ["c8y-modal-title", ""], [1, "c8y-icon", "c8y-icon-device-connect"], [4, "ngIf", "ngIfElse"], ["loadingOrError", ""], [3, "steps", "labels", "contextPath"], ["loading", ""], ["type", "error", 1, "lead", "text-regular", 3, "text", "size", "vertical"], [1, "modal-footer"], ["type", "button", "translate", "", 1, "btn", "btn-default", 3, "title", "click"], [1, "modal-header", "modal-header-info", "text-center"], [1, "text-center", 3, "message"]], template: function ExtensibleDeviceRegistrationModalComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-modal", 0);
        ɵngcc0.ɵɵelementContainerStart(1, 1);
        ɵngcc0.ɵɵelement(2, "span", 2);
        ɵngcc0.ɵɵelementStart(3, "h4");
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementContainerEnd();
        ɵngcc0.ɵɵtemplate(6, ExtensibleDeviceRegistrationModalComponent_ng_container_6_Template, 2, 3, "ng-container", 3);
        ɵngcc0.ɵɵtemplate(7, ExtensibleDeviceRegistrationModalComponent_ng_template_7_Template, 3, 2, "ng-template", null, 4, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r1 = ɵngcc0.ɵɵreference(8);
        ɵngcc0.ɵɵproperty("headerClasses", "dialog-header")("customFooter", true);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(5, 5, ctx.title));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.pages)("ngIfElse", _r1);
    } }, directives: [ɵngcc2.ModalComponent, ɵngcc4.NgIf, ɵngcc5.ExtensibleDeviceRegistrationStepperComponent, ɵngcc2.OperationResultComponent, ɵngcc2.C8yTranslateDirective, ɵngcc2.ProgressBarComponent], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
ExtensibleDeviceRegistrationModalComponent.ctorParameters = () => [
    { type: BsModalRef },
    { type: C8yJSONSchema },
    { type: ExtensibleDeviceRegistrationService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ExtensibleDeviceRegistrationModalComponent, [{
        type: Component,
        args: [{
                selector: 'extensible-device-registration-modal',
                template: "<c8y-modal [headerClasses]=\"'dialog-header'\" [customFooter]=\"true\">\n  <ng-container c8y-modal-title>\n    <span class=\"c8y-icon c8y-icon-device-connect\"></span>\n    <h4>{{ title | translate }}</h4>\n  </ng-container>\n  <ng-container *ngIf=\"pages as steps; else loadingOrError\">\n    <c8y-extensible-device-registration-stepper\n      [steps]=\"steps\"\n      [labels]=\"labels\"\n      [contextPath]=\"contextPath\"\n    ></c8y-extensible-device-registration-stepper>\n  </ng-container>\n  <ng-template #loadingOrError>\n    <ng-container *ngIf=\"loadingErrorMessage; else loading\">\n      <c8y-operation-result\n        text=\"{{ loadingErrorMessage | translate }}\"\n        [size]=\"84\"\n        [vertical]=\"true\"\n        type=\"error\"\n        class=\"lead text-regular\"\n      ></c8y-operation-result>\n      <div class=\"modal-footer\">\n        <button\n          title=\"{{ 'Close' | translate }}\"\n          (click)=\"bsModalRef.hide()\"\n          type=\"button\"\n          class=\"btn btn-default\"\n          translate\n        >\n          Close\n        </button>\n      </div>\n    </ng-container>\n    <ng-template #loading>\n      <div class=\"modal-header modal-header-info text-center\">\n        <c8y-progress-bar\n          [message]=\"'Loading the registration form definition\u2026' | translate\"\n          class=\"text-center\"\n        ></c8y-progress-bar>\n      </div>\n    </ng-template>\n  </ng-template>\n</c8y-modal>\n"
            }]
    }], function () { return [{ type: ɵngcc1.BsModalRef }, { type: ɵngcc2.C8yJSONSchema }, { type: ɵngcc3.ExtensibleDeviceRegistrationService }]; }, null); })();
