// @c8y/ngx-components/esm2015/register-device/extensible/single/extensible-device-registration-stepper.component.js
import { __awaiter } from "tslib";
import { Component, Input, ViewChild } from '@angular/core';
import { C8yStepper, gettext } from '@c8y/ngx-components';
import { FormGroup } from '@angular/forms';
import { BsModalRef } from 'ngx-bootstrap/modal';
import { ExtensibleDeviceRegistrationService } from './extensible-device-registration.service';
import { RegisterDeviceService } from '../../register-device.service';
import { STEP_STATE } from '@angular/cdk/stepper';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './extensible-device-registration.service';
import * as ɵngcc2 from '../../register-device.service';
import * as ɵngcc3 from 'ngx-bootstrap/modal';
import * as ɵngcc4 from '@c8y/ngx-components';
import * as ɵngcc5 from '@angular/common';
import * as ɵngcc6 from '@angular/cdk/stepper';
import * as ɵngcc7 from '@ngx-formly/core';

const _c0 = function () { return { cancel: true, back: true, next: true }; };
const _c1 = function () { return { next: "Register" }; };
function ExtensibleDeviceRegistrationStepperComponent_cdk_step_1_c8y_stepper_buttons_3_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-stepper-buttons", 11);
    ɵngcc0.ɵɵlistener("onCancel", function ExtensibleDeviceRegistrationStepperComponent_cdk_step_1_c8y_stepper_buttons_3_Template_c8y_stepper_buttons_onCancel_0_listener() { ɵngcc0.ɵɵrestoreView(_r9); const ctx_r8 = ɵngcc0.ɵɵnextContext(2); return ctx_r8.close(); })("onNext", function ExtensibleDeviceRegistrationStepperComponent_cdk_step_1_c8y_stepper_buttons_3_Template_c8y_stepper_buttons_onNext_0_listener() { ɵngcc0.ɵɵrestoreView(_r9); const ctx_r10 = ɵngcc0.ɵɵnextContext(2); return ctx_r10.save(); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const i_r5 = ɵngcc0.ɵɵnextContext().index;
    const ctx_r6 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("showButtons", ɵngcc0.ɵɵpureFunction0(4, _c0))("labels", ɵngcc0.ɵɵpureFunction0(5, _c1))("pending", ctx_r6.pendingStatus)("disabled", ctx_r6.forms[i_r5].invalid);
} }
const _c2 = function (a1) { return { cancel: true, back: a1, next: true }; };
function ExtensibleDeviceRegistrationStepperComponent_cdk_step_1_c8y_stepper_buttons_4_Template(rf, ctx) { if (rf & 1) {
    const _r13 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-stepper-buttons", 12);
    ɵngcc0.ɵɵlistener("onCancel", function ExtensibleDeviceRegistrationStepperComponent_cdk_step_1_c8y_stepper_buttons_4_Template_c8y_stepper_buttons_onCancel_0_listener() { ɵngcc0.ɵɵrestoreView(_r13); const ctx_r12 = ɵngcc0.ɵɵnextContext(2); return ctx_r12.close(); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const i_r5 = ɵngcc0.ɵɵnextContext().index;
    const ctx_r7 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("showButtons", ɵngcc0.ɵɵpureFunction1(2, _c2, i_r5 !== 0))("disabled", ctx_r7.forms[i_r5].invalid);
} }
function ExtensibleDeviceRegistrationStepperComponent_cdk_step_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "cdk-step", 6);
    ɵngcc0.ɵɵelementStart(1, "c8y-form-group", 7);
    ɵngcc0.ɵɵelement(2, "formly-form", 8);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(3, ExtensibleDeviceRegistrationStepperComponent_cdk_step_1_c8y_stepper_buttons_3_Template, 1, 6, "c8y-stepper-buttons", 9);
    ɵngcc0.ɵɵtemplate(4, ExtensibleDeviceRegistrationStepperComponent_cdk_step_1_c8y_stepper_buttons_4_Template, 1, 4, "c8y-stepper-buttons", 10);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const step_r4 = ctx.$implicit;
    const i_r5 = ctx.index;
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("stepControl", ctx_r0.forms[i_r5])("label", ctx_r0.labels[i_r5]);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("form", ctx_r0.forms[i_r5])("fields", step_r4)("model", ctx_r0.model);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.isRegistrationStep());
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r0.isRegistrationStep());
} }
function ExtensibleDeviceRegistrationStepperComponent_c8y_operation_result_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-operation-result", 13);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("text", ɵngcc0.ɵɵpipeBind1(1, 3, ctx_r1.message));
    ɵngcc0.ɵɵproperty("size", 84)("vertical", true);
} }
function ExtensibleDeviceRegistrationStepperComponent_ng_template_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-operation-result", 14);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("text", ɵngcc0.ɵɵpipeBind1(1, 3, ctx_r3.errorMessage));
    ɵngcc0.ɵɵproperty("size", 84)("vertical", true);
} }
const _c3 = function () { return { edit: true, done: false }; };
const _c4 = function () { return ["m-40", "m-t-16", "m-b-16"]; };
const _c5 = function (a0, a1, a2) { return { cancel: a0, back: a1, next: a2 }; };
const _c6 = function () { return { back: "Fix issue", next: "Close" }; };
export class ExtensibleDeviceRegistrationStepperComponent {
    constructor(customProviderService, registerDeviceService, bsModalRef) {
        this.customProviderService = customProviderService;
        this.registerDeviceService = registerDeviceService;
        this.bsModalRef = bsModalRef;
        this.pendingStatus = false;
        this.forms = [];
        this.model = {};
        this.SERVER_ERROR = gettext('Cannot register your device.');
        this.SUCCESSFUL_REGISTRATION = gettext('Your device was successfully registered.');
        this.NO_CONNECTION = gettext('Unable to reach the microservice.');
    }
    ngOnInit() {
        this.steps.forEach(step => this.forms.push(new FormGroup({})));
    }
    save() {
        return __awaiter(this, void 0, void 0, function* () {
            this.pendingStatus = true;
            this.errorMessage = null;
            const { res, data } = yield this.customProviderService.registerDevice(this.contextPath, this.model)
                .catch(err => {
                return {
                    res: undefined,
                    data: undefined
                };
            })
                .finally(() => {
                this.pendingStatus = false;
                this.stepper.next();
            });
            if (res && data) {
                if (res.status >= 400 && res.status < 500) {
                    this.handleError(data.message || this.SERVER_ERROR);
                }
                else if (res.status >= 500) {
                    this.handleError(this.SERVER_ERROR);
                }
                else {
                    this.message = this.SUCCESSFUL_REGISTRATION;
                }
            }
            else {
                this.handleError(this.NO_CONNECTION);
            }
        });
    }
    close() {
        this.bsModalRef.hide();
    }
    complete() {
        this.registerDeviceService.list();
        this.bsModalRef.hide();
    }
    goToFirstStep() {
        this.stepper.selectedIndex = 0;
    }
    isRegistrationStep() {
        return this.stepper.selectedIndex === this.steps.length - 1;
    }
    finalStepStatus() {
        if (this.stepper.selectedIndex === this.steps.length) {
            if (this.errorMessage) {
                return STEP_STATE.ERROR;
            }
            else {
                return STEP_STATE.DONE;
            }
        }
        else {
            return STEP_STATE.NUMBER;
        }
    }
    handleError(message) {
        this.errorMessage = message;
    }
}
ExtensibleDeviceRegistrationStepperComponent.ɵfac = function ExtensibleDeviceRegistrationStepperComponent_Factory(t) { return new (t || ExtensibleDeviceRegistrationStepperComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ExtensibleDeviceRegistrationService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.RegisterDeviceService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.BsModalRef)); };
ExtensibleDeviceRegistrationStepperComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ExtensibleDeviceRegistrationStepperComponent, selectors: [["c8y-extensible-device-registration-stepper"]], viewQuery: function ExtensibleDeviceRegistrationStepperComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(C8yStepper, 7);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.stepper = _t.first);
    } }, inputs: { contextPath: "contextPath", labels: "labels", steps: "steps" }, decls: 8, vars: 19, consts: [["linear", "", 3, "disableProgressButtons", "disableDefaultIcons", "customClasses"], [3, "stepControl", "label", 4, "ngFor", "ngForOf"], [3, "state", "label"], ["type", "success", "class", "lead", 3, "text", "size", "vertical", 4, "ngIf", "ngIfElse"], ["errorInfo", ""], [1, "sticky-bottom", "d-block", "p-t-16", "p-b-16", "separator-top", "bg-white", 3, "showButtons", "labels", "onCancel", "onNext", "onBack"], [3, "stepControl", "label"], [1, "form-group", "p-24", "p-b-0", "p-t-16"], [3, "form", "fields", "model"], ["class", "sticky-bottom d-block p-t-16 p-b-16 separator-top bg-white", 3, "showButtons", "labels", "pending", "disabled", "onCancel", "onNext", 4, "ngIf"], ["class", "sticky-bottom d-block p-t-16 p-b-16 separator-top bg-white", 3, "showButtons", "disabled", "onCancel", 4, "ngIf"], [1, "sticky-bottom", "d-block", "p-t-16", "p-b-16", "separator-top", "bg-white", 3, "showButtons", "labels", "pending", "disabled", "onCancel", "onNext"], [1, "sticky-bottom", "d-block", "p-t-16", "p-b-16", "separator-top", "bg-white", 3, "showButtons", "disabled", "onCancel"], ["type", "success", 1, "lead", 3, "text", "size", "vertical"], ["type", "error", 1, "lead", 3, "text", "size", "vertical"]], template: function ExtensibleDeviceRegistrationStepperComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-stepper", 0);
        ɵngcc0.ɵɵtemplate(1, ExtensibleDeviceRegistrationStepperComponent_cdk_step_1_Template, 5, 7, "cdk-step", 1);
        ɵngcc0.ɵɵelementStart(2, "cdk-step", 2);
        ɵngcc0.ɵɵpipe(3, "translate");
        ɵngcc0.ɵɵtemplate(4, ExtensibleDeviceRegistrationStepperComponent_c8y_operation_result_4_Template, 2, 5, "c8y-operation-result", 3);
        ɵngcc0.ɵɵtemplate(5, ExtensibleDeviceRegistrationStepperComponent_ng_template_5_Template, 2, 5, "ng-template", null, 4, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementStart(7, "c8y-stepper-buttons", 5);
        ɵngcc0.ɵɵlistener("onCancel", function ExtensibleDeviceRegistrationStepperComponent_Template_c8y_stepper_buttons_onCancel_7_listener() { return ctx.close(); })("onNext", function ExtensibleDeviceRegistrationStepperComponent_Template_c8y_stepper_buttons_onNext_7_listener() { return ctx.complete(); })("onBack", function ExtensibleDeviceRegistrationStepperComponent_Template_c8y_stepper_buttons_onBack_7_listener() { return ctx.goToFirstStep(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r2 = ɵngcc0.ɵɵreference(6);
        ɵngcc0.ɵɵproperty("disableProgressButtons", true)("disableDefaultIcons", ɵngcc0.ɵɵpureFunction0(12, _c3))("customClasses", ɵngcc0.ɵɵpureFunction0(13, _c4));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.steps);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("label", ɵngcc0.ɵɵpipeBind1(3, 10, "Summary"));
        ɵngcc0.ɵɵproperty("state", ctx.finalStepStatus());
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.errorMessage)("ngIfElse", _r2);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("showButtons", ɵngcc0.ɵɵpureFunction3(14, _c5, ctx.errorMessage, ctx.errorMessage, !ctx.errorMessage))("labels", ɵngcc0.ɵɵpureFunction0(18, _c6));
    } }, directives: [ɵngcc4.C8yStepper, ɵngcc5.NgForOf, ɵngcc6.CdkStep, ɵngcc5.NgIf, ɵngcc4.C8yStepperButtons, ɵngcc4.FormGroupComponent, ɵngcc7.FormlyForm, ɵngcc4.OperationResultComponent], pipes: [ɵngcc4.C8yTranslatePipe], encapsulation: 2 });
ExtensibleDeviceRegistrationStepperComponent.ctorParameters = () => [
    { type: ExtensibleDeviceRegistrationService },
    { type: RegisterDeviceService },
    { type: BsModalRef }
];
ExtensibleDeviceRegistrationStepperComponent.propDecorators = {
    stepper: [{ type: ViewChild, args: [C8yStepper, { static: true },] }],
    contextPath: [{ type: Input }],
    labels: [{ type: Input }],
    steps: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ExtensibleDeviceRegistrationStepperComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-extensible-device-registration-stepper',
                template: "<c8y-stepper\n  [disableProgressButtons]=\"true\"\n  [disableDefaultIcons]=\"{ edit: true, done: false }\"\n  [customClasses]=\"['m-40', 'm-t-16', 'm-b-16']\"\n  linear\n>\n  <cdk-step *ngFor=\"let step of steps; let i = index\" [stepControl]=\"forms[i]\" [label]=\"labels[i]\">\n    <c8y-form-group class=\"form-group p-24 p-b-0 p-t-16\">\n      <formly-form [form]=\"forms[i]\" [fields]=\"step\" [model]=\"model\"></formly-form>\n    </c8y-form-group>\n    <c8y-stepper-buttons\n      class=\"sticky-bottom d-block p-t-16 p-b-16 separator-top bg-white\"\n      *ngIf=\"isRegistrationStep()\"\n      [showButtons]=\"{ cancel: true, back: true, next: true }\"\n      [labels]=\"{ next: 'Register' }\"\n      (onCancel)=\"close()\"\n      (onNext)=\"save()\"\n      [pending]=\"pendingStatus\"\n      [disabled]=\"forms[i].invalid\"\n    ></c8y-stepper-buttons>\n    <c8y-stepper-buttons\n      class=\"sticky-bottom d-block p-t-16 p-b-16 separator-top bg-white\"\n      *ngIf=\"!isRegistrationStep()\"\n      [showButtons]=\"{ cancel: true, back: i !== 0, next: true }\"\n      [disabled]=\"forms[i].invalid\"\n      (onCancel)=\"close()\"\n    ></c8y-stepper-buttons>\n  </cdk-step>\n  <cdk-step [state]=\"finalStepStatus()\" label=\"{{ 'Summary' | translate }}\">\n    <c8y-operation-result\n      *ngIf=\"!errorMessage; else errorInfo\"\n      text=\"{{ message | translate }}\"\n      [size]=\"84\"\n      [vertical]=\"true\"\n      type=\"success\"\n      class=\"lead\"\n    ></c8y-operation-result>\n    <ng-template #errorInfo>\n      <c8y-operation-result\n        text=\"{{ errorMessage | translate }}\"\n        [size]=\"84\"\n        [vertical]=\"true\"\n        type=\"error\"\n        class=\"lead\"\n      ></c8y-operation-result>\n    </ng-template>\n    <c8y-stepper-buttons\n      class=\"sticky-bottom d-block p-t-16 p-b-16 separator-top bg-white\"\n      [showButtons]=\"{ cancel: errorMessage, back: errorMessage, next: !errorMessage }\"\n      [labels]=\"{ back: 'Fix issue', next: 'Close' }\"\n      (onCancel)=\"close()\"\n      (onNext)=\"complete()\"\n      (onBack)=\"goToFirstStep()\"\n    ></c8y-stepper-buttons>\n  </cdk-step>\n</c8y-stepper>\n"
            }]
    }], function () { return [{ type: ɵngcc1.ExtensibleDeviceRegistrationService }, { type: ɵngcc2.RegisterDeviceService }, { type: ɵngcc3.BsModalRef }]; }, { stepper: [{
            type: ViewChild,
            args: [C8yStepper, { static: true }]
        }], contextPath: [{
            type: Input
        }], labels: [{
            type: Input
        }], steps: [{
            type: Input
        }] }); })();
