// @c8y/ngx-components/esm2015/context-dashboard/device-info-dashboard/device-info-dashboard.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { CoreModule, HOOK_ROUTE, ViewContext, gettext } from '@c8y/ngx-components';
import { ContextDashboardModule } from '../context-dashboard.module';
import { DeviceInfoDashboardComponent } from './device-info-dashboard.component';
import * as ɵngcc0 from '@angular/core';
const ɵ0 = [
    {
        context: ViewContext.Device,
        path: 'device-info',
        component: DeviceInfoDashboardComponent,
        label: gettext('Info'),
        priority: 10000,
        icon: 'asterisk'
    }
];
export class DeviceInfoDashboardModule {
}
DeviceInfoDashboardModule.ɵfac = function DeviceInfoDashboardModule_Factory(t) { return new (t || DeviceInfoDashboardModule)(); };
DeviceInfoDashboardModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DeviceInfoDashboardModule });
DeviceInfoDashboardModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        {
            provide: HOOK_ROUTE,
            useValue: ɵ0,
            multi: true
        }
    ], imports: [[CommonModule, CoreModule, ContextDashboardModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DeviceInfoDashboardModule, [{
        type: NgModule,
        args: [{
                declarations: [DeviceInfoDashboardComponent],
                imports: [CommonModule, CoreModule, ContextDashboardModule],
                exports: [],
                entryComponents: [DeviceInfoDashboardComponent],
                providers: [
                    {
                        provide: HOOK_ROUTE,
                        useValue: ɵ0,
                        multi: true
                    }
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DeviceInfoDashboardModule, { declarations: function () { return [DeviceInfoDashboardComponent]; }, imports: function () { return [CommonModule, CoreModule, ContextDashboardModule]; } }); })();
export { ɵ0 };
