// @c8y/ngx-components/esm2015/datapoint-selector/pipes/datapoint-label.pipe.js
import { Pipe } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class DatapointLabelPipe {
    transform(datapoint, options = { includeDevice: false, doNotUseLabel: false }) {
        var _a, _b, _c, _d;
        let label = '';
        try {
            if (!(options === null || options === void 0 ? void 0 : options.doNotUseLabel) && datapoint.label) {
                label = datapoint.label;
            }
            else {
                label = `${datapoint.fragment} → ${datapoint.series}`;
            }
            if ((options === null || options === void 0 ? void 0 : options.includeDevice) && (((_a = datapoint.__target) === null || _a === void 0 ? void 0 : _a.name) || ((_b = datapoint.__target) === null || _b === void 0 ? void 0 : _b.id))) {
                label += ` (${((_c = datapoint.__target) === null || _c === void 0 ? void 0 : _c.name) || ((_d = datapoint.__target) === null || _d === void 0 ? void 0 : _d.id)})`;
            }
        }
        catch (e) {
            // nothing to do
        }
        return label;
    }
}
DatapointLabelPipe.ɵfac = function DatapointLabelPipe_Factory(t) { return new (t || DatapointLabelPipe)(); };
DatapointLabelPipe.ɵpipe = /*@__PURE__*/ ɵngcc0.ɵɵdefinePipe({ name: "datapointLabel", type: DatapointLabelPipe, pure: true });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DatapointLabelPipe, [{
        type: Pipe,
        args: [{
                name: 'datapointLabel'
            }]
    }], null, null); })();
