// @c8y/ngx-components/esm2015/datapoint-selector/pipes/includes-datapoint.pipe.js
import { Pipe } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class IncludesDatapointPipe {
    transform(datapointList, datapoint) {
        if (!Array.isArray(datapointList) || !datapoint) {
            return false;
        }
        return datapointList.some(tmp => {
            var _a, _b;
            return tmp.fragment === datapoint.fragment &&
                tmp.series === datapoint.series &&
                ((_a = tmp.__target) === null || _a === void 0 ? void 0 : _a.id) === ((_b = datapoint.__target) === null || _b === void 0 ? void 0 : _b.id);
        });
    }
}
IncludesDatapointPipe.ɵfac = function IncludesDatapointPipe_Factory(t) { return new (t || IncludesDatapointPipe)(); };
IncludesDatapointPipe.ɵpipe = /*@__PURE__*/ ɵngcc0.ɵɵdefinePipe({ name: "includesDatapoint", type: IncludesDatapointPipe, pure: true });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(IncludesDatapointPipe, [{
        type: Pipe,
        args: [{
                name: 'includesDatapoint'
            }]
    }], null, null); })();
