// @c8y/ngx-components/esm2015/operations/operations-timeline/operations-timeline.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { CoreModule } from '@c8y/ngx-components';
import { OperationsSharedModule } from '@c8y/ngx-components/operations/shared';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { OperationsTimelineComponent } from './operations-timeline.component';
import * as ɵngcc0 from '@angular/core';
export class OperationsTimelineModule {
}
OperationsTimelineModule.ɵfac = function OperationsTimelineModule_Factory(t) { return new (t || OperationsTimelineModule)(); };
OperationsTimelineModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: OperationsTimelineModule });
OperationsTimelineModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule, CoreModule, CollapseModule, TooltipModule, OperationsSharedModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OperationsTimelineModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, CoreModule, CollapseModule, TooltipModule, OperationsSharedModule],
                declarations: [OperationsTimelineComponent],
                exports: [OperationsTimelineComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(OperationsTimelineModule, { declarations: function () { return [OperationsTimelineComponent]; }, imports: function () { return [CommonModule, CoreModule, CollapseModule, TooltipModule, OperationsSharedModule]; }, exports: function () { return [OperationsTimelineComponent]; } }); })();
