// @c8y/ngx-components/esm2015/core/header/header-bar/header-bar.component.js
import { Component, Input } from '@angular/core';
import { ActionService } from '../../action/action.service';
import { BreadcrumbService } from '../../breadcrumb/breadcrumb.service';
import { SearchService } from '../../search/search.service';
import { AppStateService } from '../../common/ui-state.service';
import { UserMenuService } from '../../user/user-menu.service';
import { HeaderService } from '../header.service';
import { DocsService } from '../../docs/docs.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../header.service';
import * as ɵngcc2 from '../../action/action.service';
import * as ɵngcc3 from '../../breadcrumb/breadcrumb.service';
import * as ɵngcc4 from '../../search/search.service';
import * as ɵngcc5 from '../../common/ui-state.service';
import * as ɵngcc6 from '../../user/user-menu.service';
import * as ɵngcc7 from '../../docs/docs.service';
import * as ɵngcc8 from '@angular/common';
import * as ɵngcc9 from '../application/app-icon.component';
import * as ɵngcc10 from '../title/title-outlet.component';
import * as ɵngcc11 from '../application/app-switcher.component';
import * as ɵngcc12 from '../../user/user-menu-outlet.component';
import * as ɵngcc13 from '../../common/icon.directive';
import * as ɵngcc14 from '../../breadcrumb/breadcrumb-outlet.component';
import * as ɵngcc15 from '../../search/search-outlet.component';
import * as ɵngcc16 from '../../action/action-outlet.component';
import * as ɵngcc17 from '../right-drawer/right-drawer.component';
import * as ɵngcc18 from '../../i18n/c8y-translate.pipe';

function HeaderBarComponent_button_3_i_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 14);
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("c8yIcon", "outdent");
} }
function HeaderBarComponent_button_3_i_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 14);
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("c8yIcon", "dedent-right");
} }
function HeaderBarComponent_button_3_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 17);
    ɵngcc0.ɵɵlistener("click", function HeaderBarComponent_button_3_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r9); const ctx_r8 = ɵngcc0.ɵɵnextContext(); return ctx_r8.headerService.toggleNavigator(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtemplate(2, HeaderBarComponent_button_3_i_2_Template, 1, 1, "i", 18);
    ɵngcc0.ɵɵpipe(3, "async");
    ɵngcc0.ɵɵtemplate(4, HeaderBarComponent_button_3_i_4_Template, 1, 1, "i", 18);
    ɵngcc0.ɵɵpipe(5, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 3, "Toggle navigator"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !ɵngcc0.ɵɵpipeBind1(3, 5, ctx_r0.headerService.navigatorOpen$));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(5, 7, ctx_r0.headerService.navigatorOpen$));
} }
function HeaderBarComponent_c8y_breadcrumb_outlet_11_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-breadcrumb-outlet", 19);
    ɵngcc0.ɵɵpipe(1, "async");
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("breadcrumbs", ɵngcc0.ɵɵpipeBind1(1, 1, ctx_r1.breadcrumbService.items$));
} }
function HeaderBarComponent_c8y_search_outlet_12_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-search-outlet", 20);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵpipe(2, "async");
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, "Search"));
    ɵngcc0.ɵɵproperty("search", ɵngcc0.ɵɵpipeBind1(2, 4, ctx_r2.searchService.items$));
} }
function HeaderBarComponent_c8y_action_outlet_13_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-action-outlet", 21);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵpipe(2, "async");
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, "Add"));
    ɵngcc0.ɵɵproperty("items", ɵngcc0.ɵɵpipeBind1(2, 4, ctx_r3.actionService.items$));
} }
function HeaderBarComponent_button_18_i_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 14);
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("c8yIcon", "indent");
} }
function HeaderBarComponent_button_18_i_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 14);
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("c8yIcon", "dedent");
} }
function HeaderBarComponent_button_18_Template(rf, ctx) { if (rf & 1) {
    const _r13 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 22);
    ɵngcc0.ɵɵlistener("click", function HeaderBarComponent_button_18_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r13); const ctx_r12 = ɵngcc0.ɵɵnextContext(); return ctx_r12.drawerOpen = !ctx_r12.drawerOpen; });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtemplate(2, HeaderBarComponent_button_18_i_2_Template, 1, 1, "i", 18);
    ɵngcc0.ɵɵtemplate(3, HeaderBarComponent_button_18_i_3_Template, 1, 1, "i", 18);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 3, "Toggle side drawer"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r4.drawerOpen);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r4.drawerOpen);
} }
function HeaderBarComponent_c8y_right_drawer_24_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-right-drawer", 23);
    ɵngcc0.ɵɵpipe(1, "async");
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵpipe(3, "async");
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("app", ɵngcc0.ɵɵpipeBind1(1, 3, ctx_r5.appState$).app)("quickLinks", ɵngcc0.ɵɵpipeBind1(2, 5, ctx_r5.docs.items$))("helpAndSupport", ɵngcc0.ɵɵpipeBind1(3, 7, ctx_r5.docs.items$));
} }
const _c0 = function (a0, a1) { return { open: a0, drawerOpen: a1 }; };
const _c1 = function (a0) { return { active: a0 }; };
export class HeaderBarComponent {
    constructor(headerService, actionService, breadcrumbService, searchService, appState, userMenuService, docs) {
        this.headerService = headerService;
        this.actionService = actionService;
        this.breadcrumbService = breadcrumbService;
        this.searchService = searchService;
        this.appState = appState;
        this.userMenuService = userMenuService;
        this.docs = docs;
        this.canToggle$ = this.headerService.hasItemsInNavigator$;
        this.appState$ = this.appState.state$;
        this.drawerOpen = false;
        this.headerOpen = false;
        /**
         * Can be set to show an header bar with only
         * title, app-switcher and user-menu
         */
        this.simple = false;
    }
    ngAfterViewInit() {
        this.headerService.verifyIfNavOpen();
    }
}
HeaderBarComponent.ɵfac = function HeaderBarComponent_Factory(t) { return new (t || HeaderBarComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.HeaderService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ActionService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.BreadcrumbService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.SearchService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.AppStateService), ɵngcc0.ɵɵdirectiveInject(ɵngcc6.UserMenuService), ɵngcc0.ɵɵdirectiveInject(ɵngcc7.DocsService)); };
HeaderBarComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: HeaderBarComponent, selectors: [["c8y-header-bar"]], inputs: { simple: "simple" }, decls: 28, vars: 39, consts: [[1, "app-main-header", 3, "ngClass"], [1, "header-bar"], ["type", "button", "class", "navigator-toggle main-header-button", 3, "title", "click", 4, "ngIf"], [1, "app-view"], [3, "name", "contextPath"], [1, "page-header"], ["class", "app-breadcrumbs", 3, "breadcrumbs", 4, "ngIf"], ["class", "main-header-item", 3, "title", "search", 4, "ngIf"], [3, "title", "items", 4, "ngIf"], [1, "main-header-item", 3, "title"], [1, "main-header-item", 3, "items"], ["class", "drawer-toggle main-header-button", 3, "title", "click", 4, "ngIf"], [1, "head-toggler"], ["type", "button", 3, "title", "click"], [3, "c8yIcon"], [3, "app", "quickLinks", "helpAndSupport", 4, "ngIf"], [1, "loading-bar", 3, "ngClass"], ["type", "button", 1, "navigator-toggle", "main-header-button", 3, "title", "click"], [3, "c8yIcon", 4, "ngIf"], [1, "app-breadcrumbs", 3, "breadcrumbs"], [1, "main-header-item", 3, "title", "search"], [3, "title", "items"], [1, "drawer-toggle", "main-header-button", 3, "title", "click"], [3, "app", "quickLinks", "helpAndSupport"]], template: function HeaderBarComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵpipe(1, "async");
        ɵngcc0.ɵɵelementStart(2, "div", 1);
        ɵngcc0.ɵɵtemplate(3, HeaderBarComponent_button_3_Template, 6, 9, "button", 2);
        ɵngcc0.ɵɵpipe(4, "async");
        ɵngcc0.ɵɵelementStart(5, "div", 3);
        ɵngcc0.ɵɵelement(6, "c8y-app-icon", 4);
        ɵngcc0.ɵɵpipe(7, "async");
        ɵngcc0.ɵɵpipe(8, "async");
        ɵngcc0.ɵɵelementStart(9, "span", 5);
        ɵngcc0.ɵɵelement(10, "c8y-title-outlet");
        ɵngcc0.ɵɵtemplate(11, HeaderBarComponent_c8y_breadcrumb_outlet_11_Template, 2, 3, "c8y-breadcrumb-outlet", 6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(12, HeaderBarComponent_c8y_search_outlet_12_Template, 3, 6, "c8y-search-outlet", 7);
        ɵngcc0.ɵɵtemplate(13, HeaderBarComponent_c8y_action_outlet_13_Template, 3, 6, "c8y-action-outlet", 8);
        ɵngcc0.ɵɵelement(14, "c8y-app-switcher", 9);
        ɵngcc0.ɵɵpipe(15, "translate");
        ɵngcc0.ɵɵelement(16, "c8y-user-menu-outlet", 10);
        ɵngcc0.ɵɵpipe(17, "async");
        ɵngcc0.ɵɵtemplate(18, HeaderBarComponent_button_18_Template, 4, 5, "button", 11);
        ɵngcc0.ɵɵpipe(19, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(20, "div", 12);
        ɵngcc0.ɵɵelementStart(21, "button", 13);
        ɵngcc0.ɵɵlistener("click", function HeaderBarComponent_Template_button_click_21_listener() { return ctx.headerService.toggle(); });
        ɵngcc0.ɵɵpipe(22, "translate");
        ɵngcc0.ɵɵelement(23, "i", 14);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(24, HeaderBarComponent_c8y_right_drawer_24_Template, 4, 9, "c8y-right-drawer", 15);
        ɵngcc0.ɵɵpipe(25, "async");
        ɵngcc0.ɵɵelement(26, "div", 16);
        ɵngcc0.ɵɵpipe(27, "async");
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(34, _c0, ɵngcc0.ɵɵpipeBind1(1, 14, ctx.headerService.navigatorOpen$) && !ctx.simple, ctx.drawerOpen));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(4, 16, ctx.canToggle$) && !ctx.simple);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("name", ɵngcc0.ɵɵpipeBind1(7, 18, ctx.appState$).app.name)("contextPath", ɵngcc0.ɵɵpipeBind1(8, 20, ctx.appState$).app.contextPath);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.simple);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.simple);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.simple);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(15, 22, "Application switcher"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("items", ɵngcc0.ɵɵpipeBind1(17, 24, ctx.userMenuService.state$).userMenuItems);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(19, 26, ctx.appState$).showRightDrawer && !ctx.simple);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(22, 28, "Toggle"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yIcon", "angle-right");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(25, 30, ctx.appState$).showRightDrawer);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(37, _c1, ɵngcc0.ɵɵpipeBind1(27, 32, ctx.appState$).isLoading));
    } }, directives: [ɵngcc8.NgClass, ɵngcc8.NgIf, ɵngcc9.AppIconComponent, ɵngcc10.TitleOutletComponent, ɵngcc11.AppSwitcherComponent, ɵngcc12.UserMenuOutletComponent, ɵngcc13.IconDirective, ɵngcc14.BreadcrumbOutletComponent, ɵngcc15.SearchOutletComponent, ɵngcc16.ActionOutletComponent, ɵngcc17.RightDrawerComponent], pipes: [ɵngcc8.AsyncPipe, ɵngcc18.C8yTranslatePipe], encapsulation: 2 });
HeaderBarComponent.ctorParameters = () => [
    { type: HeaderService },
    { type: ActionService },
    { type: BreadcrumbService },
    { type: SearchService },
    { type: AppStateService },
    { type: UserMenuService },
    { type: DocsService }
];
HeaderBarComponent.propDecorators = {
    simple: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(HeaderBarComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-header-bar',
                template: "<div\n  class=\"app-main-header\"\n  [ngClass]=\"{\n    open: (headerService.navigatorOpen$ | async) && !simple,\n    drawerOpen: drawerOpen\n  }\"\n>\n  <div class=\"header-bar\">\n    <button\n      *ngIf=\"(canToggle$ | async) && !simple\"\n      type=\"button\"\n      class=\"navigator-toggle main-header-button\"\n      title=\"{{ 'Toggle navigator' | translate }}\"\n      (click)=\"headerService.toggleNavigator()\"\n    >\n      <i [c8yIcon]=\"'outdent'\" *ngIf=\"!(headerService.navigatorOpen$ | async)\"></i>\n      <i [c8yIcon]=\"'dedent-right'\" *ngIf=\"headerService.navigatorOpen$ | async\"></i>\n    </button>\n    <div class=\"app-view\">\n      <c8y-app-icon\n        [name]=\"(appState$ | async).app.name\"\n        [contextPath]=\"(appState$ | async).app.contextPath\"\n      ></c8y-app-icon>\n\n      <span class=\"page-header\">\n        <c8y-title-outlet></c8y-title-outlet>\n        <c8y-breadcrumb-outlet\n          *ngIf=\"!simple\"\n          [breadcrumbs]=\"breadcrumbService.items$ | async\"\n          class=\"app-breadcrumbs\"\n        ></c8y-breadcrumb-outlet>\n      </span>\n    </div>\n    <c8y-search-outlet\n      *ngIf=\"!simple\"\n      title=\"{{ 'Search' | translate }}\"\n      [search]=\"searchService.items$ | async\"\n      class=\"main-header-item\"\n    ></c8y-search-outlet>\n    <c8y-action-outlet\n      *ngIf=\"!simple\"\n      title=\"{{ 'Add' | translate }}\"\n      [items]=\"actionService.items$ | async\"\n    ></c8y-action-outlet>\n    <c8y-app-switcher\n      title=\"{{ 'Application switcher' | translate }}\"\n      class=\"main-header-item\"\n    ></c8y-app-switcher>\n    <c8y-user-menu-outlet\n      [items]=\"(userMenuService.state$ | async).userMenuItems\"\n      class=\"main-header-item\"\n    ></c8y-user-menu-outlet>\n\n    <button\n      title=\"{{ 'Toggle side drawer' | translate }}\"\n      class=\"drawer-toggle main-header-button\"\n      *ngIf=\"(appState$ | async).showRightDrawer && !simple\"\n      (click)=\"drawerOpen = !drawerOpen\"\n    >\n      <i [c8yIcon]=\"'indent'\" *ngIf=\"drawerOpen\"></i>\n      <i [c8yIcon]=\"'dedent'\" *ngIf=\"!drawerOpen\"></i>\n    </button>\n  </div>\n  <div class=\"head-toggler\">\n    <button title=\"{{ 'Toggle' | translate }}\" type=\"button\" (click)=\"headerService.toggle()\">\n      <i [c8yIcon]=\"'angle-right'\"></i>\n    </button>\n  </div>\n  <c8y-right-drawer\n    *ngIf=\"(appState$ | async).showRightDrawer\"\n    [app]=\"(appState$ | async).app\"\n    [quickLinks]=\"docs.items$ | async\"\n    [helpAndSupport]=\"docs.items$ | async\"\n  >\n  </c8y-right-drawer>\n  <div class=\"loading-bar\" [ngClass]=\"{ active: (appState$ | async).isLoading }\"></div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.HeaderService }, { type: ɵngcc2.ActionService }, { type: ɵngcc3.BreadcrumbService }, { type: ɵngcc4.SearchService }, { type: ɵngcc5.AppStateService }, { type: ɵngcc6.UserMenuService }, { type: ɵngcc7.DocsService }]; }, { simple: [{
            type: Input
        }] }); })();
