// @c8y/ngx-components/esm2015/core/provider-configuration/provider-configuration.component.js
import { __awaiter } from "tslib";
import { Component } from '@angular/core';
import { FormGroup } from '@angular/forms';
import { ActivatedRoute } from '@angular/router';
import { ɵdefineHiddenProp } from '@ngx-formly/core';
import { find, forOwn, get, mapValues, pick } from 'lodash-es';
import { BehaviorSubject, combineLatest, from, merge, of, Subject } from 'rxjs';
import { catchError, map, shareReplay, switchMap, tap } from 'rxjs/operators';
import { AlertService } from '../alert/alert.service';
import { Permissions, Status } from '../common/index';
import { C8yJSONSchema } from '../dynamic-forms/json-schema/c8y-json-schema.service';
import { ModalService } from '../modal/modal.service';
import { ProviderConfigurationService } from './service/provider-configuration.service';
import { ProviderDefinitionsService } from './service/provider-definitions.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../common/index';
import * as ɵngcc2 from '@angular/router';
import * as ɵngcc3 from '../modal/modal.service';
import * as ɵngcc4 from '../alert/alert.service';
import * as ɵngcc5 from './service/provider-definitions.service';
import * as ɵngcc6 from './service/provider-configuration.service';
import * as ɵngcc7 from '../dynamic-forms/json-schema/c8y-json-schema.service';
import * as ɵngcc8 from '../header/title/title.component';
import * as ɵngcc9 from '@angular/forms';
import * as ɵngcc10 from '@angular/common';
import * as ɵngcc11 from '../forms/form-group.component';
import * as ɵngcc12 from '../select/typeahead.component';
import * as ɵngcc13 from '../forms/messages.component';
import * as ɵngcc14 from '../forms/message.directive';
import * as ɵngcc15 from '../common/if-allowed.directive';
import * as ɵngcc16 from '../list-group/list-item.component';
import * as ɵngcc17 from '../search/highlight.component';
import * as ɵngcc18 from '@ngx-formly/core';
import * as ɵngcc19 from '../i18n/c8y-translate.pipe';

function ProviderConfigurationComponent_p_14_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p", 14);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵpipe(3, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    let tmp_0_0;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, (tmp_0_0 = ɵngcc0.ɵɵpipeBind1(3, 3, ctx_r0.layout$)) == null ? null : tmp_0_0.description), " ");
} }
function ProviderConfigurationComponent_c8y_li_26_Template(rf, ctx) { if (rf & 1) {
    const _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 15);
    ɵngcc0.ɵɵlistener("click", function ProviderConfigurationComponent_c8y_li_26_Template_c8y_li_click_0_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r6); const provider_r4 = restoredCtx.$implicit; const ctx_r5 = ɵngcc0.ɵɵnextContext(); ctx_r5.changeProvider$.next(provider_r4); return ctx_r5.providerInput$.next(""); });
    ɵngcc0.ɵɵpipe(1, "async");
    ɵngcc0.ɵɵelement(2, "c8y-highlight", 16);
    ɵngcc0.ɵɵpipe(3, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const provider_r4 = ctx.$implicit;
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("active", ɵngcc0.ɵɵpipeBind1(1, 3, ctx_r1.selectedProvider$) === provider_r4);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("text", provider_r4.displayName || "--")("pattern", ɵngcc0.ɵɵpipeBind1(3, 5, ctx_r1.providerInput$));
} }
function ProviderConfigurationComponent_formly_form_32_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "formly-form", 17);
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("form", ctx_r2.form)("fields", ctx_r2.fields)("model", ctx_r2.model)("options", ctx_r2.options);
} }
function ProviderConfigurationComponent_div_34_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 21);
    ɵngcc0.ɵɵlistener("click", function ProviderConfigurationComponent_div_34_button_1_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r10); const ctx_r9 = ɵngcc0.ɵɵnextContext(2); return ctx_r9.deleteProviderConfiguration(); });
    ɵngcc0.ɵɵpipe(1, "async");
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵpipe(4, "async");
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵpipe(7, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext(2);
    let tmp_0_0;
    let tmp_1_0;
    let tmp_2_0;
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(3, 7, (tmp_0_0 = ɵngcc0.ɵɵpipeBind1(4, 9, ctx_r7.layout$)) == null ? null : tmp_0_0.deleteBtnLabel));
    ɵngcc0.ɵɵproperty("disabled", !((tmp_1_0 = ɵngcc0.ɵɵpipeBind1(1, 3, ctx_r7.configuration$)) == null ? null : tmp_1_0.provider) && !((tmp_1_0 = ɵngcc0.ɵɵpipeBind1(2, 5, ctx_r7.configuration$)) == null ? null : tmp_1_0.providerName));
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(6, 11, (tmp_2_0 = ɵngcc0.ɵɵpipeBind1(7, 13, ctx_r7.layout$)) == null ? null : tmp_2_0.deleteBtnLabel), " ");
} }
function ProviderConfigurationComponent_div_34_button_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "button", 22);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵpipe(5, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r8 = ɵngcc0.ɵɵnextContext(2);
    let tmp_0_0;
    let tmp_2_0;
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 3, (tmp_0_0 = ɵngcc0.ɵɵpipeBind1(2, 5, ctx_r8.layout$)) == null ? null : tmp_0_0.saveBtnLabel));
    ɵngcc0.ɵɵproperty("disabled", ctx_r8.form.invalid || ctx_r8.form.pristine);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 7, (tmp_2_0 = ɵngcc0.ɵɵpipeBind1(5, 9, ctx_r8.layout$)) == null ? null : tmp_2_0.saveBtnLabel), " ");
} }
function ProviderConfigurationComponent_div_34_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 18);
    ɵngcc0.ɵɵtemplate(1, ProviderConfigurationComponent_div_34_button_1_Template, 8, 15, "button", 19);
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵtemplate(3, ProviderConfigurationComponent_div_34_button_3_Template, 6, 11, "button", 20);
    ɵngcc0.ɵɵpipe(4, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    let tmp_0_0;
    let tmp_1_0;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yIfAllowed", (tmp_0_0 = ɵngcc0.ɵɵpipeBind1(2, 2, ctx_r3.layout$)) == null ? null : tmp_0_0.deleteRoles);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIfAllowed", (tmp_1_0 = ɵngcc0.ɵɵpipeBind1(4, 4, ctx_r3.layout$)) == null ? null : tmp_1_0.saveRoles);
} }
const _c0 = function () { return []; };
export class ProviderConfigurationComponent {
    constructor(permissions, activatedRoute, modalService, alertService, providerDefinitionsService, providerConfigurationService, jsonschema) {
        this.permissions = permissions;
        this.activatedRoute = activatedRoute;
        this.modalService = modalService;
        this.alertService = alertService;
        this.providerDefinitionsService = providerDefinitionsService;
        this.providerConfigurationService = providerConfigurationService;
        this.jsonschema = jsonschema;
        this.layout$ = this.activatedRoute.data.pipe(map((config) => config.layout), tap((layout) => (this.layout = layout)), tap((layout) => {
            this.options.formState.disabled = !this.permissions.hasAllRoles(layout.saveRoles || []);
            this.beforeSaveHook = layout.beforeSaveHook;
        }));
        this.allRoles$ = this.layout$.pipe(map((layout) => [
            ...(layout.deleteRoles || []),
            ...(layout.saveRoles || [])
        ]));
        this.changeProvider$ = new BehaviorSubject(null);
        this.providerInput$ = new BehaviorSubject('');
        this.form = new FormGroup({});
        this.fields = [];
        this.options = {
            formState: {
                disabled: false
            }
        };
        this.reload$ = new BehaviorSubject(null);
        this.updatedConfiguration$ = new Subject();
    }
    ngOnInit() {
        const allProviders$ = from(this.providerDefinitionsService.list()).pipe(map(result => result.data), shareReplay(1));
        this.providers$ = combineLatest(allProviders$, this.providerInput$).pipe(map(([providers, input]) => input
            ? providers.filter(el => el.displayName.toLowerCase().indexOf(input.toLowerCase()) >= 0)
            : providers), shareReplay(1));
        this.configuration$ = merge(this.updatedConfiguration$, this.reload$.pipe(switchMap(() => from(this.providerConfigurationService.detail()).pipe(catchError(() => of({})))), map(result => result.data))).pipe(map(this.removeEncryptedValues), shareReplay(1));
        this.selectedProvider$ = combineLatest(allProviders$, this.configuration$, this.changeProvider$).pipe(tap(([_, configuration, newProvider]) => (this.model = newProvider
            ? pick(this.model, 'sms.senderName', 'sms.senderAddress')
            : configuration)), map(([providers, configuration, newProvider]) => newProvider ||
            find(providers, (provider) => get(configuration, 'provider') === provider.id)), tap((provider) => {
            if (provider) {
                const config = this.jsonschema.toFieldConfig(get(provider, 'schema'));
                if (config.fieldGroup) {
                    config.fieldGroup.forEach((fieldConfig) => {
                        ɵdefineHiddenProp(fieldConfig, '_keyPath', {
                            key: fieldConfig.key,
                            path: [fieldConfig.key]
                        });
                        fieldConfig.expressionProperties = {
                            'templateOptions.disabled': 'formState.disabled'
                        };
                    });
                }
                this.fields = [config];
                this.form = new FormGroup({});
            }
        }), shareReplay(1));
    }
    saveProviderConfiguration() {
        return __awaiter(this, void 0, void 0, function* () {
            const modelToSave = !!this.beforeSaveHook
                ? yield this.beforeSaveHook(this.model, this.fields)
                : this.model;
            forOwn(modelToSave, (value, key) => {
                if (Array.isArray(value)) {
                    modelToSave[key] = value.filter(item => !!item || item === 0);
                }
            });
            try {
                const res = yield this.providerConfigurationService.update(modelToSave);
                this.changeProvider$.next(null);
                this.updatedConfiguration$.next(res.data);
                this.alertService.success(this.layout.configurationUpdatedSuccessMsg);
                this.form.markAsPristine();
            }
            catch (err) {
                this.alertService.addServerFailure(err);
            }
        });
    }
    deleteProviderConfiguration() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.modalService.confirm(this.layout.deleteConfigurationModalTitle, this.layout.deleteConfigurationModalBody, Status.DANGER, {
                    ok: this.layout.deleteConfigurationModalOkBtnLabel,
                    cancel: this.layout.deleteConfigurationModalCancelBtnLabel
                });
                yield this.providerConfigurationService.delete();
                this.reload$.next();
                this.alertService.success(this.layout.configurationDeletedSuccessMsg);
            }
            catch (err) {
                if (err) {
                    this.alertService.addServerFailure(err);
                }
            }
        });
    }
    removeEncryptedValues(configuration) {
        return mapValues(configuration, value => (value === '<<Encrypted>>' ? undefined : value));
    }
}
ProviderConfigurationComponent.ɵfac = function ProviderConfigurationComponent_Factory(t) { return new (t || ProviderConfigurationComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Permissions), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ActivatedRoute), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.ModalService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.ProviderDefinitionsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc6.ProviderConfigurationService), ɵngcc0.ɵɵdirectiveInject(ɵngcc7.C8yJSONSchema)); };
ProviderConfigurationComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ProviderConfigurationComponent, selectors: [["c8y-sms-gateway"]], features: [ɵngcc0.ɵɵProvidersFeature([ProviderConfigurationService, ProviderDefinitionsService])], decls: 36, vars: 48, consts: [[1, "row"], [1, "col-md-8", "col-xs-12"], [1, "card", "card--fullpage", 3, "ngSubmit"], [1, "card-header", "separator"], [1, "card-title"], [1, "inner-scroll"], [1, "card-block"], ["class", "m-b-8", 4, "ngIf"], ["for", "providerName"], ["name", "providerName", 3, "disabled", "ngModel", "displayProperty", "placeholder", "allowFreeEntries", "required", "container", "onSearch"], ["class", "p-l-8 p-r-8 c8y-list__item--link", 3, "active", "click", 4, "ngFor", "ngForOf"], ["name", "notExisting", 3, "text"], [3, "form", "fields", "model", "options", 4, "ngIf"], ["class", "card-footer separator", 4, "c8yIfAllowedAllowAny", "c8yIfAllowed"], [1, "m-b-8"], [1, "p-l-8", "p-r-8", "c8y-list__item--link", 3, "active", "click"], [3, "text", "pattern"], [3, "form", "fields", "model", "options"], [1, "card-footer", "separator"], ["class", "btn btn-default", "type", "button", 3, "disabled", "title", "click", 4, "c8yIfAllowed"], ["class", "btn btn-primary", "type", "submit", 3, "disabled", "title", 4, "c8yIfAllowed"], ["type", "button", 1, "btn", "btn-default", 3, "disabled", "title", "click"], ["type", "submit", 1, "btn", "btn-primary", 3, "disabled", "title"]], template: function ProviderConfigurationComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵpipe(2, "translate");
        ɵngcc0.ɵɵpipe(3, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "div", 0);
        ɵngcc0.ɵɵelementStart(5, "div", 1);
        ɵngcc0.ɵɵelementStart(6, "form", 2);
        ɵngcc0.ɵɵlistener("ngSubmit", function ProviderConfigurationComponent_Template_form_ngSubmit_6_listener() { return ctx.saveProviderConfiguration(); });
        ɵngcc0.ɵɵelementStart(7, "div", 3);
        ɵngcc0.ɵɵelementStart(8, "h4", 4);
        ɵngcc0.ɵɵtext(9);
        ɵngcc0.ɵɵpipe(10, "translate");
        ɵngcc0.ɵɵpipe(11, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(12, "div", 5);
        ɵngcc0.ɵɵelementStart(13, "div", 6);
        ɵngcc0.ɵɵtemplate(14, ProviderConfigurationComponent_p_14_Template, 4, 5, "p", 7);
        ɵngcc0.ɵɵpipe(15, "async");
        ɵngcc0.ɵɵelementStart(16, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(17, "label", 8);
        ɵngcc0.ɵɵtext(18);
        ɵngcc0.ɵɵpipe(19, "translate");
        ɵngcc0.ɵɵpipe(20, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(21, "c8y-typeahead", 9);
        ɵngcc0.ɵɵlistener("onSearch", function ProviderConfigurationComponent_Template_c8y_typeahead_onSearch_21_listener($event) { return ctx.providerInput$.next($event); });
        ɵngcc0.ɵɵpipe(22, "async");
        ɵngcc0.ɵɵpipe(23, "async");
        ɵngcc0.ɵɵpipe(24, "translate");
        ɵngcc0.ɵɵpipe(25, "async");
        ɵngcc0.ɵɵtemplate(26, ProviderConfigurationComponent_c8y_li_26_Template, 4, 7, "c8y-li", 10);
        ɵngcc0.ɵɵpipe(27, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(28, "c8y-messages");
        ɵngcc0.ɵɵelement(29, "c8y-message", 11);
        ɵngcc0.ɵɵpipe(30, "translate");
        ɵngcc0.ɵɵpipe(31, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(32, ProviderConfigurationComponent_formly_form_32_Template, 1, 4, "formly-form", 12);
        ɵngcc0.ɵɵpipe(33, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(34, ProviderConfigurationComponent_div_34_Template, 5, 6, "div", 13);
        ɵngcc0.ɵɵpipe(35, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        let tmp_0_0;
        let tmp_1_0;
        let tmp_2_0;
        let tmp_3_0;
        let tmp_4_0;
        let tmp_5_0;
        let tmp_12_0;
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 15, (tmp_0_0 = ɵngcc0.ɵɵpipeBind1(3, 17, ctx.layout$)) == null ? null : tmp_0_0.pageTitle), "\n");
        ɵngcc0.ɵɵadvance(8);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(10, 19, (tmp_1_0 = ɵngcc0.ɵɵpipeBind1(11, 21, ctx.layout$)) == null ? null : tmp_1_0.cardTitle), " ");
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("ngIf", !!((tmp_2_0 = ɵngcc0.ɵɵpipeBind1(15, 23, ctx.layout$)) == null ? null : tmp_2_0.description));
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(19, 25, (tmp_3_0 = ɵngcc0.ɵɵpipeBind1(20, 27, ctx.layout$)) == null ? null : tmp_3_0.providerName));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(24, 33, (tmp_4_0 = ɵngcc0.ɵɵpipeBind1(25, 35, ctx.layout$)) == null ? null : tmp_4_0.providerNamePlaceholder));
        ɵngcc0.ɵɵproperty("disabled", !ctx.permissions.hasAllRoles(((tmp_5_0 = ɵngcc0.ɵɵpipeBind1(22, 29, ctx.layout$)) == null ? null : tmp_5_0.saveRoles) || ɵngcc0.ɵɵpureFunction0(47, _c0)))("ngModel", ɵngcc0.ɵɵpipeBind1(23, 31, ctx.selectedProvider$))("displayProperty", "displayName")("allowFreeEntries", false)("required", true)("container", "body");
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("ngForOf", ɵngcc0.ɵɵpipeBind1(27, 37, ctx.providers$));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("text", ɵngcc0.ɵɵpipeBind1(30, 39, (tmp_12_0 = ɵngcc0.ɵɵpipeBind1(31, 41, ctx.layout$)) == null ? null : tmp_12_0.providerNameNoMatchesHint));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(33, 43, ctx.selectedProvider$));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yIfAllowed", ɵngcc0.ɵɵpipeBind1(35, 45, ctx.allRoles$));
    } }, directives: [ɵngcc8.TitleComponent, ɵngcc9.ɵNgNoValidate, ɵngcc9.NgControlStatusGroup, ɵngcc9.NgForm, ɵngcc10.NgIf, ɵngcc11.FormGroupComponent, ɵngcc12.TypeaheadComponent, ɵngcc9.NgControlStatus, ɵngcc9.NgModel, ɵngcc9.RequiredValidator, ɵngcc10.NgForOf, ɵngcc13.MessagesComponent, ɵngcc14.MessageDirective, ɵngcc15.IfAllowedDirective, ɵngcc16.ListItemComponent, ɵngcc17.HighlightComponent, ɵngcc18.FormlyForm], pipes: [ɵngcc19.C8yTranslatePipe, ɵngcc10.AsyncPipe], encapsulation: 2 });
ProviderConfigurationComponent.ctorParameters = () => [
    { type: Permissions },
    { type: ActivatedRoute },
    { type: ModalService },
    { type: AlertService },
    { type: ProviderDefinitionsService },
    { type: ProviderConfigurationService },
    { type: C8yJSONSchema }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ProviderConfigurationComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-sms-gateway',
                template: "<c8y-title>\n  {{ (layout$ | async)?.pageTitle | translate }}\n</c8y-title>\n\n<div class=\"row\">\n  <div class=\"col-md-8 col-xs-12\">\n    <form class=\"card card--fullpage\" (ngSubmit)=\"saveProviderConfiguration()\">\n      <div class=\"card-header separator\">\n        <h4 class=\"card-title\">\n          {{ (layout$ | async)?.cardTitle | translate }}\n        </h4>\n      </div>\n      <div class=\"inner-scroll\">\n        <div class=\"card-block\">\n          <p *ngIf=\"!!(layout$ | async)?.description\" class=\"m-b-8\">\n            {{ (layout$ | async)?.description | translate }}\n          </p>\n          <c8y-form-group>\n            <label for=\"providerName\">{{ (layout$ | async)?.providerName | translate }}</label>\n            <c8y-typeahead\n              [disabled]=\"!permissions.hasAllRoles((layout$ | async)?.saveRoles || [])\"\n              [ngModel]=\"selectedProvider$ | async\"\n              [displayProperty]=\"'displayName'\"\n              name=\"providerName\"\n              placeholder=\"{{ (layout$ | async)?.providerNamePlaceholder | translate }}\"\n              (onSearch)=\"providerInput$.next($event)\"\n              [allowFreeEntries]=\"false\"\n              [required]=\"true\"\n              [container]=\"'body'\"\n            >\n              <c8y-li\n                *ngFor=\"let provider of providers$ | async\"\n                class=\"p-l-8 p-r-8 c8y-list__item--link\"\n                (click)=\"changeProvider$.next(provider); providerInput$.next('')\"\n                [active]=\"(selectedProvider$ | async) === provider\"\n              >\n                <c8y-highlight\n                  [text]=\"provider.displayName || '--'\"\n                  [pattern]=\"providerInput$ | async\"\n                ></c8y-highlight>\n              </c8y-li>\n            </c8y-typeahead>\n            <c8y-messages\n              ><c8y-message\n                name=\"notExisting\"\n                [text]=\"(layout$ | async)?.providerNameNoMatchesHint | translate\"\n              ></c8y-message>\n            </c8y-messages>\n          </c8y-form-group>\n          <formly-form\n            *ngIf=\"selectedProvider$ | async\"\n            [form]=\"form\"\n            [fields]=\"fields\"\n            [model]=\"model\"\n            [options]=\"options\"\n          ></formly-form>\n        </div>\n      </div>\n      <div class=\"card-footer separator\" *c8yIfAllowed=\"allRoles$ | async; allowAny\">\n        <button\n          *c8yIfAllowed=\"(layout$ | async)?.deleteRoles\"\n          class=\"btn btn-default\"\n          type=\"button\"\n          (click)=\"deleteProviderConfiguration()\"\n          [disabled]=\"\n            !(configuration$ | async)?.provider && !(configuration$ | async)?.providerName\n          \"\n          title=\"{{ (layout$ | async)?.deleteBtnLabel | translate }}\"\n        >\n          {{ (layout$ | async)?.deleteBtnLabel | translate }}\n        </button>\n        <button\n          *c8yIfAllowed=\"(layout$ | async)?.saveRoles\"\n          class=\"btn btn-primary\"\n          type=\"submit\"\n          [disabled]=\"form.invalid || form.pristine\"\n          title=\"{{ (layout$ | async)?.saveBtnLabel | translate }}\"\n        >\n          {{ (layout$ | async)?.saveBtnLabel | translate }}\n        </button>\n      </div>\n    </form>\n  </div>\n</div>\n",
                providers: [ProviderConfigurationService, ProviderDefinitionsService]
            }]
    }], function () { return [{ type: ɵngcc1.Permissions }, { type: ɵngcc2.ActivatedRoute }, { type: ɵngcc3.ModalService }, { type: ɵngcc4.AlertService }, { type: ɵngcc5.ProviderDefinitionsService }, { type: ɵngcc6.ProviderConfigurationService }, { type: ɵngcc7.C8yJSONSchema }]; }, null); })();
