// @c8y/ngx-components/esm2015/core/help/help.service.js
import { Injectable } from '@angular/core';
import { TranslateService } from '@ngx-translate/core';
import { BehaviorSubject } from 'rxjs';
import { OptionsService } from '../common/options.service';
import * as i0 from "@angular/core";
import * as i1 from "@ngx-translate/core";
import * as i2 from "../common/options.service";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@ngx-translate/core';
import * as ɵngcc2 from '../common/options.service';
export class HelpService {
    constructor(translateService, options) {
        this.translateService = translateService;
        this.options = options;
        this.contextHelp = new BehaviorSubject(false);
        this.isWarningAlertViewed = false;
        this.SUPPORTED_LANGUAGES = ['en'];
        this.handleContextHelp();
    }
    isSupportedLanguage() {
        return this.SUPPORTED_LANGUAGES.indexOf(this.translateService.currentLang) > -1;
    }
    handleContextHelp() {
        const option = this.getContextHelpOption();
        if (option) {
            this.contextHelp.next(option);
            return;
        }
        this.contextHelp.next(false);
    }
    getContextHelpOption() {
        return this.options.contextHelp;
    }
}
HelpService.ɵfac = function HelpService_Factory(t) { return new (t || HelpService)(ɵngcc0.ɵɵinject(ɵngcc1.TranslateService), ɵngcc0.ɵɵinject(ɵngcc2.OptionsService)); };
HelpService.ɵprov = i0.ɵɵdefineInjectable({ factory: function HelpService_Factory() { return new HelpService(i0.ɵɵinject(i1.TranslateService), i0.ɵɵinject(i2.OptionsService)); }, token: HelpService, providedIn: "root" });
HelpService.ctorParameters = () => [
    { type: TranslateService },
    { type: OptionsService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(HelpService, [{
        type: Injectable,
        args: [{ providedIn: 'root' }]
    }], function () { return [{ type: ɵngcc1.TranslateService }, { type: ɵngcc2.OptionsService }]; }, null); })();
