// @c8y/ngx-components/esm2015/core/help/help.component.js
import { Component, Input } from '@angular/core';
import { DocsService } from '../docs/docs.service';
import { HelpService } from './help.service';
/**
 * A component which shows a context help in
 * the action bar.
 *
 * @example
 * ```html
 * <c8y-help src="/users-guide/cockpit/#dashboards"></c8y-help>
 * ```
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../docs/docs.service';
import * as ɵngcc2 from './help.service';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from 'ngx-bootstrap/collapse';
import * as ɵngcc5 from '../action-bar/action-bar-item.component';
import * as ɵngcc6 from '../product-experience/product-experience.directive';
import * as ɵngcc7 from '../common/icon.directive';
import * as ɵngcc8 from '../common/loading.component';
import * as ɵngcc9 from '../i18n/c8y-translate.directive';
import * as ɵngcc10 from '../i18n/c8y-translate.pipe';

const _c0 = function (a0) { return { src: a0 }; };
function HelpComponent_c8y_action_bar_item_0_Template(rf, ctx) { if (rf & 1) {
    const _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-action-bar-item", 6);
    ɵngcc0.ɵɵelementStart(1, "button", 7);
    ɵngcc0.ɵɵlistener("click", function HelpComponent_c8y_action_bar_item_0_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r5 = ɵngcc0.ɵɵnextContext(); return ctx_r5.toggle(); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 8);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("placement", "right")("priority", ctx_r0.priority);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(2, 7, "Open help"))("actionName", ctx_r0.isCollapsed ? "helpOpened" : "helpClosed")("actionData", ɵngcc0.ɵɵpureFunction1(9, _c0, ctx_r0.src));
    ɵngcc0.ɵɵattribute("aria-expanded", !ctx_r0.isCollapsed);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", "question-circle-o");
} }
function HelpComponent_div_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelement(1, "c8y-loading");
    ɵngcc0.ɵɵelementEnd();
} }
function HelpComponent_div_6_div_3_div_1_Template(rf, ctx) { if (rf & 1) {
    const _r11 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 17);
    ɵngcc0.ɵɵelementStart(1, "button", 18);
    ɵngcc0.ɵɵlistener("click", function HelpComponent_div_6_div_3_div_1_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r11); const ctx_r10 = ɵngcc0.ɵɵnextContext(3); ctx_r10.showLangWarning = !ctx_r10.showLangWarning; return ctx_r10.helpService.isWarningAlertViewed = !ctx_r10.helpService.isWarningAlertViewed; });
    ɵngcc0.ɵɵelementStart(2, "span", 19);
    ɵngcc0.ɵɵtext(3, "\u00D7");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "span", 20);
    ɵngcc0.ɵɵtext(5, "Close");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "p");
    ɵngcc0.ɵɵtext(7);
    ɵngcc0.ɵɵpipe(8, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(7);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(8, 1, "Help content is only available in English."), " ");
} }
function HelpComponent_div_6_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 13);
    ɵngcc0.ɵɵtemplate(1, HelpComponent_div_6_div_3_div_1_Template, 9, 3, "div", 14);
    ɵngcc0.ɵɵelementStart(2, "h4", 15);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(4, "div", 16);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r7.helpService.isSupportedLanguage() && !ctx_r7.showLangWarning && !ctx_r7.helpService.isWarningAlertViewed);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r7.title, " ");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("innerHTML", ctx_r7.sectionContent, ɵngcc0.ɵɵsanitizeHtml);
} }
function HelpComponent_div_6_div_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 21);
    ɵngcc0.ɵɵelementStart(1, "h4", 22);
    ɵngcc0.ɵɵtext(2, "Sorry, that didn't work");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div", 23);
    ɵngcc0.ɵɵelementStart(4, "p", 24);
    ɵngcc0.ɵɵtext(5, "The content couldn't be loaded.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
const _c1 = function (a0, a1) { return { "text-warning": a0, "text-muted": a1 }; };
function HelpComponent_div_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "div", 9);
    ɵngcc0.ɵɵelement(2, "i", 10);
    ɵngcc0.ɵɵtemplate(3, HelpComponent_div_6_div_3_Template, 5, 3, "div", 11);
    ɵngcc0.ɵɵtemplate(4, HelpComponent_div_6_div_4_Template, 6, 0, "div", 12);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", !ctx_r3.hasError ? ctx_r3.icon : "unlink")("ngClass", ɵngcc0.ɵɵpureFunction2(4, _c1, ctx_r3.hasError, !ctx_r3.hasError));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r3.hasError);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.hasError);
} }
function HelpComponent_div_7_span_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 24);
    ɵngcc0.ɵɵtext(1, " Open the ");
    ɵngcc0.ɵɵelementStart(2, "span");
    ɵngcc0.ɵɵtext(3, "User guide`KEEP_ORIGINAL`");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function HelpComponent_div_7_span_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 24);
    ɵngcc0.ɵɵtext(1, " Check the ");
    ɵngcc0.ɵɵelementStart(2, "span");
    ɵngcc0.ɵɵtext(3, "User guide`KEEP_ORIGINAL`");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function HelpComponent_div_7_Template(rf, ctx) { if (rf & 1) {
    const _r15 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 25);
    ɵngcc0.ɵɵelementStart(1, "button", 26);
    ɵngcc0.ɵɵlistener("click", function HelpComponent_div_7_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r15); const ctx_r14 = ɵngcc0.ɵɵnextContext(); return ctx_r14.toggle(); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵtext(3, " Close ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "a", 27);
    ɵngcc0.ɵɵtemplate(5, HelpComponent_div_7_span_5_Template, 4, 0, "span", 28);
    ɵngcc0.ɵɵtemplate(6, HelpComponent_div_7_span_6_Template, 4, 0, "span", 28);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(2, 5, "Close help"));
    ɵngcc0.ɵɵattribute("aria-expanded", !ctx_r4.isCollapsed);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("href", ctx_r4.hasError ? ctx_r4.getUrl() : ctx_r4.getUrl(ctx_r4.src), ɵngcc0.ɵɵsanitizeUrl);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r4.hasError);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r4.hasError);
} }
const _c2 = function (a0) { return { "c8y-help-drawer-block": a0 }; };
export class HelpComponent {
    /**
     * @ignore Only private DI
     */
    constructor(docsService, helpService) {
        this.docsService = docsService;
        this.helpService = helpService;
        /**
         * The source of the documentation. Used to link to the documentation as well as
         * to parse the source to display.
         */
        this.src = '';
        /**
         * Indicates if the help dialog is collapsed.
         */
        this.isCollapsed = true;
        /**
         * The priority where the help icon should be shown in the action bar
         */
        this.priority = Infinity;
        /**
         * A title. Set in open by passing the source.
         */
        this.title = '';
        /**
         * The section heading in the doc which is going to be displayed.
         */
        this.sectionHeading = '';
        /**
         * The section content in the doc which is going to be displayed.
         */
        this.sectionContent = '';
        /**
         * Indicates if the component is loading.
         */
        this.isLoading = true;
        /**
         * Indicates if the component failed loading the source.
         */
        this.hasError = false;
        /**
         * Indicates if a warning should be shown.
         */
        this.showLangWarning = false;
        /**
         * @ignore
         */
        this.isInit = false;
    }
    /**
     * The component is shown by default and therefore breaks e2e test. This is
     * to prevent the visibility on first navigation.
     * @ignore
     */
    onCollapsed() {
        this.isInit = true;
    }
    /**
     * Builds the URL based on the src. The Base URL can be set in the application options docBaseUrl.
     * @param src The source of the help on the guide.
     * @param index This flag is used to call the index.json content of a guide. For example, "https://www.cumulocity.com/guides/users-guide/cockpit/index.json".
     */
    getUrl(src = '', index = false) {
        let docsUrl;
        try {
            docsUrl =
                typeof this.helpService.contextHelp.value === 'string'
                    ? new URL(this.docsService.getUrlWithDocsVersion(this.helpService.contextHelp.value))
                    : new URL(this.docsService.getBaseUrl());
        }
        catch (error) {
            docsUrl = new URL(this.docsService.getBaseUrl());
            console.warn(error);
        }
        const [url, hashFragment] = src.split('#');
        this.sectionHeading = hashFragment;
        if (index) {
            src = `${url}index.json`;
        }
        return `${docsUrl.href}${src}`;
    }
    /**
     * Toggles the visibility of the help dialog.
     */
    toggle() {
        if (this.isCollapsed) {
            this.open();
            return;
        }
        this.close();
    }
    /**
     * Closes the help dialog.
     */
    close() {
        this.isCollapsed = true;
        this.clean();
    }
    /**
     * Opens the help dialog.
     */
    open() {
        this.isCollapsed = false;
        this.isLoading = true;
        this.requestContent();
        if (!this.icon) {
            this.icon = this.resolveIcon();
        }
    }
    requestContent() {
        const req = new XMLHttpRequest();
        req.onreadystatechange = () => this.render(req);
        req.addEventListener('load', () => this.render(req));
        req.open('GET', this.getUrl(this.src, true));
        req.responseType = 'json';
        req.setRequestHeader('Accept', 'text/html');
        req.send();
    }
    clean() {
        this.title = '';
        this.hasError = false;
        this.sectionContent = '';
    }
    resolveIcon() {
        try {
            const icon = Array.from(document.querySelector('nav .active i').classList).find(classes => classes.startsWith('c8y-icon-') || classes.startsWith('dlt-c8y-icon-'));
            return icon.replace('dlt-c8y-icon-', '').replace('c8y-icon-', 'c8y-');
        }
        catch (ex) {
            return 'life-saver';
        }
    }
    render(req) {
        if (req.readyState === 4) {
            this.isLoading = false;
            if (req.status === 200) {
                this.hasError = false;
                const sectionData = req.response[this.sectionHeading];
                if (sectionData) {
                    this.title = sectionData.title;
                    this.sectionContent = sectionData.helpcontent;
                }
            }
            else {
                this.hasError = true;
            }
        }
    }
}
HelpComponent.ɵfac = function HelpComponent_Factory(t) { return new (t || HelpComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.DocsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.HelpService)); };
HelpComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: HelpComponent, selectors: [["c8y-help"]], inputs: { src: "src", isCollapsed: "isCollapsed", priority: "priority", icon: "icon" }, decls: 8, vars: 11, consts: [["itemClass", "pull-right", 3, "placement", "priority", 4, "ngIf"], ["id", "collapseHelp", 1, "c8y-help-drawer", 3, "collapse", "isAnimated", "collapsed"], [3, "ngClass"], ["docOutlet", ""], [4, "ngIf"], ["class", "c8y-help-drawer-footer", 4, "ngIf"], ["itemClass", "pull-right", 3, "placement", "priority"], ["aria-controls", "collapseHelp", "c8yProductExperience", "", 1, "btn", "btn-help", 3, "title", "actionName", "actionData", "click"], [1, "text-info", 3, "c8yIcon"], [1, "d-flex"], [1, "c8y-icon-duocolor", "icon-48", 3, "c8yIcon", "ngClass"], ["class", "p-l-16 flex-grow", 4, "ngIf"], ["class", "p-l-16 p-t-16 flex-grow", 4, "ngIf"], [1, "p-l-16", "flex-grow"], ["class", "alert alert-info m-b-16", 4, "ngIf"], [1, "text-bold", "text-primary", "p-t-16", "m-b-16"], ["id", "helpContent", 1, "help-content", 3, "innerHTML"], [1, "alert", "alert-info", "m-b-16"], ["type", "button", 1, "close", 3, "click"], ["aria-hidden", "true"], [1, "sr-only"], [1, "p-l-16", "p-t-16", "flex-grow"], ["translate", "", 1, "text-bold", "m-b-16"], [1, "help-content"], ["translate", ""], [1, "c8y-help-drawer-footer"], ["aria-controls", "collapseHelp", "translate", "", 1, "btn", "btn-default", 3, "title", "click"], ["target", "_blank", "rel", "noopener noreferrer", 1, "btn", "btn-primary", 3, "href"], ["translate", "", 4, "ngIf"]], template: function HelpComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, HelpComponent_c8y_action_bar_item_0_Template, 4, 11, "c8y-action-bar-item", 0);
        ɵngcc0.ɵɵpipe(1, "async");
        ɵngcc0.ɵɵelementStart(2, "div", 1);
        ɵngcc0.ɵɵlistener("collapsed", function HelpComponent_Template_div_collapsed_2_listener() { return ctx.onCollapsed(); });
        ɵngcc0.ɵɵelementStart(3, "div", 2, 3);
        ɵngcc0.ɵɵtemplate(5, HelpComponent_div_5_Template, 2, 0, "div", 4);
        ɵngcc0.ɵɵtemplate(6, HelpComponent_div_6_Template, 5, 7, "div", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(7, HelpComponent_div_7_Template, 7, 7, "div", 5);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", !!ɵngcc0.ɵɵpipeBind1(1, 7, ctx.helpService.contextHelp));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("collapse", ctx.isCollapsed)("isAnimated", true);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(9, _c2, ctx.isInit));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isLoading);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isLoading);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isLoading && !ctx.hasError);
    } }, directives: [ɵngcc3.NgIf, ɵngcc4.CollapseDirective, ɵngcc3.NgClass, ɵngcc5.ActionBarItemComponent, ɵngcc6.ProductExperienceDirective, ɵngcc7.IconDirective, ɵngcc8.LoadingComponent, ɵngcc9.C8yTranslateDirective], pipes: [ɵngcc3.AsyncPipe, ɵngcc10.C8yTranslatePipe], encapsulation: 2 });
HelpComponent.ctorParameters = () => [
    { type: DocsService },
    { type: HelpService }
];
HelpComponent.propDecorators = {
    src: [{ type: Input }],
    isCollapsed: [{ type: Input }],
    priority: [{ type: Input }],
    icon: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(HelpComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-help',
                template: "<c8y-action-bar-item\n  [placement]=\"'right'\"\n  itemClass=\"pull-right\"\n  [priority]=\"priority\"\n  *ngIf=\"!!(helpService.contextHelp | async)\"\n>\n  <button\n    class=\"btn btn-help\"\n    [title]=\"'Open help' | translate\"\n    (click)=\"toggle()\"\n    [attr.aria-expanded]=\"!isCollapsed\"\n    aria-controls=\"collapseHelp\"\n    c8yProductExperience\n    [actionName]=\"isCollapsed ? 'helpOpened' : 'helpClosed'\"\n    [actionData]=\"{ src: src }\"\n  >\n    <i [c8yIcon]=\"'question-circle-o'\" class=\"text-info\"></i>\n  </button>\n</c8y-action-bar-item>\n\n<div\n  id=\"collapseHelp\"\n  class=\"c8y-help-drawer\"\n  [collapse]=\"isCollapsed\"\n  [isAnimated]=\"true\"\n  (collapsed)=\"onCollapsed()\"\n>\n  <div [ngClass]=\"{ 'c8y-help-drawer-block': isInit }\" #docOutlet>\n    <div *ngIf=\"isLoading\">\n      <c8y-loading></c8y-loading>\n    </div>\n\n    <div *ngIf=\"!isLoading\">\n      <div class=\"d-flex\">\n        <i\n          [c8yIcon]=\"!hasError ? icon : 'unlink'\"\n          [ngClass]=\"{ 'text-warning': hasError, 'text-muted': !hasError }\"\n          class=\"c8y-icon-duocolor icon-48\"\n        ></i>\n        <div class=\"p-l-16 flex-grow\" *ngIf=\"!hasError\">\n          <div\n            class=\"alert alert-info m-b-16\"\n            *ngIf=\"\n              !helpService.isSupportedLanguage() &&\n              !showLangWarning &&\n              !helpService.isWarningAlertViewed\n            \"\n          >\n            <button\n              class=\"close\"\n              type=\"button\"\n              (click)=\"\n                showLangWarning = !showLangWarning;\n                helpService.isWarningAlertViewed = !helpService.isWarningAlertViewed\n              \"\n            >\n              <span aria-hidden=\"true\">\u00D7</span>\n              <span class=\"sr-only\">Close</span>\n            </button>\n            <p>\n              {{ 'Help content is only available in English.' | translate }}\n            </p>\n          </div>\n          <h4 class=\"text-bold text-primary p-t-16 m-b-16\">\n            {{ title }}\n          </h4>\n          <div id=\"helpContent\" class=\"help-content\" [innerHTML]=\"sectionContent\"></div>\n        </div>\n\n        <div class=\"p-l-16 p-t-16 flex-grow\" *ngIf=\"hasError\">\n          <h4 class=\"text-bold m-b-16\" translate>Sorry, that didn't work</h4>\n          <div class=\"help-content\">\n            <p translate>The content couldn't be loaded.</p>\n          </div>\n        </div>\n      </div>\n    </div>\n  </div>\n  <div class=\"c8y-help-drawer-footer\" *ngIf=\"!isLoading && !hasError\">\n    <button\n      class=\"btn btn-default\"\n      (click)=\"toggle()\"\n      [title]=\"'Close help' | translate\"\n      [attr.aria-expanded]=\"!isCollapsed\"\n      aria-controls=\"collapseHelp\"\n      translate\n    >\n      Close\n    </button>\n    <a\n      href=\"{{ hasError ? getUrl() : getUrl(src) }}\"\n      class=\"btn btn-primary\"\n      target=\"_blank\"\n      rel=\"noopener noreferrer\"\n    >\n      <span translate *ngIf=\"!hasError\"> Open the <span>User guide`KEEP_ORIGINAL`</span> </span>\n      <span translate *ngIf=\"hasError\"> Check the <span>User guide`KEEP_ORIGINAL`</span> </span>\n    </a>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.DocsService }, { type: ɵngcc2.HelpService }]; }, { src: [{
            type: Input
        }], isCollapsed: [{
            type: Input
        }], priority: [{
            type: Input
        }], icon: [{
            type: Input
        }] }); })();
